---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(ComplexHeatmap)
  library(circlize)
  library(readxl)
  library(ggh4x)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  
})
```


```{r setup}
sample_set = "Mali2"

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```

```{r}
published_dset_dir <- "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/"

read_dir_scvi2seu <- "data/processed/Pf/m2_all/scvi2seu/"
read_dir_stage_DE_markers <- "data/processed/Pf/m2_all/stage_DE_markers/"
save_dir_stage_DE_markers <- "data/processed/Pf/m2_all/stage_DE_markers/"


```

#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```


```{r}
published_dset_dir <- "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/"

read_dir_stage_DE_markers <- "data/processed/Pf/m2_all/stage_DE_markers/"
save_dir_stage_DE_markers <- "data/processed/Pf/m2_all/stage_DE_markers/"

read_dir_integrated_scvi_cleanup <- "data/processed/Pf/m2_all/integrated_scvi_cleanup/"


```


#####!!! DE below in seurat. Not sure which one to go with


```{r}
# Clean integrated scvi object with the asex+activated_fem doublets removed - Generated in 'scvi_leiden_clusters_doublet_investigations.Rmd'
all_jcC_seu_scvi_noM55_afm_cln <- readRDS(paste0(read_dir_integrated_scvi_cleanup, "all_jcC_seu_scvi_noM55_afm_cln.RDS"))

```


```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~.x@meta.data %>% count(stage_afm, leiden) %>% arrange(desc(n)))
```

```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~.x@meta.data %>% count(leiden) %>% arrange(desc(n)))

```

```{r}
# clusters2keep4DE <- map(all_jcC_seu_scvi_noM55_afm_cln, ~.x@meta.data %>% count(leiden) %>% arrange(desc(n)) %>% filter(n>150) %>% pull(leiden))

```


```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~.x@meta.data %>% count(leiden) %>% arrange(desc(n)))
```

```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~.x@meta.data %>% count(stage_afm, leiden) %>% arrange(desc(n)))
```

```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~DimPlot(.x, group.by = "leiden", reduction = "scvi_umap", label = T))

```

```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~.x@meta.data %>% count(stage_afm, leiden, orig.ident) %>% group_by(leiden) %>% arrange(desc(n), .by_group = T) %>% filter(n>50))

```

```{r}
all_jcC_seu_scvi_noM55_afm_cln <- map(all_jcC_seu_scvi_noM55_afm_cln, ~SetIdent(.x, value = "leiden"))

```

#### Get cluster markers using wilcoxon ranks sum test DE 
```{r, eval=T}
seu_scviLeiden_markers <- map(all_jcC_seu_scvi_noM55_afm_cln, ~FindAllMarkers(.x ))
```


```{r, warning=FALSE, message=FALSE, eval=T}
seu_scviLeiden_markers_anot <- map(seu_scviLeiden_markers, ~.x %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

# seu_scviLeiden_markers_anot
```


```{r, eval=T}
# saveRDS(seu_scviLeiden_markers_anot, "data/processed/Pf/m2_all/seu_scviLeiden_markers_anot.RDS")
# saveRDS(seu_scviLeiden_markers_anot, "data/processed/Pf/m2_all/scvi2seu/seu_scviLeiden_markers_anot3.RDS")
# saveRDS(seu_scviLeiden_markers_anot,paste0(save_dir_stage_DE_markers, 'seu_scviLeiden_markers_anot3.RDS'))
saveRDS(seu_scviLeiden_markers_anot, paste0(save_dir_stage_DE_markers, 'seu_scviLeiden_markers_anot.RDS'))

```

```{r, eval=T}
## Read in DE results to avoid recalculating
seu_scviLeiden_markers_anot <- readRDS(paste0(save_dir_stage_DE_markers, 'seu_scviLeiden_markers_anot.RDS'))
```


```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~FeaturePlot(.x, features = "nFeature_RNA", reduction = "scvi_umap", label = T))
map(all_jcC_seu_scvi_noM55_afm_cln, ~DimPlot(.x, group.by = "leiden", reduction = "scvi_umap", label = T))

```


```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~DimPlot(.x, group.by = "leiden", reduction = "scvi_umap", label = T))

```

```{r, warning=FALSE, message=FALSE}
map(seu_scviLeiden_markers_anot, ~.x %>% 
      # filter(gene %in% gns) %>% 
      arrange(desc(avg_log2FC)) %>% 
      filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
      group_by(cluster) %>% 
      slice_min(order_by = p_val_adj, n = 30) %>% 
      slice_max(order_by = avg_log2FC, n = 30))
```


```{r, warning=FALSE, message=FALSE}
top_de_tbl <- map(seu_scviLeiden_markers_anot, ~.x %>% 
                    # filter(gene %in% gns) %>% 
                    arrange(desc(avg_log2FC)) %>% 
                    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
                    group_by(cluster) %>% 
                    slice_min(order_by = p_val_adj, n = 40) %>% 
                    slice_max(order_by = avg_log2FC, n = 20) %>%
                    ungroup() %>%
                    mutate(across(cluster, ~droplevels(factor(.x, paste(1:50))))) %>%
                    arrange(cluster) %>%
                    mutate(across(cluster, as.character))
)

top_de_tbl
```

```{r}
map(all_jcC_seu_scvi_noM55_afm_cln, ~DimPlot(.x, group.by = "leiden", reduction = "scvi_umap", label = T))

```



## Annotation based on the DE results
8=Activated/ookinete like



```{r, warning=FALSE, message=FALSE}
top10gns <- map(top_de_tbl, ~.x %>% select(cluster, Gene, gene) %>% split(., f=factor(.$cluster)) %>% map(.,\(obj) obj %>% ungroup() %>% select(Gene, gene) %>% deframe()))

top10gns[["Female"]]
```


```{r, fig.width=18, fig.height=10}
imap(top10gns[["Female"]], ~{
  gn <- .x
  
  map(all_jcC_seu_scvi_noM55_afm_cln[1], ~FeaturePlot(.x, features = gn[1:2], reduction = "scvi_umap", raster = T, order = T))
  })

```


```{r, warning=FALSE, message=FALSE}
top_de_tbl_strict <- map(seu_scviLeiden_markers_anot, ~.x %>% 
                           # filter(gene %in% gns) %>% 
                           arrange(desc(avg_log2FC)) %>% 
                           filter(p_val_adj< 0.05 & avg_log2FC>0 & (pct.1 - pct.2) >0.2 & pct.2 < 0.5) %>%
                           group_by(cluster) %>% 
                           slice_min(order_by = p_val_adj, n = 40) %>% 
                           slice_max(order_by = avg_log2FC, n = 5))

top_de_tbl_strict
```

```{r, warning=FALSE, message=FALSE}
top10gns_strict <- map(top_de_tbl_strict, ~.x %>% select(cluster, Gene, gene) %>% split(., f=factor(.$cluster)) %>% map(.,\(obj) obj %>% ungroup() %>% select(Gene, gene) %>% deframe()))

top10gns_strict[["Female"]]
```


```{r, fig.width=18, fig.height=10}
imap(top10gns_strict[["Female"]], ~{
  gn <- .x
  
  map(all_jcC_seu_scvi_noM55_afm_cln[1], ~FeaturePlot(.x, features = gn[1:2], reduction = "scvi_umap", raster = T, order = T))
  })

```



```{r}
# kahrp
map(all_jcC_seu_scvi_noM55_afm_cln, ~FeaturePlot(.x, features = "PF3D7-0202000", reduction = "scvi_umap", pt.size = 0.01, raster = F, order = T, label = T))
```


```{r}
#cdpk1
map(all_jcC_seu_scvi_noM55_afm_cln, ~FeaturePlot(.x, features = "PF3D7-0217500", reduction = "scvi_umap", label = T, raster = F, order = T))

```

####################START Compare to published datasets ##################

Cluster 9 - activated like 

Gamete top marker - PF3D7_0825800 https://www.nature.com/articles/s41586-025-09653-0
Ngwa activated gametocytes https://opus.bibliothek.uni-wuerzburg.de/opus4-wuerzburg/frontdoor/deliver/index/docId/6822/file/PhD_thesis_Che_Julius_Ngwa_final.pdf

```{r, warning=FALSE, message=FALSE}
## Ngwa activated gametocytes https://opus.bibliothek.uni-wuerzburg.de/opus4-wuerzburg/frontdoor/deliver/index/docId/6822/file/PhD_thesis_Che_Julius_Ngwa_final.pdf
ngwa_up_activ_gam <- c("PF3D7_0827800", "Pf3D7_1246200", "PF3D7_1218800", "PF3D7_1239400", "PF3D7_0704100", "PF3D7_0417400", "PF3D7_1321000", "PF3D7_131670") %>% str_replace(., "_", "-")

map(seu_scviLeiden_markers_anot, ~.x %>% 
      filter(gene %in% ngwa_up_activ_gam) %>% 
      filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
      arrange(desc(avg_log2FC)) )
```

```{r, warning=FALSE, message=FALSE}
## Gamete top marker - PF3D7_0825800 https://www.nature.com/articles/s41586-025-09653-0
map(all_jcC_seu_scvi_noM55_afm_cln, ~FeaturePlot(.x, features = c("PF3D7-0825800"), order =T, ncol = 1))

```


```{r, warning=FALSE, message=FALSE, fig.width=16, fig.height=6}
map(all_jcC_seu_scvi_noM55_afm_cln, ~FeaturePlot(.x, features = c("PF3D7-0829900", "PF3D7-1229200", "PF3D7-1222600", "PF3D7-1102500", "PF3D7-0818900"), order =T, ncol = 3))

```

cold-activated noncoding RNAs Temperature Regulated Untranslated Lncrna (tru-lncRNA) 
https://www.sciencedirect.com/science/article/pii/S0378111923003578 
https://www.nature.com/articles/s41564-021-00940-w
https://pmc.ncbi.nlm.nih.gov/articles/PMC12235521/
https://onlinelibrary.wiley.com/doi/10.1111/1744-7917.13383
```{r, warning=FALSE, message=FALSE}
cold_activated_RNAs <- c("PF3D7-1148100", "PF3D7-1148200", "PF3D7-1148500", "PF3D7-1370500", "PF3D-1370800", "PF3D7-1370900")
```


```{r, fig.width=14, fig.height=8}
map(all_jcC_seu_scvi_noM55_afm_cln, ~FeaturePlot(.x, features = cold_activated_RNAs, order =T, ncol = 3))
```



#### P.berghei annotations 
```{r, message=FALSE}
##ortholog file to translate to falciparum IDs
ortho <- read_csv("../../Pf3D7_genomes_gtfs_indexs/ortho4.csv")

```


#### NGwa activated gametocyte genes

```{r, message=FALSE}
ngwa_activated_gams <- read_csv('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/ngwa_2013_activated_gametocytes_t_ome/initial_phase_of_mosquito_transcriptome.csv') %>% mutate(gene_id = str_replace(`Plasmodb ID`, "_", "-"))

ngwa_activated_gams
```

```{r, warning=FALSE, message=FALSE}
top_de_tbl8 <- map(seu_scviLeiden_markers_anot, ~.x %>% 
                     # filter(gene %in% gns) %>% 
                     arrange(desc(avg_log2FC)) %>% 
                     filter(cluster == 8 & p_val_adj< 1e-10 & avg_log2FC>0.5 & pct.1 > pct.2 & pct.2 < 0.2& pct.1 > 0.1) 
)

top_de_tbl8
```

```{r, warning=FALSE, message=FALSE}
top_de_tbl8[[1]]$gene %>% str_replace(., "-", "_") %>% paste0(collapse = " ") 
```


```{r, warning=FALSE, message=FALSE}
gns <-top_de_tbl8[[1]]$Gene %>% str_replace(., "-", "_") %>% dput()
gns
```

```{r, warning=FALSE, message=FALSE}
table(top_de_tbl8[[1]]$gene %in% ngwa_activated_gams$gene_id)
```


#### Rios widespread release of TR genes 2024 plos

```{r, message=FALSE}
rios_tr <- read_xlsx(paste0(published_dset_dir, 'rios_releaseOfTRGeneHostToVector_2024/journal.ppat.1012823.s010.xlsx'), sheet = "TR198 List")

rios_tr_gns <- rios_tr %>% left_join(., ortho, by=c('Gene ID'='yoelii') ) %>% filter(!is.na(falciparum)) %>% dplyr::pull(falciparum) %>% str_replace(., '_', '-')

```


```{r}
rios_tr_gns
```

```{r, message=FALSE}
rios_femgam_up <- read_xlsx(paste0(published_dset_dir, 'rios_releaseOfTRGeneHostToVector_2024/media_3_female_gam_n_zygote_enriched_genes.xlsx'), sheet = "3. FG Enriched (<log2 2.3)") %>% left_join(., ortho, by=c('Gene Name'='yoelii') ) %>% filter(!is.na(falciparum)) %>% mutate(across(falciparum, ~str_replace(., "_", "-")))

rios_zygote_up <- read_xlsx(paste0(published_dset_dir, 'rios_releaseOfTRGeneHostToVector_2024/media_3_female_gam_n_zygote_enriched_genes.xlsx'), sheet = "4. ZYG Enriched (>log2 2.3)") %>% left_join(., ortho, by=c('Gene Name'='yoelii') ) %>% filter(!is.na(falciparum)) %>% mutate(across(falciparum, ~str_replace(., "_", "-")))

```

```{r}
rios_femgam_up
```


```{r}
rios_zygote_up
```


```{r, warning=FALSE, message=FALSE}
seu_scviLeiden_markers_anot_rios <- map(seu_scviLeiden_markers_anot, ~{
  .x %>% 
    filter(p_val_adj< 0.05 & avg_log2FC > 0.5 & pct.1 > pct.2) %>%
    mutate("up_rios_fem" = ifelse(gene %in% rios_femgam_up$falciparum, "yes", "no"),
           "up_rios_zyg" = ifelse(gene %in% rios_zygote_up$falciparum, "yes", "no")) 
})
```

```{r, warning=FALSE, message=FALSE}
map(seu_scviLeiden_markers_anot_rios, ~{
  .x %>% count(cluster, up_rios_fem) %>%
    group_by(cluster) %>%
    mutate(prop = round(n / sum(n), digits = 2))
})
```


```{r, warning=FALSE, message=FALSE}
map(seu_scviLeiden_markers_anot_rios, ~{
  .x %>% count(cluster, up_rios_zyg) %>%
    group_by(cluster) %>%
    mutate(prop = round(n / sum(n), digits = 2))
})
```


#### Average expression of 511 Lasonder translationally repressed genes in MSC {.tabset}
```{r, message=FALSE}
lasonder_tr <- read_xlsx(paste0(published_dset_dir, 'lasonder/supp_512_TR_genes.xlsx'), skip = 4)
tr_gns <- lasonder_tr %>% dplyr::rename('gene_id' = `Gene ID...1`) %>% pull(gene_id) %>% str_replace(.,'_', '-')
tr_gns <- tr_gns[str_detect(tr_gns, 'PF3D7-')]
```

```{r, warning=FALSE, message=FALSE}
seu_scviLeiden_markers_anot_lasonder <- map(seu_scviLeiden_markers_anot, ~{
  .x %>% 
    filter(p_val_adj< 0.05 & avg_log2FC > 0.5 & pct.1 > pct.2) %>%
    mutate("lasonder" = ifelse(gene %in% tr_gns, "yes", "no"))
})
```

```{r, warning=FALSE, message=FALSE}
map(seu_scviLeiden_markers_anot_lasonder, ~{
  .x %>% count(cluster, lasonder) %>%
    group_by(cluster) %>%
    mutate(prop = round(n / sum(n), digits = 2))
})
```

Male and female fertility genes https://www.cell.com/cell-systems/fulltext/S2405-4712(24)00305-3
```{r}
sayers_Pb_fertility = readxl::read_xlsx(paste0(published_dset_dir, "sayers_PbFertilityGenes_2024/fertility_genes_screen.xlsx"), sheet = "(A) Fertility screen results", .name_repair = "universal", na = c("", "NA")) %>% rename_all(., ~str_remove(., "\\.\\."))
```

```{r}
sayers_Pb_fertility_cln <- sayers_Pb_fertility %>% select(Pberghei.ANKA.gene.ID, Pfalciparum.3D7.orthologue..ID, Female.fertilitylog2FC...input.sample.log2.barcode.relative.abundance.is.subtracted.from.output.sample.log2.barcode.relative.abundance.and.then.normalised.to.WT.controls., Female.p.valuelikelihood.that.fertility..log2FC..occurs.by.chance.where.a.p.value..0.05.indicates.statistical.significance., Female.fertility.phenotypemutants.with.a.fertility..log2FC..plus.2xSD.of.less.than..2.are.called.Reduced.and.greater.than..2.are.called.Not.reduced., Male.fertilitylog2FC...input.sample.log2.barcode.relative.abundance.is.subtracted.from.output.sample.log2.barcode.relative.abundance.then.normalised.to.WT.controls., Male.p.valuelikelihood.that.fertility..log2FC..occurs.by.chance.where.a.p.value..0.05.indicates.statistical.significance., Male.fertility.phenotypemutants.with.a.fertility..log2FC..plus.2xSD.of.less.than..2.are.called.Reduced.and.greater.than..2.are.called.Not.reduced., contains("gametocyte.log2FCPMID..36634679"), contains("phenotypePMID..36634679")) %>% rename_all(., ~str_remove_all(., "...input.sample.*|likelihood..*|mutants.*|PMID.*")) %>% mutate(gene_id = str_replace(Pfalciparum.3D7.orthologue..ID, "_", "-"))
```


```{r}
sayers_Pb_fertility_cln 

```


```{r}
sayers_sig <- sayers_Pb_fertility_cln %>% 
  mutate(stage_reduced = case_when(Female.fertility.phenotype == "Reduced" & Male.fertility.phenotype == "Reduced" ~ "FemMale",
                                   Female.fertility.phenotype == "Reduced" & Male.fertility.phenotype == "Reduced" ~ "Fem",
                                   Female.fertility.phenotype != "Reduced" & Male.fertility.phenotype == "Reduced" ~ "Male")) %>%
  filter(Female.fertility.phenotype == "Reduced" & Male.fertility.phenotype != "Reduced" & !is.na(Pfalciparum.3D7.orthologue..ID) & Female.p.value < 1e-10) %>% mutate(across(Pfalciparum.3D7.orthologue..ID, ~str_replace(., "_", "-")))

sayers_m_sig <- sayers_Pb_fertility_cln %>% 
  filter(Male.fertility.phenotype == "Reduced" & !is.na(Pfalciparum.3D7.orthologue..ID) & Male.p.value < 1e-10) %>% mutate(across(Pfalciparum.3D7.orthologue..ID, ~str_replace(., "_", "-")))

sayers_f_sig <- sayers_Pb_fertility_cln %>% 
  filter(Female.fertility.phenotype == "Reduced" & !is.na(Pfalciparum.3D7.orthologue..ID) & Female.p.value < 1e-10) %>% mutate(across(Pfalciparum.3D7.orthologue..ID, ~str_replace(., "_", "-")))

```


```{r, warning=FALSE, message=FALSE}
seu_scviLeiden_markers_anot_sayer <- map(seu_scviLeiden_markers_anot, ~{
  .x %>% 
    filter(p_val_adj< 0.05 & avg_log2FC > 0.5 & pct.1 > pct.2) %>%
    mutate("up_sayer_fem" = ifelse(gene %in% sayers_f_sig$Pfalciparum.3D7.orthologue..ID, "yes", "no"),
           "up_sayer_male" = ifelse(gene %in% sayers_m_sig$Pfalciparum.3D7.orthologue..ID, "yes", "no"))
})
```

```{r, warning=FALSE, message=FALSE}
map(seu_scviLeiden_markers_anot_sayer, ~{
  .x %>% count(cluster, up_sayer_fem) %>%
    group_by(cluster) %>%
    mutate(prop = round(n / sum(n), digits = 2))
})
```


```{r, warning=FALSE, message=FALSE}
map(seu_scviLeiden_markers_anot_sayer, ~{
  .x %>% count(cluster, up_sayer_male) %>%
    group_by(cluster) %>%
    mutate(prop = round(n / sum(n), digits = 2))
})
```

####################END Compare to published datasets ##################


