---
title: "Stage DE figs"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(ComplexHeatmap)
  library(circlize)
  library(readxl)
  library(ggh4x)
  library(EnhancedVolcano)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::as.factor)
  
})
```


```{r setup}
sample_set = "Mali2"

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```


#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```


```{r}
# read_dir <- "data/processed/Pf/m2_all/scvi2seu/"
read_dir_stage_DE_markers <- "data/processed/Pf/m2_all/stage_DE_markers/"
read_dir_integration <- "data/processed/Pf/m2_all/integration/"
read_dir_scvi2seu <- "data/processed/Pf/m2_all/scvi2seu/"

published_dset_dir <- "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/"


```

#####!!! DE below in seurat. Already done using scvi. Not sure which one to go with
```{r, message=FALSE, warning=FALSE}
# Generated in '' 
# Original umap with asx dbts etc

# all_jcC_seu_scvi_noM55 <- readRDS("data/processed/Pf/m2_all/scvi2seu/all_jcC_seu_scvino_M55.RDS")
# all_jcC_seu_scvi_noM55 <- readRDS(paste0(read_dir_scvi2seu, 'all_jcC_seu_scvino_M55.RDS'))

```

```{r}
# Clean integrated scvi object with the asex+activated_fem doublets removed - Generated in 'scvi_leiden_clusters_doublet_investigations.Rmd'
all_jcC_seu_scvi_noM55_afm_cln <- readRDS(paste0(read_dir_integration, "all_jcC_seu_scvi_noM55_afm_cln.RDS"))

```


```{r}
# # Generated in "scvi_leiden_clusters_seuratDE.Rmd"
# all_jcC_seu_scvi_stg_noM55_afm_cln <- readRDS(paste0(read_dir_stage_DE_markers, 'all_jcC_seu_scvi_stg_noM55_afm_cln.RDS'))

```

```{r, eval=T}
# Generated in "scvi_leiden_clusters_seuratDE.Rmd"
seu_scviLeiden_markers_anot <- readRDS(paste0(read_dir_stage_DE_markers, 'seu_scviLeiden_markers_anot3.RDS'))

```

```{r}
# Generated in "scvi_leiden_clusters_seuratDE.Rmd" - This was generated by running pseudobulk DE on all_jcC_seu_scvi_noM55_afm_cln read in above
bulk_de_all_stgs_anot <- readRDS(paste0(read_dir_stage_DE_markers, 'bulk_de_all_stgs_anot.RDS'))

```

```{r}
# generated in 'm2_raw_integration_trial_qcd_DE.ipynb'
scvi_leiden_de_markers <- read_csv("data/processed/Pf/m2_all/scvi/m2_all_scvi_leiden_marker_genes.csv") %>% 
  rename("gene_id" = `...1`) %>%
  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene_id" = "gene_id"))

```


```{r}
# Generated in 'scvi_leiden_clusters_seurat_pbulkDE_n_DE_comparisons'
de_wilcox_bulk_scvi <- readRDS(paste0(read_dir_stage_DE_markers, 'de_wilcox_bulk_scvi.RDS'))

```


#####################START#### Dotplots for markers for all stages #################################

```{r}
# go()
# Generated in "scvi_leiden_clusters_seuratDE.Rmd" - This was generated by running pseudobulk DE on all_jcC_seu_scvi_noM55_afm_cln read in above
bulk_de_all_stgs_anot <- readRDS(paste0(read_dir_stage_DE_markers, 'bulk_de_all_stgs_anot.RDS'))

```

```{r}
# Generated in 'scvi_leiden_clusters_seurat_pbulkDE_n_DE_comparisons'
de_wilcox_bulk_scvi <- readRDS(paste0(read_dir_stage_DE_markers, 'de_wilcox_bulk_scvi.RDS'))

```




### Check top markers for each leiden cluster obtained using pseudobulk and plot dotplots

```{r}
bulk_de_all_stgs_anot[["all"]] 
  

```

```{r, warning=FALSE, message=FALSE}
map(bulk_de_all_stgs_anot["all"], possibly(~.x %>% 
      filter(p_val_adj < 0.05 & avg_log2FC >0) %>% 
      arrange(desc(avg_log2FC)) %>% 
      group_by(cluster) %>% 
      slice_min(order_by = p_val_adj, n = 40) %>% 
      slice_min(p_val_adj, n =4) %>%
      ungroup() %>%
      mutate(across(cluster, ~droplevels(factor(.x, levels = c(0:30))))) %>%
      arrange(cluster)
)
)

```

Top genes but only with a gene id for easy literature search as marker
```{r, warning=FALSE, message=FALSE}
top_gns_w_ids <- map(bulk_de_all_stgs_anot["all"], possibly(~.x %>% 
      # filter(p_val_adj < 0.05 & avg_log2FC >0 & !str_detect(Gene, "PF3D")) %>% 
      filter(p_val_adj < 0.05 & avg_log2FC >1 & !str_detect(Gene, "VAR") ) %>% 
      arrange(desc(avg_log2FC)) %>% 
      group_by(cluster) %>% 
      slice_min(order_by = p_val_adj, n = 20) %>% 
      slice_min(p_val_adj, n =2) %>%
      ungroup() %>%
      mutate(across(cluster, ~droplevels(factor(.x, levels = c(0:30))))) %>%
      arrange(cluster)
)
)

top_gns_w_ids

```

```{r, warning=FALSE, message=FALSE}
all_jcC_seu_scvi_noM55_cln <- all_jcC_seu_scvi_noM55_afm_cln[["all"]]
```

Add a combination of stage leiden cluster labels for dotplots
```{r, warning=FALSE, message=FALSE}
leiden = sort(unique(all_jcC_seu_scvi_noM55_cln$leiden))
stages = sort(unique(all_jcC_seu_scvi_noM55_cln$stage_afm))
stage_asx_c = sort(unique(all_jcC_seu_scvi_noM55_cln$stage_asx_c))

stage_leiden_levels <- paste0(rep(stages, each = length(leiden)), "_", leiden)
stage_asx_c_leiden_levels <- paste0(rep(stage_asx_c, each = length(leiden)), "_", leiden)

stage_leiden_levels
```

Add the ordered factors to the metadata for plotting the dotplots
```{r, warning=FALSE, message=FALSE}
all_jcC_seu_scvi_noM55_cln <- all_jcC_seu_scvi_noM55_cln@meta.data %>%
  mutate(stage_afm_leiden = droplevels(factor(paste0(stage_afm, "_", leiden), levels = stage_leiden_levels)),
         stage_leiden = droplevels(factor(paste0(stage_asx_c, "_", leiden), levels = stage_asx_c_leiden_levels))) %>%
  select(stage_afm_leiden, stage_leiden) %>%
  AddMetaData(all_jcC_seu_scvi_noM55_cln, .)
```


Order or group leiden cluster by stage 
```{r, warning=FALSE, message=FALSE, fig.width= 14, fig.height= 5}
stage_asx_c_leiden_grouping <- all_jcC_seu_scvi_noM55_cln@meta.data %>%
  distinct(stage_asx_c, leiden) %>% 
  arrange(stage_asx_c) %>% 
  pull(leiden) %>% 
  unique()

stage_afm_leiden_grouping <- all_jcC_seu_scvi_noM55_cln@meta.data %>%
  distinct(stage_afm, leiden) %>% 
  arrange(stage_afm) %>% 
  pull(leiden) %>% 
  unique()
  

```


```{r, warning=FALSE, message=FALSE, fig.width= 14, fig.height= 6.5}
## Visualise the few misplaced leiden cluster cells - i.e. those with less than 50 cells in the asexual, female or male cells
all_jcC_seu_scvi_noM55_cln@meta.data %>%
  # count(stage_afm, stage_fld, leiden) %>% 
  count(stage_afm, leiden) 
```



```{r, warning=FALSE, message=FALSE}
DimPlot(all_jcC_seu_scvi_noM55_cln, group.by = "leiden", label = T)
```

```{r, warning=FALSE, message=FALSE, fig.width= 14, fig.height= 5}
top_gns_w_stage_asx_c <- top_gns_w_ids[[1]] %>% select(cluster, gene) %>% mutate(across(cluster, ~droplevels(factor(., levels = stage_asx_c_leiden_grouping)))) %>% arrange(cluster)
top_gns_w_stage_asx_c

top_gns_w_stage_afm <- top_gns_w_ids[[1]] %>% select(cluster, gene) %>% mutate(across(cluster, ~droplevels(factor(., levels = stage_afm_leiden_grouping)))) %>% arrange(cluster)
top_gns_w_stage_afm
```

```{r, warning=FALSE, message=FALSE, fig.width= 14, fig.height= 5}
DotPlot(all_jcC_seu_scvi_noM55_cln, features = top_gns_w_stage_afm %>% distinct(gene, .keep_all = T) %>% deframe, group.by = "stage_afm_leiden") +
  theme(axis.text.x = element_text(angle = 20, hjust = 1))
```

```{r, warning=FALSE, message=FALSE, fig.width= 14, fig.height= 5}
DotPlot(all_jcC_seu_scvi_noM55_cln, features = top_gns_w_stage_asx_c %>% distinct(gene, .keep_all = T) %>% deframe, group.by = "stage_leiden") +
  theme(axis.text.x = element_text(angle = 20, hjust = 1))
```


### Wilcoxon test DE markers - Dotplot of top markers per cluster using the correctly clustered cells




```{r, warning=FALSE, message=FALSE}
## Get top 20 genes per clusters with p value 0 but ranked by average log fold and remove var genes
wilcox_top_gns_w_ids <- de_wilcox_bulk_scvi %>% 
      filter(p_val_adj_seuWilcox == 0 & avg_log2FC_seuWilcox >0 & !str_detect(Gene_seuWilcox, "VAR") ) %>% 
      arrange(desc(avg_log2FC_seuWilcox)) %>% 
      group_by(cluster) %>% 
      slice_min(order_by = p_val_adj_seuWilcox, n = 50) %>% 
      slice_max(avg_log2FC_seuWilcox, n =50) 

wilcox_top_gns_w_ids %>% 
      slice_max(avg_log2FC_seuWilcox, n =10) %>%
      ungroup() %>%
      # mutate(across(cluster, ~droplevels(factor(.x, levels = stage_asx_c_leiden_grouping)))) %>%
      arrange(cluster)


```

```{r, warning=FALSE, message=FALSE}
wilcox_top_gns_w_ids %>% 
      slice_max(avg_log2FC_seuWilcox, n =2)
```

```{r, warning=FALSE, message=FALSE}
# top_gns_w_ids <- map(bulk_de_all_stgs_anot["all"], possibly(~.x %>% 
#       # filter(p_val_adj < 0.05 & avg_log2FC >0 & !str_detect(Gene, "PF3D")) %>% 
#       filter(p_val_adj < 0.05 & avg_log2FC >1 & !str_detect(Gene, "VAR") ) %>% 
#       arrange(desc(avg_log2FC)) %>% 
#       group_by(cluster) %>% 
#       slice_min(order_by = p_val_adj, n = 20) %>% 
#       slice_min(p_val_adj, n =2) %>%
#       ungroup() %>%
#       mutate(across(cluster, ~droplevels(factor(.x, levels = c(0:30))))) %>%
#       arrange(cluster)
# )
# )
# 
# top_gns_w_ids

```

```{r, warning=FALSE, message=FALSE, fig.width= 12, fig.height= 5}
wilcox_top_gns_w_ids_gns <- wilcox_top_gns_w_ids %>% slice_max(avg_log2FC_seuWilcox, n =2) %>% select(cluster, gene)  %>% distinct(gene, .keep_all = T) %>% deframe
wilcox_top_gns_w_ids_gns
```


```{r, warning=FALSE, message=FALSE, fig.width= 12, fig.height= 5}
DotPlot(all_jcC_seu_scvi_noM55_cln, features = wilcox_top_gns_w_ids_gns, group.by = "stage_afm_leiden") +
  scale_x_discrete(labels = wilcox_top_gns_w_ids %>% select(gene, Gene_seuWilcox) %>% deframe()) +
  theme(axis.text.x = element_text(angle = 40, hjust = 1),
        legend.position = "bottom")
```


### Dotplots with colour stage
```{r, warning=FALSE, message=FALSE, fig.width= 12, fig.height= 6.5}
# Get the different stage colours
dotplot_stg_asx_cols <- sp_fld_colors2[c("Asexual", "Female", "Female (LE)", "Male", "Male (LE)")]
dotplot_stg_asx_cols
```

```{r, warning=FALSE, message=FALSE, fig.width= 12, fig.height= 6.5}
DotPlot(all_jcC_seu_scvi_noM55_cln, features = wilcox_top_gns_w_ids_gns, group.by = "stage_leiden", split.by = "stage_afm", cols = c("#0B4B83FF", "#CB5CF5FF", "#ED9E05FF")) +
  scale_x_discrete(labels = wilcox_top_gns_w_ids %>% select(gene, Gene_seuWilcox) %>% deframe()) +
  scale_y_discrete(labels = function(x) str_replace_all(str_remove_all(x, "_[A-Za-z]+$| \\(|\\)"), c("Male" = "M", "Female" = "F", "Asexual" = "A"))) +
  theme(axis.text.x = element_text(angle = 40, hjust = 1),
        legend.position = "bottom")
```


```{r, warning=FALSE, message=FALSE, fig.width= 12, fig.height= 5}
wilcox_top5_gns <- wilcox_top_gns_w_ids %>% 
  slice_max(avg_log2FC_seuWilcox, n =4) %>% 
  ungroup() %>%
  select(cluster, gene, Gene_seuWilcox)  %>% 
  distinct(gene, .keep_all = T) %>%
  mutate(across(cluster, ~droplevels(factor(.x, levels = stage_asx_c_leiden_grouping)))) %>%
  arrange(cluster)

wilcox_top5_gns_w_ids_gns <- wilcox_top5_gns %>% select(cluster, gene)  %>%  deframe()

wilcox_top5_gns_w_ids_gn_names <- wilcox_top5_gns %>% select(gene, Gene_seuWilcox) %>% mutate(across(Gene_seuWilcox, ~str_replace(., ".*_PF3D7", "PF3D7")))  %>%  deframe()

wilcox_top5_gns_w_ids_gn_names
```



```{r, warning=FALSE, message=FALSE, fig.width= 15, fig.height= 7}
DotPlot(all_jcC_seu_scvi_noM55_cln, features = wilcox_top5_gns_w_ids_gns, group.by = "stage_leiden", split.by = "stage_afm", cols = c("#0B4B83FF", "#CB5CF5FF", "#ED9E05FF")) +
  scale_x_discrete(labels = wilcox_top5_gns_w_ids_gn_names) +
  scale_y_discrete(labels = function(x) str_replace_all(str_remove_all(x, "_[A-Za-z]+$| \\(|\\)"), c("Male" = "M", "Female" = "F", "Asexual" = "A"))) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.position = "bottom")
```

Remove stage_asx_c and leiden combinations that are less than 50 - There are some stages when grouped at a higher resolutions such as Female (LE) that end up misplaced in other clusters

Tabulate
```{r, warning=FALSE, message=FALSE, fig.width= 15, fig.height= 7}
all_jcC_seu_scvi_noM55_cln@meta.data %>% count(stage_asx_c, leiden)

all_jcC_seu_scvi_noM55_cln@meta.data %>% count(stage_asx_c, leiden) %>% filter(n < 50)
```

Remove
```{r, warning=FALSE, message=FALSE, fig.width= 15, fig.height= 7}
all_jcC_seu_scvi_noM55_cln_atleast50_4_stage_asx_c_leiden <- all_jcC_seu_scvi_noM55_cln[,all_jcC_seu_scvi_noM55_cln@meta.data %>% rownames_to_column("bcode") %>% group_by(stage_asx_c, leiden)  %>% filter(n() >50) %>% pull(bcode) ]
```

Plot
```{r, warning=FALSE, message=FALSE, fig.width= 15, fig.height= 6}
DotPlot(all_jcC_seu_scvi_noM55_cln_atleast50_4_stage_asx_c_leiden, features = wilcox_top5_gns_w_ids_gns, group.by = "stage_leiden", split.by = "stage_afm", cols = c("#0B4B83FF", "#CB5CF5FF", "#ED9E05FF")) +
  scale_x_discrete(labels = wilcox_top5_gns_w_ids_gn_names) +
  scale_y_discrete(labels = function(x) str_replace_all(str_remove_all(x, "_[A-Za-z]+$| \\(|\\)"), c("Male" = "M", "Female" = "F", "Asexual" = "A"))) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.position = "bottom")
```


##### Plot pf genes that have names not conserved to allow for easy literature search

Get the top 4 genes with names
```{r, warning=FALSE, message=FALSE, fig.width= 12, fig.height= 5}
wilcox_top5_NamedGns <- wilcox_top_gns_w_ids %>% 
  filter(!str_detect(Gene_seuWilcox, "PF3D7")) %>% 
  slice_max(avg_log2FC_seuWilcox, n =4) %>% 
  ungroup() %>%
  select(cluster, gene, Gene_seuWilcox)  %>% 
  distinct(gene, .keep_all = T) %>%
  mutate(across(cluster, ~droplevels(factor(.x, levels = stage_asx_c_leiden_grouping)))) %>%
  arrange(cluster)

wilcox_top5_NamedGns_w_ids_gns <- wilcox_top5_NamedGns %>% select(cluster, gene)  %>%  deframe()

wilcox_top5_NamedGns_w_ids_gn_names <- wilcox_top5_NamedGns %>% select(gene, Gene_seuWilcox) %>% mutate(across(Gene_seuWilcox, ~str_replace(., ".*_PF3D7", "PF3D7")))  %>%  deframe()

wilcox_top5_NamedGns_w_ids_gn_names
```

Plot in the object with stage leiden combination of more than 50
```{r, warning=FALSE, message=FALSE, fig.width= 15, fig.height= 6}
DotPlot(all_jcC_seu_scvi_noM55_cln_atleast50_4_stage_asx_c_leiden, features = wilcox_top5_NamedGns_w_ids_gns, group.by = "stage_leiden", split.by = "stage_afm", cols = c("#0B4B83FF", "#CB5CF5FF", "#ED9E05FF")) +
  scale_x_discrete(labels = wilcox_top5_NamedGns_w_ids_gn_names) +
  scale_y_discrete(labels = function(x) str_replace_all(str_remove_all(x, "_[A-Za-z]+$| \\(|\\)"), c("Male" = "M", "Female" = "F", "Asexual" = "A"))) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.position = "bottom")
```


```{r, warning=FALSE, message=FALSE, fig.width= 14, fig.height= 5}

leiden_dotplot_clstr_cols <- c("#7E9104FF", "#02E04CFF", "#ED9E05FF", "#ABE884FF", "#428E01FF", 
"#B90D43FF", "#EE4F7BFF", "#D9106CFF", "#F660E3FF", "#89F4AEFF", 
"#0B4B83FF", "#06F58FFF", "#2BB89FFF", "#048988FF", "#CB5CF5FF", 
"#C5D4FDFF", "#EDF057FF") %>% set_names(as.character(c(0:16)))

```

#####################END#### Dotplots for markers for all stages #################################


```{r, warning=FALSE, message=FALSE, fig.width= 7, fig.height=2.5}
map(all_jcC_seu_scvi_noM55_afm_cln[1:3], ~.x@meta.data %>% 
      filter(stage_afm != "Asexual" & stage_fld != "Male (LE)") %>%
      count(stage_fld, leiden, orig.ident) %>%
      filter(n>50) %>%
      group_by(stage_fld) %>%
      mutate(n_dons_all = n_distinct(`orig.ident`))%>%
      group_by(stage_fld, leiden) %>%
      mutate(n_dons_leiden = n_distinct(`orig.ident`),
             propn = n_dons_leiden/n_dons_all) %>%
      ungroup())
```

```{r, warning=FALSE, message=FALSE, fig.width= 6, fig.height=2}
imap(all_jcC_seu_scvi_noM55_afm_cln[1:3], ~.x@meta.data %>% 
       filter(stage_afm != "Asexual" & stage_fld != "Male (LE)") %>%
       count(stage_fld, leiden, orig.ident) %>%
       filter(n>50) %>%
       group_by(stage_fld) %>%
       mutate(n_dons_all = n_distinct(`orig.ident`))%>%
       group_by(stage_fld, leiden) %>%
       mutate(n_dons_leiden = n_distinct(`orig.ident`),
              propn = n_dons_leiden/n_dons_all,
              ) %>%
       ungroup() %>%
       mutate(across(stage_fld, ~str_replace(., "Female \\(LE\\)", "FLE"))) %>%
       ggplot(aes(y = propn, x = factor(leiden), fill = factor(leiden))) +
       geom_point() +
       facet_nested(.~stage_fld,  scales = "free_x", space = "free_x") +
       geom_hline(yintercept = 0.40) +
       labs(y = "Proportion", x = "Leiden clusters") +
       theme_classic() +
       theme(text = element_text(size = 18),
             legend.position = "none", 
             strip.text.x.top = element_text(size = 18))
)
```

```{r, fig.height=6, fig.width=6}
vlcano_mrg <- map(c("bulkDseq", "seuWilcox"), ~{
  EnhancedVolcano(de_wilcox_bulk_scvi %>% filter(cluster == "8"),
                  lab = de_wilcox_bulk_scvi %>% filter(cluster == "8") %>% pull(!!rlang::sym(paste0("Gene_", .x))),
                  # selectLab = .y,
                  x = paste0("avg_log2FC_", .x),
                  y = paste0("p_val_adj_", .x),
                  pCutoff = 1e-10,
                  FCcutoff = 1.5,
                  pointSize = 1.5,
                  labSize = 4.0,
                  title = '',
                  subtitle = "",
                  titleLabSize = 0,
                  subtitleLabSize = 0,
                  axisLabSize = 22,
                  drawConnectors = TRUE,
                  # boxedLabels = TRUE,
                  # legendLabels =c( "p_val_adj"),
                  gridlines.minor = FALSE
  ) +
    theme(legend.position = "none")
})

vlcano_mrg

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```


```{r}
de_wilcox_bulk_scvi %>% ggplot(aes(x = avg_log2FC_seuWilcox, y = avg_log2FC_bulkDseq)) + geom_point(size = 0.001) + facet_wrap(cluster ~ .)
```

```{r, fig.height=6, fig.width=6}
# vlcano_mrg <- map(c("bulkDseq", "seuWilcox"), ~{
#   EnhancedVolcano(de_wilcox_bulk_scvi %>% filter(cluster == "8"),
#                 lab = de_wilcox_bulk_scvi %>% filter(cluster == "8") %>% pull(!!rlang::sym(paste0("Gene_", .x))),

vlcano_mrg_bulk_all <- EnhancedVolcano(bulk_de_all_stgs_anot[[1]] %>% filter(cluster == "8"),
                                       lab = bulk_de_all_stgs_anot[[1]] %>% filter(cluster == "8") %>% pull(!!rlang::sym(paste0("Gene"))),
                                       # selectLab = .y,
                                       x = paste0("avg_log2FC"),
                                       y = paste0("p_val_adj"),
                                       pCutoff = 1e-10,
                                       FCcutoff = 1.5,
                                       pointSize = 1.5,
                                       labSize = 4.0,
                                       title = '',
                                       subtitle = "",
                                       titleLabSize = 0,
                                       subtitleLabSize = 0,
                                       axisLabSize = 22,
                                       drawConnectors = TRUE,
                                       # boxedLabels = TRUE,
                                       # legendLabels =c( "p_val_adj"),
                                       gridlines.minor = FALSE
) +
  theme(legend.position = "none")


vlcano_mrg_bulk_all

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```

```{r, fig.height=6, fig.width=5.5}
vlcano_mrg_bulk_all +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16, angle = 90, hjust = 0.5),
        axis.title = element_text(size = 16))
```


```{r}
# generated in 'm2_raw_integration_trial_qcd_DE.ipynb'
bulk_de_anot <- readRDS(paste0(read_dir_integration, 'bulk_de_anot.RDS'))

```

```{r, fig.height=6, fig.width=6}
vlcano_mrg <- map2(bulk_de_anot[c("all.fLE", "w.5.f_no_asx_dbt")], c("2", "2"), ~{
  EnhancedVolcano(.x %>% filter(cluster == .y),
                  lab = .x %>% filter(cluster == .y) %>% pull(!!rlang::sym(paste0("Gene"))),
                  
                  # bulk_de_anot_all2 <- EnhancedVolcano(bulk_de_anot[["all.fLE"]] %>% filter(cluster == "2"),
                  #                 lab = bulk_de_anot[["all.fLE"]] %>% filter(cluster == "2") %>% pull(!!rlang::sym(paste0("Gene"))),
                  # selectLab = .y,
                  x = paste0("avg_log2FC"),
                  y = paste0("p_val_adj"),
                  pCutoff = 1e-10,
                  FCcutoff = 1.5,
                  pointSize = 1.5,
                  labSize = 4.0,
                  title = '',
                  subtitle = "",
                  titleLabSize = 0,
                  subtitleLabSize = 0,
                  axisLabSize = 22,
                  drawConnectors = TRUE,
                  # boxedLabels = TRUE,
                  # legendLabels =c( "p_val_adj"),
                  gridlines.minor = FALSE
  ) +
    theme(legend.position = "none")
  
})
vlcano_mrg

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```


```{r, fig.height=6, fig.width=5.5}
map(vlcano_mrg, ~.x  +
      theme(axis.text.x = element_text(size = 16),
            axis.text.y = element_text(size = 16, angle = 90, hjust = 0.5),
            axis.title = element_text(size = 16))
)
```

```{r, fig.height=6, fig.width=5.5}
activated_de_fle_canonical <- map(bulk_de_anot[c("all.fLE", "w.5.f_no_asx_dbt")], ~.x %>%
                                    filter(p_val_adj < 0.05 & cluster == "2") ) %>%
  purrr::reduce(inner_join, by = c("gene"), suffix = c("_FLE", "_Canoncl"))
```

```{r, fig.height=6, fig.width=5.5}
activated_de_fle_canonical %>%
  ggplot(aes(x = avg_log2FC_Canoncl, y = avg_log2FC_FLE)) +
  geom_point() +
  geom_abline()
```


```{r, warning=FALSE, message=FALSE}
mca_lp_comt_genes <- activated_de_fle_canonical %>%
  filter(p_val_adj_Canoncl < 0.0001 & p_val_adj_FLE < 0.0001)

mca_lp_comt_genes
```


```{r, warning=FALSE, message=FALSE}
top_labl_genes <- mca_lp_comt_genes |>
  filter(abs(avg_log2FC_Canoncl) > 2.5 & abs(avg_log2FC_FLE) > 2.5) #|>
#arrange(desc(abs(lfc_mean))) #|>
# slice_head(n = 10)


```




```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=5}
mca_lp_comt_genes %>%
  ggplot(aes(x = avg_log2FC_Canoncl, y = avg_log2FC_FLE)) +
  geom_point() +
  geom_point(data = top_labl_genes, size = 1) +
  ggpubr::stat_cor(method = "pearson", label.x =1, label.y = 0, size = 5)+ 
  geom_text_repel(
    data = top_labl_genes,
    aes(label = Gene_Canoncl),
    size = 5,
    segment.size = 0.3,  force = 100
  ) +
  geom_abline() +
  labs(x = "avg Log2FC Canonical", y = "avg Log2FC LE") +
  theme_bw() +
  theme(text = element_text(size = 18)) 

```

```{r, warning=FALSE, message=FALSE, fig.width=6, fig.height=6}
mca_lp_comt_genes %>%
  ggplot(aes(x = avg_log2FC_Canoncl, y = avg_log2FC_FLE)) +
  geom_point() +
  geom_point(data = top_labl_genes, size = 1) +
  ggpubr::stat_cor(method = "pearson", label.x =3, label.y = 0, size = 5)+ 
  geom_text_repel(
    data = top_labl_genes,
    aes(label = Gene_Canoncl),
    size = 5,
    # segment.size = 0.3,  force = 100
  ) +
  geom_abline() +
  labs(x = "avg Log2FC MCA", y = "avg Log2FC Lasse") +
  theme_bw() +
  theme(text = element_text(size = 18)) 

```

```{r, warning=FALSE, message=FALSE}
activated_de_fle_canonical %>%
  ggplot(aes(x = avg_log2FC_Canoncl, y = avg_log2FC_FLE)) +
  geom_point() +
  geom_point(data = top_labl_genes, size = 1) +
  ggpubr::stat_cor(method = "pearson", label.x =4, label.y = 0)+ 
  geom_text_repel(
    data = top_labl_genes %>% filter(avg_log2FC_Canoncl > 0 & avg_log2FC_FLE > 0),
    aes(label = Gene_Canoncl),
    size = 5,
    segment.size = 0.3, segment.color = "black", nudge_x = -2, nudge_y = 0, force = 10
  ) +
  geom_text_repel(
    data = top_labl_genes %>% filter(avg_log2FC_Canoncl < 0 & avg_log2FC_FLE < 0),
    aes(label = Gene_Canoncl),
    size = 5,
    segment.size = 0.3, segment.color = "black", nudge_x = 2, nudge_y = 0, force = 10
  ) +
  geom_abline() +
  theme_bw() +
  theme(text = element_text(size = 18)) 

```



```{r}
de_wilcox_bulk_scvi %>% ggplot(aes(x = -log10(p_val_adj_seuWilcox), y = -log10(p_val_adj_bulkDseq))) + geom_point(size = 0.001)
```

```{r}
de_wilcox_bulk_scvi %>% filter(dset_seuWilcox == "seuWilcox" & dset_bulkDseq == "bulkDseq" )  %>% filter(cluster == "8")
```

####### START Plot scvi UMAP ##############
```{r, message=FALSE, warning=FALSE, eval=T}
# Generated in 'scvi_leiden_clusters_doublet_investigations.Rmd' 
all_jcC_seu_scvi_noM55_afm_cln_umap_mdata <- readRDS(paste0(read_dir_integration, "all_jcC_seu_scvi_noM55_afm_cln_umap_mdata.RDS"))
all_jcC_seu_scvi_noM55_afm_cln_umap_mdata <- all_jcC_seu_scvi_noM55_afm_cln_umap_mdata[["all"]]
```


```{r, message=FALSE, warning=FALSE, eval=T}
# # Generated in 'm2_scvi_to_seurat.Rmd' 
# all_jcC_seu_scvi_no_M55_umap_mdata <- readRDS("data/processed/Pf/m2_all/scvi2seu/all_jcC_seu_scvi_no_M55_umap_mdata_3.RDS")
```

```{r}
all_jcC_seu_scvi_noM55_afm_cln_umap_mdata <- all_jcC_seu_scvi_noM55_afm_cln_umap_mdata %>% 
  mutate("strain" = Strain_DN) 

```


```{r}
msc_w_cln_strns_mdata <- all_jcC_seu_scvi_noM55_afm_cln_umap_mdata %>% 
  mutate(donor = str_remove_all(orig.ident, "SC|.mrna"),
         bcode_don = paste0(donor, "_", bcode),
         # across(donor, ~factor(str_remove(., "C$"), levels = don_order)),
         stage_fld = str_replace_all(stage_fld, c("Mid trophozoite" = "Early trophozoite", ".*schizont" = "Schizont", "Gametocyte \\(developing\\)" = "Early ring")),
         across(stage_fld, ~factor(., levels = stg_refnd_lvls)),
         stage = stage_fld,
         leiden = factor(leiden)#,
         # stage_ag = ifelse(stage_c == "Asexual", "Asexual", "Gametocyte")) 
         # stage_ag = ifelse(str_detect(stage, "ale"), "Gametocyte", "Asexual")
  )  

```

```{r, fig.width=15, fig.height=5}
donor_mdata_slct <- readRDS("data/processed/Pf/m2_all/donor_mdata_slct.RDS")

```


```{r, fig.width=15, fig.height=5}
donor_mdata_slct_deets <- donor_mdata_slct %>%
  mutate(condition_detld = case_when(short_name %in% c("MSC12") ~ "Pf_culture",
                                     condition == "Symptomatic" ~ "Pf_symptomatic_22", 
                                     condition == "Asymptomatic" ~ "Pf_asymptomatic_22", 
                                     is.na(condition) & short_name %in% c("MSC1", "MSC3", "MSC13", "MSC14") ~ "Pf_asymptomatic_21",  
                                     is.na(condition) & short_name %in% c("MSC25", "MSC41", "MSC51") ~ "Pf_mxd_symptomatic_21"),
         groups = factor(case_when(str_detect(condition_detld, "_symptomatic") ~ "Symptomatic",
                                   str_detect(condition_detld, "_asymptomatic") ~ "Asymptomatic",
                                   str_detect(condition_detld, "culture") ~ "Culture"), 
                         levels = c("Asymptomatic", "Symptomatic", "Culture")),
         Group = str_replace_all(condition_detld, "_symptomatic_|_asymptomatic_", " "),
         donor_cln = str_remove(short_name, "SC")
  )


```


```{r}
msc_w_cln_strns_w_don_mdata <- msc_w_cln_strns_mdata  %>% 
  mutate("donor_cln" = str_remove_all(donor, "SC|_SLO|_MACS|C$")) %>% 
  left_join(., donor_mdata_slct_deets %>% 
              select(condition_detld, groups, Group, donor_cln), 
            by = c("donor_cln" = "donor_cln"))

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=10, fig.height=6.5}

msc_w_cln_strns_w_don_mdata %>%
  ggplot(aes(x = umap_1, y = umap_2, colour = stage)) + 
  scale_color_manual(values = sp_fld_colors3)+
  geom_point(size = 0.2) +
  theme_classic() +
  theme(strip.text = element_text(size = 13, face = "bold"),
        text = element_text(size = 13),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank()) +
  guides(colour = guide_legend(title = "Cell QC label", nrow = 2, override.aes = list(size = 4)))

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=5.5, fig.height=6.5}

msc_w_cln_strns_w_don_mdata %>%
  ggplot(aes(x = umap_1, y = umap_2, colour = stage)) + 
  scale_color_manual(values = sp_fld_colors)+
  geom_point(size = 0.2) +
  # labs(x = "UMAP 1", y = "UMAP 2") +
  theme_classic() +
  theme(text = element_text(size = 25),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.line = element_blank(),,
        axis.title = element_blank(),
        legend.title = element_text(size = 25, face = "bold")) +
  guides(colour = guide_legend(title = "Stage", title.position = "top", nrow = 5, override.aes = list(size = 4)))

```


```{r, message=FALSE, warning=FALSE, eval=T}

cluster_centers <- msc_w_cln_strns_w_don_mdata %>%
  group_by(!!rlang::sym("leiden")) %>%
  dplyr::summarize(umap_cntr_1 = median(!!rlang::sym("umap_1")), umap_cntr_2 = median(!!rlang::sym("umap_2")))

msc_w_cln_strns_w_don_mdata %>%
  ggplot(aes(x = !!rlang::sym("umap_1"), y = !!rlang::sym("umap_2"), colour = !!rlang::sym("stage"))) +
  geom_point(size = 0.01) +
  geom_label(data = cluster_centers,      # Add labels at centers
             aes(x = umap_cntr_1, y = umap_cntr_2, label = !!rlang::sym("leiden")), 
             color = "black", 
             fontface = "bold",
             size = 5) +
  scale_color_manual(values = sp_fld_colors)+
  theme_classic() +
  theme(strip.text = element_text(size = 13, face = "bold"),
        text = element_text(size = 13),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank()) +
  guides(colour = guide_colorbar(title = "Stage", direction = "horizontal"))


```

```{r, message=FALSE, warning=FALSE, eval=T}
scvi_umap_cols <- c("#E69F00", "#56B4E9", "#009E73", "#f7f1a0", "#0072B2", 
                    "#D55E00", "#CC79A7", "#845422", "#882255", "#117733", 
                    "#88CCEE", "#DDCC77", "#AA4499", "#44AA99", "#8d97f4", "#8d97f4") %>% set_names(c(0:15))

green_shades <- c("#D9F0A3", "#ADDD8E", "#41AB5D", "#45AB8D", "#238443", "#005A32", "#009E73") %>% set_names(c(15, 10,9,4,2,8, 14))
medium_purple_shades <- c("#8A66D2", "#6A4DBA", "#8d97f4") %>% set_names(c(0, 6, 11))
deep_purple_shades <- c("#F5DEB3","#E3C58F", "#C6A15B", "#A67C00", "#8B5A2B", "#845422") %>% set_names(c(1, 5,7,12,13,16)) 
femLE <- c("#CC79A7") %>% set_names(c(3))
# femLE <- c("#7cf4ec") %>% set_names(c(3))

leiden_cols <- c(green_shades, medium_purple_shades, deep_purple_shades, femLE)
```



```{r, message=FALSE, warning=FALSE, eval=T, fig.width=11, fig.height=3.2}
umap_col_plot_fn <- function(dset = asx_gns_mdata_umap, col_nm = "Stage", clstr_labl = "leiden", group_by_var = "leiden", dim1 = "umap_1", dim2 = "umap_2", pt_size = 0.01, clustr_cols = leiden_cols, is_continuous = "no") {
  
  cluster_centers <- dset %>%
    group_by(!!rlang::sym(clstr_labl)) %>%
    dplyr::summarize(umap_cntr_1 = median(!!rlang::sym(dim1)), umap_cntr_2 = median(!!rlang::sym(dim2)))
  
  dset %>%
    # group_by(!!!rlang::syms(grp_var)) %>%
    # arrange(!!rlang::sym(gene)) %>%
    ggplot(aes(x = !!rlang::sym(dim1), y = !!rlang::sym(dim2), colour = !!rlang::sym(group_by_var))) +
    geom_point(size = pt_size) +
    geom_text(data = cluster_centers,      # Add labels at centers
              aes(x = umap_cntr_1, y = umap_cntr_2, label = !!rlang::sym(clstr_labl)), 
              color = "black", 
              fontface = "bold",
              size = 7) +
    {if(is_continuous == "yes") {
      scale_color_distiller(direction = 1) 
    } else {
      scale_color_manual(values = clustr_cols)
    }} +
    # facet_nested_wrap(vars(!!!rlang::syms(grp_var)), nrow = row_num) +
    labs(x = "UMAP 1", y = "UMAP 2") +
    theme_classic() +
    theme(text = element_text(size = 18),
          legend.position = "bottom",
          axis.text = element_blank(),
          axis.ticks = element_blank()) +
    guides(colour = guide_colorbar(title = col_nm, direction = "horizontal"))
}

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
umap_col_plot_fn(dset = msc_w_cln_strns_w_don_mdata) 
```


################  START ### Get cells from the intergrated object  ################
```{r}
## Generated in 'm2_integration_asx_clustr_gn_explrtn.Rmd'

# all_jcC_seu_scvi_noM55 <- readRDS("data/processed/Pf/m2_all/scvi2seu/all_jcC_seu_scvino_M55.RDS")

```

```{r}
# DimPlot(all_jcC_seu_scvi_noM55, reduction = "scvi_umap", group.by = "stage_fld", cols = sp_col_new)

```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=3}
# DimPlot(all_jcC_seu_scvi_noM55, label = T, reduction = "scvi_umap", group.by = "stage_fld", label.size = 6, label.box = F, label = T) +
#   scale_color_manual(values = seu_cols) +
#   scale_fill_manual(values = seu_cols) +
#   theme_classic() +
#   labs(title = "", x = "UMAP1", y= "UMAP2") +
#   theme(text = element_text(size = 15),
#         legend.position = "none",
#         axis.text = element_blank(),
#         axis.ticks = element_blank(),
#         # axis.title = element_blank()
#         ) 

```

####### END Plot UMAP ##############


####### Female UMAPs Plot UMAP ##############

```{r}
# dsets to plot
fm_dsets <- c("w_pt4_m", "w_pt5_f_no_asx_dbt", "all_fLE")

```


```{r}
# Generated in 'm22_raw_integration_fm.Rmd'
fm_leiden_w_donors_norm_harmony_jnd_umap_mdata <- map(fm_dsets, ~readRDS(paste0("data/processed/Pf/m2_all/integration/fm_leiden_w_donors_norm_harmony_jnd_", .x, "_umap_mdata.RDS")))

```

```{r}
# Generated in 'm22_raw_integration_fm_pseudotime.Rmd'
fm_leiden_w_donors_norm_harmony_jnd_pt_umap_mdata <- map(fm_dsets, ~readRDS(paste0("data/processed/Pf/m2_all/integration/fm_leiden_w_donors_norm_harmony_jnd_", .x, "_pt_umap_mdata.RDS")))

```


```{r}
fm_leiden_w_donors_norm_harmony_jnd_umap_mdata <- map(fm_leiden_w_donors_norm_harmony_jnd_umap_mdata, ~mutate(., across(leiden, factor)))
fm_leiden_w_donors_norm_harmony_jnd_pt_umap_mdata <- map(fm_leiden_w_donors_norm_harmony_jnd_pt_umap_mdata, ~mutate(., across(leiden, factor)))

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(fm_leiden_w_donors_norm_harmony_jnd_umap_mdata, ~umap_col_plot_fn(dset = .x, dim1 = "umapharmonyPCman_1", dim2 = "umapharmonyPCman_2", pt_size = 0.2))
map(fm_leiden_w_donors_norm_harmony_jnd_pt_umap_mdata, ~umap_col_plot_fn(dset = .x, dim1 = "umapharmonyPCman_1", dim2 = "umapharmonyPCman_2", pt_size = 0.2))

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=4}
map(fm_leiden_w_donors_norm_harmony_jnd_pt_umap_mdata, ~umap_col_plot_fn(dset = .x, group_by_var = "slingPseudotime_1", dim1 = "umapharmonyPCman_1", dim2 = "umapharmonyPCman_2", pt_size = 0.2, is_continuous = "yes"))

```
```{r}
# dsets to plot
fm_dsets2 <- c("w.5.f_no_asx_dbt", "all.fLE")
```

```{r}
# Generated in 'm2_integration_fm_clustr_gn_explrtn.Rmd'
fm_leiden_w_donors_norm_harmony_jnd_cln_plots <- map(fm_dsets2, ~readRDS(paste0("data/processed/Pf/m2_all/integration/fm_leiden_w_donors_norm_harmony_jnd_cln_", .x, ".RDS"))) %>% set_names(fm_dsets2)

```

```{r, warning=FALSE, message=FALSE}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata <- map(fm_leiden_w_donors_norm_harmony_jnd_cln_plots, ~{
  .x@meta.data %>% 
    rownames_to_column("bcode") %>%
    left_join(., .x@reductions$umap.harmony_PC_man@cell.embeddings %>% data.frame() %>% rownames_to_column("bcode"), by = "bcode") 
})
```

```{r, warning=FALSE, message=FALSE}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata <- map(fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata, ~.x %>%
                                                                  mutate(across(harmony_clusters_coarse_bulkDE, as.factor))
)
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata, ~umap_col_plot_fn(dset = .x, dim1 = "umapharmonyPCman_1", dim2 = "umapharmonyPCman_2", pt_size = 0.2, clustr_cols = scvi_umap_cols, clstr_labl = "harmony_clusters_coarse_bulkDE", is_continuous = "no", group_by_var = "harmony_clusters_coarse_bulkDE"))

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>%
  ggplot(aes(y = nFeature_RNA, x = harmony_clusters_coarse_bulkDE)) +
  geom_sina(size = 0.01) +
  facet_wrap(. ~ dset)


```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>%
  ggplot(aes(y = log10(nCount_RNA), x = harmony_clusters_coarse_bulkDE)) +
  geom_sina(size = 0.01) +
  facet_wrap(. ~ dset)


```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>%
  ggplot(aes(y = log10(scf_score_norm), x = harmony_clusters_coarse_bulkDE)) +
  geom_sina(size = 0.01) +
  facet_wrap(. ~ dset)


```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>% 
  filter(harmony_clusters_coarse_bulkDE == "2") %>%
  count(dset, donor, harmony_clusters_coarse_bulkDE, Strain_DN) %>%
  ggplot(aes(y = n, x = Strain_DN, fill = Strain_DN)) +
  geom_col() +
  facet_nested(. ~ dset + harmony_clusters_coarse_bulkDE + donor, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))


```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>% 
  # filter(harmony_clusters_coarse_bulkDE == "2") %>%
  count(dset, donor, harmony_clusters_coarse_bulkDE) %>%
  ggplot(aes(y = n, x = donor, fill = donor)) +
  geom_col() +
  facet_nested(. ~ dset + harmony_clusters_coarse_bulkDE, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))


```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>% 
  # filter(harmony_clusters_coarse_bulkDE == "2") %>%
  count(dset, donor, harmony_clusters_coarse_bulkDE) %>%
  group_by(dset, harmony_clusters_coarse_bulkDE)  %>%
  mutate(don_prop = round(n/sum(n), 2)) %>%
  # count(dset, donor, harmony_clusters_coarse_bulkDE) %>%
  ggplot(aes(y = don_prop, x = donor, fill = donor)) +
  geom_col() +
  facet_nested(. ~ dset + harmony_clusters_coarse_bulkDE, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))


```

```{r, fig.width=14, fig.height=4}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>% 
  # filter(harmony_clusters_coarse_bulkDE == "2") %>%
  count(dset, donor, harmony_clusters_coarse_bulkDE) %>%
  group_by(dset, harmony_clusters_coarse_bulkDE)  %>%
  mutate(don_prop = round(n/sum(n), 2)) %>%
  ggplot(aes(y = don_prop, x = donor, fill = donor)) +
  geom_col() +
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5.5, show.legend = F, fontface ='bold', angle = 90)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.4))) +
  facet_nested(. ~ dset + harmony_clusters_coarse_bulkDE, scales = "free_x", space = "free_x") + 
  theme_classic() +
  labs(y = "Proportion", x= "Strain") +
  theme(axis.text.x = element_text(angle = 90),
        strip.text = element_text(size = 18, face = "bold"),
        text = element_text(size = 18),
        legend.position = "none") 

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
fm_leiden_w_donors_norm_harmony_jnd_cln_plots_umap_mdata %>%
  bind_rows(., .id = "dset") %>% 
  filter(harmony_clusters_coarse_bulkDE == "2") %>%
  ggplot(aes(y = sR_actl_score, x = donor, fill = donor)) +
  geom_sina() +
  facet_nested(. ~ dset + harmony_clusters_coarse_bulkDE, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))


```

##### Read UMAPs and metadatas for the female subclusters

```{r, warning=FALSE, message=FALSE}
fem_nms <- c("f")
```

```{r, warning=FALSE, message=FALSE}
# Generated in 'fem_subclusters_pbulkDE.Rmd'
## All cells in the female subclusters
seu_scvi_fem_sub_clusters_umap_mdata <- map(fem_nms, ~readRDS(paste0(read_dir_stage_DE_markers, 'seu_scvi_fem_sub_clusters_umap_mdata_', .x,'.RDS')))
seu_scvi_fem_sub_clusters<- map(fem_nms, ~readRDS(paste0(read_dir_stage_DE_markers, 'seu_scvi_fem_sub_clusters_', .x,'.RDS')))

## Only those clean cells that are taken forward for pseudobulk DE
seu_scvi_fem_sub_clusters_cln4pbulkDE_umap_mdata <- map(fem_nms, ~readRDS(paste0(read_dir_stage_DE_markers, 'seu_scvi_fem_sub_clusters_cln4pbulkDE_umap_mdata_', .x,'.RDS')))
seu_scvi_fem_sub_clusters_cln4pbulkDE<- map(fem_nms, ~readRDS(paste0(read_dir_stage_DE_markers, 'seu_scvi_fem_sub_clusters_cln4pbulkDE_', .x,'.RDS')))

## Pseudobulk objects
pseudo_fem <- map(fem_nms, ~readRDS(paste0(read_dir_stage_DE_markers, 'pseudo_fem_', .x,'.RDS')))
pseudo_fem_hiresfem <- map(fem_nms, ~readRDS(paste0(read_dir_stage_DE_markers, 'pseudo_fem_hiresfem_', .x,'.RDS')))

```

```{r, fig.width=15, fig.height=4}
## convert variables for female subclusters for easy plotting
fem_subcluster_levels <- c("Female", "Female4",  "Female2", "ActivFem", "Female (LE)",  "ActivFemLE")

seu_scvi_fem_sub_clusters[[1]]$fem_subclusters <- droplevels(factor(seu_scvi_fem_sub_clusters[[1]]$fem_subclusters, levels = fem_subcluster_levels))
seu_scvi_fem_sub_clusters_cln4pbulkDE[[1]]$fem_subclusters <- droplevels(factor(seu_scvi_fem_sub_clusters_cln4pbulkDE[[1]]$fem_subclusters, levels = fem_subcluster_levels))

seu_scvi_fem_sub_clusters[[1]]$fem_hires_subclusters <- droplevels(factor(seu_scvi_fem_sub_clusters[[1]]$fem_hires_subclusters, levels = fem_subcluster_levels))
seu_scvi_fem_sub_clusters_cln4pbulkDE[[1]]$fem_hires_subclusters <- droplevels(factor(seu_scvi_fem_sub_clusters_cln4pbulkDE[[1]]$fem_hires_subclusters, levels = fem_subcluster_levels))

pseudo_fem[[1]]$fem_subclusters <- droplevels(factor(pseudo_fem[[1]]$fem_subclusters, levels = fem_subcluster_levels))
pseudo_fem_hiresfem[[1]]$fem_hires_subclusters <- droplevels(factor(pseudo_fem_hiresfem[[1]]$fem_hires_subclusters, levels = fem_subcluster_levels))

```

##### Read DE results


```{r}
# read from 'fem_subclusters_pbulkDE.Rmd'
fem_subclusters_bulk_pwDE_anot <- readRDS(paste0(read_dir_stage_DE_markers, 'fem_subclusters_bulk_pwDE_anot.RDS'))

names(fem_subclusters_bulk_pwDE_anot) <- str_remove(names(fem_subclusters_bulk_pwDE_anot), "f.")
```

##### Preprocess DE results for correlation plots

```{r, fig.height=6, fig.width=5.5}
sufix_2_rename <- c("FLE_inert_v_activ", "Fem_inert_v_activ", "inert_FLE_v_Fem", "actv_FLE_v_Fem", "inertFLE_v_activFem")


activated_innert_de_pw <- 
  fem_subclusters_bulk_pwDE_anot[c( "Female (LE)vsActivFemLE", "FemalevsActivFem", "Female (LE)vsFemale",  "ActivFemLEvsActivFem", "Female (LE)vsActivFem" )] %>%
  map2(., sufix_2_rename, function(df, nm) {
    df %>%
      # filter(p_val_adj < 0.05) %>%
      rename_with(
        ~ paste0(nm, "_", .x),
        .cols = -gene
      )
  }) %>%
  purrr::reduce(inner_join, by = "gene") %>%
  mutate(sig_group_FemDE = case_when(inert_FLE_v_Fem_p_val_adj < 0.05 & actv_FLE_v_Fem_p_val_adj < 0.05 ~ "Sig in both",
                                     inert_FLE_v_Fem_p_val_adj < 0.05 ~ "Sig in inert",
                                     actv_FLE_v_Fem_p_val_adj < 0.05 ~ "Sig in activ",
                                     T ~ "Non sig"),
         sig_group_activDE = case_when(Fem_inert_v_activ_p_val_adj < 0.05 & FLE_inert_v_activ_p_val_adj < 0.05 ~ "Sig in both",
                                       Fem_inert_v_activ_p_val_adj < 0.05 ~ "Sig in Fem",
                                       FLE_inert_v_activ_p_val_adj < 0.05 ~ "Sig in FLE",
                                       T ~ "Non sig"),
         sig_group_inertFLE_v_activFem = case_when(inertFLE_v_activFem_p_val_adj < 0.05 & FLE_inert_v_activ_p_val_adj < 0.05 ~ "Sig in both",
                                                   inertFLE_v_activFem_p_val_adj < 0.05 ~ "Sig in inertFLE_v_activFem",
                                                   FLE_inert_v_activ_p_val_adj < 0.05 ~ "Sig in FLE_inert_v_activ",
                                                   T ~ "Non sig"))

```

```{r, fig.height=6, fig.width=5.5}
activated_innert_de_pw %>%
  filter(if_any(contains("inert_v_activ_p_val_adj"), ~ .x < 0.05)) %>%
  ggplot(aes(x = FLE_inert_v_activ_avg_log2FC, y = Fem_inert_v_activ_avg_log2FC, colour = sig_group_activDE)) +
  geom_point() +
  geom_abline()
```


```{r, fig.height=6, fig.width=5.5}
activated_innert_de_pw %>%
  filter(if_any(contains("FLE_v_Fem_p_val_adj"), ~ .x < 0.05))  %>%
  ggplot(aes(x = inert_FLE_v_Fem_avg_log2FC, y = actv_FLE_v_Fem_avg_log2FC, colour = sig_group_FemDE)) +
  geom_point(size = 0.4) +
  geom_abline()
```


```{r, warning=FALSE, message=FALSE}
mca_lp_comt_genes <- activated_innert_de_pw %>%
  filter(FLE_inert_v_activ_p_val_adj < 0.0001 | Fem_inert_v_activ_p_val_adj < 0.0001)

mca_lp_comt_genes
```


```{r, warning=FALSE, message=FALSE}
top_labl_genes <- mca_lp_comt_genes |>
  filter((abs(FLE_inert_v_activ_avg_log2FC) > 2.5 | abs(Fem_inert_v_activ_avg_log2FC) > 2.5) & !str_detect(actv_FLE_v_Fem_Gene, "PF3D7")) #|>
#arrange(desc(abs(lfc_mean))) #|>
# slice_head(n = 10)

top_labl_genes

```




```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=5}
mca_lp_comt_genes %>%
  ggplot(aes(x = Fem_inert_v_activ_avg_log2FC, y = FLE_inert_v_activ_avg_log2FC, colour = sig_group_activDE)) +
  geom_point() +
  geom_point(data = top_labl_genes, size = 1) +
  ggpubr::stat_cor(method = "pearson", label.x =1, label.y = 0, size = 5)+ 
  geom_text_repel(
    data = top_labl_genes,
    aes(label = Fem_inert_v_activ_Gene),
    size = 5,
    segment.size = 0.3,  force = 100
  ) +
  geom_abline() +
  labs(x = "avg Log2FC Canonical", y = "avg Log2FC LE") +
  theme_bw() +
  theme(text = element_text(size = 18)) 

```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=6}
mca_lp_comt_genes %>%
  ggplot(aes(x = -Fem_inert_v_activ_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC, colour = sig_group_activDE)) +
  geom_point() +
  ggpubr::stat_cor(inherit.aes = FALSE, mapping = aes(x = -Fem_inert_v_activ_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC), method = "pearson", label.x =2, label.y = -0.5, size = 5) + 
  geom_point(data = top_labl_genes, size = 1) +
  geom_text_repel(
    data = top_labl_genes,
    aes(label = Fem_inert_v_activ_Gene),
    size = 6,
    fontface = "bold"
    # segment.size = 0.3,  force = 100
  ) +
  scale_color_brewer(palette = "Set2") +
  geom_abline() +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  labs(x = "Female (Activated VS inert) avg Log2FC", y = "FLE (Activated VS inert) avg Log2FC", title = "Activated VS inert") +
  theme_bw() +
  theme(text = element_text(size = 18),
        legend.position = "bottom") 

```

```{r, fig.width=15, fig.height=4}
map(seu_scvi_fem_sub_clusters_cln4pbulkDE, ~VlnPlot(.x, features = top_labl_genes %>% filter(sig_group_activDE == "Sig in both") %>% pull(gene) %>% .[1:7], group.by = "fem_subclusters", ncol = 7))

```

```{r, fig.width=15, fig.height=4}
map(pseudo_fem, ~VlnPlot(.x, features = top_labl_genes %>% filter(sig_group_activDE == "Sig in both") %>% pull(gene) %>% .[1:7] , group.by = "fem_subclusters", ncol = 7))

```


```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=6}
mca_lp_comt_genes %>%
  ggplot(aes(x = -Fem_inert_v_activ_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC, colour = sig_group_activDE)) +
  geom_point() +
  ggpubr::stat_cor(inherit.aes = FALSE, mapping = aes(x = -Fem_inert_v_activ_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC), method = "pearson", label.x =2, label.y = -0.5, size = 5) + 
  geom_point(data = top_labl_genes, size = 1) +
  geom_text_repel(
    data = top_labl_genes %>% filter(-Fem_inert_v_activ_avg_log2FC > 0 & -FLE_inert_v_activ_avg_log2FC > 0),
    aes(label = Fem_inert_v_activ_Gene),
    size = 5,
    segment.size = 0.3, segment.color = "black", segment.alpha = 0.3, nudge_x = -5, nudge_y = 0, force = 10
  ) +
  geom_text_repel(
    data = top_labl_genes %>% filter(-Fem_inert_v_activ_avg_log2FC < 0 & -FLE_inert_v_activ_avg_log2FC < 0),
    aes(label = Fem_inert_v_activ_Gene),
    size = 5,
    segment.size = 0.3, segment.color = "black", segment.alpha = 0.3, nudge_x = 5, nudge_y = 0, force = 10
  ) +
  geom_abline() +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  labs(x = "Female (Activated VS inert) avg Log2FC", y = "FLE (Activated VS inert) avg Log2FC", title = "Activated VS inert") +
  theme_bw() +
  theme(text = element_text(size = 18),
        legend.position = "bottom") 
```

##### Discrepant genes that are significantly upregulated in FLE activeVSinert but not in Fem activeVSinert

```{r, warning=FALSE, message=FALSE}
discrepant_activeVSinert_FLEup_genes <- mca_lp_comt_genes %>% filter(sig_group_activDE == "Sig in FLE") 

discrepant_activeVSinert_FLEup_genes

```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=6}
mca_lp_comt_genes %>%
  ggplot(aes(x = -Fem_inert_v_activ_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC, colour = sig_group_activDE)) +
  geom_point() +
  ggpubr::stat_cor(inherit.aes = FALSE, mapping = aes(x = -Fem_inert_v_activ_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC), method = "pearson", label.x =2, label.y = -0.5, size = 5) + 
  geom_point(data = discrepant_activeVSinert_FLEup_genes, size = 1) +
  geom_text_repel(
    data = discrepant_activeVSinert_FLEup_genes,
    aes(label = Fem_inert_v_activ_Gene),
    size = 5,
    segment.size = 0.3, segment.color = "black", segment.alpha = 0.3, nudge_x = -1, nudge_y = 1.5, force = 20
  ) +
  geom_abline() +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  labs(x = "Female (Activated VS inert) avg Log2FC", y = "FLE (Activated VS inert) avg Log2FC", title = "Activated VS inert") +
  theme_bw() +
  theme(text = element_text(size = 18),
        legend.position = "bottom") 
```

```{r, fig.width=15, fig.height=8}
map(seu_scvi_fem_sub_clusters_cln4pbulkDE, ~VlnPlot(.x, features = discrepant_activeVSinert_FLEup_genes$gene[1:4] ))

```

```{r, fig.width=15, fig.height=4}
map(seu_scvi_fem_sub_clusters_cln4pbulkDE, ~VlnPlot(.x, features = discrepant_activeVSinert_FLEup_genes$gene[1:7] , group.by = "fem_subclusters", ncol = 7))

```

```{r, fig.width=15, fig.height=4}
map(pseudo_fem, ~VlnPlot(.x, features = discrepant_activeVSinert_FLEup_genes$gene[1:7] , group.by = "fem_subclusters", ncol = 7))

```


```{r, fig.width=15, fig.height=3}
map(seu_scvi_fem_sub_clusters_cln4pbulkDE, ~FeaturePlot(.x, features = discrepant_activeVSinert_FLEup_genes$gene[1:4], reduction = "scvi_umap", order = T, ncol = 4, ))

```

#### Plots for Fem VS FLE in active and also innert
With significance of atleast 0.0001 in either and with some of those with absolute LFC of >2.5 being labelled

```{r, warning=FALSE, message=FALSE}
FemVFLE_genes <- activated_innert_de_pw %>%
  filter(inert_FLE_v_Fem_p_val_adj < 0.0001 | actv_FLE_v_Fem_p_val_adj < 0.0001)


top_FemVFLE_genes <- FemVFLE_genes |>
  filter((abs(inert_FLE_v_Fem_avg_log2FC) > 2.5 | abs(actv_FLE_v_Fem_avg_log2FC) > 2.5) & !str_detect(actv_FLE_v_Fem_Gene, "PF3D7"))  #|>
#arrange(desc(abs(lfc_mean))) #|>
# slice_head(n = 10)

top_FemVFLE_genes

```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=6}
FemVFLE_genes %>%
  ggplot(aes(x = actv_FLE_v_Fem_avg_log2FC, y = inert_FLE_v_Fem_avg_log2FC, colour = sig_group_FemDE)) +
  geom_point() +
  ggpubr::stat_cor(inherit.aes = FALSE, mapping = aes(x = actv_FLE_v_Fem_avg_log2FC, y = inert_FLE_v_Fem_avg_log2FC), method = "pearson", label.x =0.1, label.y = -0.2, size = 5) + 
  geom_point(data = top_FemVFLE_genes, size = 1) +
  geom_text_repel(
    data = top_FemVFLE_genes,
    aes(label = actv_FLE_v_Fem_Gene),
    size = 5,
    # segment.size = 0.3,  force = 100
  ) +
  geom_abline() +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  labs(x = "avg Log2FC actv FLE upregulated", y = "avg Log2FC inert FLE upregulated") +
  theme_bw() +
  theme(text = element_text(size = 18)) 

```


```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=6}
FemVFLE_genes %>%
  ggplot(aes(x = -actv_FLE_v_Fem_avg_log2FC, y = -inert_FLE_v_Fem_avg_log2FC, colour = sig_group_FemDE)) +
  geom_point() +
  ggpubr::stat_cor(inherit.aes = FALSE, mapping = aes(x = -actv_FLE_v_Fem_avg_log2FC, y = -inert_FLE_v_Fem_avg_log2FC), method = "pearson", label.x =0.1, label.y = -0.2, size = 5) + 
  geom_point(data = top_FemVFLE_genes, size = 1) +
  geom_text_repel(
    data = top_FemVFLE_genes %>% filter(-actv_FLE_v_Fem_avg_log2FC > 0 & -inert_FLE_v_Fem_avg_log2FC > 0),
    aes(label = actv_FLE_v_Fem_Gene),
    size = 5,
    fontface = "bold",
    segment.size = 0.3, segment.color = "black", segment.alpha = 0.3, nudge_x = -3, nudge_y = 0, force = 10
  ) +
  geom_text_repel(
    data = top_FemVFLE_genes %>% filter(-actv_FLE_v_Fem_avg_log2FC < 0 & -inert_FLE_v_Fem_avg_log2FC < 0),
    aes(label = actv_FLE_v_Fem_Gene),
    size = 6,
    fontface = "bold",
    segment.size = 0.3, segment.color = "black", segment.alpha = 0.3, nudge_x = 3, nudge_y = 0, force = 10
  ) +
  scale_color_brewer(palette = "Set1") +
  geom_abline() +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  labs(x = "activated (Fem vs FLE) avg Log2FC", y = "inert (Fem vs FLE) avg Log2FC", title = "Female VS Female (LE) ") +
  theme_bw() +
  theme(text = element_text(size = 18),
        legend.position = "bottom") 
```


##### Correlation between the avgLFC from activated versus inert FLEs ie -FLE_inert_v_activ and activated female versus inert FLE ie inertFLE_v_activFem_p_val_adj
If activatedFLE is doublet of (Activated female + FLE) then

FLE_inert_v_activ - is the activate female (Activated female + inert FLE) signal since we remove the inert FLE signal

inertFLE_v_activFem_p_val_adj - is the inert FLE signal (Activated female + inert FLE) since we remove the Activated female signal
The activated female signal from above is the combination of canonical Fem and activation signal which should be captured in the inertFLE_v_activFem and also inertFem VS inert FLE

if there is discrepancy in the correlation between inertFLE_v_activFem and FLE_inert_v_activ then it means the canonical female signal expected from doublet is missing and therefore no doublets

```{r, fig.height=6, fig.width=5.5}
activated_innert_de_pw %>%
  filter(if_any(c("inertFLE_v_activFem_p_val_adj", "FLE_inert_v_activ_p_val_adj"), ~ .x < 0.05))  %>%
  ggplot(aes(x = inertFLE_v_activFem_avg_log2FC, y = FLE_inert_v_activ_avg_log2FC, colour = sig_group_inertFLE_v_activFem)) +
  geom_point(size = 0.4) +
  geom_abline()
```


```{r, warning=FALSE, message=FALSE}
doublet_investigation_genes <- activated_innert_de_pw %>%
  filter(inertFLE_v_activFem_p_val_adj < 0.0001 | FLE_inert_v_activ_p_val_adj < 0.0001)

doublet_investigation_genes
```

```{r, warning=FALSE, message=FALSE}
FLE_inert_v_activ_genes <- doublet_investigation_genes %>% filter(sig_group_inertFLE_v_activFem == "Sig in FLE_inert_v_activ") 

FLE_inert_v_activ_genes

```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=6}
doublet_investigation_genes %>% 
  ggplot(aes(x = -inertFLE_v_activFem_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC, colour = sig_group_inertFLE_v_activFem)) +
  geom_point() +
  ggpubr::stat_cor(inherit.aes = FALSE, mapping = aes(x = -inertFLE_v_activFem_avg_log2FC, y = -FLE_inert_v_activ_avg_log2FC), method = "pearson", label.x =2, label.y = -0.5, size = 5) + 
  geom_point(data = FLE_inert_v_activ_genes, size = 1) +
  geom_text_repel(
    data = FLE_inert_v_activ_genes,
    aes(label = Fem_inert_v_activ_Gene),
    size = 5,
    segment.size = 0.3, segment.color = "black", segment.alpha = 0.3, nudge_x = -1, nudge_y = 1.5, force = 20
  ) +
  geom_abline() +
  geom_vline(xintercept = 0) +
  geom_hline(yintercept = 0) +
  labs(x = "Activated Female VS inert FLE avg Log2FC", y = "FLE (Activated VS inert) avg Log2FC", title = "Activated VS inert") +
  theme_bw() +
  theme(text = element_text(size = 18),
        legend.position = "bottom") 
```



#### Plots to assess the likelihood of activated FLE being doublets

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
seu_scvi_fem_sub_clusters_umap_mdata_list <- list(seu_scvi_fem_sub_clusters_umap_mdata[[1]], seu_scvi_fem_sub_clusters_cln4pbulkDE_umap_mdata[[1]]) %>% set_names(c("fem_clstrs", "fem_clstrs_cln4pbulkDE")) %>%
  map(., ~.x %>% mutate(across(c(fem_subclusters, fem_hires_subclusters), ~droplevels(factor(., levels = fem_subcluster_levels)))))

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=4, fig.height=2.5}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      ggplot(aes(y = nFeature_RNA, x = fem_hires_subclusters)) +
      geom_sina(size = 0.01) +
      labs(x = "Female clusters", y = "Gene count") +
      theme_classic() +
      theme(text = element_text(size = 18),
            axis.text.x =element_text(size = 18, angle = 20, hjust = 1) )
    
)

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      ggplot(aes(y = log10(nCount_RNA), x = fem_hires_subclusters)) +
  geom_sina(size = 0.01) 
)

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      ggplot(aes(y = log10(scf_score_norm), x = fem_hires_subclusters)) +
  geom_sina(size = 0.01) 

)
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      filter(str_detect(fem_hires_subclusters, "Activ")) %>%
  count(donor, fem_hires_subclusters, Strain_DN) %>%
  ggplot(aes(y = n, x = Strain_DN, fill = Strain_DN)) +
  geom_col() +
  facet_nested(. ~ fem_hires_subclusters + donor, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))
)

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      count(donor, fem_hires_subclusters) %>%
  ggplot(aes(y = n, x = donor, fill = donor)) +
  geom_col() +
  facet_nested(. ~ fem_hires_subclusters, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))

)
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      count(donor, fem_hires_subclusters) %>%
  group_by(fem_hires_subclusters)  %>%
  mutate(don_prop = round(n/sum(n), 2)) %>%
  # count(donor, fem_hires_subclusters) %>%
  ggplot(aes(y = don_prop, x = donor, fill = donor)) +
  geom_col() +
  facet_nested(. ~ fem_hires_subclusters, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))
)

```

```{r, fig.width=12, fig.height=2.6}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      count(donor, fem_hires_subclusters) %>%
  group_by(fem_hires_subclusters)  %>%
  mutate(across(donor, ~str_remove(., "SC")),
         don_prop = round(n/sum(n), 2)) %>%
  ggplot(aes(y = don_prop, x = donor, fill = donor)) +
    scale_fill_manual(values = donor_cols) +
  geom_col() +
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5.5, show.legend = F, fontface ='bold', angle = 90)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.4))) +
  facet_nested(. ~ fem_hires_subclusters, scales = "free_x", space = "free_x") + 
  theme_classic() +
  labs(y = "Proportion", x= "Strain") +
  theme(axis.text.x = element_text(angle = 90, size = 16),
        axis.text.y = element_text( size = 16),
        strip.text = element_text(size = 18, face = "bold"),
        text = element_text(size = 18),
        legend.position = "none") 
)
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=16, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~.x %>%
      filter(str_detect(fem_hires_subclusters, "Activ")) %>%
  ggplot(aes(y = sR_actl_score, x = donor, fill = donor)) +
  geom_sina() +
  facet_nested(. ~ fem_hires_subclusters, 
               scales = "free_x", 
               space = "free_x") + 
  theme(axis.text.x = element_text(angle = 90))

)
```

Umap plot for female subclusters
```{r, fig.width=4.5, fig.height=4.5}
map(seu_scvi_fem_sub_clusters, ~DimPlot(.x, group.by = "fem_subclusters") + scale_color_manual(values = sp_fld_colors2) +
      theme(legend.position = "bottom",
            plot.title = element_blank(), 
            text = element_text(size = 14),
            axis.title = element_blank(),
            axis.text = element_blank(),
            axis.ticks = element_blank()) +
      guides(colour = guide_legend(title = "Female subclusters", title.position = "top", nrow = 1, override.aes = list(size = 4)))) 

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~umap_col_plot_fn(dset = .x, is_continuous = "no", clstr_labl = "leiden", group_by_var = "fem_subclusters", col_nm = "fem_subclusters", clustr_cols = sp_fld_colors2) )
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(seu_scvi_fem_sub_clusters_umap_mdata_list, ~umap_col_plot_fn(dset = .x, is_continuous = "no", clstr_labl = "leiden", group_by_var = "fem_hires_subclusters", col_nm = "fem_hires_subclusters", clustr_cols = sp_fld_colors2) )
```

```{r}
## Comparisons to plot
fem_clusters_comp_nms <- c("FLE activated VS inert", "F activated VS inert", "Inert F V FLE", "Activated F V FLE")

# 1. Define the pairs of clusters you want to turn "lightgrey"
grey_pairs <- list(
  c("ActivFem", "Female"),
  c("ActivFemLE", "Female (LE)"),
  c("ActivFem", "ActivFemLE"),
  c("Female (LE)", "Female")
)

# 2. Run the nested loop
all_plots <- map2(grey_pairs, fem_clusters_comp_nms, function(pair, nms) {
  # Create a temporary color vector for this specific iteration
  tmp_colors <- replace(sp_fld_colors2, pair, "lightgrey")
  
  # Generate the plots for all objects in seu_scvi_fem_sub_clusters
  map(seu_scvi_fem_sub_clusters, ~DimPlot(.x, group.by = "fem_subclusters") + 
        scale_color_manual(values = tmp_colors) + 
        NoLegend() +
        labs(title = nms) +
        theme(axis.text = element_blank(),
              axis.title = element_blank())
  )
}) %>% 
  list_flatten() 

# all_plots
```

```{r, fig.width=10, fig.height=2.5}
plot_grid(plotlist = all_plots, nrow = 1)
```

### Leiden average expression plots

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
avg_exprsn_fem_clusters_data <- map(seu_scvi_fem_sub_clusters_cln4pbulkDE, ~AverageExpression(.x, group.by = "fem_hires_subclusters", layer = "data") %>% .$RNA %>% as.data.frame() %>% rownames_to_column("gene"))

avg_exprsn_fem_clusters_counts <- map(seu_scvi_fem_sub_clusters_cln4pbulkDE, ~AverageExpression(.x, group.by = "fem_hires_subclusters", layer = "counts") %>% .$RNA %>% as.data.frame() %>% rownames_to_column("gene"))

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
avg_exprsn_fem_clusters_data
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map(list(avg_exprsn_fem_clusters_data[[1]], avg_exprsn_fem_clusters_counts[[1]]), ~.x %>%
  ggplot(aes(x = `Female2`, y = `Female4`)) +
  geom_point(size = 0.01) +
  geom_abline()
)

```



#### GGgally for pairwise comparisons across all clusters

```{r}
lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")
}

# Add this custom function to your environment
cor_heatmap <- function(data, mapping, ...) {
  GGally::ggally_cor(data, mapping, ...) + 
    theme(panel.background = element_rect(fill = "white")) # Ensures background doesn't override colors
}

```

```{r}
## pairwise correlation between female leiden subclusters
map(list(avg_exprsn_fem_clusters_data[[1]], avg_exprsn_fem_clusters_counts[[1]]), ~.x %>%
  column_to_rownames("gene") %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)
)
```

```{r}
# Define a helper to ensure panel borders are visible
my_ggpairs_theme <- theme_bw() + # Starting with theme_bw provides the panel boxes
  theme(
    axis.text.x = element_text(angle = 90, size = 11),
    axis.text.y = element_text(size = 11),
    strip.text = element_text(size = 16, face = "bold"),
    text = element_text(size = 16),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.border = element_rect(colour = "black", fill = NA, linewidth = 0.5), # Grid surrounding facets
    legend.position = "none"
  )
```


```{r, fig.width=8, fig.height=7}
## pairwise correlation between female leiden subclusters
map(list(avg_exprsn_fem_clusters_data[[1]], avg_exprsn_fem_clusters_counts[[1]]), ~.x %>%
  column_to_rownames("gene") %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          # upper = list(continuous = wrap("cor", size = 5)), # 'cor' automatically handles heat colors if mapped,
          upper = list(continuous = wrap(cor_heatmap, size = 5)),
          progress = FALSE) +
    theme_bw() +
  my_ggpairs_theme
    
)


```

```{r}
## pairwise correlation of each gene's expression averaged across all cells between female leiden subclusters - less than 50 transcripts
map(list(avg_exprsn_fem_clusters_data[[1]], avg_exprsn_fem_clusters_counts[[1]]), ~.x %>%
  column_to_rownames("gene") %>%
  filter(if_all(everything(), ~ .x <= 50)) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)
)
```

```{r}
## pairwise correlation of each gene's expression averaged across all cells between female leiden subclusters - log transformed
map(list(avg_exprsn_fem_clusters_data[[1]], avg_exprsn_fem_clusters_counts[[1]]), ~.x %>%
  column_to_rownames("gene") %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE) +
    
  scale_x_log10() +
  scale_y_log10() 
    
)
```


!! NOTE - PASTED from a different notebook and may not work - will need to be tweaked to work
```{r, warning=FALSE, message=FALSE}

cor_fun <- function(data, mapping, method="pearson", use="pairwise", ndp=2, sz=5, stars=TRUE, ...){

# grab data
x <- eval_data_col(data, mapping$x)
y <- eval_data_col(data, mapping$y)

# calculate correlation: for significance stars
corr <- cor.test(x, y, method=method)
est <- corr$estimate
lb.size <- sz* abs(est)

# get significance stars
if(stars){
  stars <- c("***", "**", "*", "")[findInterval(corr$p.value, c(0, 0.001, 0.01, 0.05, 1))]
  lbl <- paste0(round(est, ndp), stars)
}else{
  lbl <- round(est, ndp)
}
# calculate correlation: for colored tiles
corr <- cor(x, y, method=method, use=use)

# calculate color based on correlation value
# corr = -1 => blue, 
# corr =  0 => white, 
# corr = +1 => red, 
colFn <- colorRampPalette(c("blue","white", "red"), interpolate ='spline')
fill <- colFn(100)[findInterval(corr, seq(0, 1, length=100))]

ggplot(data = data, mapping = mapping, ...) + 
  theme_void() +
  annotate("text",
           x=mean(x, na.rm=TRUE),
           y=mean(y, na.rm=TRUE),
           label=lbl,
           # size=lb.size,
           size=4,
           ...) +
  theme(panel.background = element_rect(fill=fill,  # to fill background of panel with color
                                        colour=NA), # to remove border of panel
        panel.grid.major = element_blank())
}


## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Early trophozoite', 'Late ring', 'Female (LE)', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = c('Female (LE)', 'Male (LE)', 'Early trophozoite', 'Late ring')) %>% 
  select(where(~!all(is.na(.x)))) %>% 
  column_to_rownames('gene') %>% 
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          # upper = list(continuous = wrap("cor", size = 5)))
          upper = list(continuous = cor_fun),
          progress = FALSE)

```
#### UMAPs for asexual + active female doublets

```{r, warning=FALSE, message=FALSE}
# Generated in 'fem_subclusters_DE.Rmd'
all_jcC_seu_scvi_stg_noM55_ActivAsexDbts_umap_mdata <- readRDS(paste0(save_dir_stage_DE_markers, 'all_jcC_seu_scvi_stg_noM55_ActivAsexDbts_umap_mdata.RDS'))

```


#### UMAPs for activated females in the independent seurat objects
```{r, warning=FALSE, message=FALSE}
# Generated in 'fem_subclusters_DE.Rmd'
all_anotd_stg_cln_norm_w_fem_subclusters_umap_mdata <- readRDS(paste0(save_dir_stage_DE_markers, 'all_anotd_stg_cln_norm_w_fem_subclusters_umap_mdata.RDS'))

```



