---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(SummarizedExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(ComplexHeatmap)
  library(circlize)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  conflict_prefer("colData", "SummarizedExperiment")
  
})
```


```{r setup}
sample_set = "Mali2"

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```


#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/14_strain_DE/strain_DE_common_fns.R")
```

```{r}
optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)

optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]
```


```{r}
read_dir <- "data/processed/Pf/m2_all/strain_DE_afm/"
save_dir <- "data/processed/Pf/m2_all/strain_DE_afm/"
```

Integrated asexual
```{r}
## subset where field parasites are for slingshot
v3_afm_harmony_ss_res_fld_pt <- readRDS(paste0(save_dir, 'm2_afm_bal_seu.RDS'))

```

```{r, warning=FALSE, message=FALSE}
# names(v3_afm_harmony_ss_res_fld_pt) <- c("f", "m", "a")
```

## MAST DE of interesting clusters and also for strains


```{r}

map(v3_afm_harmony_ss_res_fld_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_ss", pt.size = 0.5, dims = c(1,2), label = T))

```

```{r, fig.width=30, fig.height=6}
map(v3_afm_harmony_ss_res_fld_pt, ~FeaturePlot(.x, features = "PF3D7-0925700", reduction = "umap.harmony_PC_man", split.by = "donor", pt.size = 2, ncol = 7))
```



```{r}
st_dist <- map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, condition, StrainDon) %>% filter(n>20) %>% group_by(donor) %>% filter(n() > 1 ))

```

```{r}
map(st_dist, ~.x %>% 
  # mutate(across(condition, ~str_replace_all(., c("No" = "Asymptomatic", "Yes" = "Symptomatic")))) %>%
  ggplot(aes(x = donor, fill = StrainDon, y = n)) + 
  geom_col(position = position_dodge2()) + facet_wrap(condition~., scales = "free_x")
)
```


```{r}
map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% 
  count(donor, condition, StrainDon) %>% 
  # mutate(across(Symptomatic, ~str_replace_all(., c("No" = "Asymptomatic", "Yes" = "Symptomatic")))) %>% 
  ggplot(aes(x = donor, fill = StrainDon, y = n)) + geom_col()) 
```


```{r, fig.width= 10, fig.height= 5}
map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>%
  count(donor, Symptomatic, StrainDon) %>% filter(n>20) %>% 
  # mutate(across(Symptomatic, ~str_replace_all(., c("No" = "Asymptomatic", "Yes" = "Symptomatic")))) %>% 
  ggplot(aes(x = donor, fill = StrainDon, y = n)) + geom_col(position = position_dodge2()) + facet_wrap(Symptomatic~., scales = "free_x") + theme_classic()+ theme(legend.position = "right", axis.text.x = element_text(angle = 45, hjust = 1), text = element_text(size = 20))) 
```

######################### Read in slingshot SCEs needed for tradeseq #########################
```{r, message=F, warning=F, eval=T}
## Get file list
fm_leiden_w_donors_pt_list_files <- Sys.glob("data/processed/Pf/m2_all/integration/fm_leiden_w_donors_*_sce_ss.RDS")

fm_leiden_w_donors_pt_list_files
```

```{r, message=F, warning=F, eval=T}
## Subset to candidate ones
v3_fm_harmony_sce_ss_list_files <- fm_leiden_w_donors_pt_list_files[str_detect(fm_leiden_w_donors_pt_list_files, "w_pt7_f_sce_ss|w_pt4_m_sce_ss|w_pt5_f_sce_ss|w_pt5_f_sce_ss")]

v3_fm_harmony_sce_ss_list_files
```


```{r, message=F, warning=F, eval=T}
## read filtered barcodes
v3_fm_harmony_sce_ss <- map(v3_fm_harmony_sce_ss_list_files, ~.x %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/m2_all/integration/fm_leiden_w_donors_|_sce_ss.RDS')))) %>% 
  unlist(., recursive = F) %>% 
  # .[sample_name] %>%
  map(readRDS)

```


```{r, eval=T}
# read asex sce
v3_m2_asex_harmony_sce_ss <- readRDS("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/asex_noM55_jcC_harmony_no_comtd_sce_ss3.RDS")

```


```{r, warning=FALSE, message=FALSE}
v3_afm_harmony_sce_ss <- c(list(v3_m2_asex_harmony_sce_ss), v3_fm_harmony_sce_ss[c("w_pt7_f", "w_pt4_m", "w_pt5_f", "w_pt5_f")])
```


```{r, warning=FALSE, message=FALSE}
names(v3_afm_harmony_sce_ss) <- c("a", "f", "m", "f2", "f3")
```
######################### Read in slingshot SCEs needed for tradeseq #########################

Subset SCE with the seurat balanced object used for MAST DE to have the same cells that went into MASt DE also go into tradeseq DE
```{r, eval=T}
## !!! NOTE - label MSC14 as asymptomatic
v3_afm_harmony_sce_ss <- map2(v3_afm_harmony_sce_ss, v3_afm_harmony_ss_res_fld_pt[names(v3_afm_harmony_sce_ss)], ~.x[, colnames(.y)])

```

```{r, warning=FALSE, message=FALSE}
for (nm in names(v3_afm_harmony_sce_ss)) {
  SummarizedExperiment::colData(v3_afm_harmony_sce_ss[[nm]])[,colnames(v3_afm_harmony_ss_res_fld_pt[[nm]]@meta.data)] <- v3_afm_harmony_ss_res_fld_pt[[nm]]@meta.data
}
```

```{r, eval=T}
# ## !!! NOTE - Make same as the seurat object processed above
# new_seu_mdata = colnames(v3_afm_harmony_ss_res_fld_pt@meta.data)[!colnames(v3_afm_harmony_ss_res_fld_pt@meta.data) %in% colnames(v3_afm_harmony_sce_ss@colData)]

# colData(v3_afm_harmony_sce_ss)[,new_seu_mdata] <- v3_afm_harmony_ss_res_fld_pt@meta.data[,new_seu_mdata]
# colData(v3_afm_harmony_sce_ss)[,colnames(v3_afm_harmony_ss_res_fld_pt@meta.data)] <- v3_afm_harmony_ss_res_fld_pt@meta.data

```

Save SCE for pseudobulk DE attempt

```{r}
## Save for pseudotime correlatation with bulk in notebook m2_corr2pubd_bulk.Rmd
imap(v3_afm_harmony_sce_ss, ~saveRDS(.x, paste0(save_dir, "v3_afm_harmony_sce_ss_4pbulkDE_", .y, ".RDS")))

```

```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
# v3_afm_harmony_sce_ss_df <- ptime_chunk_calc_df_fn(sc_df = v3_afm_harmony_sce_ss@colData, ptime = "pseudotime", chunk_num = 20, p_cat_nm = "ptime_cat", p_bal_cat_nm = "ptime_bal_cat")
```


```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
fm_tr_sce_pt_chunks_sshot <- v3_afm_harmony_sce_ss
# fm_tr_sce_pt_chunks_sshot@colData[,colnames(v3_afm_harmony_sce_ss_df)] <- v3_afm_harmony_sce_ss_df

```


```{r}
map(fm_tr_sce_pt_chunks_sshot, ~plotReducedDim(.x, dimred = "UMAP.HARMONY_PC_MAN", colour_by = "ptime_bal_cat"))

```

```{r}
## Save for pseudotime correlatation with bulk in notebook m2_corr2pubd_bulk.Rmd
## Subset seurat with pseudotime and add it and save
v3_afm_harmony_ss_res_fld_pt_ptime1 <- map(v3_afm_harmony_ss_res_fld_pt, ~.x[,colnames(.x)])

```


```{r}
## Save for pseudotime correlatation with bulk in notebook m2_corr2pubd_bulk.Rmd
imap(v3_afm_harmony_ss_res_fld_pt_ptime1, ~saveRDS(.x, paste0(save_dir, "v3_afm_harmony_ss_res_fld_pt_ptime1_", .y, ".RDS")))

rm(v3_afm_harmony_ss_res_fld_pt_ptime1)
```
Before removal of strain grous with less than 100
```{r, fig.width=10, fig.height=3}
plt_count_fn <- function(dset ){
  dset %>%
  mutate(group = str_remove(don_stage, ".MSC.*")) %>%
  count(group, donor,  strain) %>% 
  mutate(across(donor, ~str_remove_all(., "SC|SLO|C|_"))) %>% 
  group_by( donor) %>% 
  mutate(strn_n = n_distinct(strain)) %>% 
  filter(strn_n >1) %>%
  ggplot(aes(x = strain, y = n)) +
  geom_col() +
  # geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.45))) +
  facet_nested_wrap(vars(group, donor), nrow = 1, scales = "free_x") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        strip.text = element_text(size = 14),
        text = element_text(size = 14)) 
}
# )

map(fm_tr_sce_pt_chunks_sshot, ~colData(.x)[, c("donor", "strain")] %>% 
  data.frame()) %>%
  bind_rows(.id = "don_stage") %>%
  plt_count_fn(.)
```

```{r}
# Count the donors with atleast a 100 for atleast one strain
don_100n <- map(v3_afm_harmony_ss_res_fld_pt, ~{
  .x@meta.data %>% 
    count(orig.ident, StrainDon) %>% 
    filter(n>100) %>% 
    group_by(orig.ident) %>% 
    filter(n_distinct(StrainDon) > 1) %>% 
    pull(orig.ident) %>% 
    unique() %>% 
    str_remove(., ".mrna")
  })

dons_bal_sset_nms <- don_100n
dons_bal_sset_nms
```


```{r, eval=T}
# Subset donors with atleast 2 strain per donor into independent objects for DE analysis between strains per donor.
# dons_bal_sset_nms <- str_remove_all(str_replace_all(dons_bal_sset, "M", "MSC"), "_$|C_$|\\|")
dons_bal_sset_nms <- don_100n

# fm_sce <- map(dons_bal_sset_nms, ~fm_tr_sce_pt_chunks_sshot[, str_detect(colnames(fm_tr_sce_pt_chunks_sshot),.x) & !str_detect(fm_tr_sce_pt_chunks_sshot@colData$StrainDon, "Negative|Doublet")]) %>% set_names(dons_bal_sset_nms)

fm_sce <- map2(fm_tr_sce_pt_chunks_sshot, dons_bal_sset_nms[names(fm_tr_sce_pt_chunks_sshot)], ~{
  dset <- .x
  dons <- .y

  map(dons, ~dset[, dset@colData$orig.ident == paste0(.x, ".mrna")]) %>%
    set_names(dons)

}) %>% 
  unlist(recursive = T)

```


```{r, warning=FALSE, message=FALSE}
ds_grp <- rep('StrainDon', length(unlist(dons_bal_sset_nms)))

```



```{r, warning=FALSE, message=FALSE}
######################### !!!!!NOTE - Unbalanced alternative to increase power replacing balanced with unbalanced - switch to return balanced or viceversa 
fm_bal_sce <- fm_sce
# fm_bal_sce <- fm_tr_sce_pt_chunks_sshot
```

```{r, eval=T}
map(fm_bal_sce, ~.x@colData[,c("ptime_bal_cat", "StrainDon")] %>% data.frame() %>% dplyr::count(ptime_bal_cat, StrainDon))
```

```{r, eval=FALSE}
map(fm_bal_sce, ~.x@colData[,c("ptime_bal_cat", "StrainDon")] %>% data.frame() %>% dplyr::count(ptime_bal_cat, StrainDon))
```

```{r}
map(fm_bal_sce, ~plotReducedDim(.x, dimred = "UMAP.HARMONY_PC_MAN", colour_by = "ptime_bal_cat"))

```

```{r}
map(fm_bal_sce, ~.x@colData[,c("ptime_bal_cat", "StrainDon")] %>% data.frame() %>% dplyr::count(ptime_bal_cat, StrainDon))
```

Before removal of 100 below
```{r, fig.width=10, fig.height=3, eval=FALSE}
map(fm_bal_sce, ~colData(.x)[, c("donor", "strain")] %>% 
  data.frame()) %>%
  bind_rows(.id = "don_stage") %>%
  plt_count_fn(.)

```

```{r, fig.width=10, fig.height=3, eval=T}
map(fm_bal_sce, ~colData(.x)[, c("donor", "strain")] %>% 
  data.frame()) %>%
  bind_rows(.id = "don_stage") %>%
  plt_count_fn(.)

```

### Retain cells with strains having more than 10 in each stage


```{r, eval=FALSE}
##Remove groups with less than 10

fm_bal_sce <- map2(fm_bal_sce, ds_grp, ~{
  .x[,.x@colData[, .y, drop = F] %>% 
       data.frame()  %>%
       rownames_to_column('bcode') %>% 
       group_by(!!rlang::sym(.y)) %>% 
       filter(n() >=100) %>% 
       pull(bcode)])

map2(fm_bal_sce, ds_grp, ~table(.x@colData[, .y]) )

```

After removal of 100 
```{r, fig.width=10, fig.height=3}
map(fm_bal_sce, ~colData(.x)[, c("donor", "strain")] %>% 
  data.frame()) %>%
  bind_rows(.id = "don_stage") %>%
  plt_count_fn(.)
```

```{r}
map(fm_bal_sce, ~.x@colData[,c("ptime_bal_cat", "StrainDon")] %>% data.frame() %>% dplyr::count(ptime_bal_cat, StrainDon))
```

```{r}
# Count the donors with atleast 2 strains each with atleast 100 cells
# fm_bal_sce <- fm_bal_sce[unlist(map2(fm_bal_sce, ds_grp, ~length(table(.x@colData[, .y]) ) > 1))]

```


```{r}
## Get only the asexuals
dons_bal_sset_nms_flat <- unlist(dons_bal_sset_nms, recursive = F)
names(dons_bal_sset_nms_flat) <- str_remove(names(dons_bal_sset_nms_flat), "[0-9]+$")
dons_bal_sset_nms_flat_nms_val <- paste0(names(dons_bal_sset_nms_flat), ".", dons_bal_sset_nms_flat)

```

```{r}
## Get only the asexuals
#¡¡¡¡¡NOTE - NEED TO FIND OUT WHAT THIS DOES - SEEMS TO REPEAT WHAT IS DONE IN LINE 261
# fm_bal_sce <- fm_bal_sce[names(fm_bal_sce) %in% dons_bal_sset_nms_flat_nms_val]

```

## Strain VS strain
### Balance (subsample) the strain cells within pseudotime chunks


```{r, results="asis", fig.width=4, fig.height=3}
library("ggpubr")

umap_meta_df <- map(v3_afm_harmony_sce_ss, ~{
  dset = .x
  
  # print(str(dset))
  
  reducedDims(dset)$UMAP.HARMONY_PC_MAN[,c(1,2)] %>% as.data.frame() %>% rownames_to_column("cell_bc") %>% 
    left_join(., dset@colData[,c("stage", "slingPseudotime_1", "orig.ident", "StrainDon", "Symptomatic", "donor", "strain", "dsetC", "ptime_bal_cat")] %>%
                as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc") #%>% 
# mutate(across(strain_dset, ~factor(., levels = rev(strain_dset_lvls_don)))) %>% arrange(strain_dset
})

##Remove groups with less than 20
# umap_meta_df <- umap_meta_df %>% group_by(StrainDon) %>% filter(n() >=20) %>% filter(!is.na(strain))

tm <- theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 15) ,
        legend.title = element_text(size =15, face = "bold") )

gbox_theme<- theme(axis.text=element_text(size=15),
        axis.title=element_text(size=15,face="bold"), 
        legend.text = element_text(size = 15) ,
        legend.title = element_text(size = 15, face = "bold"),
        legend.position = "bottom")
        
dens_thme <- theme(axis.text=element_text(size=22),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=22,face="bold"), 
        legend.text = element_text(size = 15) ,
        legend.title = element_text(size = 15, face = "bold"),
        legend.position = "bottom")

# smpl = map(fm_sce, ~umap_meta_df[umap_meta_df$cell_bc %in% colnames(.x),]  %>% 
#   count(StrainDon) %>% pull(n) %>% min())
# 
# tbl = map2(fm_sce,smpl, ~umap_meta_df[umap_meta_df$cell_bc %in% colnames(.x),]   %>%
#   group_by(StrainDon) %>% 
#   slice_sample(n = .y) )

tbl = imap(fm_bal_sce, ~umap_meta_df[[str_remove(.y, "\\..*")]][umap_meta_df[[str_remove(.y, "\\..*")]]$cell_bc %in% colnames(.x),])
all_cells_tbl = map(names(fm_bal_sce), ~umap_meta_df[[str_remove(.x, "\\..*")]][umap_meta_df[[str_remove(.x, "\\..*")]]$donor == str_remove(.x, ".*\\."), ])


## !!!!!!!!NOTE - Check later
ds_grp <- rep('StrainDon', length(fm_bal_sce))

map2(c(tbl, all_cells_tbl), rep(ds_grp, 2), ~.x%>%
  ggplot(., aes(x = slingPseudotime_1, col=!!rlang::sym(.y), fill=!!rlang::sym(.y))) + 
  geom_density(aes(y = ..count..*0.3), size=2, bw = 0.3, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  dens_thme  +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))
)

map2(c(tbl, all_cells_tbl), rep(ds_grp, 2), ~.x%>%
  ggboxplot(., y = "slingPseudotime_1",x= .y, col=.y) + 
  stat_compare_means(aes(label = paste0("P = ", after_stat(p.format))), method = "wilcox.test", label.x = 1.25,
  label.y = 4) +
    gbox_theme+
  theme(legend.position = "none")+
  labs(x = "Strain", y="Pseudotime")
)
  


```


```{r, results="asis", fig.width=4, fig.height=3}
# bal100_de_cells <- map(fm_bal_sce, ~.x@colData[, c("ptime_bal_cat", "StrainDon")] %>% as.data.frame()) %>% bind_rows(.id = "donor") %>% mutate(bal_de = "yes") %>% rownames_to_column("cell_bc")
bal100_de_cells <- map(fm_bal_sce, ~.x@colData[, c("ptime_bal_cat", "StrainDon")] %>% as.data.frame() %>% rownames_to_column("cell_bc")) %>% 
  bind_rows(.id = "donor") %>% 
  mutate(bal_de = "yes")  %>% 
  separate_wider_delim(donor, delim = ".", names = c("stage", "donor")) %>% 
  split(., f = factor(.$stage))

# umap_meta_df_bound <- umap_meta_df %>% bind_rows(.id = "donor") %>% mutate(across(donor, ~paste0(.x, ".", str_remove(orig.ident, ".mrna"))))
umap_meta_df_bound <- umap_meta_df

# umap_meta_df_mask <-  left_join(umap_meta_df_bound, bal100_de_cells , by = c("cell_bc", "donor", "ptime_bal_cat", "StrainDon"))
umap_meta_df_mask <-  map2(umap_meta_df_bound, bal100_de_cells[names(umap_meta_df_bound)], ~left_join(.x, .y , by = c("cell_bc", "donor", "ptime_bal_cat", "StrainDon")))

```



```{r, fig.width=8, fig.height=4}
# solo_don <- c("MSC33", "MSC49", "MSC55", "MSC60", "MSC66")
# solo_don <- c("MSC33", "MSC49", "MSC60", "MSC66")
# solo_don <- c("MSC49", "MSC66")
# solo_don <- names(fm_bal_sce)
solo_don <- str_remove(names(fm_bal_sce), ".*\\.")

# solo_don <- c("MSC49", "MSC66")
# strn_comps = list(c("M33_SC2", "M33_SC7"), c("M49_SC1", "M49_SC5"), c("M55_SC4", "M55_SC6"), c("M60_SC3", "M60_SC7"), c("M66_SC5", "M66_SC8"))
# strn_comps = list(c("M33_SC2", "M33_SC7"), c("M49_SC1", "M49_SC5"), c("M60_SC3", "M60_SC7"), c("M66_SC5", "M66_SC8"))
strn_comps = unlist(map(fm_bal_sce, ~unique(.x@colData$StrainDon)))

strn_bx_tbl <- map(umap_meta_df, ~.x %>% 
  filter(donor %in% solo_don) %>%
  arrange(StrainDon) %>%
  group_by(StrainDon) %>% 
  # filter(n() >=20) %>% 
  # filter(n() >=100) %>% 
  ungroup() %>%
  add_count(donor, name = "don_n") %>%
  add_count(StrainDon, name = "strn_n") %>% 
  mutate(StrainProp = round(strn_n/don_n, 2)) 
)

strn_bx_props <- map(strn_bx_tbl, ~.x %>% select(StrainDon, StrainProp) %>% distinct())

map2(strn_bx_tbl, strn_bx_props, ~.x %>%
  # ggboxplot(., y = "slingPseudotime_1",x= "StrainDon", col="StrainDon", facet.by = 'donor', scales = "free_x", space = "free", nrow = 1) +
  ggboxplot(., y = "slingPseudotime_1",x= "StrainDon", col="StrainDon",  nrow = 1) + 
  stat_compare_means(comparisons = strn_comps, label.y = c(18, 18, 18,18,18), method = "wilcox.test", size = 4.5) +
  # facet_grid(. ~ donor, scales = "free_x", space = "free") +
  scale_color_manual(values = strain_don_cols) +
  geom_text(data= .y, aes(x = StrainDon, y = 1, label=StrainProp), inherit.aes = FALSE, size = 4.5, show.legend = F, fontface ='plain') +
  labs(x = "Strain", y = "Pseudotime") +
  theme(axis.text.x = element_text(angle = 20, hjust = 1),
        text = element_text(size = 14),
        legend.text = element_text(size = 14),
        legend.position = "bottom") +
  guides(color = guide_legend(title.position = "top", nrow =2, title = "Strain")) 
)
  
# geom_text(
#     data = proportions, 
#     aes(x = strain, y = min(data$slingPseudotime_1) - 0.5, 
#         label = paste0("Prop: ", round(proportion, 2))),
#     inherit.aes = FALSE,
#     color = "black"
#   )
```




```{r, fig.width=3, fig.height=2.8}
# custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)

all_cells_tbl[[1]] %>%
  ggplot(., aes(x = umapharmonyPCman_1, y = umapharmonyPCman_2, colour = stage))+
  geom_point(size = 1) +
  facet_wrap(StrainDon ~.,  nrow = 1) +
  theme_classic()+
  theme(#legend.position = "bottom", 
        legend.position = "none",
        axis.text = element_blank(),
        # strip.text =element_text(size=16,face="bold", colour = 'black'),
        strip.text =element_text(size=16,face="bold", colour = 'black'))+
  # guides(x = "axis_nested") +
  labs(x = "UMAP_1", y="UMAP_2") 
```

```{r, fig.width=2.5, fig.height=2}
# custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)

all_cells_tbl[[1]] %>%
  ggplot(., aes(x = umapharmonyPCman_1, y = umapharmonyPCman_2, colour = stage))+
  geom_point(size = 1) +
  facet_wrap(StrainDon ~.,  nrow = 1) +
  theme_classic()+
  scale_color_manual(values = sp_fld_colors) +
  theme(#legend.position = "bottom", 
        legend.position = "none",
        axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        # strip.text =element_text(size=16,face="bold", colour = 'black'),
        strip.text =element_text(size=16,face="bold", colour = 'black'))+
  # guides(x = "axis_nested") +
  labs(x = "UMAP_1", y="UMAP_2") 
```



```{r, fig.width=3, fig.height=2.8}
strn_sizes <- ifelse(strain_nms == "Other", 0.01, 0.5) %>% set_names(strain_nms)

```




```{r, fig.width=10, fig.height=2.8}
# background cells
umap_meta_df_mask_slim <- map(umap_meta_df_mask, ~.x %>% select(c("umapharmonyPCman_1", "umapharmonyPCman_2", "donor", "strain", "ptime_bal_cat", "bal_de")))
```

```{r, fig.width=10, fig.height=2.8}
# donors per stage
dons_bal_sset_nms
```


```{r, fig.width=16, fig.height=12}
# custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)

umap_meta_df_bal_cells_plot2 <- map2(dons_bal_sset_nms, umap_meta_df_mask_slim, ~{
  dons = .x
  dset = .y
  # print(stage)
  
  map(dons, ~{
  don = .x
  # don = str_remove(.x, ".*\\.")
  print(don)

  
  pt = dset %>% filter(bal_de == "yes" & donor == don) %>% pull(ptime_bal_cat) %>% unique()
  
  map(pt, ~{
    pts = .x
  sd = dset %>% filter(bal_de == "yes" & donor == don & ptime_bal_cat == pts) %>% pull(strain) %>% unique()
  
  print(don)
  print(pts)
  print(sd)
  
     map(sd, ~{
    
      dset %>%
        mutate(Strain = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == pts & strain == .x ~ strain, 
                                         T ~ "Other"), 
                               levels = strain_nms_lvls),
               strn_rank = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == pts & strain == .x ~ "Strain_rnk", 
                                            T ~ "Other"), 
                                  levels = c("Strain_rnk", "Other")),
               pt_st_set = paste0(don, "_", pts, "_", .x)) %>%
        arrange(desc(strn_rank))
  })
  })
  }) %>%
  unlist(recursive = F) %>%
  bind_rows() #%>%
  # mutate(pt_bg = paste0(donor, "_", pt_set ))
})

```


```{r, fig.width=16, fig.height=12}
# # custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)
# 
# umap_meta_df_bal_cells_plot2 <- map(names(fm_bal_sce), ~{
#   don = .x
#   # don = str_remove(.x, ".*\\.")
#   print(don)
# 
#   
#   pt = umap_meta_df_mask_slim %>% filter(bal_de == "yes" & donor == don) %>% pull(ptime_bal_cat) %>% unique()
#   
#   map(pt, ~{
#     pts = .x
#   sd = umap_meta_df_mask_slim %>% filter(bal_de == "yes" & donor == don & ptime_bal_cat == pts) %>% pull(strain) %>% unique()
#   
#   print(don)
#   print(pts)
#   print(sd)
#   
#      map(sd, ~{
#     
#       umap_meta_df_mask_slim %>%
#         mutate(Strain = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == pts & strain == .x ~ strain, 
#                                          T ~ "Other"), 
#                                levels = strain_nms_lvls),
#                strn_rank = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == pts & strain == .x ~ "Strain_rnk", 
#                                             T ~ "Other"), 
#                                   levels = c("Strain_rnk", "Other")),
#                pt_st_set = paste0(don, "_", pts, "_", .x)) %>%
#         arrange(desc(strn_rank))
#   })
#   })
#   }) %>%
#   unlist(recursive = F) %>%
#   bind_rows() #%>%
#   # mutate(pt_bg = paste0(donor, "_", pt_set ))


  
```


```{r, fig.width=16, fig.height=12}
map(umap_meta_df_bal_cells_plot2, ~.x %>%
      ggplot(., aes(x = umapharmonyPCman_1, y = umapharmonyPCman_2, colour = Strain, size = Strain))+
      geom_point() +
      scale_color_manual(values = strain_cols_n)+
      scale_size_manual(values = strn_sizes) +
      facet_wrap(pt_st_set ~ .) +
      theme_classic()+
      # scale_color_manual(values = sp_fld_colors) +
      theme(#legend.position = "bottom", 
        legend.position = "none",
        axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        # strip.text =element_text(size=16,face="bold", colour = 'black'),
        strip.text =element_text(size=10,face="bold", colour = 'black'))+
  # guides(x = "axis_nested") +
  labs(x = "UMAP_1", y="UMAP_2")
)
  
```

```{r, results="asis", fig.width=4, fig.height=3}
map(bal100_de_cells, ~.x %>% count(donor, ptime_bal_cat, StrainDon))
bal100_de_counts <- map(bal100_de_cells, ~.x %>% count(donor, ptime_bal_cat, StrainDon) %>% pull(n))

```



```{r, fig.width=10, fig.height=12}
# custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)

umap_meta_df_bal_cells_plot <- map2(dons_bal_sset_nms, umap_meta_df_mask_slim, ~{
  dons = .x
  dset = .y
  # print(stage)
  
  map(dons, ~{
  don = .x
  pt = dset %>% filter(bal_de == "yes" & donor == don) %>% pull(ptime_bal_cat) %>% unique()
  
  map(pt, ~{
    
    strain_num = dset %>% filter(bal_de == "yes" & donor == don & ptime_bal_cat == .x) %>%
      count(donor, ptime_bal_cat, strain) %>% select(strain, n) %>% mutate(nm = paste(strain, n, sep=":")) %>% pull(nm) %>% paste0(., collapse = "\n")
    
    dset %>%
        mutate(Strain = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == .x ~ strain, 
                                         T ~ "Other"), 
                               levels = strain_nms_lvls),
               strn_rank = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == .x  ~ "Strain_rnk", 
                                            T ~ "Other"), 
                                  levels = c("Strain_rnk", "Other")),
                pt_set = paste0(don, "_", .x)#,
                # "strain_n" = strain_num,
               ) %>%
        add_count(bal_de, donor, ptime_bal_cat, strain, name = "strain_n") %>%
        mutate(strain_n = ifelse(Strain == strain, paste0(strain, "_", strain_n), NA)) %>%
        arrange(desc(strn_rank))
  })
  }) %>%
  unlist(recursive = F) %>%
  bind_rows() #%>%
  # mutate(pt_bg = paste0(donor, "_", pt_set ))
})
```


```{r, fig.width=10, fig.height=12}
# # custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)
# 
# umap_meta_df_bal_cells_plot <- map(names(fm_bal_sce), ~{
#   don = .x
#   pt = umap_meta_df_mask_slim %>% filter(bal_de == "yes" & donor == don) %>% pull(ptime_bal_cat) %>% unique()
#   map(pt, ~{
#     
#     strain_num = umap_meta_df_mask_slim %>% filter(bal_de == "yes" & donor == don & ptime_bal_cat == .x) %>%
#       count(donor, ptime_bal_cat, strain) %>% select(strain, n) %>% mutate(nm = paste(strain, n, sep=":")) %>% pull(nm) %>% paste0(., collapse = "\n")
#     
#     umap_meta_df_mask_slim %>%
#         mutate(Strain = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == .x ~ strain, 
#                                          T ~ "Other"), 
#                                levels = strain_nms_lvls),
#                strn_rank = factor(case_when(bal_de == "yes" & donor == don & ptime_bal_cat == .x  ~ "Strain_rnk", 
#                                             T ~ "Other"), 
#                                   levels = c("Strain_rnk", "Other")),
#                 pt_set = paste0(don, "_", .x)#,
#                 # "strain_n" = strain_num,
#                ) %>%
#         add_count(bal_de, donor, ptime_bal_cat, strain, name = "strain_n") %>%
#         mutate(strain_n = ifelse(Strain == strain, paste0(strain, "_", strain_n), NA)) %>%
#         arrange(desc(strn_rank))
#   })
#   }) %>%
#   unlist(recursive = F) %>%
#   bind_rows() #%>%
#   # mutate(pt_bg = paste0(donor, "_", pt_set ))

```



```{r, fig.width=10, fig.height=10}
map(umap_meta_df_bal_cells_plot, ~.x %>%
      ggplot(., aes(x = umapharmonyPCman_1, y = umapharmonyPCman_2, colour = Strain, size = Strain))+
      geom_point() +
      scale_color_manual(values = strain_cols_n)+
      scale_size_manual(values = strn_sizes) +
      geom_text(data = .x %>% 
                  filter(bal_de == "yes" & Strain != "Other") %>% 
                  distinct(donor, pt_set,Strain, .keep_all = T) %>%
                  group_by(donor, pt_set) %>%
                  mutate(y_pos = rank(-as.numeric(Strain)) +3), 
                aes(x = 1, y = y_pos, label = strain_n), 
                size = 5#, 
                # color = "black"
                ) +
      facet_wrap(pt_set ~ .) +
      theme_classic()+
      # scale_color_manual(values = sp_fld_colors) +
      theme(#legend.position = "bottom", 
        legend.position = "none",
        axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        # strip.text =element_text(size=16,face="bold", colour = 'black'),
        strip.text =element_text(size=10,face="bold", colour = 'black'))+
  # guides(x = "axis_nested") +
  labs(x = "UMAP_1", y="UMAP_2")
)
```


```{r, fig.width=10, fig.height=12}
# save table fot umaps with downsampled strains highlighted
saveRDS(umap_meta_df_bal_cells_plot, paste0(save_dir, "umap_meta_df_bal_cells_plot.RDS"))
```


#### Put subseted objects into a list and calculate metrics need for plotting and tradeseq input
```{r}

str_cols = c("SC1" = "#dcbeff", "SC5" = "#008080")
lins = rep(1,length(fm_bal_sce))
# lins_m = rep(c(1,2),4)

fm_bal_sds <- map(fm_bal_sce, ~SlingshotDataSet(.x))
fm_bal_ptime <- pmap(list(fm_bal_sce, lins),  ~slingPseudotime(..1)[,..2] )
fm_bal_cw <- pmap(list(fm_bal_sce,  lins),  ~slingCurveWeights(..1)[,..2] )

```


```{r}
## create directory
map(names(fm_bal_sce), ~system(paste0("mkdir -p /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/", str_extract(.x, "MSC.*"),"/tradeseq_afm/")))

imap(fm_bal_sce, ~saveRDS(.x, paste0('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/', str_extract(.y, "MSC.*"),'/tradeseq_afm/', str_remove(.y, ".MSC.*"),'_afm_bal_sce.RDS')))

imap(fm_bal_ptime, ~saveRDS(.x, paste0('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/', str_extract(.y, "MSC.*"),'/tradeseq_afm/', str_remove(.y, ".MSC.*"),'_afm_bal_ptime.RDS')))

imap(fm_bal_cw, ~saveRDS(.x, paste0('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/', str_extract(.y, "MSC.*"),'/tradeseq_afm/', str_remove(.y, ".MSC.*"),'_afm_bal_cw.RDS')))

```


```{r}
## Conditions for DE in nextflow tradeseq
cond_tbl = data.frame( "don_dset" = str_extract(names(fm_bal_sce), "MSC.*"), "condtn" = ds_grp, knots = 4, gns_cut = 5)

system(paste0("mkdir -p /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/tradeseq_afm/"))
write_csv(cond_tbl, paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/tradeseq_afm/cond_tbl_afm.csv"))
```


```{r}
# ## Conditions for DE in nextflow tradeseq
# cond_tbl = map(list(
#   names(fm_bal_sce)[str_detect(names(fm_bal_sce), "w_pt4_m")],
#   names(fm_bal_sce)[str_detect(names(fm_bal_sce), "w_pt7_f")]
# ), ~data.frame( "don_dset" = str_extract(.x, "MSC.*"), "condtn" = ds_grp, knots = 4, gns_cut = 5))
# 
# write_csv(cond_tbl, paste0("data/processed/Pf/tradeseq/cond_tbl_afm_", str_remove(.y, "MSC.*"), ".csv"))
```

```{r}
## plot balanced symptomatic vs assymptomatic by pseudotime
# v3_afm_harmony_ss_res_fld_pt_bal@meta.data$sympto <- ifelse(v3_afm_harmony_ss_res_fld_pt_bal@meta.data$Symptomatic == "Yes", "Symptomatic", "Asymptomatic")
# 
# FeaturePlot(v3_afm_harmony_ss_res_fld_pt_bal, reduction = "umap.harmony_PC_man", features = "pseudotime", pt.size = 0.5, dims = c(1,2), split.by = "sympto ")

```

```{r}
dput(names(fm_bal_sce))
```


```{r, results="asis", eval=T}
gogo()
## Tradeseq script
# bsub < /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code_lnk/bash_scripts/submsn_tseq_condtn.sh

## Female and male
# --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/tradeseq_afm/*fm_bal"  \
# --cond_tbl "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/tradeseq_afm/cond_tbl_afm.csv"

## Female and male
# --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/tradeseq_afm/*fm_bal"  \
# --cond_tbl "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/tradeseq_afm/cond_tbl_afm.csv"

## Asexual, Female and male
# --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/tradeseq_afm/*afm_bal"  \
# --cond_tbl "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/tradeseq_afm/cond_tbl_afm.csv"


```


Read in tradeseq output from above run
```{r, message=F, warning=F, eval=T}
## Get file list
fm_bal_ts_files <- Sys.glob("data/processed/Pf/*/tradeseq_afm/[a|f|m]*_afm_bal_ts.RDS")
fm_bal_ts_files
```

```{r, message=F, warning=F, eval=T}
## read filtered barcodes
fm_bal_ts <- map(fm_bal_ts_files, ~.x %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|tradeseq_afm/|_afm_.*') %>% 
                    # str_remove_all(.,'data/processed/Pf/|tradeseq_afm/|\\._afm_.*') %>% 
                    gsub("(.*)/(.*)", "\\2/\\1", .) %>% 
                    str_replace(., "/", ".")))) %>% 
  unlist(., recursive = F) %>% 
  .[names(fm_bal_sce)] %>%
  map(readRDS)
```

Read in tradeseq condition DE files from local
```{r, message=F, warning=F, eval=T}
## Get file list
fm_bal_de_files <- Sys.glob("data/processed/Pf/*/tradeseq_afm/[a|f|m]*_afm_bal_de.RDS")
fm_bal_de_files
```

```{r, message=F, warning=F, eval=T}
## read filtered barcodes
fm_bal_de <- map(fm_bal_de_files, ~.x %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|tradeseq_afm/|_afm_.*') %>% 
                    gsub("(.*)/(.*)", "\\2/\\1", .) %>% 
                    str_replace(., "/", ".")))) %>% 
  unlist(., recursive = F) %>% 
  .[names(fm_bal_sce)] %>%
  map(readRDS)
```


```{r}
# fm_bal_ts <- list.files('data/processed/Pf/pseudotime_tradeseq_op', pattern = "tseq_obj.*_M.*DS$", full.names = T) %>% 
#   str_sort(., numeric = T)  %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/.*obj_|\\.RDS') %>% str_replace_all(., c( 'StrainDon'='StrainsDE')) )) %>%
#   map(., ~readRDS(.x))
```

```{r}
# de_order <- c(paste0("FieldVlab_", m_stgs[1:8]), paste0("lf_StrainsDE_", m_stgs[9:28]))
# # fm_bal_ts <- fm_bal_ts[c(de_order, paste0(de_order[1:10], '_unbal'))]
# fm_bal_ts <- fm_bal_ts[c(de_order)]

```

Read in tradeseq condition DE files from local
```{r}
# fm_bal_de <- list.files('data/processed/Pf/pseudotime_tradeseq_op', pattern = "tseq_cond_obj.*_M.*DS$", full.names = T) %>% str_sort(., numeric = T)  %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/.*obj_|\\.RDS') %>% str_replace_all(., c( 'StrainDon'='StrainsDE')) )) %>%
#   map(., ~readRDS(.x))
```

```{r}
# fm_bal_de <- fm_bal_de[c(de_order)]
fm_bal_de <- fm_bal_de[!is.na(names(fm_bal_de))]
fm_bal_ts <- fm_bal_ts[names(fm_bal_de)]

```

Read in tradeseq files from local
```{r}
# msc_y21_ref_afm_bal_ts_de <- readRDS('data/processed/Pf/pseudotime_tradeseq/msc_y21_ref_afm_bal_ts_de.RDS') %>% 
#   set_names(paste0(names(.), '_', c(rep('FieldVlab', 5), rep('StrainsDE', 5))))
```

#### Visualize cells tradeseq trajectories {.tabset}

#### Read in and format tradeseq condition test results (lab VS field, field VS field) (Performed in farm)  


```{r}
##NOTE!!! replace the conditional test file with the newly generated one which applied 7 clusters
##Ajusting the global and pairwise pvalues - old p values discarded
fm_bal_de_adj <- map(fm_bal_de, ~.x %>%
                             mutate(across(starts_with('pvalue'), ~p.adjust(., "fdr"))) %>%
                             arrange(desc(waldStat)) %>%
                             rownames_to_column('gene_id'))

```



```{r}
##save condition results for GSEA
saveRDS(fm_bal_de_adj, paste0(save_dir, "fm_bal_de_adj.RDS"))
```

```{r}
# fm_bal_ts <- fm_bal_ts
# cond
```


```{r}
### based on mean smoother
### All these genes are in the top 22 with waldstat > 20
cond_gns_005_ub_de_afm <- pmap(list(fm_bal_de_adj, fm_bal_ts),~{
  
  b = levels(..2$tradeSeq$conditions)
  
  print(b)
  
  if (dim(..1)[2] > 4) {
     ..1 %>% rename_with(., ~str_replace_all(., c('s9' = b[9],
                                               's8' = b[8],
                                               's7' = b[7],
                                               's6' = b[6],
                                               's5' = b[5],
                                               's4' = b[4],
                                               's3' = b[3],
                                               's2' = b[2],
                                               's1' = b[1],
                                               # ' ' = '',
                                               'vm'='_vs_m',
                                               'vS'='_vs_S',
                                               'vL'='_vs_L',
                                               'vN'='_vs_N',
                                               'v7'='_vs_7',
                                               #'4 \\(O'='4\\(O',
                                               'cond'=''))
  )
  } else {
  
    
    ..1 %>% mutate(across(.cols = -c("gene_id"), .names = '{.col}_2')) %>% rename_with(., ~str_replace_all(., c('2' = paste0(b[1], "_vs_", b[2]),
                                               #'4 \\(O'='4\\(O',
                                               'cond'='')))
  
  }
  
})


```

```{r}

cond_gns_de_afm <- c(cond_gns_005_ub_de_afm) %>%
  map(., ~left_join(., Pf3D7_genes_plasmoDB %>% dplyr::select(gene_id, Gene, Ka_Ks, `Transcript Product Description`, TM_domains, `Computed GO Functions`), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)))

```


```{r}
##save condition results for GSEA
saveRDS(cond_gns_de_afm, paste0(save_dir, 'cond_gns_de_afm.RDS'))
```


```{r}
# msc_y21_ref_afm_bal_ts_de <- msc_y21_ref_afm_bal_ts_de[c(1:15)]
# cond_gns_de_afm <- cond_gns_de_afm[c(1:15)]
```

#### Plot smoothers function


```{r, results='asis'}
# cat('## Visualize field parasites {.tabset}   \n')
tde_smooth_plt <- function(tde_obj_lst = fm_bal_ts[1],
                           cond_obj_lst = cond_gns_de_afm[1],
                           pw_comp= 'waldStat_M33_SC5vM33_SC7', 
                           n_gns = c(1:5),
                           lege_pos = "bottom",
                           wstat = 20,
                           pval_cond = "pvalue",
                           pval = 0.05
)
{
  
  list(tde_obj_lst, cond_obj_lst) %>% 
    pmap(.,possibly(~{
      
      gns = ..2 %>% filter(!!rlang::sym(pw_comp) > wstat & !!rlang::sym(pval_cond)  < pval) %>% filter(!is.na(gene_id)) %>%  arrange(desc(!!rlang::sym(pw_comp)))  %>% pull(gene_id) %>% .[n_gns]
      
      gn_names = ..2 %>% filter(!!rlang::sym(pw_comp) > wstat & !!rlang::sym(pval_cond)  < pval) %>% filter(!is.na(gene_id))  %>% arrange(desc(!!rlang::sym(pw_comp))) %>% pull(Gene) %>% .[n_gns]
      print(gn_names)
      
      tde = ..1
      
      list(gns, gn_names) %>% 
        pmap(., possibly(~{
          plotSmoothers(tde, 
                        assays(tde)$counts,
                        gene = ..1,
                        alpha = 1, 
                        border = TRUE) +
            labs(title = ..2) +
            theme(text = element_text(size = 14),
                  legend.position = lege_pos) +
            guides(color = guide_legend(override.aes = list(size = 2), title.position = "top", ncol = 1))
        }, 'missing'))
    }, 'missing'))
}

```


```{r, results="asis", eval=T}
cond_gns_de_afm[[1]] %>% arrange(pvalue)
```

#### smoothers for select genes that are differentially expressed between strains {.tabset}

```{r, results='asis', warning=F, message=F, fig.width=5.5, fig.height=4}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC33")],
               cond_obj_lst = cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC33")],
               pw_comp= 'waldStat_M33_SC5vM33_SC7',
               n_gns = c(1:10),
               lege_pos = "right"
)

```

```{r, results='asis', warning=F, message=F, fig.width=3, fig.height=4}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC60C")],
               cond_obj_lst = cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC60C")],
               pw_comp= 'waldStat_M60_SC3_vs_M60_SC7',
               n_gns = c(1:20)
)

```


```{r, results='asis', warning=F, message=F, fig.width=3, fig.height=4}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC49")],
               cond_obj_lst = cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC49")],
               pw_comp= 'waldStat_M49_SC1_vs_M49_SC5',
               n_gns = c(1:20)
)

```


```{r, results='asis', warning=F, message=F, fig.width=3, fig.height=4}
## var genes
tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC48")],
               cond_obj_lst = cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC48")][[1]] %>% filter(str_detect(Gene, "VAR")) %>% list(),
               pw_comp= 'waldStat_M48_SC1_vs_M48_SC5',
               n_gns = c(1:20),
               wstat = 10,
               pval = 0.05
)

```

```{r, fig.width=7.5, fig.height=2}
## Direction of upregulation
# strain_up <- c('PF3D7-0400400' ='SC1', "PF3D7-0412400" ='SC1', 'PF3D7-0421300' ='SC5', "PF3D7-0412700" ='SC5', 'PF3D7-0712900' ='SC5', 'PF3D7-1041300' ='SC5', "PF3D7-0632800" ='SC5', 'PF3D7-0937800' ='SC5', 'PF3D7-1240600' ='SC5', 'PF3D7-1240400' ='SC5', 'PF3D7-0808700' ='SC5', 'PF3D7-0900100' ='N')
strain_up <- c('PF3D7-0400400' ='SC1', "PF3D7-0412400" ='SC1', 'PF3D7-0421300' ='SC5', "PF3D7-0412700" ='SC5', 'PF3D7-0712900' ='SC5', 'PF3D7-1041300' ='SC5', "PF3D7-0632800" ='SC5', 'PF3D7-0937800' ='SC5', 'PF3D7-1240600' ='SC5', 'PF3D7-1240400' ='SC5', 'PF3D7-0808700' ='SC5', 'PF3D7-0900100' ='SC1')
```

 PF3D7_0412700 Expressed at high levels all the time in a clone https://www-ncbi-nlm-nih-gov.ezp.lib.cam.ac.uk/pmc/articles/PMC10151525/
 PF3D7_0412400 Expressed at high levels all the time in a clone
PF3D7_0937800 in malaria naive individuals https://pubmed.ncbi.nlm.nih.gov/31295334/
PF3D7_0900100 in malaria naive individuals https://pubmed.ncbi.nlm.nih.gov/31295334/
```{r, fig.width=7.5, fig.height=2}
## Ups group

var_grps <- c('PF3D7-0400400' ='A', "PF3D7-0412400" ='B/C', 'PF3D7-0421300' ='C', "PF3D7-0412700" ='C', 'PF3D7-0712900' ='B', 'PF3D7-1041300' ='B', "PF3D7-0632800" ='B', 'PF3D7-0937800' ='B', 'PF3D7-1240600' ='C', 'PF3D7-1240400' ='B/C', 'PF3D7-0808700' ='B/C', 'PF3D7-0900100' ='B')

var_gn_info <- data.frame(strain_up) %>% 
  rownames_to_column("gene_id") %>% 
  left_join(., data.frame(var_grps) %>% rownames_to_column("gene_id"))
```

```{r, results='asis', warning=F, message=F, fig.width=3, fig.height=4}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC49")],
               cond_obj_lst = cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC49")],
               pw_comp= 'waldStat_M49_SC1_vs_M49_SC5',
               n_gns = c(1:20)
)

```

```{r, results='asis', warning=F, message=F, fig.width=5.5, fig.height=4}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC66")],
               cond_obj_lst = cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC66")],
               pw_comp= 'waldStat_M66_SC7vM66_SC8',
               n_gns = c(1:10),
               lege_pos = "right"
)

```

Heatmap plotting function
```{r}
### Annotation groups for the header ie column
tseq_de_hmap <- function(tseq = fm_bal_ts[[1]], de_tbl , pnts = 30, rnames = F, comp_grps, Pv = 0.05,  wstat =30){
  
  # gns = de_tbl %>% filter(if_any(matches("pvalue.*"), ~. < 0.05)) %>% pull(gene_id)
  gns = de_tbl %>% filter(pvalue < Pv & waldStat >= wstat) %>% pull(gene_id)

  # print(str(gns))
  ##Edited from the original version to allow the labelling of rows with gene names
  msc_yhatSmooth_w_annot <- predictSmooth(tseq, gene = gns, nPoints = pnts, tidy = FALSE) %>% 
        t() %>%scale() %>% t() %>% as.data.frame()%>% 
        rownames_to_column('gene_id') %>% 
        left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
        dplyr::select(-c(gene_id)) %>% 
        column_to_rownames('Gene')
  
  msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
    dplyr::select(-c(Ka_Ks)) 
  
  # print(str(msc_yhatSmooth))
  msc_col_annot <- predictSmooth(tseq, gene = gns, nPoints = pnts, tidy = T)  %>% 
        mutate(point = paste0("point", rep(rep(1:pnts,length(unique(tseq@colData$tradeSeq$conditions))),length(gns))), 
               lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
        rownames_to_column('No')  %>% 
        dplyr::select(lin_id,time,condition)%>% 
        distinct()%>% 
        column_to_rownames('lin_id') %>%
        rename('Pseudotime' = time, 'Strain' = condition)
  # print(str(msc_col_annot))

  hmap<- ComplexHeatmap::Heatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix(),
                         cluster_columns = FALSE,
                         show_row_names = {{rnames}}, 
                         show_column_names = F, 
                         # main = comp_grps,
                         show_heatmap_legend = T,
                         # silent = TRUE,
                         # left_annotation = rowAnnotation(df= msc_yhatSmooth_w_annot %>% select(Ka_Ks)),
                         top_annotation = HeatmapAnnotation(df= msc_col_annot %>% dplyr::select( Strain,Pseudotime) ,
                                                            # cols = list(Strain = msc_col_annot %>% mutate(cols = ifelse(Strain == 'M49_SC1') ~ '#e7a06b', '#9cb4cf') %>%  pull(cols) ), 
                                                            annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                           direction = "horizontal", 
                                                                                           title_gp = gpar(fontsize = 14), labels_gp = gpar(fontsize = 14)),
                                                            show_legend = c(TRUE, FALSE) ##https://support.bioconductor.org/p/82447/
                                                            ),
                         # fontsize = 14,
                         # heatmap_legend_param = list(ncol = 1, nrow = 4, legend_direction = "vertical", by_row = T),
            # heatmap_legend_param = list(direction = "horizontal", legend_gp = gpar(fontsize = lgsize)),
            heatmap_legend_param = list(title = "Expression",
                                        title_gp = gpar(fontsize = 14),
                                        labels_gp = gpar(fontsize = 14),
                                        direction = "horizontal",
                                        labels_rot = 90#,
                                        # grid_width = unit(6, "mm"),
                                        # legend_width = unit(lg_width, "cm")
                                        ),
            row_names_gp = grid::gpar(fontsize = 14)
                         # 
      ) %>%
  ##Merging legends into 1 column. Could be set to F
  draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")
    
  return(hmap)
  # print(hmap)
  # print(pheatmap::pheatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix()))
      # cat('\n')
      # cat("\n")
    
}
```

Heatmap plotting function
```{r}
### Annotation groups for the header ie column
tseq_de_hmap <- function(tseq = fm_bal_ts[[1]], de_tbl , pnts = 30, rnames = F, comp_grps, Pv = 0.05,  wstat =30){
  
  # gns = de_tbl %>% filter(if_any(matches("pvalue.*"), ~. < 0.05)) %>% pull(gene_id)
  gns = de_tbl %>% filter(pvalue < Pv & waldStat >= wstat) %>% pull(gene_id)

  # print(str(gns))
  ##Edited from the original version to allow the labelling of rows with gene names
  msc_yhatSmooth_w_annot <- predictSmooth(tseq, gene = gns, nPoints = pnts, tidy = FALSE) %>% 
        t() %>%scale() %>% t() %>% as.data.frame()%>% 
        rownames_to_column('gene_id') %>% 
        left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
        dplyr::select(-c(gene_id)) %>% 
        column_to_rownames('Gene')
  
  msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
    dplyr::select(-c(Ka_Ks)) 
  
  # print(str(msc_yhatSmooth))
  msc_col_annot <- predictSmooth(tseq, gene = gns, nPoints = pnts, tidy = T)  %>% 
        mutate(point = paste0("point", rep(rep(1:pnts,length(unique(tseq@colData$tradeSeq$conditions))),length(gns))), 
               lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
        rownames_to_column('No')  %>% 
        dplyr::select(lin_id,time,condition)%>% 
        distinct()%>% 
        column_to_rownames('lin_id') %>%
        rename('Pseudotime' = time, 'Condition' = condition)
  # print(str(msc_col_annot))

  hmap<- ComplexHeatmap::pheatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix(),
                         cluster_cols = FALSE,
                         show_rownames = {{rnames}}, 
                         show_colnames = F, 
                         main = comp_grps,
                         legend = T,
                         silent = TRUE,
                         annotation_row = msc_yhatSmooth_w_annot %>% select(Ka_Ks), 
                         annotation_col = (msc_col_annot %>% dplyr::select(Pseudotime, Condition)),
                         fontsize = 14
                         # 
      )
  # str(hmap)
  return(hmap)
  # print(hmap)
  # print(pheatmap::pheatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix()))
      # cat('\n')
      # cat("\n")
    
}
```


#### heatmaps for DE between field and lab in the asexual, female and males {.tabset}

```{r, results='asis', warning=F, message=F, fig.width=8, fig.height=10}

hmap_plt<- list(fm_bal_ts[1], 
                # map(cond_gns_de_afm, ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)), 
                cond_gns_de_afm[1],
                rep(30, length(cond_gns_de_afm[1])), 
                rep(T, length(cond_gns_de_afm[1])), 
                names(fm_bal_ts[1])) %>% 
  pmap(., #possibly(
    ~{
  tseq_de_hmap(tseq = ..1, de_tbl= ..2, pnts = ..3, rnames = ..4, comp_grps = ..5)
}#, 'missing genes')
)
  
  # tseq_de_hmap(tseq = fm_bal_ts[[1]], gns= map(cond_gns_de_afm[1], ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)) %>% flatten() %>% as.character(), pnts = 30, rnames = F, title = "Strain DE")
  #                      
hmap_plt

```

```{r, results='asis', warning=F, message=F}

hmap_plt<- list(fm_bal_ts, 
                # map(cond_gns_de_afm, ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)), 
                cond_gns_de_afm,
                rep(30, length(cond_gns_de_afm)), 
                rep(T, length(cond_gns_de_afm)), 
                names(fm_bal_ts)) %>% 
  pmap(., possibly(
    ~{
  tseq_de_hmap(tseq = ..1, de_tbl= ..2, pnts = ..3, rnames = ..4, comp_grps = ..5)
}, 'missing genes')
)
  
  # tseq_de_hmap(tseq = fm_bal_ts[[1]], gns= map(cond_gns_de_afm[1], ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)) %>% flatten() %>% as.character(), pnts = 30, rnames = F, title = "Strain DE")
  #                      

```


```{r, fig.height=8, fig.width=6}

hmap_plt

```


```{r, fig.height=14, fig.width=7}
str_cols = c("#dcbeff", "#008080")
hmap_plt[2]

```


Heatmap plotting function
```{r}
gene_annot <- var_gn_info %>% mutate(across(strain_up, ~paste0("M49_", .))) %>% rename("Up" = strain_up, VAR = "var_grps") %>% select(-c(Up))

var_cols <- c("A" = "#8343b7" , "B/C" = "#a8253a", "B" =  "#b0921d", "C" = "#fd3a17", "NV" = "lightgrey")
```

```{r}
### Annotation groups for the header ie column
# gns = cond_gns_de_afm[["MSC49"]] %>% filter(pvalue < 0.05) %>% arrange(desc(waldStat)) %>%   mutate(wstat_rank = rank(-waldStat)) %>% filter(., wstat_rank < 20 | str_detect(Gene, "VAR")) %>% pull(gene_id)

gns = cond_gns_de_afm[["f.MSC48C"]] %>% 
  filter(pvalue < 0.001) %>% 
  filter(waldStat >= 20) %>% 
  arrange(desc(waldStat)) %>%
  mutate(wstat_rank = rank(-waldStat)) %>%
  # filter(., wstat_rank < 20 ) %>%
  pull(gene_id)

# print(str(gns))
##Edited from the original version to allow the labelling of rows with gene names
msc_yhatSmooth_w_annot <- predictSmooth(fm_bal_ts[["f.MSC48C"]], gene = gns, nPoints = 30, tidy = FALSE) %>% 
  t() %>%scale() %>% t() %>% as.data.frame()%>% 
  rownames_to_column('gene_id') %>% 
  left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
  left_join(., gene_annot, by = 'gene_id') %>% mutate(across(c("VAR"), ~case_when(is.na(.) ~ "NV", T ~ .))) %>%
  mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
  dplyr::select(-c(gene_id)) %>% 
  column_to_rownames('Gene')

msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
  dplyr::select(-c(Ka_Ks, VAR)) 

# print(str(msc_yhatSmooth))
msc_col_annot <- predictSmooth(fm_bal_ts[["f.MSC48C"]], gene = gns, nPoints = 30, tidy = T)  %>% 
  mutate(point = paste0("point", rep(rep(1:30,length(unique(fm_bal_ts[["f.MSC48C"]]@colData$tradeSeq$conditions))),length(gns))), 
         lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
  rownames_to_column('No')  %>% 
  dplyr::select(lin_id,time,condition)%>% 
  distinct()%>% 
  column_to_rownames('lin_id') %>%
  rename('Pseudotime' = time, 'Strain' = condition)
# print(str(msc_col_annot))

hmap<- ComplexHeatmap::Heatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix(),
                               cluster_columns = FALSE,
                               show_row_names = T, 
                               show_column_names = F, 
                               # main = comp_grps,
                               show_heatmap_legend = T,
                               # silent = TRUE,
                               left_annotation = rowAnnotation(df= msc_yhatSmooth_w_annot %>% select(VAR),
                                                               col = list(VAR = var_cols ), 
                                                               annotation_name_gp = gpar(fontsize = 14, fontface = "bold"),
                                                               annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                               direction = "horizontal", 
                                                                                               title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                                               labels_gp = gpar(fontsize = 12),
                                                                                               ncol = 3),
                                                                  show_legend = c(FALSE)),
                               top_annotation = HeatmapAnnotation(df= msc_col_annot %>% dplyr::select( Strain,Pseudotime) ,
                                                                  col = list(Strain = strain_don_cols[unique(msc_col_annot[, "Strain"])],
                                                                             Pseudotime = colorRamp2(c(min(msc_col_annot[,"Pseudotime"]), max(msc_col_annot[,"Pseudotime"])), c("white", "#281e5d"))), 
                                                                  annotation_name_gp = gpar(fontsize = 14, fontface = "bold"), 
                                                                  annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                                 direction = "horizontal", 
                                                                                                 title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                                                 labels_gp = gpar(fontsize = 12)),
                                                                  show_legend = c(T, FALSE) ##https://support.bioconductor.org/p/82447/
                               ),
                               # fontsize = 14,
                               # heatmap_legend_param = list(ncol = 1, nrow = 4, legend_direction = "vertical", by_row = T),
                               # heatmap_legend_param = list(direction = "horizontal", legend_gp = gpar(fontsize = lgsize)),
                               heatmap_legend_param = list(title = "Expression",
                                                           title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                           labels_gp = gpar(fontsize = 12),
                                                           direction = "horizontal"#,
                                                           # direction = "vertical",
                                                           # labels_rot = 90#,
                                                           # grid_width = unit(6, "mm"),
                                                           # legend_width = unit(lg_width, "cm")
                               ),
                               row_names_gp = grid::gpar(fontsize = 14)
                               # 
) %>%
  ##Merging legends into 1 column. Could be set to F
  draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")



```

```{r, fig.height=8.5, fig.width=4.3}
hmap
# c("#b7e577", "#3ce07e")
```


```{r}
### Annotation groups for the header ie column
# gns = cond_gns_de_afm[["f.MSC48C"]] %>% filter(pvalue < 0.05) %>% arrange(desc(waldStat)) %>%   mutate(wstat_rank = rank(-waldStat)) %>% filter(., wstat_rank < 20 | str_detect(Gene, "VAR")) %>% pull(gene_id)

gns = cond_gns_de_afm[["f.MSC48C"]] %>% 
  filter(pvalue < 0.001) %>% 
  filter(waldStat >= 20) %>% 
  arrange(desc(waldStat)) %>%
  mutate(wstat_rank = rank(-waldStat)) %>%
  # filter(., wstat_rank < 20 ) %>%
  pull(gene_id)

# print(str(gns))
##Edited from the original version to allow the labelling of rows with gene names
msc_yhatSmooth_w_annot <- predictSmooth(fm_bal_ts[["f.MSC48C"]], gene = gns, nPoints = 30, tidy = FALSE) %>% 
  t() %>%scale() %>% t() %>% as.data.frame()%>% 
  rownames_to_column('gene_id') %>% 
  left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
  left_join(., gene_annot, by = 'gene_id') %>% mutate(across(c("VAR"), ~case_when(is.na(.) ~ "NV", T ~ .))) %>%
  mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
  dplyr::select(-c(gene_id)) %>% 
  column_to_rownames('Gene')

msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
  dplyr::select(-c(Ka_Ks, VAR)) 

# print(str(msc_yhatSmooth))
msc_col_annot <- predictSmooth(fm_bal_ts[["f.MSC48C"]], gene = gns, nPoints = 30, tidy = T)  %>% 
  mutate(point = paste0("point", rep(rep(1:30,length(unique(fm_bal_ts[["f.MSC48C"]]@colData$tradeSeq$conditions))),length(gns))), 
         lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
  rownames_to_column('No')  %>% 
  dplyr::select(lin_id,time,condition)%>% 
  distinct()%>% 
  column_to_rownames('lin_id') %>%
  rename('Pseudotime' = time, 'Strain' = condition)
# print(str(msc_col_annot))

hmap2<- ComplexHeatmap::Heatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix(),
                               cluster_columns = FALSE,
                               show_row_names = T, 
                               show_column_names = F, 
                               # main = comp_grps,
                               show_heatmap_legend = T,
                               # silent = TRUE,
                               # left_annotation = rowAnnotation(df= msc_yhatSmooth_w_annot %>% select(VAR),
                               #                                 col = list(VAR = var_cols ), 
                               #                                 annotation_name_gp = gpar(fontsize = 14, fontface = "bold"),
                               #                                 annotation_legend_param = list(Strain = list(title = "Strain"), 
                               #                                                                 direction = "horizontal", 
                               #                                                                 title_gp = gpar(fontsize = 12, fontface = "bold"), 
                               #                                                                 labels_gp = gpar(fontsize = 12),
                               #                                                                 ncol = 3),
                               #                                    show_legend = c(FALSE)),
                               top_annotation = HeatmapAnnotation(df= msc_col_annot %>% dplyr::select( Strain,Pseudotime) ,
                                                                  col = list(Strain = strain_don_cols[unique(msc_col_annot[, "Strain"])],
                                                                             Pseudotime = colorRamp2(c(min(msc_col_annot[,"Pseudotime"]), max(msc_col_annot[,"Pseudotime"])), c("white", "#281e5d"))), 
                                                                  annotation_name_gp = gpar(fontsize = 14, fontface = "bold"), 
                                                                  annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                                 direction = "horizontal", 
                                                                                                 title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                                                 labels_gp = gpar(fontsize = 12)),
                                                                  show_legend = c(T, FALSE) ##https://support.bioconductor.org/p/82447/
                               ),
                               # fontsize = 14,
                               # heatmap_legend_param = list(ncol = 1, nrow = 4, legend_direction = "vertical", by_row = T),
                               # heatmap_legend_param = list(direction = "horizontal", legend_gp = gpar(fontsize = lgsize)),
                               heatmap_legend_param = list(title = "Expression",
                                                           title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                           labels_gp = gpar(fontsize = 12),
                                                           direction = "horizontal"#,
                                                           # direction = "vertical",
                                                           # labels_rot = 90#,
                                                           # grid_width = unit(6, "mm"),
                                                           # legend_width = unit(lg_width, "cm")
                               ),
                               row_names_gp = grid::gpar(fontsize = 14)
                               # 
) %>%
  ##Merging legends into 1 column. Could be set to F
  draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")



```

```{r, fig.height=8, fig.width=4.3}
hmap2
# c("#b7e577", "#3ce07e")
```

```{r}
### Annotation groups for the header ie column
# gns = cond_gns_de_afm[["f.MSC48C"]] %>% filter(pvalue < 0.05) %>% arrange(desc(waldStat)) %>%   mutate(wstat_rank = rank(-waldStat)) %>% filter(., wstat_rank < 20 | str_detect(Gene, "VAR")) %>% pull(gene_id)

gns = cond_gns_de_afm[["f.MSC66C"]] %>% 
  filter(pvalue < 0.001) %>% 
  filter(waldStat >= 20) %>% 
  arrange(desc(waldStat)) %>%
  mutate(wstat_rank = rank(-waldStat)) %>%
  # filter(., wstat_rank < 20 ) %>%
  pull(gene_id)

# print(str(gns))
##Edited from the original version to allow the labelling of rows with gene names
msc_yhatSmooth_w_annot <- predictSmooth(fm_bal_ts[["f.MSC66C"]], gene = gns, nPoints = 30, tidy = FALSE) %>% 
  t() %>%scale() %>% t() %>% as.data.frame()%>% 
  rownames_to_column('gene_id') %>% 
  left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
  left_join(., gene_annot, by = 'gene_id') %>% mutate(across(c("VAR"), ~case_when(is.na(.) ~ "NV", T ~ .))) %>%
  mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
  dplyr::select(-c(gene_id)) %>% 
  column_to_rownames('Gene')

msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
  dplyr::select(-c(Ka_Ks, VAR)) 

# print(str(msc_yhatSmooth))
msc_col_annot <- predictSmooth(fm_bal_ts[["f.MSC66C"]], gene = gns, nPoints = 30, tidy = T)  %>% 
  mutate(point = paste0("point", rep(rep(1:30,length(unique(fm_bal_ts[["f.MSC66C"]]@colData$tradeSeq$conditions))),length(gns))), 
         lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
  rownames_to_column('No')  %>% 
  dplyr::select(lin_id,time,condition)%>% 
  distinct()%>% 
  column_to_rownames('lin_id') %>%
  rename('Pseudotime' = time, 'Strain' = condition)
# print(str(msc_col_annot))

hmap66<- ComplexHeatmap::Heatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix(),
                               cluster_columns = FALSE,
                               show_row_names = T, 
                               show_column_names = F, 
                               # main = comp_grps,
                               show_heatmap_legend = T,
                               # silent = TRUE,
                               # left_annotation = rowAnnotation(df= msc_yhatSmooth_w_annot %>% select(VAR),
                               #                                 col = list(VAR = var_cols ), 
                               #                                 annotation_name_gp = gpar(fontsize = 14, fontface = "bold"),
                               #                                 annotation_legend_param = list(Strain = list(title = "Strain"), 
                               #                                                                 direction = "horizontal", 
                               #                                                                 title_gp = gpar(fontsize = 12, fontface = "bold"), 
                               #                                                                 labels_gp = gpar(fontsize = 12),
                               #                                                                 ncol = 3),
                               #                                    show_legend = c(FALSE)),
                               top_annotation = HeatmapAnnotation(df= msc_col_annot %>% dplyr::select( Strain,Pseudotime) ,
                                                                  col = list(Strain = strain_don_cols[unique(msc_col_annot[, "Strain"])],
                                                                             Pseudotime = colorRamp2(c(min(msc_col_annot[,"Pseudotime"]), max(msc_col_annot[,"Pseudotime"])), c("white", "#281e5d"))), 
                                                                  annotation_name_gp = gpar(fontsize = 14, fontface = "bold"), 
                                                                  annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                                 direction = "horizontal", 
                                                                                                 title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                                                 labels_gp = gpar(fontsize = 12)),
                                                                  show_legend = c(T, FALSE) ##https://support.bioconductor.org/p/82447/
                               ),
                               # fontsize = 14,
                               # heatmap_legend_param = list(ncol = 1, nrow = 4, legend_direction = "vertical", by_row = T),
                               # heatmap_legend_param = list(direction = "horizontal", legend_gp = gpar(fontsize = lgsize)),
                               heatmap_legend_param = list(title = "Expression",
                                                           title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                           labels_gp = gpar(fontsize = 12),
                                                           direction = "horizontal"#,
                                                           # direction = "vertical",
                                                           # labels_rot = 90#,
                                                           # grid_width = unit(6, "mm"),
                                                           # legend_width = unit(lg_width, "cm")
                               ),
                               row_names_gp = grid::gpar(fontsize = 14)
                               # 
) %>%
  ##Merging legends into 1 column. Could be set to F
  draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")



```

```{r, fig.height=6, fig.width=4.3}
hmap66
# c("#b7e577", "#3ce07e")
```

```{r}
### Annotation groups for the header ie column
# gns = cond_gns_de_afm[["f.MSC48C"]] %>% filter(pvalue < 0.05) %>% arrange(desc(waldStat)) %>%   mutate(wstat_rank = rank(-waldStat)) %>% filter(., wstat_rank < 20 | str_detect(Gene, "VAR")) %>% pull(gene_id)
gns = cond_gns_de_afm[["f.MSC48C"]] %>% filter(pvalue < 0.05) %>% arrange(desc(waldStat)) %>%
  mutate(wstat_rank = rank(-waldStat)) %>%
  # filter(., wstat_rank < 20 ) %>%
  pull(gene_id)

# print(str(gns))
##Edited from the original version to allow the labelling of rows with gene names
msc_yhatSmooth_w_annot <- predictSmooth(fm_bal_ts[["f.MSC48C"]], gene = gns, nPoints = 30, tidy = FALSE) %>% 
  t() %>%scale() %>% t() %>% as.data.frame()%>% 
  rownames_to_column('gene_id') %>% 
  left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
  left_join(., gene_annot, by = 'gene_id') %>% mutate(across(c("VAR"), ~case_when(is.na(.) ~ "NV", T ~ .))) %>%
  mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
  mutate(across(Gene, ~str_remove(., "VAR_|RPS19_|HSP90_"))) %>%
  dplyr::select(-c(gene_id)) %>% 
  column_to_rownames('Gene')

msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
  dplyr::select(-c(Ka_Ks, VAR)) 

# print(str(msc_yhatSmooth))
msc_col_annot <- predictSmooth(fm_bal_ts[["f.MSC48C"]], gene = gns, nPoints = 30, tidy = T)  %>% 
  mutate(point = paste0("point", rep(rep(1:30,length(unique(fm_bal_ts[["f.MSC48C"]]@colData$tradeSeq$conditions))),length(gns))), 
         lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
  rownames_to_column('No')  %>% 
  dplyr::select(lin_id,time,condition)%>% 
  distinct()%>% 
  column_to_rownames('lin_id') %>%
  rename('Pseudotime' = time, 'Strain' = condition)
# print(str(msc_col_annot))

hmap_t<- ComplexHeatmap::Heatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix() %>% t(),
                               cluster_rows = FALSE,
                               show_row_names = F, 
                               # main = comp_grps,
                               show_heatmap_legend = T,
                               # silent = TRUE,
                               bottom_annotation = HeatmapAnnotation(df= msc_yhatSmooth_w_annot %>% select(VAR),
                                                               col = list(VAR = var_cols ), 
                                                               annotation_name_gp = gpar(fontsize = 15, fontface = "bold"),
                                                               annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                               direction = "horizontal",
                                                                                               title_position = "lefttop", 
                                                                                               title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                                               labels_gp = gpar(fontsize = 12),
                                                                                               ncol = 5)),
                               column_names_rot = 90,
                               column_names_gp = grid::gpar(fontsize = 15),
                               left_annotation = rowAnnotation(df= msc_col_annot %>% dplyr::select( Strain,Pseudotime) ,
                                                                  col = list(Strain = strain_don_cols[unique(msc_col_annot[, "Strain"])],
                                                                             Pseudotime = colorRamp2(c(min(msc_col_annot[,"Pseudotime"]), max(msc_col_annot[,"Pseudotime"])), c("white", "#281e5d"))), 
                                                                  annotation_name_gp = gpar(fontsize = 15, fontface = "bold"), 
                                                                  annotation_name_rot = 90,
                                                                  annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                                 direction = "horizontal", 
                                                                                                 title_gp = gpar(fontsize = 12, fontface = "bold"), 
                                                                                                 labels_gp = gpar(fontsize = 12)),
                                                                  show_legend = c(FALSE, FALSE) ##https://support.bioconductor.org/p/82447/
                               ),
                               # fontsize = 14,
                               # heatmap_legend_param = list(ncol = 1, nrow = 4, legend_direction = "vertical", by_row = T),
                               # heatmap_legend_param = list(direction = "horizontal", legend_gp = gpar(fontsize = lgsize)),
                               heatmap_legend_param = list(title = "Expression",
                                                           title_gp = gpar(fontsize = 12, fontface = "bold"),
                                                           labels_gp = gpar(fontsize = 12),
                                                           direction = "horizontal",
                                                           title_position = "lefttop"#,
                                                           # direction = "vertical",
                                                           # labels_rot = 90#,
                                                           # grid_width = unit(6, "mm"),
                                                           # legend_width = unit(lg_width, "cm")
                               ),
                               row_names_gp = grid::gpar(fontsize = 14)
                               # 
) %>%
  ##Merging legends into 1 column. Could be set to F
  draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom", padding = unit(c(2, 15, 2, 2), "mm"))



```

```{r, fig.height=5, fig.width=12}
hmap_t
# c("#b7e577", "#3ce07e")
```

```{r, results="asis", eval=T}
# gogo()
```

```{r, results='asis', warning=F, message=F}

hmap_plt<- list(fm_bal_ts, 
                # map(cond_gns_de_afm, ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)), 
                cond_gns_de_afm,
                rep(30, length(cond_gns_de_afm)), 
                rep(T, length(cond_gns_de_afm)), 
                names(fm_bal_ts)) %>% 
  pmap(., possibly(
    ~{
  tseq_de_hmap(tseq = ..1, de_tbl= ..2, pnts = ..3, rnames = ..4, comp_grps = ..5)
}, 'missing genes')
)
  
  # tseq_de_hmap(tseq = fm_bal_ts[[1]], gns= map(cond_gns_de_afm[1], ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)) %>% flatten() %>% as.character(), pnts = 30, rnames = F, title = "Strain DE")
  #                      

```

```{r, results='asis', warning=F, message=F}
# > cond_gns_de_afm[[1]] %>% filter(pvalue<0.05 &  gene_id %in% cond_gns_de_afm[[2]][cond_gns_de_afm[[2]]$pvalue < 0.05,]$gene_id) %>% filter(!is.na(gene_id))
# overlap_gns <- cond_gns_de_afm[["f.MSC48C"]] %>% filter(pvalue<0.05 & !is.na(gene_id)) %>% inner_join(cond_gns_de_afm[["MSC60"]][cond_gns_de_afm[["MSC60"]]$pvalue < 0.05,], by = "gene_id") 
overlap_gns <- cond_gns_de_afm[["f.MSC48C"]] %>% filter(pvalue<0.05 & !is.na(gene_id)) %>% inner_join(cond_gns_de_afm[["f.MSC66C"]][cond_gns_de_afm[["f.MSC66C"]]$pvalue < 0.05,], by = "gene_id") 

```


```{r, results='asis', warning=F, message=F}

all_sig_f_gns_afm <- map(cond_gns_de_afm[str_detect(names(cond_gns_de_afm), paste0(solo_don, collapse = "|"))], ~{
# all_sig_f_gns_afm <- map(cond_gns_de_afm[solo_don2], ~{
  .x %>% 
  select(-c(starts_with("df"))) %>%
  rename("wstat_overal"=waldStat, "p_overal"=pvalue) %>%
  pivot_longer(.,cols = starts_with(c('waldStat_','pvalue_')), 
                     # names_pattern = "(.*)_(SC._vs.*)",
                     names_pattern = "(waldStat|pvalue)_(.*)",
                     names_to = c( ".value", "ts_pw_comp")) %>%
    filter(waldStat > 30 & pvalue < 0.05)
}) 

```


```{r, results='asis', warning=F, message=F}
all_sig_f_gns_df_afm <- list(
  all_sig_f_gns_afm[str_detect(names(cond_gns_de_afm), "a.MSC")] %>% bind_rows(., .id = "donor"), 
  all_sig_f_gns_afm[str_detect(names(cond_gns_de_afm), "f.MSC")] %>% bind_rows(., .id = "donor"), 
  all_sig_f_gns_afm[str_detect(names(cond_gns_de_afm), "m.MSC")] %>% bind_rows(., .id = "donor"), 
  all_sig_f_gns_afm[str_detect(names(cond_gns_de_afm), "f2.MSC")] %>% bind_rows(., .id = "donor"), 
  all_sig_f_gns_afm[str_detect(names(cond_gns_de_afm), "f3.MSC")] %>% bind_rows(., .id = "donor")) %>%
  set_names(c("a", "f", "m", "f2", "f3"))

all_sig_f_gns_df_afm <- map(all_sig_f_gns_df_afm, ~.x %>%
  group_by(gene_id) %>%
  mutate("freq_btwn_don" = n_distinct(donor)) %>%
  group_by(donor, gene_id) %>%
  mutate("freq_wthn_don" = n_distinct(ts_pw_comp)) %>%
  ungroup()
)


```

```{r, results='asis', warning=F, message=F}
map(all_sig_f_gns_df_afm, ~.x %>% 
  arrange(desc(freq_btwn_don), Gene) %>%
  distinct(donor, gene_id, Gene, freq_btwn_don)
)
```


```{r, results='asis', warning=F, message=F, fig.width=2.5, fig.height=2}
map(all_sig_f_gns_df_afm, ~.x %>% 
  arrange(desc(freq_btwn_don), Gene) %>%
  distinct(donor, gene_id, Gene, freq_btwn_don) %>%
  mutate(across(freq_btwn_don, factor)) %>% 
  distinct(gene_id, .keep_all = T) %>%
  count(freq_btwn_don) %>%
  ggplot(aes(x = freq_btwn_don, y = n)) +
  geom_col() +
  theme_classic() +
  labs(x = "Frequency across donors", y = "DE gene count")
)

```

```{r, results='asis', warning=F, message=F, eval=FALSE}

hmap_plt<- list(fm_bal_ts, 
                # map(cond_gns_de_afm, ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)), 
                map(cond_gns_de_afm, ~.x %>% filter(gene_id %in% overlap_gns$gene_id)),
                rep(30, length(cond_gns_de_afm)), 
                rep(T, length(cond_gns_de_afm)), 
                names(fm_bal_ts)) %>% 
  pmap(., #possibly(
    ~{
  tseq_de_hmap(tseq = ..1, de_tbl= ..2, pnts = ..3, rnames = ..4, comp_grps = ..5)
}#, 'missing genes')
)
  
  # tseq_de_hmap(tseq = fm_bal_ts[[1]], gns= map(cond_gns_de_afm[1], ~.x %>% filter(if_any(matches("pvalue"), ~. < 0.05)) %>% pull(gene_id)) %>% flatten() %>% as.character(), pnts = 30, rnames = F, title = "Strain DE")
  #                      
hmap_plt

```


```{r, results='asis', warning=F, message=F}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC49C")],
               cond_obj_lst = map(cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC49C")], ~.x %>% filter(gene_id %in% overlap_gns$gene_id)),
               pw_comp= 'waldStat_M49_SC1_vs_M49_SC5',
               n_gns = c(1:4)
)

```

```{r, results='asis', warning=F, message=F}

tde_smooth_plt(tde_obj_lst = fm_bal_ts[str_detect(names(fm_bal_ts), "MSC60")],
               cond_obj_lst = map(cond_gns_de_afm[str_detect(names(cond_gns_de_afm), "MSC60")], ~.x %>% filter(gene_id %in% overlap_gns$gene_id)),
               pw_comp= 'waldStat_M60_SC3_vs_M60_SC7',
               n_gns = c(1:4)
)

```

###### MSC49 strain DE assessing var genes


```{r, fig.width=7.5, fig.height=2}
## convert SCE to seurat
# fm_bal_seu <- map(fm_bal_sce, as.Seurat)
```

```{r, fig.width=7.5, fig.height=2}
# m49_var_gns_exp <- FetchData(object = fm_harmony_tr_pt_sset, vars = c('PF3D7-0400400', "PF3D7-0412400", 'PF3D7-0421300', "PF3D7-0412700", 'PF3D7-0712900', 'PF3D7-1041300', "PF3D7-0632800", 'PF3D7-0937800', 'PF3D7-1240600', 'PF3D7-1240400', 'PF3D7-0808700', 'PF3D7-0900100'), slot = "data") 

# m49_var_gns_exp <- map(fm_bal_sce, ~.x@assays@data$normcounts[c('PF3D7-0400400', "PF3D7-0412400", 'PF3D7-0421300', "PF3D7-0412700", 'PF3D7-0712900', 'PF3D7-1041300', "PF3D7-0632800", 'PF3D7-0937800', 'PF3D7-1240600', 'PF3D7-1240400', 'PF3D7-0808700', 'PF3D7-0900100'),] %>% t() %>% as.data.frame() )
m49_var_gns_exp <- map(fm_bal_sce, ~.x@assays@data$logcounts[c('PF3D7-0400400', "PF3D7-0412400", 'PF3D7-0421300', "PF3D7-0412700", 'PF3D7-0712900', 'PF3D7-1041300', "PF3D7-0632800", 'PF3D7-0937800', 'PF3D7-1240600', 'PF3D7-1240400', 'PF3D7-0808700', 'PF3D7-0900100'),] %>% t() %>% as.data.frame() )

```


```{r, fig.width=7.5, fig.height=2}
msc_bal_umap_gns <- map2(fm_bal_sce, m49_var_gns_exp, ~{
  reducedDims(.x)$UMAP.HARMONY_PC_MAN %>% 
    as.data.frame() %>% 
    rownames_to_column("bcode") %>% 
    left_join(., .x@colData[,c("StrainDon", "strain", "donor")] %>% data.frame() %>% rownames_to_column("bcode")) %>%  
    left_join(., .y %>% rownames_to_column("bcode"))
  })

```

```{r, fig.height=18, fig.width=2.5}
# custom_colors <- colorRampPalette(c("lightgrey", "blue"))(100)

msc_bal_umap_gns[["f.MSC48C"]] %>% 
  filter(donor == "MSC48C") %>%
  pivot_longer(cols = starts_with("PF"), names_to = "gene_id", values_to = "Expression") %>%
  left_join(., var_gn_info) %>%
  mutate(across(donor, ~str_remove(., "SC")),
         # across(strain, ~factor(.x, levels = c("SC1", "SC2")))
         ) %>%
  arrange(gene_id,strain,Expression) %>%
  # ggplot(., aes(x = weave_factors(Strain, donor), y = Counts, fill = donor)) +
  ggplot(., aes(x = umapharmonyPCman_1, y = umapharmonyPCman_2, colour = Expression))+
  geom_point(size = 1.5) +
  scale_color_distiller( direction = 1) +
  # ggh4x::facet_nested_wrap(vars(Vars, strain),  nrow = 1, scales = "free_x") +
  ggh4x::facet_nested(strain_up+gene_id+ var_grps ~  strain) +
  # ggh4x::facet_grid2(gene_id ~  strain)) +
  theme_classic()+
  theme(legend.position = "bottom", 
        axis.text = element_blank(),
        strip.text =element_text(size=12,face="bold", colour = 'black'))+
  # guides(x = "axis_nested") +
  labs(x = "UMAP_1", y="UMAP_2", title = "Var genes") 
```

###### END MSC49 strain DE assessing var genes

```{r, results="asis", eval=T}
gogo()
```

## Symptomatic VS asymptomatic

### Balance symptomatic and asymptomatic


Remove lab and MSC12 and those lacking symptomatic condition
```{r}

fm_tr_sce_pt_chunks_sshot_no_dbt <- fm_tr_sce_pt_chunks_sshot[, !is.na(fm_tr_sce_pt_chunks_sshot@colData$StrainDon) & !str_detect(fm_tr_sce_pt_chunks_sshot@colData$StrainDon, "Negative|Doublet")]
```


```{r}

fm_tr_sce_pt_chunks_sshot_no_dbt@colData[, c("Symptomatic", "dset", "donor")] %>%
  as.data.frame() %>%
  dplyr::count(Symptomatic, dset, donor)
```

```{r}
## Get only the asexuals
# fm_seu <- map(msc_w_cln_strns[c("MSC49C", "MSC60")], ~.x[,.x@meta.data$stage %in% c("Early trophozoite", "Late ring")] %>% AddMetaData(., ))

fm_tr_sce_pt_chunks_sshot_no_dbt <- fm_tr_sce_pt_chunks_sshot[, !is.na(fm_tr_sce_pt_chunks_sshot@colData$StrainDon) & !str_detect(fm_tr_sce_pt_chunks_sshot@colData$StrainDon, "Negative|Doublet")]
```

### Balance (subsample) the strain cells within pseudotime chunks
```{r}
##Subsample the cells within pseudotime chunks calculated above to ensure balanced representation of field and lab cells along trajectories 

##Declaring the variables to group by for the different datasets 
# grps <- c('ptime_cat', 'ptime_cat', 'ptime_cat')
# grps2 <- list(c('ptime_cat', 'lf_strain'), c('ptime_cat', 'lf_strain'), c('ptime_cat', 'lf_strain'))


##- Asexual stages retaining pseudotime by stage groups containing MSC14 field cells and grouping by pseudotime chunk and stage : then sampling within pseudotime and stage group with sample size equal to the minimum dataset between field and lab for each stage within each pseudotime group
##- Gams retaining pseudotime groups containing MSC14 field cells and grouping by pseudotime chunk only : then sampling within pseudotime group with sample size equal to the minimum dataset between field and lab

##Tabulate balanced cells
set.seed(3783822)

##Remove lab cells - can be included in future if necessary by uncommenting line below 'filter(strain_dset != 'lab') %>%'

## get the group to downsample by
ds_grp <- c(rep('Symptomatic', 1))

##Revert order in filter https://stackoverflow.com/questions/70951525/filter-dataframe-within-a-group-with-one-column-meeting-an-and-condition-in-r
# balanced_cells_tbl_str <- list(msc14_sub_sce_unbal_ptchunks, grps, grps2)  %>%
balanced_cells_tbl_str <- list(list(fm_tr_sce_pt_chunks_sshot_no_dbt), ds_grp)  %>%
  pmap(~..1@colData[,c('StrainDon', 'ptime_cat', 'ptime_bal_cat', 'Symptomatic', 'donor')] %>% 
    as.data.frame() %>% 
    rownames_to_column('cell_bc') %>% 
    group_by(across(all_of(c('ptime_bal_cat', ..2)))) %>% 
    mutate(cnt  = sum(!is.na(!!rlang::sym(..2)))) %>% 
             group_by(across(all_of(c('ptime_bal_cat')))) %>% 
             mutate(max_cnt = max(cnt)) %>% 
             mutate(max2_cnt = max(cnt[cnt != max_cnt], na.rm = TRUE)) %>%
             group_by(ptime_bal_cat, Symptomatic) %>% 
             mutate(samp = sample(n()))%>%
             filter(samp <= max2_cnt) %>%
             ungroup() #%>%
             # mutate(across(ptime_bal_cat, droplevels))
  )

balanced_cells_cnts_str <- balanced_cells_tbl_str %>%
  map(. %>% dplyr::count(ptime_bal_cat,  Symptomatic, max2_cnt, cnt))

balanced_cells_tbl_str %>%
  map(. %>% dplyr::count(ptime_bal_cat,Symptomatic))

##Get balanced cells

# set.seed(989778)
balanced_cells_str <- balanced_cells_tbl_str %>% 
  map(. %>% dplyr::select(cell_bc, ptime_bal_cat))

##Subset the balanced cells for the 3 objects
fm_sympto_bal_sce <- pmap(list(list(fm_tr_sce_pt_chunks_sshot_no_dbt), balanced_cells_str), ~..1[,..2$cell_bc])


```

### Retain cells with strains having more than 10 in each stage
```{r}
##Remove groups with less than 10
# fm_sce_bal <- map(fm_sce_bal, ~.x[,.x@colData[, 'StrainDon', drop = F] %>% data.frame()  %>% rownames_to_column('bcode') %>% group_by(StrainDon) %>% filter(n() >=20) %>% pull(bcode)])
# 
# map(fm_sce_bal, ~table(.x@colData[, 'StrainDon']) )

```

```{r, results="asis", eval=T}
balanced_cells_tbl_str[[1]] %>% count(ptime_bal_cat, Symptomatic)
```



```{r, results="asis", fig.width=5, fig.height=3.5}
library("ggpubr")
tbl = map(fm_sympto_bal_sce, ~umap_meta_df[umap_meta_df$cell_bc %in% colnames(.x),])

map(tbl, ~.x%>%
  ggplot(., aes(x = slingPseudotime_1, col=Symptomatic, fill=Symptomatic)) + 
  geom_density(aes(y = ..count..*0.3), size=2, bw = 0.3, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  dens_thme +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))
)

map(tbl, ~.x%>%
  ggboxplot(., y = "slingPseudotime_1",x="Symptomatic", col="Symptomatic") + 
  stat_compare_means(method = "wilcox.test") +gbox_theme
)
  


```

#### Put subseted objects into a list and calculate metrics need for plotting and tradeseq input
```{r}
lins = rep(1,1)
# lins_m = rep(c(1,2),4)

fm_sympto_bal_sds <- map(fm_sympto_bal_sce, ~SlingshotDataSet(.x))
fm_sympto_bal_ptime <- pmap(list(fm_sympto_bal_sce, lins),  ~slingPseudotime(..1)[,..2] )
fm_sympto_bal_cw <- pmap(list(fm_sympto_bal_sce,  lins),  ~slingCurveWeights(..1)[,..2] )

```


```{r}
## Batch design correction for tradeseq
u_don <- model.matrix(~  donor, fm_sympto_bal_sce[[1]]@colData[, c("donor", "Symptomatic")] %>% data.frame())

# ## tradeseq manual example - Batch design correction for tradeseq
# counts <- as.matrix(countMatrix)
# 
# batch <- factor(rep(c("A", "B"), each = ncol(counts)/2))
# U <- model.matrix(~batch)
# sceBatch <- fitGAM(counts = counts,
#                    U = U,
#                    sds = crv, 
#                    nknots = 6, 
#                    verbose = FALSE)
```


```{r}
## create directory
system(paste0("mkdir -p /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/"))


saveRDS(fm_sympto_bal_sce[[1]], '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_sce.RDS')

saveRDS(fm_sympto_bal_ptime[[1]], '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_ptime.RDS')

saveRDS(fm_sympto_bal_cw[[1]], '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_cw.RDS')

```


```{r}
## plot balanced symptomatic vs assymptomatic by pseudotime
v3_afm_harmony_ss_res_fld_pt_bal <- v3_afm_harmony_ss_res_fld_pt[,colnames(fm_sympto_bal_sce[[1]])]
v3_afm_harmony_ss_res_fld_pt_bal@meta.data$pseudotime <- fm_sympto_bal_sce[[1]]$slingPseudotime_1
v3_afm_harmony_ss_res_fld_pt_bal@meta.data$ptime_cat <- fm_sympto_bal_sce[[1]]$ptime_cat

```

```{r}
## plot balanced symptomatic vs assymptomatic by pseudotime
v3_afm_harmony_ss_res_fld_pt_bal@meta.data$sympto <- ifelse(v3_afm_harmony_ss_res_fld_pt_bal@meta.data$Symptomatic == "Yes", "Symptomatic", "Asymptomatic")

map(v3_afm_harmony_ss_res_fld_pt_bal, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", features = "pseudotime", pt.size = 0.5, dims = c(1,2), split.by = "sympto"))

```

```{r}
## plot balanced symptomatic vs assymptomatic by pseudotime
v3_afm_harmony_ss_res_fld_pt_bal <- v3_afm_harmony_ss_res_fld_pt[,colnames(fm_sympto_bal_sce[[1]])]
v3_afm_harmony_ss_res_fld_pt_bal@meta.data$pseudotime <- fm_sympto_bal_sce[[1]]$slingPseudotime_1
v3_afm_harmony_ss_res_fld_pt_bal@meta.data$ptime_cat <- fm_sympto_bal_sce[[1]]$ptime_cat

DimPlot(v3_afm_harmony_ss_res_fld_pt_bal, reduction = "umap.harmony_PC_man", group.by = "ptime_cat", pt.size = 0.5, dims = c(1,2), label = T, split.by = "Symptomatic")

```

```{r, results="asis", eval=T}
gogo()
## Tradeseq script
# nextflow run ../mali22_code_lnk/nf_scripts/submsn_tseq.nf --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal" --scrpt /nfs/users/nfs_j/jr35/multipurpose_scripts/tradeseq_nf.R --condtn "Symptomatic" --knots "4" --gns_cut "5"  --resume
```


```{r, results="asis", eval=T}
fm_sympto_bal_sce_ts <- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_sce_ts.RDS")
fm_sympto_bal_sce_de <- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_sce_de.RDS")
```


```{r, results="asis", eval=T}
fm_sympto_bal_sce_de %>% arrange(desc(waldStat))
```

```{r}
##NOTE!!! replace the conditional test file with the newly generated one which applied 7 clusters
##Ajusting the global and pairwise pvalues - old p values discarded
fm_sympto_bal_sce_de <- fm_sympto_bal_sce_de %>%
                             mutate(across(starts_with('pvalue'), ~p.adjust(., "fdr"))) %>%
                             arrange(desc(waldStat)) %>%
                             rownames_to_column('gene_id')

```


```{r}

fm_sympto_bal_sce_de <- left_join(fm_sympto_bal_sce_de, Pf3D7_genes_plasmoDB %>% dplyr::select(gene_id, Gene, Ka_Ks, `Transcript Product Description`, TM_domains, `Computed GO Functions`), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene))

```


```{r, results="asis", eval=T}
fm_sympto_bal_sce_de %>% arrange(pvalue)
```

#### smoothers for select genes that are differentially expressed between strains {.tabset}


```{r, results='asis', warning=F, message=F}

tde_smooth_plt(tde_obj_lst = list(fm_sympto_bal_sce_ts),
               cond_obj_lst = list(fm_sympto_bal_sce_de),
               pw_comp= 'waldStat',
               n_gns = c(1:20)
)

```


```{r, results='asis', warning=F, message=F, fig.width=6, fig.height=6}

tseq_de_hmap(tseq = fm_sympto_bal_sce_ts, de_tbl= fm_sympto_bal_sce_de, pnts = 30, rnames = T, comp_grps = "Symptomatic", Pv = 1e-20)

```

### accounting for batch

```{r, results="asis", eval=T}
balanced_cells_tbl_str[[1]] #%>% count(Symptomatic, donor)
```


#### Put subseted objects into a list and calculate metrics need for plotting and tradeseq input


```{r}
## Batch design correction for tradeseq
fm_sympto_bal_u_don <- model.matrix(~  donor, fm_sympto_bal_sce[[1]]@colData[, c("donor", "Symptomatic")] %>% data.frame())

# ## tradeseq manual example - Batch design correction for tradeseq
# counts <- as.matrix(countMatrix)
# 
# batch <- factor(rep(c("A", "B"), each = ncol(counts)/2))
# U <- model.matrix(~batch)
# sceBatch <- fitGAM(counts = counts,
#                    U = U,
#                    sds = crv, 
#                    nknots = 6, 
#                    verbose = FALSE)
```


```{r}
## create directory
saveRDS(fm_sympto_bal_u_don, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_udon.RDS')

```



```{r, results="asis", eval=F}
# fm_bal_ts_de  <- fitGAM(counts = counts(fm_sympto_bal_sce[[1]]), 
#                     pseudotime = fm_sympto_bal_ptime[[1]], 
#                     cellWeights = fm_sympto_bal_cw[[1]], 
#                     nknots =4,
#                     genes = rownames(fm_sympto_bal_sce[[1]])[rowSums(counts(fm_sympto_bal_sce[[1]]) != 0) > 5],
#                     conditions = factor(SummarizedExperiment::colData(fm_sympto_bal_sce[[1]])[, "Symptomatic"]), 
#                     verbose = T, 
#                     parallel = T,
#                     U = fm_sympto_bal_u_don
#                     )
## Tradeseq script
# nextflow run submsn_tseq.nf --sce_obj "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_{sce,ptime,cw}.RDS" --condtn "Symptomatic" --knots "4" --gns_cut "5"  --resume
```



```{r, results="asis", eval=T}
gogo()
## Tradeseq script
# nextflow run ../mali22_code_lnk/nf_scripts/submsn_tseq_batch.nf --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal" --scrpt /nfs/users/nfs_j/jr35/multipurpose_scripts/tradeseq_nf.R --condtn "Symptomatic" --knots "4" --gns_cut "5"  --resume

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_tseq.nf --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal" --scrpt /nfs/users/nfs_j/jr35/multipurpose_scripts/tradeseq_batch_nf.R --condtn "Symptomatic" --knots "4" --gns_cut "5"  --resume
```


```{r, results="asis", eval=T}
fm_sympto_bal_batch_ts <- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_batch_ts.RDS")
fm_sympto_bal_batch_de <- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_sympto_bal_batch_de.RDS")
```


```{r, results="asis", eval=T}
fm_sympto_bal_batch_de %>% arrange(desc(waldStat))
```

```{r}
##NOTE!!! replace the conditional test file with the newly generated one which applied 7 clusters
##Ajusting the global and pairwise pvalues - old p values discarded
fm_sympto_bal_batch_de <- fm_sympto_bal_batch_de %>%
                             mutate(across(starts_with('pvalue'), ~p.adjust(., "fdr"))) %>%
                             arrange(desc(waldStat)) %>%
                             rownames_to_column('gene_id')

```


```{r}

fm_sympto_bal_batch_de <- left_join(fm_sympto_bal_batch_de, Pf3D7_genes_plasmoDB %>% dplyr::select(gene_id, Gene, Ka_Ks, `Transcript Product Description`, TM_domains, `Computed GO Functions`), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene))

```


```{r, results="asis", eval=T}
fm_sympto_bal_batch_de %>% arrange(pvalue)
```

#### smoothers for select genes that are differentially expressed between strains {.tabset}


```{r, results='asis', warning=F, message=F}

tde_smooth_plt(tde_obj_lst = list(fm_sympto_bal_batch_ts),
               cond_obj_lst = list(fm_sympto_bal_batch_de),
               pw_comp= 'waldStat',
               n_gns = c(1:20)
)

```


```{r, results='asis', warning=F, message=F, fig.width=6, fig.height=6}

tseq_de_hmap(tseq = fm_sympto_bal_batch_ts, de_tbl= fm_sympto_bal_batch_de, pnts = 30, rnames = T, comp_grps = "Symptomatic", Pv = 0.05)

```


trial with batch

```{r, results="asis"}
## Save for tradeseq
summary(fm_tr_sce_pt_chunks_sshot@colData$slingPseudotime_1)
fm_tr_sce_pt_chunks_3 <- fm_tr_sce_pt_chunks_sshot[,fm_tr_sce_pt_chunks_sshot@colData$slingPseudotime_1 > 13 & fm_tr_sce_pt_chunks_sshot@colData$slingPseudotime_1 < 14 ]
```

```{r}
set.seed(5656)
bcodes <- fm_tr_sce_pt_chunks_3@colData[,c('Symptomatic', 'donor')]%>%
    as.data.frame()  %>% 
  filter(str_detect(donor, "MSC")) %>% 
    rownames_to_column('cell_bc') %>% 
    group_by(Symptomatic, donor) %>% 
    mutate(samp = sample(n())) %>%
    filter(samp <= 5) %>%
  ungroup() %>%
  pull(cell_bc)
    
fm_tr_sce_pt_chunks_3_sample <- fm_tr_sce_pt_chunks_3[,bcodes]
```

```{r}
## Batch design correction for tradeseq
u_don <- model.matrix(~  donor, fm_tr_sce_pt_chunks_3_sample@colData[, c("donor", "Symptomatic")] %>% data.frame())

# ## tradeseq manual example - Batch design correction for tradeseq
# counts <- as.matrix(countMatrix)
# 
# batch <- factor(rep(c("A", "B"), each = ncol(counts)/2))
# U <- model.matrix(~batch)
# sceBatch <- fitGAM(counts = counts,
#                    U = U,
#                    sds = crv, 
#                    nknots = 6, 
#                    verbose = FALSE)
```

```{r}
fm_tr_sce_pt_chunks_3_sample_sds <- SlingshotDataSet(fm_tr_sce_pt_chunks_3_sample)
fm_tr_sce_pt_chunks_3_sample_ptime <- slingPseudotime(fm_tr_sce_pt_chunks_3_sample)[,1]
fm_tr_sce_pt_chunks_3_sample_cw <- slingCurveWeights(fm_tr_sce_pt_chunks_3_sample)[,1]

```


```{r}
## create directory
system(paste0("mkdir -p /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/"))


saveRDS(fm_tr_sce_pt_chunks_3_sample, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_tr_sce_pt_chunks_3_sample_obj.RDS')

saveRDS(fm_tr_sce_pt_chunks_3_sample_ptime, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_tr_sce_pt_chunks_3_sample_ptime.RDS')

saveRDS(fm_tr_sce_pt_chunks_3_sample_cw, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_tr_sce_pt_chunks_3_sample_cw.RDS')

```


```{r, results="asis", eval=T}
gogo()
## Tradeseq script
# nextflow run submsn_tseq.nf --params.sce_obj "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/symptos/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_tr_sce_pt_chunks_3_sample_*{obj,ptime,cw}.RDS" --params.condtn "Symptomatic" --params.knots "4" --params.gns_cut "5"  --resume
```


```{r, results="asis", eval=T}
fm_tr_sce_pt_chunks_3_sample_obj_ts<- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_tr_sce_pt_chunks_3_sample_obj_ts.RDS")
fm_tr_sce_pt_chunks_3_sample_obj_de<- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/sympto/fm_tr_sce_pt_chunks_3_sample_obj_de.RDS")
```



```{r}
gogogo()
```

```{r}
m48_wc <- str_remove(colnames(v3_afm_harmony_ss_res_fld_pt[,v3_afm_harmony_ss_res_fld_pt@meta.data$harmony_clusters_asex == "2" & v3_afm_harmony_ss_res_fld_pt@meta.data$donor == "MSC48"]), "M48_")
```

```{r}
DimPlot(msc_w_cln_strns[["MSC48"]], cells.highlight = m48_wc)

```

```{r}
m49_wc <- str_remove(colnames(v3_afm_harmony_ss_res_fld_pt[,v3_afm_harmony_ss_res_fld_pt@meta.data$harmony_clusters_asex == "2" & v3_afm_harmony_ss_res_fld_pt@meta.data$donor == "MSC49"]), "M49_")

```

```{r}
DimPlot(msc_w_cln_strns[["MSC49"]], cells.highlight = m49_wc)

```

```{r}
msc49 <- msc_w_cln_strns[["MSC49"]]@meta.data %>% rownames_to_column("bcode") %>% mutate(w_clst_stgs = case_when(bcode %in% m49_wc ~ "wrd_asex", T ~ stage)) %>% column_to_rownames("bcode") %>% select(w_clst_stgs) %>% AddMetaData(msc_w_cln_strns[["MSC49"]], .) 

# m49_asex <- msc_w_cln_strns[["MSC49"]][,msc_w_cln_strns[["MSC49"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]
# 
# 
# m49_asex <- m49_asex@meta.data %>% rownames_to_column("bcode") %>% mutate(w_clst = case_when(bcode %in% m49_wc ~ "wrd_clstr", T ~ "other_clstr")) %>% column_to_rownames("bcode") %>% select(w_clst) %>% AddMetaData(m49_asex, .)
# 
# DimPlot(m49_asex, group.by = "w_clst")
# 

```

```{r}
VlnPlot(msc49, features = "PF3D7-0204800", group.by = "w_clst_stgs")
VlnPlot(msc49, features = "PF3D7-0204900", group.by = "w_clst_stgs")
```


```{r}
VlnPlot(msc49, features = "PF3D7-0406200", group.by = "w_clst_stgs")
```

```{r}

m49_asex <- msc_w_cln_strns[["MSC49"]][,msc_w_cln_strns[["MSC49"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]
m49_asex <- m49_asex@meta.data %>% rownames_to_column("bcode") %>% mutate(w_clst = case_when(bcode %in% m49_wc ~ "wrd_clstr", T ~ "other_clstr")) %>% column_to_rownames("bcode") %>% select(w_clst) %>% AddMetaData(m49_asex, .)

DimPlot(m49_asex, group.by = "w_clst")


```


```{r}
Idents(m49_asex) <- "w_clst"
# w_markrs <- FindMarkers(m49_asex, ident.1 = )

m49_afm_seu <- m49_asex

m49_afm_seu[["RNA"]]$data <- as(object = m49_afm_seu[["RNA"]]$data, Class = "dgCMatrix")
all.markers <- FindAllMarkers(object = m49_afm_seu, test.use = "MAST")
head(all.markers, 10)

```


```{r}

m48_asex <- msc_w_cln_strns[["MSC48"]][,msc_w_cln_strns[["MSC48"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]
m48_asex <- m48_asex@meta.data %>% rownames_to_column("bcode") %>% mutate(w_clst = case_when(bcode %in% m48_wc ~ "wrd_clstr", T ~ "other_clstr")) %>% column_to_rownames("bcode") %>% select(w_clst) %>% AddMetaData(m48_asex, .)
DimPlot(m48_asex, group.by = "w_clst")


```

```{r}
Idents(m48_asex) <- "w_clst"
# w_markrs <- FindMarkers(m48_asex, ident.1 = )

m48_afm_seu <- m48_asex

m48_afm_seu[["RNA"]]$data <- as(object = m48_afm_seu[["RNA"]]$data, Class = "dgCMatrix")
w48_all.markers <- FindMarkers(object = m48_afm_seu, test.use = "MAST", ident.1 = "wrd_clstr", ident.2 = "other_clstr")
head(w48_all.markers, 10)

```

```{r}
m48_afm_seu@meta.data %>% count(StrainDon)
m49_afm_seu@meta.data %>% count(StrainDon)

```


```{r}
m48_afm_seu <- m48_afm_seu[, !str_detect(m48_afm_seu@meta.data$StrainDon, "Negative|Doublet")]

m49_afm_seu <- m49_afm_seu[, !str_detect(m49_afm_seu@meta.data$StrainDon, "Negative|Doublet") ]
```

```{r}
Idents(m48_afm_seu) <- "StrainDon"

m48_afm_seu@meta.data %>% count(StrainDon)
m48_strn_d_markers <- FindMarkers(object = m48_afm_seu, test.use = "MAST", ident.1 = "M48_SC2", ident.2 = "M48_SC4")
m48_strn_d_markers_df <- m48_strn_d_markers %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m48_strn_d_markers_df %>% filter(p_val_adj < 0.05)

# PF3D7_1016300 PF3D7_1127000 https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8527989/
# PF3D7_1127000 https://elifesciences.org/reviewed-preprints/95879
# PF3D7_1223900 https://www.biorxiv.org/content/10.1101/2023.02.13.528367v1.full
# PF3D7_1324900 https://www.sciencedirect.com/science/article/pii/S2213716523001376
```


```{r}
Idents(m49_afm_seu) <- "StrainDon"

m49_afm_seu@meta.data %>% count(StrainDon)
m49_strn_d_markers <- FindMarkers(object = m49_afm_seu, test.use = "MAST", ident.1 = "M49_SC1", ident.2 = "M49_SC5")
m49_strn_d_markers_antd <- m49_strn_d_markers  %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))

m49_strn_d_markers_antd %>% filter(p_val_adj < 1e-20)

## vps3 extracellular vessicle content https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8764524/
## ΔHsp70x - virulence cytoadhesion - dispensable 
##VAR_PF3D7-1041300: Mapping immune variation and var gene switching in naive hosts infected with Plasmodium falciparum - PMC.html
##CBP2 ΔGEXP07 parasites have a growth advantage compared to wild-type parasites https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7078486/
##CBP2 ΔGEXP07 parasites have a growth advantage compared to wild-type parasites https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7078486/
## MAHRP2 interacts with pfemp1 rers cleft  https://www.nature.com/articles/ncomms16044, https://journals.asm.org/doi/10.1128/msphere.00755-21
```


```{r}
m33_afm_seu <- msc_w_cln_strns[["MSC33"]][,msc_w_cln_strns[["MSC33"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]
m33_afm_seu <- m33_afm_seu[, str_detect(m33_afm_seu@meta.data$StrainDon, "SC2|SC5|SC6|SC7")]

m33_afm_seu[["RNA"]]$data <- as(object = m33_afm_seu[["RNA"]]$data, Class = "dgCMatrix")

Idents(m33_afm_seu) <- "StrainDon"

m33_afm_seu@meta.data %>% count(StrainDon)
m33_strn_d_markers <- FindAllMarkers(object = m33_afm_seu, test.use = "MAST")
m33_strn_d_markers_df <- m33_strn_d_markers  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m33_strn_d_markers_df %>% filter(p_val_adj < 1e-5)

```

```{r}
m55_afm_seu <- msc_w_cln_strns[["MSC55"]][,msc_w_cln_strns[["MSC55"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]

m55_afm_seu[["RNA"]]$data <- as(object = m55_afm_seu[["RNA"]]$data, Class = "dgCMatrix")

Idents(m55_afm_seu) <- "StrainDon"

m55_afm_seu@meta.data %>% count(StrainDon)
m55_strn_d_markers <- FindMarkers(object = m55_afm_seu, test.use = "MAST", ident.1 = "M55_SC6", ident.2 = "M55_SC8")
m55_strn_d_markers_df <- m55_strn_d_markers  %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m55_strn_d_markers_df %>% filter(p_val_adj < 1e-5)

```


```{r}
m60_afm_seu <- msc_w_cln_strns[["MSC60"]][,msc_w_cln_strns[["MSC60"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]

m60_afm_seu[["RNA"]]$data <- as(object = m60_afm_seu[["RNA"]]$data, Class = "dgCMatrix")

Idents(m60_afm_seu) <- "StrainDon"

m60_afm_seu@meta.data %>% count(StrainDon)
m60_strn_d_markers <- FindMarkers(object = m60_afm_seu, test.use = "MAST", ident.1 = "M60_SC3", ident.2 = "M60_SC7")
m60_strn_d_markers_df <- m60_strn_d_markers  %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m60_strn_d_markers_df %>% filter(p_val_adj < 1e-5)

```


```{r}

# m60_strn_d_markers_ds <- FindMarkers(object = m60_afm_seu, test.use = "MAST", ident.1 = "M60_SC3", ident.2 = "M60_SC7", max.cells.per.ident = 90)
# m60_strn_d_markers_ds_df <- m60_strn_d_markers_ds  %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
# m60_strn_d_markers_ds_df %>% filter(p_val_adj < 0.05)
```

```{r}
m66_afm_seu <- msc_w_cln_strns[["f.MSC66C"]][,msc_w_cln_strns[["f.MSC66C"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]

m66_afm_seu[["RNA"]]$data <- as(object = m66_afm_seu[["RNA"]]$data, Class = "dgCMatrix")

Idents(m66_afm_seu) <- "StrainDon"

m66_afm_seu@meta.data %>% count(StrainDon)
m66_strn_d_markers1 <- FindMarkers(object = m66_afm_seu, test.use = "MAST", ident.1 = "M66_SC5", ident.2 = "M66_SC8")
m66_strn_d_markers1_df <- m66_strn_d_markers1 %>% rownames_to_column("gene")  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m66_strn_d_markers1_df %>% filter(p_val_adj < 1e-5)

```


```{r}
Idents(m66_afm_seu) <- "StrainDon"

m66_afm_seu@meta.data %>% count(StrainDon)
m66_strn_d_markers2 <- FindMarkers(object = m66_afm_seu, test.use = "MAST", ident.1 = "M66_SC7", ident.2 = "M66_SC8")
m66_strn_d_markers2_df <- m66_strn_d_markers2 %>% rownames_to_column("gene")  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m66_strn_d_markers2_df %>% filter(p_val_adj < 1e-5)

```


```{r}
Idents(m66_afm_seu) <- "StrainDon"

m66_afm_seu@meta.data %>% count(StrainDon)
m66_strn_d_markers3 <- FindMarkers(object = m66_afm_seu, test.use = "MAST", ident.1 = "M66_SC5", ident.2 = "M66_SC7")
m66_strn_d_markers3_df <- m66_strn_d_markers3 %>% rownames_to_column("gene")  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m66_strn_d_markers3_df %>% filter(p_val_adj < 1e-5)

```


```{r}
m14_afm_seu <- msc_qcd_anotd[["MSC14"]][,msc_qcd_anotd[["MSC14"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]

m14_afm_seu@meta.data %>% count(strain_qcd)

```

```{r}

Idents(m14_afm_seu) <- "strain_qcd"

m14_strn_d_markers <- FindMarkers(object = m14_afm_seu, test.use = "MAST", ident.1 = "SC1", ident.2 = "SC2")
m14_strn_d_markers_df <- m14_strn_d_markers %>% rownames_to_column("gene")  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m14_strn_d_markers_df %>% filter(p_val_adj < 0.05)

```


```{r}
m14_strn_d_markers2 <- FindMarkers(object = m14_afm_seu, test.use = "MAST", ident.1 = "SC1", ident.2 = "SC6")
m14_strn_d_markers2_df <- m14_strn_d_markers2 %>% rownames_to_column("gene")  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))
m14_strn_d_markers2_df %>% filter(p_val_adj < 0.05)

```

```{r}
# M48, M49, M60
ggvenn(list("M48" = m48_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M49" = m49_strn_d_markers_antd %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M60" = m60_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)))

```


```{r, fig.height=3.8, fig.width=3.8}
# M48, M49, M60
ggvenn(list("M48" = m48_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M49" = m49_strn_d_markers_antd %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M60" = m60_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)),
       show_percentage = F,
       text_size = 6)

```

```{r}
# M48, M49, M60
ggvenn(list("M48" = m48_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M49" = m49_strn_d_markers_antd %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M60" = m60_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M14" = m14_strn_d_markers2_df %>% filter(p_val_adj < 0.05) %>% pull(gene)))

```


```{r}
# M48, M49, M60
ggvenn(list("M48" = m48_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M49" = m49_strn_d_markers_antd %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M60" = m60_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene),
            "M14" = m14_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)))

```

```{r}
# M48, M49, M60
m48_g <- m48_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)
m49_g <- m49_strn_d_markers_antd %>% filter(p_val_adj < 0.05) %>% pull(gene)
m60_g <- m60_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)


m48_m60_g_ovlp <- m48_g[m48_g %in% m60_g]
m49_m60_g_ovlp <- m49_g[m49_g %in% m60_g]

```

```{r}
# M48, M49, M60
m49_m60_g_ovlp <- m49_g[m49_g %in% m60_g]

m49_strn_d_markers_antd %>% filter(gene %in% m49_m60_g_ovlp) %>% mutate(don = "M49") %>% bind_rows(., m60_strn_d_markers_df %>% filter(gene %in% m49_m60_g_ovlp) %>% mutate(don = "M60")) %>% ggplot(aes(x = avg_log2FC, y = Gene, fill = don)) + geom_col(position = position_dodge2()) 
```
```{r, fig.height=5.5, fig.width=4}
# M48, M49, M60
m49_m60_g_ovlp <- m49_g[m49_g %in% m60_g]

m49_strn_d_markers_antd %>% filter(gene %in% m49_m60_g_ovlp) %>% mutate(don = "M49") %>% bind_rows(., m60_strn_d_markers_df %>% filter(gene %in% m49_m60_g_ovlp) %>% mutate(don = "M60")) %>% ggplot(aes(x = avg_log2FC, y = Gene, fill = don)) + geom_col() 
```


```{r}
# M48, M49, M60
m48_g <- m48_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)
m49_g <- m49_strn_d_markers_antd %>% filter(p_val_adj < 0.05) %>% pull(gene)
m60_g <- m60_strn_d_markers_df %>% filter(p_val_adj < 0.05) %>% pull(gene)


m48_m60_g_ovlp <- m48_g[m48_g %in% m60_g]
m49_m60_g_ovlp <- m49_g[m49_g %in% m60_g]
m48_m49_g_ovlp <- m48_g[m48_g %in% m49_g]

```

```{r, fig.width=4, fig.height=2}
# M48, M49, M60

m49_strn_d_markers_antd %>% filter(gene %in% m48_m49_g_ovlp) %>% mutate(don = "M49") %>% bind_rows(., m48_strn_d_markers_df %>% filter(gene %in% m48_m49_g_ovlp) %>% mutate(don = "M48")) %>% ggplot(aes(x = avg_log2FC, y = Gene, fill = don)) + geom_col() 
```
```{r}
# M48, M49, M60

m49_strn_d_markers_antd %>% filter(gene %in% m48_m49_g_ovlp) %>% mutate(don = "M49") %>% bind_rows(., m48_strn_d_markers_df %>% filter(gene %in% m48_m49_g_ovlp) %>% mutate(don = "M48")) %>% ggplot(aes(x = avg_log2FC, y = Gene, fill = don)) + geom_col() 
```

```{r}
Idents(m48_asex) <- "StrainDon"
w48_markrs <- FindAllMarkers(m48_asex)

```

```{r}
m48_asex <- msc_w_cln_strns[["MSC48"]][,msc_w_cln_strns[["MSC48"]]@meta.data$stage %in% c("Early trophozoite", "Late ring")]

```

```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
map(v3_afm_harmony_ss_res_fld_pt, ~FeaturePlot(.x, features = "scf_class_rand_dbt_prop", reduction = "umap.harmony_PC_man", order = T))

```


```{r, fig.width=15, fig.height=3}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
FeaturePlot(harmony_int, features = c("PF3D7-0204800", "PF3D7-0204900", "PF3D7-0113000", "PF3D7-0316600", "PF3D7-0220000"), reduction = "umap.harmony_PC_man", order = T, ncol = 5)

```


```{r, fig.width=15, fig.height=3}
## 
VlnPlot(v3_afm_harmony_ss_res_fld_pt, features = c("PF3D7-0204800", "PF3D7-0204900", "PF3D7-0113000", "PF3D7-0316600", "PF3D7-0220000"), ncol = 5)

```
```{r, fig.width=6, fig.height=3}
## 
VlnPlot(v3_afm_harmony_ss_res_fld_pt, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2, split.by = "stage", pt.size = 0)

```

```{r, fig.width=18, fig.height=3}
## 
VlnPlot(v3_afm_harmony_ss_res_fld_pt, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2, split.by = "StrainDon", pt.size = 0)

```

```{r, fig.width=6, fig.height=3}
## 
map(v3_afm_harmony_ss_res_fld_pt, ~DimPlot(.x,  split.by = "dset", reduction = "umap.harmony_PC_man", dims = c(2,3)))

```

```{r, fig.width=6, fig.height=3}
## 
map(v3_afm_harmony_ss_res_fld_pt, ~DimPlot(.x,  split.by = "dset", reduction = "umap.harmony_PC_man", dims = c(2,3)))

```

```{r, fig.width=15, fig.height=3}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
FeaturePlot(harmony_int, features = c("PF3D7-0204800", "PF3D7-0204900", "PF3D7-0113000", "PF3D7-0316600", "PF3D7-0220000"), reduction = "umap.harmony_PC_man", order = T, ncol = 5)

```



```{r, fig.width=18, fig.height=6}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
VlnPlot(harmony_int[, harmony_int@meta.data$StagePrelimC == "Asexual"], features = c("PF3D7-0204800", "PF3D7-0204900", "PF3D7-0113000", "PF3D7-0316600", "PF3D7-0220000"), ncol = 5, pt.size = 0, group.by = "harmony_clusters")

```

```{r, fig.width=12, fig.height=6}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
FeaturePlot(harmony_int[, harmony_int@meta.data$StagePrelimC == "Asexual"], features = c("PF3D7-0204800", "PF3D7-0204900"), order = T, ncol = 2, pt.size = 1, reduction = "umap.harmony_PC_man")

```

```{r, fig.width=18, fig.height=6}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
harmony_int <- harmony_int@meta.data %>% rownames_to_column('bcode') %>% mutate(wd_cluster = case_when(bcode %in% cbs ~ "weird_afm", T ~ stage)) %>% select(bcode, wd_cluster) %>% column_to_rownames('bcode') %>% AddMetaData(harmony_int, .)


```

```{r, fig.width=18, fig.height=6}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
VlnPlot(harmony_int[, harmony_int@meta.data$StagePrelimC == "Asexual"], features = c("PF3D7-0204800", "PF3D7-0204900", "PF3D7-0113000", "PF3D7-0316600", "PF3D7-0220000"), ncol = 5, pt.size = 0, group.by = "wd_cluster")
VlnPlot(harmony_int, features = c("PF3D7-0204800", "PF3D7-0204900", "PF3D7-0113000", "PF3D7-0316600", "PF3D7-0220000"), ncol = 5, pt.size = 0, group.by = "wd_cluster")

```


```{r, fig.width=12, fig.height=6}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
DotPlot(harmony_int, features = c("PF3D7-0204800", "PF3D7-0204900"), group.by = "stage")

```



## Assess distribution of strains over pseudotime

```{r, results="asis", fig.width=5, fig.height=3.5}
library("ggpubr")

smpl = umap_meta_df %>% 
  filter(StrainDon %in% c("M49_SC1", "M49_SC5")) %>% 
  count(StrainDon) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(StrainDon %in% c("M49_SC1", "M49_SC5")) %>% 
  group_by(StrainDon) %>% 
  slice_sample(n = smpl) 

tbl%>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.3), size=2, bw = 0.3, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  dens_thme +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="StrainDon", col="StrainDon") + 
  stat_compare_means(method = "wilcox.test") +gbox_theme
  


```



```{r, results="asis", fig.width=5, fig.height=3.5}
smpl = umap_meta_df %>% 
  filter(StrainDon %in% c("M60_SC3", "M60_SC7")) %>% 
  count(StrainDon) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(StrainDon %in% c("M60_SC3", "M60_SC7")) %>% 
  group_by(StrainDon) %>% 
  slice_sample(n = smpl) 

tbl %>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.3), size=2, bw = 0.3, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="StrainDon", col="StrainDon") + stat_compare_means(method = "wilcox.test")+gbox_theme


```


```{r, results="asis", fig.width=5, fig.height=3.5}
smpl = umap_meta_df %>% 
  filter(StrainDon %in% c("M48_SC2", "M48_SC4")) %>% 
  count(StrainDon) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(StrainDon %in% c("M48_SC2", "M48_SC4")) %>% 
  group_by(StrainDon) %>% 
  slice_sample(n = smpl) 

tbl %>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.48), size=2, bw = 0.48, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))


tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="StrainDon", col="StrainDon") + stat_compare_means(method = "wilcox.test")+gbox_theme


```


```{r, results="asis", fig.width=5, fig.height=4}
smpl = umap_meta_df %>% 
  filter(StrainDon %in% c("M55_SC6", "M55_SC8")) %>% 
  count(StrainDon) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(StrainDon %in% c("M55_SC6", "M55_SC8")) %>% 
  group_by(StrainDon) %>% 
  slice_sample(n = smpl) 

tbl%>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.48), size=2, bw = 0.48, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))



tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="StrainDon", col="StrainDon") + stat_compare_means(method = "wilcox.test")+gbox_theme

```


```{r, results="asis", fig.width=5, fig.height=4}
smpl = umap_meta_df %>% 
  filter(StrainDon %in% c("M66_SC5", "M66_SC7")) %>% 
  count(StrainDon) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(StrainDon %in% c("M66_SC5", "M66_SC7", "M66_SC8")) %>% 
  group_by(StrainDon) %>% 
  slice_sample(n = smpl) 

tbl%>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.48), size=2, bw = 0.48, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))


tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="StrainDon", col="StrainDon") +stat_compare_means(method = "anova", aes(label = ..p.signif..))+gbox_theme

```


```{r, results="asis"}
smpl = umap_meta_df %>% 
  filter(StrainDon %in% c("M33_SC6", "M33_SC7")) %>% 
  count(StrainDon) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(StrainDon %in% c("M33_SC2","M33_SC5", "M33_SC6", "M33_SC7")) %>% 
  group_by(StrainDon) %>% 
  slice_sample(n = smpl) 

tbl%>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.48), size=2, bw = 0.48, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))


tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="StrainDon", col="StrainDon") +stat_compare_means(method = "kruskal.test", aes(label = ..p.signif..))+gbox_theme

```

```{r, results="asis"}
smpl = umap_meta_df %>% 
  filter(!is.na(Symptomatic)) %>% 
  count(Symptomatic) %>% pull(n) %>% min()

tbl = umap_meta_df %>% 
  filter(!is.na(Symptomatic)) %>% 
  group_by(Symptomatic) %>% 
  slice_sample(n = smpl) 

tbl%>%
  ggplot(., aes(x = slingPseudotime_1, col=Symptomatic, fill=Symptomatic)) + 
  geom_density(aes(y = ..count..*0.48), size=2, bw = 0.48, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

tbl %>% 
  ggboxplot(., y = "slingPseudotime_1",x="Symptomatic", col="Symptomatic") +stat_compare_means(method = "wilcox.test")


```



## MAST DE of interesting clusters and also for strains


```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
# DimPlot(v3_afm_harmony_ss_res_fld_pt, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_ss", pt.size = 0.5, dims = c(1,2), label = T)

```

```{r, fig.width=30, fig.height=6}
# FeaturePlot(v3_afm_harmony_ss_res_fld_pt, features = "PF3D7-0925700", reduction = "umap.harmony_PC_man", split.by = "donor", pt.size = 2, ncol = 7)
```

```{r}
## Weird cluster 2 markers
# v3_afm_harmony_ss_res_fld_pt@meta.data %>% filter(harmony_clusters_asex == "2") %>% count(harmony_clusters_asex, donor)

```


```{r}
## Weird cluster 2 markers
# v3_afm_harmony_ss_res_fld_pt@meta.data %>% filter(harmony_clusters_asex == "2") %>% count(harmony_clusters_asex, StrainDon) %>% arrange(desc(n))

```


```{r}
## Weird cluster 2 markers
# v3_afm_harmony_ss_res_fld_pt@meta.data %>% filter(donor %in% c("MSC48", "MSC49")) %>% count(harmony_clusters_asex, StrainDon) %>% arrange(desc(n))

```



```{r}
## Weird cluster 2 markers
# v3_afm_harmony_ss_res_fld_pt@meta.data %>% filter(dset != "Lab" & str_detect(StrainDon, "Doublet|Negative")) %>% count(donor, Symptomatic, StrainDon)
# 
```


```{r}
## Weird cluster 2 markers
# v3_afm_harmony_ss_res_fld_pt@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, Symptomatic, StrainDon) %>% filter(n>20) %>% group_by(donor) %>% filter(n() > 1 )

```



```{r}
m49_afm_seu_pt <- m49_afm_seu %>% AddMetaData(., umap_meta_df %>% filter(orig.ident == "MSC49.mrna") %>% mutate(across(cell_bc, ~str_remove(., "M[0-9]*_")))%>% select(cell_bc, slingPseudotime_1) %>% column_to_rownames("cell_bc") %>% filter(!is.na(slingPseudotime_1)))

m49_afm_seu_pt <- m49_afm_seu_pt[,!is.na(m49_afm_seu_pt@meta.data$slingPseudotime_1)]

Idents(m49_afm_seu_pt) <- "StrainDon"

m49_afm_seu_pt@meta.data %>% count(StrainDon)
m49_strn_d_markers_pt <- FindMarkers(object = m49_afm_seu_pt, test.use = "MAST", ident.1 = "M49_SC1", ident.2 = "M49_SC5", latent.vars = "slingPseudotime_1")
m49_strn_d_markers_antd_pt <- m49_strn_d_markers_pt  %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))

m49_strn_d_markers_antd_pt %>% filter(p_val_adj < 1e-20)

## vps3 extracellular vessicle content https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8764524/
## ΔHsp70x - virulence cytoadhesion - dispensable 
##VAR_PF3D7-1041300: Mapping immune variation and var gene switching in naive hosts infected with Plasmodium falciparum - PMC.html
##CBP2 ΔGEXP07 parasites have a growth advantage compared to wild-type parasites https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7078486/
##CBP2 ΔGEXP07 parasites have a growth advantage compared to wild-type parasites https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7078486/
## MAHRP2 interacts with pfemp1 rers cleft  https://www.nature.com/articles/ncomms16044, https://journals.asm.org/doi/10.1128/msphere.00755-21
```


```{r}
m60_afm_seu_pt <- m60_afm_seu %>% AddMetaData(., umap_meta_df %>% filter(orig.ident == "MSC60.mrna") %>% mutate(across(cell_bc, ~str_remove(., "M[0-9]*_")))%>% select(cell_bc, slingPseudotime_1) %>% column_to_rownames("cell_bc") %>% filter(!is.na(slingPseudotime_1)))

m60_afm_seu_pt <- m60_afm_seu_pt[,!is.na(m60_afm_seu_pt@meta.data$slingPseudotime_1)]

Idents(m60_afm_seu_pt) <- "StrainDon"

m60_afm_seu_pt@meta.data %>% count(StrainDon)
m60_strn_d_markers_pt <- FindMarkers(object = m60_afm_seu_pt, test.use = "MAST", ident.1 = "M60_SC3", ident.2 = "M60_SC7", latent.vars = "slingPseudotime_1")
m60_strn_d_markers_antd_pt <- m60_strn_d_markers_pt  %>% rownames_to_column("gene") %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene), by = c("gene"="gene_id"))

m60_strn_d_markers_antd_pt %>% filter(p_val_adj < 1e-5)

```

```{r, results="asis"}
plt = umap_meta_df %>% 
  filter(StrainDon %in% c("M60_SC1", "M60_SC5")) %>%
  ggplot(., aes(x = slingPseudotime_1, col=StrainDon, fill=StrainDon)) + 
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)

```

Explore particular clusters 


```{r}
## Weird cluster 2 markers
Idents(v3_afm_harmony_ss_res_fld_pt) <- "harmony_clusters_asex"
cl2_markrs <- FindMarkers(v3_afm_harmony_ss_res_fld_pt, ident.1 = "2")

```

```{r}
## Weird cluster 2 markers
cl2_markrs %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene,`Genomic Location (Gene)`))

## PF3D7-0204900 https://www.sciencedirect.com/science/article/pii/S0006291X20321744#abs0010
## PF3D7-0204900 https://academic.oup.com/femsre/article/30/4/596/2367720
## Possible copy number variant since they are in the same location in the genome

## GARP - antigarp induces pf death and present in relatively resistant individuals https://pubmed.ncbi.nlm.nih.gov/32427965/
## GARP high expression in K13580Y mutant as well as pfemp1 - https://www.biorxiv.org/content/biorxiv/early/2023/12/06/2023.12.06.570387.full.pdf 

## FNT Drug target - due to its housekeeping role in lactate efflux during the intraerythrocytic stage https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8428694/

## FNT Drug target - due to its housekeeping role in lactate efflux during the intraerythrocytic stage https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8428694/

## PfA-M17 essential gene and drug target (compound 3) implicated in the final stage of hemoglobin digestion - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9470162/ https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1010926 

##PF3D7-1223900 Contained in drug resistance gene duplication https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7774857/ https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-019-0708-9, https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9948948/

```

```{r}
## Expression of AF1 genes in unique cluster - https://www.biorxiv.org/content/10.1101/2024.01.20.576496v1.full.pdf
cl2_markrs %>% rownames_to_column('gene_id') %>% mutate(across(gene_id, ~str_replace(., "-", "_"))) %>% filter(gene_id %in% c("PF3D7_1035500", "PF3D7_1035600", "PF3D7_1035400", "PF3D7_1035300", "PF3D7_0930300", "PF3D7_0207600", "PF3D7_0207500", "PF3D7_1149200", "PF3D7_0401800", "PF3D7_1370300", "PF3D7_1149000", "PF3D7_0935900", "PF3D7_0936000", "PF3D7_1039000", "PF3D7_1149300", "PF3D7_0424400", "PF3D7_0830800", "PF3D7_1477600", "PF3D7_0935800")) %>% mutate(across(gene_id, ~str_replace(., "_", "-"))) %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene,`Genomic Location (Gene)`))

```

```{r}
## Weird cluster 2 markers
map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2") %>% count(harmony_clusters_asex, donor))
map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2") %>% count(seurat_clusters, donor, Symptomatic))
```

```{r}
## Weird cluster 2 markers
harmony_int <- readRDS("data/processed/Pf/m2_all/harmony_int.RDS")
```

```{r}
## Weird cluster 2 markers
harmony_int@meta.data %>% filter(harmony_clusters == "2") %>% count(harmony_clusters, seurat_clusters, donor)
map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2") %>% count(harmony_clusters_asex, seurat_clusters, donor))
```


```{r}
## Weird cluster 2 markers
harmony_int@meta.data %>% filter(harmony_clusters == "2") %>% count(harmony_clusters, seurat_clusters, donor)
map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2") %>% count(harmony_clusters_asex, seurat_clusters, donor, Strain_DN) %>% arrange(desc(n)))
```

```{r}
## Weird cluster 2 markers
cbs<- map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2") %>% rownames())

harmony_int@meta.data %>% rownames_to_column("bcode") %>% filter(bcode %in% cbs) %>% count(harmony_clusters, seurat_clusters, donor) %>% arrange(desc(n))

```

```{r}

DimPlot(harmony_int[,harmony_int@meta.data$StagePrelimC == "Asexual"], reduction = "umap.harmony_PC_man", cells.highlight = cbs, pt.size = 0.5, dims = c(1,2), label = F)
map(v3_afm_harmony_ss_res_fld_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man",cells.highlight = cbs, pt.size = 0.5, dims = c(1,2), label = F))

cbs_49 <- map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2" & donor == "MSC49") %>% rownames() %>% str_remove(., "M49_"))
DimPlot(msc_w_cln_strns[["MSC49"]], cells.highlight = cbs_49, pt.size = 0.5, dims = c(1,2), label = F)

cbs_48 <- map(v3_afm_harmony_ss_res_fld_pt, ~.x@meta.data %>% filter(harmony_clusters_asex == "2") %>% rownames() %>% str_remove(., "M48_"))
DimPlot(msc_w_cln_strns[["MSC48"]], cells.highlight = cbs_48, pt.size = 0.5, dims = c(1,2), label = F)

```

