---
title: "MSC QC cbender_comparisons"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(scran)
  library(SingleR)
  library(scDblFinder)
  library(ggsankey)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(dplyr::desc)
  conflicts_prefer(dplyr::count)
})
```



# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/project_plotting_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")

```


###################START- RAW READING#############

```{r, eval=T}
pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
```

## Genotype clustering for MSC

initialize variable for the sample names



```{r}

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
# slct_sample_names <-c("MSC68", "MSC14", "MSC49", "MSC55") %>% str_sort(., numeric = T)


# ## poor quality samples
# sample_rm <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39", "MSC50_SLO", "MSC33")
sample_rm <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39")

sample_mxd <- pf_solo_mixed_mdata_decode %>% filter(mixed_infection2 == "mixed") %>% pull(sample_nm)
# 
# # slct_sample_names[-c(sample_rm)]
slct_sample_names <- slct_sample_names[!(slct_sample_names %in% c(sample_rm, sample_mxd))]

## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```


```{r, message=F, warning=F, eval=F}
## read filtered barcodes
barcodes <- list.files("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/", pattern = "barcodes.tsv.gz$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes//|outs/filtered.*'))) %>% 
  .[source_irods] %>%
  set_names(names(source_irods)) %>%
  read_tsv()
```


```{r, message=F, warning=F, eval=F}
## Get the number of cells that pass cellranger filter
crangr_filt_num <- system("zgrep -c '' /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz", intern = TRUE) 
```


```{r, message=F, warning=F, eval=F}
## Df of cellranger pass cells
crangr_filt_df <- data.frame('irods_id' = str_extract(crangr_filt_num, "5736STDY[0-9]+"),
                             'cranger_filt_n' = str_extract(crangr_filt_num, "[0-9]+$"))

pf_mdata_decode_cranger_filt_n <- pf_solo_mixed_mdata_decode %>% full_join(crangr_filt_df, by = "irods_id")

pf_mdata_decode_cranger_filt_n %>% write_csv(., "data/raw/pf_mdata_decode_cranger_filt_n.csv")
```



```{r, fig.width=15, fig.height=5}
donor_10x_mdata_sk21_summ <- readRDS("data/processed/Pf/m2_all/donor_10x_mdata_sk21_summ.RDS") %>%
  filter(donor %in% slct_sample_names) 

```


```{r, fig.width=15, fig.height=5}

donor_10x_mdata_sk21_summ %>% count(frac_load, overload, frac_over, targeted_cells)
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_rmHs <- paste0(sample_name, "_rm_HS")
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_yascpCranger <- paste0(sample_name, "_yascpCr")
```

```{r, message=F, warning=F, eval=T}
## Names for cellbender with learning rate of 0.000025
sample_name_LRate_mod <- paste0(sample_name, "_LRpt000025")
```


```{r, message=F, warning=F, eval=T}
sample_name_LR_mod <- paste0(sample_name, "_LRpt000005e250")
sample_name_LR <- paste0(sample_name, "_LRpt0001")
sample_name_pf_LR_mod <- paste0(sample_name, "_rmHS_LRpt000005e250")
sample_name_pf <- paste0(sample_name, "_rmHS_LRpt0001")
```


```{r, message=F, warning=F, eval=T}
# sample_name_LR_mod <- paste0(sample_name, "_LRpt000005e250")
# sample_name_LR <- paste0(sample_name, "_LRpt0001")
# sample_name_pf_LR_mod <- paste0(sample_name, "_rmHS_LRpt000005e250")
# sample_name_pf <- paste0(sample_name, "_rmHS_LRpt0001")

sample_name_LR_mod <- paste0(sample_name, "_yascp_default_pfhs_dir")
sample_name_LR <- paste0(sample_name, "_LRpt0001")
sample_name_pf_LR_mod <- paste0(sample_name, "_yascp_default_pf_dir")
sample_name_pf <- paste0(sample_name, "_rmHS_LRpt0001")

```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
cbender_mdata_std_yascp_syst <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*|/'))) %>% 
  # set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|/cb_yascp_default|_dir|_gns|/cb_mdata.*'))) %>% 
  .[c(sample_name_LR, sample_name_LR_mod, sample_name_pf, sample_name_pf_LR_mod)] %>%
  .[!is.na(.)] %>%
  map(read_csv) 
```

#########################START - Read in MSC68 manual (expected cells = 10,000 and expected droplets = 50,000) #########################

```{r, message=F, warning=F, eval=T}
sample_name_LR_mod_man_msc68 <- paste0("MSC68", "_man_LRpt000005e250")
sample_name_LR_msc68_man_msc68 <- paste0("MSC68", "_rmHS_man_LRpt000005e250")
```

```{r, message=F, warning=F, eval=T}
##Read cellbender output
cbender_mdata_std_yascp_syst_msc68 <- list.files("data/processed/Pf/MSC68", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*'))) %>% 
  .[c(sample_name_LR_mod_man_msc68, sample_name_LR_msc68_man_msc68)] %>%
  .[!is.na(.)] %>%
  map(read_csv)

```

## Replace MSC68 yascp learning rate and epochs settings with manual
```{r, message=F, warning=F, eval=T}
map(cbender_mdata_std_yascp_syst_msc68, ~.x %>% count(cell_probability > 0.5) )

```

## Replace MSC68 yascp learning rate and epochs settings with manual
```{r, message=F, warning=F, eval=T}
map(cbender_mdata_std_yascp_syst_msc68, ~.x %>% count(cell_probability > 0.5) )

```


## !!!!NOTE Replace MSC68 yascp learning rate and epochs settings with manual
```{r, message=F, warning=F, eval=F}
cbender_mdata_std_yascp_syst[["MSC68_LRpt000005e250"]] <- cbender_mdata_std_yascp_syst_msc68[["MSC68_man_LRpt000005e250"]]
cbender_mdata_std_yascp_syst[["MSC68_rmHS_LRpt000005e250"]] <- cbender_mdata_std_yascp_syst_msc68[["MSC68_rmHS_man_LRpt000005e250"]]
```


#########################START - Read in MSC68 manual (expected cells = 10,000 and expected droplets = 50,000) #########################


```{r, message=F, warning=F, eval=T}

cbender_std_yascpCr_mrg_syst <- pmap(
  list(cbender_mdata_std_yascp_syst[sample_name_LR_mod], 
       cbender_mdata_std_yascp_syst[sample_name_LR], 
       cbender_mdata_std_yascp_syst[sample_name_pf_LR_mod], 
       cbender_mdata_std_yascp_syst[sample_name_pf] ), ~{
       full_join(..1, ..2, by = "barcode", suffix = c("_lhs", "_yhs")) #%>%
         # full_join(., ..3 %>% rename_with(~ paste0(.x, "_lpf"), -all_of("barcode")), by = "barcode") %>%
         # full_join(., ..4 %>% rename_with(~ paste0(.x, "_ypf"), -all_of("barcode")), by = "barcode")
       }) 

names(cbender_std_yascpCr_mrg_syst) <- str_remove_all(names(cbender_std_yascpCr_mrg_syst), "_LRpt000005e250|_yascp_default_pfhs_dir|_yascp_default_pf_dir")
```
########################################################################### JUMP to where it says "PICK UP FROM HERE"

```{r, message=F, warning=F, eval=T}
gogo()
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
# cbender_mdata_std_yascp <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*'))) %>% 
#   .[c(sample_name, sample_name_rmHs, sample_name_yascpCranger, sample_name_LRate_mod)] %>%
#   .[!is.na(.)] %>%
#   map(read_csv) 
```

```{r, message=F, warning=F, eval=F}
all_cbender_mdata_mrg <- map2(cbender_mdata_std_yascp[sample_name],
                              cbender_mdata_std_yascp[sample_name_rmHs], ~{
                                full_join(.x, .y, by = "barcode", suffix = c("_wHs", "_rmHs"))
                                })

```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_yascpCranger_nms <- names(cbender_mdata_std_yascp)[str_detect(names(cbender_mdata_std_yascp), "yascpCr")]
sample_name_yascpCranger_nms <- sample_name_yascpCranger_nms[!str_detect(sample_name_yascpCranger_nms, "MSC50_SLO_yascpCr")]

```

```{r, message=F, warning=F, eval=T}
cbender_std_yascpCr_mrg <- pmap(
  list(cbender_mdata_std_yascp[str_remove(sample_name_yascpCranger_nms, "_yascpCr")],
       cbender_mdata_std_yascp[paste0(str_remove(sample_name_yascpCranger_nms, "_yascpCr"), "_rm_HS")],
       cbender_mdata_std_yascp[sample_name_yascpCranger_nms],
       cbender_mdata_std_yascp[paste0(str_remove(sample_name_yascpCranger_nms, "_yascpCr"), "_LRpt000025")]), ~{
       full_join(..1, ..2, by = "barcode", suffix = c("_wHs", "_rmHs")) %>%
         full_join(., ..3 %>% rename_with(~ paste0(.x, "_yCr"), -all_of("barcode")), by = "barcode") %>%
         full_join(., ..4 %>% rename_with(~ paste0(.x, "_lrt"), -all_of("barcode")), by = "barcode")
       }) 

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_mrg %>%
  bind_rows(., .id = "donor") %>%
  # filter(!(cell_probability_wHs == 1 & cell_probability_rmHs ==1)) %>%
  ggplot(aes(x=cell_probability_wHs, y = cell_probability_rmHs)) +
  geom_point(size = 0.0001) +
  geom_density2d() +
  facet_wrap(. ~ donor)

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by using the pf+Hs gene counts
all_cbender_mdata_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter((cell_probability_wHs < 0.9 & cell_probability_rmHs > 0.9) | (is.na(cell_probability_wHs)  & cell_probability_rmHs > 0.9)) %>%
  # filter(cell_size_rmHs >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by using the pf+Hs gene counts
all_cbender_mdata_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(cell_probability_wHs < 0.9 & cell_probability_rmHs > 0.9) %>%
  filter(cell_size_rmHs >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(!(cell_probability_wHs == 1 & cell_probability_rmHs ==1)) %>%
  ggplot(aes(x=cell_probability_wHs, y = cell_probability_rmHs)) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  theme_bw() +
  facet_wrap(. ~ donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_mrg %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(x=cell_size_wHs, y = cell_size_rmHs)) +
  geom_point(size = 0.01) +
  geom_abline() +
  facet_wrap(. ~ donor)

```

```{r, message=F, warning=F, eval=T}
cbender_mdata_std_yascp[[1]] %>% str

```

```{r, message=F, warning=F, eval=T}
cbender_mdata_std_yascp[[1]] %>% 
  pivot_longer()

```


```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_probability)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```


```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_probability)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```


```{r, message=F, warning=F, eval=T, fig.height=9, fig.width=7}
cbender_mdata_std_yascp %>%
  bind_rows(., .id = "donor") %>% 
      filter(cell_probability > 0.8) %>%
  group_by(donor) %>% 
  mutate(bg_50 = sum(background_fraction > 0.5)) %>% 
  ungroup() %>%
  ggplot(aes(x = background_fraction)) +
    geom_histogram() +
    geom_text(aes(label=bg_50, x = 0.8, y = 200),  size = 3)+
    geom_vline(xintercept = 0.5) +
    facet_wrap(donor ~ ., scales = "free_y", ncol = 5)

```

```{r, message=F, warning=F, eval=T, fig.height=8, fig.width=12}
cbender_mdata_std_yascp %>%
  bind_rows(., .id = "donor") %>% 
      filter(cell_probability > 0.8) %>%
  count(donor, background_fraction>0.5)

```

```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_size)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```

```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = cell_probability, y = cell_size)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.9) +
    labs(title = .y)
)

```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_cbender_mdata_pass.5 <- map(cbender_mdata_std_yascp, ~.x %>% filter(cell_probability>0.5) %>% pull(barcode))

```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_yascp <- paste0(sample_name, "_yascp")
```

```{r, message=F, warning=F, eval=T}
##Read output from cellbender standard pipeline and cellbender yascp pipeline
all_cbender_mdata_std_yascp <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*'))) %>% 
  .[c(sample_name, sample_name_yascp)] %>%
  .[!is.na(.)] %>%
  map(read_csv) 
```

```{r, message=F, warning=F, eval=T}
all_cbender_mdata_std_yascp_mrg <- map2(all_cbender_mdata_std_yascp[sample_name], all_cbender_mdata_std_yascp[sample_name_yascp], ~{
  full_join(.x, .y, by = "barcode", suffix = c("_std", "_yascp")) 
}) 

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  # filter(!(cell_probability_std == 1 & cell_probability_yascp ==1)) %>%
  ggplot(aes(x=cell_probability_std, y = cell_probability_yascp)) +
  geom_point(size = 0.0001) +
  geom_density2d() +
  facet_wrap(. ~ donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by standard cellbender
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(cell_probability_yascp > 0.9) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by standard cellbender
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(cell_probability_std > 0.9) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by standard cellbender
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  # filter((cell_probability_std < 0.9 & cell_probability_yascp > 0.9)) %>%
  filter((cell_probability_std < 0.9 & cell_probability_yascp > 0.9) | (is.na(cell_probability_std)  & cell_probability_yascp > 0.9)) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by yascp cellbender
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  # filter((cell_probability_std < 0.9 & cell_probability_yascp > 0.9)) %>%
  filter((cell_probability_yascp < 0.9 & cell_probability_std > 0.9) | (is.na(cell_probability_yascp)  & cell_probability_std > 0.9)) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by using the pf+Hs gene counts
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(cell_probability_std < 0.9 & cell_probability_yascp > 0.9) %>%
  filter(cell_size_yascp >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(!(cell_probability_std == 1 & cell_probability_yascp ==1)) %>%
  ggplot(aes(x=cell_probability_std, y = cell_probability_yascp)) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  theme_bw() +
  facet_wrap(. ~ donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_std_yascp_mrg %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(x=cell_size_std, y = cell_size_yascp)) +
  geom_point(size = 0.01) +
  geom_abline() +
  facet_wrap(. ~ donor)

```



```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_yascpCranger <- paste0(sample_name, "_yascpCr")
```

```{r, message=F, warning=F, eval=T}
##Read output from cellbender standard pipeline and cellbender yascp pipeline
all_cbender_mdata_std_yascpCr <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*'))) %>% 
  .[c(sample_name_yascp, sample_name_yascpCranger)] %>%
  .[!is.na(.)] %>%
  map(read_csv) 
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_yascpCranger_nms <- names(all_cbender_mdata_std_yascpCr)[str_detect(names(all_cbender_mdata_std_yascpCr), "yascpCr")]
```


```{r, message=F, warning=F, eval=T}
all_cbender_mdata_std_yascpCr_mrg <- map2(all_cbender_mdata_std_yascpCr[str_remove(sample_name_yascpCranger_nms, "Cr")], all_cbender_mdata_std_yascpCr[sample_name_yascpCranger_nms], ~{
  full_join(.x, .y, by = "barcode", suffix = c("_std", "_yascpCr")) 
}) 

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
all_cbender_mdata_std_yascpCr_mrg %>%
  bind_rows(., .id = "donor") %>%
  # filter(!(cell_probability_std == 1 & cell_probability_yascp ==1)) %>%
  ggplot(aes(x=cell_probability_std, y = cell_probability_yascpCr)) +
  geom_point(size = 0.0001) +
  geom_density2d() +
  facet_wrap(. ~ donor)

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by standard cellbender
all_cbender_mdata_std_yascpCr_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(cell_probability_yascpCr > 0.9) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by standard cellbender
all_cbender_mdata_std_yascpCr_mrg %>%
  bind_rows(., .id = "donor") %>%
  filter(cell_probability_std > 0.9) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```
########################################################################### PICK UP FROM here
```{r, message=F, warning=F, eval=T}
##Read output from cellbender standard pipeline and cellbender yascp pipeline
# cbender_cutoffs <- list.files("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/yascp_trial_dir/results/preprocessing/cellbender", pattern = "umi_count_estimates.*_cutoff.tsv.gz*", full.names = T, recursive = T, include.dirs = T) %>% 

cbender_cutoffs <- list.files("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/yascp_default_pf_dir/results/preprocessing/cellbender", pattern = "umi_count_estimates.*_cutoff.tsv.gz*", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'/lustre.*ing/cellbender/|/cellbender-.*estimates-|_cutoff.tsv.gz'))) %>% 
  # .[c(sample_name, sample_name_yascp)] %>%
  .[!is.na(.)] %>%
  map(read_tsv) 

```


```{r, message=F, warning=F, eval=T}
##Read output from cellbender standard pipeline and cellbender yascp pipeline
cbender_cutoffs_all <- cbender_cutoffs %>%
  bind_rows(., .id = "donor_cutoff") %>%
  mutate(across(method, ~str_remove(., "dropletutils::barcoderanks::")))

```


```{r, message=F, warning=F, eval=T}

cbender_cutoffs_all_ncells_ndrops <- cbender_cutoffs_all %>%
  filter(method %in% c('estimated_ncells', 'estimated_ndroplets', 'knee', 'inflection')) 

```

```{r, message=F, warning=F, eval=T}
cbender_cutoffs_all_ncells_ndrops %>% 
  filter(donor_cutoff == paste0("MSC1", "total_droplets"))  %>% 
  filter(method == "knee") %>% pull(n_cells)
```


```{r, message=F, warning=F, eval=T}
# cbender_cutoffs_all_ncells_ndrops %>% filter(str_match(donor_cutoff, "MSC12.*")) %>% filter(method == "knee") %>% pull(n_cells)
```

```{r, message=F, warning=F, eval=T}
##Read bcrank results matrices
all_bcrank <- list.files("data/processed/Pf", pattern = "bcrank_1.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/edrops/bcrank_1.*'))) %>% 
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(readRDS) 
```


```{r, message=F, warning=F, eval=T}
##Read edrops results matrices
map(c("knee", "inflection"), ~{
  knee_or_inlctn <- .x
  imap(all_bcrank, ~{
  hlines_df = data.frame(lims = c(cbender_cutoffs_all_ncells_ndrops %>% filter(donor_cutoff == paste0(.y, "cell_estimate")) %>% filter(method == knee_or_inlctn) %>% pull(umi_counts_cutoff),
                                  cbender_cutoffs_all_ncells_ndrops %>% filter(donor_cutoff == paste0(.y, "total_droplets")) %>% filter(method == knee_or_inlctn)  %>% pull(umi_counts_cutoff)),
                         labls = c("cell_estimate", "total_droplets"),
                         pos = rep(500, 2))
  print(hlines_df)
  
  .x %>%
    data.frame() %>%
  # filter(total < 500) %>% 
  filter(!(duplicated(rank))) %>%
  # rename("total_umi" = detected) %>%
  # mutate(category = case_when(total_umi <= 50 ~ "Below_Limit",
  #                             is_cell_ed == "CellEd" ~ "CellEd",
  #                             TRUE ~ "Non-Cell")) %>%
  ggplot(., aes(x = rank, y = total)) +
      geom_point(size = 0.3, alpha = 0.6) +
      # scale_colour_manual(values = rank_colours) +
      scale_x_log10() +
      scale_y_log10() +
    geom_hline(yintercept = hlines_df$lims) +
    labs(title = paste0(.y, knee_or_inlctn)) +
      geom_text(data = hlines_df, mapping = aes(x = pos, y = lims, label = labls), vjust = -0.2)

    })
})
```

```{r, message=F, warning=F, eval=T}
##Read edrops results matrices
imap(all_bcrank, ~{
  hlines_df = data.frame(lims = c(cbender_cutoffs_all_ncells_ndrops %>% filter(donor_cutoff == paste0(.y, "cell_estimate")) %>% filter(method == "estimated_ncells") %>% pull(n_cells),
                                  cbender_cutoffs_all_ncells_ndrops %>% filter(donor_cutoff == paste0(.y, "total_droplets")) %>% filter(method == "estimated_ndroplets") %>% pull(n_cells)),
                         labls = c("cell_estimate", "total_droplets"),
                         pos = rep(5, 2))
  # print(hlines_df)
  
  .x %>%
    data.frame() %>%
  # filter(total < 500) %>% 
  filter(!(duplicated(rank))) %>%
  # rename("total_umi" = detected) %>%
  # mutate(category = case_when(total_umi <= 50 ~ "Below_Limit",
  #                             is_cell_ed == "CellEd" ~ "CellEd",
  #                             TRUE ~ "Non-Cell")) %>%
  ggplot(., aes(x = rank, y = total)) +
      geom_point(size = 0.3, alpha = 0.6) +
      # scale_colour_manual(values = rank_colours) +
      scale_x_log10() +
      scale_y_log10() +
    geom_vline(xintercept = hlines_df$lims) +
    labs(title = .y) +
      geom_text(data = hlines_df, mapping = aes(y = pos, x = lims, label = labls), vjust = -0.2, angle = 90)+
  theme_classic()

    }) 

```


### Compare cellbender iterations and also with cellranger

```{r, message=F, warning=F, eval=T}
## Read Cellranger filtered barcodes
cranger_bcodes <- list.files("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/", pattern = "barcodes.tsv.gz$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes//|/outs/filtered_feature_bc_matrix.*'))) %>% 
  .[source_irods] %>%
  set_names(names(source_irods)) %>%
  # .[sample_name] %>%
  map(.,~read_tsv(., col_names = "bcode")) 

```


```{r, message=F, warning=F, eval=T}
cranger_bcodes <- map(cranger_bcodes, ~.x %>% mutate(is_cell_cr = "Cr"))

```


```{r}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_hmnz_syst_snky <- map2(cbender_std_yascpCr_mrg_syst, cranger_bcodes[str_remove(names(cbender_std_yascpCr_mrg_syst), "_LRpt000005e250")], ~.x %>%
    full_join(., .y, by = c("barcode"="bcode")) %>%
    mutate(cr = case_when(is_cell_cr == "Cr" ~ "Cell")) %>%
    mutate(across(starts_with("cell_probability_"), ~case_when(.x > 0.5  ~ "Cell", .x <= 0.5  ~ "nonCell"), .names = "{sub('^cell_probability_', 'cb_', .col)}"))
)
         
      
```


```{r}
## Read raw cellranger metadata with the sum of Human and Pf reads
all_msc_bp_raw_noHsGns_mdata <- readRDS("data/processed/Pf/m2_all/all_msc_bp_raw_noHsGns_mdata.RDS") %>% .[sample_name]
```


```{r, message=F, warning=F, eval=T}
cr_cbender_mdata_hmnz_syst_snky_gns <- map2(cr_cbender_mdata_hmnz_syst_snky[intersect(names(cr_cbender_mdata_hmnz_syst_snky), names(all_msc_bp_raw_noHsGns_mdata))],
                            all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_hmnz_syst_snky), names(all_msc_bp_raw_noHsGns_mdata))],
                            ~left_join(.x, 
                                       .y %>% rownames_to_column("barcode"), 
                                       by = "barcode")# %>%
                              # filter(pf_umi_count >= 100)
                            )

```

```{r, message=FALSE, warning=FALSE}

imap(cr_cbender_mdata_hmnz_syst_snky_gns, ~snk_pt_fn(., strns_2_plt = c("cr", "cb_yhs", "cb_lhs", "cb_ypf", "cb_lpf"), labls = c("cr", "cb_yhs", "cb_lhs", "cb_ypf", "cb_lpf"), bp_face = c("bold", "bold", "bold", "bold", "bold"), ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')))

```

```{r, message=FALSE, warning=FALSE}

imap(cr_cbender_mdata_hmnz_syst_snky_gns, ~snk_pt_fn(., strns_2_plt = c("cr", "cb_dhs", "cb_lhs", "cb_dpf", "cb_lpf"), labls = c("cr", "cb_dhs", "cb_lhs", "cb_dpf", "cb_lpf"), bp_face = c("bold", "bold", "bold", "bold", "bold"), ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')))


```




```{r}
sample_cbender_select <- c("MSC12_LRpt0001", "MSC13_LRpt0001", "MSC14_LRpt000005e250", "MSC3_LRpt000005e250", "MSC33_LRpt000005e250", "MSC40_LRpt000005e250", "MSC45_LRpt0001", "MSC48_LRpt000005e250", "MSC49_LRpt000005e250", "MSC50_MACS_LRpt000005e250", "MSC50_SLO_LRpt000005e250", "MSC53_LRpt0001", "MSC54_LRpt000005e250", "MSC55_LRpt000005e250", "MSC57_LRpt0001", "MSC60_LRpt000005e250", "MSC66_LRpt000005e250", "MSC67_LRpt000005e250", "MSC68_LRpt000005e250", "MSC70_LRpt0001")
```


```{r, message=F, warning=F, eval=T}
sample_cbender_select_ptns <- ifelse(str_detect(sample_cbender_select, "LRpt000005e250"), c("lhs|lpf"), c("dhs|dpf")) %>% set_names(str_remove(sample_cbender_select, "_LR.*"))
```

```{r, message=F, warning=F, eval=T}
cbender_std_yascpCr_mrg_syst_slct <- map2(cbender_std_yascpCr_mrg_syst,  sample_cbender_select_ptns[names(cbender_std_yascpCr_mrg_syst)], ~.x %>%
       select(barcode, contains(str_remove(.y, "\\|.*")), contains(str_remove(.y, ".*\\|"))) %>%
         rename_with(~ str_replace_all(.x, c("lhs"="hs", "dhs"="hs", "lpf"="pf", "dpf"="pf")), -all_of("barcode"))
       )
```


```{r, message=F, warning=F, eval=T}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_slct_hmnz <- map2(cbender_std_yascpCr_mrg_syst_slct, cranger_bcodes[names(cbender_std_yascpCr_mrg_syst_slct)], ~.x %>%
    mutate(#is_cell_cr = case_when(barcode %in% .y$bcode ~ "Cr"),
           is_cell_cb_hs = case_when(cell_probability_hs> 0.5  ~ "Cbhs", cell_probability_hs <= 0.5  ~ "nonCell"),
           is_cell_cb_pf = case_when(cell_probability_pf > 0.5  ~ "Cbpf", cell_probability_pf <= 0.5  ~ "nonCell")
           ) %>%
      full_join(., .y, by = c("barcode"="bcode")))
         
      
```

```{r}
go()
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_hmnz <- map2(cbender_std_yascpCr_mrg, cranger_bcodes[names(cbender_std_yascpCr_mrg)], ~.x %>%
    mutate(#is_cell_cr = case_when(barcode %in% .y$bcode ~ "Cr"),
           is_cell_cb_wHs = case_when(cell_probability_wHs > 0.5  ~ "Cbhs", cell_probability_wHs <= 0.5  ~ "nonCell"),
           is_cell_cb_rmHs = case_when(cell_probability_rmHs > 0.5  ~ "Cbpf", cell_probability_rmHs <= 0.5  ~ "nonCell"),
           is_cell_cb_yCr = case_when(cell_probability_yCr > 0.5  ~ "Cbycr", cell_probability_yCr <= 0.5  ~ "nonCell"),
           is_cell_cb_lrt = case_when(cell_probability_lrt > 0.5  ~ "Cblrt", cell_probability_lrt <= 0.5  ~ "nonCell")
           )  %>%
      full_join(., .y, by = c("barcode"="bcode")))
         
      
```

```{r}
go()
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_hmnz <- map2(cbender_std_yascpCr_mrg, cranger_bcodes[names(cbender_std_yascpCr_mrg)], ~.x %>%
    mutate(#is_cell_cr = case_when(barcode %in% .y$bcode ~ "Cr"),
           is_cell_cb_wHs = case_when(cell_probability_wHs > 0.5  ~ "Cbhs", cell_probability_wHs <= 0.5  ~ "nonCell"),
           is_cell_cb_rmHs = case_when(cell_probability_rmHs > 0.5  ~ "Cbpf", cell_probability_rmHs <= 0.5  ~ "nonCell"),
           is_cell_cb_yCr = case_when(cell_probability_yCr > 0.5  ~ "Cbycr", cell_probability_yCr <= 0.5  ~ "nonCell"),
           is_cell_cb_lrt = case_when(cell_probability_lrt > 0.5  ~ "Cblrt", cell_probability_lrt <= 0.5  ~ "nonCell")
           ) %>%
      full_join(., .y, by = c("barcode"="bcode")))
         
      
```


```{r}
## Read raw cellranger metadata with the sum of Human and Pf reads
all_msc_bp_raw_noHsGns_mdata <- readRDS("data/processed/Pf/m2_all/all_msc_bp_raw_noHsGns_mdata.RDS") %>% .[sample_name]
```

```{r, message=F, warning=F, eval=T}
cr_cbender_mdata_slct_hmnz_gns <- map2(cr_cbender_mdata_slct_hmnz[intersect(names(cr_cbender_mdata_slct_hmnz), names(all_msc_bp_raw_noHsGns_mdata))],
                            all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_slct_hmnz), names(all_msc_bp_raw_noHsGns_mdata))],
                            ~left_join(.x, 
                                       .y %>% rownames_to_column("barcode"), 
                                       by = "barcode")
                            )

```


```{r, message=F, warning=F, eval=T}
go()
cr_cbender_mdata_hmnz_gns <- map2(cr_cbender_mdata_hmnz[intersect(names(cr_cbender_mdata_hmnz), names(all_msc_bp_raw_noHsGns_mdata))],
                            all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_hmnz), names(all_msc_bp_raw_noHsGns_mdata))],
                            ~left_join(.x, 
                                       .y %>% rownames_to_column("barcode"), 
                                       by = "barcode")
                            )

```

## Sankey plots of cell calls

```{r, message=FALSE, warning=FALSE}
cr_cbender_mdata_uni_sanky <- map(cr_cbender_mdata_hmnz_gns, ~.x %>% 
      filter(pf_umi_count >= 100) %>% 
      select(barcode, contains("is_cell")) %>%
      mutate(across(-c(barcode), ~ifelse(str_detect(.x, "Cr|Cb"), "Cell", .x))) %>%
      rename_with(~ str_remove(.x, "is_cell_"), -all_of("barcode"))
)
  
```


```{r, message=FALSE, warning=FALSE}

imap(cr_cbender_mdata_uni_sanky, ~snk_pt_fn(., strns_2_plt = c("cr", "cb_wHs", "cb_rmHs", "cb_yCr", "cb_lrt"), labls = c("cr", "cb_wHs", "cb_rmHs", "cb_yCr", "cb_lrt"), bp_face = c("bold", "bold", "bold", "bold", "bold"), ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')))
  
```

```{r}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
# cr_cbender_mdata_uni <- map(cr_cbender_mdata_hmnz, ~.x %>%
cr_cbender_mdata_slct_uni <- map(cr_cbender_mdata_slct_hmnz_gns, ~.x %>%
    mutate(mthds_union = paste0(is_cell_cr, is_cell_cb_hs, is_cell_cb_pf),
           mthds_union = case_when(str_detect(mthds_union, "Cb|Cr") ~ str_remove_all(mthds_union, "nonCell|NA"),
                                   !str_detect(mthds_union, "Cb|Cr") ~ str_remove_all(mthds_union, "NA"),
                                   T ~ mthds_union),
           mthds_union = case_when(str_detect(mthds_union, "nonCell") ~  "nonCell",
                                   T ~ mthds_union)
           ))
         
      
```

```{r}
go()
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_uni <- map(cr_cbender_mdata_hmnz, ~.x %>%
    # filter() %>%
    mutate(mthds_union = paste0(is_cell_cr, is_cell_cb_wHs, is_cell_cb_rmHs, is_cell_cb_yCr),
           mthds_union = case_when(str_detect(mthds_union, "Cb|Cr") ~ str_remove_all(mthds_union, "nonCell|NA"),
                                   !str_detect(mthds_union, "Cb|Cr") ~ str_remove_all(mthds_union, "NA"),
                                   T ~ mthds_union),
           mthds_union = case_when(str_detect(mthds_union, "nonCell") ~  "nonCell",
                                   T ~ mthds_union)
           ))
         
      
```

```{r}
## Convert mthds_union to factor
## Get all possible combinations of factors factoring in the order in which the variables were introduced
c_mthds <- c("Cr", "Cbhs", "Cbpf", "Cbycr")
c_combos <- map(1:length(c_mthds), ~ combn(c_mthds, .x, paste0, collapse = "")) %>% unlist() %>% dput()
c_combos <- c(rev(c_combos), "nonCell")

set.seed(12)
c_combos_cols <- sample(c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#fabebe', '#008080', '#e6beff', '#9a6324',  '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075'), length(c_combos)) %>% set_names(c_combos)

c_combos_cols <- ifelse(names(c_combos_cols) == "nonCell", "lightgrey", c_combos_cols)  %>% set_names(c_combos)

```


```{r}
go()
## Convert mthds_union to factor

cr_cbender_mdata_uni <- map(cr_cbender_mdata_uni, ~.x %>%
    mutate(across(mthds_union, ~factor(., levels = c_combos)))
)

```

```{r}
## Convert mthds_union to factor
cr_cbender_mdata_slct_uni <- map(cr_cbender_mdata_slct_uni, ~.x %>%
    mutate(across(mthds_union, ~factor(., levels = c_combos)))
)

```

```{r}
## Convert mthds_union to factor
# cr_cbender_mdata_uni <- map2(cbender_std_yascpCr_mrg, cranger_bcodes[names(cbender_std_yascpCr_mrg)], ~.x %>%
#     mutate(is_cell_cr = case_when(barcode %in% .y$bcode ~ "CellCr"),
#            is_cell_cb_wHs = case_when(cell_probability_wHs > 0.5  ~ "CellCb", cell_probability_wHs <= 0.5  ~ "nonCell"),
#            is_cell_cb_rmHs = case_when(cell_probability_rmHs > 0.5  ~ "CellCb", cell_probability_rmHs <= 0.5  ~ "nonCell"),
#            is_cell_cb_yCr = case_when(cell_probability_yCr > 0.5  ~ "CellCb", cell_probability_yCr <= 0.5  ~ "nonCell")) %>%
#     mutate(mthds_union = case_when(is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" & is_cell_cb_rmHs == "CellCb" & is_cell_cb_yCr == "CellCb" ~ "CrCbpfCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" ~ "CrCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_rmHs == "CellCb" ~ "CrCbpf",
#                                    is.na(is_cell_cr) & is_cell_cb_rmHs == "CellCb" & is_cell_cb_wHs == "CellCb" ~ "CbpfCbhs",
#                                    is_cell_cr == "CellCr" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs))  ~ "Cr",
#                                    is.na(is_cell_cr)  & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) & is_cell_cb_wHs == "CellCb" ~ "Cbhs",
#                                    is.na(is_cell_cr)  & is_cell_cb_rmHs == "CellCb" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) ~ "Cbpf",
#                                    is.na(is_cell_cr) & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) ~ "Cbnoncell"),
#            across(mthds_union, ~factor(., levels = c("CrCbpfCbhs", "CrCbpf", "CrCbhs", "CbpfCbhs", "Cr", "Cbpf", "Cbhs","Cbnoncell"))))
# )


```

```{r}
# cr_cbender_mdata_uni <- map2(all_cbender_mdata_mrg, cranger_bcodes[names(all_cbender_mdata_mrg)], ~.x %>%
#     mutate(is_cell_cr = case_when(barcode %in% .y$bcode ~ "CellCr"),
#            is_cell_cb_wHs = case_when(cell_probability_wHs > 0.5  ~ "CellCb", cell_probability_wHs <= 0.5  ~ "nonCell"),
#            is_cell_cb_rmHs = case_when(cell_probability_rmHs > 0.5  ~ "CellCb", cell_probability_wHs <= 0.5  ~ "nonCell")) %>%
#     mutate(mthds_union = case_when(is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" & is_cell_cb_rmHs == "CellCb" ~ "CrCbpfCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" ~ "CrCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_rmHs == "CellCb" ~ "CrCbpf",
#                                    is.na(is_cell_cr) & is_cell_cb_rmHs == "CellCb" & is_cell_cb_wHs == "CellCb" ~ "CbpfCbhs",
#                                    is_cell_cr == "CellCr" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs))  ~ "Cr",
#                                    is.na(is_cell_cr)  & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) & is_cell_cb_wHs == "CellCb" ~ "Cbhs",
#                                    is.na(is_cell_cr)  & is_cell_cb_rmHs == "CellCb" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) ~ "Cbpf",
#                                    is.na(is_cell_cr) & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) ~ "Cbnoncell"),
#            across(mthds_union, ~factor(., levels = c("CrCbpfCbhs", "CrCbpf", "CrCbhs", "CbpfCbhs", "Cr", "Cbpf", "Cbhs","Cbnoncell"))))
# )


```


```{r, message=F, warning=F, eval=T}
# cr_cbender_mdata_uni <- map2(cr_cbender_mdata_uni[intersect(names(cr_cbender_mdata_uni), names(all_msc_bp_raw_noHsGns_mdata))],
#                             all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_uni), names(all_msc_bp_raw_noHsGns_mdata))],
#                             ~left_join(.x, 
#                                        .y %>% rownames_to_column("barcode"), 
#                                        by = "barcode")
#                             )

```


```{r}
map(cr_cbender_mdata_uni, ~.x %>% count(is_cell_cr, is_cell_cb_wHs, is_cell_cb_rmHs, mthds_union))

```

```{r}
# saveRDS(cr_cbender_mdata_uni, "data/processed/Pf/m2_all/cr_cbender_mdata_uni.RDS")

```

```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_slct_uni, ~.x %>% 
      # filter(pf_umi_count > 100) %>%
      count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  # filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  scale_fill_manual(values = c_combos_cols) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_slct_uni, ~.x %>% 
      filter(pf_umi_count > 100) %>%
      count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  # filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  scale_fill_manual(values = c_combos_cols) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```



```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(map(cr_cbender_mdata_slct_uni, ~filter(., pf_umi_count > 100)), .id = "donor") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    # geom_density_2d()  +
    scale_colour_manual(values = c_combos_cols) +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 4), ncol = 1))
})

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_mdata_slct_uni, .id = "donor") %>%
    filter(., pf_umi_count > 100 & mthds_union != "nonCell") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    # geom_density_2d()  +
    scale_colour_manual(values = c_combos_cols) +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 4), ncol = 1))
})

```

######## Get clean cells and create a variable for this
```{r}
cr_cbender_pf100_mdata_slct_uni <- map(cr_cbender_mdata_slct_uni, ~.x %>% filter(pf_umi_count >= 100))

cr_cbender_pf100_mdata_slct_uni <- map(cr_cbender_pf100_mdata_slct_uni, ~.x %>%
    # filter() %>%
    mutate(fnl_call = case_when(mthds_union %in%  c("CrCbpf", "CbhsCbpf", "CrCbhsCbpf") ~ "Cell",
                                mthds_union %in%  c("nonCell") ~ "nonCell",
                                T ~ "other_calls"))
)
         
      
```

```{r}
map(cr_cbender_pf100_mdata_slct_uni, ~.x %>% count(mthds_union, fnl_call))
```

```{r}
## Save first pass cellbender and cellranger cells
saveRDS(cr_cbender_pf100_mdata_slct_uni, "data/processed/Pf/m2_all/cr_cbender_pf100_mdata_slct_uni.RDS")
```
###################################### START - Plot knee ################################

```{r}
# Add emptydrops metadata
all_raw_cbcr_mdata <- pmap(list(cr_cbender_pf100_mdata_slct_uni, all_msc_bp_raw_noHsGns_mdata[names(cr_cbender_pf100_mdata_slct_uni)]), ~{
  ..1 %>% 
    select(barcode, is_cell_cr, is_cell_cb_hs, is_cell_cb_pf, cell_probability_pf, cell_probability_hs, mthds_union, fnl_call) %>%
    full_join(..2 %>% rownames_to_column("barcode"), by = "barcode") %>% 
                           filter(nCount_RNA > 1) %>%
                           column_to_rownames("barcode")  %>%
                           mutate(across(c(contains("is_cell"), "fnl_call", "mthds_union"), ~replace_na(as.character(.), "empty_bc")))
    
})

```


```{r}
## read manual thresholds
edrops_thresh <- readRDS("data/processed/Pf/m2_all/man_thresh.RDS")
```

```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb_hspf <- map2(all_raw_cbcr_mdata, edrops_thresh[names(all_raw_cbcr_mdata)], ~{
  .x %>% 
    # filter(!is.na(is_cell_cb_hs) & !is.na(is_cell_cb_hs) ) %>%
    mutate(rank = rank(desc(nCount_RNA), ties.method = "average"),
           man_thresh = .y) %>% 
    filter(!(duplicated(rank) & is_cell_cb_hs %in% c("empty_bc", "nonCell") & is_cell_cb_pf %in% c("empty_bc", "nonCell")))
}) %>%
  bind_rows(., .id = "donor")


```


```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb_hspf_mrg <- map2(all_raw_cbcr_mdata, edrops_thresh[names(all_raw_cbcr_mdata)], ~{
  .x %>% 
    # filter(!is.na(is_cell_cb)) %>%
    mutate(rank = rank(desc(nCount_RNA), ties.method = "average"),
           pf_rank = rank(desc(pf_umi_count), ties.method = "average"),
           hs_rank = rank(desc(hs_umi_count), ties.method = "average"),
           man_thresh = .y) #%>% 
    # filter(!(duplicated(rank) & duplicated(pf_rank) & duplicated(hs_rank) ))
}) %>%
  bind_rows(., .id = "donor")


```

```{r, fig.height= 10, fig.width= 10, eval=T}
all_rank_ncount_cb_hspf_mrg %>%
  filter(!(duplicated(pf_rank) & is_cell_cb_hs %in% c("empty_bc", "nonCell") & is_cell_cb_pf %in% c("empty_bc", "nonCell")), .by = "donor") %>%
  kneeplt_facet_fn(ncount_tbl = ., cell_call_methd = "fnl_call", mthd_col = cell_call_mthd_col, donor_group = "donor", numbr_of_cols = 5, rank_var = "pf_rank", cell_var = "pf_umi_count", cut_offs = c(50, 100))

```


```{r, fig.height= 10, fig.width= 10, eval=T}
# kneeplt_facet_fn(ncount_tbl = all_rank_ncount_cb_hspf_mrg %>% filter(!(duplicated(pf_rank)), .by = "donor") , 
#                  cell_call_methd = "donor", donor_group = "donor", numbr_of_cols = 5, rank_var = "hs_rank", cell_var = "hs_umi_count")

```


```{r, fig.height= 10, fig.width= 10, eval=T}
all_rank_ncount_cb_hspf_mrg %>% 
  filter(nCount_RNA > 5) %>%
  filter((duplicated(pf_rank) & duplicated(hs_rank) & (duplicated(rank))), .by = "donor") %>%
  arrange(rank)

```

```{r, fig.height= 10, fig.width= 10, eval=T}
# all_rank_ncount_cb_hspf_mrg %>% 
#   filter(!(duplicated(pf_rank) & duplicated(hs_rank) & (duplicated(rank))), .by = "donor") %>%
#   rename("all_rank" = rank) %>%
#   rename("all_umi_count" = nCount_RNA) %>%
#   pivot_longer(cols = c(pf_rank, hs_rank, all_rank, pf_umi_count, hs_umi_count, all_umi_count), names_to = c("species", ".value"), names_pattern = "(pf|hs|all)_(rank|umi_count)") %>%
#   kneeplt_facet_fn(ncount_tbl = . , donor_group = "donor", numbr_of_cols = 5, rank_var = "rank", cell_var = "umi_count", cell_call_methd = "species", mthd_col = c("pf" = "#007BFF", "all" = "#FFC107", "hs" = "#b2d8d8"), cut_offs = c(20,40,50))

```
######################################END - Plot knee ################################


######################################START - Write souporcell barcodes################################

```{r}

resc_bcodes_hspf <- map(cr_cbender_pf100_mdata_slct_uni, ~{
  .x  %>%
    filter(fnl_call %in%  c("Cell")) %>% 
    pull(barcode) 
  })  
```

```{r}
map(resc_bcodes_hspf, length) %>% bind_rows(., .id = "donor")
```

```{r}
map(resc_bcodes_hspf, length) %>% bind_rows(., .id = "donor")
```

```{r, message=F, warning=F, eval=T}
# gogo()
```
Write out barcodes to rerun souporcell including all possible barcode that are potentially actual cells

```{r}

imap(resc_bcodes_hspf, ~{
  system(paste0('mkdir data/processed/Pf/', .y, '/soupc_resc_hspf'))
  
  .x %>% write(., gzfile(paste0('data/processed/Pf/', .y, '/soupc_resc_hspf/barcodes.tsv.gz')))
  }) 
```

Run souporcell like below for the filtered barcodes of pf in mixed infections

```{r, eval=T}
gogo()
# bsub < /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code_lnk/bash_scripts/soupc_v25_m2_bsub.sh
# --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/5736STDY*/outs/possorted_genome_bam.bam" \
# --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/soupc_resc_hspf/barcodes.tsv.gz" \
# --soup_dir "soupc_resc_hspf" \
```


######################################END - Write souporcell barcodes################################

```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")

cr_cbender_mdata_uni[["MSC12"]] %>% filter(is.na(mthds_union))
```

```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(cr_cbender_mdata_uni[1:4], ~.x %>% filter(is.na(mthds_union)))

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 4), ncol = 1))
})

```

```{r}
cr_cbender_pf100_mdata_uni <- map(cr_cbender_mdata_uni, ~.x %>% filter(pf_umi_count >= 100))
```

```{r}
# saveRDS(cr_cbender_pf100_mdata_uni, "data/processed/Pf/m2_all/cr_cbender_pf100_mdata_uni.RDS")
```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_pf100_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 3), ncol = 1))
})

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(cr_cbender_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")



```



```{r, fig.width=10, fig.height=12}
map(cr_cbender_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, fig.width=10, fig.height=5}
map(cr_cbender_pf100_mdata_uni, ~.x %>% 
      filter(!(mthds_union %in% c("Cr", "Cbnoncell"))) %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cr")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_pf100_mdata_uni, .id = "donor") %>%
    filter(!(mthds_union %in% c("Cbycr", "nonCell"))) %>%
    mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cbycr")) %>% 
    filter(!(mthds_union_CbCr %in% c("CrCbhsCbpf", "CbhsCbpf"))) %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union_CbCr)) +
    geom_point(size = 0.5) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    theme_classic() +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 3), ncol = 1))
})

```

```{r, fig.width=10, fig.height=5}
map(cr_cbender_pf100_mdata_uni, ~.x %>% 
      filter(!(mthds_union %in% c("Cbpf", "Cbnoncell"))) %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cbpf")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, fig.width=10, fig.height=5}
map(cr_cbender_pf100_mdata_uni, ~.x %>% 
      filter(!(mthds_union %in% c("Cbpf", "Cbnoncell"))) %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cbpf")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```
