---
title: "MSC QC cbender_comparisons"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(scran)
  library(SingleR)
  library(scDblFinder)
  library(ggsankey)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(dplyr::desc)
  conflicts_prefer(dplyr::count)
  conflicts_prefer(purrr::reduce)
})
```



# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/project_plotting_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")

```

###################START- RAW READING#############

```{r, eval=T}
# pf_all_id_decode <- read_csv("data/raw/pf_all_id_decode.csv")
pf_all_id_decode <- read_csv("data/raw/pf_all_id_decode_minus_noJC.csv")

pf_all_id_decode$sample_nm
```


```{r, eval=T}
# pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
pf_solo_mixed_mdata_decode <- pf_all_id_decode 

```

## Genotype clustering for MSC

initialize variable for the sample names



```{r}

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
# slct_sample_names <-c("MSC68", "MSC14", "MSC49", "MSC55") %>% str_sort(., numeric = T)


# ## poor quality samples
# sample_rm <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39", "MSC50_SLO", "MSC33")
sample_rm <- c("MSC1", "MSC24", "MSC36", "MSC37", "MSC38", "MSC39")

sample_mxd <- pf_solo_mixed_mdata_decode %>% filter(mixed_infection2 == "mixed") %>% pull(sample_nm)
# 
## slct_sample_names[-c(sample_rm)]
# slct_sample_names <- slct_sample_names[!slct_sample_names %in% c(sample_rm, sample_mxd)]

all_samples2rm <- c(sample_rm, sample_mxd)
all_samples2rm <- c(all_samples2rm, paste0(all_samples2rm, "C"), paste0(all_samples2rm, "noJC"))
slct_sample_names <- slct_sample_names[!slct_sample_names %in% all_samples2rm]

## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```

```{r, message=F, warning=F, eval=F}
## Get the number of cells that pass cellranger filter
crangr_filt_num <- system("zgrep -c '' /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_run*/Pf_all_genes/*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz", intern = TRUE) 
```


```{r, message=F, warning=F, eval=F}
## Df of cellranger pass cells
crangr_filt_df <- data.frame('irods_id' = str_extract(crangr_filt_num, "5736STDY[0-9]+"),
                             'cranger_filt_n' = str_extract(crangr_filt_num, "[0-9]+$"))

pf_mdata_decode_cranger_filt_n <- pf_solo_mixed_mdata_decode %>% full_join(crangr_filt_df, by = "irods_id")

pf_mdata_decode_cranger_filt_n %>% write_csv(., "data/raw/pf_mdata_decode_cranger_filt_n.csv")
```



```{r, fig.width=15, fig.height=5}
donor_10x_mdata_sk21_summ <- readRDS("data/processed/Pf/m2_all/donor_10x_mdata_sk21_summ.RDS") %>%
  filter(donor %in% unique(str_remove(slct_sample_names, "noJC$|C$"))) 

```


```{r, fig.width=15, fig.height=5}

donor_10x_mdata_sk21_summ %>% count(frac_load, overload, frac_over, targeted_cells)
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_rmHs <- paste0(sample_name, "_rm_HS")
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
sample_name_yascpCranger <- paste0(sample_name, "_yascpCr")
```

```{r, message=F, warning=F, eval=T}
## Names for cellbender with learning rate of 0.000025
sample_name_LRate_mod <- paste0(sample_name, "_LRpt000025")
```


```{r, message=F, warning=F, eval=T}
sample_name_LR_mod <- paste0(sample_name, "_LRpt000005e250")
sample_name_LR <- paste0(sample_name, "_LRpt0001")
sample_name_pf_LR_mod <- paste0(sample_name, "_rmHS_LRpt000005e250")
sample_name_pf <- paste0(sample_name, "_rmHS_LRpt0001")
```


```{r, message=F, warning=F, eval=T}
# sample_name_LR_mod <- paste0(sample_name, "_LRpt000005e250")
# sample_name_LR <- paste0(sample_name, "_LRpt0001")
# sample_name_pf_LR_mod <- paste0(sample_name, "_rmHS_LRpt000005e250")
# sample_name_pf <- paste0(sample_name, "_rmHS_LRpt0001")

# sample_name_LR_mod <- paste0(sample_name, "_yascp_default_pfhs_dir")
# sample_name_LR <- paste0(sample_name, "_LRpt0001")
# sample_name_pf_LR_mod <- paste0(sample_name, "_yascp_default_pf_dir")
# sample_name_pf <- paste0(sample_name, "_rmHS_LRpt0001")

# sample_name_LR_mod <- paste0(sample_name, "_wHsPf_LR0.000005_E250")
# sample_name_LR <- paste0(sample_name, "_wHsPf_LR0.0001_E150")
# sample_name_pf_LR_mod <- paste0(sample_name, "_rmHs_LR0.000005_E250")
# sample_name_pf <- paste0(sample_name, "_rmHs_LR0.0001_E150")

sample_hsLRloEhi <- paste0(sample_name, "_wHsPf_LR0.000005_E250")
sample_hsLRloElo <- paste0(sample_name, "_wHsPf_LR0.000005_E150")
sample_hsLRhiEhi <- paste0(sample_name, "_wHsPf_LR0.0001_E250")
sample_hsLRhiElo <- paste0(sample_name, "_wHsPf_LR0.0001_E150")

sample_pfLRloEhi <- paste0(sample_name, "_rmHs_LR0.000005_E250")
sample_pfLRloElo <- paste0(sample_name, "_rmHs_LR0.000005_E150")
sample_pfLRhiEhi <- paste0(sample_name, "_rmHs_LR0.0001_E250")
sample_pfLRhiElo <- paste0(sample_name, "_rmHs_LR0.0001_E150")

sample_dpfLRloElo <- paste0(sample_name, "_drmHs_LR0.000005_E150")
sample_dhsLRloElo <- paste0(sample_name, "_dwHsPf_LR0.000005_E150")

sample_sufx_cb <- c("_hsLRloEhi", "_hsLRloElo", "_hsLRhiEhi", "_hsLRhiElo", "_pfLRloEhi", "_pfLRloElo", "_pfLRhiEhi", "_pfLRhiElo", "_dpfLRloElo", "_dhsLRloElo")

```

```{r, message=F, warning=F, eval=T}
dirs <- Sys.glob("data/processed/Pf/MSC*/cbender_custom2_*/cb*")
class(dirs)

```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
cbender_mdata_std_yascp_syst <- map(dirs, ~list.files(.x, pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender_custom2|_gns|/cb_mdata.*|/cb|/|efault_'))))  %>% unlist(., recursive = F) %>% 
  # set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*|/'))) %>% 
  # set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|/cb_yascp_default|_dir|_gns|/cb_mdata.*'))) %>% 
  .[c(sample_hsLRloEhi, sample_hsLRloElo, sample_hsLRhiEhi, sample_hsLRhiElo, sample_pfLRloEhi, sample_pfLRloElo, sample_pfLRhiEhi, sample_pfLRhiElo, sample_dpfLRloElo, sample_dhsLRloElo)] %>%
  .[!is.na(.)] %>%
  map(., ~read_csv(.x, show_col_types = FALSE)) 
```


```{r, message=F, warning=F, eval=T}
# ##Read BPcells matrices
# cbender_mdata_std_yascp_syst <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender_custom|_gns|/cb_mdata.*|/'))) %>% 
#   # set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*|/'))) %>% 
#   # set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|/cb_yascp_default|_dir|_gns|/cb_mdata.*'))) %>% 
#   .[c(sample_name_LR, sample_name_LR_mod, sample_name_pf, sample_name_pf_LR_mod)] %>%
#   .[!is.na(.)] %>%
#   map(read_csv) 
```

```{r, message=F, warning=F, eval=T}
# Create a list of dataframes to join, with their corresponding suffixes
dfs_to_join <- list(
  cbender_mdata_std_yascp_syst[sample_hsLRloEhi],
  cbender_mdata_std_yascp_syst[sample_hsLRloElo],
  cbender_mdata_std_yascp_syst[sample_hsLRhiEhi],
  cbender_mdata_std_yascp_syst[sample_hsLRhiElo],
  cbender_mdata_std_yascp_syst[sample_pfLRloEhi],
  cbender_mdata_std_yascp_syst[sample_pfLRloElo],
  cbender_mdata_std_yascp_syst[sample_pfLRhiEhi],
  cbender_mdata_std_yascp_syst[sample_pfLRhiElo],
  cbender_mdata_std_yascp_syst[sample_dpfLRloElo],
  cbender_mdata_std_yascp_syst[sample_dhsLRloElo]
)

# Get the number of samples
n_samples <- length(dfs_to_join[[1]])

# For each sample, join all 8 dataframes with appropriate suffixes
cbender_std_yascpCr_mrg_syst <- map(1:n_samples, function(i) {
  # Extract the i-th dataframe from each of the 8 lists and add suffix
  map2(dfs_to_join, sample_sufx_cb, function(df_list, suffix) {
    df_list[[i]] %>% rename_with(~ paste0(.x, suffix), -all_of("barcode"))
  }) %>%
    reduce(full_join, by = "barcode")
})

# Assign names from the first list (they should all have the same names)
names(cbender_std_yascpCr_mrg_syst) <- names(dfs_to_join[[1]])

names(cbender_std_yascpCr_mrg_syst) <- str_remove_all(names(cbender_std_yascpCr_mrg_syst), "_LRpt000005e250|_yascp_default_pfhs_dir|_yascp_default_pf_dir|_wHsPf_LR0.000005_E250")
# names(cbender_std_yascpCr_mrg_syst) <- str_remove_all(names(cbender_std_yascpCr_mrg_syst), "_LRpt000005e250|_yascp_default_pfhs_dir|_yascp_default_pf_dir")
```



########################################################################### JUMP to where it says "PICK UP FROM HERE"


########################################################################### PICK UP FROM here

### Compare cellbender iterations and also with cellranger

```{r, message=F, warning=F, eval=T}
bcodes_dirs <- Sys.glob("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz")
class(bcodes_dirs)

```


```{r, message=F, warning=F, eval=T}
## read filtered barcodes
cranger_bcodes <- map(bcodes_dirs, ~.x %>% 
  set_names(nm = (str_remove_all(.,'/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/|/outs/filtered.*')))) %>% 
  unlist(., recursive = F) %>% 
  .[source_irods] %>%
  set_names(names(source_irods)) %>%
  map(.,~read_tsv(., col_names = "bcode")) 
```


```{r, message=F, warning=F, eval=T}
cranger_bcodes <- map(cranger_bcodes, ~.x %>% mutate(is_cell_cr = "Cr"))

```


```{r}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_hmnz_syst_snky <- map2(cbender_std_yascpCr_mrg_syst, cranger_bcodes[str_remove(names(cbender_std_yascpCr_mrg_syst), "_LRpt000005e250")], ~.x %>%
    full_join(., .y, by = c("barcode"="bcode")) %>%
    mutate(cr = case_when(is_cell_cr == "Cr" ~ "Cell")) %>%
    mutate(across(starts_with("cell_probability_"), ~case_when(.x > 0.5  ~ "Cell", .x <= 0.5  ~ "nonCell"), .names = "{sub('^cell_probability_', 'cb_', .col)}"))
)
         
      
```

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5, eval=F}
# Set up parallel processing
plan(multisession, workers = 7)

# Parallelize plot generation
cb_snky_plts <- future_imap(cr_cbender_mdata_hmnz_syst_snky, ~snk_pt_fn(., strns_2_plt = c("cr", paste0("cb", sample_sufx_cb)), labls = c("cr", paste0("cb", sample_sufx_cb)), ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')), .options = furrr_options(seed = TRUE))

# Clean up
plan(sequential)

```

```{r, message=FALSE, warning=FALSE, fig.width=10, fig.height=5, eval=F}
map(cb_snky_plts, ~.x + 
      theme(axis.text.x = element_text(angle = 20, hjust = 0.8, face = "plain")))

```


```{r}
## Read raw cellranger metadata with the sum of Human and Pf reads
# all_msc_bp_raw_noHsGns_mdata <- readRDS("data/processed/Pf/m2_all/all_msc_bp_raw_noHsGns_mdata.RDS") %>% .[sample_name]
# all_msc_bp_raw_noHsGns_mdata <- readRDS("data/processed/Pf/m2_all/all_msc_bp_raw_noHsGns_mdata.RDS") %>% .[sample_name]

```


```{r, message=F, warning=F, eval=T}
## Get mdata file list
mdata_file_list <- Sys.glob("data/processed/Pf/*/seu_obj/all_msc_bp_raw_noHsGns_mdata.RDS")
mdata_file_list
```

```{r, message=F, warning=F, eval=T}
## read filtered barcodes
all_msc_bp_raw_noHsGns_mdata <- map(mdata_file_list, ~.x %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj.*')))) %>% 
  unlist(., recursive = F) %>% 
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(readRDS) 

```

```{r, message=F, warning=F, eval=T}
cr_cbender_mdata_hmnz_syst_snky_pfumi <- map2(cr_cbender_mdata_hmnz_syst_snky[intersect(names(cr_cbender_mdata_hmnz_syst_snky), names(all_msc_bp_raw_noHsGns_mdata))],
                            all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_hmnz_syst_snky), names(all_msc_bp_raw_noHsGns_mdata))],
                            ~left_join(.x, 
                                       .y %>% rownames_to_column("barcode"), 
                                       by = "barcode") #%>%
                              #filter(pf_umi_count >= 100 & !is.na(pf_umi_count))
                            )

```

```{r, message=FALSE, warning=FALSE, eval=T}
# Parallelize plot generation
plan(multisession, workers = 7)

cb_snky_plts_pf100 <- future_imap(cr_cbender_mdata_hmnz_syst_snky_pfumi, ~snk_pt_fn(.x %>% filter(pf_umi_count >= 100 & !is.na(pf_umi_count)), strns_2_plt = c("cr", paste0("cb", sample_sufx_cb)), labls = c("cr", paste0("cb", sample_sufx_cb)), ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')), .options = furrr_options(seed = TRUE))

plan(sequential)

```

```{r, message=FALSE, warning=FALSE, eval=T}
map(cb_snky_plts_pf100, ~.x + 
      theme(axis.text.x = element_text(angle = 20, hjust = 0.8, face = "plain")))

```

```{r, message=FALSE, warning=FALSE}
runs2plot <- c("cr", paste0("cb", c("_hsLRloEhi", "_hsLRhiElo", "_pfLRloEhi", "_pfLRhiElo")))
runs2plot <- c("cr", paste0("cb", c("_hsLRloElo", "_pfLRloElo", "_dpfLRloElo", "_dhsLRloElo")))

```

```{r, message=FALSE, warning=FALSE, eval=T}
# Parallelize plot generation
plan(multisession, workers = 7)

cb_snky_plts_pf100_set <- future_imap(cr_cbender_mdata_hmnz_syst_snky_pfumi, ~snk_pt_fn(.x %>% filter(pf_umi_count >= 100 & !is.na(pf_umi_count)), strns_2_plt = runs2plot, labls = runs2plot, ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')), .options = furrr_options(seed = TRUE))

plan(sequential)

```

```{r, message=FALSE, warning=FALSE, eval=T}
cb_snky_plts_pf100_set

```

```{r}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_hmnz <- map2(cr_cbender_mdata_hmnz_syst_snky_pfumi, cranger_bcodes[names(cr_cbender_mdata_hmnz_syst_snky_pfumi)], ~.x %>%
    mutate(#is_cell_cr = case_when(barcode %in% .y$bcode ~ "Cr"),
           is_cell_cb_hs = case_when(cell_probability_hsLRhiElo > 0.5  ~ "CbHs", cell_probability_hsLRhiElo <= 0.5  ~ "nonCell"),
           is_cell_cb_pf = case_when(cell_probability_pfLRhiElo > 0.5  ~ "CbPf", cell_probability_pfLRhiElo <= 0.5  ~ "nonCell"),
           is_cell_cb_yhs = case_when(cell_probability_hsLRloEhi > 0.5  ~ "CbyHs", cell_probability_hsLRloEhi <= 0.5  ~ "nonCell"),
           is_cell_cb_ypf = case_when(cell_probability_pfLRloEhi > 0.5  ~ "CbyPf", cell_probability_pfLRloEhi <= 0.5  ~ "nonCell"),
           is_cell_cb_mhs = case_when(cell_probability_hsLRloElo > 0.5  ~ "CbmHs", cell_probability_hsLRloElo <= 0.5  ~ "nonCell"),
           is_cell_cb_mpf = case_when(cell_probability_pfLRloElo > 0.5  ~ "CbmPf", cell_probability_pfLRloElo <= 0.5  ~ "nonCell")
           ) #%>%
      #full_join(., .y, by = c("barcode"="bcode"))
    )
         
      
```

```{r}
# go()
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_uni <- map(cr_cbender_mdata_hmnz, function(val) { val %>%
    # filter() %>%
    mutate(mthds_union = paste0(is_cell_cr, is_cell_cb_yhs, is_cell_cb_ypf, is_cell_cb_pf, is_cell_cb_mpf, is_cell_cb_mhs),
           mthds_union_slim = paste0(is_cell_cr, is_cell_cb_yhs, is_cell_cb_ypf),
           mthds_union_slim2 = paste0(is_cell_cr, is_cell_cb_ypf, is_cell_cb_pf),
           mthds_union_slim3 = paste0(is_cell_cr, is_cell_cb_mpf, is_cell_cb_mhs)) %>%
    mutate(across(starts_with("mthds_union"), ~case_when(
      str_detect(., "Cb|Cr") ~ str_remove_all(., "nonCell|NA"),
      !str_detect(., "Cb|Cr") ~ str_remove_all(., "NA"),
      TRUE ~ .
    ))) %>%
    mutate(across(contains("mthds_union"), ~case_when(
      !str_detect(., "Cb|Cr") & str_detect(., "nonCell") ~ "nonCell",
      !str_detect(., "Cb|Cr") & mthds_union == "nonCell" ~ "nonCell",
      !str_detect(., "Cb|Cr") | is.na(.) ~ "nonCell",
      TRUE ~ .
    )))
})
         
# sample_sufx_cb <- c("_hsLRloEhi", "_hsLRloElo", "_hsLRhiEhi", "_hsLRhiElo", "_pfLRloEhi", "_pfLRloElo", "_pfLRhiEhi", "_pfLRhiElo")
      
```


```{r}
cr_cbender_mdata_uni[[1]] %>% glimpse()
```

```{r}
## Convert mthds_union to factor
## Get all possible combinations of factors factoring in the order in which the variables were introduced
# c_mthds <- c("Cr", "Cbhs", "Cbpf", "Cbycr")
# c_mthds <- c("Cr", "CbyHs", "CbyPf", "CbPf")
c_mthds <- c("Cr", "CbyHs", "CbyPf", "CbPf", "CbmPf", "CbmHs")
c_mthds <- str_replace_all(c_mthds, "Cb", "")

c_combos <- map(1:length(c_mthds), ~ combn(c_mthds, .x, paste0, collapse = "_")) %>% unlist() %>% dput()
c_combos <- c(rev(c_combos), "nonCell")

set.seed(12)
# c_combos_cols <- sample(c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#fabebe', '#008080', '#e6beff', '#9a6324',  '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075'), length(c_combos)) %>% set_names(c_combos)
c_combos_cols <- sample(c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#fabebe', '#008080', '#e6beff', '#9a6324',  '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075','#000000','#FFFFFF','#FF0000','#00FF00','#0000FF','#FFFF00','#00FFFF','#FF00FF',
  '#800080','#808080','#C0C0C0','#FFA500','#A52A2A','#008000','#ADD8E6','#00008B',
  '#B22222','#FF1493','#2E8B57','#DAA520','#4B0082','#7FFF00','#D2691E','#F4A460',
  '#20B2AA','#87CEEB','#6A5ACD','#DB7093','#F0E68C','#9ACD32','#00CED1','#4682B4',
  '#5F9EA0','#FF69B4','#CD5C5C','#191970','#FFE4B5','#8B4513','#2F4F4F','#556B2F',
  '#FFB6C1','#E9967A','#8A2BE2','#00BFFF','#7CFC00','#F5DEB3','#9AED92'), length(c_combos)) %>% set_names(c_combos)

c_combos_cols <- ifelse(names(c_combos_cols) == "nonCell", "lightgrey", c_combos_cols)  %>% set_names(c_combos)

```


```{r}
# go()
## Convert mthds_union to factor

cr_cbender_mdata_uni <- map(cr_cbender_mdata_uni, ~.x %>%
    # mutate(across(mthds_union, ~factor(., levels = c_combos)))%>%
    mutate(across(contains("mthds_union"), ~factor(str_replace_all(., c("^Cb" = "", "Cb" = "_")), levels = c_combos)))
)

```

```{r}
map(cr_cbender_mdata_uni[1:2], ~.x %>% count(is_cell_cr,is_cell_cr, is_cell_cb_pf, is_cell_cb_yhs, is_cell_cb_ypf, mthds_union, mthds_union_slim))

```


#########################################################################START## RUN for custom selected samples based on the cellbender convergence plot


```{r}
sample_cbender_select <- c("MSC12_LRpt0001", "MSC13_LRpt0001", "MSC14_LRpt000005e250", "MSC3_LRpt000005e250", "MSC33_LRpt000005e250", "MSC40_LRpt000005e250", "MSC45_LRpt0001", "MSC48_LRpt000005e250", "MSC49_LRpt000005e250", "MSC50_MACS_LRpt000005e250", "MSC50_SLO_LRpt000005e250", "MSC53_LRpt0001", "MSC54_LRpt000005e250", "MSC55_LRpt000005e250", "MSC57_LRpt0001", "MSC60_LRpt000005e250", "MSC66_LRpt000005e250", "MSC67_LRpt000005e250", "MSC68_LRpt000005e250", "MSC70_LRpt0001")
```


```{r, message=F, warning=F, eval=F}
sample_cbender_select_ptns <- ifelse(str_detect(sample_cbender_select, "LRpt000005e250"), c("lhs|lpf"), c("dhs|dpf")) %>% set_names(str_remove(sample_cbender_select, "_LR.*"))
```

```{r, message=F, warning=F, eval=F}
cbender_std_yascpCr_mrg_syst_slct <- map2(cbender_std_yascpCr_mrg_syst,  sample_cbender_select_ptns[names(cbender_std_yascpCr_mrg_syst)], ~.x %>%
       select(barcode, contains(str_remove(.y, "\\|.*")), contains(str_remove(.y, ".*\\|"))) %>%
         rename_with(~ str_replace_all(.x, c("lhs"="hs", "dhs"="hs", "lpf"="pf", "dpf"="pf")), -all_of("barcode"))
       )
```


```{r, message=F, warning=F, eval=F}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
cr_cbender_mdata_slct_hmnz <- map2(cbender_std_yascpCr_mrg_syst_slct, cranger_bcodes[names(cbender_std_yascpCr_mrg_syst_slct)], ~.x %>%
    mutate(#is_cell_cr = case_when(barcode %in% .y$bcode ~ "Cr"),
           is_cell_cb_hs = case_when(cell_probability_hs> 0.5  ~ "Cbhs", cell_probability_hs <= 0.5  ~ "nonCell"),
           is_cell_cb_pf = case_when(cell_probability_pf > 0.5  ~ "Cbpf", cell_probability_pf <= 0.5  ~ "nonCell")
           ) %>%
      full_join(., .y, by = c("barcode"="bcode")))
         
      
```

```{r, message=F, warning=F, eval=F}
cr_cbender_mdata_slct_hmnz_gns <- map2(cr_cbender_mdata_slct_hmnz[intersect(names(cr_cbender_mdata_slct_hmnz), names(all_msc_bp_raw_noHsGns_mdata))],
                            all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_slct_hmnz), names(all_msc_bp_raw_noHsGns_mdata))],
                            ~left_join(.x, 
                                       .y %>% rownames_to_column("barcode"), 
                                       by = "barcode")
                            )

```


## Sankey plots of cell calls

```{r, message=FALSE, warning=FALSE, eval=F}
cr_cbender_mdata_slct_uni_sanky <- map(cr_cbender_mdata_slct_hmnz_gns, ~.x %>% 
      filter(pf_umi_count >= 100) %>% 
      select(barcode, contains("is_cell")) %>%
      mutate(across(-c(barcode), ~ifelse(str_detect(.x, "Cr|Cb"), "Cell", .x))) %>%
      rename_with(~ str_remove(.x, "is_cell_"), -all_of("barcode"))
)
  
```


```{r, message=FALSE, warning=FALSE, eval=F}

imap(cr_cbender_mdata_slct_uni_sanky, ~snk_pt_fn(., strns_2_plt = c("cr", "cb_wHs", "cb_rmHs", "cb_yCr", "cb_lrt"), labls = c("cr", "cb_wHs", "cb_rmHs", "cb_yCr", "cb_lrt"), bp_face = c("bold", "bold", "bold", "bold", "bold"), ttl = .y, strain_colours = c('Cell' ='#911eb4', 'nonCell' ='#008080', 'NA' = 'lightgrey')))
  
```

```{r, eval=F}
## Create a column summarising the agreement of call methods for each cell - called 'mthds_union'
# cr_cbender_mdata_uni <- map(cr_cbender_mdata_hmnz, ~.x %>%
cr_cbender_mdata_slct_uni <- map(cr_cbender_mdata_slct_hmnz_gns, ~.x %>%
    mutate(mthds_union = paste0(is_cell_cr, is_cell_cb_hs, is_cell_cb_pf),
           mthds_union = case_when(str_detect(mthds_union, "Cb|Cr") ~ str_remove_all(mthds_union, "nonCell|NA"),
                                   !str_detect(mthds_union, "Cb|Cr") ~ str_remove_all(mthds_union, "NA"),
                                   T ~ mthds_union),
           mthds_union = case_when(str_detect(mthds_union, "nonCell") ~  "nonCell",
                                   T ~ mthds_union)
           ))
         
      
```


```{r, eval=F}
ls()
```

```{r, eval=F}
## Convert mthds_union to factor
cr_cbender_mdata_slct_uni <- map(cr_cbender_mdata_slct_uni, ~.x %>%
    mutate(across(mthds_union, ~factor(., levels = c_combos)))
)

```



#########################################################################END## RUN for custom selected samples based on the cellbender convergence plot


```{r}
## Convert mthds_union to factor
# cr_cbender_mdata_uni <- map2(cbender_std_yascpCr_mrg, cranger_bcodes[names(cbender_std_yascpCr_mrg)], ~.x %>%
#     mutate(is_cell_cr = case_when(barcode %in% .y$bcode ~ "CellCr"),
#            is_cell_cb_wHs = case_when(cell_probability_wHs > 0.5  ~ "CellCb", cell_probability_wHs <= 0.5  ~ "nonCell"),
#            is_cell_cb_rmHs = case_when(cell_probability_rmHs > 0.5  ~ "CellCb", cell_probability_rmHs <= 0.5  ~ "nonCell"),
#            is_cell_cb_yCr = case_when(cell_probability_yCr > 0.5  ~ "CellCb", cell_probability_yCr <= 0.5  ~ "nonCell")) %>%
#     mutate(mthds_union = case_when(is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" & is_cell_cb_rmHs == "CellCb" & is_cell_cb_yCr == "CellCb" ~ "CrCbpfCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" ~ "CrCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_rmHs == "CellCb" ~ "CrCbpf",
#                                    is.na(is_cell_cr) & is_cell_cb_rmHs == "CellCb" & is_cell_cb_wHs == "CellCb" ~ "CbpfCbhs",
#                                    is_cell_cr == "CellCr" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs))  ~ "Cr",
#                                    is.na(is_cell_cr)  & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) & is_cell_cb_wHs == "CellCb" ~ "Cbhs",
#                                    is.na(is_cell_cr)  & is_cell_cb_rmHs == "CellCb" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) ~ "Cbpf",
#                                    is.na(is_cell_cr) & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) ~ "Cbnoncell"),
#            across(mthds_union, ~factor(., levels = c("CrCbpfCbhs", "CrCbpf", "CrCbhs", "CbpfCbhs", "Cr", "Cbpf", "Cbhs","Cbnoncell"))))
# )


```

```{r}
# cr_cbender_mdata_uni <- map2(all_cbender_mdata_mrg, cranger_bcodes[names(all_cbender_mdata_mrg)], ~.x %>%
#     mutate(is_cell_cr = case_when(barcode %in% .y$bcode ~ "CellCr"),
#            is_cell_cb_wHs = case_when(cell_probability_wHs > 0.5  ~ "CellCb", cell_probability_wHs <= 0.5  ~ "nonCell"),
#            is_cell_cb_rmHs = case_when(cell_probability_rmHs > 0.5  ~ "CellCb", cell_probability_wHs <= 0.5  ~ "nonCell")) %>%
#     mutate(mthds_union = case_when(is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" & is_cell_cb_rmHs == "CellCb" ~ "CrCbpfCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_wHs == "CellCb" ~ "CrCbhs",
#                                    is_cell_cr == "CellCr" & is_cell_cb_rmHs == "CellCb" ~ "CrCbpf",
#                                    is.na(is_cell_cr) & is_cell_cb_rmHs == "CellCb" & is_cell_cb_wHs == "CellCb" ~ "CbpfCbhs",
#                                    is_cell_cr == "CellCr" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs))  ~ "Cr",
#                                    is.na(is_cell_cr)  & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) & is_cell_cb_wHs == "CellCb" ~ "Cbhs",
#                                    is.na(is_cell_cr)  & is_cell_cb_rmHs == "CellCb" & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) ~ "Cbpf",
#                                    is.na(is_cell_cr) & (is_cell_cb_wHs != "CellCb" | is.na(is_cell_cb_wHs)) & (is_cell_cb_rmHs != "CellCb" | is.na(is_cell_cb_rmHs)) ~ "Cbnoncell"),
#            across(mthds_union, ~factor(., levels = c("CrCbpfCbhs", "CrCbpf", "CrCbhs", "CbpfCbhs", "Cr", "Cbpf", "Cbhs","Cbnoncell"))))
# )


```


```{r, message=F, warning=F, eval=T}
# cr_cbender_mdata_uni <- map2(cr_cbender_mdata_uni[intersect(names(cr_cbender_mdata_uni), names(all_msc_bp_raw_noHsGns_mdata))],
#                             all_msc_bp_raw_noHsGns_mdata[intersect(names(cr_cbender_mdata_uni), names(all_msc_bp_raw_noHsGns_mdata))],
#                             ~left_join(.x, 
#                                        .y %>% rownames_to_column("barcode"), 
#                                        by = "barcode")
#                             )

```


```{r}
# saveRDS(cr_cbender_mdata_uni, "data/processed/Pf/m2_all/cr_cbender_mdata_uni.RDS")

```

```{r}
cr_cbender_mdata_slct_uni <- cr_cbender_mdata_uni
```


```{r, fig.width=12, fig.height=14}
call_methods_overlap_barplot_fn <- function(tbl = cr_cbender_mdata_slct_uni, mthd = "mthds_union", pf_umi_filt = 0, min_grp_size =0) {
map(tbl, ~.x %>% 
      filter(pf_umi_count >= pf_umi_filt) %>%
      count(!!rlang::sym(mthd)) %>%
      filter(n > min_grp_size)
    ) %>%
  bind_rows(., .id = "donor") %>%
  # filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = !!rlang::sym(mthd), fill = !!rlang::sym(mthd))) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.8))) +
  scale_fill_manual(values = c_combos_cols) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 11),
        legend.position = "bottom")

}
```


```{r, fig.width=12, fig.height=14}

call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union", pf_umi_filt = 0, min_grp_size = 10)
```

```{r, fig.width=12, fig.height=14, eval=FALSE}

call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union", pf_umi_filt = 100, min_grp_size = 10)
```

```{r, fig.width=12, fig.height=14}

call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union_slim") 

```

```{r, fig.width=12, fig.height=14}

call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union_slim", pf_umi_filt = 100) 

```

```{r, fig.width=12, fig.height=14, eval=F}

call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union_slim", pf_umi_filt = 100) 

```


```{r}
map(cr_cbender_pf100_mdata_slct_uni, ~.x %>% count(mthds_union_slim, mthds_union) %>% filter(n>10)) 
```


```{r, fig.width=12, fig.height=14}
call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union_slim2", pf_umi_filt = 100) 
```

```{r, fig.width=12, fig.height=14}
call_methods_overlap_barplot_fn(cr_cbender_mdata_slct_uni, "mthds_union_slim3", pf_umi_filt = 100) 
```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
cr_cbender_uni_pf100_mrg <- bind_rows(map(cr_cbender_mdata_slct_uni, ~filter(., pf_umi_count > 100)), .id = "donor")
```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
call_methods_overlap_scatter_fn <- function(tbl = cr_cbender_uni_pf100_mrg, gn_or_umi = "umi",  mthd = "mthds_union") {

  ggplot(tbl, aes(x = log10(!!rlang::sym(paste0('pf_',gn_or_umi,'_count'))), y = log10(!!rlang::sym(paste0('hs_',gn_or_umi,'_count'))), colour = !!rlang::sym(mthd))) +
    geom_point(size = 0.01) +
    # geom_density_2d()  +
    scale_colour_manual(values = c_combos_cols) +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 4), ncol = 1))
}

```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~call_methods_overlap_scatter_fn(gn_or_umi = .x))

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~call_methods_overlap_scatter_fn(gn_or_umi = .x))

```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
tbl = bind_rows(map(cr_cbender_mdata_slct_uni, ~filter(., pf_umi_count > 100 & mthds_union != "nonCell")), .id = "donor") 

map(c("gn", "umi"), ~call_methods_overlap_scatter_fn(tbl = tbl, gn_or_umi = .x))

```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
tbl = bind_rows(map(cr_cbender_mdata_slct_uni, ~filter(., pf_umi_count > 100 & mthds_union_slim != "nonCell")), .id = "donor") 

map(c("gn", "umi"), ~call_methods_overlap_scatter_fn(tbl = tbl, gn_or_umi = .x, mthd = "mthds_union_slim"))

```

######## Get clean cells and create a variable for this
```{r}
cr_cbender_mdata_slct_uni <- map(cr_cbender_mdata_slct_uni, ~.x %>%
    # filter() %>%
    # mutate(fnl_call = case_when(mthds_union %in%  c("CrCbpf", "CbhsCbpf", "CrCbhsCbpf") ~ "Cell",
    mutate(fnl_call = case_when(mthds_union_slim %in%  c("Cr_yPf", "yHs_yPf", "Cr_yHs_yPf", "yPf") ~ "Cell",
                                mthds_union_slim == "yHs" & hs_umi_count < 100 & pf_umi_count > 100 ~ "Cell",
                                if_any(matches("^(cb_|cr$)"), ~ .x == "Cell") ~ "other_calls",
                                if_any(matches("^(cb_|cr$)"), ~ .x == "nonCell") ~ "nonCell",
                                T ~ "missing"),
           fnl_call2 = case_when(mthds_union_slim2 %in%  c("Cr_yPf_Pf", "Cr_yPf", "Cr_Pf", "yPf_Pf", "yPf", "Pf") ~ "Cell",
                                if_any(matches("^(cb_|cr$)"), ~ .x == "Cell") ~ "other_calls",
                                if_any(matches("^(cb_|cr$)"), ~ .x == "nonCell") ~ "nonCell",
                                T ~ "missing"),
           fnl_call3 = case_when(mthds_union_slim3 %in%  c("Cr_mPf_mHs", "Cr_mPf", "mPf_mHs", "mPf") ~ "Cell",
                                mthds_union_slim3 == "mHs" & hs_umi_count < 100 & pf_umi_count > 100 ~ "Cell",
                                if_any(matches("^(cb_|cr$)"), ~ .x == "Cell") ~ "other_calls",
                                if_any(matches("^(cb_|cr$)"), ~ .x == "nonCell") ~ "nonCell",
                                T ~ "missing"))
)
         
      
```

```{r, eval=T}
# saveRDS(cr_cbender_mdata_slct_uni, "data/processed/Pf/m2_all/cr_cbender_mdata_slct_uni.RDS")
imap(cr_cbender_mdata_slct_uni, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/cr_cbender_mdata_slct_uni.RDS")))

```

```{r}
cr_cbender_pf100_mdata_slct_uni <- map(cr_cbender_mdata_slct_uni, ~.x %>% filter(pf_umi_count >= 100))
go()    
```

```{r}
map(cr_cbender_pf100_mdata_slct_uni[19], ~.x %>% filter(fnl_call2 == "missing"))
```

```{r}
map(cr_cbender_pf100_mdata_slct_uni[1:2], ~.x %>% filter(fnl_call == "nonCell"))
```

```{r}
map(cr_cbender_pf100_mdata_slct_uni, ~.x %>% count(mthds_union, mthds_union_slim, fnl_call))
```



```{r, fig.width=12, fig.height=14}

call_methods_overlap_barplot_fn(cr_cbender_pf100_mdata_slct_uni, "fnl_call")
```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
tbl = bind_rows(map(cr_cbender_pf100_mdata_slct_uni, ~filter(., fnl_call == "Cell")), .id = "donor") 

map(c("gn", "umi"), ~call_methods_overlap_scatter_fn(tbl = tbl, gn_or_umi = .x, mthd = "mthds_union_slim"))

```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
tbl = bind_rows(map(cr_cbender_pf100_mdata_slct_uni, ~filter(., mthds_union_slim == "yHs" & fnl_call == "Cell")), .id = "donor") 

map(c("gn", "umi"), ~call_methods_overlap_scatter_fn(tbl = tbl, gn_or_umi = .x, mthd = "mthds_union_slim"))

```

```{r}
## Save first pass cellbender and cellranger cells
# saveRDS(cr_cbender_pf100_mdata_slct_uni, "data/processed/Pf/m2_all/cr_cbender_pf100_mdata_slct_uni.RDS")
# imap(cr_cbender_pf100_mdata_slct_uni, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/cr_cbender_pf100_mdata_slct_uni.RDS")))
imap(cr_cbender_pf100_mdata_slct_uni, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/cr_cbender_pf100_mdata_slct_uni3.RDS")))


```


```{r}
cr_cbender_pf100_mdata_slct_uni_m49c <- readRDS("data/processed/Pf/MSC49C/seu_obj/cr_cbender_pf100_mdata_slct_uni.RDS")

```


```{r}
table(cr_cbender_pf100_mdata_slct_uni_m49c$fnl_call2)
table(cr_cbender_pf100_mdata_slct_uni[['MSC49C']]$fnl_call2)
table(cr_cbender_pf100_mdata_slct_uni[['MSC49C']]$fnl_call3)
```


######################################START - Write souporcell barcodes################################

```{r}

resc_bcodes_hspf <- map(cr_cbender_pf100_mdata_slct_uni, ~{
  .x  %>%
    filter(fnl_call %in%  c("Cell")) %>% 
    pull(barcode) 
  })  

resc_bcodes_hspf2 <- map(cr_cbender_pf100_mdata_slct_uni, ~{
  .x  %>%
    filter(fnl_call2 %in%  c("Cell")) %>% 
    pull(barcode) 
  })  

resc_bcodes_hspf3 <- map(cr_cbender_pf100_mdata_slct_uni, ~{
  .x  %>%
    filter(fnl_call3 %in%  c("Cell")) %>% 
    pull(barcode) 
  })  
```

```{r}
map(resc_bcodes_hspf, length) %>% bind_rows(., .id = "donor")
```

```{r}
map(resc_bcodes_hspf2, length) %>% bind_rows(., .id = "donor")
```

```{r}
map(cr_cbender_pf100_mdata_slct_uni, ~.x %>% count(mthds_union_slim2, fnl_call2))
```

```{r, message=F, warning=F, eval=T}
# gogo()
```
Write out barcodes to rerun souporcell including all possible barcode that are potentially actual cells

```{r}

imap(resc_bcodes_hspf, ~{
  system(paste0('mkdir data/processed/Pf/', .y, '/soupc_resc_hspf'))
  
  .x %>% write(., gzfile(paste0('data/processed/Pf/', .y, '/soupc_resc_hspf/barcodes.tsv.gz')))
  }) 
```

```{r}

imap(resc_bcodes_hspf2, ~{
  system(paste0('mkdir data/processed/Pf/', .y, '/soupc_resc_hspf2'))
  
  .x %>% write(., gzfile(paste0('data/processed/Pf/', .y, '/soupc_resc_hspf2/barcodes.tsv.gz')))
  }) 
```

```{r}

imap(resc_bcodes_hspf3, ~{
  system(paste0('mkdir data/processed/Pf/', .y, '/soupc_resc_hspf3'))
  
  .x %>% write(., gzfile(paste0('data/processed/Pf/', .y, '/soupc_resc_hspf3/barcodes.tsv.gz')))
  }) 
```
Run souporcell like below for the filtered barcodes of pf in mixed infections

```{r, eval=T}
gogo()
# bsub < /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code_lnk/bash_scripts/soupc_v25_m2_bsub.sh
# --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf_all_genes/5736STDY*/outs/possorted_genome_bam.bam" \
# --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/soupc_resc_hspf/barcodes.tsv.gz" \
# --soup_dir "soupc_resc_hspf" \
```


######################################END - Write souporcell barcodes################################


###################################### START - Plot knee ################################

```{r}
# Add emptydrops metadata
all_raw_cbcr_mdata <- pmap(list(cr_cbender_pf100_mdata_slct_uni, all_msc_bp_raw_noHsGns_mdata[names(cr_cbender_pf100_mdata_slct_uni)]), ~{
  ..1 %>% 
    mutate(cell_probability_pf = cell_probability_pfLRhiElo,
           cell_probability_hs = cell_probability_hsLRhiElo) %>% 
    select(barcode, is_cell_cr, is_cell_cb_hs, is_cell_cb_pf, cell_probability_pf, cell_probability_hs, mthds_union, fnl_call) %>%
    full_join(..2 %>% rownames_to_column("barcode"), by = "barcode") %>% 
                           filter(nCount_RNA > 1) %>%
                           column_to_rownames("barcode")  %>%
                           mutate(across(c(contains("is_cell"), "fnl_call", "mthds_union"), ~replace_na(as.character(.), "empty_bc")))
    
})

```


```{r, eval=FALSE}
## read manual thresholds
edrops_thresh <- readRDS("data/processed/Pf/m2_all/man_thresh.RDS")
```

```{r, eval=T}
## read edrops thresholds
yascp_cbender_cutoffs_all_ncells_ndrops <- readRDS("data/processed/Pf/m2_all/yascp_cbender_cutoffs_all_ncells_ndrops.RDS")
```

```{r, eval=T}
yascp_cbender_cutoffs_all_ncells_ndrops %>% glimpse

```

```{r, fig.height= 4, fig.width= 4, eval=FALSE}
yascp_ncells_estimates <- yascp_cbender_cutoffs_all_ncells_ndrops %>% filter() %>% select(sample, umi_counts_cutoff_total_droplets_knee, n_cells_total_droplets_estimated_ndroplets, n_cells_cell_estimate_estimated_ncells) %>% separate_wider_regex(sample, c(donor = ".*", "_" , species = "(?:rmHs|wHsPf)"), names_repair = "unique", cols_remove = TRUE) %>%
 pivot_wider(names_from = species, values_from = c(umi_counts_cutoff_total_droplets_knee, n_cells_total_droplets_estimated_ndroplets, n_cells_cell_estimate_estimated_ncells)) %>%
  rename_with(~ str_remove_all(.x, "_cutoff_total_droplets|n_cells_total_droplets_estimated_|n_cells_cell_estimate_estimated_"), everything())
```

```{r, fig.height= 4, fig.width= 4, eval=FALSE}
all_rank_ncount_cb_hspf <- map(all_raw_cbcr_mdata, ~{
  .x %>% 
    # filter(!is.na(is_cell_cb_hs) & !is.na(is_cell_cb_hs) ) %>%
    mutate(rank = rank(desc(nCount_RNA), ties.method = "average")) %>% 
    filter(!(duplicated(rank) & is_cell_cb_hs %in% c("empty_bc", "nonCell") & is_cell_cb_pf %in% c("empty_bc", "nonCell")))
}) %>%
  bind_rows(., .id = "donor") %>%
  left_join(yascp_ncells_estimates, by = "donor")


```


```{r, fig.height= 4, fig.width= 4, eval=FALSE}
all_rank_ncount_cb_hspf_mrg <- map(all_raw_cbcr_mdata, ~{
  .x %>% 
    # filter(!is.na(is_cell_cb)) %>%
    mutate(rank = rank(desc(nCount_RNA), ties.method = "average"),
           pf_rank = rank(desc(pf_umi_count), ties.method = "average"),
           hs_rank = rank(desc(hs_umi_count), ties.method = "average")) #%>% 
    # filter(!(duplicated(rank) & duplicated(pf_rank) & duplicated(hs_rank) ))
}) %>%
  bind_rows(., .id = "donor") %>%
  left_join(yascp_ncells_estimates, by = "donor")


```


```{r, fig.height= 10, fig.width= 10, eval=F}
all_rank_ncount_cb_hspf_mrg %>%
  filter(!(duplicated(pf_rank) & is_cell_cb_hs %in% c("empty_bc", "nonCell") & is_cell_cb_pf %in% c("empty_bc", "nonCell")), .by = "donor") %>%
  kneeplt_facet_fn(ncount_tbl = ., cell_call_methd = "fnl_call", mthd_col = cell_call_mthd_col, donor_group = "donor", numbr_of_cols = 5, rank_var = "pf_rank", cell_var = "pf_umi_count", cut_offs = c(50, 100), thresh = "umi_counts_knee_rmHs")

```


```{r, fig.height= 10, fig.width= 10, eval=T}
# kneeplt_facet_fn(ncount_tbl = all_rank_ncount_cb_hspf_mrg %>% filter(!(duplicated(pf_rank)), .by = "donor") , 
#                  cell_call_methd = "donor", donor_group = "donor", numbr_of_cols = 5, rank_var = "hs_rank", cell_var = "hs_umi_count")

```


```{r, fig.height= 10, fig.width= 10, eval=F}
all_rank_ncount_cb_hspf_mrg %>% 
  filter(nCount_RNA > 5) %>%
  filter((duplicated(pf_rank) & duplicated(hs_rank) & (duplicated(rank))), .by = "donor") %>%
  arrange(rank)

```

```{r, fig.height= 10, fig.width= 10, eval=T}
# all_rank_ncount_cb_hspf_mrg %>% 
#   filter(!(duplicated(pf_rank) & duplicated(hs_rank) & (duplicated(rank))), .by = "donor") %>%
#   rename("all_rank" = rank) %>%
#   rename("all_umi_count" = nCount_RNA) %>%
#   pivot_longer(cols = c(pf_rank, hs_rank, all_rank, pf_umi_count, hs_umi_count, all_umi_count), names_to = c("species", ".value"), names_pattern = "(pf|hs|all)_(rank|umi_count)") %>%
#   kneeplt_facet_fn(ncount_tbl = . , donor_group = "donor", numbr_of_cols = 5, rank_var = "rank", cell_var = "umi_count", cell_call_methd = "species", mthd_col = c("pf" = "#007BFF", "all" = "#FFC107", "hs" = "#b2d8d8"), cut_offs = c(20,40,50))

```
######################################END - Plot knee ################################


```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")

# cr_cbender_mdata_uni[["MSC12"]] %>% filter(is.na(mthds_union))
```

```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, fig.width=10, fig.height=12}

map(cr_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(cr_cbender_mdata_uni[1:4], ~.x %>% filter(is.na(mthds_union)))

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 4), ncol = 1))
})

```

```{r}
cr_cbender_pf100_mdata_uni <- map(cr_cbender_mdata_uni, ~.x %>% filter(pf_umi_count >= 100))
```

```{r}
# saveRDS(cr_cbender_pf100_mdata_uni, "data/processed/Pf/m2_all/cr_cbender_pf100_mdata_uni.RDS")
```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_pf100_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 3), ncol = 1))
})

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(cr_cbender_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  filter(n > 100) %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 5, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")



```



```{r, fig.width=10, fig.height=12}
map(cr_cbender_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, fig.width=10, fig.height=5}
map(cr_cbender_pf100_mdata_uni, ~.x %>% 
      filter(!(mthds_union %in% c("Cr", "Cbnoncell"))) %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cr")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(cr_cbender_pf100_mdata_uni, .id = "donor") %>%
    filter(!(mthds_union %in% c("Cbycr", "nonCell"))) %>%
    mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cbycr")) %>% 
    filter(!(mthds_union_CbCr %in% c("CrCbhsCbpf", "CbhsCbpf"))) %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union_CbCr)) +
    geom_point(size = 0.5) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    theme_classic() +
    guides(color = guide_legend(title = 'Cell call', override.aes = list(size = 3), ncol = 1))
})

```

```{r, fig.width=10, fig.height=5}
map(cr_cbender_pf100_mdata_uni, ~.x %>% 
      filter(!(mthds_union %in% c("Cbpf", "Cbnoncell"))) %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cbpf")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, fig.width=10, fig.height=5}
map(cr_cbender_pf100_mdata_uni, ~.x %>% 
      filter(!(mthds_union %in% c("Cbpf", "Cbnoncell"))) %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Cbpf")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


#########################################################################START## Need to investigate what happens here"
```{r, message=F, warning=F, eval=T}
gogo()
```

```{r, message=F, warning=F, eval=F}
##Read BPcells matrices
# cbender_mdata_std_yascp <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender|_gns|/cb_mdata.*'))) %>% 
#   .[c(sample_name, sample_name_rmHs, sample_name_yascpCranger, sample_name_LRate_mod)] %>%
#   .[!is.na(.)] %>%
#   map(read_csv) 
```


```{r, message=F, warning=F, eval=F}
##Read BPcells matrices
sample_name_yascpCranger_nms <- names(cbender_mdata_std_yascp)[str_detect(names(cbender_mdata_std_yascp), "yascpCr")]
sample_name_yascpCranger_nms <- sample_name_yascpCranger_nms[!str_detect(sample_name_yascpCranger_nms, "MSC50_SLO_yascpCr")]
# sample_sufx_cb <- c("_hsLRloEhi", "_hsLRloElo", "_hsLRhiEhi", "_hsLRhiElo", "_pfLRloEhi", "_pfLRloElo", "_pfLRhiEhi", "_pfLRhiElo")

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
cbender_std_yascpCr_mrg_syst_don <- cbender_std_yascpCr_mrg_syst %>%
  bind_rows(., .id = "donor")# %>%
  # filter(!(cell_probability_lpf == 1 & cell_probability_ypf ==1 & cell_probability_lhs ==1 & cell_probability_yhs ==1))

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by standard cellbender
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>%
  filter(cell_probability_pfLRloEhi > 0.9) %>%
  # filter(cell_size_yascp >= 100) %>%
  count(donor)

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by using the pf+Hs gene counts
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>%
  filter(cell_probability_pfLRhiElo < 0.9 & cell_probability_pfLRloEhi > 0.9) %>%
  filter(cell_probability_pfLRloEhi >= 100) %>%
  count(donor)

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
cbender_std_yascpCr_mrg_syst_don %>%
  ggplot(aes(x=cell_probability_pfLRhiElo, y = cell_probability_pfLRloEhi)) +
  geom_point(size = 0.0001) +
  geom_density2d() +
  facet_wrap(. ~ donor)

```


```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by using the pf+Hs gene counts
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>%
  filter((cell_probability_hsLRloEhi < 0.9 & cell_probability_pfLRloEhi > 0.9) | (is.na(cell_probability_hsLRloEhi)  & cell_probability_pfLRloEhi > 0.9)) %>%
  # filter(cell_size_pfLRloEhi >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
## Count pf cells that are potentially missed by using the pf+Hs gene counts
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>%
  filter(cell_probability_hsLRloEhi < 0.9 & cell_probability_pfLRloEhi > 0.9) %>%
  filter(cell_size_pfLRloEhi >= 100) %>%
  count(donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>%
  filter(!(cell_probability_hsLRloEhi == 1 & cell_probability_pfLRloEhi ==1)) %>%
  ggplot(aes(x=cell_probability_hsLRloEhi, y = cell_probability_pfLRloEhi)) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  theme_bw() +
  facet_wrap(. ~ donor)

```

```{r, message=F, warning=F, fig.height=10, fig.width=10, eval=T}
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>%
  ggplot(aes(x=cell_size_hsLRloEhi, y = cell_size_pfLRloEhi)) +
  geom_point(size = 0.01) +
  geom_abline() +
  facet_wrap(. ~ donor)

```


```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp_syst, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_probability)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```


```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp_syst, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_probability)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```


```{r, message=F, warning=F, eval=T, fig.height=9, fig.width=7}
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>% 
      filter(cell_probability_pfLRloEhi > 0.8) %>%
  group_by(donor) %>% 
  mutate(bg_50 = sum(background_fraction_pfLRloEhi > 0.5)) %>% 
  ungroup() %>%
  ggplot(aes(x = background_fraction_pfLRloEhi)) +
    geom_histogram() +
    geom_text(aes(label=bg_50, x = 0.8, y = 200),  size = 3)+
    geom_vline(xintercept = 0.5) +
    facet_wrap(donor ~ ., scales = "free_y", ncol = 5)

```

```{r, message=F, warning=F, eval=T, fig.height=8, fig.width=12}
cbender_std_yascpCr_mrg_syst_don %>%
  # bind_rows(., .id = "donor") %>% 
      filter(cell_probability_pfLRloEhi > 0.8) %>%
  count(donor, background_fraction_pfLRloEhi>0.5)

```

```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp_syst, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_size)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```

```{r, message=F, warning=F, eval=T}
imap(cbender_mdata_std_yascp_syst, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = cell_probability, y = cell_size)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.9) +
    labs(title = .y)
)

```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_cbender_mdata_pass.5 <- map(cbender_mdata_std_yascp_syst, ~.x %>% filter(cell_probability>0.5) %>% pull(barcode))

```



#########################################################################END## Need to investigate what happens here"
