---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
})
```







# MSC Preprocessing


## setup, read input files and exploratory visualizations and general functions


```{r, setup}
# library(reticulate)
# # use_condaenv("/Users/jr35/envs/scv_jupcon/", required = TRUE)
# # matplotlib <- reticulate::import("matplotlib")
# # matplotlib$use("Agg", force = TRUE)
# 
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali')
```

### Variable declarations

```{r}

##Strain cluster colours
strain_cols_n <-c("#910000", "#ff1a8d", "#007c71", "#0ea300", "#00deca", "#9a96c7", "#5d57a4", "#a8875b","#708db3", "#c39a00", "#ffd94d", "#9770b3", "#C9E0F1", "#a40000" , "#00b7a7", "#ffcd12","thistle","#a40000" , "#00b7a7", "#e7f1f9", 'black', '#050301','grey')
#, 'lavender', 'lightgray', 'grey'
names(strain_cols_n) <- c('SC2', 'SC1','SC5',  'SC4', 'SC3',  'SC6', 'SC7', 'SC8', 'SC9', 'SC10', 'SC11', 'SC12', 'S3_b', 'S1', 'S2', 'Singlet', 'Negative', 'SC1_3i','Doublet', 'Lab', 'Less5', 'PoorQC', NA)

strain_cols <- c(met.brewer(name = 'Thomas')[c(1:8)],'black', 'light gray')

names(strain_cols) = c('G1', 'G2', 'G3', 'G4','G5', 'G6','G7', 'G8', 'Doublet', 'Negative')

strain_cols =c(strain_cols,strain_cols_n)

dblt_cols <- c(brewer.pal(name = 'Set1', n = 8)[c(1:5,7:8)], 'lightblue')
names(dblt_cols) <- c('ScDfSt', 'ScDf', 'ScSt', 'DfSt', 'Sc', 'Df','St', 'Singlet')

cell_qc_filt_nm_lvls <- c("sc_neg", "lo_gn_cnt",  "hi_gn_cnt",  "hi_mtp", "ss_db",  "strn_db", "stg_db", "pass")
cell_qc_cols <- c(brewer.pal(name = 'Set1', n = 8)[c(1:5,7:8)], 'lightblue') %>% set_names(cell_qc_filt_nm_lvls)

dblt_ss_cols <- c(brewer.pal(name = 'Set1', n = 8)[c(1:3)], 'lightgrey')
names(dblt_ss_cols) <- c('Strain & stage', 'Strain', 'Stage', 'Singlet')

cluster_cols <- brewer.pal(8, name = 'Set2')[c(1:4,7,8)] %>% set_names(as.character(c(0:5)))

stg_refnd_lvls = c("Early ring", "Late ring", "Early trophozoite", "Late trophozoite", "Early schizont", "Late schizont", "Gametocyte (committed)", "Gametocyte (developing)", "Gametocyte (branching)", "Female (HE)", "Female (LE)", "Gametocyte (female)", "Gametocyte (early female)", "Gametocyte (late female)", "Gametocyte (male)", "Male (HE)", "Male (LE)", "Gametocyte (early male)", "Gametocyte (late male)", "Unassigned", "Failed qc", "Not assigned", "Atlas")

sample_name_nms <- c("MSC1", "MSC3","MSC13","MSC14","MSC1272") %>% str_sort(numeric = T)


sp_colors <- c('Early ring'='#D1EC9F',
                     'Late ring'='#78C679',
                     'Early trophozoite'='#FEEEAA',
                     'Late trophozoite'='#FEB24C',
                     'Early schizont'='#C9E8F1',
                     'Late schizont'='#85B1D3',
                     'Gametocyte (developing)'='thistle',
                     'Gametocyte (female)'='#551A8B',
                     'Gametocyte (male)'='mediumpurple',
               'Unassigned'='red', 
               'Not assigned'='#D6CFC7',
               'Female (LE)'='#208cf7',
               'Female (HE)'='#551A8B',
               'Male (LE)'='#a894d1',
               'Male (HE)'='mediumpurple',
               'Failed QC' = 'grey',
               'Gametocyte (committed)' = '#C399A2', 
               'Gametocyte (developing)' = '#66444C', 
               'Gametocyte (branching)' = '#66444C', 
               'Gametocyte (early female)' = '#749E89', 
               'Gametocyte (early male)' = '#7D87B2', 
               'Gametocyte (late female)' = '#4E6D58', 
               'Gametocyte (late male)' = '#41507B'
)

sp_fld_colors <- c('Late ring'= '#78C679',#met.brewer('Java')[2],
                   'Early trophozoite'='#FEEEAA',#met.brewer('Java')[3],
                   'Female (LE)'= met.brewer('Ingres')[5],
                   'Female (HE)'= '#551A8B', #met.brewer('Ingres')[7],
                   # 'Male (LE)'=met.brewer('Nizami')[6],
                   # 'Male (HE)'=met.brewer('Nizami')[8],
                   # 'Gametocyte (male)'=met.brewer('Austria')[5],
                   'Male (LE)'=met.brewer('Ingres')[7],#'#a2e19e',
                   'Male (HE)'= 'mediumpurple',#'#039911',
                   'Gametocyte (male)'='#3DED97',
                   'Gametocyte (female)'=met.brewer('Ingres')[6],
                   'Lab' = '#D3D3D3',
                   'Unassigned'='#DCDCDC',
                   'Failed QC' = 'grey',
                   'Not assigned' = 'grey',
                   'Atlas' = 'grey',
                   'Early ring'='#D1EC9F',
                     'Late trophozoite'='#FEB24C',
                     'Early schizont'='#C9E8F1',
                     'Late schizont'='#85B1D3',
                   'Gametocyte (developing)'='thistle',
                   'early female' = '#749E89',
                   'late female'= '#4E6D58'
)

donor_cols_m = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" ), '.mrna'))
# donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))
donor_cols = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))


```


### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```

### Common functions
#### Average expression of gene sets

Function 
```{r}
makeSetScore <- function(seur, gene.set,assay='RNA') {
  # Get mean expression of genes of interest per cell
  mean.exp <- colMeans(x = as.matrix(GetAssayData(seur, slot = 'data', assay = assay))[gene.set, ], na.rm = TRUE)
  
  # Add mean expression values in 'object@meta.data$gene.set.score'
  if (all(names(x = mean.exp) == rownames(x = seur@meta.data))) {
    return(mean.exp)
  }
}

```

#### Seurat normalization function from Elias
```{r}

seurat_norm_stand = function(Seurat_object, nHVG = 500){
  Seurat_object = NormalizeData(Seurat_object, verbose = F)
  Seurat_object = FindVariableFeatures(Seurat_object, nfeatures = nHVG, verbose = F)
  Seurat_object = ScaleData(Seurat_object, verbose = F)
  Seurat_object = RunPCA(Seurat_object, verbose = F)
  return(Seurat_object)
}
```

```{r}
##Boxplot labeling function
stat_box_data <- function(y) {
  return( 
    data.frame(
      y=quantile(y,probs=1)*1.4,  #may need to modify this depending on your data
      label = paste(
        'n:', length(y), '\n',
        'mu:', round(mean(y), 1)
      )
    )
  )
}

```

initialize variable for the sample names

```{r}
# sample_id <- c("MSC1", "MSC3", "MSC1272", "MSC13", "MSC14" )
sample_name <-c("MSC1", "MSC3", "MSC1272", "MSC13", "MSC14" ) %>% str_sort(., numeric = T)
mt_filt <- c(5,1.2,0.9,0.45,2)

# source_irods <- c("5736STDY11771545", "5736STDY11771544", "5736STDY11771536", "5736STDY11771546", "5736STDY11771535")
# sample_name <- c("MSC1272", "MSC13", "MSC3", "MSC14", "MSC1")
```


Save paths to raw MSC files in list - also needed for soupX below
```{r}
all_msc_raw_paths <- list.files("data/raw", pattern = "raw_feature_bc_matrix$", full.names = T, recursive = T, include.dirs = T) %>% 
  str_sort(., numeric = T) %>%
  .[str_detect(., 'DB52e_star')]
```

MSC raw matrices BP
```{r}
## List filtered cellranger h5 matrices into R
list.files("data/raw/", pattern = "raw_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
  str_sort(., numeric = T) %>%
  .[str_detect(., 'DB52e_star')]
```

```{r}
## Write Pf barcodes for mixed species donor for souporcell
map(sample_name, ~{
    system(paste0("mkdir -p data/processed/Pf/", .x, "/bpcells/raw/counts_mtx"))
    system(paste0("mkdir -p data/processed/Pf/", .x, "/bpcells/filtered/counts_mtx"))
    # system(paste0("rmdir data/processed/Pf/", .x, "/bpcells/counts_mtx"))
  })

```

```{r}
## List filtered cellranger h5 matrices into R
list.files("data/raw/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
  str_sort(., numeric = T) %>%
  .[str_detect(., 'DB52e_star')]
```

```{r, message=F, warning=F, eval=T}
##Write BPcells matrices
map(c("raw", "filtered"), ~{
  dset = .x
  list.files("data/raw/", pattern = paste0(dset, "_feature_bc_matrix.h5$"), full.names = T, recursive = T, include.dirs = T) %>% 
  str_sort(., numeric = T) %>%
  .[str_detect(., 'DB52e_star')] %>%
  set_names(nm = (str_remove_all(.,'data/raw//|/5736STDY.*'))) %>% 
  .[sample_name] %>%
  map(~open_matrix_10x_hdf5(path = .)) %>% 
  imap(., ~write_matrix_dir(mat = .x, dir = paste0('data/processed/Pf/', .y, '/bpcells/',dset,'/counts_mtx'), overwrite = T))
})
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_bp_raw <- list.files("data/processed/Pf", pattern = "counts_mtx$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/bpcells.*'))) %>%
  .[str_detect(., 'raw')]  %>% 
  .[sample_name] %>%
  map(~open_matrix_dir(dir = .)) %>% 
  imap(., ~CreateSeuratObject(counts = .x, project = paste0(.y, ".mrna"))) 
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_bp_filt <- list.files("data/processed/Pf", pattern = "counts_mtx$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/bpcells.*'))) %>%
  .[str_detect(., 'filtered')]  %>% 
  .[sample_name] %>%
  map(~open_matrix_dir(dir = .)) %>% 
  imap(., ~CreateSeuratObject(counts = .x, project = paste0(.y, ".mrna"))) 
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_raw_mdata_all <- map2(all_msc_bp_raw, all_msc_bp_filt, ~.x@meta.data %>% rownames_to_column("bcode") %>% mutate(cr_filt = case_when(bcode %in% rownames(.y@meta.data) ~ "Yes", T ~ "No"))) 
```


Find gene count cut offs
```{r, results='asis'}
library(scales)
strn_knee_p <- imap(all_msc_raw_mdata_all, ~{
  dset = .x
  don = .y
  head(dset)
  map2(c("nFeature_RNA", "nCount_RNA"), c("Gene", "Transcript"), ~{
    dset %>% 
      arrange(desc(!!rlang::sym(.x))) %>%
      mutate(rank = row_number()) %>% 
      ggplot(., aes(y = !!rlang::sym(.x), x= rank, colour = cr_filt)) +
      ggrastr::rasterise(geom_point(size=2), dpi=1000) +
      geom_hline(yintercept = 200) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
      # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=20), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=20,face="bold"),
            legend.position = "none"
    )+ 
      annotation_logticks() +
      labs(x= 'Cell rank', y= paste0(.y, 'count')#,
           #title = don
           )
    
})
})
```


```{r, fig.height= 4, fig.width= 4}
strn_knee_p
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_raw_mdata <- map2(all_msc_bp_raw, all_msc_bp_filt, ~.x@meta.data %>% filter(nFeature_RNA > 10) %>% rownames_to_column("bcode") %>% mutate(cr_filt = case_when(bcode %in% rownames(.y@meta.data) ~ "Yes", T ~ "No"))) 
```




```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_bp_150 <- list() 
```


```{r, eval=T}
##write out for normalization processing in farm
for (i in sample_name) {
## Subset those passing gene count cut off
  print(i)
  all_msc_bp_150[i] <- subset(all_msc_bp_raw[[i]], subset = nFeature_RNA >= 150)
}

```


```{r, eval=T}
all_msc_bp_150

```

```{r, message=F, warning=F, eval=F}
# Convert to dense object - might be necessary in future instances
# all_msc <- all_msc_bp %>%
#   map(., ~{
#     .x[["RNA"]]$counts <- as(object = .x[["RNA"]]$counts, Class = "dgCMatrix")
#     return(.x)
#   })

```


```{r, eval=T}
##write out for normalization processing in farm

for (i in 1:length(sample_name)) {
## Create new directories
  system(paste0('mkdir -p data/processed/Pf/', sample_name[i], '/seu_obj'))
}


```

FARM CODE

```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
### filt
imap(all_msc_bp_filt, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu.RDS")))

### gn count cutoff 150
imap(all_msc_bp_150, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_150_seu.RDS")))
```

!!!NB - FARM CODE - DONT DELETE

```{r, eval=T}
gogo()
## Run standard normalisation on farm like below
## !!NB - reducing the number of hypervariable genes results in scattered and disjointed clusters. higher hvg number better

# nextflow run submsn_seu_norm.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/MSC*/seu_obj/bpc_seu.RDS" --o_file_name "seu_norm.RDS" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --elbo_nm "seu_norm_elbo.RDS"

# nextflow run submsn_seu_norm.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/MSC*/seu_obj/bpc_150_seu.RDS" --o_file_name "g150_seu_norm.RDS" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --elbo_nm "g150_seu_norm_elbo.RDS"
```

Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read BPcells matrices
sample_name_all <- c(sample_name, paste0(sample_name, "g150"))

```

```{r, message=F, warning=F}
##Read BPcells matrices
all_msc_gn150 <- list.files("data/processed/Pf", pattern = "*seu_norm.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|seu.*'))) %>% 
  .[sample_name_all] %>%
  map(readRDS)

```


Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read elbo plots
all_msc_gn150_elbo <- list.files("data/processed/Pf", pattern = "*seu_norm_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|seu.*'))) %>% 
  .[sample_name_all] %>%
  map(readRDS)

all_msc_gn150_elbo
```

```{r, warning=FALSE, message=FALSE}
imap(all_msc_gn150, ~DimPlot(.x, dims = c(1,2), label = T) + labs(title = .y))
```


## scmap annotation

### Preprocessing references

#### msc_combi reference dataset
```{r, eval=FALSE}
##Read MCA V3 blood stage reference dataset
# msc_combi <- readRDS("../sc_velo/data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.4.RDS")
raw_combined_cut_0.4_nodoublets_stagenew <- readRDS("data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")
# msc_combi <- readRDS("../Mali/data/processed/Pf/msc_combi.RDS")
DefaultAssay(raw_combined_cut_0.4_nodoublets_stagenew) <- "RNA"

## !! IMPORTANT - Create a new stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

```

```{r, message=FALSE, warning=FALSE, eval=FALSE}
##Convert to reference to sce
raw_combined_cut_0.4_nodoublets_stagenew.sce <- as.SingleCellExperiment(raw_combined_cut_0.4_nodoublets_stagenew)
# raw_combined_cut_0.4_nodoublets_stagenew.sce@colData[,"stageHL_ref"] <- raw_combined_cut_0.4_nodoublets_stagenew.sce@colData[,"stagenewx"]

##Place counts appropriately into single cell experiment object https://github.com/hemberg-lab/scmap/issues/9
# SingleCellExperiment::logcounts(raw_combined_cut_0.4_nodoublets_stagenew.sce) <- as(log2(SingleCellExperiment::counts(raw_combined_cut_0.4_nodoublets_stagenew.sce) + 1), "sparseMatrix")
rowData(raw_combined_cut_0.4_nodoublets_stagenew.sce)$feature_symbol <- rownames(raw_combined_cut_0.4_nodoublets_stagenew.sce)
# raw_combined_cut_0.4_nodoublets_stagenew.sce <- raw_combined_cut_0.4_nodoublets_stagenew.sce[!duplicated(rownames(raw_combined_cut_0.4_nodoublets_stagenew.sce)), ]

# library(Matrix)
# normcounts(raw_combined_cut_0.4_nodoublets_stagenew.sce)
```



```{r, message=FALSE, warning=FALSE, eval = F}
scater::plotReducedDim(raw_combined_cut_0.4_nodoublets_stagenew.sce, dimred = "PCA", colour_by = "stagenewx", ncomponents = c(2,3))
```


```{r, message=FALSE, warning=FALSE, eval = F}
saveRDS(raw_combined_cut_0.4_nodoublets_stagenew.sce, "data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.sce.RDS")
```


```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run submsn_scmap.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.sce.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/MSC*/seu_obj/seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stagenewx"

# nextflow run submsn_scmap.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.sce.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/MSC*/seu_obj/g150_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "g150_prelim_anotd.RDS" --stg_labl "stagenewx"
```


```{r, message=F, warning=F}
##Read the annotations
m21_gn150_anots <- list.files("data/processed/Pf", pattern = "*prelim_anotd.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_pre.*|pre.*'))) %>% 
  .[sample_name_all] %>%
  .[!is.na(.)] %>%
  map(readRDS)
```

```{r}
## Adding annotations to query cells
m21_gn150_anotd_preQC <- list(m21_gn150_anots, all_msc_gn150) %>%
  pmap(~ ..1 %>%
         column_to_rownames("cell_bc") %>%
         mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
         mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
         mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
         AddMetaData(..2, .) 
  )

```

```{r, warning=FALSE, message=FALSE}

imap(m21_gn150_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'Stage_MSC', cols = sp_fld_colors) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

imap(m21_gn150_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'stage_lab1', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}

map(m21_gn150_anotd_preQC, ~..1@meta.data %>% dplyr::count(Stage_MSC, stage_lab1))

```


```{r}
##Modify the stage labels for MSC14
# m21_gn150_anotd_preQC[[4]] <- m21_gn150_anotd_preQC[[4]]@meta.data %>%
m21_gn150_anotd_preQC <- pmap(list(m21_gn150_anotd_preQC, rep(0.4,length(sample_name_all)), c(rep('ring|trophozoite|developing', length(sample_name_all)))), ~{
  ..1@meta.data %>%
    mutate(across(Stage_MSC, as.character)) %>%
    # mutate(Stage_MSC = Stage_MSC) %>%
    mutate(StagePrelim_lf = case_when(
      is.na(Stage_MSC) & sim_lab1 > 0.35 & str_detect(stage_lab1, ..3) & str_detect(stage_lab2, ..3) & str_detect(stage_lab3, ..3) ~ stage_lab1, ## lower cos sim of 0.35 for asexuals instead of 0.4
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Male|male') & str_detect(stage_lab2, 'Male|male') & str_detect(stage_lab3, 'Male|male') ~ stage_lab1,
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Female|female') & str_detect(stage_lab2, 'Female|female') & str_detect(stage_lab3, 'Female|female') ~ stage_lab1,
      TRUE ~ Stage_MSC)#,
      # across(StagePrelim, str_to_sentence)
    )%>%
    mutate(StagePrelim = case_when(
      str_detect(StagePrelim_lf, "developing") ~ "Early trophozoite", ## !!NB - label developing gams as early trophs the closest stage
      str_detect(StagePrelim_lf, "schizont") & str_detect(stage_lab2, "ring|trophozoite")  & str_detect(stage_lab3, "ring|trophozoite") ~ str_to_sentence(stage_lab2),
      str_detect(StagePrelim_lf, "late|early") ~ str_to_sentence(StagePrelim_lf),
      str_detect(StagePrelim_lf, "gametocyte") ~ str_to_sentence(str_remove_all(StagePrelim_lf, "gametocyte \\(|\\)")),
      is.na(StagePrelim_lf) ~ "Unassigned",
      TRUE ~ StagePrelim_lf)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(StagePrelimC = case_when(
      str_detect(StagePrelim, "ring|trophozoite|developing|schizont") ~ "Asexual",
      TRUE ~ StagePrelim)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(across(c(StagePrelim, StagePrelimC), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    dplyr::select(StagePrelim, StagePrelim_lf, StagePrelimC) %>%
    # mutate(across(starts_with('Stage_V'), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    AddMetaData(.x, .)
})

```


```{r, warning=FALSE, message=FALSE}

map(m21_gn150_anotd_preQC, ~.x@meta.data %>% dplyr::count(Stage_MSC, StagePrelim, stage_lab1))
# map(m21_gn150_anotd_preQC, ~.x@meta.data %>% dplyr::count(StagePrelim, StagePrelimC))
```

```{r, warning=FALSE, message=FALSE}
imap(m21_gn150_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = F, group.by = 'StagePrelim', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}
m21_gn150_anots[["MSC51"]]
```


```{r, warning=FALSE, message=FALSE}
imap(m21_gn150_anotd_preQC, ~FeaturePlot(.x, dims = c(1,2), features = "nFeature_RNA")+ labs(title = .y))
```

```{r, results='asis',  fig.width = 8, fig.height = 4}
##adding clusters that are used to display mitochondrial count

plts <- list(m21_gn150_anotd_preQC, names(m21_gn150_anotd_preQC)) %>%
  pmap(., ~{
    
    cell_qc_plot <-  ..1 %>% DimPlot(., pt.size = 1, reduction = 'umap', dims = c(1,2), group.by = 'StagePrelim', cols = sp_fld_colors) +labs(title = ..2)#,
    
    cell_qc_plot$data <- cell_qc_plot$data %>% rownames_to_column('bcode') %>% left_join(., ..1@meta.data[,'seurat_clusters', drop = F] %>% rownames_to_column('bcode'), by='bcode') %>% column_to_rownames('bcode')
    
    # print(str(cell_qc_plot$data))
    
    cell_qc_plot<- LabelClusters(cell_qc_plot, id = "seurat_clusters", size = 9, repel = T)
    
    return(cell_qc_plot)

  })
```


```{r, results='asis',  fig.width = 8, fig.height = 4}
plts

```

compare filt and raw with gn count cut off of 150
```{r, message=F, warning=F}
# go()
##Read the annotations
m21_all_anots_mdata <- map2(m21_gn150_anotd_preQC[1:5], m21_gn150_anotd_preQC[6:10], ~{
  .x@meta.data %>% 
    rownames_to_column("bcode") %>%
    full_join(., .y@meta.data %>% rownames_to_column("bcode"), suffix = c("_filt", "_gn_150"), by = "bcode") %>%
    mutate(dset = case_when(!is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "both",
                            !is.na(nFeature_RNA_filt) & is.na(nFeature_RNA_gn_150)  ~ "filt",
                            is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "gn_150"))
})
```


```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% count(dset))

```

```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% filter(dset == "gn_150") %>% count(StagePrelim_gn_150, Stage_MSC_gn_150, stage_lab1_gn_150, stage_lab2_gn_150, stage_lab3_gn_150))

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = sim_lab1_gn_150) ) +
      geom_boxplot() +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```

Umap with stage labels in legend

```{r, fig.width = 7, fig.height = 5, eval=F}
trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.2, "cm")
)

# Assign the female High_low stage of the projected cell to the nearest cell in the reference dataset
# msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(msc_combi.sce))), length(m21_gn150_jc))
msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(v3labmscs.ref.sce))), length(m21_gn150_jc))


for (i in 1:length(msc_qcd_anotd_mali_mapped14HL)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m21_gn150_anotd_preQC[[i]]$StagePrelim
}


for (i in 1:length(m21_gn150_anotd_preQC)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m21_gn150_anotd_preQC[[i]]$StagePrelim
  # msc_combi@meta.data[, paste0("stage_", names(m21_gn150_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  v3labmscs.ref.sce@colData[, paste0("stage_", names(m21_gn150_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  
}

labs_order <- c('Late ring', 'Early trophozoite','Female (HE)',  'Male (HE)', 'Female (LE)', 'Male (LE)','Gametocyte (developing)', 'Gametocyte (male)','Gametocyte (female)',  'Atlas')
names(labs_order)<- labs_order

umap_theme <- theme(legend.position = 'right',
                    axis.title=element_text(size=12,face="bold",hjust = 0),
                    axis.text=element_text(size=18), 
                    plot.title = element_blank(),
                    legend.title=element_text(size=14,face="bold"),
                    legend.text=element_text(size=14),
                    legend.spacing.y = unit(0.25, "cm"),
                    axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed"))
)  


# Idents(msc_combi) <- 'MSC14_V3'
# m14_scmap_mca_umap <- map(names(m21_gn150_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), 
#         order = labs_order) +
m14_scmap_mca_umap <- map(names(m21_gn150_anotd_preQC), ~{
  cbind(reducedDims(v3labmscs.ref.sce)$PCA, colData(v3labmscs.ref.sce)[, c("stageHL_ref", paste0("stage_",.x)), drop=F]) %>% 
    data.frame() %>% 
    mutate(od_stg = droplevels(factor(!!rlang::sym(paste0("stage_",.x)), levels = rev(stg_refnd_lvls)))) %>%
    arrange(od_stg) %>%
    ggplot(aes(y = PC2, x=PC3, color = od_stg)) +
    geom_point() +
    # scale_color_manual(breaks=labs_order, values = sp_fld_colors, labels = str_wrap(labs_order, 10))+
    scale_color_manual(values = sp_fld_colors)+
    umap_theme+
    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
           x= trunc_axis,
           y= trunc_axis)+
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(breaks = NULL)
})

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap, 'plots/fig4/m14_scmap_mca_umap')

m14_scmap_mca_umap


```

Umap with stage labels inside
```{r, fig.width = 7, fig.height = 5, eval=F}

m14_scmap_mca_umap_labs_in <- map(names(m21_gn150_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), label = T, order = labs_order, repel = T, label.size = 5) +
                                    scale_color_manual(values = sp_fld_colors)+ 
                                    umap_theme+
                                    theme(axis.title=element_text(size=18,face="bold"), 
                                          plot.title = element_blank(),
                                          legend.position='None'
                                    ) +
                                    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
                                           x= trunc_axis,
                                           y= trunc_axis)+
                                    scale_x_continuous(breaks = NULL) +
                                    scale_y_continuous(breaks = NULL)
)

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap_labs_in, 'plots/fig4/m14_scmap_mca_umap_labs_in.RDS')


m14_scmap_mca_umap_labs_in


```


```{r, warning=FALSE, message=FALSE}

map(m21_gn150_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim, Stage_MSC, stage_lab1, stage_lab2))

```


```{r, warning=FALSE, message=FALSE}
saveRDS(m21_gn150_anotd_preQC, "data/processed/Pf/m21_gn150_anotd_preQC.RDS")
```


```{r, warning=FALSE, message=FALSE}

map(m21_gn150_anotd_preQC, ~..1@meta.data) %>% saveRDS(., "data/processed/Pf/m21_gn150_anotd_preQC_mdata.RDS")

```

```{r, warning=FALSE, message=FALSE}

map(m21_gn150_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim))

```



```{r, fig.width = 12, fig.height = 4}

##Plotting top two scmap stages among the top 10 with their similarities for MSC14
# 
# stage_fct_labs <- msc_qcd_anotd[[1]]@meta.data  %>% dplyr::group_by(stageHL, seurat_clusters) %>% dplyr::group_keys() %>% dplyr::mutate(across(c(stageHL, seurat_clusters), as.character), across(seurat_clusters, ~paste0("Cluster ",., ':\n',stageHL))) %>% deframe()
# 
# scm_barpt <- msc_qcd_anotd[[1]]@meta.data %>% 
#   dplyr::select(matches('s.*lab.'), stageHL) %>% 
#   rownames_to_column('cell_bc')%>%
#   pivot_longer(-c('cell_bc', 'stageHL'),
#                names_to = c( ".value", "Label"),
#                names_pattern = "(s.*)_(l.*)")%>% 
#   filter(!is.na(sim)) %>%
#   mutate(across(sim, as.numeric)) %>% 
#   mutate(across(Label, ~str_replace(., 'lab','Label '))) %>%
#   mutate(across(stage, ~droplevels(factor(., stg_refnd_lvls)))) %>%
#   ggplot(., aes(y = sim, x = stage, color = Label)) +
#   geom_boxplot(aes(fill = Label))+
#   scale_fill_manual(values =  c('Label 1' = '#007FFF', 'Label 2' = '#FFFF9F'), name = "Label",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 2)
#   )+
#   scale_color_manual(values =  c('Label 1' = 'black', 'Label 2' = 'black'), name = "Label")+ 
#   geom_point(aes(group=Label), size = 0.5, position = position_jitterdodge(jitter.width = 0.2))+
#   new_scale_fill()+
#   geom_tile(data = data.frame(x = c( 'Late ring', 'Early trophozoite', 'Gametocyte (developing)', 'Gametocyte (female)', 'Gametocyte (male)')), inherit.aes = F, aes(x = x, y = 0.37, fill = x, height = 0.03)) +
#   scale_fill_manual(values =  sp_fld_colors, name = "MCA stage",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 3)
#                     # guide = 'none'
#   )+
#   scale_y_continuous(expand = expansion(mult = c(0.01, 0.05))) +
#   facet_grid(.~stageHL, labeller = labeller(stageHL = stage_fct_labs)) +
#   geom_hline(yintercept = 0.4)+
#   labs(y = 'scmap \nsimilarity scores', x = 'MCA stage') +
#   theme_classic() +
#   theme(axis.text.x = element_blank(),
#         axis.title=element_text(size=16,face="plain"), 
#         axis.text = element_text(size=15), 
#         axis.ticks.x = element_blank(),
#         strip.text =element_text(size=16,face="plain", colour = 'black'),
#         panel.border=element_rect(colour="black",fill='transparent'),
#         legend.title = element_text(size=16,face="bold"),
#         legend.text = element_text(size=15),
#         legend.position = 'bottom')
# 
# scm_barpt
```


Run souporcell like below for the filtered barcodes of all infections
```{r}
gogo()

# nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/MSC*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```



Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
gogo()

#  nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373{63,72,89}/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/Pf_mxd/MSC*/outs/sk21_bcodes_pp/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```
