---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(ComplexHeatmap)
  library(circlize)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  
})
```


```{r setup}
sample_set = "Mali2"

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```


#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```

```{r}
# optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)
# 
# optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]
```


```{r}
# m2_seu_raw_cb_norm_harmony_PC_manl <-  readRDS("data/processed/Pf/m2_all/m2_seu_raw_cb_norm_harmony_PC_manl.RDS")
# all_msc_cb_no_dbt_mrg_noM12_sct_rpca_sct_PC10 <-  readRDS("data/processed/Pf/m2_all/all_msc_cb_no_dbt_mrg_noM12_sct_rpca_sct_PC10.RDS")
# all_msc_cb_no_dbt_mrg_noM12_sct_rpca_sct_PC10@meta.data[,'donor'] <- str_remove(all_msc_cb_no_dbt_mrg_noM12_sct_rpca_sct_PC10@meta.data[,'orig.ident'], ".mrna")
```


```{r}
m2_all_raw_mrg <- readRDS("data/processed/Pf/m2_all/m2_all_raw_mrg.RDS")
```

```{r}
DimPlot(m2_all_raw_mrg, reduction = "umap.rpca_sct_PC10", group.by = "sR_fine_stg", cols = sp_col_new)

```

```{r}
DimPlot(m2_all_raw_mrg, reduction = "umap.rpca_sct_PC10", group.by = "sR_fine_stg", cols = sp_col_new, dims = c(1,2))

```

```{r}
DimPlot(m2_all_raw_mrg, reduction = "umap.rpca_sct_PC10", group.by = "sR_fine_stg", cols = sp_col_new, dims = c(1,3))

```
```{r, warning=FALSE, message=FALSE}
VlnPlot(m2_all_raw_mrg, features = "nFeature_RNA", group.by = "stage", pt.size = 0) + geom_hline(yintercept = c(200, 400, 500, 1000))
```

```{r}
all_sct_intgrtd_male <- m2_all_raw_mrg[, m2_all_raw_mrg$sR_fine_stg  == "Male"]
all_sct_intgrtd_maleLE <- m2_all_raw_mrg[, m2_all_raw_mrg$sR_fine_stg  == "Male (LE)"]

```

```{r}
all_sct_intgrtd_male_lst <- list(all_sct_intgrtd_male, all_sct_intgrtd_maleLE) %>% set_names(c("male", "maleLE"))

```

```{r}
# names(all_sct_intgrtd_male_lst)

```
##############################START maleale DE ##########################

```{r}
map(all_sct_intgrtd_male_lst, ~DimPlot(.x, reduction = "umap.rpca_sct_PC10", group.by = "rpca_sct_PC10_clusters", label = T))
```


```{r}
pmap(list(all_sct_intgrtd_male_lst, c(-2,-3), c(-4,2.5)),  ~DimPlot(..1, reduction = "umap.rpca_sct_PC10", group.by = "rpca_sct_PC10_clusters", label = F, pt.size = 0.4) + geom_vline(xintercept =  ..2) + geom_hline(yintercept =  ..3))
```

```{r}
all_sct_intgrtd_male <- pmap(list(all_sct_intgrtd_male_lst, c(-2,-3), c(-4,2.5)),  ~..1[, ..1@reductions[["umap.rpca_sct_PC10"]]@cell.embeddings[, "umaprpcasctPC10_1"] < ..2 & ..1@reductions[["umap.rpca_sct_PC10"]]@cell.embeddings[, "umaprpcasctPC10_2"] > ..3]
)
```

```{r, eval=F}
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "umap.rpca_sct_PC10", cells.highlight = colnames(.x[,.x$rpca_sct_PC10_clusters == "25"])))
```

```{r}
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "umap.rpca_sct_PC10", group.by = "rpca_sct_PC10_clusters", label = T))
```

```{r, warning=FALSE, message=FALSE}
for (nm in names(all_sct_intgrtd_male)) {
  DefaultAssay(all_sct_intgrtd_male[[nm]]) <- "RNA"
  all_sct_intgrtd_male[[nm]][["SCT"]] <- NULL
}
```


```{r}
# DefaultAssay(all_sct_intgrtd_male) <- "RNA"
# all_sct_intgrtd_male[["SCT"]] <- NULL
```


```{r}
map(all_sct_intgrtd_male, ~table(.x$orig.ident))
```

```{r}
## Donors with more than 50 cells
don_50 <- map(all_sct_intgrtd_male, ~table(.x$donor) %>% data.frame() %>% filter(Freq >= 40) %>% pull(Var1) %>% as.character)
```

```{r}
## Retain only donors with more than 50 cells
all_sct_intgrtd_male <- map2(all_sct_intgrtd_male, don_50, ~subset(.x, subset = donor %in% .y))
```

## !!NOTE - NOT SURE WHY, I have to remove donors with less than 100 cells for FindVariableFeatures to work
```{r, warning=FALSE, message=FALSE}
# run standard seurat normalization
all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~NormalizeData(.x))
all_sct_intgrtd_male <- map2(all_sct_intgrtd_male, c(1000, 100), ~FindVariableFeatures(.x, nfeatures = .y))
all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~ScaleData(.x, verbose = F))

```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~RunPCA(.x, verbose = FALSE, npcs = 20))

```

```{r, warning=FALSE, message=FALSE}

all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~FindNeighbors(.x, dims = 1:10, reduction = "pca"))
all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~FindClusters(.x, resolution = 0.3))
```


```{r, warning=FALSE, message=FALSE}
# run sctransform
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "pca"))
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "pca", group.by = "orig.ident", label = T))


```



```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~IntegrateLayers(
  object = .x,
  method = HarmonyIntegration,
  orig.reduction = "pca", 
  new.reduction = "harmony",
  # normalization.method = "SCT",
  verbose = F,
  # k.weight = 98,
))


```


```{r, warning=FALSE, message=FALSE}

all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~FindNeighbors(.x, dims = 1:10, reduction = "harmony"))
all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~FindClusters(.x, resolution = 0.3))
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_sct_intgrtd_male <- map(all_sct_intgrtd_male, ~RunUMAP(.x, dims = 1:6, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "harmony"))

```


```{r, warning=FALSE, message=FALSE}
# run sctransform
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "umap"))
map(all_sct_intgrtd_male, ~DimPlot(.x, group.by = "orig.ident"))

map(all_sct_intgrtd_male, ~FeaturePlot(.x, reduction = "umap", features = "nCount_RNA"))

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "harmony"))
map(all_sct_intgrtd_male, ~DimPlot(.x, reduction = "harmony", dims = c(1,3)))

# FeaturePlot(all_sct_intgrtd_male, reduction = "umap", features = "nCount_RNA")

```



```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_sct_intgrtd_male[[1]]@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_sct_intgrtd_male[[1]]@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE, eval=F}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_sct_intgrtd_male@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_sct_intgrtd_male@meta.data$donor,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
map(all_sct_intgrtd_male, ~.x@meta.data %>% filter(seurat_clusters %in% c(4,6)) %>% count(seurat_clusters, orig.ident) %>% group_by(seurat_clusters) %>% arrange(desc(n), .by_group = T))
```

```{r, warning=FALSE, message=FALSE}
map(all_sct_intgrtd_male, ~.x@meta.data %>% count(seurat_clusters, orig.ident) %>% group_by(seurat_clusters) %>% arrange(desc(n), .by_group = T))
```



```{r, warning=FALSE, message=FALSE}
all_sct_intgrtd_male_jnd <- map(all_sct_intgrtd_male, ~JoinLayers(.x)) 
```

```{r, warning=FALSE, message=FALSE}
for (nm in names(all_sct_intgrtd_male_jnd)) {
  all_sct_intgrtd_male_jnd[[nm]][["RNA"]]$counts <- as(all_sct_intgrtd_male_jnd[[nm]][["RNA"]]$counts, Class = "dgCMatrix")
  all_sct_intgrtd_male_jnd[[nm]][["RNA"]]$data <- as(all_sct_intgrtd_male_jnd[[nm]][["RNA"]]$data, Class = "dgCMatrix")
  Idents(all_sct_intgrtd_male_jnd[[nm]]) <- "seurat_clusters"
}
```

```{r, warning=FALSE, message=FALSE}
all_sct_intgrtd_male_jnd <- map(all_sct_intgrtd_male_jnd, ~FindClusters(.x, resolution = 0.35))
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
map(all_sct_intgrtd_male_jnd, ~DimPlot(.x, reduction = "umap", pt.size = 0.2, label = T, dims = c(1,3)))

```

```{r, warning=FALSE, message=FALSE}
male_mks <- map(all_sct_intgrtd_male_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data"))
```


```{r, warning=FALSE, message=FALSE}
# run sctransform
map(all_sct_intgrtd_male_jnd, ~DimPlot(.x, reduction = "umap", pt.size = 0.1, label = T))

```


```{r, warning=FALSE, message=FALSE}
male_mks_anot <- map(male_mks, ~.x %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

male_mks_anot
```

```{r, warning=FALSE, message=FALSE}
map(male_mks_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 10))
```


```{r, warning=FALSE, message=FALSE}
male_mks_anot[["male"]] %>% filter(cluster == "3" & avg_log2FC >0)# %>% filter( avg_log2FC >1)
male_mks_anot[["male"]] %>% filter(cluster == "3" & avg_log2FC >0 & pct.2 < 0.1)# %>% filter( avg_log2FC >1)

```

```{r, warning=FALSE, message=FALSE}
male_mks_anot[["male"]] %>% filter(cluster == "2" & avg_log2FC >0)

```


```{r, warning=FALSE, message=FALSE}
map(male_mks_anot, ~.x %>% 
  filter(gene %in% gns) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 10))
```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=4}
#KAHRP, HSP70x, GARP, HRPIII, 
map(all_sct_intgrtd_male_jnd, ~FeaturePlot(.x, reduction = "umap", dims = c(1,3), pt.size = 0.15, label = T, features = c("PF3D7-0202000", "PF3D7-0831700", "PF3D7-0113000", "PF3D7-1372200"), ncol = 4))

```

```{r, warning=FALSE, message=FALSE}
male_mks_anot[["male"]] %>% filter(cluster == "4" & avg_log2FC >0)
```

```{r, warning=FALSE, message=FALSE}
# ap2-fg, pfg
male_mks_anot[["male"]] %>% filter(gene %in% c("PF3D7-1317200", "PF3D7-1146800") )
```

```{r, warning=FALSE, message=FALSE}
male_mks_anot[["male"]] %>% filter(cluster == "5" & avg_log2FC >0)# %>% filter( avg_log2FC >1)
# male_mks_anot[["male"]] %>% filter(cluster == "5" & avg_log2FC >0 & p_val_adj < 0.05) %>% arrange(desc(avg_log2FC))# %>% filter( avg_log2FC >1)

## cluster 5 could be activated-ish https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1010510#ppat-1010510-g002 
```

```{r, warning=FALSE, message=FALSE, fig.width=18, fig.height=4}
map(all_sct_intgrtd_male_jnd, ~FeaturePlot(.x, reduction = "umap", dims = c(1,3), pt.size = 0.15, label = T, features = c("PF3D7-0411000", "PF3D7-0628100", "PF3D7-0715200", "PF3D7-1143100", "nFeature_RNA"), ncol = 5))

```

```{r, warning=FALSE, message=FALSE}
activated_gams_ngwa <- c("PF3D7-0827800", "PF3D7-1246200", "PF3D7-1218800", "PF3D7-1239400", "PF3D7-0704100", "PF3D7-0417400", "PF3D7-1321000")

male_mks_anot[["male"]] %>% filter(gene %in% activated_gams_ngwa & avg_log2FC > 0)

```


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=4}
map(all_sct_intgrtd_male_jnd, ~FeaturePlot(.x, reduction = "umap", dims = c(1,3), pt.size = 0.15, label = T, features = c("PF3D7-0411000", "PF3D7-1143100", "PF3D7-1107800", "PF3D7-0209000", "nFeature_RNA"), ncol = 4))

```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=4}
map(all_sct_intgrtd_male_jnd, ~FeaturePlot(.x, reduction = "umap", dims = c(1,3), pt.size = 0.15, label = T, features = c("PF3D7-1143100", "PF3D7-1107800", "PF3D7-0209000"), ncol = 3))

```

```{r, warning=FALSE, message=FALSE}
male_mks_anot[["maleLE"]] %>% filter(cluster == "3" & avg_log2FC > 0) 

```

```{r, warning=FALSE, message=FALSE}
# map(male_mks_anot, ~.x %>% filter(cluster == "6" & avg_log2FC >1))

```

```{r, warning=FALSE, message=FALSE}
# map(male_mks_anot, ~.x %>% filter(cluster == "6" & avg_log2FC >1 & pct.2 < 0.1))

```


```{r, warning=FALSE, message=FALSE}
male_clustrs_w_asexuals <- c("6", "3") %>% set_names(c("male", "maleLE"))
```

```{r, warning=FALSE, message=FALSE}
saveRDS(all_sct_intgrtd_male_jnd, 'data/processed/Pf/m2_all/all_sct_intgrtd_male_jnd.RDS')

```

```{r}
# REmove poor QC asexual like cluster 
all_sct_intgrtd_male_jnd_no_asx_dblts_cln <- map2(all_sct_intgrtd_male_jnd, male_clustrs_w_asexuals, ~.x[, .x@meta.data %>% 
                                               rownames_to_column('bcode') %>% 
                                               filter(seurat_clusters != .y)%>% 
                                               pull(bcode)])


```

```{r, warning=FALSE, message=FALSE}
saveRDS(all_sct_intgrtd_male_jnd_no_asx_dblts_cln, 'data/processed/Pf/m2_all/all_sct_intgrtd_male_jnd_no_asx_dblts_cln.RDS')

```

```{r}
# m2_asex_ap2g_bal_sce <- m2_asex_ap2g_bal_sce[, m2_asex_ap2g_bal_sce@colData[, c('ap2g_exp_all', 'ptime_bal_cat')] %>% 
all_sct_intgrtd_m_jnd_no_asx_dblts <- all_sct_intgrtd_m_jnd[, all_sct_intgrtd_m_jnd@meta.data %>% 
                                               rownames_to_column('bcode') %>% 
                                               filter(seurat_clusters != "6") %>%
                                               group_by(seurat_clusters, donor, Strain_DN) %>% 
                                               filter(n() >=50) %>%
                                               group_by(seurat_clusters, donor) %>% 
                                               filter(n_distinct(Strain_DN) >1) %>%
                                               # filter(n() >=100) %>% 
                                               pull(bcode)]


```



```{r}
## subset where field parasites are for slingshot
all_sct_intgrtd_m_jnd_no_asx_dblts@meta.data %>% count(seurat_clusters, donor, Strain_DN) %>% group_by(seurat_clusters, donor) %>% mutate(strn_n = n_distinct(Strain_DN)) %>% filter(strn_n >1)
```



```{r}
DimPlot(all_sct_intgrtd_m_jnd_no_asx_dblts, reduction = "umap", group.by = "seurat_clusters" )
DimPlot(all_sct_intgrtd_m_jnd_no_asx_dblts, reduction = "umap", group.by = "seurat_clusters", dims = c(1,3) )

```





## MAST DE adjusting for pseudotime

MAST DE on the entire dataset before subsampling. Subsampling is not necessary for MAST and also for tradeseq
### Prepare dataset
Normalizing and scaling the data in the subseted seurat objects - may have been done before on the larger parent object but just repeating since will not change anything as it starts from the raw counts.

Adding pseudotime metadata from the single cell experiment object 
Scaling is for visualization


```{r}
##REMEMBER TO REMOVE CLUSTERS WITH VERY FEW CELLS AND DOUBLETS SEURAT
# msc_gam_combi_noLE_de <- msc_gam_combi_noLE_de[, cells_de]
```


```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
# all_sct_intgrtd_m_jnd_no_asx_dblts <- all_sct_intgrtd_m_jnd_no_asx_dblts
all_sct_intgrtd_m_jnd_no_asx_dblts[["SCT"]] <- NULL
all_sct_intgrtd_m_jnd_no_asx_dblts[["RNA"]]$scale.data <- NULL
DefaultAssay(all_sct_intgrtd_m_jnd_no_asx_dblts) <- "RNA"

```

```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
all_sct_intgrtd_m_jnd_no_asx_dblts@meta.data[, "don_seu_clutrs"] <- paste0(all_sct_intgrtd_m_jnd_no_asx_dblts$donor, "_", all_sct_intgrtd_m_jnd_no_asx_dblts$seurat_clusters)

```

```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
all_m_de_bal_nosct <- SplitObject(all_sct_intgrtd_m_jnd_no_asx_dblts, split.by = "don_seu_clutrs")

```

```{r, message=FALSE, warning=FALSE}
all_gns = map(all_m_de_bal_nosct, ~unique(rownames(.x)))

m_m2_bal_seu <- map2(all_m_de_bal_nosct,all_gns, ~{
  .x %>%
         NormalizeData(., normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE) %>%
         ScaleData(., features = .y,  verbose = FALSE)
  })

# m2_bal_seu <- list(m2_bal_seu) %>% set_names("gam")

```

### Actual MAST DE between pairwise strains and fieldvslab 

MAST DE adjusting for pseudotime as a latent variable

```{r, message=FALSE, warning=FALSE}
de_idents <- rep("Strain_DN", length = length(m_m2_bal_seu))
m_m2_bal_mast <- map2(m_m2_bal_seu, de_idents, ~SetIdent(.x, value=.y))

```

```{r, message=FALSE, warning=FALSE}
saveRDS(m_m2_bal_mast, 'data/processed/Pf/m2_all/m_m2_bal_mast.RDS')

```

```{r, message=FALSE, warning=FALSE}
m_m2_bal_mast <- readRDS('data/processed/Pf/m2_all/m_m2_bal_mast.RDS')

```

```{r}
# a<- merge(m_m2_bal_mast[[1]], m_m2_bal_mast[2:length(m_m2_bal_mast)])

```

```{r}
# DimPlot(a,  group.by = "seurat_clusters" )

```

```{r, message=FALSE, warning=FALSE}
##shortcut to run relevant donors subsets only
# m2_bal_mast_noM55M54 <- subset(m_m2_bal_mast[["gam"]], subset = donor %in% c("MSC68", "MSC66", "MSC70", "MSC60", "MSC49"))
```


```{r, message=FALSE, warning=FALSE, eval=T}
##!! NB Rerun only when the DE inputs change - takes looooonng

## Mast DE of dataset with male low (ml) included - shouldnt have any contribution as each stage handled differently
## Including lab and field strain DE
## !!!min.cells.group added for mh in the strain_dset where there's only balancing in lab VS field - we need to remove later

pts <- "slingPseudotime_1"
pts_all<- c(rep(pts,1))
pts_all<- pts

### !!! NOTE - RUNNING FOR NECESSARY DONORS, WILL REVERT LATER
m_pairwise_mkrs_seu_clustr <- list(m_m2_bal_mast, pts, de_idents, names(m_m2_bal_mast)) %>%
# m_pairwise_mkrs_seu_clustr <- list(list(m2_bal_mast_noM55M54), pts, de_idents, names(m2_bal_mast)) %>%
  pmap(.,  ~{
    # ide = ..3
    # a = combn(levels(droplevels(..1@meta.data[,..3])), 2,simplify = F) 
    a = combn(unique(..1@meta.data[,..3]), 2,simplify = F) 
    obj = ..1
    pseudos = ..2
    
    print(class(obj))
    print(paste("running ...", ..4))
    map(a, ~print(paste(.x[1], "n ", .x[2])))
    
    # map(a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], latent.vars = pseudos, logfc.threshold = 0, test.use = 'MAST', min.cells.group = 1) %>% rownames_to_column('gene_id'))
    map(a, ~FindMarkers(obj, 
                        ident.1 = .x[1], 
                        ident.2 = .x[2], 
                        # latent.vars = c("donor", pseudos), 
                        # latent.vars = c(pseudos), 
                        logfc.threshold = 0, 
                        # test.use = 'MAST', 
                        min.cells.group = 1) %>% 
          rownames_to_column('gene_id')) %>%
      set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
      bind_rows(., .id = 'pw_comparison')
    
  }#, 'missing genes')
  )

saveRDS(m_pairwise_mkrs_seu_clustr,'data/processed/Pf/m2_all/m_pairwise_mkrs_seu_clustr.RDS')
# saveRDS(m_pairwise_mkrs_seu_clustr_don,'data/processed/Pf/m2_all/m_pairwise_mkrs_seu_clustr_don.RDS')
# c(m_pairwise_mkrs_seu_clustr_lf, m_pairwise_mkrs_seu_clustr)



```

```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr <- readRDS('data/processed/Pf/m2_all/m_pairwise_mkrs_seu_clustr.RDS')

```


```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot <- map(m_pairwise_mkrs_seu_clustr, ~.x %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")]))

m_pairwise_mkrs_seu_clustr_anot
```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more gam UMI
map(m_pairwise_mkrs_seu_clustr_anot, ~.x  %>% filter(p_val_adj < 0.05 & avg_log2FC > 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in gam positive cells

```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more gam UMI
imap(m_pairwise_mkrs_seu_clustr_anot, ~.x %>% filter(p_val_adj < 0.05 & avg_log2FC < 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in gam positive cells

```


```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot_all <- m_pairwise_mkrs_seu_clustr_anot %>% 
  bind_rows(., .id = "sample") %>%
  filter(p_val_adj < 0.05) %>%
  group_by(gene_id) %>%
  mutate(donor = str_remove(sample, "_.*"),
         gene_DE_sample_prevalence = n_distinct(sample, pw_comparison),
         gene_DE_donor_prevalence = n_distinct(donor)) %>%
  ungroup() %>%
  mutate(gene_DE_sample_prevalence_cat = cut(gene_DE_sample_prevalence, breaks = c(3), labels = c("Low", "Medium", "High"), right = FALSE), 
         gene_DE_donor_prevalence_cat = cut(gene_DE_donor_prevalence, breaks = c(3), labels = c("Low", "Medium", "High"), right = FALSE))   %>% 
  filter(!str_detect(Gene, "VAR_"))
```

```{r, warning=FALSE, message=FALSE}
write_csv(m_pairwise_mkrs_seu_clustr_anot_all, "../colleagues/chiyun/m_pairwise_mkrs_seu_clustr_anot_all.csv")
```

```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T)
```

```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T)
```


```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot_all %>% 
  filter(!str_detect(Gene, "VAR_")) %>%
  filter(Gene == "MDR1")
```


```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-5) %>% 
  ggplot(aes(x = avg_log2FC)) +
  geom_histogram(binwidth = 0.08) + 
  geom_vline(xintercept = c(-3, -2, -1, 1, 2, 3))
```

```{r, warning=FALSE, message=FALSE}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10) %>% 
  ggplot(aes(y = avg_log2FC, x = sample)) +
  geom_violin() + 
  geom_hline(yintercept = c(-3, -2, -1, 1, 2, 3)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10 ) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.1) + 
  geom_hline(yintercept = seq(-9,9,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 28))

```


```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=5}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10 ) %>% 
  ungroup() %>%
  ggplot(aes(y = avg_log2FC, x = gene_DE_donor_prevalence_cat, colour = gene_DE_donor_prevalence_cat)) +
  geom_violin() +
  geom_sina(size = 0.1) + 
  geom_hline(yintercept = seq(-8,8,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```
##############################END Male DE ##########################

##############################START Compare with published CNVs ##########################


```{r, fig.width=30, fig.height=6}
## NB - This needs to go to the common scripts file
## read alias file first time round
pf_pdb68_alias <- read_tsv("../../Pf3D7_genomes_gtfs_indexs/release_68/PlasmoDB-68_Pfalciparum3D7_GeneAliases.txt", col_names = F)

## Warnings present due to column disparity across rows "Warning: One or more parsing issues, call `problems()` on your data frame for details, e.g.:" Use problems to check maximum number of rows

problems(pf_pdb68_alias) %>% count(col)

## Sort by separating the last column based on \t
pf_pdb68_alias <- pf_pdb68_alias %>% separate_wider_delim(X2, delim = "\t", names_sep = "_", too_few = "align_start") 
```


#### Field VS lab lit comparisons
```{r}
pfdb_old_ids <- Pf3D7_genes_plasmoDB %>% separate(col = `Previous ID(s)`, sep=';|: ', into = c('a', 'id_mal', 'id_pfa1','id_pfa2','id_pfa3', 'extra_ids'), extra = 'merge') %>% pivot_longer(., cols = c('id_mal', 'id_pfa1','id_pfa2','id_pfa3'), names_to = 'id_typ', values_to = 'gene_id_old', values_drop_na = T)

genes_20_up_in_fld_mknon <- c('PF13_0011|PFI1720w|PFL0795c|PF10_0350|PF13_0196|PFL1085w|PFI1735c|PFI1740c|PFI1755c|PF07_0005|PF07_0006|PF07_0007|PF14_0748|PFI1770w|PFI1780w|MAL7P1.7|PF14_0644|PF14_0708|PF14_0735|PF14_0736|PFE0940c')

pfdb_up_in_mknon <- Pf3D7_genes_plasmoDB %>% filter(str_detect(`Previous ID(s)`,genes_20_up_in_fld_mknon))

genes_20_low_in_fld_mknon <- c('MAL7P1.225|PFB0105c|PFD1140w|PF14_0616|PFB0115w|PFI0085c|PFL1370w|PF10_0383|PF11_0033|MYXSPDY|PFB0350c|PFL1140w|PF11_0478|PFB0095c|PFB0100c|PF07_0076|MAL8P1.210|PFL0325w|PFL0600w|PFL0655w')

pfdb_low_in_mknon <- Pf3D7_genes_plasmoDB %>% filter(str_detect(`Previous ID(s)`,genes_20_low_in_fld_mknon))

pfdb_up_in_mknon %>% filter(gene_id %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id)
pfdb_up_in_mknon %>% filter(gene_id %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id)
```

#### mknon malaria field vs lab DE
```{r}
DE_preg_mal_mknon <- read_csv('../published_dsets/field_vs_lab/mckinon_2009/DE_preg_mal_s013.csv') %>% rename('gene_id_old' = `...1`)

pfdb_pg_in_mknon <- DE_preg_mal_mknon  %>% left_join(., pfdb_old_ids, by='gene_id_old')

pfdb_pg_in_mknon %>% filter( gene_id %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id) 
pfdb_pg_in_mknon %>% filter( gene_id %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id) 

```

```{r}
ls()
```

#### field vs lab DE within cnv (mknon)
```{r}
# df <- pdftools::pdf_data("../published_dsets/field_vs_lab/mckinon_2009/CNV_list_inc_litrev.1000644.s012.pdf")[1:16]
df <- readRDS("../published_dsets/field_vs_lab/mckinon_2009/CNV_list_inc_litrev.1000644.s012.RDS")[1:16]

df[[1]] <- df[[1]] %>% filter(y>140)

df[[16]] <- df[[16]] %>% filter(y<267)


mknon_cnv <- df %>% bind_rows(.id = 'tbl') %>%
    group_by(y) %>%
    arrange(x, by_group = TRUE)%>% 
  pivot_wider(., names_from = 'x', id_cols = c('y', 'tbl'), values_from = 'text') %>% 
  select_if(~any(!is.na(.))) %>%
  mutate('chrom' = coalesce(`310`, `316`),
         'Bp_gn_midpoint' = coalesce(`339`, `345`, `350`), 
         'mknon' = coalesce(`714`, `717`, `721`, `723`, `727`)) %>% 
  dplyr::rename('Kidgell' = `389`,'Ribacke'= `437`, 'Bozdech' = `495`, 'Carret'= `550`, 'Others' = `597`,  'CNV_name' = `72`, 'SD_name' = `126`, 'gene_id_old' = `171`) %>%
  ungroup() %>%
  dplyr::select(c("CNV_name", "SD_name", "gene_id_old","chrom", "Bp_gn_midpoint", "Kidgell", "Ribacke", "Bozdech", "Carret", "Others", "mknon")) %>%
  .[1:449,] 

pfdb_mknon_cnv <- mknon_cnv  %>% left_join(., pfdb_old_ids, by='gene_id_old')


# mknon_cnv_pdb <-mknon_cnv %>% left_join(., Pf3D7_genes_plasmoDB, by = c('gene_id_old'='Previous ID(s)')) 

# asx_cnv <- pfdb_mknon_cnv %>% filter( gene_id %in% cond_gns_001_de[[1]]$gene_id) %>% filter(duplicated(CNV_name))
# fem_cnv <- pfdb_mknon_cnv %>% filter( gene_id %in% cond_gns_001_de[[2]]$gene_id) %>% filter(duplicated(CNV_name))
pfdb_mknon_cnv %>% filter( gene_id %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name))
pfdb_mknon_cnv %>% filter( gene_id %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name))

pfdb_mknon_cnv %>% filter( gene_id %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name))
pfdb_mknon_cnv %>% filter( gene_id %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name))

```


#### field vs lab DE within PDB cnv
```{r}
pfdb_cnv <- read_csv('/lustre/scratch126/tol/teams/lawniczak/users/jr35/Pf3D7_genomes_gtfs_indexs/GenesByCopyNumberComparison_Summary091022.csv') %>% mutate(across(`Gene ID`, ~str_replace(., '_', '-')))

pfdb_cnv %>% filter( `Gene ID` %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id)
pfdb_cnv %>% filter( `Gene ID` %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id)
```

#### mknon CNV in PDB cnv
```{r}
cnv_pbd_mknon <- pfdb_mknon_cnv %>% mutate(gn_in_cnv_mk = 'mk_yes') %>% full_join(.,pfdb_cnv %>% mutate(gn_in_cnv_pbd = 'pdb_yes'), by=c('gene_id' = 'Gene ID')) %>% mutate(gn_in_cnv = coalesce(gn_in_cnv_mk, gn_in_cnv_pbd))

cnv_pbd_mknon %>% dplyr::count(gn_in_cnv_mk, gn_in_cnv_pbd)

cnv_pbd_mknon %>% filter( gene_id %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name)) %>% dplyr::count(CNV_name,gene_id, gn_in_cnv_mk, gn_in_cnv_pbd)
cnv_pbd_mknon %>% filter( gene_id %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name)) %>% dplyr::count(CNV_name, gene_id, gn_in_cnv_mk, gn_in_cnv_pbd)
cnv_pbd_mknon %>% filter( gene_id %in% asx_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name)) %>% dplyr::count(CNV_name, gene_id, gn_in_cnv_mk, gn_in_cnv_pbd)

cnv_gn_lst <- list('f_don' = cnv_pbd_mknon %>% filter( gene_id %in% fem_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name)) %>% dplyr::count(CNV_name,gene_id, gn_in_cnv_mk, gn_in_cnv_pbd) %>% pull(gene_id),
                'm_don' = cnv_pbd_mknon %>% filter( gene_id %in% m_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name)) %>% dplyr::count(CNV_name,gene_id, gn_in_cnv_mk, gn_in_cnv_pbd) %>% pull(gene_id),
                'asx_don' = cnv_pbd_mknon %>% filter( gene_id %in% asx_pairwise_mkrs_seu_clustr_anot_all$gene_id) %>% filter(duplicated(CNV_name)) %>% dplyr::count(CNV_name,gene_id, gn_in_cnv_mk, gn_in_cnv_pbd) %>% pull(gene_id)
                )

```

```{r}
cnv_gn_lst
```


```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=5}
m_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10 & gene_id %in% cnv_gn_lst$m_don) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.5) + 
  geom_hline(yintercept = seq(-8,8,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=5}
fem_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10 & gene_id %in% cnv_gn_lst$f_don) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.5) + 
  geom_hline(yintercept = seq(-8,8,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=5}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-5 & gene_id %in% cnv_gn_lst$asx_don) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin()+
  geom_sina(size = 0.5) +
  # geom_point(size = 0.5, position = position_jitter()) + 
  geom_hline(yintercept = seq(-13,13,1))  +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=5}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-5 & gene_id %in% cnv_gn_lst$asx_don) 
```

##############################END Compare with published CNVs ##########################



```{r, warning=FALSE, message=FALSE}
gogo()
```


#### female high vs female low DE visualization
```{r, fig.width=6, fig.height=6}
map(list(fem_pairwise_mkrs_seu_clustr_anot[[1]] %>% filter(Gene != "AP2-G"), 
         fem_pairwise_mkrs_seu_clustr_anot[[1]] %>% filter(Gene != "AP2-G") %>% filter(p_val_adj < 0.05), 
         fem_pairwise_mkrs_seu_clustr_anot[[1]] %>% filter(Gene != "AP2-G") %>% filter(!(abs(avg_log2FC) > 5 & p_val_adj > 0.05)),
         fem_pairwise_mkrs_seu_clustr_anot[[1]]), ~{
  EnhancedVolcano(.x,
                  lab = .x$Gene,
                  selectLab =.x %>% 
                    # filter(p_val_adj < 0.05 & (avg_log2FC > 1 | avg_log2FC < -1) & !str_detect(Gene, 'PF3D7')) %>%
                    filter(p_val_adj < 0.05 & (avg_log2FC > 1 | avg_log2FC < -1) ) %>%
                    pull(Gene),
                  x = 'avg_log2FC',
                  y = 'p_val_adj',
                  pCutoff = 0.05,
                  FCcutoff = 1,
                  pointSize = 3.0,
                  labSize = 7.0,
                  title = '',
                  subtitle = "",
                  titleLabSize = 0,
                  subtitleLabSize = 0,
                  axisLabSize = 24,
                  drawConnectors = TRUE,
                  legendLabels =c(expression(Log[2] ~ FC), "p-value"),
                  gridlines.minor = FALSE)
})
```

