---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(ComplexHeatmap)
  library(circlize)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  
})
```


```{r setup}
sample_set = "Mali2"

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```


#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```

```{r}
# optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)
# 
# optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]
```

Integrated asexual
```{r}
## subset where field parasites are for slingshot
# all_asex_merged_int_no_actv_harmony_pt2 <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int_no_actv_harmony_pt2.RDS")
# m2_scvi_asex_noM12_harmony_pt <- readRDS("data/processed/Pf/m2_scvi_asex_noM12_harmony_pt.RDS")
# m2_scvi_asex_noM12_harmony <- readRDS("data/processed/Pf/m2_scvi_asex_noM12_harmony.RDS")
# m2_scvi_asex_noM12_harmony_pt <- readRDS("data/processed/Pf/v3_m2_asex_cr_filt_norm_harmony_pt.RDS")

```


```{r}
## subset where field parasites are for slingshot
# m2_scvi_asex_noM12_harmony_pt2 <- readRDS("data/processed/Pf/m2_all/m2_scvi_asex_noM12_harmony_pt2.RDS")
# m2_scvi_asex_noM12_harmony_sce_ss <- readRDS("data/processed/Pf/m2_scvi_asex_noM12_harmony_sce_ss.RDS")
# v3_m2_asex_cr_filt_norm_harmony_pt <- readRDS("data/processed/Pf/v3_m2_asex_cr_filt_norm_harmony_pt.RDS")
v3_m2_asex_cr_filt_norm_harmony_pt <- readRDS("data/processed/Pf/asex_noM55_jcC_harmony_no_comtd_pt3.RDS")

v3_m2_asex_cr_filt_norm_harmony_pt <- v3_m2_asex_cr_filt_norm_harmony_pt[, !v3_m2_asex_cr_filt_norm_harmony_pt$Strain_DN %in% c("Doublet", "Negative")]
```


```{r}
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_ss" )
```


```{r}
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man", group.by = "stage" )

```

```{r}
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man", features = "slingPseudotime_1")
```

```{r, warning=FALSE, message=FALSE}
# ap2-g pfs16 gexp02 gene_count
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man",  label = T, features = "PF3D7-1222600", order = T)
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man",  label = T, features = "PF3D7-0406200", order = T)
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man",  label = T, features = "PF3D7-1102500", order = T)
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man",  label = T, features = "nFeature_RNA", order = T)
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man",  label = T, features = "nCount_RNA", order = T)


```


```{r, warning=FALSE, message=FALSE}
##Generate pseudotime chunks categories with range of highest and lowest pseudotime values and breaks numbers to ensure balanced representation of cells and drop the cells without strain information

ptime_chunk_calc_df_fn <- function(df, ptime, chunk_num) {
  # v3_m2_asex_harmony_fld_sset_ptime1_fld@meta.data[,c('ptime_cat', 'ptime_bal_cat')] 
  df %>% 
          mutate(ptime_cat = cut(!!rlang::sym(ptime),  
                                 breaks = c(-Inf, seq(min(!!rlang::sym(ptime)),  max(!!rlang::sym(ptime)), round((max(!!rlang::sym(ptime)) - min(!!rlang::sym(ptime)))/chunk_num, 2)), Inf)),
                 pt_rank = dense_rank(!!rlang::sym(ptime)),
                 ptime_bal_cat = cut(pt_rank,  
                                 breaks = c(-Inf, seq(min(pt_rank),  max(pt_rank), round((max(pt_rank) - min(pt_rank))/chunk_num, 2)), Inf))
                 ) %>%
    mutate(ptime_grp_rnk = dense_rank(factor(ptime_cat)),
           ptime_bal_grp_rnk = dense_rank(factor(ptime_bal_cat)),
           ptime_cat = factor(paste0("ptime_", ptime_grp_rnk), levels = paste0("ptime_", 1:max(ptime_grp_rnk)) ),
           ptime_bal_cat = factor(paste0("ptime_", ptime_bal_grp_rnk), levels = paste0("ptime_", 1:max(ptime_bal_grp_rnk)) )
           ) %>%
          dplyr::select(ptime_cat, ptime_bal_cat)
  
}

```


```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
bal_df <- ptime_chunk_calc_df_fn(df = v3_m2_asex_cr_filt_norm_harmony_pt@meta.data, ptime = "slingPseudotime_1", chunk_num = 6)
```


```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
v3_m2_asex_cr_filt_norm_harmony_pt <- AddMetaData(v3_m2_asex_cr_filt_norm_harmony_pt, bal_df) 
```

```{r}
## subset where field parasites are for slingshot
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man", group.by = "ptime_cat" )
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" )
```

```{r}
## subset rings - Why did I remove rings initially? Is it because this is where commitment happens? 
# v3_m2_asex_cr_filt_norm_harmony_pt_no_rings <- v3_m2_asex_cr_filt_norm_harmony_pt[, !v3_m2_asex_cr_filt_norm_harmony_pt$harmony_clusters_ss %in% c("1")]
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings <- v3_m2_asex_cr_filt_norm_harmony_pt

```

```{r}
## subset where field parasites are for slingshot
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings@meta.data %>% count(ptime_bal_cat, donor, Strain_DN)
```


```{r}
# m2_asex_ap2g_bal_sce <- m2_asex_ap2g_bal_sce[, m2_asex_ap2g_bal_sce@colData[, c('ap2g_exp_all', 'ptime_bal_cat')] %>% 
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings <- v3_m2_asex_cr_filt_norm_harmony_pt_no_rings[, v3_m2_asex_cr_filt_norm_harmony_pt_no_rings@meta.data %>% 
                                               rownames_to_column('bcode') %>% 
                                               group_by(ptime_bal_cat, donor, Strain_DN) %>% 
                                               filter(n() >=20) %>%
                                               group_by(ptime_bal_cat, donor) %>% 
                                               filter(n_distinct(Strain_DN) >1) %>%
                                               # filter(n() >=100) %>% 
                                               pull(bcode)]


```



```{r}
## subset where field parasites are for slingshot
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings@meta.data %>% count(ptime_bal_cat, donor, Strain_DN) %>% group_by(ptime_bal_cat, donor) %>% mutate(strn_n = n_distinct(Strain_DN)) %>% filter(strn_n >1)
```



```{r}
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" )
```





## MAST DE adjusting for pseudotime

MAST DE on the entire dataset before subsampling. Subsampling is not necessary for MAST and also for tradeseq
### Prepare dataset
Normalizing and scaling the data in the subseted seurat objects - may have been done before on the larger parent object but just repeating since will not change anything as it starts from the raw counts.

Adding pseudotime metadata from the single cell experiment object 
Scaling is for visualization


```{r}
##REMEMBER TO REMOVE CLUSTERS WITH VERY FEW CELLS AND DOUBLETS SEURAT
# msc_gam_combi_noLE_de <- msc_gam_combi_noLE_de[, cells_de]
```


```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
# v3_m2_asex_cr_filt_norm_harmony_pt_no_rings <- v3_m2_asex_cr_filt_norm_harmony_pt_no_rings
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings[["SCT"]] <- NULL
# v3_m2_asex_cr_filt_norm_harmony_pt_no_rings[["RNA"]]$scale.data <- NULL
DefaultAssay(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings) <- "RNA"

```


```{r, message=FALSE, warning=FALSE}
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings <- JoinLayers(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings, layers = c("counts", "data", "scale.data"))
```

```{r, message=FALSE, warning=FALSE}
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings[["RNA"]]$scale.data <- NULL

```

```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
v3_m2_asex_cr_filt_norm_harmony_pt_no_rings@meta.data[, "don_ptime"] <- paste0(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings$donor, "_", v3_m2_asex_cr_filt_norm_harmony_pt_no_rings$ptime_bal_cat)

```

```{r, message=FALSE, warning=FALSE}
# Save for tradeseq 
saveRDS(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings, 'data/processed/Pf/m2_all/v3_m2_asex_cr_filt_norm_harmony_pt_no_rings.RDS')

```

```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
all_asex_de_bal_nosct <- SplitObject(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings, split.by = "don_ptime")

```

```{r, message=FALSE, warning=FALSE}
all_gns = map(all_asex_de_bal_nosct, ~unique(rownames(.x)))

asx_m2_bal_seu <- map2(all_asex_de_bal_nosct,all_gns, ~{
  .x %>%
         NormalizeData(., normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE) %>%
         ScaleData(., features = .y,  verbose = FALSE)
  })

# m2_bal_seu <- list(m2_bal_seu) %>% set_names("gam")

```

### Actual MAST DE between pairwise strains and fieldvslab 

MAST DE adjusting for pseudotime as a latent variable

```{r, message=FALSE, warning=FALSE}
de_idents <- rep("Strain_DN", length = length(asx_m2_bal_seu))
asx_m2_bal_mast <- map2(asx_m2_bal_seu, de_idents, ~SetIdent(.x, value=.y))

```

```{r, message=FALSE, warning=FALSE}
saveRDS(asx_m2_bal_mast, 'data/processed/Pf/m2_all/asx_m2_bal_mast3.RDS')

```


```{r, message=FALSE, warning=FALSE}
##shortcut to run relevant donors subsets only
# m2_bal_mast_noM55M54 <- subset(m2_bal_mast[["gam"]], subset = donor %in% c("MSC68", "MSC66", "MSC70", "MSC60", "MSC49"))
```

```{r, warning=FALSE, message=FALSE}
for (nm in names(asx_m2_bal_mast)) {
  asx_m2_bal_mast[[nm]][["RNA"]]$data <- as(asx_m2_bal_mast[[nm]][["RNA"]]$data, Class = "dgCMatrix")
}
```

```{r, message=FALSE, warning=FALSE, eval=T}
##!! NB Rerun only when the DE inputs change - takes looooonng

## Mast DE of dataset with male low (ml) included - shouldnt have any contribution as each stage handled differently
## Including lab and field strain DE
## !!!min.cells.group added for mh in the strain_dset where there's only balancing in lab VS field - we need to remove later

pts <- "slingPseudotime_1"
pts_all<- c(rep(pts,1))
pts_all<- pts

### !!! NOTE - RUNNING FOR NECESSARY DONORS, WILL REVERT LATER
asx_pairwise_mkrs_seu_clustr <- list(asx_m2_bal_mast, pts, de_idents, names(asx_m2_bal_mast)) %>%
# ap2g_ub_pairwise_mkrs_pt_ml <- list(list(m2_bal_mast_noM55M54), pts, de_idents, names(m2_bal_mast)) %>%
  pmap(.,  ~{
    # ide = ..3
    # a = combn(levels(droplevels(..1@meta.data[,..3])), 2,simplify = F) 
    a = combn(unique(..1@meta.data[,..3]), 2,simplify = F) 
    obj = ..1
    pseudos = ..2
    
    print(class(obj))
    print(paste("running ...", ..4))
    map(a, ~print(paste(.x[1], "n ", .x[2])))
    
    # map(a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], latent.vars = pseudos, logfc.threshold = 0, test.use = 'MAST', min.cells.group = 1) %>% rownames_to_column('gene_id'))
    map(a, ~FindMarkers(obj, 
                        ident.1 = .x[1], 
                        ident.2 = .x[2], 
                        # latent.vars = c("donor", pseudos), 
                        latent.vars = c(pseudos), 
                        logfc.threshold = 0, 
                        test.use = 'MAST', 
                        min.cells.group = 1) %>% 
          rownames_to_column('gene_id')) %>%
      set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
      bind_rows(., .id = 'pw_comparison')
    
  }#, 'missing genes')
  )

saveRDS(asx_pairwise_mkrs_seu_clustr,'data/processed/Pf/m2_all/asx_pairwise_mkrs_seu_clustr3.RDS')
# saveRDS(ap2g_ub_pairwise_mkrs_pt_ml_don,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml_don.RDS')
# c(ap2g_ub_pairwise_mkrs_pt_ml_lf, ap2g_ub_pairwise_mkrs_pt_ml)



```


```{r, message=FALSE, warning=FALSE, eval=F}
##!! NB Rerun only when the DE inputs change - takes looooonng
### Getting all genes for gsea - saved as RDS and analysis done in functional_GO_n_lit_comparison.RMD

# ap2g_ub_pairwise_mkrs_pt_ml_gsea_full <- list(m2_bal_mast, rep(pts,2), de_idents) %>%
#   pmap( ~{
#     # ide = ..3
#     a = combn(levels(..1@meta.data[,..3]), 2,simplify = F) 
#     obj = ..1
#     pseudos = ..2
#     map(a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], latent.vars = pseudos, logfc.threshold = 0, test.use = 'MAST', min.pct = -Inf, min.diff.pct = -Inf) %>% rownames_to_column('gene_id')) %>%
#       set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
#       bind_rows(., .id = 'pw_comparison')
#     
#   })
# saveRDS(ap2g_ub_pairwise_mkrs_pt_ml_gsea_full,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml_gsea_full.RDS')

```


```{r, message=FALSE, warning=FALSE}
asx_pairwise_mkrs_seu_clustr<- readRDS('data/processed/Pf/m2_all/asx_pairwise_mkrs_seu_clustr3.RDS')

```


```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot <- map(asx_pairwise_mkrs_seu_clustr, ~.x %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")]))

asx_pairwise_mkrs_seu_clustr_anot
```


```{r, warning=FALSE, message=FALSE}
# Another amplification of 3x on chromosome 12
# was observed immediately before GTP cyclohydrolase I (pfgch1,
# PF3D7_1224000) across two genes (PF3D7_1223800-PF3D7_1223900), which
# was also in accordance to previous studies 
# https://link.springer.com/article/10.1186/gb-2009-10-2-r21
# https://escholarship.org/content/qt1bc2p97r/qt1bc2p97r_noSplash_c86b54d668a56b24f34dfe70cb02191e.pdf?t=o8x13q
```


```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more gam UMI
map(asx_pairwise_mkrs_seu_clustr_anot, ~.x  %>% filter(p_val_adj < 0.05 & avg_log2FC > 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in gam positive cells

```


```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more gam UMI
imap(asx_pairwise_mkrs_seu_clustr_anot, ~.x %>% filter(p_val_adj < 0.05 & avg_log2FC < 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in gam positive cells

```


```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot_all <- asx_pairwise_mkrs_seu_clustr_anot %>% 
  bind_rows(., .id = "sample") %>%
  filter(p_val_adj < 0.05) %>%
  group_by(gene_id) %>%
  mutate(across(sample, ~str_remove(., "_ptime")),
         donor = str_remove(sample, "_.*"),
         gene_DE_sample_prevalence = n_distinct(sample, pw_comparison),
         gene_DE_donor_prevalence = n_distinct(donor, pw_comparison)) %>%
  ungroup() %>%
  mutate(gene_DE_sample_prevalence_cat = cut(gene_DE_sample_prevalence, breaks = c(3), labels = c("Low", "Medium", "High"), right = FALSE), 
         gene_DE_donor_prevalence_cat = cut(gene_DE_donor_prevalence, breaks = c(3), labels = c("Low", "Medium", "High"), right = FALSE))  %>% 
  filter(!str_detect(Gene, "VAR_"))
```


```{r, warning=FALSE, message=FALSE}
write_csv(asx_pairwise_mkrs_seu_clustr_anot_all, "../colleagues/chiyun/asx_pairwise_mkrs_seu_clustr_anot_all.csv")
```

```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T)
```

```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T)
```


```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot_all %>% 
  filter(!str_detect(Gene, "VAR_")) %>%
  filter(Gene == "MDR1")
```


```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot_all %>% 
  filter(!str_detect(Gene, "VAR_")) %>%
  distinct(Gene, .keep_all = T)
```


```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more gam UMI
c("PF3D7-0113600", "PF3D7-1301800", "PF3D7-1222600")
```


```{r, warning=FALSE, message=FALSE}
asx_pairwise_mkrs_seu_clustr_anot[[1]] %>% filter(gene_id %in% c("PF3D7-0113600", "PF3D7-1301800", "PF3D7-1222600"))
```



```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10 ) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.1) + 
  geom_hline(yintercept = seq(-12,12,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 28))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
## malariagen_CNVs gch1, crt, mdr1, plasmepsin 2/3, hrp2, hrp3

malariagen_CNVs <- c("PF3D7-1224000", "PF3D7-0709000", "PF3D7-0523000", "PF3D7-1408000", "PF3D7-0831800", "PF3D7-1372200")

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(gene_id %in% malariagen_CNVs) 

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(gene_id %in% malariagen_CNVs) %>%
  filter(p_val_adj < 0.0001 ) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.1) + 
  geom_hline(yintercept = seq(-7,7,1)) +
  facet_wrap(Gene ~ ., nrow = 1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 28))
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
likely_cnvs <- asx_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(gene_id %in% malariagen_CNVs) %>%
  filter(p_val_adj < 0.0001 ) %>% 
  filter(abs(avg_log2FC)>1) 

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
likely_cnvs_don <- likely_cnvs  %>% 
    distinct(donor, gene_id, pw_comparison, .keep_all = T) %>%
  mutate(labl = paste0(donor, "_", Gene, "_", pw_comparison)) %>%
  select(donor, gene_id, Gene, pw_comparison, labl, avg_log2FC)
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt, split.by = "donor", reduction = "umap.harmony_PC_man")
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings, split.by = "donor", reduction = "umap.harmony_PC_man")
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
imap(likely_cnvs_don, possibly(~FeaturePlot(subset(v3_m2_asex_cr_filt_norm_harmony_pt_no_rings, subset = donor == .y), features = .x, split.by = "Strain_DN", reduction = "umap.harmony_PC_man", order = T)))
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
likely_cnvs_don
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
pmap(list(likely_cnvs_don$donor, likely_cnvs_don$gene_id, likely_cnvs_don$labl), possibly(~DotPlot(subset(v3_m2_asex_cr_filt_norm_harmony_pt, subset = donor == ..1), features = ..2, group.by = "Strain_DN") + labs(title = ..3)))
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=3.5}
pmap(list(likely_cnvs_don$donor, likely_cnvs_don$gene_id, likely_cnvs_don$labl), possibly(~FeaturePlot(subset(v3_m2_asex_cr_filt_norm_harmony_pt, subset = donor == ..1), features = ..2, split.by = "Strain_DN", reduction = "umap.harmony_PC_man", order = F) + labs(title = ..3)))
```


```{r, warning=FALSE, message=FALSE}
top_prevalent_gns <- asx_pairwise_mkrs_seu_clustr_anot_all   %>%
  filter(p_val_adj < 1e-20 ) %>% 
  filter(abs(avg_log2FC)>1) %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  # filter(gene_DE_sample_prevalence >= 5) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T) %>% 
  filter(donor != "MSC55") %>%
  arrange(desc(abs(avg_log2FC)))

```


```{r, warning=FALSE, message=FALSE}
# as.list(split(top_prevalent_gns[,"donor"], top_prevalent_gns[,"gene_id"]))
top_gns_list <- tapply(top_prevalent_gns$gene_id, top_prevalent_gns$donor, c)
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
imap(top_gns_list, possibly(~DotPlot(subset(v3_m2_asex_cr_filt_norm_harmony_pt, subset = donor == .y), features = .x, group.by = "Strain_DN") + labs(title = .y)))
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
imap(top_gns_list, possibly(~DotPlot(subset(v3_m2_asex_cr_filt_norm_harmony_pt, subset = donor == .y), features = .x, group.by = "Strain_DN") + labs(title = .y)))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
asx_m2_bal_mast
```
