---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  # library(patchwork)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  # library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(knitr)
  # library(readxl)
  # library(SCpubr)
  # library(ggh4x)
  # library(htmlwidgets)
  library(slingshot)
  library(harmony)
  library(ggforce)
  library(tradeSeq)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::as.factor)
  
})
```


#  integration of all MSC


```{r setup}
sample_set = "Mali2"

knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("mali22_code_lnk/pf_common_vars.R")
```


```{r}
optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =6, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =2, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1, "MSC41"= 3, "MSC51"= 1, "MSC25"= 2)

optimum_k_qcd_21 <- c("MSC1" =5, "MSC3" =2, "MSC13" =2, "MSC14" =8, "MSC12" =2)

optimum_k_qcd <- c(optimum_k_qcd_21, optimum_k_qcd)

optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]
```



```{r, eval=T}
pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
```

## Genotype clustering for MSC

Read in clusters.tsv files with assignment of cells to strains and PCA plots metrics

Transfer clusters.tsv files with assignment of cells to strains and PCA plots metrics for the best K from farm to my local Mac

```{r}

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm

## poor quality samples
sample_rm <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39")
sample_mxd <- pf_solo_mixed_mdata_decode %>% filter(mixed_infection2 == "mixed") %>% pull(sample_nm)

# slct_sample_names[-c(sample_rm)]
slct_sample_names <- slct_sample_names[!(slct_sample_names %in% c(sample_rm, sample_mxd))]

## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```

Read in seurat objects processed in farm
```{r, message=F, warning=F}
# slct_sample_names <-c("MSC24", "MSC68", "MSC14", "MSC49", "MSC55") %>% str_sort(., numeric = T)
# slct_sample_names <-c("MSC24", "MSC68", "MSC14", "MSC55") %>% str_sort(., numeric = T)
sample_name <- slct_sample_names
# sample_name_ed_cr <- c(paste0(sample_name, "_edrops50"), paste0(sample_name, "_cbendr"))
sample_name_ed_cr <- c( paste0(sample_name, "_cb"))

sample_name_ed_cr
```

```{r, message=FALSE, warning=FALSE}
## save for running stdseurat in farm
asex_opts <- c('all', 'no_m12', 'no_m12_m55')
asex_opts_no_m12 <- asex_opts[2]
```

```{r, message=F, warning=F, eval=T}
##Read raw objects with metadata
m2_seu_cb_scvi_asex_norm_harmony <- list.files("data/processed/Pf/m2_all", pattern = "m2_seu_cb_scvi_asex.*_norm_harmony.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/m2_all|/|m2_seu_cb_scvi_asex_|_norm_harmony.RDS'))) %>% 
  # .[sample_name_ed_cr] %>%
  .[asex_opts_no_m12] %>%
  .[!is.na(.)] %>%
  map(readRDS)

```


```{r, message=F, warning=F, eval=T}
m2_scvi_asex_noM12_harmony <- m2_seu_cb_scvi_asex_norm_harmony[["no_m12"]]
```

################## START DE findmarkers


```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(m2_scvi_asex_noM12_harmony) <- "RNA"
m2_scvi_asex_noM12_harmony <- JoinLayers(m2_scvi_asex_noM12_harmony)

```

```{r, warning=FALSE, message=FALSE}
m2_scvi_asex_noM12_harmony[["RNA"]]$counts <- as(object = m2_scvi_asex_noM12_harmony[["RNA"]]$counts, Class = "dgCMatrix")
m2_scvi_asex_noM12_harmony[["RNA"]]$data <- as(object = m2_scvi_asex_noM12_harmony[["RNA"]]$data, Class = "dgCMatrix")
```

```{r, warning=FALSE, message=FALSE}
m2_scvi_asex_noM12_harmony[["RNA"]]["counts"]
```


```{r, warning=FALSE, message=FALSE}
DimPlot(m2_scvi_asex_noM12_harmony, reduction = "umap.harmony", group.by = "harmony_clusters", order = T, pt.size = 0.01) 

```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=4}
VlnPlot(m2_scvi_asex_noM12_harmony, features = c("nFeature_RNA", "nCount_RNA"), group.by = "harmony_clusters",pt.size = 0.001)

```

```{r, warning=FALSE, message=FALSE}
## DE
Idents(m2_scvi_asex_noM12_harmony) <- "harmony_clusters"
de_gns_fmall_cln <- FindAllMarkers(m2_scvi_asex_noM12_harmony)
# de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% slice_head(n=5)
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_cln_anot %>% filter(avg_log2FC > 0, cluster == .x))

```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE}
## Dormant (d5MDPs) genes after artemisinin - https://www-nature-com.ezp.lib.cam.ac.uk/articles/s41467-024-51846-0
d5MDP_up_gns <- c("PF3D7_1430900", "PF3D7_0928000", "PF3D7_0217100", "PF3D7_0412100", "PF3D7_1341100") %>% str_replace(., "_", "-")

d5MDP_down_gns<- c("PF3D7_0202500", "PF3D7_1102800", "PF3D7_1401400", "PF3D7_0202000") %>% str_replace(., "_", "-")


```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_up_gns)
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_down_gns)
```




################## END DE findmarkers


Slingshot
```{r, warning=FALSE, message=FALSE}
DefaultAssay(m2_scvi_asex_noM12_harmony) <- "RNA"
```

```{r, warning=FALSE, message=FALSE}
DimPlot(m2_scvi_asex_noM12_harmony, reduction = "umap.harmony", group.by = "harmony_clusters", order = T, pt.size = 0.01) 

```

```{r, warning=FALSE, message=FALSE}
m2_scvi_asex_noM12_harmony <- FindClusters(m2_scvi_asex_noM12_harmony, resolution = 0.1, cluster.name = "harmony_clusters_ss") 
```


```{r, warning=FALSE, message=FALSE}
DimPlot(m2_scvi_asex_noM12_harmony, reduction = "umap.harmony", group.by = "harmony_clusters_ss", order = T, pt.size = 0.01) 

```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(m2_scvi_asex_noM12_harmony@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= m2_scvi_asex_noM12_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, eval=T}
# gogo()
m2_scvi_asex_noM12_harmony_sce <- as.SingleCellExperiment(m2_scvi_asex_noM12_harmony)

```

```{r, eval=T}
rowData(m2_scvi_asex_noM12_harmony_sce)$feature_symbol <- rownames(m2_scvi_asex_noM12_harmony_sce)
```

!!!! NOTE -BEEN DOING SLINGSHOT WRONG. LOGCOUNTS SHOULD HAVE data values from seurat
```{r, eval=T}
# gogo()
# Seurat_object <- m2_asex_harmony_res

## Convert seurat object to SCE
m2_scvi_asex_noM12_harmony_sce_ss <- slingshot(m2_scvi_asex_noM12_harmony_sce,
            clusterLabels = "harmony_clusters_ss",
            reducedDim = "UMAP.HARMONY",
            start.clus = 2,
            end.clus = 3,
            # shrink =1,
            # stretch = 1,
            # allow.break = F
            )


```


```{r}
##Prepare slingshot dataset (SDS) object
m2_scvi_asex_noM12_harmony_sce_ss_sds <- SlingshotDataSet(m2_scvi_asex_noM12_harmony_sce_ss)

# plot(reducedDims(m2_scvi_asex_noM12_harmony_sce_ss)$UMAP[,1:2], col = brewer.pal(9,"Set1")[as.factor(m2_scvi_asex_noM12_harmony_sce_ss$stage)], pch=16, asp = 0.7)
plot(reducedDims(m2_scvi_asex_noM12_harmony_sce_ss)$UMAP.HARMONY[,c(1,2)], col = brewer.pal(9,"Set1")[as.factor(m2_scvi_asex_noM12_harmony_sce_ss$harmony_clusters_ss)], pch=16, asp = 0.7, )

lines(m2_scvi_asex_noM12_harmony_sce_ss_sds, lwd=2, col="black",dims = c(1,2))
```


```{r}
## Add pseudotime metadata to the object
m2_scvi_asex_noM12_harmony_pt <- AddMetaData(m2_scvi_asex_noM12_harmony, 
                                                      data.frame(m2_scvi_asex_noM12_harmony_sce_ss@colData[, c("slingPseudotime_1"), drop = F]) #%>%
                                                        # rename("pseudotime" = slingPseudotime_1)
                                                      ) 

```


```{r}
FeaturePlot(m2_scvi_asex_noM12_harmony_pt, features = "slingPseudotime_1", reduction = "umap.harmony")

```

```{r}
## Save pseudotime objects
saveRDS(m2_scvi_asex_noM12_harmony_pt, "data/processed/Pf/m2_scvi_asex_noM12_harmony_pt.RDS")
saveRDS(m2_scvi_asex_noM12_harmony_sce_ss, "data/processed/Pf/m2_scvi_asex_noM12_harmony_sce_ss.RDS")
saveRDS(m2_scvi_asex_noM12_harmony_sce_ss_sds, "data/processed/Pf/m2_scvi_asex_noM12_harmony_sce_ss_sds.RDS")

### GOTO m22_raw_tseq_MAST_intradon_DE for tradeseq
```









######################## END integrate cellranger filtered with rescued asexuals ########################

########END TRY SCT ########


```{r, warning=FALSE, message=FALSE}
## mget

imap(all_ed, ~FeaturePlot(.x, features = "PF3D7-1328800", order = T) + labs(title = .y))
imap(all_ed, ~FeaturePlot(.x, features = "PF3D7-1328800", reduction = "pca", order = T) + labs(title = .y))
```




##############START - Painter EArly asexual markers##############
```{r}
painter_raw = readxl::read_xlsx("../published_dsets/painter_29985403_48hr_dynamics/Expression_Peak_Timing_48h_IDC.xlsx", sheet = "Estimated Total Abundance", .name_repair = "universal") %>% mutate(across(Gene.ID, ~str_replace(., "_", "-"))) %>% rename_all(., ~str_remove(., "\\.\\.")) %>% rename_with(.fn = ~str_remove(paste0("hpi_", .x), ".hpi"), .cols = contains("hpi")) %>% rename("peak_hpi" = `hpi_Peak.Tim.`)
```

```{r}
painter_raw 
```


```{r}

# Get the column names that start with 'hpi'
hpi_cols <- grep("^hpi_[0-9]+$", names(painter_raw), value = TRUE)

# Dynamically split the columns into groups of 5
col_groups <- split(hpi_cols, ceiling(seq_along(hpi_cols) / 10))

# Compute the row-wise means for each group of 5 columns
means_list <- lapply(col_groups, function(cols) {
  rowMeans(painter_raw[, cols], na.rm = TRUE)
})

# Combine the results into a single data frame
means_df <- data.frame(means_list)
names(means_df)<- paste0("mean_", seq_along(means_df))

# Combine the original dataframe with the means
painter_raw_summary <- cbind(painter_raw[, c("Gene.ID", "Annotation")], means_df)

```


```{r}
painter_raw_summary  %>% arrange(desc(mean_1))
```

```{r}
painter_raw %>% filter(Gene.ID %in% c("PF3D7-1102800", "PF3D7-0202500", "PF3D7-1370300")) %>% 
  pivot_longer(cols = starts_with("hpi_"), names_to = "time", values_to = "expresion") %>% 
  mutate(across(time, ~as.numeric(str_extract(., "[0-9]+")))) %>% 
  ggplot(aes(x = time, y = expresion, color = Gene.ID)) + 
  geom_point()
```


```{r, warning=FALSE, message=FALSE}
## mget
imap(all_ed, ~FeaturePlot(.x, features = c("PF3D7-0501300", "PF3D7-1102800"), order = T) + labs(title = .y))
imap(all_ed, ~FeaturePlot(.x, features = c("PF3D7-0501300", "PF3D7-1102800"), reduction = "pca", order = T) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
## mget
FeaturePlot(all_msc_100_700, features = c("PF3D7-1370300", "PF3D7-1102800"), order = T, pt.size = 0.5)
FeaturePlot(all_msc_100_700, features = c("PF3D7-1370300", "PF3D7-1102800"), reduction = "pca", order = T, pt.size = 0.5) 
```

##############END - Painter EArly asexual markers##############

## scmap annotation

### Preprocessing references

#### msc_combi reference dataset
```{r, message=FALSE, warning=FALSE, eval = T}
## v3labmscs.ref.sce.stgasx.RDS generated in m2_QC_prelim_stage_anno 
v3labmscs.ref.sce.stgasx <- readRDS("data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS")
```


```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_edrops50_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd_edrops50.RDS" --stg_labl "stageHL_ref"

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```
SCT for raw
```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_edrops50_sct.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd_edrops50.RDS" --stg_labl "stageHL_ref"

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_sct.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```

```{r, message=F, warning=F}
##Read the annotations
sample_name_ed <- sample_name_all[!str_detect(sample_name_all, "_150")]
sample_name_ed <- sample_name_all
```

```{r, message=F, warning=F}
##Read the annotations
m2_raw_anots <- list.files("data/processed/Pf", pattern = "*prelim_anotd.*RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|prelim_anotd|.RDS'))) %>% 
  # .[sample_name_ed] %>%
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(readRDS)
```

```{r}
## Adding annotations to query cells
m2_raw_anotd_preQC <- list(m2_raw_anots, all_ed[sample_name_ed]) %>%
  pmap(~ ..1 %>%
         column_to_rownames("cell_bc") %>%
         mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
         # mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
         mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_replace_all(.x, c("gametocyte" = "Gametocyte", "late" = "Late", "early" = "Early")))) %>%
         mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
         AddMetaData(..2, .) 
  )

```

```{r}
## Adding annotations to query cells
m2_raw_anotd_preQC <- list(m2_raw_anotd_preQC) %>%
  pmap(~ ..1@meta.data %>%
         mutate(gene100 = ifelse(nFeature_RNA > 100, "yes", "no"),
                trancript100 = ifelse(nCount_RNA > 100, "yes", "no")) %>%
         AddMetaData(..1, .) 
  )

```

```{r}
# ## Adding annotations to query cells
# m2_raw_anotd_preQC <- list(m21_gn150_anots, all_ed) %>%
#   pmap(~ ..1 %>%
#          column_to_rownames("cell_bc") %>%
#          mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
#          mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
#          mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
#          AddMetaData(..2, .) 
#   )

```

```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'Stage_MSC', cols = sp_fld_colors) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'stage_lab1', cols = sp_fld_colors2)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'gene100')+ labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(Stage_MSC, stage_lab1))

```


```{r}
##Modify the stage labels for MSC14
# m2_raw_anotd_preQC[[4]] <- m2_raw_anotd_preQC[[4]]@meta.data %>%
m2_raw_anotd_preQC <- pmap(list(m2_raw_anotd_preQC, rep(0.4,length(sample_name_all)), c(rep('ring|trophozoite|developing', length(sample_name_all)))), ~{
  ..1@meta.data %>%
    mutate(across(Stage_MSC, as.character)) %>%
    # mutate(Stage_MSC = Stage_MSC) %>%
    mutate(StagePrelim_lf = case_when(
      is.na(Stage_MSC) & sim_lab1 > 0.35 & str_detect(stage_lab1, ..3) & str_detect(stage_lab2, ..3) & str_detect(stage_lab3, ..3) ~ stage_lab1, ## lower cos sim of 0.35 for asexuals instead of 0.4
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Male|male') & str_detect(stage_lab2, 'Male|male') & str_detect(stage_lab3, 'Male|male') ~ stage_lab1,
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Female|female') & str_detect(stage_lab2, 'Female|female') & str_detect(stage_lab3, 'Female|female') ~ stage_lab1,
      TRUE ~ Stage_MSC)#,
      # across(StagePrelim, str_to_sentence)
    )%>%
    mutate(StagePrelim = case_when(
      str_detect(StagePrelim_lf, "developing") ~ "Early trophozoite", ## !!NB - label developing gams as early trophs the closest stage
      str_detect(StagePrelim_lf, "schizont") & str_detect(stage_lab2, "ring|trophozoite")  & str_detect(stage_lab3, "ring|trophozoite") ~ str_to_sentence(stage_lab2),
      str_detect(StagePrelim_lf, "late|early") ~ str_to_sentence(StagePrelim_lf),
      str_detect(StagePrelim_lf, "gametocyte") ~ str_to_sentence(str_remove_all(StagePrelim_lf, "gametocyte \\(|\\)")),
      is.na(StagePrelim_lf) ~ "Unassigned",
      TRUE ~ StagePrelim_lf)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(StagePrelimC = case_when(
      str_detect(StagePrelim, "ring|trophozoite|developing|schizont") ~ "Asexual",
      TRUE ~ StagePrelim)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(across(c(StagePrelim, StagePrelimC), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    dplyr::select(StagePrelim, StagePrelim_lf, StagePrelimC) %>%
    # mutate(across(starts_with('Stage_V'), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    AddMetaData(.x, .)
})

```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~.x@meta.data %>% dplyr::count(Stage_MSC, StagePrelim, stage_lab1))
# map(m2_raw_anotd_preQC, ~.x@meta.data %>% dplyr::count(StagePrelim, StagePrelimC))
```

```{r, warning=FALSE, message=FALSE}
imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = F, group.by = 'StagePrelim', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}
# m21_gn150_anots[["MSC51"]]
```


```{r, warning=FALSE, message=FALSE}
imap(m2_raw_anotd_preQC, ~FeaturePlot(.x, dims = c(1,2), features = "nFeature_RNA")+ labs(title = .y))
```

```{r, results='asis',  fig.width = 8, fig.height = 4}
##adding clusters that are used to display mitochondrial count

plts <- list(m2_raw_anotd_preQC, names(m2_raw_anotd_preQC)) %>%
  pmap(., ~{
    
    cell_qc_plot <-  ..1 %>% DimPlot(., pt.size = 1, reduction = 'umap', dims = c(1,2), group.by = 'StagePrelim', cols = sp_fld_colors) +labs(title = ..2)#,
    
    cell_qc_plot$data <- cell_qc_plot$data %>% rownames_to_column('bcode') %>% left_join(., ..1@meta.data[,'seurat_clusters', drop = F] %>% rownames_to_column('bcode'), by='bcode') %>% column_to_rownames('bcode')
    
    # print(str(cell_qc_plot$data))
    
    cell_qc_plot<- LabelClusters(cell_qc_plot, id = "seurat_clusters", size = 9, repel = T)
    
    return(cell_qc_plot)

  })
```


```{r, results='asis',  fig.width = 8, fig.height = 4}
plts

```

compare filt and raw with gn count cut off of 150
```{r, message=F, warning=F}
# go()
##Read the annotations
m21_all_anots_mdata <- map2(m2_raw_anotd_preQC[1:3], m2_raw_anotd_preQC[4:6], ~{
  .x@meta.data %>% 
    rownames_to_column("bcode") %>%
    full_join(., .y@meta.data %>% rownames_to_column("bcode"), suffix = c("_filt", "_gn_150"), by = "bcode") %>%
    mutate(dset = case_when(!is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "both",
                            !is.na(nFeature_RNA_filt) & is.na(nFeature_RNA_gn_150)  ~ "filt",
                            is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "gn_150"))
})
```


```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% count(dset))
map(m21_all_anots_mdata, ~.x %>% count(dset, stage_lab1_gn_150))

```

```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% filter(dset == "gn_150") %>% count(StagePrelim_gn_150, Stage_MSC_gn_150, stage_lab1_gn_150, stage_lab2_gn_150, stage_lab3_gn_150))

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = sim_lab1_gn_150) ) +
      geom_boxplot() +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = nFeature_RNA_gn_150) ) +
      geom_boxplot() +
      geom_hline(yintercept = 100) +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```

Umap with stage labels in legend

```{r, fig.width = 7, fig.height = 5, eval=F}
trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.2, "cm")
)

# Assign the female High_low stage of the projected cell to the nearest cell in the reference dataset
# msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(msc_combi.sce))), length(m21_gn150_jc))
msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(v3labmscs.ref.sce.stgasx))), length(m21_gn150_jc))


for (i in 1:length(msc_qcd_anotd_mali_mapped14HL)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m2_raw_anotd_preQC[[i]]$StagePrelim
}


for (i in 1:length(m2_raw_anotd_preQC)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m2_raw_anotd_preQC[[i]]$StagePrelim
  # msc_combi@meta.data[, paste0("stage_", names(m2_raw_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  v3labmscs.ref.sce.stgasx@colData[, paste0("stage_", names(m2_raw_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  
}

labs_order <- c('Late ring', 'Early trophozoite','Female (HE)',  'Male (HE)', 'Female (LE)', 'Male (LE)','Gametocyte (developing)', 'Gametocyte (male)','Gametocyte (female)',  'Atlas')
names(labs_order)<- labs_order

umap_theme <- theme(legend.position = 'right',
                    axis.title=element_text(size=12,face="bold",hjust = 0),
                    axis.text=element_text(size=18), 
                    plot.title = element_blank(),
                    legend.title=element_text(size=14,face="bold"),
                    legend.text=element_text(size=14),
                    legend.spacing.y = unit(0.25, "cm"),
                    axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed"))
)  


# Idents(msc_combi) <- 'MSC14_V3'
# m14_scmap_mca_umap <- map(names(m2_raw_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), 
#         order = labs_order) +
m14_scmap_mca_umap <- map(names(m2_raw_anotd_preQC), ~{
  cbind(reducedDims(v3labmscs.ref.sce.stgasx)$PCA, colData(v3labmscs.ref.sce.stgasx)[, c("stageHL_ref", paste0("stage_",.x)), drop=F]) %>% 
    data.frame() %>% 
    mutate(od_stg = droplevels(factor(!!rlang::sym(paste0("stage_",.x)), levels = rev(stg_refnd_lvls)))) %>%
    arrange(od_stg) %>%
    ggplot(aes(y = PC2, x=PC3, color = od_stg)) +
    geom_point() +
    # scale_color_manual(breaks=labs_order, values = sp_fld_colors, labels = str_wrap(labs_order, 10))+
    scale_color_manual(values = sp_fld_colors)+
    umap_theme+
    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
           x= trunc_axis,
           y= trunc_axis)+
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(breaks = NULL)
})

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap, 'plots/fig4/m14_scmap_mca_umap')

m14_scmap_mca_umap


```

Umap with stage labels inside
```{r, fig.width = 7, fig.height = 5, eval=F}

m14_scmap_mca_umap_labs_in <- map(names(m2_raw_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), label = T, order = labs_order, repel = T, label.size = 5) +
                                    scale_color_manual(values = sp_fld_colors)+ 
                                    umap_theme+
                                    theme(axis.title=element_text(size=18,face="bold"), 
                                          plot.title = element_blank(),
                                          legend.position='None'
                                    ) +
                                    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
                                           x= trunc_axis,
                                           y= trunc_axis)+
                                    scale_x_continuous(breaks = NULL) +
                                    scale_y_continuous(breaks = NULL)
)

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap_labs_in, 'plots/fig4/m14_scmap_mca_umap_labs_in.RDS')


m14_scmap_mca_umap_labs_in


```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim, Stage_MSC, stage_lab1, stage_lab2))

```


```{r, warning=FALSE, message=FALSE}
saveRDS(m2_raw_anotd_preQC, "data/processed/Pf/m2_raw_anotd_preQC.RDS")
```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data) %>% saveRDS(., "data/processed/Pf/m2_raw_anotd_preQC_mdata.RDS")

```

```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim))

```




```{r, fig.width = 12, fig.height = 4}

##Plotting top two scmap stages among the top 10 with their similarities for MSC14
# 
# stage_fct_labs <- msc_qcd_anotd[[1]]@meta.data  %>% dplyr::group_by(stageHL, seurat_clusters) %>% dplyr::group_keys() %>% dplyr::mutate(across(c(stageHL, seurat_clusters), as.character), across(seurat_clusters, ~paste0("Cluster ",., ':\n',stageHL))) %>% deframe()
# 
# scm_barpt <- msc_qcd_anotd[[1]]@meta.data %>% 
#   dplyr::select(matches('s.*lab.'), stageHL) %>% 
#   rownames_to_column('cell_bc')%>%
#   pivot_longer(-c('cell_bc', 'stageHL'),
#                names_to = c( ".value", "Label"),
#                names_pattern = "(s.*)_(l.*)")%>% 
#   filter(!is.na(sim)) %>%
#   mutate(across(sim, as.numeric)) %>% 
#   mutate(across(Label, ~str_replace(., 'lab','Label '))) %>%
#   mutate(across(stage, ~droplevels(factor(., stg_refnd_lvls)))) %>%
#   ggplot(., aes(y = sim, x = stage, color = Label)) +
#   geom_boxplot(aes(fill = Label))+
#   scale_fill_manual(values =  c('Label 1' = '#007FFF', 'Label 2' = '#FFFF9F'), name = "Label",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 2)
#   )+
#   scale_color_manual(values =  c('Label 1' = 'black', 'Label 2' = 'black'), name = "Label")+ 
#   geom_point(aes(group=Label), size = 0.5, position = position_jitterdodge(jitter.width = 0.2))+
#   new_scale_fill()+
#   geom_tile(data = data.frame(x = c( 'Late ring', 'Early trophozoite', 'Gametocyte (developing)', 'Gametocyte (female)', 'Gametocyte (male)')), inherit.aes = F, aes(x = x, y = 0.37, fill = x, height = 0.03)) +
#   scale_fill_manual(values =  sp_fld_colors, name = "MCA stage",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 3)
#                     # guide = 'none'
#   )+
#   scale_y_continuous(expand = expansion(mult = c(0.01, 0.05))) +
#   facet_grid(.~stageHL, labeller = labeller(stageHL = stage_fct_labs)) +
#   geom_hline(yintercept = 0.4)+
#   labs(y = 'scmap \nsimilarity scores', x = 'MCA stage') +
#   theme_classic() +
#   theme(axis.text.x = element_blank(),
#         axis.title=element_text(size=16,face="plain"), 
#         axis.text = element_text(size=15), 
#         axis.ticks.x = element_blank(),
#         strip.text =element_text(size=16,face="plain", colour = 'black'),
#         panel.border=element_rect(colour="black",fill='transparent'),
#         legend.title = element_text(size=16,face="bold"),
#         legend.text = element_text(size=15),
#         legend.position = 'bottom')
# 
# scm_barpt
```


Run souporcell like below for the filtered barcodes of all infections
```{r}
gogo()

# nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/MSC*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```



Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
gogo()

#  nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373{63,72,89}/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/Pf_mxd/MSC*/outs/sk21_bcodes_pp/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```
