---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
})
```

# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}
# library(reticulate)
# # use_condaenv("/Users/jr35/envs/scv_jupcon/", required = TRUE)
# # matplotlib <- reticulate::import("matplotlib")
# # matplotlib$use("Agg", force = TRUE)
# 
# knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali')
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```


## Read in entire object and subset asexuals

```{r, warning=FALSE, message=FALSE, eval=F}
all_msc_cb_no_dbt_mrg_noM12 <- readRDS("data/processed/Pf/m2_all/all_msc_cb_no_dbt_mrg_noM12.RDS")
```


```{r, warning=FALSE, message=FALSE, eval=F}
all_msc_cb_no_dbt_mrg_noM12@meta.data %>% count(sR_fine_stg)
```

```{r, warning=FALSE, message=FALSE, eval=F}
all_msc_cb_no_dbt_mrg_noM12@meta.data %>% count(donor, sR_fine_stg)
```


```{r, eval=F}
## Making list of objects and renaming before integrating
all_msc_cb_no_dbt_mrg_noM12_asex <- subset(all_msc_cb_no_dbt_mrg_noM12, subset = sR_fine_stg %in% c("Early ring", "Late ring", "Early trophozoite", "Gametocyte (developing)"))
  
## Join layers before conversion
DefaultAssay(all_msc_cb_no_dbt_mrg_noM12_asex) <- "RNA"
saveRDS(all_msc_cb_no_dbt_mrg_noM12_asex, "data/processed/Pf/m2_all/all_msc_cb_no_dbt_mrg_noM12_asex.RDS")

```

```{r, warning=FALSE, message=FALSE}
# readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony_res_fld_pt.RDS")
```

########START TRY SCT ########

Integrated asexual to integrate with the rescued parasites
```{r}
## subset where field parasites are for slingshot
# v3_m2_asex_harmony_ss_res_fld_pt <- readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony_ss_res_fld_pt.RDS")
# v3_m2_asex_harmony_ss_res_fld_pt <- readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony_res_fld_pt.RDS")
v3_m2_asex_harmony_ss_res_fld_pt <- readRDS("data/processed/Pf/m2_all/all_msc_cb_no_dbt_mrg_noM12_asex.RDS")

# %in% c("MSC25" ,"MSC41", "MSC51")

```



```{r, warning=FALSE, message=FALSE}
v3_m2_asex_harmony_ss_res_fld_pt@meta.data %>% count(sR_fine_stg)
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_harmony_ss_res_fld_pt <- v3_m2_asex_harmony_ss_res_fld_pt@meta.data %>% 
  rownames_to_column('bcode') %>%
  mutate(dset_labmrg = "Field",
         stage = sR_fine_stg,
         donor = str_remove_all(orig.ident, ".mrna")) %>% 
  add_count(donor, name = "don_count") %>%
  select(bcode, dset_labmrg, donor, don_count, stage) %>%
  column_to_rownames('bcode') %>%
  AddMetaData(v3_m2_asex_harmony_ss_res_fld_pt, .)
  
```



```{r, warning=FALSE, message=FALSE}
v3_m2_asex_harmony_ss_res_fld_pt@meta.data %>% dplyr::count(donor, dset_labmrg)
```

```{r}
## remove lab and M12
v3_m2_asex_cr_filt <- v3_m2_asex_harmony_ss_res_fld_pt[, !v3_m2_asex_harmony_ss_res_fld_pt@meta.data$dset_labmrg %in% c("Lab", "MSC12")]
```


```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% count(donor, sR_fine_stg) %>% group_by(donor) %>% mutate(prop = round(n/sum(n), 3)) %>% filter(sR_fine_stg == "Gametocyte (developing)")
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% add_count(sR_fine_stg, name = "stg_n") %>% group_by() %>% dplyr::count(donor)
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% filter(sR_fine_stg == "Gametocyte (developing)") %>% dplyr::count(donor)
```

```{r, eval=T}
## Making list of objects and renaming before integrating
v3_m2_asex_cr_filt <- subset(v3_m2_asex_cr_filt, subset = donor != "MSC55")

```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% dplyr::count(donor, dset_labmrg)
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% add_count(donor) %>% dplyr::filter( n < 100)
```

```{r}
## remove weird 2 cells with >9000 ncountRNA
VlnPlot(v3_m2_asex_cr_filt, features = "nCount_RNA") + geom_hline(yintercept = 9000)
```

```{r}
## remove weird 2 cells with >9000 ncountRNA
table(v3_m2_asex_cr_filt@meta.data$nCount_RNA >9000)
v3_m2_asex_cr_filt <- v3_m2_asex_cr_filt[, !v3_m2_asex_cr_filt@meta.data$nCount_RNA >9000]
```

```{r, warning=FALSE, message=FALSE}
# split layers
# v3_m2_asex_cr_filt[["RNA"]] <- split(v3_m2_asex_cr_filt[["RNA"]], f = v3_m2_asex_cr_filt$donor)

```


```{r}
## remove any donor with less than 100 cells
v3_m2_asex_cr_filt <- v3_m2_asex_cr_filt[, !v3_m2_asex_cr_filt@meta.data$don_count < 150]
```


```{r, warning=FALSE, message=FALSE}
# run sctransform
v3_m2_asex_cr_filt_sct <- SCTransform(v3_m2_asex_cr_filt, verbose = FALSE, variable.features.n = 400)
# v3_m2_asex_cr_filt_sct <- SCTransform(v3_m2_asex_cr_filt, verbose = FALSE, variable.features.n = 300)

```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3_m2_asex_cr_filt_sct <- RunPCA(v3_m2_asex_cr_filt_sct, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(v3_m2_asex_cr_filt_sct, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
# Error in getGlobalsAndPackages(expr, envir = envir, globals = globals) : 
options(future.globals.maxSize = 1000 * 1024^2)  # 1000 MiB (or 1 GiB)

v3_m2_asex_cr_filt_sct_int <- IntegrateLayers(
  object = v3_m2_asex_cr_filt_sct,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  # k.weight = 90,
  # k.weight = 82,
)


```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% dplyr::count(donor)
```

```{r, warning=FALSE, message=FALSE}

v3_m2_asex_cr_filt_sct_int <- FindNeighbors(v3_m2_asex_cr_filt_sct_int, dims = 1:5, reduction = "integrated.dr")
v3_m2_asex_cr_filt_sct_int <- FindClusters(v3_m2_asex_cr_filt_sct_int, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3_m2_asex_cr_filt_sct_int <- RunUMAP(v3_m2_asex_cr_filt_sct_int, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "integrated.dr")
v3_m2_asex_cr_filt_sct_int <- FindNeighbors(v3_m2_asex_cr_filt_sct_int, dims = 1:5, reduction = "integrated.dr", verbose = FALSE)
v3_m2_asex_cr_filt_sct_int <- FindClusters(v3_m2_asex_cr_filt_sct_int, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "sR_fine_stg",  dims = c(1,2)) # + scale_color_manual(values = sp_fld_colors)

```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_cr_filt_sct_int@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= v3_m2_asex_cr_filt_sct_int@meta.data$sR_fine_stg,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, reduction = "pca")
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T)
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "sR_fine_stg")# + scale_color_manual(values = sp_fld_colors)
DimPlot(v3_m2_asex_cr_filt_sct_int, group.by = "donor")
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "nFeature_RNA")
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "nCount_RNA")
VlnPlot(v3_m2_asex_cr_filt_sct_int, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "sR_fine_stg", reduction = "integrated.dr", dims = c(1,2))# + scale_color_manual(values = sp_fld_colors)

```


```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(2,3))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, dims = c(1,3))

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", order = T) 
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", reduction = "pca", order = T) 
```
```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(v3_m2_asex_cr_filt_sct_int) <- "RNA"
v3_m2_asex_cr_filt_sct_int <- NormalizeData(v3_m2_asex_cr_filt_sct_int) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(v3_m2_asex_cr_filt_sct_int) <- "RNA"
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", order = T) 
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", reduction = "pca", order = T) 
```



```{r, warning=FALSE, message=FALSE}
## ap2g
v3_m2_asex_cr_filt_sct_int_ap2g <- subset(v3_m2_asex_cr_filt_sct_int, subset = `PF3D7-1222600` > 0, slot = 'counts')

table(v3_m2_asex_cr_filt_sct_int_ap2g$donor)
```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1215500", order = T) 
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1215500", reduction = "pca", order = T) 
```


```{r, warning=FALSE, message=FALSE}

DefaultAssay(v3_m2_asex_cr_filt_sct_int) <- "SCT"
```

######################## START integrate cellranger filtered with rescued asexuals ########################
```{r, warning=FALSE, message=FALSE}
# run sctransform
all_resc_asex_rna <-  all_resc_asex_sct
```

```{r, warning=FALSE, message=FALSE}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(all_resc_asex_sct_rna)) {
  DefaultAssay(all_resc_asex_rna[[i]]) <- "RNA"
  all_resc_asex_rna[[i]]@assays$SCT <- NULL
  all_resc_asex_rna[[i]]@project.name <- "SeuratProject"
}

all_resc_asex_rna
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
v3_m2_asex_cr_filt_list <-  SplitObject(v3_m2_asex_cr_filt, split.by = "donor")
```


```{r, warning=FALSE, message=FALSE}
## Donors with cells in the rescued object
resc_dons <- names(v3_m2_asex_cr_filt_list)[names(v3_m2_asex_cr_filt_list) %in% names(all_resc_asex_rna)]
non_resc_dons <- names(v3_m2_asex_cr_filt_list)[!names(v3_m2_asex_cr_filt_list) %in% names(all_resc_asex_rna)]

```

```{r, warning=FALSE, message=FALSE}
## Donors with cells in the rescued object
all_asex_rna <- list()
```


```{r, warning=FALSE, message=FALSE}
## for those with rescued cells merge these with cell ranger qcd cells
for (i in resc_dons) {
  all_asex_rna[i] <- merge(v3_m2_asex_cr_filt_list[[i]], all_resc_asex_rna[[i]])
  
}

all_asex_rna
```

```{r, warning=FALSE, message=FALSE}
## Add those without rescued cells to the object above created after merging
v3_m2_asex_cr_filt_no_resc <- v3_m2_asex_cr_filt_list[non_resc_dons]
all_asex_merged_list <- c(v3_m2_asex_cr_filt_no_resc, all_asex_rna)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- merge(all_asex_merged_list[[1]], all_asex_merged_list[2:length(all_asex_merged_list)])
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- JoinLayers(all_asex_merged)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged[["RNA"]] <- split(all_asex_merged[["RNA"]], f = all_asex_merged$donor)
all_asex_merged
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
all_asex_merged <- SCTransform(all_asex_merged, verbose = FALSE, variable.features.n = 200)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged <- RunPCA(all_asex_merged, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int <- IntegrateLayers(
  object = all_asex_merged,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int <- FindNeighbors(all_asex_merged_int, dims = 1:6, reduction = "integrated.dr")
all_asex_merged_int <- FindClusters(all_asex_merged_int, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int <- RunUMAP(all_asex_merged_int, dims = 1:6, verbose = FALSE, n.components = 3, seed.use = 867)
all_asex_merged_int <- FindNeighbors(all_asex_merged_int, dims = 1:6, verbose = FALSE)
all_asex_merged_int <- FindClusters(all_asex_merged_int, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, reduction = "pca")
DimPlot(all_asex_merged_int, label = T)
DimPlot(all_asex_merged_int, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int, group.by = "donor")
FeaturePlot(all_asex_merged_int, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int, features = "nCount_RNA")
VlnPlot(all_asex_merged_int, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(2,3))
DimPlot(all_asex_merged_int, label = T, dims = c(2,3))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, dims = c(1,3))

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```
```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int) <- "RNA"
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int) <- "SCT"
```



```{r, warning=FALSE, message=FALSE}
## ap2g
all_asex_merged_int_ap2g <- subset(all_asex_merged_int, subset = `PF3D7-1222600` > 0, slot = 'counts')

table(all_asex_merged_int_ap2g$donor)
```

```{r, warning=FALSE, message=FALSE}
# ## change ident to ap2g
# for (i in 1:length(all_asex_merged_int)) {
#   Idents(all_asex_merged_int[[i]]) <- "ap2_clst_finer"
# }

```

```{r, warning=FALSE, message=FALSE}
## ap2g
all_asex_merged_int <- JoinLayers(all_asex_merged_int)
DefaultAssay(all_asex_merged_int) <- "RNA"

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int)
de_gns_fmall_all <- de_gns_fmall_all %>% arrange(p_val_adj)

```


```{r, warning=FALSE, message=FALSE}
de_gns_sig_anot <- de_gns_fmall_all %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_anot
```

```{r, warning=FALSE, message=FALSE}
VlnPlot(all_asex_merged_int, features = c("nFeature_RNA", "nCount_RNA"))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T)
```




```{r, warning=FALSE, message=FALSE, eval=TRUE}
all_asex_merged_int@meta.data %>% filter(seurat_clusters == 4) %>% count(donor, seurat_clusters)

```


```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

# Cluster 4 highly expressess truRNA which is switched on when temperature is lowered from 37C to 26C activates!!! https://www.sciencedirect.com/science/article/pii/S0378111923003578
# https://pmc.ncbi.nlm.nih.gov/articles/PMC11597781/

```

```{r, warning=FALSE, message=FALSE}
gogo()
```

############ SCT2 -without activated asexuals
Remove the low temperature activate asexuals and cluster the rest
```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv <- all_asex_merged_int[, all_asex_merged_int$seurat_clusters != 4]
DimPlot(all_asex_merged_int_no_actv, label = T)
```


```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv[["RNA"]] <- split(all_asex_merged_int_no_actv[["RNA"]], f = all_asex_merged_int_no_actv$donor)
all_asex_merged_int_no_actv
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
all_asex_merged_int_no_actv <- SCTransform(all_asex_merged_int_no_actv, verbose = FALSE, variable.features.n = 100)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv <- RunPCA(all_asex_merged_int_no_actv, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged_int_no_actv, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int_no_actv <- IntegrateLayers(
  object = all_asex_merged_int_no_actv,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv <- FindNeighbors(all_asex_merged_int_no_actv, dims = 1:5, reduction = "integrated.dr")
all_asex_merged_int_no_actv <- FindClusters(all_asex_merged_int_no_actv, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv <- RunUMAP(all_asex_merged_int_no_actv, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867)
all_asex_merged_int_no_actv <- FindNeighbors(all_asex_merged_int_no_actv, dims = 1:5, verbose = FALSE)
all_asex_merged_int_no_actv <- FindClusters(all_asex_merged_int_no_actv, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv, label = T, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv, label = T)
DimPlot(all_asex_merged_int_no_actv, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int_no_actv, group.by = "donor")
FeaturePlot(all_asex_merged_int_no_actv, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int_no_actv, features = "nCount_RNA")
VlnPlot(all_asex_merged_int_no_actv, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv) <- "RNA"
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int_no_actv) <- "SCT"
```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv) <- "RNA"
all_asex_merged_int_no_actv <- JoinLayers(all_asex_merged_int_no_actv)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_cln <- FindAllMarkers(all_asex_merged_int_no_actv)
de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv, label = T)
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

```

```{r, warning=FALSE, message=FALSE}
## Dormant (d5MDPs) genes after artemisinin - https://www-nature-com.ezp.lib.cam.ac.uk/articles/s41467-024-51846-0
d5MDP_up_gns <- c("PF3D7_1430900", "PF3D7_0928000", "PF3D7_0217100", "PF3D7_0412100", "PF3D7_1341100") %>% str_replace(., "_", "-")

d5MDP_down_gns<- c("PF3D7_0202500", "PF3D7_1102800", "PF3D7_1401400", "PF3D7_0202000") %>% str_replace(., "_", "-")


```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_up_gns)
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_down_gns)
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE}
VlnPlot(all_asex_merged_int, features = c("nFeature_RNA", "nCount_RNA"))

```

############ Harmony -without activated asexuals
Remove the low temperature activate asexuals and cluster the rest
```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony <- all_asex_merged_int[, all_asex_merged_int$seurat_clusters != 4]
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
```


```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony[["RNA"]] <- split(all_asex_merged_int_no_actv_harmony[["RNA"]], f = all_asex_merged_int_no_actv_harmony$donor)
all_asex_merged_int_no_actv_harmony
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
# all_asex_merged_int_no_actv_harmony <- SCTransform(all_asex_merged_int_no_actv_harmony, verbose = FALSE, variable.features.n = 100)

all_asex_merged_int_no_actv_harmony <- NormalizeData(all_asex_merged_int_no_actv_harmony)
all_asex_merged_int_no_actv_harmony <- FindVariableFeatures(all_asex_merged_int_no_actv_harmony, selection.method = "vst", nfeatures = 300)
all_asex_merged_int_no_actv_harmony <- ScaleData(all_asex_merged_int_no_actv_harmony, verbose = F)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv_harmony <- RunPCA(all_asex_merged_int_no_actv_harmony, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "pca", group.by = "stage")


```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int_no_actv_harmony <- IntegrateLayers(
  object = all_asex_merged_int_no_actv_harmony,
  method = HarmonyIntegration,
  orig.reduction = "pca", 
  new.reduction = "harmony",
  # normalization.method = "SCT",
  verbose = F,
  # k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, reduction = "harmony")
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv_harmony <- RunUMAP(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, label = T, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
DimPlot(all_asex_merged_int_no_actv_harmony, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int_no_actv_harmony, group.by = "donor")
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "nCount_RNA")
VlnPlot(all_asex_merged_int_no_actv_harmony, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "SCT"
```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
all_asex_merged_int_no_actv_harmony <- JoinLayers(all_asex_merged_int_no_actv_harmony)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_cln <- FindAllMarkers(all_asex_merged_int_no_actv_harmony)
de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```

```{r, warning=FALSE, message=FALSE}
gogo()
```


######################## END integrate cellranger filtered with rescued asexuals ########################

########END TRY SCT ########


```{r, warning=FALSE, message=FALSE}
## mget

imap(all_msc_ed, ~FeaturePlot(.x, features = "PF3D7-1328800", order = T) + labs(title = .y))
imap(all_msc_ed, ~FeaturePlot(.x, features = "PF3D7-1328800", reduction = "pca", order = T) + labs(title = .y))
```




##############START - Painter EArly asexual markers##############
```{r}
painter_raw = readxl::read_xlsx("../published_dsets/painter_29985403_48hr_dynamics/Expression_Peak_Timing_48h_IDC.xlsx", sheet = "Estimated Total Abundance", .name_repair = "universal") %>% mutate(across(Gene.ID, ~str_replace(., "_", "-"))) %>% rename_all(., ~str_remove(., "\\.\\.")) %>% rename_with(.fn = ~str_remove(paste0("hpi_", .x), ".hpi"), .cols = contains("hpi")) %>% rename("peak_hpi" = `hpi_Peak.Tim.`)
```

```{r}
painter_raw 
```


```{r}

# Get the column names that start with 'hpi'
hpi_cols <- grep("^hpi_[0-9]+$", names(painter_raw), value = TRUE)

# Dynamically split the columns into groups of 5
col_groups <- split(hpi_cols, ceiling(seq_along(hpi_cols) / 10))

# Compute the row-wise means for each group of 5 columns
means_list <- lapply(col_groups, function(cols) {
  rowMeans(painter_raw[, cols], na.rm = TRUE)
})

# Combine the results into a single data frame
means_df <- data.frame(means_list)
names(means_df)<- paste0("mean_", seq_along(means_df))

# Combine the original dataframe with the means
painter_raw_summary <- cbind(painter_raw[, c("Gene.ID", "Annotation")], means_df)

```


```{r}
painter_raw_summary  %>% arrange(desc(mean_1))
```

```{r}
painter_raw %>% filter(Gene.ID %in% c("PF3D7-1102800", "PF3D7-0202500", "PF3D7-1370300")) %>% 
  pivot_longer(cols = starts_with("hpi_"), names_to = "time", values_to = "expresion") %>% 
  mutate(across(time, ~as.numeric(str_extract(., "[0-9]+")))) %>% 
  ggplot(aes(x = time, y = expresion, color = Gene.ID)) + 
  geom_point()
```


```{r, warning=FALSE, message=FALSE}
## mget
imap(all_msc_ed, ~FeaturePlot(.x, features = c("PF3D7-0501300", "PF3D7-1102800"), order = T) + labs(title = .y))
imap(all_msc_ed, ~FeaturePlot(.x, features = c("PF3D7-0501300", "PF3D7-1102800"), reduction = "pca", order = T) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
## mget
FeaturePlot(all_msc_100_700, features = c("PF3D7-1370300", "PF3D7-1102800"), order = T, pt.size = 0.5)
FeaturePlot(all_msc_100_700, features = c("PF3D7-1370300", "PF3D7-1102800"), reduction = "pca", order = T, pt.size = 0.5) 
```

##############END - Painter EArly asexual markers##############

## scmap annotation

### Preprocessing references

#### msc_combi reference dataset
```{r, message=FALSE, warning=FALSE, eval = T}
## v3labmscs.ref.sce.stgasx.RDS generated in m2_QC_prelim_stage_anno 
v3labmscs.ref.sce.stgasx <- readRDS("data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS")
```


```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_edrops100_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd_edrops100.RDS" --stg_labl "stageHL_ref"

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```
SCT for raw
```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_edrops100_sct.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd_edrops100.RDS" --stg_labl "stageHL_ref"

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_sct.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```

```{r, message=F, warning=F}
##Read the annotations
sample_name_ed <- sample_name_all[!str_detect(sample_name_all, "_150")]
sample_name_ed <- sample_name_all
```

```{r, message=F, warning=F}
##Read the annotations
m2_raw_anots <- list.files("data/processed/Pf", pattern = "*prelim_anotd.*RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|prelim_anotd|.RDS'))) %>% 
  # .[sample_name_ed] %>%
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(readRDS)
```

```{r}
## Adding annotations to query cells
m2_raw_anotd_preQC <- list(m2_raw_anots, all_msc_ed[sample_name_ed]) %>%
  pmap(~ ..1 %>%
         column_to_rownames("cell_bc") %>%
         mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
         # mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
         mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_replace_all(.x, c("gametocyte" = "Gametocyte", "late" = "Late", "early" = "Early")))) %>%
         mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
         AddMetaData(..2, .) 
  )

```

```{r}
## Adding annotations to query cells
m2_raw_anotd_preQC <- list(m2_raw_anotd_preQC) %>%
  pmap(~ ..1@meta.data %>%
         mutate(gene100 = ifelse(nFeature_RNA > 100, "yes", "no"),
                trancript100 = ifelse(nCount_RNA > 100, "yes", "no")) %>%
         AddMetaData(..1, .) 
  )

```

```{r}
# ## Adding annotations to query cells
# m2_raw_anotd_preQC <- list(m21_gn150_anots, all_msc_ed) %>%
#   pmap(~ ..1 %>%
#          column_to_rownames("cell_bc") %>%
#          mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
#          mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
#          mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
#          AddMetaData(..2, .) 
#   )

```

```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'Stage_MSC', cols = sp_fld_colors) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'stage_lab1', cols = sp_fld_colors2)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'gene100')+ labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(Stage_MSC, stage_lab1))

```


```{r}
##Modify the stage labels for MSC14
# m2_raw_anotd_preQC[[4]] <- m2_raw_anotd_preQC[[4]]@meta.data %>%
m2_raw_anotd_preQC <- pmap(list(m2_raw_anotd_preQC, rep(0.4,length(sample_name_all)), c(rep('ring|trophozoite|developing', length(sample_name_all)))), ~{
  ..1@meta.data %>%
    mutate(across(Stage_MSC, as.character)) %>%
    # mutate(Stage_MSC = Stage_MSC) %>%
    mutate(StagePrelim_lf = case_when(
      is.na(Stage_MSC) & sim_lab1 > 0.35 & str_detect(stage_lab1, ..3) & str_detect(stage_lab2, ..3) & str_detect(stage_lab3, ..3) ~ stage_lab1, ## lower cos sim of 0.35 for asexuals instead of 0.4
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Male|male') & str_detect(stage_lab2, 'Male|male') & str_detect(stage_lab3, 'Male|male') ~ stage_lab1,
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Female|female') & str_detect(stage_lab2, 'Female|female') & str_detect(stage_lab3, 'Female|female') ~ stage_lab1,
      TRUE ~ Stage_MSC)#,
      # across(StagePrelim, str_to_sentence)
    )%>%
    mutate(StagePrelim = case_when(
      str_detect(StagePrelim_lf, "developing") ~ "Early trophozoite", ## !!NB - label developing gams as early trophs the closest stage
      str_detect(StagePrelim_lf, "schizont") & str_detect(stage_lab2, "ring|trophozoite")  & str_detect(stage_lab3, "ring|trophozoite") ~ str_to_sentence(stage_lab2),
      str_detect(StagePrelim_lf, "late|early") ~ str_to_sentence(StagePrelim_lf),
      str_detect(StagePrelim_lf, "gametocyte") ~ str_to_sentence(str_remove_all(StagePrelim_lf, "gametocyte \\(|\\)")),
      is.na(StagePrelim_lf) ~ "Unassigned",
      TRUE ~ StagePrelim_lf)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(StagePrelimC = case_when(
      str_detect(StagePrelim, "ring|trophozoite|developing|schizont") ~ "Asexual",
      TRUE ~ StagePrelim)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(across(c(StagePrelim, StagePrelimC), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    dplyr::select(StagePrelim, StagePrelim_lf, StagePrelimC) %>%
    # mutate(across(starts_with('Stage_V'), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    AddMetaData(.x, .)
})

```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~.x@meta.data %>% dplyr::count(Stage_MSC, StagePrelim, stage_lab1))
# map(m2_raw_anotd_preQC, ~.x@meta.data %>% dplyr::count(StagePrelim, StagePrelimC))
```

```{r, warning=FALSE, message=FALSE}
imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = F, group.by = 'StagePrelim', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}
# m21_gn150_anots[["MSC51"]]
```


```{r, warning=FALSE, message=FALSE}
imap(m2_raw_anotd_preQC, ~FeaturePlot(.x, dims = c(1,2), features = "nFeature_RNA")+ labs(title = .y))
```

```{r, results='asis',  fig.width = 8, fig.height = 4}
##adding clusters that are used to display mitochondrial count

plts <- list(m2_raw_anotd_preQC, names(m2_raw_anotd_preQC)) %>%
  pmap(., ~{
    
    cell_qc_plot <-  ..1 %>% DimPlot(., pt.size = 1, reduction = 'umap', dims = c(1,2), group.by = 'StagePrelim', cols = sp_fld_colors) +labs(title = ..2)#,
    
    cell_qc_plot$data <- cell_qc_plot$data %>% rownames_to_column('bcode') %>% left_join(., ..1@meta.data[,'seurat_clusters', drop = F] %>% rownames_to_column('bcode'), by='bcode') %>% column_to_rownames('bcode')
    
    # print(str(cell_qc_plot$data))
    
    cell_qc_plot<- LabelClusters(cell_qc_plot, id = "seurat_clusters", size = 9, repel = T)
    
    return(cell_qc_plot)

  })
```


```{r, results='asis',  fig.width = 8, fig.height = 4}
plts

```

compare filt and raw with gn count cut off of 150
```{r, message=F, warning=F}
# go()
##Read the annotations
m21_all_anots_mdata <- map2(m2_raw_anotd_preQC[1:3], m2_raw_anotd_preQC[4:6], ~{
  .x@meta.data %>% 
    rownames_to_column("bcode") %>%
    full_join(., .y@meta.data %>% rownames_to_column("bcode"), suffix = c("_filt", "_gn_150"), by = "bcode") %>%
    mutate(dset = case_when(!is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "both",
                            !is.na(nFeature_RNA_filt) & is.na(nFeature_RNA_gn_150)  ~ "filt",
                            is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "gn_150"))
})
```


```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% count(dset))
map(m21_all_anots_mdata, ~.x %>% count(dset, stage_lab1_gn_150))

```

```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% filter(dset == "gn_150") %>% count(StagePrelim_gn_150, Stage_MSC_gn_150, stage_lab1_gn_150, stage_lab2_gn_150, stage_lab3_gn_150))

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = sim_lab1_gn_150) ) +
      geom_boxplot() +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = nFeature_RNA_gn_150) ) +
      geom_boxplot() +
      geom_hline(yintercept = 100) +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```

Umap with stage labels in legend

```{r, fig.width = 7, fig.height = 5, eval=F}
trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.2, "cm")
)

# Assign the female High_low stage of the projected cell to the nearest cell in the reference dataset
# msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(msc_combi.sce))), length(m21_gn150_jc))
msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(v3labmscs.ref.sce.stgasx))), length(m21_gn150_jc))


for (i in 1:length(msc_qcd_anotd_mali_mapped14HL)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m2_raw_anotd_preQC[[i]]$StagePrelim
}


for (i in 1:length(m2_raw_anotd_preQC)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m2_raw_anotd_preQC[[i]]$StagePrelim
  # msc_combi@meta.data[, paste0("stage_", names(m2_raw_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  v3labmscs.ref.sce.stgasx@colData[, paste0("stage_", names(m2_raw_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  
}

labs_order <- c('Late ring', 'Early trophozoite','Female (HE)',  'Male (HE)', 'Female (LE)', 'Male (LE)','Gametocyte (developing)', 'Gametocyte (male)','Gametocyte (female)',  'Atlas')
names(labs_order)<- labs_order

umap_theme <- theme(legend.position = 'right',
                    axis.title=element_text(size=12,face="bold",hjust = 0),
                    axis.text=element_text(size=18), 
                    plot.title = element_blank(),
                    legend.title=element_text(size=14,face="bold"),
                    legend.text=element_text(size=14),
                    legend.spacing.y = unit(0.25, "cm"),
                    axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed"))
)  


# Idents(msc_combi) <- 'MSC14_V3'
# m14_scmap_mca_umap <- map(names(m2_raw_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), 
#         order = labs_order) +
m14_scmap_mca_umap <- map(names(m2_raw_anotd_preQC), ~{
  cbind(reducedDims(v3labmscs.ref.sce.stgasx)$PCA, colData(v3labmscs.ref.sce.stgasx)[, c("stageHL_ref", paste0("stage_",.x)), drop=F]) %>% 
    data.frame() %>% 
    mutate(od_stg = droplevels(factor(!!rlang::sym(paste0("stage_",.x)), levels = rev(stg_refnd_lvls)))) %>%
    arrange(od_stg) %>%
    ggplot(aes(y = PC2, x=PC3, color = od_stg)) +
    geom_point() +
    # scale_color_manual(breaks=labs_order, values = sp_fld_colors, labels = str_wrap(labs_order, 10))+
    scale_color_manual(values = sp_fld_colors)+
    umap_theme+
    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
           x= trunc_axis,
           y= trunc_axis)+
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(breaks = NULL)
})

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap, 'plots/fig4/m14_scmap_mca_umap')

m14_scmap_mca_umap


```

Umap with stage labels inside
```{r, fig.width = 7, fig.height = 5, eval=F}

m14_scmap_mca_umap_labs_in <- map(names(m2_raw_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), label = T, order = labs_order, repel = T, label.size = 5) +
                                    scale_color_manual(values = sp_fld_colors)+ 
                                    umap_theme+
                                    theme(axis.title=element_text(size=18,face="bold"), 
                                          plot.title = element_blank(),
                                          legend.position='None'
                                    ) +
                                    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
                                           x= trunc_axis,
                                           y= trunc_axis)+
                                    scale_x_continuous(breaks = NULL) +
                                    scale_y_continuous(breaks = NULL)
)

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap_labs_in, 'plots/fig4/m14_scmap_mca_umap_labs_in.RDS')


m14_scmap_mca_umap_labs_in


```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim, Stage_MSC, stage_lab1, stage_lab2))

```


```{r, warning=FALSE, message=FALSE}
saveRDS(m2_raw_anotd_preQC, "data/processed/Pf/m2_raw_anotd_preQC.RDS")
```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data) %>% saveRDS(., "data/processed/Pf/m2_raw_anotd_preQC_mdata.RDS")

```

```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim))

```




```{r, fig.width = 12, fig.height = 4}

##Plotting top two scmap stages among the top 10 with their similarities for MSC14
# 
# stage_fct_labs <- msc_qcd_anotd[[1]]@meta.data  %>% dplyr::group_by(stageHL, seurat_clusters) %>% dplyr::group_keys() %>% dplyr::mutate(across(c(stageHL, seurat_clusters), as.character), across(seurat_clusters, ~paste0("Cluster ",., ':\n',stageHL))) %>% deframe()
# 
# scm_barpt <- msc_qcd_anotd[[1]]@meta.data %>% 
#   dplyr::select(matches('s.*lab.'), stageHL) %>% 
#   rownames_to_column('cell_bc')%>%
#   pivot_longer(-c('cell_bc', 'stageHL'),
#                names_to = c( ".value", "Label"),
#                names_pattern = "(s.*)_(l.*)")%>% 
#   filter(!is.na(sim)) %>%
#   mutate(across(sim, as.numeric)) %>% 
#   mutate(across(Label, ~str_replace(., 'lab','Label '))) %>%
#   mutate(across(stage, ~droplevels(factor(., stg_refnd_lvls)))) %>%
#   ggplot(., aes(y = sim, x = stage, color = Label)) +
#   geom_boxplot(aes(fill = Label))+
#   scale_fill_manual(values =  c('Label 1' = '#007FFF', 'Label 2' = '#FFFF9F'), name = "Label",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 2)
#   )+
#   scale_color_manual(values =  c('Label 1' = 'black', 'Label 2' = 'black'), name = "Label")+ 
#   geom_point(aes(group=Label), size = 0.5, position = position_jitterdodge(jitter.width = 0.2))+
#   new_scale_fill()+
#   geom_tile(data = data.frame(x = c( 'Late ring', 'Early trophozoite', 'Gametocyte (developing)', 'Gametocyte (female)', 'Gametocyte (male)')), inherit.aes = F, aes(x = x, y = 0.37, fill = x, height = 0.03)) +
#   scale_fill_manual(values =  sp_fld_colors, name = "MCA stage",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 3)
#                     # guide = 'none'
#   )+
#   scale_y_continuous(expand = expansion(mult = c(0.01, 0.05))) +
#   facet_grid(.~stageHL, labeller = labeller(stageHL = stage_fct_labs)) +
#   geom_hline(yintercept = 0.4)+
#   labs(y = 'scmap \nsimilarity scores', x = 'MCA stage') +
#   theme_classic() +
#   theme(axis.text.x = element_blank(),
#         axis.title=element_text(size=16,face="plain"), 
#         axis.text = element_text(size=15), 
#         axis.ticks.x = element_blank(),
#         strip.text =element_text(size=16,face="plain", colour = 'black'),
#         panel.border=element_rect(colour="black",fill='transparent'),
#         legend.title = element_text(size=16,face="bold"),
#         legend.text = element_text(size=15),
#         legend.position = 'bottom')
# 
# scm_barpt
```


Run souporcell like below for the filtered barcodes of all infections
```{r}
gogo()

# nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/MSC*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```



Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
gogo()

#  nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373{63,72,89}/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/Pf_mxd/MSC*/outs/sk21_bcodes_pp/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```
