---
title: "Exploration of LE genes using correlation and core genes"
date: "2024-01-06"
author: 'Jesse Rop'
output: 
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---
```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(ggpubr)
  library(tidyr)
  library(cowplot)
  library(SeuratData)
  library(SeuratDisk)
  library(pheatmap)
  library(plotly)
  library(rmarkdown)
  library(RColorBrewer)
  library(MAST)
  library(EnhancedVolcano)
  # library(org.Pf.plasmo.db)
  library(clusterProfiler) ##github installation newest version https://github.com/YuLab-SMU/clusterProfiler/issues/434#issuecomment-1048633162 remotes::install_github(c("YuLab-SMU/DOSE", "YuLab-SMU/clusterProfiler", "YuLab-SMU/enrichplot" ))
  # library(Platypus)
  library(ggvenn)
  library(tradeSeq)
  library(MetBrewer)
  library(ComplexHeatmap)
  library(fgsea)
  # library(msigdbr)
  library(slingshot)
  library(enrichplot)
  # library(chromoMap)
  library(pdftools)
  library(SCpubr)
  library(GGally)
  library(circlize)
  library(readxl)
  library("ggVennDiagram")
  library(UpSetR)
  library(ComplexUpset)
})
```

# DE within MSC14 field parasites and DE on MSC14 VS MCA V3 parasites

## Variable declarations, read in input files and exploratory visualizations

```{r setup}
##Set working directory
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali1_thesis')
```

```{r}
##Assign colour variables for the different plots

sp_colors <- c('Early ring'='#D1EC9F',
                     'Late ring'='#78C679',
                     'Early trophozoite'='#FEEEAA',
                     'Late trophozoite'='#FEB24C',
                     'Early schizont'='#C9E8F1',
                     'Late schizont'='#85B1D3',
                     'Gametocyte (developing)'='thistle',
                     'Gametocyte (female)'='#551A8B',
                     'Gametocyte (male)'='mediumpurple',
               'Unassigned'='red', 
               'Not assigned'='#D6CFC7',
               'Female (LE)'='#208cf7',
               # 'Female (HE)'='#551A8B',
               'Female'='#551A8B',
               'Male (LE)'='#a894d1',
               'Male'='mediumpurple',
               # 'Male (HE)'='mediumpurple',
               'Failed QC' = 'grey',
               'Gametocyte (committed)' = '#C399A2', 
               'Gametocyte (developing)' = '#66444C', 
               'Gametocyte (branching)' = '#66444C', 
               'Gametocyte (early female)' = '#749E89', 
               'Gametocyte (early male)' = '#7D87B2', 
               'Gametocyte (late female)' = '#4E6D58', 
               'Gametocyte (late male)' = '#41507B'
)

sp_fld_colors <- c('Late ring'= '#78C679',#met.brewer('Java')[2],
                   'Early trophozoite'='#FEEEAA',#met.brewer('Java')[3],
                   'Female (LE)'= met.brewer('Ingres')[5],
                   # 'Female (HE)'= '#551A8B', #met.brewer('Ingres')[7],
                   'Female'= '#551A8B', #met.brewer('Ingres')[7],
                   # 'Male (LE)'=met.brewer('Nizami')[6],
                   # 'Male (HE)'=met.brewer('Nizami')[8],
                   # 'Gametocyte (male)'=met.brewer('Austria')[5],
                   'Male (LE)'=met.brewer('Ingres')[7],#'#a2e19e',
                   'Male'= 'mediumpurple',#'#039911',
                   ## 'Male (HE)'= 'mediumpurple',#'#039911',
                   'Gametocyte (male)'='#3DED97',
                   'Gametocyte (female)'=met.brewer('Ingres')[6],
                   'Lab' = '#D3D3D3',
                   'Unassigned'='#DCDCDC',
                   'Failed QC' = 'grey',
                   'Not assigned' = 'grey',
                   'Early ring'='#D1EC9F',
                     'Late trophozoite'='#FEB24C',
                     'Early schizont'='#C9E8F1',
                     'Late schizont'='#85B1D3',
                   'Gametocyte (developing)'='thistle'
)

donor_cols = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))

# strain_dset_lvls = c( "G1", "G2", "G3", "Lab")
# strain_dset_lvls = c( 'SC1', 'SC2','SC3',  'SC4', 'SC5',  'SC6', 'S7', 'Lab' )
strain_dset_lvls = c( 'SC1', 'SC2','SC3',  'SC4', 'SC5',  'SC6', 'SC7', 'SC8', 'Lab', 'Doublet')

```


```{r}
##Copy souporcell files for the optimum clusters

##Sample names
# sample_name_nms <- c("msc33", "msc48","msc49", "msc55", "msc60", "msc66") %>% sort()
sample_name_nms <- c("MSC1", "MSC3","MSC13") %>% str_sort(numeric = T)
sample_name_nms_all <- c("MSC1", "MSC3","MSC13","MSC14") %>% str_sort(numeric = T)

##Sample irods IDs
# source_irods_nms <- list("5736STDY13782211","5736STDY13782212","5736STDY13782213","5736STDY13782214","5736STDY13782215", "5736STDY13437393") %>% set_names(sample_name_nms)

source_irods_nms <- list("5736STDY11771535","5736STDY11771536","5736STDY11771544") %>% set_names(sample_name_nms)


# sample_ids <- paste0(sample_name_nms, '_', source_irods_nms) %>% set_names(sample_name_nms)
sample_ids <- paste0(sample_name_nms, '_', source_irods_nms) %>% set_names(sample_name_nms)

trip = 'Mali'


```



```{r, message=FALSE}
##Metadata from all p.falciparum genes downloaded from plasmoDB
Pf3D7_genes_plasmoDB <- read_csv("../../Pf3D7_genomes_gtfs_indexs/Pf3d7_Genes_Summary_270922.csv") %>% 
  dplyr::rename("gene_id" = `Gene ID` ) %>% 
  group_by(gene_id) %>% 
  add_count(name='gn_iso') %>% 
  ungroup() %>%
  mutate(across(gene_id, ~case_when(gn_iso>1 ~ source_id,
                                    TRUE ~ as.character(.)))) %>% 
  mutate(across(gene_id, ~str_replace(.,"_", "-")))%>%
  dplyr::rename('Gene_name' = `Gene Name or Symbol`, 'Ka_Ks' = `NonSyn/Syn SNP Ratio All Strains`, 'TM_domains'=`# TM Domains`) %>% 
  mutate(across(where(is.character), ~na_if(., "N/A")),
         Gene = coalesce(`Gene_name`, gene_id)
  )  %>%
  group_by(Gene) %>% 
  add_count() %>% 
  ungroup() %>%
  mutate(across(Gene, ~case_when(n>1 ~ paste0(.,'_', gene_id),
                                 TRUE ~ as.character(.)))) 

```


```{r}
##Field datasets (MSC1,3,13,14,1272C)

msc_qcd_anotd <- readRDS("data/processed/DB52e_star/msc_qcd_anotd.RDS") 

```


### Exploratory Visualization
#### Field parasites Umap split by strain {.tabset}
```{r, results='asis', message=FALSE, warning=FALSE, fig.height=4, fig.width=18}
# cat('## Visualize field parasites {.tabset}   \n')
# msc_qcd_anotd[1:4]
# msc_qcd_anotd[1:4]  %>% 
#   purrr::iwalk(.,~{
#     # create tabset for each group 
#     cat('##### ',.y,'   \n')
#     p <- DimPlot(.x, reduction = "umap",  split.by = 'strain_qcd', group.by = 'stageHL', 
#                  cols = sp_fld_colors[.@meta.data %>% pull(stageHL) %>% unique()], pt.size = 1)  + 
#       theme(axis.text=element_text(size=18), 
#             axis.title=element_text(size=20,face="bold"), 
#             # legend.text = element_blank(), 
#             plot.title = element_blank(),
#             strip.text.x = element_text(size = 24, face = "bold") )
#     print(p)
#     cat('\n  ')
#     cat("\n")
#   })

```


#### Field parasites gene counts {.tabset}
```{r, results='asis', message=FALSE, warning=FALSE}
# cat('## Visualize field parasites {.tabset}   \n')
# msc_qcd_anotd[1:4]  %>%
msc_qcd_anotd[1:4]  %>% 
  purrr::iwalk(.,~{
    # create tabset for each group 
    cat('##### ',.y,'   \n')
    p <- VlnPlot(., features = c("nFeature_RNA"), group.by = 'stageHL', ncol = 5, cols = sp_fld_colors[.@meta.data %>% pull(stageHL) %>% unique()]) + 
      theme(legend.position="none", axis.text.x = element_text(angle = 90)) +
      xlab('stageHL')
    print(p)
    cat('\n \n \n ')
    cat("\n\n\n")
  })

```


## V3 and MSC 14 gene absence or absence

#### tabulate lab stages ovlp
```{r}
##read MCA
EF_V3_ref <- readRDS("../Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle
##female
mca_r <- rownames(EF_V3_ref)[rowSums(EF_V3_ref[,EF_V3_ref$stagenewx %in% c('early ring', 'late ring')]@assays$RNA@counts > 0) > (sum(EF_V3_ref$stagenewx %in% c('early ring', 'late ring'))*0.25)]

##male
mca_t <- rownames(EF_V3_ref)[rowSums(EF_V3_ref[,EF_V3_ref$stagenewx %in% c('early trophozoite', 'late trophozoite')]@assays$RNA@counts > 0) > (sum(EF_V3_ref$stagenewx %in% c('early trophozoite', 'late trophozoite'))*0.25)]


##asex
mca_s <- rownames(EF_V3_ref)[rowSums(EF_V3_ref[,EF_V3_ref$stagenewx %in% c('early schizont', 'late schizont')]@assays$RNA@counts > 0) > (sum(EF_V3_ref$stagenewx %in% c('early schizont', 'late schizont'))*0.25)]

##asex
mca_gd <- rownames(EF_V3_ref)[rowSums(EF_V3_ref[,EF_V3_ref$stagenewx == 'gametocyte (developing)']@assays$RNA@counts > 0) > (sum(EF_V3_ref$stagenewx == 'gametocyte (developing)')*0.25)]

##male
mca_f <- rownames(EF_V3_ref)[rowSums(EF_V3_ref[,EF_V3_ref$stagenewx == 'gametocyte (female)']@assays$RNA@counts > 0) > (sum(EF_V3_ref$stagenewx == 'gametocyte (female)')*0.25)]

mca_m <- rownames(EF_V3_ref)[rowSums(EF_V3_ref[,EF_V3_ref$stagenewx == 'gametocyte (male)']@assays$RNA@counts > 0) > (sum(EF_V3_ref$stagenewx == 'gametocyte (male)')*0.25)]

```


```{r}
mca_gns <- list('MCA_R' = mca_r,
     'MCA_T' = mca_t,
     'MCA_S' = mca_s,
     'MCA_GD' = mca_gd,
     'MCA_F' = mca_f,
    'MCA_M' = mca_m) %>% 
  purrr::map(., data.frame) %>% 
  bind_rows(., .id='stage') %>% 
  mutate(present = 'y') %>% 
  pivot_wider(., names_from = 'stage', values_from = 'present') %>% 
  rename('gene_id' =`.x..i..`) %>% 
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>% 
  dplyr::select('gene_id', starts_with('MCA'), 'Gene' )


```

```{r}
c_upst_fn <- function(tbl, c_names) {
  ComplexUpset::upset(tbl, 
                      c_names, 
                          height_ratio=0.8, 
                          width_ratio=0.3, 
                          set_sizes=FALSE,
                          base_annotations=list(
                            'Intersection size'=intersection_size(
                              text=list(
                                size = 4.8,
                                vjust=0.5,
                                hjust=-0.1,
                                angle=90
                              ),
                              bar_number_threshold = 1 
                            ) + 
                              labs(y= 'Gene count')+
                              scale_y_continuous(expand = expansion(mult = c(0, 0.45)))),
                          matrix=(
                            intersection_matrix()), 
                          name= 'Stage genes sets',
                          themes=upset_modify_themes(
                            list(
                              'intersections_matrix'=theme(axis.text.y=element_text(size=15),
                                                           axis.title.x=element_text(size=16, face = 'plain')),
                              'Intersection size' = theme(axis.text.y=element_text(size=15), 
                                                          axis.title.y=element_text(size=16, face = 'plain'))
                            )
                          )
      )
}
```


UpSet plot
```{r, fig.width= 12, fig.height=4.4}
# list(mast_tde_corr_tbls[6:10]) %>%
upst_mca_pt <- mca_gns %>% 
      dplyr::select(contains('MCA_'))%>% 
      mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
      c_upst_fn(., colnames(.)) 

upst_mca_pt

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(upst_mca_pt, "plots/figS26/upst_mca_pt.RDS")
```


#### tabulate female high vs female low ovlp
```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle
##female


# msc_fhe_p <- map(msc_qcd_anotd[1:4], ~{
msc_f <- map(msc_qcd_anotd[1:4], ~{
rownames(.x)[rowSums(.x[,.x$StageMCA == 'Gametocyte (female)']@assays$RNA@counts > 0) > (sum(.x$StageMCA == 'Gametocyte (female)')*0.25)]
})

msc_m <- map(msc_qcd_anotd[1:4], ~{
rownames(.x)[rowSums(.x[,.x$StageMCA == 'Gametocyte (male)']@assays$RNA@counts > 0) > (sum(.x$StageMCA == 'Gametocyte (male)')*0.25)]
})

msc_lr <- map(msc_qcd_anotd[c(1,4)], ~{
rownames(.x)[rowSums(.x[,.x$StageMCA == 'Late ring']@assays$RNA@counts > 0) > (sum(.x$StageMCA == 'Late ring')*0.25)]
})

msc_et <- map(msc_qcd_anotd[c(1,4)], ~{
rownames(.x)[rowSums(.x[,.x$StageMCA == 'Early trophozoite']@assays$RNA@counts > 0) > (sum(.x$StageMCA == 'Early trophozoite')*0.25)]
})

```

#### overlap between stgs
```{r}
m_f_gns_all <- map(list(msc_f, msc_m, msc_lr, msc_et), ~{ 
  stg_lst <- .x
  purrr::map(stg_lst, data.frame) %>% 
        bind_rows(., .id='donor') 
  })  %>%
  set_names(c('F_genes', 'M_genes', 'LR_genes', 'ET_genes')) %>%
  bind_rows(., .id='stage') %>%
  mutate(present = 'y') %>%
  rename('gene_id' =`.x..i..`) %>%
  # group_by(gene_id, stage) %>%
  # mutate(donor = case_when(n_distinct(donor) == 2 ~ "Both", n_distinct(donor) == 1 ~ donor)) %>%
  # ungroup() %>%
  distinct(gene_id, stage, .keep_all = T) 
  

m_f_gns_all %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>%
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id')
  
```
```{r, fig.width= 12, fig.height=4.4}
gn_up_all <- m_f_gns_all %>% pivot_wider(., names_from = 'stage', values_from = c('present','donor')) %>%
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
      dplyr::select(contains('present_')) %>% 
      mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
      c_upst_fn(., colnames(.))

gn_up_all
```

#### tabulate female high vs female low ovlp
```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle
##female


# msc_fhe_p <- map(msc_qcd_anotd[1:4], ~{
msc_fhe_p <- map(msc_qcd_anotd[3:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Female']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Female')*0.25)]
})

msc_fle_p <- map(msc_qcd_anotd[3:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Female (LE)']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Female (LE)')*0.25)]
})

msc_mhe_p <- map(msc_qcd_anotd[3:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Male']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Male')*0.25)]
})

msc_mle_p <- map(msc_qcd_anotd[3:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Male (LE)']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Male (LE)')*0.25)]
})


# Proper naming
msc_3_13_14_fhe_p <- map(msc_qcd_anotd[2:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Female']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Female')*0.25)]
})

msc_3_13_14_fle_p <- map(msc_qcd_anotd[2:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Female (LE)']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Female (LE)')*0.25)]
})

msc_13_14_mhe_p <- map(msc_qcd_anotd[3:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Male']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Male')*0.25)]
})

msc_13_14_mle_p <- map(msc_qcd_anotd[3:4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Male (LE)']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Male (LE)')*0.25)]
})

msc_14_lr_p <- map(msc_qcd_anotd[4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Late ring']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Late ring')*0.25)]
})

msc_14_et_p <- map(msc_qcd_anotd[4], ~{
rownames(.x)[rowSums(.x[,.x$stageHL == 'Early trophozoite']@assays$RNA@counts > 0) > (sum(.x$stageHL == 'Early trophozoite')*0.25)]
})

```



#### overlap between top 500 mh,ml,fh,fl above
```{r}
gams_vn<-pmap(list(msc_fhe_p, msc_fle_p,msc_mhe_p,msc_mle_p, names(msc_fhe_p)), ~{ggvenn(
  list('F' = ..1,
       'FLE' = ..2,
       'M' = ..3,
       'MLE' = ..4),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
  stroke_size = 0.5, set_name_size = 5,show_percentage = F, text_size = 5
)+ 
    labs(title = ..5)+
  theme(#plot.background = element_rect(fill = 'white', colour = 'darkgrey'),
        panel.background = element_rect(fill = 'white', colour = 'white'),
        plot.title = element_text(hjust = 0.5, size = 12, face = "bold"),
        # plot.margin=margin(b=-1.5,t=-2.5,l=-1.,r=-1,unit="cm"))
        # plot.margin=margin(b=-0.4,t=-0.4,l=0.3,r=0.3,unit="cm"),
        aspect.ratio = 0.8,
        
        )})

gams_vn
```



#### overlap between top 500 mh,ml,fh,fl above
```{r}
gams_3_13_14_vn<-pmap(list(msc_3_13_14_fhe_p, msc_3_13_14_fle_p, names(msc_3_13_14_fle_p)), ~{ggvenn(
  list('F' = ..1,
       'FLE' = ..2),
       fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
  stroke_size = 0.5, set_name_size = 5,show_percentage = F, text_size = 5.5
)+ 
    labs(title = ..3)+
  theme(#plot.background = element_rect(fill = 'white', colour = 'darkgrey'),
        panel.background = element_rect(fill = 'white', colour = 'white'),
        plot.title = element_text(hjust = 0.5, size = 12, face = "bold"),
        # plot.margin=margin(b=-1.5,t=-2.5,l=-1.,r=-1,unit="cm"))
        # plot.margin=margin(b=-0.4,t=-0.4,l=0.3,r=0.3,unit="cm"),
        aspect.ratio = 0.8,
        
        )})

gams_3_13_14_vn
```

#### overlap between top 500 mh,ml,fh,fl above
```{r}
cgns_list <- list(msc_fhe_p, msc_fle_p,msc_mhe_p,msc_mle_p) %>% 
  set_names(c("F", "FLE", "M", "MLE"))

gams_vn_d<-imap(cgns_list, ~ggvenn(.x,
       # fill_color = c("#0037C2FF", "#EFC3FF", "#2FF000", "#A48000"),
       fill_color = as.character(donor_cols),
  stroke_size = 0.5, set_name_size = 5,show_percentage = F, text_size = 6
)+ 
    labs(title = .y)+
  theme(#plot.background = element_rect(fill = 'white', colour = 'darkgrey'),
        panel.background = element_rect(fill = 'white', colour = 'white'),
        plot.title = element_text(hjust = 0.5),
        text = element_text(size = 20),
        # plot.margin=margin(b=-1.5,t=-2.5,l=-1.,r=-1,unit="cm"))
        # plot.margin=margin(b=-0.4,t=-0.4,l=0.3,r=0.3,unit="cm"),
        aspect.ratio = 0.8,
        
        ))

gams_vn_d
```

```{r}
##15 MSC1 FLE genes
saveRDS(gams_vn, "plots/figS_HEvLE/gams_vn.RDS")
```

```{r}
##15 MSC1 FLE genes
msc1_fle_uniq <- msc_fle_p[["MSC13"]][!(msc_fle_p[["MSC13"]] %in% c(msc_fhe_p[["MSC13"]], msc_mle_p[["MSC13"]], msc_mhe_p[["MSC13"]] ))]

map(msc1_fle_uniq, ~FeaturePlot(msc_qcd_anotd[[3]], features = .x, slot = "counts"))
```


```{r}
##15 MSC1 FLE genes
Pf3D7_genes_plasmoDB %>% filter(`gene_id` %in% msc1_fle_uniq) %>% select(`Product Description`, Gene, `Curated GO Functions`, `Curated GO Processes`)
```

```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle
##female
# gn_exp <- map(msc_qcd_anotd[1:4], ~AverageExpression(.x, group.by = "stageHL", assays = "RNA") %>% .[[1]] %>% as.data.frame() %>% rownames_to_column('gene')) %>% bind_rows(., .id = "donor") %>% select(-c('Gametocyte (female)', 'Gametocyte (male)', 'Gametocyte (developing)'))

gn_exp <- map(msc_qcd_anotd[1:4], ~AverageExpression(.x, group.by = "stageHL", assays = "RNA") %>% .[[1]] %>% as.data.frame() %>% rownames_to_column('gene')) %>% bind_rows(., .id = "donor")

gn_exp
```


```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle
##female

gn_exp %>% pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% glimpse 

```



```{r}
library(GGally)

lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")
}
## correlation between genes by stage
gn_exp %>% filter(donor == 'MSC1') %>% select(-c(donor)) %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Male (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Female (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female')) %>% pivot_wider(names_from = 'donor', values_from = 'Female') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)


```


```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male')) %>% pivot_wider(names_from = 'donor', values_from = 'Male') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```

```{r, warning=FALSE, message=FALSE}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male (LE)', 'Male')) %>% pivot_wider(names_from = 'donor', values_from = c('Male (LE)','Male')) %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r, warning=FALSE, message=FALSE}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female (LE)', 'Female')) %>% pivot_wider(names_from = 'donor', values_from = c('Female (LE)','Female')) %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r, warning=FALSE, message=FALSE}

cor_fun <- function(data, mapping, method="pearson", use="pairwise", ndp=2, sz=5, stars=TRUE, ...){

# grab data
x <- eval_data_col(data, mapping$x)
y <- eval_data_col(data, mapping$y)

# calculate correlation: for significance stars
corr <- cor.test(x, y, method=method)
est <- corr$estimate
lb.size <- sz* abs(est)

# get significance stars
if(stars){
  stars <- c("***", "**", "*", "")[findInterval(corr$p.value, c(0, 0.001, 0.01, 0.05, 1))]
  lbl <- paste0(round(est, ndp), stars)
}else{
  lbl <- round(est, ndp)
}
# calculate correlation: for colored tiles
corr <- cor(x, y, method=method, use=use)

# calculate color based on correlation value
# corr = -1 => blue, 
# corr =  0 => white, 
# corr = +1 => red, 
colFn <- colorRampPalette(c("blue","white", "red"), interpolate ='spline')
fill <- colFn(100)[findInterval(corr, seq(0, 1, length=100))]

ggplot(data = data, mapping = mapping, ...) + 
  theme_void() +
  annotate("text",
           x=mean(x, na.rm=TRUE),
           y=mean(y, na.rm=TRUE),
           label=lbl,
           # size=lb.size,
           size=4,
           ...) +
  theme(panel.background = element_rect(fill=fill,  # to fill background of panel with color
                                        colour=NA), # to remove border of panel
        panel.grid.major = element_blank())
}


## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Early trophozoite', 'Late ring', 'Female (LE)', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = c('Female (LE)', 'Male (LE)', 'Early trophozoite', 'Late ring')) %>% 
  select(where(~!all(is.na(.x)))) %>% 
  column_to_rownames('gene') %>% 
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          # upper = list(continuous = wrap("cor", size = 5)))
          upper = list(continuous = cor_fun),
          progress = FALSE)

```



```{r}

## correlation between genes by stage

base_thm<-theme(axis.title=element_text(size=16,face="plain"), 
        axis.text = element_text(size=15), 
        legend.title = element_text(size=16,face="bold"),
        legend.text = element_text(size=15))
 
scattr_thm <- base_thm +
   theme(#aspect.ratio = 1, 
          legend.position = "none",
         plot.title = element_text(size=16, hjust = .5, face = "bold"),
        )

gam_corrs <- map(c("Female", "Female (LE)", "Male", "Male (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC13, y= MSC14, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  ggpubr::stat_cor(method = "pearson", label.x =150, label.y = 10)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(title = .x)+
    theme_classic()+
    scattr_thm
}) 

gam_corrs
```


```{r}

## correlation between genes by stage

gam_corrs3_13 <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC13/100, y= MSC3/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC13", y= "MSC3", title = .x)+
    theme_classic()+
    scattr_thm
}) 


gam_corrs3_14 <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC3/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC3", title = .x)+
    theme_classic()+
    scattr_thm
}) 

gam_corrs14_13 <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC13/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC13", title = .x)+
    theme_classic()+
    scattr_thm
}) 


gam_corrsm14_13 <- map(c("Male", "Male (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC13/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC13", title = .x)+
    theme_classic()+
    scattr_thm
}) 

gam_corrs3_14
gam_corrs3_13
gam_corrs14_13
gam_corrsm14_13
```


```{r}
lowerfun_le_corr <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
    stat_cor(method = "pearson", label.x =150, label.y = 10)+ 
    # labs(title = .x)+
    theme_classic()+
    theme(axis.text=element_text(size=20), 
          axis.title=element_text(size=18,face="bold"), 
          legend.position = "none",
          panel.border = element_rect(colour = "black",fill = NA))
}

## correlation between genes by stage
gam_corrsf_ga <- map(c("Female", "Female (E)"), ~{
  gn_exp %>% select(c('gene','donor', 'Female (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Female (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
GGally::ggduo(., 1:2, 3, types = list(continuous = lowerfun_le_corr)) +
  labs(title = 'Female (LE)')
})
gam_corrsf_ga

```


```{r}

gam_corrsm_ga <- map(c("Male", "Male (LE)"), ~{
  gn_exp %>% select(c('gene','donor', .x)) %>% pivot_wider(names_from = 'donor', values_from = .x) %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  # mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1, 2, types = list(continuous = lowerfun_le_corr))
}) 

gam_corrsm_ga
```

```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Male (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1, 2, types = list(continuous = lowerfun))

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male')) %>% pivot_wider(names_from = 'donor', values_from = 'Male') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1, 2, types = list(continuous = lowerfun))
```

```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Female (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1:2, 3, types = list(continuous = lowerfun))

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female')) %>% pivot_wider(names_from = 'donor', values_from = 'Female') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1:2, 3, types = list(continuous = lowerfun))
```


```{r}
##15 MSC1 FLE genes
# AverageExpression(msc_qcd_anotd[[1]], features = msc1_fle_uniq, group.by = 'stageHL')
# AverageExpression(msc_qcd_anotd[[1]], features = msc1_fle_uniq, group.by = 'stageHL', assays = "RNA", slot = "counts")
```



#### overlap between top 500 mh,ml,fh,fl above
```{r}
# low_gam_asex_vn <- ggvenn(
#   list('FL' = m14_fle_p,
#        'ML' = m14_mle_p,
#        'LR' = m14_lr_p,
#        'ET' = m14_et_p),
#        fill_color = c("#0073C2FF", "#EFC000FF", "#868686FF", "#A46868"),
#   stroke_size = 0.5, set_name_size = 5,show_percentage = F, text_size = 4.5
# ) +
#   theme(panel.background = element_rect(fill = 'white', colour = 'white'),
#         # plot.margin=margin(b=-0.4,t=-0.4,l=0.3,r=0.3,unit="cm"),
#         aspect.ratio = 0.65)
# 
# low_gam_asex_vn
```

```{r}
map(list(msc_fhe_p, msc_fle_p, msc_mhe_p, msc_mle_p, msc_lr, msc_et), ~{ 
  stg_lst <- .x
  purrr::map(stg_lst, data.frame) %>% 
        bind_rows(., .id='donor') 
  })  %>%
  set_names(c('F_genes', 'FLE_genes', 'M_genes', 'MLE_genes', 'LR_genes', 'ET_genes')) %>%
  bind_rows(., .id='stage') %>%
  mutate(present = 'y') %>%
  rename('gene_id' =`.x..i..`) %>%
  group_by(gene_id, stage) %>%
  mutate(d_n=n_distinct(donor)) %>% arrange(desc(d_n)) %>%
  mutate(donor = case_when(d_n == 2 ~ "Both", d_n == 1 ~ donor)) %>%
  ungroup() %>%
  distinct() 
  

```

#### overlap between core genes
```{r}
mf_gns_3.13.14 <- map(list(msc_3_13_14_fhe_p, msc_3_13_14_fle_p, msc_13_14_mhe_p, msc_13_14_mle_p), ~{ 
  stg_lst <- .x
  purrr::map(stg_lst, data.frame) %>% 
        bind_rows(., .id='donor') 
  })  %>%
  set_names(c('F_genes', 'FLE_genes', 'M_genes', 'MLE_genes')) %>%
  bind_rows(., .id='stage') %>%
  mutate(present = 'y') %>%
  rename('gene_id' =`.x..i..`) #%>%
  # group_by(gene_id, stage) %>%
  # mutate(donor = case_when(n_distinct(donor) == 2 ~ "Both", n_distinct(donor) == 1 ~ donor)) %>%
  # ungroup() %>%
  # distinct(stage, gene_id, .keep_all = T) 
  

mf_gns_3.13.14 %>% select(-c(donor)) %>% distinct() %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>%
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id')

```


UpSet plot

```{r, fig.width= 12, fig.height=4.4}
# list(mast_tde_corr_tbls[6:10]) %>%
upst_3_13_14_strn_de_plts <- mf_gns_3.13.14 %>% select(-c(donor)) %>% distinct() %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>%
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
      dplyr::select(contains('_genes')) %>% 
      mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
      ComplexUpset::upset(., colnames(.), 
                          height_ratio=0.8, 
                          width_ratio=0.3, 
                          set_sizes=FALSE,
                          base_annotations=list(
                            'Intersection size'=intersection_size(
                              text=list(
                                size = 4.8,
                                vjust=0.5,
                                hjust=-0.1,
                                angle=90
                              ),
                              bar_number_threshold = 1 
                            ) + 
                              labs(y= 'Gene count')+
                              scale_y_continuous(expand = expansion(mult = c(0, 0.45)))),
                          matrix=(
                            intersection_matrix()), 
                          name= 'Stage genes sets',
                          themes=upset_modify_themes(
                            list(
                              'intersections_matrix'=theme(axis.text.y=element_text(size=15),
                                                           axis.title.x=element_text(size=16, face = 'plain')),
                              'Intersection size' = theme(axis.text.y=element_text(size=15), 
                                                          axis.title.y=element_text(size=16, face = 'plain'))
                            )
                          )
      ) 

upst_3_13_14_strn_de_plts

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(upst_strn_de_plts, "plots/figS26/fld_gn_expression_specificity_upset.RDS")
```

#### overlap between top 500 mh,ml,fh,fl above
```{r}
## Include resolution on overlap for the donors
mf_gns_3.13.14 %>%  pivot_wider(., names_from = c('stage','donor'), values_from = c('present'))
```


```{r, fig.width= 12, fig.height=4.4}
# list(mast_tde_corr_tbls[6:10]) %>%
upst_3_13_14_strn_de_plts_sep <- mf_gns_3.13.14 %>%  pivot_wider(., names_from = c('stage','donor'), values_from = c('present'))  %>%
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
      dplyr::select(contains('genes')) %>% 
      mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
      ComplexUpset::upset(., colnames(.), 
                          height_ratio=0.8, 
                          width_ratio=0.3, 
                          set_sizes=FALSE,
                          base_annotations=list(
                            'Intersection size'=intersection_size(
                              text=list(
                                size = 4.8,
                                vjust=0.5,
                                hjust=-0.1,
                                angle=90
                              ),
                              bar_number_threshold = 1 
                            ) + 
                              labs(y= 'Gene count')+
                              scale_y_continuous(expand = expansion(mult = c(0, 0.45)))),
                          matrix=(
                            intersection_matrix()), 
                          name= 'Stage genes sets',
                          themes=upset_modify_themes(
                            list(
                              'intersections_matrix'=theme(axis.text.y=element_text(size=15),
                                                           axis.title.x=element_text(size=16, face = 'plain')),
                              'Intersection size' = theme(axis.text.y=element_text(size=15), 
                                                          axis.title.y=element_text(size=16, face = 'plain'))
                            )
                          )
      ) 

upst_3_13_14_strn_de_plts_sep

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(upst_strn_de_plts, "plots/figS26/fld_gn_expression_specificity_upset.RDS")
```

```{r}
mfet_gns_3.13.14 <- map(list(msc_3_13_14_fhe_p, msc_3_13_14_fle_p, msc_13_14_mhe_p, msc_13_14_mle_p, msc_14_lr_p, msc_14_et_p), ~{ 
  stg_lst <- .x
  purrr::map(stg_lst, data.frame) %>% 
        bind_rows(., .id='donor') 
  })  %>%
  set_names(c('F_genes', 'FLE_genes', 'M_genes', 'MLE_genes', 'LR_genes', 'ET_genes')) %>%
  bind_rows(., .id='stage') %>%
  mutate(present = 'y') %>%
  rename('gene_id' =`.x..i..`) #%>%
  # group_by(gene_id, stage) %>%
  # mutate(donor = case_when(n_distinct(donor) == 2 ~ "Both", n_distinct(donor) == 1 ~ donor)) %>%
  # ungroup() %>%
  # distinct(stage, gene_id, .keep_all = T) 
  

mfet_gns_3.13.14 %>% select(-c(donor)) %>% distinct() %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>%
  left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id')

```


UpSet plot

```{r, fig.width= 12, fig.height=4.4}
# list(mast_tde_corr_tbls[6:10]) %>%
upst_3_13_14_strn_de_plts_mfet <- mfet_gns_3.13.14 %>% select(-c(donor)) %>% distinct() %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>%
      dplyr::select(contains('_genes')) %>% 
      mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
      ComplexUpset::upset(., colnames(.), 
                          height_ratio=0.7, 
                          width_ratio=0.3, 
                          set_sizes=FALSE,
                          base_annotations=list(
                            'Intersection size'=intersection_size(
                              text=list(
                                size = 4.8,
                                vjust=0.5,
                                hjust=-0.1,
                                angle=90
                              ),
                              bar_number_threshold = 1 
                            ) + 
                              labs(y= 'Gene count')+
                              scale_y_continuous(expand = expansion(mult = c(0, 0.45)))),
                          matrix=(
                            intersection_matrix()), 
                          name= 'Stage genes sets',
                          themes=upset_modify_themes(
                            list(
                              'intersections_matrix'=theme(axis.text.y=element_text(size=15),
                                                           axis.title.x=element_text(size=16, face = 'plain')),
                              'Intersection size' = theme(axis.text.y=element_text(size=15), 
                                                          axis.title.y=element_text(size=16, face = 'plain'))
                            )
                          )
      ) 

upst_3_13_14_strn_de_plts_mfet

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(upst_strn_de_plts, "plots/figS26/fld_gn_expression_specificity_upset.RDS")
```


```{r, fig.width= 12, fig.height=4.4}
# list(mast_tde_corr_tbls[6:10]) %>%
upst_3_13_14_strn_de_plts_mfet_sep <- mfet_gns_3.13.14 %>%  pivot_wider(., names_from = c('stage','donor'), values_from = c('present'))  %>%
      dplyr::select(contains('_genes')) %>% 
      mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
      ComplexUpset::upset(., colnames(.), 
                          height_ratio=0.8, 
                          width_ratio=0.3, 
                          set_sizes=FALSE,
                          min_size =7,
                          base_annotations=list(
                            'Intersection size'=intersection_size(
                              text=list(
                                size = 4.8,
                                vjust=0.5,
                                hjust=-0.1,
                                angle=90
                              ),
                              bar_number_threshold = 1 
                            ) + 
                              labs(y= 'Gene count')+
                              scale_y_continuous(expand = expansion(mult = c(0, 0.45)))),
                          matrix=(
                            intersection_matrix()), 
                          name= 'Stage genes sets',
                          themes=upset_modify_themes(
                            list(
                              'intersections_matrix'=theme(axis.text.y=element_text(size=15),
                                                           axis.title.x=element_text(size=16, face = 'plain')),
                              'Intersection size' = theme(axis.text.y=element_text(size=15), 
                                                          axis.title.y=element_text(size=16, face = 'plain'))
                            )
                          )
      ) 

upst_3_13_14_strn_de_plts_mfet_sep

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(upst_strn_de_plts, "plots/figS26/fld_gn_expression_specificity_upset.RDS")
```

```{r, fig.width= 6, fig.height=4.8}
# list(mast_tde_corr_tbls[6:10]) %>%

#   
# m_f_gns %>% pivot_wider(., names_from = 'stage', values_from = c('present','donor')) %>%
#   left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
#       dplyr::select(contains('present_')) %>% 
#       mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
#     dim
#     
# m_f_gns %>% pivot_wider(., names_from = 'stage', values_from = c('donor')) %>%
#   left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
#       dplyr::select(contains('_genes')) 
```




```{r, fig.width= 12, fig.height=4.4}
# list(mast_tde_corr_tbls[6:10]) %>%
# 
# tb <- m_f_gns %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>%
#   left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>% 
#   select(c(donor, contains("_genes"))) %>%
#       mutate(across(contains("_genes"), ~ifelse(is.na(.), F, T)))
#   
# don_upst <- ComplexUpset::upset(tb, 
#                         tb %>% dplyr::select(contains('_genes')) %>% colnames(), 
#                           height_ratio=1.1, 
#                           width_ratio=0.3, 
#                           set_sizes=FALSE,
#                           base_annotations=list(
#                             'Intersection size'=intersection_size(
#                               text=list(
#                                 size = 4.8,
#                                 vjust=0.5,
#                                 hjust=-0.1,
#                                 angle=90
#                               ),
#                               bar_number_threshold = 1,
#                               # counts=FALSE,
#                               mapping=aes(fill=donor) 
#                             ) + 
#                               labs(y= 'DE genes\nintersection')+
#                               scale_y_continuous(expand = expansion(mult = c(0, 0.38)))+
#                               # scale_fill_manual(name = "Lab & Field", values = c('No' = '#d2a7d2', 'Yes'= '#40e0d0'))+
#                               theme(legend.position = 'top',
#                                     legend.margin = margin(b=-9),
#                                     legend.text = element_text(size = 13),
#                                     legend.title = element_text(size = 14))),
#                           matrix=(
#                             intersection_matrix()), 
#                           name= 'Pairwise strain comparison',
#                           themes=upset_modify_themes(
#                             list(
#                               'intersections_matrix'=theme(axis.text.y=element_text(size=14),
#                                                            axis.title.x=element_text(size=15)),
#                               'Intersection size' = theme(axis.text.y=element_text(size=14), 
#                                                           axis.title.y=element_text(size=15, face = 'plain'))
#                             )
#                           )
#       ) 
# don_upst    

```

UpSet plot

```{r, fig.width= 12, fig.height=4.4}
# 
# m_f_gns %>% filter(donor == "Both") %>% pivot_wider(., names_from = 'stage', values_from = c('present','donor')) %>%
#   left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
#       dplyr::select(contains('present_')) %>% 
#       mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
#       c_upst_fn(., colnames(.))
# 
# m_f_gns %>% filter(donor == "msc14") %>% pivot_wider(., names_from = 'stage', values_from = c('present','donor')) %>%
#   left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
#       dplyr::select(contains('present_')) %>% 
#       mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
#       c_upst_fn(., colnames(.))
# 
# m_f_gns %>% filter(donor == "msc13") %>% pivot_wider(., names_from = 'stage', values_from = c('present','donor')) %>%
#   left_join(., Pf3D7_genes_plasmoDB, by = 'gene_id') %>%
#       dplyr::select(contains('present_')) %>% 
#       mutate(across(everything(), ~ifelse(is.na(.), F, T))) %>% 
#       c_upst_fn(., colnames(.))

```

```{r, fig.width= 12, fig.height=16}
##Put together the plots for annotation
# stg_lgnd <- cowplot::get_legend(
#   scm_barpt_s[[4]] +
#     guides(fill=guide_legend(title.position = "top", nrow = 2)))

plot_grid(plot_grid(plot_grid(plotlist = gams_vn, nrow = 2),
          plot_grid(plotlist = gam_corrs, ncol = 2, nrow = 2), ncol = 2),
          upst_3_13_14_strn_de_plts_mfet_sep,
          # plot_grid(plotlist = map(scm_barptHL[1:2], ~.+ theme(legend.position = "none")), ncol = 2, rel_widths = c(.6, .4)),
          # plot_grid(plotlist = map(scm_barptHL[3:4], ~.+ theme(legend.position = "none")), ncol = 2, rel_widths = c(.4, .6)),
          # stg_lgnd,
          nrow = 3,
            # rel_heights = c(.2,.175,.225,.175,.175,.05),
            # labels = c('A', 'B', 'C', 'D', 'E'),
            label_size = 24,
            vjust= 0.3)
```


Put together all the necessary plots
```{r, fig.width= 12, fig.height=13}
##Put together the plots for annotation
# stg_lgnd <- cowplot::get_legend(
#   scm_barpt_s[[4]] +
#     guides(fill=guide_legend(title.position = "top", nrow = 2)))

gam_hl_vn <- map(c(gams_3_13_14_vn[1], gams_vn), ~.x + theme(text = element_text(size = 14), aspect.ratio=2/4))

plot_grid(plot_grid(plotlist = map(c(gams_3_13_14_vn[1], gams_vn), ~.x + theme(text = element_text(size = 14), aspect.ratio=2/4)), nrow = 1, rel_widths = c(.3,.35,.35)),
          plot_grid(plot_grid(plotlist = c(gam_corrs3_13, gam_corrs3_14), ncol = 4), 
                              plot_grid(plotlist = c(gam_corrs14_13, gam_corrsm14_13), ncol = 4),
                              nrow = 2),
          upst_3_13_14_strn_de_plts_mfet,
          # plot_grid(plotlist = map(scm_barptHL[1:2], ~.+ theme(legend.position = "none")), ncol = 2, rel_widths = c(.6, .4)),
          # plot_grid(plotlist = map(scm_barptHL[3:4], ~.+ theme(legend.position = "none")), ncol = 2, rel_widths = c(.4, .6)),
          # stg_lgnd,
          nrow = 3,
            rel_heights = c(.2,.5,.3),
            # labels = c('A', 'B', 'C', 'D', 'E'),
            label_size = 24,
            vjust= 0.3)
```


```{r, fig.width=12, fig.height=5}
gam_hl_vn <- map(c(gams_3_13_14_vn[1], gams_vn), ~.x + theme(aspect.ratio=2.7/4, plot.title = element_text(size = 14, margin = margin(0,0,-20,0))))

fld_gn_expression_specificity_upset_vn_don3 <- ggdraw(ggdraw(upst_3_13_14_strn_de_plts_mfet)  +
      draw_plot(ggdraw(gam_hl_vn[[1]]+ theme(aspect.ratio=3/4.5)), x=-0.11, y=.625, width =1.04, height =.45) +
      draw_plot(ggdraw(gam_hl_vn[[2]]), x=.13, y=.575, width =.99, height =.5)+
      draw_plot(ggdraw(gam_hl_vn[[3]]), x=.37, y=.575, width =.97, height =.5))


##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
saveRDS(fld_gn_expression_specificity_upset_vn_don3, "plots/figS26/fld_gn_expression_specificity_upset_vn_don3.RDS")

fld_gn_expression_specificity_upset_vn_don3
```



```{r, fig.width=12, fig.height=5, eval=FALSE}
fld_gn_expression_specificity_upset_vn <- ggdraw(ggdraw(upst_strn_de_plts)  +
      draw_plot(ggdraw(gams_vn), x=.05, y=.625, width =.99, height =.45) +
      draw_plot(ggdraw(low_gam_asex_vn), x=.36, y=.625, width =.99, height =.45))


##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
saveRDS(fld_gn_expression_specificity_upset_vn, "plots/figS26/fld_gn_expression_specificity_upset_vn.RDS")

fld_gn_expression_specificity_upset_vn
```

thesis
```{r}
corr_theme <- theme(axis.title=element_text(size=20,face="plain"), 
        axis.text = element_text(size=20), 
        legend.title = element_text(size=20,face="bold"),
        legend.text = element_text(size=20)) +
   theme(#aspect.ratio = 1, 
          legend.position = "none",
         plot.title = element_text(size=20, hjust = .5, face = "bold"),
        )
## correlation between genes by stage

gam_corrs3_13_thesis <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC13/100, y= MSC3/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =10/100, label.y = 350/100, size = 6)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC13", y= "MSC3", title = .x)+
    theme_classic()+
    corr_theme
}) 


gam_corrs3_14_thesis <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC3/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =10/100, label.y = 350/100, size = 6)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC3", title = .x)+
    theme_classic()+
    corr_theme
}) 

gam_corrs14_13_thesis <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC13/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =10/100, label.y = 350/100, size =6)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC13", title = .x)+
    theme_classic()+
    corr_theme
}) 


gam_corrsm14_13_thesis <- map(c("Male", "Male (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC13/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =10/100, label.y = 350/100, size = 6)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC13", title = .x)+
    theme_classic()+
    corr_theme
}) 

gam_corrs3_14_thesis
gam_corrs3_13_thesis
gam_corrs14_13_thesis
gam_corrsm14_13_thesis
```
thesis
Put together all the necessary plots
```{r, fig.width= 12, fig.height=12}


plot_grid(plot_grid(plot_grid(plotlist = c(gam_corrs3_13_thesis, gam_corrs3_14_thesis[1]), ncol = 3), 
                    plot_grid(plotlist = c(gam_corrs3_14_thesis[1], gam_corrs14_13_thesis), ncol = 3),
                    plot_grid(NULL, plot_grid(plotlist = c(gam_corrsm14_13_thesis), ncol = 2, labels = c("B", ""), label_size = 24), NULL, rel_widths = c(1.5,6,1.5), ncol = 3),
                    nrow = 3,
                    labels = c("A", "", ""),
                    label_size = 24))
  

```

Thesis
Put together all the necessary plots
```{r, fig.width= 4, fig.height=2.7}
##Put together the plots for annotation
# stg_lgnd <- cowplot::get_legend(
#   scm_barpt_s[[4]] +
#     guides(fill=guide_legend(title.position = "top", nrow = 2)))

gam_hl_vn <- map(c(gams_3_13_14_vn[1], gams_vn), ~.x + theme(text = element_text(size = 40), aspect.ratio=2/4))


gam_hl_vn

# plot_grid(plot_grid(plotlist = map(c(gams_3_13_14_vn[1], gams_vn), ~.x + theme(text = element_text(size = 14), aspect.ratio=2/4)), nrow = 1, rel_widths = c(.3,.35,.35)),
#           plot_grid(plot_grid(plotlist = c(gam_corrs3_13, gam_corrs3_14), ncol = 4), 
#                               plot_grid(plotlist = c(gam_corrs14_13, gam_corrsm14_13), ncol = 4),
#                               nrow = 2),
#           upst_3_13_14_strn_de_plts_mfet,
#           # plot_grid(plotlist = map(scm_barptHL[1:2], ~.+ theme(legend.position = "none")), ncol = 2, rel_widths = c(.6, .4)),
#           # plot_grid(plotlist = map(scm_barptHL[3:4], ~.+ theme(legend.position = "none")), ncol = 2, rel_widths = c(.4, .6)),
#           # stg_lgnd,
#           nrow = 3,
#             rel_heights = c(.2,.5,.3),
#             # labels = c('A', 'B', 'C', 'D', 'E'),
#             label_size = 24,
#             vjust= 0.3)
```


Put together all the necessary plots
```{r, fig.width= 12, fig.height=12}
##Put together the plots for annotation
# stg_lgnd <- cowplot::get_legend(
#   scm_barpt_s[[4]] +
#     guides(fill=guide_legend(title.position = "top", nrow = 2)))

le_corr_venn_upset <- plot_grid(plot_grid(plot_grid(plotlist = c(gam_corrs3_13, gam_corrs3_14), ncol = 4), 
                              plot_grid(plotlist = c(gam_corrs14_13, gam_corrsm14_13), ncol = 4),
                              nrow = 2),
          NULL,
          fld_gn_expression_specificity_upset_vn_don3,
          nrow = 3,
            rel_heights = c(.53,.05,.42),
            labels = c('B','', 'C'),
            label_size = 24,
            vjust= 1)

le_corr_venn_upset
##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
saveRDS(le_corr_venn_upset, "plots/figS26/le_corr_venn_upset.RDS")

```

Put together all the necessary plots
```{r, fig.width= 12, fig.height=8}
##Put together the plots for annotation
# stg_lgnd <- cowplot::get_legend(
#   scm_barpt_s[[4]] +
#     guides(fill=guide_legend(title.position = "top", nrow = 2)))

le_corr_venn_upset <- plot_grid(plot_grid(plot_grid(plotlist = c(gam_corrs3_13, gam_corrs3_14), ncol = 4), 
                              plot_grid(plotlist = c(gam_corrs14_13, gam_corrsm14_13), ncol = 4),
                              nrow = 2),
          NULL,
          fld_gn_expression_specificity_upset_vn_don3,
          nrow = 3,
            rel_heights = c(.53,.05,.42),
            labels = c('B','', 'C'),
            label_size = 24,
            vjust= 1)

le_corr_venn_upset
##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
saveRDS(le_corr_venn_upset, "plots/figS26/le_corr_venn_upset.RDS")

```

```{r}
## Make table with core gene information that 
## 1. disregards donor level information from MSC3,13,14
mfet_gns_3.13.14_don_colps <- mfet_gns_3.13.14 %>% select(-c(donor)) %>% distinct() %>% pivot_wider(., names_from = 'stage', values_from = c('present'))

## 1. includes donor level information from MSC3,13,14
mfet_gns_3.13.14_don <- mfet_gns_3.13.14 %>%  pivot_wider(., names_from = c('stage','donor'), values_from = c('present'))
 
```



```{r}
##Find outlier genes for paper
##Female
fhe_gn_pcnt_14 <- rowSums(msc_qcd_anotd[[4]][,msc_qcd_anotd[[4]]$stageHL == 'Female']@assays$RNA@counts > 0) / (sum(msc_qcd_anotd[[4]]$stageHL == 'Female'))

##fle exclusive genes
# fl_gns_14 <- mf_gns_3.13.14 %>% filter( donor == "msc14" & is.na(MLE_genes) & is.na(M_genes) & FLE_genes == 'y' & is.na(F_genes) & is.na(LR_genes) & is.na(ET_genes)) %>% pull(gene_id)
fl_gns_14 <- mfet_gns_3.13.14_don %>% filter( is.na(MLE_genes_MSC14) & is.na(M_genes_MSC14) & FLE_genes_MSC14 == 'y' & is.na(F_genes_MSC14) & is.na(LR_genes_MSC14) & is.na(ET_genes_MSC14)) %>% pull(gene_id)


##Show genes
fhe_gn_pcnt_14[fl_gns_14]

##Male
mhe_gn_pcnt_14 <- rowSums(msc_qcd_anotd[[4]][,msc_qcd_anotd[[4]]$stageHL == 'Male']@assays$RNA@counts > 0) / (sum(msc_qcd_anotd[[4]]$stageHL == 'Male'))

##fle exclusive genes
ml_gns_14 <- mfet_gns_3.13.14_don %>% filter( is.na(FLE_genes_MSC14) & is.na(F_genes_MSC14) & MLE_genes_MSC14 == 'y' & is.na(M_genes_MSC14) & is.na(LR_genes_MSC14) & is.na(ET_genes_MSC14)) %>% pull(gene_id)
mhe_gn_pcnt_14[ml_gns_14]
```


```{r, message=FALSE}
##Different classes of genes
mfet_gns_3.13.14_don_colps <- mfet_gns_3.13.14_don_colps %>% 
  mutate(## Gametocyte core present in HE and LE of the same sex and absent in all others
         ##fh & fl
         fh_fl_only = case_when(is.na(MLE_genes) & is.na(M_genes) & FLE_genes == 'y' & F_genes == 'y' & is.na(LR_genes) & is.na(ET_genes) ~ 'y'),
         ##mh & ml
         mh_ml_only = case_when(is.na(FLE_genes) & is.na(F_genes) & MLE_genes == 'y' & M_genes == 'y' & is.na(LR_genes) & is.na(ET_genes) ~ 'y'),
         
         ##Present in fh or mh and absent in all the others
         ##fh
         fh_only = case_when(is.na(MLE_genes) & is.na(M_genes) & is.na(FLE_genes) & F_genes == 'y' & is.na(LR_genes) & is.na(ET_genes) ~ 'y'),
         ##mh
         mh_only = case_when(is.na(FLE_genes) & is.na(F_genes) & is.na(MLE_genes)  & M_genes == 'y' & is.na(LR_genes) & is.na(ET_genes) ~ 'y'),
         
         ##Present in fl, ml and either present or absent the others - This is because these cells express very few genes
         ##fl
         fl = case_when(FLE_genes == 'y' ~ 'y'), 
         ##ml
         ml = case_when(MLE_genes == 'y' ~ 'y'),
         
         ##Present in fl, ml and absent in asexuals but either present or absent in the others - This is because these cells express very few genes
         ##fl
         fl_noasex =  case_when(FLE_genes == 'y' & is.na(LR_genes) & is.na(ET_genes) ~ 'y'), 
         ##ml
         ml_noasex =  case_when(MLE_genes == 'y'& is.na(LR_genes) & is.na(ET_genes) ~ 'y'),
         
         ##Present in all stages
         ##fl
         all = case_when(if_all(contains('_genes'), ~!is.na(.))  ~ 'y'),
         
         ##Present in all gam stages and absent in asexual stages
         gam_noasex = case_when(if_all(matches('.[H|L]_genes'), ~!is.na(.) ) & if_all(matches('.[R|T]_genes'), ~is.na(.))  ~ 'y'),
         ##Present in all asexual stages and absent in gam stages
         asex_nogam = case_when(if_all(matches('.[R|T]_genes'), ~!is.na(.) ) & if_all(matches('.[H|L]_genes'), ~is.na(.))  ~ 'y'), 
  ) 



```

```{r}
## Visualize the genes expressed in various combinations of   FL, F, ML and M
mfet_gns_3.13.14_don_colps %>% filter(is.na(F_genes)&FLE_genes == 'y'& is.na(M_genes) & is.na(MLE_genes) & is.na(LR_genes) & is.na(ET_genes)) %>% head
mfet_gns_3.13.14_don_colps %>% filter(is.na(F_genes)& is.na(F_genes)& is.na(M_genes) & MLE_genes == 'y'& is.na(LR_genes) & is.na(ET_genes)) %>% head
mfet_gns_3.13.14_don_colps %>% filter(is.na(F_genes)& is.na(F_genes)& M_genes == 'y' & MLE_genes == 'y' & is.na(LR_genes) & is.na(ET_genes)) %>% head
mfet_gns_3.13.14_don_colps %>% filter(is.na(F_genes)& is.na(MLE_genes)& FLE_genes == 'y' & is.na(M_genes) & is.na(LR_genes) & is.na(ET_genes)) %>% head
```


```{r}
## Copy the gene lists to make do GO overrepresentation test
system("mkdir data/processed/DB52e_star/MSC14/mf_le_he/")
saveRDS(mfet_gns_3.13.14_don_colps, "data/processed/DB52e_star/MSC14/mf_le_he/mfet_gns_3.13.14_don_colps.RDS")

```

```{r}
## Separate tables for each donor for saving
mfet_gns_3.13.14_f <- mfet_gns_3.13.14 %>% mutate(across(donor, ~droplevels(factor(., levels= sample_name_nms_all)))) 
mfet_gns_3.13.14_lst <- split(mfet_gns_3.13.14_f, f = mfet_gns_3.13.14_f$donor)

mfet_gns_3.13.14_lst <- map(mfet_gns_3.13.14_lst, ~.x %>% pivot_wider(., names_from = 'stage', values_from = c('present')) %>% select(-c(donor)))

saveRDS(mfet_gns_3.13.14_lst, "data/processed/DB52e_star/MSC14/mf_le_he/mfet_gns_3.13.14_lst.RDS")

```
