---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(scran)
  library(SingleR)
  library(scDblFinder)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(dplyr::desc)
  conflicts_prefer(dplyr::count)
})
```



# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Variable declarations

```{r}

# ##Strain cluster colours
# strain_cols_n <-c("#910000", "#ff1a8d", "#007c71", "#0ea300", "#00deca", "#9a96c7", "#5d57a4", "#a8875b","#708db3", "#c39a00", "#ffd94d", "#9770b3", "#C9E0F1", "#a40000" , "#00b7a7", "#ffcd12","thistle","#a40000" , "#00b7a7", "#e7f1f9", 'black', '#050301','grey')
# #, 'lavender', 'lightgray', 'grey'
# names(strain_cols_n) <- c('SC2', 'SC1','SC5',  'SC4', 'SC3',  'SC6', 'SC7', 'SC8', 'SC9', 'SC10', 'SC11', 'SC12', 'S3_b', 'S1', 'S2', 'Singlet', 'Negative', 'SC1_3i','Doublet', 'Lab', 'Less5', 'PoorQC', NA)
# 
# strain_cols <- c(met.brewer(name = 'Thomas')[c(1:8)],'black', 'light gray')
# 
# names(strain_cols) = c('G1', 'G2', 'G3', 'G4','G5', 'G6','G7', 'G8', 'Doublet', 'Negative')
# 
# strain_cols =c(strain_cols,strain_cols_n)
# 
# dblt_cols <- c(brewer.pal(name = 'Set1', n = 8)[c(1:5,7:8)], 'lightblue')
# names(dblt_cols) <- c('ScDfSt', 'ScDf', 'ScSt', 'DfSt', 'Sc', 'Df','St', 'Singlet')
# 
# cell_qc_filt_nm_lvls <- c("sc_neg", "lo_gn_cnt",  "hi_gn_cnt",  "hi_mtp", "ss_db",  "strn_db", "stg_db", "pass")
# cell_qc_cols <- c(brewer.pal(name = 'Set1', n = 8)[c(1:5,7:8)], 'lightblue') %>% set_names(cell_qc_filt_nm_lvls)
# 
# dblt_ss_cols <- c(brewer.pal(name = 'Set1', n = 8)[c(1:3)], 'lightgrey')
# names(dblt_ss_cols) <- c('Strain & stage', 'Strain', 'Stage', 'Singlet')
# 
# cluster_cols <- brewer.pal(8, name = 'Set2')[c(1:4,7,8)] %>% set_names(as.character(c(0:5)))
# 
# stg_refnd_lvls = c("Early ring", "Late ring", "Early trophozoite", "Late trophozoite", "Early schizont", "Late schizont", "Gametocyte (committed)", "Gametocyte (developing)", "Gametocyte (branching)", "Female (HE)", "Female (LE)", "Gametocyte (female)", "Gametocyte (early female)", "Gametocyte (late female)", "Gametocyte (male)", "Male (HE)", "Male (LE)", "Gametocyte (early male)", "Gametocyte (late male)", "Unassigned", "Failed qc", "Not assigned", "Atlas")
# 
# sample_name_nms <- c("MSC1", "MSC3","MSC13","MSC14","MSC1272") %>% str_sort(numeric = T)
# 
# 
# sp_colors <- c('Early ring'='#D1EC9F',
#                      'Late ring'='#78C679',
#                      'Early trophozoite'='#FEEEAA',
#                      'Late trophozoite'='#FEB24C',
#                      'Early schizont'='#C9E8F1',
#                      'Late schizont'='#85B1D3',
#                      'Gametocyte (developing)'='thistle',
#                      'Gametocyte (female)'='#551A8B',
#                      'Gametocyte (male)'='mediumpurple',
#                'Unassigned'='red', 
#                'Not assigned'='#D6CFC7',
#                'Female (LE)'='#208cf7',
#                'Female (HE)'='#551A8B',
#                'Male (LE)'='#a894d1',
#                'Male (HE)'='mediumpurple',
#                'Failed QC' = 'grey',
#                'Gametocyte (committed)' = '#C399A2', 
#                'Gametocyte (developing)' = '#66444C', 
#                'Gametocyte (branching)' = '#66444C', 
#                'Gametocyte (early female)' = '#749E89', 
#                'Gametocyte (early male)' = '#7D87B2', 
#                'Gametocyte (late female)' = '#4E6D58', 
#                'Gametocyte (late male)' = '#41507B'
# )
# 
# sp_fld_colors <- c('Late ring'= '#78C679',#met.brewer('Java')[2],
#                    'Early trophozoite'='#FEEEAA',#met.brewer('Java')[3],
#                    'Female (LE)'= met.brewer('Ingres')[5],
#                    'Female (HE)'= '#551A8B', #met.brewer('Ingres')[7],
#                    # 'Male (LE)'=met.brewer('Nizami')[6],
#                    # 'Male (HE)'=met.brewer('Nizami')[8],
#                    # 'Gametocyte (male)'=met.brewer('Austria')[5],
#                    'Male (LE)'=met.brewer('Ingres')[7],#'#a2e19e',
#                    'Male (HE)'= 'mediumpurple',#'#039911',
#                    'Gametocyte (male)'='#3DED97',
#                    'Gametocyte (female)'=met.brewer('Ingres')[6],
#                    'Lab' = '#D3D3D3',
#                    'Unassigned'='#DCDCDC',
#                    'Failed QC' = 'grey',
#                    'Not assigned' = 'grey',
#                    'Atlas' = 'grey',
#                    'Early ring'='#D1EC9F',
#                      'Late trophozoite'='#FEB24C',
#                      'Early schizont'='#C9E8F1',
#                      'Late schizont'='#85B1D3',
#                    'Gametocyte (developing)'='thistle',
#                    'early female' = '#749E89',
#                    'late female'= '#4E6D58'
# )
# 
# donor_cols_m = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" ), '.mrna'))
# # donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))
# donor_cols = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))
# 
# "#DC143C"
#  "#B22222"
#  "#F08080"
#  "#FA8072"
```

```{r}
##colors
# sp_col_new <- c(`Late ring` = "#78C679", `Early trophozoite` = "#FEEEAA", `Female (LE)` = "#a97f2f", 
# Female = "#551A8B", `Male (LE)` = "#7e5522", Male = "mediumpurple", 
# `Gametocyte (male)` = "#1a6c8b", `Gametocyte (female)` = "#d1b252", 
# Lab = "#D3D3D3", Unassigned = "#DCDCDC", `Failed QC` = "grey", 
# `Not assigned` = "grey", Atlas = "grey", `Early ring` = "#D1EC9F", 
# `Late trophozoite` = "#FEB24C", `Early schizont` = "#C9E8F1", 
# `Late schizont` = "#85B1D3", `Gametocyte (developing)` = "thistle", 
# `early female` = "#749E89", `late female` = "#4E6D58", Asexual = "#FF6961", 
# `NULL` = "lightgrey", `Field unlabelled` = "#D2C1DC", "weak_score" = "#c9c9c9", "unclear" = "#dddddd")

```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")

```

#### Run emptydrops on all 
```{r}
## NOTE - USE raw_feature_bc_matrix.h5 and not the folder raw_feature_bc_matrix/ since nextflow can't expand to all the donor paths when using the folder after specifying 5736STDY*
##!!! NOTE The hdf5 matrix makes the running of emptydrops very slow so need to convert to dgCMatrix - https://github.com/MarioniLab/DropletUtils/issues/55

## Run emptydrops on all 
# nextflow run ../mali22_code_lnk/nf_scripts/edrops.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/edrops.R" --input_raw_mtx "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/raw_feature_bc_matrix.h5" --o_path "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --retain_nm "wo_retain,w_retain" --retain_val "Inf,NULL" --limt "20,50,100,150,200" --fdr_l "0.01" -resume


```

###################START- RAW READING#############

```{r, eval=T}
pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
```

## Genotype clustering for MSC

initialize variable for the sample names

```{r}
# select sample names
# slct_sample_names <-c("MSC48", "MSC68", "MSC14", "MSC49", "MSC60", "MSC66", "MSC67", "MSC70", "MSC24", "MSC33", "MSC40", "MSC54", "MSC55") %>% str_sort(., numeric = T)

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```

```{r}
## limit for different donors
edrops_thresh <- rep(50, length(pf_solo_mixed_mdata_decode$sample_nm)) %>% set_names(pf_solo_mixed_mdata_decode$sample_nm)
edrops_thresh[["MSC1"]] <- 150
edrops_thresh[["MSC13"]] <- 100
edrops_thresh[["MSC14"]] <- 100
edrops_thresh[["MSC24"]] <- 200
edrops_thresh[["MSC37"]] <- 200
edrops_thresh[["MSC38"]] <- 200
edrops_thresh[["MSC53"]] <- 150
edrops_thresh[["MSC57"]] <- 150
edrops_thresh[["MSC50_SLO"]] <- 100
```


!!!NOTE - Reading in the metadata to add to the mixed species samples which were missed in the initial step - Will need to redo adding to the source
```{r}
mrc_sk21_summary <- read_csv("data/raw/MRC_Sample_Analysis_Summary_Summary.csv", skip = 1) 

# mrc_sk21_summary_mxd <- mrc_sk21_summary %>% filter(Sample_Name %in% c("MSC25" ,"MSC41", "MSC51")) %>% mutate(across(Symptomatic, ~str_replace_all(., c("Y" = "Yes", "N" = "No")))) %>% select(Sample_Name,Symptomatic, Hb_Type) %>% rename("donor" = Sample_Name)
```
```{r}
mrc_sk21_summary_slct <- mrc_sk21_summary %>% 
  left_join(pf_solo_mixed_mdata_decode, ., by = c("irods_id" = "Sanger_SampleID")) %>%
  # mutate(donor = str_remove(sample_nm, "_.*")) %>%
  rename("perc_map_p" = `%map to pfpmpo`, "perc_map_hs" = `%map to hs`) %>%
  mutate(p_map_cat = case_when(perc_map_p > 0 & perc_map_p < 33 ~ "low",
                              perc_map_p >= 33 & perc_map_p < 66 ~ "medium",
                              perc_map_p >= 66 & perc_map_p <= 100 ~ "high"),
         hs_map_cat = case_when(perc_map_hs > 0 & perc_map_hs < 33 ~ "low",
                              perc_map_hs >= 33 & perc_map_hs < 66 ~ "medium",
                              perc_map_hs >= 66 & perc_map_hs <= 100 ~ "high"),
         across(c(p_map_cat, hs_map_cat), ~factor(., levels = c("low", "medium", "high")))) %>% 
  mutate(peak = factor(str_replace(str_remove(Peak, " peak"), "Very ", "V."), levels = c("No", "Small", "Medium", "Big", "V.big"))) %>%
  select(sample_nm, pf, pm, po, nStrains_Pf, nStrains_Pm, nStrains_Po, Hb_Type, mixed_infection, Jumpcode, Gel_Emulsion, Symptomatic, perc_map_p, perc_map_hs, p_map_cat, hs_map_cat, Hb_perc, `RBCs loaded`, peak, Comment)
```


```{r}
mrc_sk21_summary_slct
```


```{r}
mrc_sk21_summary_slct %>%
  dplyr::count(p_map_cat, hs_map_cat)

```


```{r, fig.width=15, fig.height=5}
# donor_mdata <- read_csv("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/MRC_Sample_Manifest_FOR_GEM_10X.csv.csv", col_types =  paste0(c(rep("c", 7), "d", "c","d","d", rep("c", 14)), collapse = ""), show_col_types = T)  
donor_10x_mdata <- read_csv("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/MRC_Sample_Manifest - FOR_GEM_10X.csv", 
                            col_types =  paste0(c(rep("c", 9), rep("d", 8), "c" , rep("d", 6), rep("c", 4), rep("d", 6)), collapse = ""),
                            show_col_types = T
                            ) 

donor_10x_mdata <- donor_10x_mdata %>% rename_all(., ~str_to_lower(str_replace_all(.,c("\\["="_", "\\]"="", " "="", "-"="",  "\\/"="_", "º"=""))))
```


```{r, fig.width=15, fig.height=5}

donor_10x_mdata %>%
  filter(!is.na(short_name) & str_detect(short_name, "MSC"))%>%
  filter(short_name %in% slct_sample_names) 

```

```{r, fig.width=15, fig.height=5}

donor_10x_mdata_slct <- donor_10x_mdata %>%
  filter(!is.na(short_name) & str_detect(short_name, "MSC"))%>%
  filter(short_name %in% slct_sample_names) %>%
  rename("donor" = short_name,
         "parasites_loaded" = totalparasitesloaded,
         "macsloaded" = macsloaded...29,
         "sloloaded" = sloloaded...30) %>% 
  mutate(frac_load = case_when(macsloaded > 0 & sloloaded > 0 ~ "MACS_SLO",
                               macsloaded > 0 & (sloloaded == 0 | is.na(sloloaded)) ~ "MACS",
                               (macsloaded == 0 | is.na(macsloaded)) & sloloaded > 0 ~ "SLO"),
         overload = case_when(parasites_loaded > 20000 ~ "Yes_20k",
                              parasites_loaded > 16000 ~ "Yes_16_20k",
                              parasites_loaded > 10000 & parasites_loaded <= 16000 ~ "No_10_16K",
                              parasites_loaded < 16000 ~ "No_under10K"),
         overload = factor(overload, levels = c("No_under10K", "No_10_16K", "Yes_16_20k", "Yes_20k")),
         frac_over = case_when(macsloaded > 16000 & sloloaded > 16000 ~ "MACS_SLO",
                               macsloaded > 16000 & (sloloaded == 0 | is.na(sloloaded)) ~ "MACS",
                               (macsloaded == 0 | is.na(macsloaded)) & sloloaded > 16000 ~ "SLO")) %>%
  select(donor, macs_smear_parasitemia_fraction, macs_parasites_per_ul, slo_smear_parasitemia_fraction, slo_parasites_per_ul, slo_volume_10x, rbcs_in_mix, parasites_loaded...23, targeted_cells, gel_emulsion, macsloaded, sloloaded, parasites_loaded, frac_load, overload, frac_over)

```


```{r, fig.width=15, fig.height=5}
donor_10x_mdata_sk21_summ <- donor_10x_mdata_slct %>% left_join(., mrc_sk21_summary_slct, by = c("donor" = "sample_nm"))

```

```{r, fig.width=15, fig.height=5}
saveRDS(donor_10x_mdata_sk21_summ, "data/processed/Pf/m2_all/donor_10x_mdata_sk21_summ.RDS")

```

```{r, fig.width=15, fig.height=5}

donor_10x_mdata_sk21_summ %>% count(frac_load, overload, frac_over, targeted_cells)
```



```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_cbender_mdata <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender/cb_mdata.*'))) %>% 
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(read_csv) 
```


```{r, message=F, warning=F, eval=T}
all_cbender_mdata
```


```{r, message=F, warning=F, eval=T}
imap(all_cbender_mdata, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_probability)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```


```{r, message=F, warning=F, eval=T, fig.height=9, fig.width=7}
all_cbender_mdata %>%
  bind_rows(., .id = "donor") %>% 
      filter(cell_probability > 0.8) %>%
  group_by(donor) %>% 
  mutate(bg_50 = sum(background_fraction > 0.5)) %>% 
  ungroup() %>%
  ggplot(aes(x = background_fraction)) +
    geom_histogram() +
    geom_text(aes(label=bg_50, x = 0.8, y = 200),  size = 3)+
    geom_vline(xintercept = 0.5) +
    facet_wrap(donor ~ ., scales = "free_y", ncol = 5)

```

```{r, message=F, warning=F, eval=T, fig.height=8, fig.width=12}
all_cbender_mdata %>%
  bind_rows(., .id = "donor") %>% 
      filter(cell_probability > 0.8) %>%
  count(donor, background_fraction>0.5)

```

```{r, message=F, warning=F, eval=T}
imap(all_cbender_mdata, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = background_fraction, y = cell_size)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.5) +
    labs(title = .y)
)
```

```{r, message=F, warning=F, eval=T}
imap(all_cbender_mdata, ~.x %>% 
      filter(cell_probability > 0.5) %>%
  ggplot(aes(x = cell_probability, y = cell_size)) +
    geom_point(size = 0.1) +
    geom_vline(xintercept = 0.9) +
    labs(title = .y)
)
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_cbender_mdata_pass.5 <- map(all_cbender_mdata, ~.x %>% filter(cell_probability>0.5) %>% pull(barcode))

```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_cbender_mdata_pass.9 <- map(all_cbender_mdata, ~.x %>% filter(cell_probability>0.9) %>% pull(barcode))

```

## Save data
```{r}
part <- "Part1"
outdir <- paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/sk_25_QC/PmML03_PowML03_QC_objects/", part, "/")

# processed_data_noretain <- readRDS(paste0(outdir, "Part1_EmptyDrops_pmpo_samples_50_to_500_wo_retain.RDS"))
# processed_data_withretain <- readRDS( paste0(outdir, "Part1_EmptyDrops_pmpo_samples_50_to_500_w_retain.RDS"))
```


```{r}
# processed_data_noretain_50 <- map(processed_data_noretain, ~.x[["50"]] %>% filter(!is.na(Barcode) )) %>% .[sample_name]
# processed_data_noretain_100 <- map(processed_data_noretain, ~.x[["100"]] %>% filter(!is.na(Barcode) )) %>% .[sample_name]
# 
# processed_data_withretain_50 <- map(processed_data_withretain, ~.x[["50"]] %>% filter(!is.na(Barcode) )) %>% .[sample_name]
# processed_data_withretain_100 <- map(processed_data_withretain, ~.x[["100"]] %>% filter(!is.na(Barcode) )) %>% .[sample_name]

```


```{r}
# processed_data_noretain_50
# processed_data_withretain_100

```

```{r}
processed_data_w_retain <- readRDS("data/processed/Pf/m2_all/processed_data_w_retain.RDS")
processed_data_wo_retain <- readRDS("data/processed/Pf/m2_all/processed_data_wo_retain.RDS")
```

```{r, fig.height=4.0, fig.width=4.0}
# M48, M49, M60
map(slct_sample_names, ~{
  print(paste0(.x, "_"))
  lst <- processed_data_wo_retain[str_detect(names(processed_data_wo_retain), paste0(.x, "_"))]
  lst <- lst[!str_detect(names(lst), "_20_")]
  names(lst) <- str_remove_all(names(lst), "MSC[0-9]+_|_retain")
  print(names(lst))
  lst <- map(lst, ~.x$Barcode)
  ggvenn(lst,
       show_percentage = F,
       text_size = 4,
       set_name_size = 4) +
    labs(title = .x)
})

```

```{r, fig.height=4.0, fig.width=4.0}
# M48, M49, M60
map(names(edrops_thresh[edrops_thresh >50]), ~{
  print(paste0(.x, "_"))
  lst <- processed_data_wo_retain[str_detect(names(processed_data_wo_retain), paste0(.x, "_"))]
  lst <- lst[!str_detect(names(lst), "_20_")]
  names(lst) <- str_remove_all(names(lst), "MSC[0-9]+_|_retain")
  print(names(lst))
  lst <- map(lst, ~.x$Barcode)
  ggvenn(lst,
       show_percentage = F,
       text_size = 4,
       set_name_size = 4) +
    labs(title = .x)
})

```

```{r}
names(processed_data_w_retain )
```


```{r}
wo_rtn_50 <- paste0(names(edrops_thresh), "_", edrops_thresh, "_wo_retain")
w_rtn_50 <- paste0(names(edrops_thresh), "_", edrops_thresh, "_w_retain")

wo_rtn_100 <- paste0(names(edrops_thresh), "_", "100", "_wo_retain")
w_rtn_100 <- paste0(names(edrops_thresh), "_", "100", "_w_retain")
```


```{r}
processed_data_noretain_50 <- processed_data_wo_retain[wo_rtn_50] %>% set_names(names(edrops_thresh)) %>% .[sample_name]
processed_data_withretain_50 <- processed_data_w_retain[w_rtn_50] %>% set_names(names(edrops_thresh))%>% .[sample_name]

processed_data_noretain_100 <- processed_data_wo_retain[wo_rtn_100] %>% set_names(names(edrops_thresh))%>% .[sample_name]
processed_data_withretain_100 <-  processed_data_w_retain[w_rtn_100] %>% set_names(names(edrops_thresh))%>% .[sample_name]

```


```{r}
# # processed_data_noretain_50 <- processed_data_wo_retain[str_detect(names(processed_data_wo_retain), "_50_wo_retain")]
# # processed_data_noretain_50 <- processed_data_noretain_50 %>% .[names(processed_data_noretain_50)[str_detect(names(processed_data_noretain_50), paste0(sample_name, collapse = "|"))]]
# # names(processed_data_noretain_50) <- str_remove(names(processed_data_noretain_50), "_50_wo_retain")
# 
# processed_data_noretain_100 <- processed_data_wo_retain[str_detect(names(processed_data_wo_retain), "_100_wo_retain")]
# processed_data_noretain_100 <- processed_data_noretain_100 %>% .[names(processed_data_noretain_100)[str_detect(names(processed_data_noretain_100), paste0(sample_name, collapse = "|"))]]
# names(processed_data_noretain_100) <- str_remove(names(processed_data_noretain_100), "_100_wo_retain")
# 
# 
# # processed_data_withretain_50 <-  processed_data_w_retain[str_detect(names(processed_data_w_retain), "_50_w_retain")]
# # processed_data_withretain_50 <- processed_data_withretain_50 %>% .[names(processed_data_withretain_50)[str_detect(names(processed_data_withretain_50), paste0(sample_name, collapse = "|"))]]
# # names(processed_data_withretain_50) <- str_remove(names(processed_data_withretain_50), "_50_w_retain")
# 
# processed_data_withretain_100 <-  processed_data_w_retain[str_detect(names(processed_data_w_retain), "_100_w_retain")]
# processed_data_withretain_100 <- processed_data_withretain_100 %>% .[names(processed_data_withretain_100)[str_detect(names(processed_data_withretain_100), paste0(sample_name, collapse = "|"))]]
# names(processed_data_withretain_100) <- str_remove(names(processed_data_withretain_100), "_100_w_retain")
# 

```


```{r}
overlap_tbl <- pmap(list(processed_data_noretain_50, processed_data_withretain_50, processed_data_noretain_100, processed_data_withretain_100), ~{
  data.frame(
  "Limit (no retain): 50" = length(..1$Barcode),
  "Limit (retain): 50" = length(..2$Barcode),
  "Overlap 50 (prop)" = length(intersect(..1$Barcode, ..2$Barcode))/ length(unique(c(..1$Barcode, ..2$Barcode))),
  "Limit (no retain): 100" = length(..3$Barcode),
  "Limit (retain): 100" = length(..4$Barcode),
  "Overlap 100 (prop)" = length(intersect(..3$Barcode, ..4$Barcode))/ length(unique(c(..3$Barcode, ..4$Barcode))),
  "Overlap 50.100 noRet (prop)" = length(intersect(..1$Barcode, ..3$Barcode))/ length(unique(c(..1$Barcode, ..3$Barcode))),
  "Overlap all (prop)" = length(Reduce(intersect, list(..1$Barcode, ..2$Barcode, ..3$Barcode, ..4$Barcode)))/ length(unique(c(..1$Barcode, ..2$Barcode, ..3$Barcode, ..4$Barcode))),
  check.names = F
)
  }) %>%
  bind_rows(.id = "donor")

overlap_tbl
```

```{r}

overlap_tbl
```


```{r}
# nms_cb <- names(all_cbender_mdata)

# overlap_tbl_cbender <- pmap(list(processed_data_noretain_50, processed_data_noretain_100, all_cbender_mdata_pass.5, all_cbender_mdata_pass.9), ~{
overlap_tbl_cbender <- pmap(list(processed_data_noretain_50, processed_data_noretain_100, all_cbender_mdata_pass.9, all_cbender_mdata_pass.5), ~{
  data.frame(
  "Limit (no retain): 50" = length(..1$Barcode),
  "Limit (no retain): 100" = length(..2$Barcode),
  "CBendr" = length(..3),
  "CBendr.9" = length(..4),
  "Overlap 50:CBendr (prop)" = length(intersect(..1$Barcode, ..3))/ length(unique(c(..1$Barcode, ..3))),
  "Overlap 100:CBendr (prop)" = length(intersect(..2$Barcode, ..3))/ length(unique(c(..2$Barcode, ..3))),
  "50 in CBendr" = sum(..1$Barcode %in% ..3),
  "Overlap all (prop)" = length(Reduce(intersect, list(..1$Barcode, ..2$Barcode, ..3)))/ length(unique(c(..1$Barcode, ..2$Barcode, ..3))),
  check.names = F
)
  }) %>%
  bind_rows(.id = "donor")

overlap_tbl_cbender

```



```{r, fig.width=10, fig.height=5}
## Function to make line plots to compare the different cell calling methods and the overlap between them

methds_ln_plt_nohline <- function(tbl, x_var = "donor", y_var, col_var = "mthds_union") {
  ggplot(data = tbl, aes(x= !!rlang::sym(x_var), y = !!rlang::sym(y_var), colour = !!rlang::sym(col_var)))  +
  geom_point(size = 4) +
  geom_line(aes(group = !!rlang::sym(col_var)), size = 1) +
  theme_classic() + 
  theme(text = element_text(size =20),
        axis.text.x = element_text(size =20, angle = 45, hjust = 1),
        legend.position = "bottom")
}

```

```{r, fig.width=14, fig.height=5}
overlap_tbl_cbender %>% 
  pivot_longer(cols = c("Limit (no retain): 50", "Limit (no retain): 100", "CBendr", "CBendr.9", "50 in CBendr"), names_to = "Cell_id_mthd", values_to = "Cells") %>%
  mutate(across(donor, ~str_remove(., "_[0-9].*")),
         Cells_log10 = log10(Cells)) %>%
  methds_ln_plt_nohline(., x_var = "donor", y_var = "Cells_log10", col_var = "Cell_id_mthd")

```



```{r, message=F, warning=F, eval=T}
# gogo()
```

```{r, fig.width=14, fig.height=4.3}
overlap_tbl_cbender %>% 
  pivot_longer(cols = c("Limit (no retain): 50", "Limit (no retain): 100", "CBendr", "CBendr.9", "50 in CBendr"), names_to = "Cell_id_mthd", values_to = "Cells") %>%
  mutate(across(donor, ~str_remove(., "_[0-9].*"))) %>%
  methds_ln_plt_nohline(., x_var = "donor", y_var = "Cells", col_var = "Cell_id_mthd")

```


```{r, fig.width=10, fig.height=5}
## Function to make line plots to compare the different cell calling methods and the overlap between them

methds_ln_plt <- function(tbl, x_var = "donor", y_var, col_var = "mthds_union") {
  ggplot(data = tbl, aes(x= !!rlang::sym(x_var), y = !!rlang::sym(y_var), colour = !!rlang::sym(col_var)))  +
  geom_point(size = 3) +
  geom_line(aes(group = !!rlang::sym(col_var)), size = 1) +
  geom_hline(yintercept = .50) +
  geom_hline(yintercept = .70) +
  theme_classic() + 
  theme(text = element_text(size =20),
        axis.text.x = element_text(size =20, angle = 45, hjust = 1),
        legend.position = "bottom")
}

```

```{r, fig.width=10, fig.height=4.3}
overlap_tbl_cbender %>% 
  pivot_longer(cols = c("Overlap 50:CBendr (prop)", "Overlap 100:CBendr (prop)"), names_to = "Cell_id_mthd", values_to = "Cells") %>%
  mutate(across(donor, ~str_remove(., "_[0-9].*")))  %>%
  methds_ln_plt(., x_var = "donor", y_var = "Cells", col_var = "Cell_id_mthd")

```



```{r}
# overlap_tbl_cbender %>% 
#   pivot_longer(cols = -c(donor), names_pattern = "(.*)\\.(S.*)", names_to = c( "Dbt_method", ".value")) %>%
#   ggplot(aes(x= donor, y = Sensitivity, color = Dbt_method))  +
#   geom_point() +
#   geom_line(aes(group = Dbt_method), linetype = "dotted")+
#   geom_point(aes(y = Specificity)) +
#   geom_line(aes(y = Specificity, group = Dbt_method))+
#   scale_y_continuous( name = 'Sensitivity', sec.axis = sec_axis(~. *1,name = "Specificity"))

```

##### Loaded versus recovered cells

```{r, fig.width=15, fig.height=5}

mdata_10x_resc <- donor_10x_mdata_sk21_summ %>% left_join(overlap_tbl_cbender, by = "donor")
```

```{r, fig.width=14, fig.height=4.3}
map(c("overload", "frac_load", "p_map_cat", "peak"), ~{

  mdata_10x_resc %>% 
    pivot_longer(cols = c("Limit (no retain): 50", "Limit (no retain): 100", "CBendr", "CBendr.9", "50 in CBendr"), names_to = "Cell_id_mthd", values_to = "Cells") %>%
    mutate(across(donor, ~str_remove(., "_[0-9].*"))) %>%
    ggplot(aes(x= donor, y = Cells, color = Cell_id_mthd))  +
    geom_point(size = 4) +
    geom_line(aes(group = Cell_id_mthd), size = 1) +
    theme_classic()  +
    facet_grid(cols = vars(!!rlang::sym(.x)), scales = "free_x", space =  "free_x") + 
    theme(text = element_text(size =20),
          axis.text.x = element_text(size =20, angle = 45, hjust = 1),
          legend.position = "bottom")
})
```


```{r, fig.width=14, fig.height=4.3}
map(c("overload", "frac_load", "p_map_cat", "peak"), ~{

  mdata_10x_resc %>% 
  pivot_longer(cols = c("Overlap 50:CBendr (prop)", "Overlap 100:CBendr (prop)"), names_to = "Cell_id_mthd", values_to = "Cells") %>%
  mutate(across(donor, ~str_remove(., "_[0-9].*"))) %>%
  ggplot(aes(x= donor, y = Cells, color = Cell_id_mthd))  +
  geom_point(size = 4) +
  geom_line(aes(group = Cell_id_mthd), size = 1) +
  geom_hline(yintercept = .50) +
  geom_hline(yintercept = .70) +
  facet_grid(cols = vars(!!rlang::sym(.x)), scales = "free_x", space =  "free_x") +
  theme_classic() + 
  theme(text = element_text(size =20),
        axis.text.x = element_text(size =20, angle = 45, hjust = 1),
        legend.position = "bottom")
})
```



```{r, fig.width=15, fig.height=5}

mdata_10x_resc %>% 
  pivot_longer(cols = c("Limit (no retain): 50", "CBendr", "Limit (no retain): 100"), names_to = "Cell_id_mthd", values_to = "cellNo") %>%
  ggplot(aes(x =  log10(parasites_loaded), y =  log10(cellNo))) +
  geom_point(size = 4) +
  geom_line(aes(group = Cell_id_mthd), size = 1) +
  # scale_y_log10() +
  # scale_x_log10() +
  facet_wrap(Cell_id_mthd ~ .)
```

```{r, fig.width=15, fig.height=5}

mdata_10x_resc %>% 
  pivot_longer(cols = c("Limit (no retain): 50", "CBendr", "Limit (no retain): 100"), names_to = "Cell_id_mthd", values_to = "cellNo") %>%
  ggplot(aes(x = log10(rbcs_in_mix), y = log10(cellNo))) +
  geom_point(size = 4) +
  geom_line(aes(group = Cell_id_mthd), size = 1) +
  facet_wrap(Cell_id_mthd ~ .)
```


```{r, fig.width=15, fig.height=5}

donor_10x_mdata_sk21_summ %>% count(frac_load, overload, frac_over, targeted_cells)
```


```{r, fig.width=9, fig.height=4}

mdata_10x_resc_cell_mthd <- mdata_10x_resc %>% 
  pivot_longer(cols = c("Limit (no retain): 50", "CBendr", "Limit (no retain): 100"), names_to = "Cell_id_mthd", values_to = "cellNo") 

```

```{r, fig.width=9, fig.height=4}
map(c("overload", "frac_load", "p_map_cat", "peak"), ~{

  mdata_10x_resc_cell_mthd%>%
  ggplot(aes(y =  cellNo, x = !!rlang::sym(.x) )) +
  geom_boxplot() +
  geom_point(size = 1) +
  facet_wrap(Cell_id_mthd ~ .)+ 
  theme(text = element_text(size =20),
        axis.text.x = element_text(angle = 25, hjust = 1))
})
```



```{r, fig.width=9, fig.height=4}

mdata_10x_resc_cell_mthd %>%
  ggplot(aes(y =  cellNo, x = frac_load )) +
  geom_boxplot() +
  geom_point(size = 1) +
  geom_text_repel(data = mdata_10x_resc_cell_mthd %>% filter(cellNo > 20000), aes(label = donor), vjust = -0.5, color = "red", size = 5) +
  facet_wrap(Cell_id_mthd ~ .)+ 
  theme(text = element_text(size =20),
        axis.text.x = element_text(angle = 25, hjust = 1))


```


```{r, fig.width=9, fig.height=4}

mdata_10x_resc_cell_mthd %>%
  ggplot(aes(y =  cellNo, x = frac_load )) +
  geom_boxplot() +
  geom_point(size = 1) +
  geom_text_repel(data = mdata_10x_resc_cell_mthd %>% filter(overload == "Yes_20k"), aes(label = donor), vjust = -0.5, color = "red", size = 5) +
  facet_wrap(Cell_id_mthd ~ .)+ 
  theme(text = element_text(size =20),
        axis.text.x = element_text(angle = 25, hjust = 1))


```

```{r, eval=FALSE}
cell_count_table <- readRDS(paste0(outdir, "Part1_EmptyDrops_pmpo_samples_50_to_500_w_wo_retain_cell_count_table.RDS"))
cell_count_table
```

```{r}

# ed_counts_lndscape
```

```{r, message=F, warning=F, eval=F}
##Write BPcells matrices
list.files("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf//|/outs.*'))) %>% 
  .[source_irods] %>%
  set_names(names(source_irods)) %>%
  # .[sample_name] %>%
  map(~open_matrix_10x_hdf5(path = .)) %>% 
  imap(., ~write_matrix_dir(mat = .x, dir = paste0('data/processed/Pf/', .y, '/bpcells/counts_mtx'), overwrite = T))
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices. - prepared in m2_QC_prelim_stage_anno.RMD

all_msc_bp_filt <- list.files("data/processed/Pf", pattern = "^counts_mtx$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/bpcells.*'))) %>%
  # .[str_detect(., 'filtered')]  %>% 
  .[sample_name] %>%
  map(~open_matrix_dir(dir = .)) %>% 
  imap(., ~CreateSeuratObject(counts = .x, project = paste0(.y, ".mrna")))

```


```{r, eval=F}
## List filtered cellranger h5 matrices into R
list.files("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/", pattern = "raw_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T)
```


```{r, message=F, warning=F, eval=F}
##Write BPcells matrices
list.files("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/", pattern = "raw_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf//|/outs.*'))) %>% 
  .[source_irods] %>%
  set_names(names(source_irods)) %>%
  # .[sample_name] %>%
  map(~open_matrix_10x_hdf5(path = .)) %>% 
  imap(., ~write_matrix_dir(mat = .x, dir = paste0('data/processed/Pf/', .y, '/bpcells/raw_counts_mtx'), overwrite = T))
```


```{r, message=F, warning=F, eval=T}
#gogo()
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_bp_raw <- list.files("data/processed/", pattern = "raw_counts_mtx$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed//Pf/|/bpcells.*'))) %>% 
  .[sample_name] %>%
  map(~open_matrix_dir(dir = .)) %>% 
  imap(., ~CreateSeuratObject(counts = .x, project = paste0(.y, ".mrna"))) 
```


```{r}
# all_msc_bp_raw_exp <- map(all_msc_bp_raw[1:5], ~.x %>% AggregateExpression)
```


```{r}
pf_hs_gene_sum_fn <- function(seurat_obj){
  
  # Identify human and pf genes
  pf_genes <- grep("^PF3D7", rownames(seurat_obj), value = TRUE)
  hs_genes <- grep("^ENSG", rownames(seurat_obj), value = TRUE)
  print(head(hs_genes))
  # Extract expression matrix and subset to those genes
  
  
  pf_matrix <- as(GetAssayData(seurat_obj, layer = "counts")[pf_genes, , drop = FALSE], Class = "dgCMatrix")  
  hs_matrix <- as(GetAssayData(seurat_obj, layer = "counts")[hs_genes, , drop = FALSE], Class = "dgCMatrix")  
  
  # sum the umis and genes across all genes per cell
  pf_umi_count <- colSums(pf_matrix)
  hs_umi_count <- colSums(hs_matrix)
  pf_gn_count <- colSums(pf_matrix>0)
  hs_gn_count <- colSums(hs_matrix>0)

  # Store in metadata
  seurat_obj[["pf_umi_count"]] <- pf_umi_count
  seurat_obj[["hs_umi_count"]] <- hs_umi_count
  seurat_obj[["pf_gn_count"]] <- pf_gn_count
  seurat_obj[["hs_gn_count"]] <- hs_gn_count
  
  # Keep only pf genes
  seurat_obj <- subset(seurat_obj, features = pf_genes)
  
  return(seurat_obj)
  
}

```


```{r}
## Count the pf and human genes and add to metadata for filtered set and also remove human genes
all_msc_bp_filt <- map(all_msc_bp_filt, ~pf_hs_gene_sum_fn(seurat_obj = .x))
```

```{r}
## Count the pf and human genes and add to metadata for raw set and also remove human genes
all_msc_bp_raw <- map(all_msc_bp_raw, ~pf_hs_gene_sum_fn(seurat_obj = .x))
```


Save paths to raw MSC files in list - also needed for soupX below
```{r}
imap(all_msc_bp_filt, ~{
  .x@meta.data %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point() +
    geom_hline(yintercept = log10(400)) +
    geom_vline(xintercept = log10(400)) +
    labs(title = .y)
})
```

```{r}
imap(all_msc_bp_raw, ~{
  .x@meta.data %>%
    filter(pf_umi_count > 50) %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    labs(title = .y)
})
```

###################END - RAW READING##############

Save paths to raw MSC files in list - also needed for soupX below
```{r}
# all_msc_raw_paths <- list.files("data/raw/Pf", pattern = "raw_feature_bc_matrix$", full.names = T, recursive = T, include.dirs = T) %>% 
#   str_sort(., numeric = T) %>%
#   .[str_detect(., 'DB52e_star')]
```


```{r}
edrops_cbender_mdata <- pmap(list(processed_data_noretain_50, all_cbender_mdata, all_msc_bp_filt, all_msc_bp_raw), ~{
  ..1 %>% 
    data.frame() %>% 
    mutate(is_cell = ifelse(FDR <= 0.01, "CellEd", "nonCell")) %>%
    full_join(., ..2[,c("barcode", "cell_probability")] %>%
                mutate(is_cell = case_when(cell_probability > 0.8 ~ "CellCb.8",
                                           cell_probability > 0.5 ~ "CellCb.5",
                                           T ~ "nonCell")),
              by = c("Barcode" = "barcode"),
              suffix = c("_ed", "_cb")) %>%
    full_join(..3@meta.data %>% 
                rownames_to_column("bcode") %>% 
                mutate("is_cell_cr" = "CellCr") %>% 
                .[,c("bcode", "is_cell_cr")],
              .,
              by = c("bcode" = "Barcode")) %>%
    left_join(., ..4@meta.data[, c("nFeature_RNA", "nCount_RNA", "pf_umi_count", "hs_umi_count", "pf_gn_count", "hs_gn_count")] %>% 
                rownames_to_column("bcode"),
              by = c("bcode" = "bcode"))
    
})
```


```{r}
saveRDS(edrops_cbender_mdata, "data/processed/Pf/m2_all/edrops_cbender_mdata.RDS")
```


```{r}
edrops_cbender_mdata <- readRDS("data/processed/Pf/m2_all/edrops_cbender_mdata.RDS")
```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
bind_rows(edrops_cbender_mdata, .id = "donor") %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(50)) +
    geom_vline(xintercept = log10(100)) 

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
    bind_rows(edrops_cbender_mdata, .id = "donor") %>%
    ggplot(aes(x = log10(pf_gn_count), y = log10(hs_gn_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) 

```



```{r}
# edrops_cbender_mdata_uni <- map(edrops_cbender_mdata, ~.x%>% 
#     filter(!(is.na(is_cell_cr) & is_cell_cb == "nonCell" & is.na(is_cell_ed))) %>% 
#     mutate(mthds_union = case_when(is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb.") & is_cell_ed == "CellEd" ~ "CrCbEd",
#                                    is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb.") & is.na(is_cell_ed) ~ "CrCb",
#                                    is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb)) & is_cell_ed == "CellEd" ~ "CrEd",
#                                    is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb.") & is_cell_ed == "CellEd" ~ "CbEd",
#                                    is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb)) & is.na(is_cell_ed) ~ "Cr",
#                                    is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb.") & is.na(is_cell_ed) ~ "Cb",
#                                    is.na(is_cell_cr) & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb)) & is_cell_ed == "CellEd" ~ "Ed"),
#            across(mthds_union, ~factor(., levels = c("CrCbEd", "CrCb", "CrEd", "CbEd", "Cr", "Cb", "Ed"))))
# ) 

edrops_cbender8_mdata_uni <- map(edrops_cbender_mdata, ~.x%>% 
    filter(!(is.na(is_cell_cr) & (is_cell_cb == "nonCell" | is_cell_cb == "CellCb.5") & is.na(is_cell_ed))) %>% 
    mutate(mthds_union = case_when(is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb.8") & is_cell_ed == "CellEd" ~ "CrCbEd",
                                   is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb.8") & is.na(is_cell_ed) ~ "CrCb",
                                   is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb) | is_cell_cb == "CellCb.5") & is_cell_ed == "CellEd" ~ "CrEd",
                                   is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb.8") & is_cell_ed == "CellEd" ~ "CbEd",
                                   is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb) | is_cell_cb == "CellCb.5") & is.na(is_cell_ed) ~ "Cr",
                                   is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb.8") & is.na(is_cell_ed) ~ "Cb",
                                   is.na(is_cell_cr) & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb) | is_cell_cb == "CellCb.5") & is_cell_ed == "CellEd" ~ "Ed"),
           across(mthds_union, ~factor(., levels = c("CrCbEd", "CrCb", "CrEd", "CbEd", "Cr", "Cb", "Ed"))))
) 

```

```{r}
# map(edrops_cbender_mdata_uni, ~.x %>% count(is_cell_cr, is_cell_ed, is_cell_cb,mthds_union))
map(edrops_cbender8_mdata_uni, ~.x %>% dplyr::count(is_cell_cr, is_cell_ed, is_cell_cb,mthds_union))

```

```{r}
saveRDS(edrops_cbender8_mdata_uni, "data/processed/Pf/m2_all/edrops_cbender8_mdata_uni.RDS")
```

```{r, fig.width=10, fig.height=12}
map(edrops_cbender8_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
    bind_rows(edrops_cbender8_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(pf_gn_count), y = log10(hs_gn_count), colour = mthds_union)) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) 

```

```{r}
edrops_cbender_pf100_mdata <- map(edrops_cbender_mdata, ~.x %>% filter(pf_umi_count >= 100))
```

```{r}
# saveRDS(edrops_cbender_pf100_mdata, "data/processed/Pf/m2_all/edrops_cbender_pf100_mdata.RDS")
```

```{r}
edrops_cbender8_pf100_mdata_uni <- map(edrops_cbender8_mdata_uni, ~.x %>% filter(pf_umi_count >= 100))
```

```{r}
# saveRDS(edrops_cbender8_pf100_mdata_uni, "data/processed/Pf/m2_all/edrops_cbender8_pf100_mdata_uni.RDS")
```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
bind_rows(edrops_cbender8_pf100_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    # geom_vline(xintercept = log10(50)) +
    geom_vline(xintercept = log10(100)) 

```

```{r, fig.width=10, fig.height=12}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```




```{r, fig.width=10, fig.height=4}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union) %>% mutate("cb_ed_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CbEd"), "yes", "no")) %>% group_by(cb_ed_ovlp) %>% mutate(prop_ovlp = sum(n)) %>% distinct(cb_ed_ovlp, .keep_all = T) %>% ungroup() %>% mutate(prop = prop_ovlp/sum(prop_ovlp)) %>% filter(cb_ed_ovlp == "yes")) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "prop", col_var = "mthds_union") +
  theme(legend.position = "none")


```

```{r, fig.width=10, fig.height=5}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "n", col_var = "mthds_union")

```

```{r, fig.width=10, fig.height=5}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union) %>% mutate(prop = n/sum(n))) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "prop", col_var = "mthds_union")


```

```{r, fig.width=10, fig.height=4}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% 
      count(mthds_union) %>% 
      mutate("cb_ed_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CbEd"), "yes", "no"),
             "cr_cb_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CrCb"), "yes", "no"),
             "cr_ed_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CrEd"), "yes", "no")) %>% 
      group_by(cb_ed_ovlp) %>% mutate(cb_ed_sum = sum(n)) %>% 
      group_by(cr_cb_ovlp) %>% mutate(cr_cb_sum = sum(n)) %>% 
      group_by(cr_ed_ovlp) %>% mutate(cr_ed_sum = sum(n)) %>% 
      ungroup() %>%
      mutate(cell_total = sum(n)) %>% 
      mutate(cb_ed_prop = cb_ed_sum/cell_total,
             cr_cb_prop = cr_cb_sum/cell_total,
             cr_ed_prop = cr_ed_sum/cell_total) %>%
      pivot_longer(cols = matches("cb_ed_.*|cr_cb_.*|cr_ed_.*"), names_to = c("mthd_pair", ".value"), names_pattern = "(.._..)_(.*)") %>%
      filter(ovlp == "yes") %>%
      distinct(mthd_pair, .keep_all = T)
    ) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "prop", col_var = "mthd_pair")



```

```{r, fig.width=10, fig.height=4}
map(edrops_cbender_pf100_mdata, ~.x %>% dim) %>%
  bind_rows(., .id = "donor")

```

```{r, fig.width=10, fig.height=4}
edrops_cbender_pf100_mdata[[1]] %>% count(is_cell_cr,is_cell_ed,is_cell_cb)
```

```{r}

resc_bcodes <- map(edrops_cbender_pf100_mdata, ~{
  .x %>% 
    filter(is_cell_ed == "CellEd" | is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_cr == "CellCr") %>% 
    pull(bcode) 
  })  
```

```{r}
map(resc_bcodes, length) %>% bind_rows(., .id = "donor")
```

```{r}
map(resc_bcodes, length) %>% bind_rows(., .id = "donor")
```

```{r, message=F, warning=F, eval=T}
gogo()
```
Write out barcodes to rerun souporcell including all possible barcode that are potentially actual cells

```{r}

imap(resc_bcodes, ~{
  system(paste0('mkdir data/processed/Pf/', .y, '/soup_resc'))
  
  .x %>% write(., gzfile(paste0('data/processed/Pf/', .y, '/soup_resc/barcodes.tsv.gz')))
  }) 
```


Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
gogo()

#  nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC40/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume

# nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC24,MSC68,MSC14,MSC49,MSC55}/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume

# nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC1,MSC12,MSC13,MSC25,MSC3,MSC33,MSC37,MSC38,MSC39,MSC40,MSC41,MSC45,MSC48,MSC50_MACS,MSC50_SLO,MSC51,MSC53,MSC54,MSC57,MSC60,MSC66,MSC67,MSC70}/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume

# nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```


```{r}
all_msc_bp_raw <- pmap(list(edrops_cbender_pf100_mdata, all_msc_bp_raw[names(edrops_cbender_pf100_mdata)]), ~{
  ..1 %>% 
    column_to_rownames("bcode") %>%
    # select(mthds_union) %>%
    AddMetaData(..2, .) 
    
})
```


```{r}
# all_msc_bp_raw <- pmap(list(edrops_cbender8_mdata_uni, all_msc_bp_raw[names(edrops_cbender8_mdata_uni)]), ~{
#   ..1 %>% 
#     column_to_rownames("bcode") %>%
#     select(mthds_union) %>%
#     AddMetaData(..2, .) 
#     
# })
```



```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
# all_msc_raw_mdata <- map2(all_msc_bp_raw, all_msc_bp_filt, ~.x@meta.data %>% filter(nFeature_RNA > 10) %>% rownames_to_column("bcode") %>% mutate(cr_filt = case_when(bcode %in% rownames(.y@meta.data) ~ "Yes", T ~ "No"))) 
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_raw_mdata_full <- map(all_msc_bp_raw, ~.x@meta.data %>% 
                           filter(nCount_RNA > 1) %>%
                           rownames_to_column("bcode") 
                         ) 
```

```{r}
# saveRDS(all_msc_raw_mdata_full, "data/processed/Pf/m2_all/all_msc_raw_mdata_full.RDS")
```

```{r}
# all_msc_raw_mdata_full <- readRDS("data/processed/Pf/m2_all/all_msc_raw_mdata_full.RDS")
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_raw_mdata <- map(all_msc_raw_mdata_full, ~.x %>% 
                           filter(nCount_RNA > 5) %>%
                           mutate(across(contains("is_cell"), ~replace_na(., "empty_bc")))
                         ) 
```




```{r, fig.width=15, fig.height=5}
# all_mdata_10x_resc <- map(all_msc_bp_raw, ~.x@meta.data %>% 
#                            left_join(., donor_10x_mdata_sk21_summ, by = "donor")
#                          ) 


```

```{r, fig.width=15, fig.height=5}

# mdata_10x_resc %>% 
#   pivot_longer(cols = c("Limit (no retain): 50", "CBendr", "Limit (no retain): 100"), names_to = "Cell_id_mthd", values_to = "cellNo") %>%
#   ggplot(aes(x = parasites_loaded, y = cellNo)) +
#   geom_point(size = 4) +
#   geom_line(aes(group = Cell_id_mthd), size = 1) +
#   facet_wrap(Cell_id_mthd ~ .)
```

Find gene count cut offs
```{r, fig.height= 4, fig.width= 4}
library(scales)
strn_knee_p_10_fn <- function(dset = seu_mdata, donor_nm,  ft_count = "nCount_RNA", ft_name = "Transcript", cell_method = "is_cell_cr") {
    dset %>% 
      arrange(desc(!!rlang::sym(ft_count))) %>%
      mutate(rank = row_number()) %>% 
      ggplot(., aes(y = !!rlang::sym(ft_count), x= rank, colour = !!rlang::sym(cell_method))) +
      ggrastr::rasterise(geom_point(size=0.5), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_hline(yintercept = 100) +
      geom_hline(yintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=20), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=20,face="bold"),
            legend.position = "none"
    )+ 
      annotation_logticks() +
      labs(x= 'Cell rank', y= paste0(ft_name, ' count'),
           title = donor_nm
           )
    
}


```

Find gene count cut offs
```{r, fig.height= 4, fig.width= 4, eval=T}
strn_knee_p_10_cr <- imap(all_msc_raw_mdata, ~{
  seu_mdata = .x
  don = .y
  # map2(c("nFeature_RNA", "nCount_RNA"), c("Gene", "Transcript"), ~{
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_cr")
    
})
})

strn_knee_p_10_cr
```

```{r, fig.height= 4, fig.width= 4, eval=T}
strn_knee_p_10_ed <- imap(all_msc_raw_mdata, ~{
  seu_mdata = .x
  don = .y
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_ed")
    
})
})

strn_knee_p_10_ed
```

```{r, fig.height= 4, fig.width= 4, eval=T}
strn_knee_p_10_cb <- imap(all_msc_raw_mdata, ~{
  seu_mdata = .x
  don = .y
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_cb")+
      theme(legend.position = "bottom",
            legend.text = element_text(size = 10),
            legend.title = element_text(size = 10))
    
})
})

strn_knee_p_10_cb
```

```{r, fig.height= 4, fig.width= 4, eval=T}
imap(all_msc_raw_mdata[2:4], ~{
  seu_mdata = .x
  don = .y
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_cb")+
      theme(legend.position = "bottom",
            legend.text = element_text(size = 10),
            legend.title = element_text(size = 10))
    
})
})

```

```{r, fig.height= 14, fig.width= 10, eval=T}

all_msc_raw_mdata[[2]] %>%
  filter(!is.na(is_cell_cb)) %>%
  count(is_cell_cb)

```


```{r, fig.height= 14, fig.width= 10, eval=T}
edrops_thresh_df <- edrops_thresh %>% as.data.frame() %>% rownames_to_column("donor") %>% rename("thresh" = ".")
```

```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb <- map2(all_msc_raw_mdata, edrops_thresh[names(all_msc_raw_mdata)], ~{
  .x %>% 
    filter(!is.na(is_cell_cb)) %>%
      arrange(desc(!!rlang::sym("nCount_RNA"))) %>%
      mutate(rank = min_rank(desc(nCount_RNA)),
             man_thresh = .y) %>% 
    filter(!(duplicated(rank) & is_cell_cb %in% c("empty_bc", "nonCell")))%>%
    filter(!(duplicated(rank) & is_cell_cb %in% c("nonCell")))# %>%
    # filter(!duplicated(rank))
    
}) %>%
  bind_rows(., .id = "donor")


```


```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb_don <- all_rank_ncount_cb %>% left_join(., donor_10x_mdata_sk21_summ, by = "donor") 

```


```{r, fig.height= 14, fig.width= 10, eval=T}

all_rank_ncount_cb_don %>%
  filter(donor == "MSC12") %>%
  count(is_cell_cb)

```

```{r, fig.height= 14, fig.width= 10, eval=T}

all_rank_ncount_cb_don %>%
  filter(!is.na(is_cell_cb)) %>%
  count(donor, is_cell_cb)

```

```{r, fig.height= 14, fig.width= 10, eval=T}

# all_rank_ncount_don %>%
#   filter(!is.na(is_cell_cb)) %>%
#   count(donor, is_cell_cb)

```


```{r, fig.height= 14, fig.width= 10, eval=T}
edrops_thresh %>% as.data.frame() %>% rownames_to_column("donor") %>% rename("thresh" = ".")
```

```{r, fig.height= 14, fig.width= 10, eval=T}
library(scales)

map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{all_rank_ncount_cb_don %>%
  filter(!is.na(is_cell_ed)) %>%
      ggplot(., aes(y = nCount_RNA, x= rank, colour = !!rlang::sym("is_cell_ed"))) +
      # ggrastr::rasterise(geom_point(size=0.5), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_hline(aes(yintercept = man_thresh), colour = "red") +
      # geom_hline(yintercept = 100) +
      geom_hline(yintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4) +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Cell rank', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
})
```


```{r, fig.height= 14, fig.width= 10, eval=T}
map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{all_rank_ncount_cb_don %>%
  filter(!is.na(is_cell_cb)) %>%
      ggplot(., aes(y = nCount_RNA, x= rank, colour = !!rlang::sym("is_cell_cb"))) +
      # ggrastr::rasterise(geom_point(size=0.5), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_hline(aes(yintercept = man_thresh), colour = "red") +
      # geom_hline(yintercept = 100) +
      # geom_hline(yintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4) +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Cell rank', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
})
```


```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_gn_umi_bp_don <- map(all_msc_raw_mdata, ~{
  .x #%>% 
    # filter(is_cell_cb != "empty_bc") %>%
    # filter(nFeature_RNA < 300)
    
}) %>%
  bind_rows(., .id = "donor") %>% 
  left_join(., donor_10x_mdata_sk21_summ, by = "donor") 


```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=14}

map(c("is_cell_cb", "is_cell_ed", "is_cell_cr")[1],  ~{ 
  mthd = .x
  
  map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload")[1], ~{
    all_rank_gn_umi_bp_don %>%
    filter(nFeature_RNA < 300) %>%
    filter(!!rlang::sym(mthd) != "empty_bc" | !!rlang::sym(mthd) != "nonCell") %>%
    ggplot(aes(x = nFeature_RNA, y= nCount_RNA, color = !!rlang::sym(mthd))) + 
    ggrastr::rasterise(geom_point(size=0.01), dpi=50) + ## Reducing the value increases the speed but makes the plot more blurry
    # geom_point(size = 0.1) + 
    # scale_color_manual(values = sp_col_new) +
    geom_hline(yintercept = 100) + 
    geom_vline(xintercept = 100) +
    geom_vline(xintercept = 80) +
    # labs(title = .y) +
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4) +
    theme_classic()+
                                     theme(legend.position = "bottom",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(),
                                           axis.text = element_blank()
                                           ) +
      labs(title = paste0(str_remove(mthd, "is_cell_"), "_", .x))
  })
})
```



```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
# Bad_cells
# system.time(
  map(c("is_cell_cb", "is_cell_ed", "is_cell_cr")[1],  ~{ 
  mthd = .x
  
  map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload")[1], ~{
    all_rank_gn_umi_bp_don %>%
    filter(nFeature_RNA < 300) %>%
    # filter(!!rlang::sym(.x) != "empty_bc") %>%
    filter(!!rlang::sym(mthd) %in% c("empty_bc", "nonCell")) %>%
    ggplot(aes(x = nFeature_RNA, y= nCount_RNA, color = !!rlang::sym(mthd))) +
      ggrastr::rasterise(geom_point(size=0.1), dpi=50) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.1) + ## Reducing the value increases the speed but makes the plot more blurry
    geom_hline(yintercept = 100) + 
    geom_hline(yintercept = 50) +
    geom_vline(xintercept = 100) +
    geom_vline(xintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4) +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )+
      labs(title = paste0(str_remove(mthd, "is_cell_"), "_", .x))
})
})
# )
```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
# Good_cells
# system.time(
  map(c("is_cell_cb", "is_cell_ed", "is_cell_cr"),  ~{ 
  mthd = .x
  
  map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{
    all_rank_gn_umi_bp_don %>%
    filter(nFeature_RNA < 300) %>%
    # filter(!!rlang::sym(.x) != "empty_bc") %>%
    filter(!(!!rlang::sym(mthd) %in% c("empty_bc", "nonCell"))) %>%
    ggplot(aes(x = nFeature_RNA, y= nCount_RNA, color = !!rlang::sym(mthd))) +
      # ggrastr::rasterise(geom_point(size=0.1), dpi=50) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_point(size=0.1) + ## Reducing the value increases the speed but makes the plot more blurry
    geom_hline(yintercept = 100) + 
    geom_hline(yintercept = 50) +
    geom_vline(xintercept = 100) +
    geom_vline(xintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4) +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )+
      labs(title = paste0(str_remove(mthd, "is_cell_"), "_", .x))
})
})
# )
```



```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}

map(c("is_cell_cb", "is_cell_ed", "is_cell_cr"), 
    ~ all_rank_gn_umi_bp_don %>%
    # filter(nFeature_RNA < 300) %>%
    # filter(!!rlang::sym(.x) != "empty_bc") %>%
    ggplot(aes(x = nFeature_RNA, y= nCount_RNA, color = !!rlang::sym(.x))) + 
      ggrastr::rasterise(geom_point(size=0.1), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry
    geom_hline(yintercept = 100) + 
    geom_hline(yintercept = 50) +
    geom_vline(xintercept = 100) +
    geom_vline(xintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(frac_load, donor), ncol = 4) +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
)
```


```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=14}

map(c("is_cell_cb", "is_cell_ed", "is_cell_cr"), 
    ~ all_rank_gn_umi_bp_don %>%
    # filter(nFeature_RNA < 300) %>%
    # filter(!!rlang::sym(.x) != "empty_bc") %>%
    ggplot(aes( y= nCount_RNA/1000, x = !!rlang::sym(.x), color = !!rlang::sym(.x))) + 
      # ggrastr::rasterise(geom_point(size=0.1), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry+
    # geom_sina()  +
      geom_boxplot()  +
    geom_hline(yintercept = 20000/1000) + 
    geom_hline(yintercept = 100/1000) + 
      theme_classic() +
     # scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
     #          labels = trans_format("log10", math_format(10^.x)))+
     # scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
     #          labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(frac_load, donor), ncol = 4, scales = "free_y") +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
)
```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=14}

map(c("is_cell_cb", "is_cell_ed", "is_cell_cr"), 
    ~ all_rank_gn_umi_bp_don %>%
    # filter(nFeature_RNA < 300) %>%
    # filter(!!rlang::sym(.x) != "empty_bc") %>%
    ggplot(aes( y= nCount_RNA, x = !!rlang::sym(.x), color = !!rlang::sym(.x))) + 
      # ggrastr::rasterise(geom_point(size=0.1), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry+
    # geom_sina()  +
      geom_boxplot()  +
    geom_hline(yintercept = 20000) + 
    geom_hline(yintercept = 100) + 
      theme_classic() +
     # scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
     #          labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(frac_load, donor), ncol = 4, scales = "free_y") +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
)
```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=14}

  map(c("is_cell_cb", "is_cell_ed", "is_cell_cr"),  ~{ 
  mthd = .x
  
  map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{
    all_rank_gn_umi_bp_don %>%
    # filter(nFeature_RNA < 300) %>%
    # filter(!!rlang::sym(.x) != "empty_bc") %>%
    ggplot(aes( y= nCount_RNA, x = !!rlang::sym(mthd), color = !!rlang::sym(mthd))) + 
      # ggrastr::rasterise(geom_point(size=0.1), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry+
    # geom_sina()  +
      geom_violin()  +
    geom_hline(yintercept = 20000) + 
    geom_hline(yintercept = 100) + 
      theme_classic() +
     # scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
     #          labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4, scales = "free_y") +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
  })
  })
```


```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
 map(c("p_map_cat", "peak", "frac_load", "overload"), ~{
    all_rank_gn_umi_bp_don %>% 
  # filter(is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>% 
  filter(is_cell_cb %in% c("CellCb.8") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>% 
  pivot_longer(cols = c(is_cell_cb, is_cell_cr, is_cell_ed), names_to = "Mthds", values_to = "CellCat") %>% 
  filter(!(CellCat %in% c("empty_bc", "nonCell")))  %>%
  mutate(across(Mthds, ~factor(., levels = c("is_cell_cb", "is_cell_ed", "is_cell_cr")))) %>%
    ggplot(aes( y= nCount_RNA, x = Mthds, color = Mthds)) + 
      # ggrastr::rasterise(geom_point(size=0.1), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry+
    # geom_sina()  +
      geom_violin()  +
    geom_hline(yintercept = 20000) + 
    geom_hline(yintercept = 100) + 
      theme_classic() +
     # scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
     #          labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(!!rlang::sym(.x), donor), ncol = 4, scales = "free_y") +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=12), 
            axis.text.x=element_text(angle=25, hjust = 1), 
            axis.title=element_text(size=16,face="bold"), 
            legend.text = element_text(size=16), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Cell Method', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
 })
```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
all_rank_gn_umi_bp_don %>% filter(is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>% pivot_longer(cols = c(is_cell_cb, is_cell_cr, is_cell_ed), names_to = "Mthds", values_to = "CellCat") %>% filter(!(CellCat %in% c("empty_bc", "nonCell")))  %>%
  mutate(across(Mthds, ~factor(., levels = c("is_cell_cb", "is_cell_ed", "is_cell_cr")))) %>%
    ggplot(aes( y= nFeature_RNA, x = Mthds, color = Mthds)) + 
      # ggrastr::rasterise(geom_point(size=0.1), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      # geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry+
    # geom_sina()  +
      geom_violin()  +
    geom_hline(yintercept = 4000) + 
    geom_hline(yintercept = 100) + 
      theme_classic() +
     # scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
     #          labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   facet_nested_wrap(vars(frac_load, donor), ncol = 4, scales = "free_y") +
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=12), 
            axis.text.x=element_text(angle=25, hjust = 1), 
            axis.title=element_text(size=16,face="bold"), 
            legend.text = element_text(size=16), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Cell Method', y= paste0("Gene", ' count'),
           # title = donor_nm
           )
```


```{r}
# Convert matrix to raw counts
# all_msc_counts <- map(all_msc_bp_raw, ~as(.x[["RNA"]]$counts, Class = "dgCMatrix"))
```

```{r}
# Run emptyDrops with the current limit
# e.out <- map(all_msc_counts, ~emptyDrops(.x, lower=100, test.ambient=TRUE))
# e.out <- map(all_msc_counts, ~emptyDrops(.x, lower=50, test.ambient=TRUE))


```


```{r}
# for(i in 1:length(e.out)) {
#   e.out[[i]]["Is_Cell"] <- ifelse(e.out[[i]][,"FDR"] <= 0.01, "Cell", "Non-Cell")
# }

```




```{r}
# all_msc_bp_edrops50 <- map(all_msc_bp_raw, ~subset(.x, subset = Is_Cell_100 == "Cell"))
all_msc_bp_edrops50 <- map(all_msc_bp_raw, ~subset(.x, subset = is_cell_ed == "CellEd"))
all_msc_bp_edrops50
```

```{r}
all_msc_bp_cbendr <- map(all_msc_bp_raw, ~subset(.x, subset = is_cell_cb == "CellCb.5" | is_cell_cb == "CellCb.8"))
```

```{r}
all_msc_bp_cbendr 
```

```{r}
map(all_msc_bp_edrops50, ~{table(.x@meta.data$nFeature > 100)})
```

```{r}
map(all_msc_bp_cbendr, ~{table(.x@meta.data$nFeature > 100)})
```


```{r, eval=T}
cell_counts <- pmap(list(all_msc_bp_raw, all_msc_bp_filt, all_msc_bp_edrops50, all_msc_bp_cbendr), ~{
  data.frame("raw" = dim(..1)[2], "cr_filt" = dim(..2)[2], "edrops50" = dim(..3)[2], "cbendr" = dim(..4)[2])
  
}) %>%
  bind_rows(., .id = "donor")

```


```{r, eval=T}
cell_counts
```


```{r, eval=T}
rm(all_msc_bp_raw)
```


```{r, message=F, warning=F, eval=F}
# Convert to dense object - might be necessary in future instances
# all_msc <- all_msc_bp %>%
#   map(., ~{
#     .x[["RNA"]]$counts <- as(object = .x[["RNA"]]$counts, Class = "dgCMatrix")
#     return(.x)
#   })

```


```{r, eval=T}
##write out for normalization processing in farm

for (i in 1:length(sample_name)) {
## Create new directories
  system(paste0('mkdir -p data/processed/Pf/', sample_name[i], '/seu_obj'))
  # system(paste0('rm data/processed/Pf/', sample_name[i], '/seu_obj/*'))
}


```

Remove cells with less than 100 pf reads
```{r, eval=T}
##write out for normalization processing in farm
# all_msc_bp_filt_pf100
all_msc_bp_filt_pf100 <- map(all_msc_bp_filt, ~subset(.x, subset = pf_umi_count >= 100))

```


FARM CODE

```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
### filt
imap(all_msc_bp_filt_pf100, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_cr.RDS")))

### cellbender
imap(all_msc_bp_cbendr, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_cbendr.RDS")))
all_msc_bp_cbendr
### emptydrops
imap(all_msc_bp_edrops50, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_edrops50.RDS")))
```


!!!NB - FARM CODE - DONT DELETE

```{r, eval=T}
gogo()
## Run standard normalisation on farm like below
## !!NB - reducing the number of hypervariable genes results in scattered and disjointed clusters. higher hvg number better

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/seu_obj/bpc_seu_cr.RDS" --normlzn "norm" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --ncores "3" --mem "10 GB" --resume

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_150.RDS" --normlzn "norm" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --resume 

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/seu_obj/bpc_seu_edrops50.RDS" --normlzn "norm" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --ncores "3" --mem "10 GB" --resume

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/seu_obj/bpc_seu_cbendr.RDS" --normlzn "norm" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --ncores "3" --mem "10 GB" --resume
```

SCT for the evaluation of raw counts
```{r, eval=T}
gogo()
## Run standard normalisation on farm like below
## !!NB - reducing the number of hypervariable genes results in scattered and disjointed clusters. higher hvg number better

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sctnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14,MSC49,MSC70}/seu_obj/bpc_seu_cr.RDS" --normlzn "sct" --umap_name "UMAP_" --hvg 400 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --resume 
# 
# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sctnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14,MSC49,MSC70}/seu_obj/bpc_seu_raw_150.RDS" --normlzn "sct" --umap_name "UMAP_" --hvg 400 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --resume
#  
# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sctnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/seu_obj/bpc_seu_edrops50.RDS" --normlzn "sct" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --ncores "3" --mem "10 GB" --resume

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sctnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/*/seu_obj/bpc_seu_cbendr.RDS" --normlzn "sct" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --ncores "3" --mem "10 GB" --resume
```

