---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(scran)
  library(SingleR)
  library(scDblFinder)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(dplyr::desc)
  conflicts_prefer(dplyr::count)
})
```



# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/project_plotting_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")

```

#### Run emptydrops on all 
```{r}
## NOTE - USE raw_feature_bc_matrix.h5 and not the folder raw_feature_bc_matrix/ since nextflow can't expand to all the donor paths when using the folder after specifying 5736STDY*
##!!! NOTE The hdf5 matrix makes the running of emptydrops very slow so need to convert to dgCMatrix - https://github.com/MarioniLab/DropletUtils/issues/55

## Run emptydrops on all 
# nextflow run ../mali22_code_lnk/nf_scripts/edrops.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/edrops.R" --input_raw_mtx "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/raw_feature_bc_matrix.h5" --o_path "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --retain_nm "wo_retain,w_retain" --retain_val "Inf,NULL" --limt "20,50,100,150,200" --fdr_l "0.01" -resume


```

###################START- RAW READING#############

```{r, eval=T}
pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
```

## Genotype clustering for MSC

initialize variable for the sample names

```{r}
# # select sample names
# # slct_sample_names <-c("MSC48", "MSC68", "MSC14", "MSC49", "MSC60", "MSC66", "MSC67", "MSC70", "MSC24", "MSC33", "MSC40", "MSC54", "MSC55") %>% str_sort(., numeric = T)
# 
# slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
# ## get relevant sample names
# pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)
# 
# ##Sample ID
# sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
# smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
# 
# ##Sample IDs
# source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```


```{r}

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
# slct_sample_names <-c("MSC68", "MSC14", "MSC49", "MSC55") %>% str_sort(., numeric = T)


# ## poor quality samples
sample_rm <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39")
sample_mxd <- pf_solo_mixed_mdata_decode %>% filter(mixed_infection2 == "mixed") %>% pull(sample_nm)
# 
# # slct_sample_names[-c(sample_rm)]
slct_sample_names <- slct_sample_names[!(slct_sample_names %in% c(sample_rm, sample_mxd))]

## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```


```{r}
# ## limit for different donors
# edrops_thresh <- rep(50, length(pf_solo_mixed_mdata_decode$sample_nm)) %>% set_names(pf_solo_mixed_mdata_decode$sample_nm)
# edrops_thresh[["MSC1"]] <- 150
# edrops_thresh[["MSC13"]] <- 100
# edrops_thresh[["MSC14"]] <- 100
# edrops_thresh[["MSC24"]] <- 200
# edrops_thresh[["MSC37"]] <- 200
# edrops_thresh[["MSC38"]] <- 200
# edrops_thresh[["MSC53"]] <- 150
# edrops_thresh[["MSC57"]] <- 150
# edrops_thresh[["MSC50_SLO"]] <- 100
```


```{r}
## read manual thresholds
edrops_thresh <- readRDS("data/processed/Pf/m2_all/man_thresh.RDS")
```

```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_cbender_mdata <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  # set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender_yascpCr/cb_mdata.*'))) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender/cb_mdata.*'))) %>% 
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(read_csv) 
```

```{r, message=F, warning=F, eval=T}
## Just work with files that have cellbender output
sample_name <- names(all_cbender_mdata)
```

```{r, message=F, warning=F, eval=T}
# ##Read BPcells matrices
# all_cbender_mdata <- list.files("data/processed/Pf", pattern = "cb_mdata.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/cbender/cb_mdata.*'))) %>% 
#   .[sample_name] %>%
#   .[!is.na(.)] %>%
#   map(read_csv) 
```

```{r}
processed_data_noretain_50 <- readRDS("data/processed/Pf/m2_all/processed_data_noretain_50.RDS") %>% .[sample_name]
```

```{r, fig.width=15, fig.height=5}
donor_10x_mdata_sk21_summ <- readRDS("data/processed/Pf/m2_all/donor_10x_mdata_sk21_summ.RDS") %>%
  filter(donor %in% sample_name) 

```


```{r, fig.width=15, fig.height=5}

donor_10x_mdata_sk21_summ %>% count(frac_load, overload, frac_over, targeted_cells)
```


```{r}
all_msc_bp_filt_noHsGns_mdata <- readRDS("data/processed/Pf/m2_all/all_msc_bp_filt_noHsGns_mdata.RDS") %>% .[sample_name]
```

```{r}
all_msc_bp_raw_noHsGns_mdata <- readRDS("data/processed/Pf/m2_all/all_msc_bp_raw_noHsGns_mdata.RDS") %>% .[sample_name]
```


Save paths to raw MSC files in list - also needed for soupX below
```{r}
imap(all_msc_bp_filt_noHsGns_mdata, ~{
  # .x@meta.data %>%
  .x %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point() +
    geom_hline(yintercept = log10(400)) +
    geom_vline(xintercept = log10(400)) +
    labs(title = .y)
})
```

```{r}
imap(all_msc_bp_raw_noHsGns_mdata, ~{
  # .x@meta.data %>%
  .x %>%
    filter(pf_umi_count > 50) %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) +
    labs(title = .y)
})
```

###################END - RAW READING##############

Save paths to raw MSC files in list - also needed for soupX below
```{r}
# all_msc_raw_paths <- list.files("data/raw/Pf", pattern = "raw_feature_bc_matrix$", full.names = T, recursive = T, include.dirs = T) %>% 
#   str_sort(., numeric = T) %>%
#   .[str_detect(., 'DB52e_star')]
```


```{r}
edrops_cbender_mdata <- pmap(list(processed_data_noretain_50, all_cbender_mdata, all_msc_bp_filt_noHsGns_mdata, all_msc_bp_raw_noHsGns_mdata), ~{
  ..1 %>% 
    data.frame() %>% 
    mutate(is_cell = ifelse(FDR <= 0.01, "CellEd", "nonCell")) %>%
    full_join(., ..2[,c("barcode", "cell_probability")] %>%
                mutate(is_cell = case_when(cell_probability > 0.8 ~ "CellCb.8",
                                           cell_probability > 0.5 ~ "CellCb.5",
                                           T ~ "nonCell")),
              by = c("Barcode" = "barcode"),
              suffix = c("_ed", "_cb")) %>%
    full_join(#..3@meta.data %>% 
              ..3 %>% 
                rownames_to_column("bcode") %>% 
                mutate("is_cell_cr" = "CellCr") %>% 
                .[,c("bcode", "is_cell_cr")],
              .,
              by = c("bcode" = "Barcode")) %>%
    left_join(., ..4[, c("nFeature_RNA", "nCount_RNA", "pf_umi_count", "hs_umi_count", "pf_gn_count", "hs_gn_count")] %>% 
                rownames_to_column("bcode"),
              by = c("bcode" = "bcode"))
    
})
```


```{r}
saveRDS(edrops_cbender_mdata, "data/processed/Pf/m2_all/edrops_cbender_mdata.RDS")
```


```{r}
edrops_cbender_mdata <- readRDS("data/processed/Pf/m2_all/edrops_cbender_mdata.RDS")
```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
bind_rows(edrops_cbender_mdata, .id = "donor") %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(50)) +
    geom_vline(xintercept = log10(100)) 

```


```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
    bind_rows(edrops_cbender_mdata, .id = "donor") %>%
    ggplot(aes(x = log10(pf_gn_count), y = log10(hs_gn_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) 

```



```{r}
edrops_cbender_mdata_uni <- map(edrops_cbender_mdata, ~.x%>%
    filter(!(is.na(is_cell_cr) & is_cell_cb == "nonCell" & is.na(is_cell_ed))) %>%
    mutate(mthds_union = case_when(is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb") & is_cell_ed == "CellEd" ~ "CrCbEd",
                                   is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb") & is.na(is_cell_ed) ~ "CrCb",
                                   is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb)) & is_cell_ed == "CellEd" ~ "CrEd",
                                   is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb") & is_cell_ed == "CellEd" ~ "CbEd",
                                   is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb)) & is.na(is_cell_ed) ~ "Cr",
                                   is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb") & is.na(is_cell_ed) ~ "Cb",
                                   is.na(is_cell_cr) & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb)) & is_cell_ed == "CellEd" ~ "Ed"),
           across(mthds_union, ~factor(., levels = c("CrCbEd", "CrCb", "CrEd", "CbEd", "Cr", "Cb", "Ed"))))
)

edrops_cbender8_mdata_uni <- map(edrops_cbender_mdata, ~.x%>% 
    filter(!(is.na(is_cell_cr) & (is_cell_cb == "nonCell" | is_cell_cb == "CellCb.5") & is.na(is_cell_ed))) %>% 
    mutate(mthds_union = case_when(is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb.8") & is_cell_ed == "CellEd" ~ "CrCbEd",
                                   is_cell_cr == "CellCr" & str_detect(is_cell_cb, "CellCb.8") & is.na(is_cell_ed) ~ "CrCb",
                                   is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb) | is_cell_cb == "CellCb.5") & is_cell_ed == "CellEd" ~ "CrEd",
                                   is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb.8") & is_cell_ed == "CellEd" ~ "CbEd",
                                   is_cell_cr == "CellCr" & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb) | is_cell_cb == "CellCb.5") & is.na(is_cell_ed) ~ "Cr",
                                   is.na(is_cell_cr) & str_detect(is_cell_cb, "CellCb.8") & is.na(is_cell_ed) ~ "Cb",
                                   is.na(is_cell_cr) & (str_detect(is_cell_cb, "nonCell") | is.na(is_cell_cb) | is_cell_cb == "CellCb.5") & is_cell_ed == "CellEd" ~ "Ed"),
           across(mthds_union, ~factor(., levels = c("CrCbEd", "CrCb", "CrEd", "CbEd", "Cr", "Cb", "Ed"))))
) 

```

```{r}
map(edrops_cbender_mdata_uni, ~.x %>% count(is_cell_cr, is_cell_ed, is_cell_cb,mthds_union))
map(edrops_cbender8_mdata_uni, ~.x %>% dplyr::count(is_cell_cr, is_cell_ed, is_cell_cb,mthds_union))

```

```{r}
saveRDS(edrops_cbender8_mdata_uni, "data/processed/Pf/m2_all/edrops_cbender8_mdata_uni.RDS")
saveRDS(edrops_cbender_mdata_uni, "data/processed/Pf/m2_all/edrops_cbender_mdata_uni.RDS")

```

```{r, fig.width=10, fig.height=12}

map(edrops_cbender8_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, fig.width=10, fig.height=12}

map(edrops_cbender_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
map(c("gn", "umi"), ~{bind_rows(edrops_cbender8_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(!!rlang::sym(paste0('pf_',.x,'_count'))), y = log10(!!rlang::sym(paste0('hs_',.x,'_count'))), colour = mthds_union)) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    geom_vline(xintercept = log10(100)) 
})

```

```{r}
edrops_cbender_pf100_mdata <- map(edrops_cbender_mdata, ~.x %>% filter(pf_umi_count >= 100))
```

```{r}
# saveRDS(edrops_cbender_pf100_mdata, "data/processed/Pf/m2_all/edrops_cbender_pf100_mdata.RDS")
```

```{r}
edrops_cbender8_pf100_mdata_uni <- map(edrops_cbender8_mdata_uni, ~.x %>% filter(pf_umi_count >= 100))
```

```{r}
# saveRDS(edrops_cbender8_pf100_mdata_uni, "data/processed/Pf/m2_all/edrops_cbender8_pf100_mdata_uni.RDS")
```

```{r, message=F, warning=F, eval=T, fig.height=14, fig.width=12}
bind_rows(edrops_cbender8_pf100_mdata_uni, .id = "donor") %>%
    ggplot(aes(x = log10(pf_umi_count), y = log10(hs_umi_count))) +
    geom_point(size = 0.01) +
    geom_density_2d() +
    facet_wrap(donor ~ .,  ncol = 5) +
    geom_hline(yintercept = log10(100)) +
    # geom_vline(xintercept = log10(50)) +
    geom_vline(xintercept = log10(100)) 

```

```{r, fig.width=10, fig.height=12}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union, fill = mthds_union)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.35))) +
  facet_wrap(. ~ donor, ncol = 7, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r}
edrops_cbender_pf100_mdata_uni <- map(edrops_cbender_mdata_uni, ~.x %>% filter(pf_umi_count >= 100))
```

```{r, fig.width=10, fig.height=5}
map(edrops_cbender_pf100_mdata_uni, ~.x %>% 
      filter(mthds_union != "Ed") %>%
      mutate("mthds_union_CbCr" = str_remove(mthds_union, "Ed")) %>% 
      count(mthds_union_CbCr)) %>%
  bind_rows(., .id = "donor") %>%
  ggplot(aes(y = n/1000, x = mthds_union_CbCr, fill = mthds_union_CbCr)) +
  geom_col() +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  geom_text(aes(label=n), angle=90, hjust=-0.1, vjust= 0.5, position = position_dodge(width = .9), size = 3.8, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.5))) +
  facet_wrap(. ~ donor, ncol = 10, scales = "free_y") +
  # facet_wrap(. ~ donor, ncol = 7) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 13),
        legend.position = "bottom")


```


```{r, fig.width=10, fig.height=4}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union) %>% mutate("cb_ed_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CbEd"), "yes", "no")) %>% group_by(cb_ed_ovlp) %>% mutate(prop_ovlp = sum(n)) %>% distinct(cb_ed_ovlp, .keep_all = T) %>% ungroup() %>% mutate(prop = prop_ovlp/sum(prop_ovlp)) %>% filter(cb_ed_ovlp == "yes")) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "prop", col_var = "mthds_union") +
  theme(legend.position = "none")


```

```{r, fig.width=10, fig.height=5}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union)) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "n", col_var = "mthds_union")

```

```{r, fig.width=10, fig.height=5}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% count(mthds_union) %>% mutate(prop = n/sum(n))) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "prop", col_var = "mthds_union")


```

```{r, fig.width=10, fig.height=4}
map(edrops_cbender8_pf100_mdata_uni, ~.x %>% 
      count(mthds_union) %>% 
      mutate("cb_ed_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CbEd"), "yes", "no"),
             "cr_cb_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CrCb"), "yes", "no"),
             "cr_ed_ovlp" = ifelse(mthds_union %in% c("CrCbEd", "CrEd"), "yes", "no")) %>% 
      group_by(cb_ed_ovlp) %>% mutate(cb_ed_sum = sum(n)) %>% 
      group_by(cr_cb_ovlp) %>% mutate(cr_cb_sum = sum(n)) %>% 
      group_by(cr_ed_ovlp) %>% mutate(cr_ed_sum = sum(n)) %>% 
      ungroup() %>%
      mutate(cell_total = sum(n)) %>% 
      mutate(cb_ed_prop = cb_ed_sum/cell_total,
             cr_cb_prop = cr_cb_sum/cell_total,
             cr_ed_prop = cr_ed_sum/cell_total) %>%
      pivot_longer(cols = matches("cb_ed_.*|cr_cb_.*|cr_ed_.*"), names_to = c("mthd_pair", ".value"), names_pattern = "(.._..)_(.*)") %>%
      filter(ovlp == "yes") %>%
      distinct(mthd_pair, .keep_all = T)
    ) %>%
  bind_rows(., .id = "donor") %>%
  methds_ln_plt(., x_var = "donor", y_var = "prop", col_var = "mthd_pair")



```

```{r, fig.width=10, fig.height=4}
map(edrops_cbender_pf100_mdata, ~.x %>% dim) %>%
  bind_rows(., .id = "donor")

```

```{r, fig.width=10, fig.height=4}
edrops_cbender_pf100_mdata[[1]] %>% count(is_cell_cr,is_cell_ed,is_cell_cb)
```

```{r}

resc_bcodes <- map(edrops_cbender_pf100_mdata, ~{
  .x %>% 
    filter(is_cell_ed == "CellEd" | is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_cr == "CellCr") %>% 
    pull(bcode) 
  })  
```

```{r}
map(resc_bcodes, length) %>% bind_rows(., .id = "donor")
```

```{r}
map(resc_bcodes, length) %>% bind_rows(., .id = "donor")
```

```{r, message=F, warning=F, eval=T}
# gogo()
```
Write out barcodes to rerun souporcell including all possible barcode that are potentially actual cells

```{r}

imap(resc_bcodes, ~{
  system(paste0('mkdir data/processed/Pf/', .y, '/soup_resc'))
  
  .x %>% write(., gzfile(paste0('data/processed/Pf/', .y, '/soup_resc/barcodes.tsv.gz')))
  }) 
```


Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
# gogo()

#  nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC40/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume

# nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC24,MSC68,MSC14,MSC49,MSC55}/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume

# nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC1,MSC12,MSC13,MSC25,MSC3,MSC33,MSC37,MSC38,MSC39,MSC40,MSC41,MSC45,MSC48,MSC50_MACS,MSC50_SLO,MSC51,MSC53,MSC54,MSC57,MSC60,MSC66,MSC67,MSC70}/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume

# nextflow run ../mali22_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/soup_resc/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc_resc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```


```{r}
# Add emptydrops metadata
all_msc_bp_raw_noHsGns_mdata <- pmap(list(edrops_cbender_pf100_mdata, all_msc_bp_raw_noHsGns_mdata[names(edrops_cbender_pf100_mdata)]), ~{
  ..1 %>% 
    select(bcode, is_cell_cr, FDR, is_cell_ed, cell_probability, is_cell_cb) %>%
    full_join(..2 %>% rownames_to_column("bcode"), by = "bcode") 
    
})
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_raw_mdata <- map(all_msc_bp_raw_noHsGns_mdata, ~.x %>% 
                           filter(nCount_RNA > 1) %>%
                           column_to_rownames("bcode")  %>%
                           mutate(across(contains("is_cell"), ~replace_na(., "empty_bc")))
                         ) 
```



```{r, message=F, warning=F, eval=T}
## save raw metadata
saveRDS(all_msc_raw_mdata, "data/processed/Pf/m2_all/all_msc_raw_mdata.RDS")

```


Find gene count cut offs
```{r, fig.height= 4, fig.width= 4}
library(scales)
strn_knee_p_10_fn <- function(dset = seu_mdata, donor_nm,  ft_count = "nCount_RNA", ft_name = "Transcript", cell_method = "is_cell_cr") {
    dset %>% 
      mutate(rank = rank(desc(!!rlang::sym(ft_count)), ties.method = "average")) %>% 
      filter(!duplicated(rank)) %>%
      ggplot(., aes(y = !!rlang::sym(ft_count), x= rank, colour = !!rlang::sym(cell_method))) +
      ggrastr::rasterise(geom_point(size=0.5), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_hline(yintercept = 100) +
      geom_hline(yintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     # scale_color_manual(name= 'Strain', values = strain_cols_n) +
      theme(axis.text=element_text(size=20), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=20,face="bold"),
            legend.position = "none"
    )+ 
      annotation_logticks() +
      labs(x= 'Cell rank', y= paste0(ft_name, ' count'),
           title = donor_nm
           )
    
}


```



```{r, fig.height= 4, fig.width= 4, eval=T, fig.width=6, fig.height=4}
map(edrops_cbender_mdata_uni, ~.x %>% 
  filter(pf_umi_count < 1000) %>%
  ggplot(aes(x = log10(hs_umi_count), y = log10(pf_umi_count), color= is_cell_cb)) +
  geom_point(size = 0.01) +
  geom_hline(yintercept = c(log10(100),log10(250)))
)

```

```{r, fig.height= 4, fig.width= 4, eval=T, fig.width=6, fig.height=4}
map(all_msc_raw_mdata, ~.x %>% 
  filter(pf_umi_count < 1000) %>%
  ggplot(aes(x = is_cell_cb, y = pf_umi_count)) +
  geom_sina(size = 0.01) +
  geom_hline(yintercept = c(100,250))
)

```


```{r, fig.height= 4, fig.width= 4, eval=T, fig.width=15, fig.height=8}
map(all_msc_raw_mdata, ~.x %>% 
  filter(nCount_RNA < 1000) %>%
  ggplot(aes(x = is_cell_cb, y = nCount_RNA)) +
  geom_sina(size = 0.01) +
  geom_hline(yintercept = 250)
)

```



Find gene count cut offs
```{r, fig.height= 4, fig.width= 4, eval=T}
strn_knee_p_10_cr <- imap(all_msc_raw_mdata, ~{
  seu_mdata = .x
  don = .y
  # map2(c("nFeature_RNA", "nCount_RNA"), c("Gene", "Transcript"), ~{
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_cr")
    
})
})

strn_knee_p_10_cr
```

```{r, fig.height= 4, fig.width= 4, eval=T}
strn_knee_p_10_ed <- imap(all_msc_raw_mdata, ~{
  seu_mdata = .x
  don = .y
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_ed")
    
})
})

strn_knee_p_10_ed
```

```{r, fig.height= 4, fig.width= 4, eval=T}
strn_knee_p_10_cb <- imap(all_msc_raw_mdata, ~{
  seu_mdata = .x
  don = .y
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_cb")+
      theme(legend.position = "bottom",
            legend.text = element_text(size = 10),
            legend.title = element_text(size = 10))
    
})
})

strn_knee_p_10_cb
```

```{r, fig.height= 4, fig.width= 4, eval=T}
imap(all_msc_raw_mdata[2:4], ~{
  seu_mdata = .x
  don = .y
  map2(c("nCount_RNA"), c("Transcript"), ~{
    strn_knee_p_10_fn(dset = seu_mdata, donor_nm = don, ft_count = .x, ft_name = .y, cell_method = "is_cell_cb")+
      theme(legend.position = "bottom",
            legend.text = element_text(size = 10),
            legend.title = element_text(size = 10))
    
})
})

```

```{r, fig.height= 14, fig.width= 10, eval=T}

all_msc_raw_mdata[[2]] %>%
  filter(!is.na(is_cell_cb)) %>%
  count(is_cell_cb)

```


```{r, fig.height= 14, fig.width= 10, eval=T}
edrops_thresh_df <- edrops_thresh %>% as.data.frame() %>% rownames_to_column("donor") %>% rename("thresh" = ".")
```


```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb <- map2(all_msc_raw_mdata, edrops_thresh[names(all_msc_raw_mdata)], ~{
  .x %>% 
    filter(!is.na(is_cell_cb)) %>%
    mutate(rank = rank(desc(nCount_RNA), ties.method = "average"),
           man_thresh = .y) %>% 
    filter(!(duplicated(rank) & is_cell_cb %in% c("empty_bc", "nonCell")))
}) %>%
  bind_rows(., .id = "donor")


```



```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb_don <- all_rank_ncount_cb %>% left_join(., donor_10x_mdata_sk21_summ, by = "donor") 

```

```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_ncount_cb_don <- all_rank_ncount_cb_don %>% mutate(is_cell_cb2 = ifelse(is_cell_cb %in% c("CellCb.5", "CellCb.8"), "CellCb", is_cell_cb))
```


```{r, message=F, warning=F, eval=T}
## save raw metadata
saveRDS(all_rank_ncount_cb_don, "data/processed/Pf/m2_all/all_rank_ncount_cb_don.RDS")
```


```{r, fig.height= 14, fig.width= 10, eval=T}

all_rank_ncount_cb_don %>%
  filter(donor == "MSC12") %>%
  count(is_cell_cb)

```

```{r, fig.height= 14, fig.width= 10, eval=T}

all_rank_ncount_cb_don %>%
  filter(!is.na(is_cell_cb)) %>%
  count(donor, is_cell_cb)

```


```{r, fig.height= 14, fig.width= 10, eval=T}
edrops_thresh %>% as.data.frame() %>% rownames_to_column("donor") %>% rename("thresh" = ".")
```


```{r, fig.height= 14, fig.width= 10, eval=T}
kneeplt_facet_fn <- function(ncount_tbl, cell_call_methd, donor_group, numbr_of_cols = 4, mthd_col = cell_call_mthd_col){ 

  ncount_tbl %>%
  filter(!is.na(!!rlang::sym(cell_call_methd))) %>%
      ggplot(., aes(y = nCount_RNA, x= rank, colour = !!rlang::sym(cell_call_methd))) +
      # ggrastr::rasterise(geom_point(size=0.5), dpi=100) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_point(size=0.5) + ## Reducing the value increases the speed but makes the plot more blurry
      geom_hline(aes(yintercept = man_thresh), colour = "red") +
      # geom_hline(yintercept = 100) +
      geom_hline(yintercept = 50) +
      theme_classic() +
     scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
     scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+
   # facet_nested_wrap(vars(!!rlang::sym(donor_group), donor), ncol = 4) +
   facet_nested_wrap(vars(!!!rlang::syms(unique(c(donor_group, "donor")))), ncol = numbr_of_cols) +
   scale_color_manual(name= 'CellCall', values = mthd_col) +
      theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Cell rank', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )
}
```

```{r, fig.height= 10, fig.width= 10, eval=T}
map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{
  kneeplt_facet_fn(ncount_tbl = all_rank_ncount_cb_don, cell_call_methd = "is_cell_ed", donor_group = .x, numbr_of_cols = 5)
})
```


```{r, fig.height= 10, fig.width= 10, eval=T}
map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{
  kneeplt_facet_fn(ncount_tbl = all_rank_ncount_cb_don, cell_call_methd = "is_cell_cb", donor_group = .x, numbr_of_cols = 5)
})
```




```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_gn_umi_bp_don <- map(all_msc_raw_mdata, ~{
  .x #%>% 
    # filter(is_cell_cb != "empty_bc") %>%
    # filter(nFeature_RNA < 300)
    
}) %>%
  bind_rows(., .id = "donor") %>% 
  left_join(., donor_10x_mdata_sk21_summ, by = "donor") 


```


```{r, fig.height= 4, fig.width= 4, eval=T}
all_rank_gn_umi_bp_don <- all_rank_gn_umi_bp_don %>% mutate(is_cell_cb2 = ifelse(is_cell_cb %in% c("CellCb.5", "CellCb.8"), "CellCb", is_cell_cb))
```

```{r, message=F, warning=F, eval=T}
## save raw metadata
saveRDS(all_rank_gn_umi_bp_don, "data/processed/Pf/m2_all/all_rank_gn_umi_bp_don.RDS")
```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
scatter_facet_fn <- function(ncount_tbl, cell_call_methd, donor_group, numbr_of_cols = 4, raster_resln = 50, use_raster = T, mthd_col = cell_call_mthd_col){ 

  ncount_tbl %>%
    ggplot(aes(x = nFeature_RNA, y= nCount_RNA, color = !!rlang::sym(cell_call_methd))) +
    (if (use_raster) ggrastr::rasterise(geom_point(size=0.1), dpi=raster_resln) else geom_point(size=0.1)) +
    geom_hline(yintercept = c(50, 100)) + 
    geom_vline(xintercept = c(50, 100)) +
    theme_classic() +
    scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x), labels = trans_format("log10", math_format(10^.x)))+
    scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x), labels = trans_format("log10", math_format(10^.x)))+
    facet_nested_wrap(vars(!!rlang::sym(donor_group), donor), ncol = 4) +
    scale_color_manual(name= 'CellCall', values = mthd_col) +
    theme(axis.text=element_text(size=15), 
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      # annotation_logticks() +
      labs(x= 'Gene count', y= paste0("Transcript", ' count'),
           # title = donor_nm
           )+
      labs(title = paste0(str_remove(cell_call_methd, "is_cell_"), "_", donor_group))
}


```

```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
# Bad_cells with less than 300 genes
map(c("is_cell_cb2", "is_cell_ed", "is_cell_cr")[1],  ~{ 
  mthd = .x
  
  tbl = all_rank_gn_umi_bp_don %>%
    filter(nFeature_RNA < 300 & (!!rlang::sym(mthd) %in% c("empty_bc", "nonCell") ))
    
  map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload"), ~{
    
    scatter_facet_fn(ncount_tbl = tbl, cell_call_methd = mthd, donor_group = .x, numbr_of_cols = 5, use_raster = T, raster_resln = 50)

})
})

```


```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
# Good cells with less than 300 genes
map(c("is_cell_cb2", "is_cell_ed", "is_cell_cr")[1],  ~{ 
  mthd = .x
  
  tbl = all_rank_gn_umi_bp_don %>%
    filter(nFeature_RNA < 300 & !(!!rlang::sym(mthd) %in% c("empty_bc", "nonCell") ))
    
  map(c("p_map_cat", "hs_map_cat", "peak", "frac_load", "overload")[1], ~{
    
    scatter_facet_fn(ncount_tbl = tbl, cell_call_methd = mthd, donor_group = .x, numbr_of_cols = 5, use_raster = F)

})
})

```




```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
# All cells 
map(c("is_cell_cb2", "is_cell_ed", "is_cell_cr")[1],  ~{ 
  mthd = .x
  
  tbl = all_rank_gn_umi_bp_don 
    
  map(c("frac_load", "p_map_cat", "hs_map_cat", "peak", "overload")[1], ~{
    
    scatter_facet_fn(ncount_tbl = tbl, cell_call_methd = mthd, donor_group = .x, numbr_of_cols = 5, use_raster = T, raster_resln = 50)

})
})

```


```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=12}
umi_box_vln_plt_fn <- function(ncount_tbl, cell_call_methd, numbr_of_cols = 4, box_or_vln, donor_group = "frac_load", gn_umi = "nCount_RNA", gn_umi_nm = "Transcript count", scales_free = "free_y", mthd_col = cell_call_mthd_col){ 

  ncount_tbl %>%
    ggplot(aes( y= !!rlang::sym(gn_umi), x = !!rlang::sym(cell_call_methd), color = !!rlang::sym(cell_call_methd))) + 
    (if (box_or_vln == "boxplot") {geom_boxplot()} else if (box_or_vln == "violin") {geom_violin()}) +
    geom_hline(yintercept = 20000) + 
    geom_hline(yintercept = 100) +
    scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
              labels = trans_format("log10", math_format(10^.x)))+ 
    theme_classic() +
   # facet_nested_wrap(vars(frac_load, donor), ncol = numbr_of_cols, scales = "free_y") +
    facet_nested_wrap(vars(!!!rlang::syms(unique(c(donor_group, "donor")))), ncol = numbr_of_cols, scales = scales_free) +
    scale_color_manual(name= 'CellCall', values = mthd_col) +
    theme(axis.text=element_text(size=15), 
          axis.text.x=element_text(angle=45, hjust = 1),
            axis.title=element_text(size=20,face="bold"), 
            legend.text = element_text(size=20), 
            legend.title = element_text(size=15,face="bold"),
            legend.position = "bottom",
            strip.text.x = element_text(size = 10, face = "bold")
    )+ 
      labs(x= 'Gene count', y= gn_umi_nm,
           title = paste0(str_remove(cell_call_methd, "is_cell_"), "_", donor_group)
           )
}


```


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=5}
# All cells 
map(c("is_cell_cb2", "is_cell_ed", "is_cell_cr"),  ~{ 
  mthd = .x
  
  map(c("frac_load"), ~{
    
    umi_box_vln_plt_fn(ncount_tbl = all_rank_gn_umi_bp_don, cell_call_methd = mthd,  numbr_of_cols = 19, box_or_vln = "boxplot", donor_group = .x, scales_free = "fixed")

})
})

```


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=5}
# All cells 
map(c("is_cell_cb2", "is_cell_ed", "is_cell_cr")[1],  ~{ 
  mthd = .x
  
  map(c("frac_load", "p_map_cat", "peak", "overload")[1], ~{
    
    umi_box_vln_plt_fn(ncount_tbl = all_rank_gn_umi_bp_don, cell_call_methd = mthd,  numbr_of_cols = 19, box_or_vln = "violin", donor_group = .x, scales_free = "fixed")

})
})

```


```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=8}
# All cells 
map(c("donor", "frac_load", "p_map_cat", "peak", "overload"), ~{
  
    tbl <- all_rank_gn_umi_bp_don %>% 
      filter(is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>%
      # filter(is_cell_cb %in% c("CellCb.8") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>% 
      pivot_longer(cols = c(is_cell_cb, is_cell_cr, is_cell_ed), names_to = "Mthds", values_to = "CellCat") %>% 
      filter(!(CellCat %in% c("empty_bc", "nonCell")))  %>%
      mutate(across(Mthds, ~factor(., levels = c("is_cell_cb", "is_cell_ed", "is_cell_cr"))))
      
  
    umi_box_vln_plt_fn(ncount_tbl = tbl, cell_call_methd = "Mthds",  numbr_of_cols = 10, box_or_vln = "violin", donor_group = .x, scales_free = "fixed")

})

```


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=5}
# All cells 
map(c("frac_load", "p_map_cat", "peak", "overload")[1:2], ~{
  
    tbl <- all_rank_gn_umi_bp_don %>% 
      filter(is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>%
      # filter(is_cell_cb %in% c("CellCb.8", "CellCb.5") | is_cell_ed == "CellEd" | is_cell_cr == "CellCr" ) %>%
      pivot_longer(cols = c(is_cell_cb, is_cell_cr, is_cell_ed), names_to = "Mthds", values_to = "CellCat") %>% 
      filter(!(CellCat %in% c("empty_bc", "nonCell")))  %>%
      mutate(across(Mthds, ~factor(., levels = c("is_cell_cb", "is_cell_ed", "is_cell_cr"))))
      
  
    umi_box_vln_plt_fn(ncount_tbl = tbl, gn_umi = "nFeature_RNA", gn_umi_nm = "Gene count", cell_call_methd = "Mthds",  numbr_of_cols = 19, box_or_vln = "violin", donor_group = .x, scales_free = "fixed")

})

```


```{r}
# all_msc_bp_edrops50 <- map(all_msc_bp_raw_noHsGns, ~subset(.x, subset = Is_Cell_100 == "Cell"))
all_msc_bp_edrops50 <- map(all_msc_bp_raw_noHsGns, ~subset(.x, subset = is_cell_ed == "CellEd"))
all_msc_bp_edrops50
```

```{r}
all_msc_bp_cbendr <- map(all_msc_bp_raw_noHsGns, ~subset(.x, subset = is_cell_cb == "CellCb.5" | is_cell_cb == "CellCb.8"))
```


```{r, message=FALSE}
# Add cellbender metadata
all_msc_bp_cbendr <- list(all_msc_bp_cbendr, all_cbender_mdata, edrops_cbender_mdata_uni[names(all_msc_bp_cbendr)]) %>%
  pmap(~ {
    ..2 %>%
      left_join(., ..3, by = c("barcode" = "bcode")) %>%
      select(barcode, cell_size, background_fraction, droplet_efficiency, mthds_union) %>%
      column_to_rownames("barcode") %>%
      AddMetaData(..1, .)
       })

all_msc_bp_edrops50 <- list(all_msc_bp_edrops50, all_cbender_mdata, edrops_cbender_mdata_uni[names(all_msc_bp_edrops50)]) %>%
  pmap(~ {
    ..2 %>%
      left_join(., ..3, by = c("barcode" = "bcode")) %>%
      select(barcode, cell_size, background_fraction, droplet_efficiency, mthds_union) %>%
      column_to_rownames("barcode") %>%
      AddMetaData(..1, .)
       })

```


```{r}
all_msc_bp_cbendr 
```

```{r}
map(all_msc_bp_edrops50, ~{table(.x@meta.data$nFeature > 100)})
```

```{r}
map(all_msc_bp_cbendr, ~{table(.x@meta.data$nFeature > 100)})
```


```{r, eval=T}
cell_counts <- pmap(list(all_msc_bp_raw_noHsGns, all_msc_bp_filt_noHsGns_mdata, all_msc_bp_edrops50, all_msc_bp_cbendr), ~{
  data.frame("raw" = dim(..1)[2], "cr_filt" = dim(..2)[2], "edrops50" = dim(..3)[2], "cbendr" = dim(..4)[2])
  
}) %>%
  bind_rows(., .id = "donor")

```


```{r, eval=T}
cell_counts
```

```{r, eval=T}
gogo()
```


```{r, eval=T}
##write out for normalization processing in farm

for (i in 1:length(sample_name)) {
## Create new directories
  system(paste0('mkdir -p data/processed/Pf/', sample_name[i], '/seu_obj'))
  # system(paste0('rm data/processed/Pf/', sample_name[i], '/seu_obj/*'))
}


```


```{r}
all_msc_bp_filt_noHsGns <- readRDS("data/processed/Pf/m2_all/all_msc_bp_filt_noHsGns.RDS")
```

```{r}
all_msc_bp_raw_noHsGns <- readRDS("data/processed/Pf/m2_all/all_msc_bp_raw_noHsGns.RDS")
```


```{r}
# Add emptydrops metadata
all_msc_bp_raw_noHsGns <- pmap(list(edrops_cbender_pf100_mdata, all_msc_bp_raw_noHsGns[names(edrops_cbender_pf100_mdata)]), ~{
  ..1 %>% 
    column_to_rownames("bcode") %>%
    # select(mthds_union) %>%
    AddMetaData(..2, .) 
    
})
```

Remove cells with less than 100 pf reads
```{r, eval=T}
##write out for normalization processing in farm
# all_msc_bp_filt_pf100
all_msc_bp_filt_pf100 <- map(all_msc_bp_filt_noHsGns, ~subset(.x, subset = pf_umi_count >= 100))

```

```{r, message=FALSE}
# Add cellbender metadata
all_msc_bp_filt_pf100 <- list(all_msc_bp_filt_pf100, all_cbender_mdata, edrops_cbender_mdata_uni[names(all_msc_bp_filt_pf100)]) %>%
  pmap(~ {
    ..2 %>%
      left_join(., ..3, by = c("barcode" = "bcode")) %>%
      select(barcode, cell_size, background_fraction, droplet_efficiency, mthds_union) %>%
      column_to_rownames("barcode") %>%
      AddMetaData(..1, .)
       })

```

```{r}
all_msc_bp_cbcr <- pmap(list(all_msc_bp_raw_noHsGns, all_msc_bp_cbendr, all_msc_bp_filt_pf100), ~subset(..1, cells = unique(c(colnames(..2), colnames(..3)))))
```

```{r, eval=T}
rm(all_msc_bp_raw_noHsGns)
```

```{r, message=FALSE}
# Add cellbender metadata
all_msc_bp_cbcr <- list(all_msc_bp_cbcr, all_cbender_mdata, edrops_cbender_mdata_uni[names(all_msc_bp_cbcr)]) %>%
  pmap(~ {
    ..2 %>%
      left_join(., ..3, by = c("barcode" = "bcode")) %>%
      select(barcode, cell_size, background_fraction, droplet_efficiency, mthds_union) %>%
      column_to_rownames("barcode") %>%
      AddMetaData(..1, .)
       })

```

FARM CODE

```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
### filt
imap(all_msc_bp_filt_pf100, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_cr.RDS")))

### cellbender
imap(all_msc_bp_cbendr, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_cbendr.RDS")))
all_msc_bp_cbendr

### emptydrops
# imap(all_msc_bp_edrops50, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_edrops50.RDS")))
imap(all_msc_bp_edrops50, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_edrops.RDS")))

# sample_name_ed <- paste0(names(all_msc_bp_edrops50), "_edrops", man_thresh[names(all_msc_bp_edrops50)])
# map2(all_msc_bp_edrops50, sample_name_ed, ~saveRDS(.x, paste0("data/processed/Pf/", str_remove(.y, "_ed.*"), "/seu_obj/bpc_seu_", str_extract(.y, "ed.*"),".RDS")))

### cbender + cr
# imap(all_msc_bp_edrops50, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_edrops50.RDS")))
imap(all_msc_bp_cbcr, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu_cbcr.RDS")))
map(all_msc_bp_cbcr, dim)
```
