---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  # library(patchwork)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  # library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(knitr)
  # library(readxl)
  # library(SCpubr)
  # library(ggh4x)
  # library(htmlwidgets)
  library(slingshot)
  library(harmony)
  library(ggforce)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  
})
```


#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/plotting_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/pf_common_vars.R")
```

```{r setup}
sample_set = "Mali2"

optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)

optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```


NB-LOCAL CODE
```{r, eval=F}

system(paste0("rsync -av jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/all_msc_filt_no_dblts.RDS /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/"
))

```

initialize variable for the sample names

```{r}
all_msc_filt_no_dblts <- readRDS("data/processed/Pf/all_msc_filt_no_dblts.RDS")
```


```{r}
strn_c_new_mdata_m22_min_qcd <- readRDS("data/processed/Pf/strn_c_new_mdata_m22_min_qcd.RDS")
```


```{r, fig.width=12, fig.height=4}
# add stage metadata
all_msc_filt_no_dblts <- pmap(list(all_msc_filt_no_dblts[names(strn_c_new_mdata_m22_min_qcd)], strn_c_new_mdata_m22_min_qcd, names(strn_c_new_mdata_m22_min_qcd)), ~{
  ..2 %>% 
  mutate(StrainDon = paste0(str_remove(..3, "SC"), "_", StrainQCd)) %>%
  select(bcode, StrainQCd, StrainDon) %>% 
    column_to_rownames("bcode") %>%
  AddMetaData(..1,.)
})

```


```{r, eval=T}

imap(all_msc_filt_no_dblts, ~DimPlot(.x, group.by = "StrainDon")  + labs(title = .y))
```


```{r}
msc_qcd_anotd <-readRDS("../Mali/data/processed/DB52e_star/msc_qcd_anotd.RDS")

```

```{r, eval=T}

msc_qcd_anotd_mdata_14 <- msc_qcd_anotd[["MSC14"]]@meta.data %>% mutate(StrainQCd= strain_qcd) 
```


```{r, fig.width = 18, fig.height = 4, message=FALSE, warning=FALSE}

strn_c_new_mdata_m22_min_qcd <-  strn_c_new_mdata_m22_min_qcd[names(optimum_k_qcd[optimum_k_qcd>1])]

ss_bar_plt <-list( c(strn_c_new_mdata_m22_min_qcd, list(msc_qcd_anotd_mdata_14 )), c(names(strn_c_new_mdata_m22_min_qcd), "MSC14"))  %>% 
  purrr::pmap(.,~{
  a_tbl <- ..1 %>%
      mutate(across(stage_afm, ~factor(., levels=c('Asexual', 'Female', 'Male', 'Developing', 'Unassigned')))) %>%
      dplyr::count(StrainQCd, stage_afm, .drop = FALSE) 
  
  ##get stage_ag labels and counts to add to the StrainQCd+stage_afm counts generated above
  stage_ag_cnt <- ..1 %>% 
    dplyr::select(StrainQCd, stage_ag, stage_afm) %>% 
    add_count(StrainQCd, stage_ag, name = 'stage_ag_n') %>% 
    distinct(pick(StrainQCd, stage_afm), .keep_all = T) %>% 
    left_join(a_tbl, ., by =c('StrainQCd', 'stage_afm')) %>% 
    rename(c('Stage'= stage_afm, 'Counts' =n, 'Strain' = StrainQCd))  
  
  print(stage_ag_cnt)
  
    b_plt <- stage_ag_cnt %>%
      # ggplot(., aes(x = Stage, y = Counts, fill = Stage, color=stage_ag_n<30)) +
      ggplot(., aes(x = Stage, y = Counts, fill = Stage)) +
      geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
      geom_text(aes(label=Counts), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 7, show.legend = F, fontface ='bold', colour='black')+
      scale_fill_manual(values = c_stg_cols) + #'Asexual' = "#FFE5B4", 'Male' = "#BFD7B5", 'Female'="#8ab6d6"
      # scale_color_manual(values = c('FALSE' = "white", 'TRUE' = "blue")) +
  geom_tile(aes(x = Stage, y = -10, fill = Stage, height = 14)) +
  facet_wrap( Strain ~ ., nrow = 1) +
      scale_y_continuous(expand = expansion(mult = c(0.01, 0.15))) +
      scale_x_discrete(expand = expansion(add = c(1, 1))) +
      theme_classic() +
  labs( y = 'Counts', title = ..2)+
      theme(
        axis.text.y  = element_text(size = 22),
        axis.title.y  = element_text(size = 22, face = "bold"),
        axis.title.x = element_blank(),
        # axis.title = element_text(size = 18, face = "bold"),
        legend.text = element_text(size = 22),
        legend.title = element_text(size = 22, face = "bold"),
        legend.position="bottom",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        strip.text =element_text(size=20,face="bold", colour = 'black'),
        panel.border=element_rect(colour="black",fill='transparent')
      )+ 
      guides(color = F)
    print(b_plt)
    return(b_plt)
})


```

```{r, fig.width=12, fig.height=4}
# add stage metadata
all_msc_filt_no_dblts <- map(all_msc_filt_no_dblts, ~{
  .x@meta.data %>% 
  rownames_to_column('bcode') %>%
  mutate(donor = str_remove(orig.ident, ".mrna"), 
         # strain_don = case_when(str_detect(orig.ident, "MSC") ~ paste0(Strain_DN, "_", donor))
         ) %>%
  left_join(., irods_id_sk21_mdata_cln %>% select(sample_nm, Symptomatic, Hb_Type), by = c("donor" = "sample_nm")) %>%
  select(bcode, donor, Symptomatic, Hb_Type) %>%
  column_to_rownames('bcode') %>% 
  AddMetaData(.x,.)
})

```



```{r, eval=T}

imap(all_msc_filt_no_dblts, ~DimPlot(.x, group.by = "StagePrelimC") + scale_color_manual(values = sp_fld_colors) + labs(title = .y))
```

```{r, eval=T}

imap(all_msc_filt_no_dblts, ~DimPlot(.x, group.by = "StagePrelimC", reduction = "pca") + scale_color_manual(values = sp_fld_colors) + labs(title = .y))
```

```{r}
all_msc_filt_no_dblts <- all_msc_filt_no_dblts[!names(all_msc_filt_no_dblts) %in% c("MSC38", "MSC39")]
```



```{r}
## Write out for Seri
all_msc_filt_no_dblts %>% .[c("MSC33", "MSC57")] %>% imap(., ~.x@meta.data[, c("StagePrelim", "Strain_DN")] %>% rownames_to_column("bcode") %>% write_csv(., paste0("data/other_students/sk25/all_msc_filt_no_dblts_", .y ,"_mdata.csv")))
```

```{r}
##Read MCA V3 blood stage reference dataset
# EF_V3_ref <- readRDS("../sc_velo/data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.4.RDS")
EF_V3_ref <- readRDS("../Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")
```

```{r}
##make refined metadata column that includes the gametocyte substages for finer resolution mapping
EF_V3_ref <- EF_V3_ref@meta.data %>%
  mutate(stage_fine = case_when(optionB_Stage == "no assignment" ~ as.character(Stage),
                                TRUE ~ as.character(optionB_Stage))
  ) %>% 
  mutate(across(stage_fine, ~str_to_sentence(str_replace_all(., c("Gametocytes"="Gametocyte", "comitted" = "committed"))))) %>%
  mutate(across(stage_fine, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
  mutate(across(stagenewx, ~droplevels(factor(str_to_sentence(.), levels = stg_refnd_lvls)))) %>% 
  dplyr::select(stage_fine, stagenewx) %>%
  AddMetaData(EF_V3_ref, .)
```


Write out annotated MSC14 dataset in .h5ad 
```{r, message=FALSE, warning=FALSE, eval=F}
##Save msc14 h5ad
all_msc_filt_no_dblts_raw <- map(all_msc_filt_no_dblts, ~DietSeurat(
  .x,
  counts = TRUE,
  data = F,
  scale.data = FALSE,
  features = NULL,
  assays = "RNA",
  dimreducs = c("pca", "umap"),
  graphs = NULL
)
)

##Change factor metadata to character as these are not handled properly when converting to h5ad
all_msc_filt_no_dblts_raw<- map(all_msc_filt_no_dblts_raw, ~{
  .x@meta.data %>% mutate(across(where(is.factor), as.character)) %>% AddMetaData(.x, .)
})

# imap(all_msc_filt_no_dblts_raw, ~system(paste0("mkdir /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/", .y)))


# ## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top
# imap(all_msc_filt_no_dblts_raw, ~SeuratDisk::SaveH5Seurat(.x, filename =  paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.y,"/",.y,"_qcd_anotd_diet.h5Seurat"), overwrite = T))
# 
# 
# imap(all_msc_filt_no_dblts_raw, ~Convert(paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.y,"/",.y,"_qcd_anotd_diet.h5Seurat"), dest = "h5ad", overwrite = T))

```


Write out annotated MSC14 dataset in .h5ad 
```{r, message=FALSE, warning=FALSE, eval=T}
##Save msc14 h5ad
# all_msc_filt_no_dblts_raw <- map(all_msc_filt_no_dblts, ~CreateSeuratObject(
#   counts = .x[["RNA"]]$counts,
#   meta.data = .x@meta.data
# )
# )
# 
# ##Change factor metadata to character as these are not handled properly when converting to h5ad
# all_msc_filt_no_dblts_raw<- map(all_msc_filt_no_dblts_raw, ~{
#   .x@meta.data %>% mutate(across(where(is.factor), as.character)) %>% AddMetaData(.x, .)
# })
# 
# imap(all_msc_filt_no_dblts_raw, ~system(paste0("mkdir /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/", .y)))

# ## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top
# imap(all_msc_filt_no_dblts_raw, ~SeuratDisk::SaveH5Seurat(.x, filename =  paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.y,"/",.y,"_qcd_anotd_diet.h5Seurat"), overwrite = T))
# 
# 
# imap(all_msc_filt_no_dblts_raw, ~Convert(paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.y,"/",.y,"_qcd_anotd_diet.h5Seurat"), dest = "h5ad", overwrite = T))

```

```{r, message=FALSE, warning=FALSE, eval=T}
##Get the raw counts

all_msc_filt_no_dblts_raw <- map(all_msc_filt_no_dblts, ~{
  obj <- CreateSeuratObject(counts = as(object = .x[["RNA"]]$counts, Class = "dgCMatrix"), meta.data = .x@meta.data)
  obj@meta.data <- obj@meta.data %>% mutate(across(where(is.factor), as.character))
  obj@reductions$pca <- CreateDimReducObject(embeddings = .x@reductions$pca@cell.embeddings, loadings = .x@reductions$pca@feature.loadings, projected = .x@reductions$pca@feature.loadings.projected, key = "PC_", assay = "RNA")
  obj@reductions$umap <- CreateDimReducObject(embeddings = .x@reductions$umap@cell.embeddings, loadings = .x@reductions$umap@feature.loadings, projected = .x@reductions$umap@feature.loadings.projected, key = "UMAP_", assay = "RNA")
  return(obj)
  
})

```

Write out raw objects into .h5ad 

```{r, message=FALSE, warning=FALSE, eval=F}
##Save msc14 h5ad

imap(all_msc_filt_no_dblts_raw, ~{
  
  ##Convert to v3 object since v5 not supported by seuratdisk 
  obj <- .x
  obj[["RNA3"]] <- as(object = obj[["RNA"]], Class = "Assay")
  DefaultAssay(obj) <- "RNA3"
  obj[["RNA"]] <- NULL
  obj <- RenameAssays(object = obj, RNA3 = 'RNA')
  
  # SeuratDisk::SaveH5Seurat(obj, filename =  paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.y,"/",.y,"_qcd_anotd_raw.h5Seurat"), overwrite = T)
  SeuratDisk::SaveH5Seurat(obj, filename =  paste0("data/processed/Pf/",.y,"/qcd_anotd_raw.h5Seurat"), overwrite = T)
  
  # Convert(paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.y,"/",.y,"_qcd_anotd_raw.h5Seurat"), dest = "h5ad", overwrite = T)
  Convert(paste0("data/processed/Pf/",.y,"/qcd_anotd_raw.h5Seurat"), dest = "h5ad", overwrite = T)
  
})

```

NB-LOCAL CODE
Transfer MCA and MSC14 .h5ad to farm
```{r, eval=F}

##Copy h5ad
map(names(all_msc_filt_no_dblts_raw), ~{
  # system(paste0("rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.x,"/",.x,"_qcd_anotd_diet.h5ad jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/",.x,"/"))
  system(paste0("rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/",.x,"/qcd_anotd_raw.h5ad jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/",.x,"/"))
  
})

```

```{r, eval=T}

##Copy h5ad
# rm(all_msc_filt_no_dblts_raw)

```

Intergrate each sample with lab one by one
```{r}
# select features that are repeatedly variable across datasets for integration

# msc_y21_mca_fem_combi <- list(unlist(list(msc_list_fnorm, msc_list_mnorm), recursive = F), rep(list(MCA_ref_fem_d10b,MCA_ref_m_d10b), each =4)) %>%
# msc_y21_mca_fem_combi <- list(msc_list_fle[c(1,3,4)], rep(list(MCA_ref_fem_d10b),3), c(200,200,100), c(100,100,50)) %>%
std_seu_int <- function(donor_obj, lab_obj=list(), k_filt = 200, k_wt =100, nf =200, fnf =500){
  msc_y21_mca_fem_features <- SelectIntegrationFeatures(object.list = unlist(list(donor_obj, lab_obj), recursive = F), verbose = F, nfeatures = nf, fvf.nfeatures = fnf)
  
  msc_y21_mca_fem_anchors <- FindIntegrationAnchors(object.list = unlist(list(donor_obj, lab_obj), recursive = F), anchor.features = msc_y21_mca_fem_features, verbose = F, k.filter = k_filt)
  
  # this command creates an "integrated" data assay
  msc_y21_mca_int <- IntegrateData(anchorset = msc_y21_mca_fem_anchors, verbose = F, k.weight =  k_wt)
  
  rm(msc_y21_mca_fem_anchors, msc_y21_mca_fem_features)
  
  return(msc_y21_mca_int)
  
}
# , "few cells")


```


### Std seurat normalization and clustering
```{r}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the "RNA" assay

std_norm_seu_umap <- function(seu_int_obj, seed_n = 123456){
  seu_int_obj %>%
    ScaleData(., features = rownames(.), verbose = FALSE)%>% 
    RunPCA(., features = VariableFeatures(object = .), verbose = FALSE) %>% 
    # RunUMAP(., reduction = "pca", dims = 1:6, seed.use = 123456, n.components = 3, min.dist = 0.3, n.neighbors = 30, spread = 1, local.connectivity = 1, a =60, b = 0.9, verbose = F) %>%
    RunUMAP(., reduction = "pca", dims = 1:6, seed.use = seed_n, n.components = 3) %>%
    FindNeighbors(k.param = 30, verbose = F)
}
```


```{r}
## Making list of objects and renaming before integrating
m2_list <- merge(x = all_msc_filt_no_dblts_raw[[1]], y = all_msc_filt_no_dblts_raw[2:length(all_msc_filt_no_dblts_raw)], add.cell.ids = str_remove(names(all_msc_filt_no_dblts_raw), "SC"))
```

```{r, fig.width=12, fig.height=4}
# add stage metadata
# gogo()

```

```{r, eval=F}
## Making list of objects and renaming before integrating
rm(all_msc_filt_no_dblts_raw)
```

Write out annotated MCA reference in .h5ad 
```{r, message=FALSE, warning=FALSE, eval=T}
##Change factor metadata to character as these are not handled properly when converting to h5ad
m2_list_raw <- m2_list
m2_list_raw@meta.data <- m2_list_raw@meta.data %>% mutate(across(where(is.factor), as.character))

## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top
m2_list_raw[["RNA3"]] <- as(object = m2_list_raw[["RNA"]], Class = "Assay")
DefaultAssay(m2_list_raw) <- "RNA3"
m2_list_raw[["RNA"]] <- NULL
m2_list_raw <- RenameAssays(object = m2_list_raw, RNA3 = 'RNA')

# SaveH5Seurat(m2_list_raw, filename =  paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/m2_list_raw.h5Seurat"), overwrite = T)
SaveH5Seurat(m2_list_raw, filename =  paste0("data/processed/Pf/m2_all/m2_list_raw.h5Seurat"), overwrite = T)

# Convert(paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/m2_list_raw.h5Seurat"), dest = "h5ad", overwrite = T)
Convert(paste0("data/processed/Pf/m2_all/m2_list_raw.h5Seurat"), dest = "h5ad", overwrite = T)

```

```{r, eval=F}

system(paste0("rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/m2_list_raw.h5ad jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/"))


```

```{r}
# m2_list[["RNA"]] <- split(m2_list[["RNA"]], f = m2_list$orig.ident)
m2_list

```

```{r, message=FALSE, warning=FALSE}
## save for running stdseurat in farm
saveRDS(m2_list, "data/processed/Pf/m2_all/m2_list.RDS")

```

!!!NB - FARM CODE - DONT DELETE
```{r, eval=T}
gogo()
## Run SNORM on farm like below
# nextflow run submsn_seu_norm.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/m2_list.RDS" --o_file_name "m2_seu_norm.RDS" --umap_name "UMAP_" --hvg 500 --cluster_resoln 1 --cl_nm "NULL" --umap_redctn_nm "umap.unintegrated" --redctn "pca" --integrtd "yes" --resume
```


```{r}
m2_seu_norm <-  readRDS("data/processed/Pf/m2_all/m2_seu_norm.RDS")
```


```{r}
# m2_list <-  seurat_norm_stand(Seurat_object = m2_list, hvg = 500)
```


```{r}
# m2_list <- FindNeighbors(m2_list, dims = 1:30, reduction = "pca")
# m2_list <- FindClusters(m2_list, resolution = 2 )

```


```{r}
# m2_list <- RunUMAP(m2_list, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")

```


```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
# DimPlot(m2_seu_norm, reduction = "umap.unintegrated", group.by = c("orig.ident", "StagePrelim"), label = F)
```

```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(m2_seu_norm, reduction = "umap.unintegrated", group.by = "StagePrelimC", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(m2_seu_norm, reduction = "umap.unintegrated", group.by = c("orig.ident", "StagePrelim"), label = F)
```


```{r}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
FeaturePlot(m2_seu_norm, features = "scf_class_norm_dbt_prop")
```


```{r, message=FALSE, warning=FALSE}
## save for running stdseurat in farm
# saveRDS(m2_list, "data/processed/Pf/m2_all/m2_list.RDS")

```

!!!NB - FARM CODE - DONT DELETE
```{r, eval=T}
# gogo()
## Run SNORM on farm like below
# nextflow run submsn_intgrtn.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_integrtn.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/m2_seu_norm.RDS" --mthd "HarmonyIntegration" --o_file_name "harmony_int.RDS" --cluster_resoln 1 --cl_nm "harmony_clusters" --umap_redctn_nm "umap.harmony" --new_redctn "integrated.harmony" --resume
```


```{r}
harmony_int <-  readRDS("data/processed/Pf/m2_all/harmony_int.RDS")
```


```{r}
harmony_int <- JoinLayers(harmony_int)
```


```{r}

DimPlot(harmony_int, reduction = "umap.harmony", group.by = c("StagePrelimAll", "harmony_clusters"), combine = F, label = T)
```

```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(harmony_int, reduction = "umap.harmony", group.by = "StagePrelim", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```




```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(harmony_int@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= harmony_int@meta.data$StagePrelimC,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot
harmony_int_asex <- harmony_int[,harmony_int@meta.data$StagePrelimC == "Asexual"]

```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot
plot_ly(data = data.frame(harmony_int_asex@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= harmony_int_asex@meta.data$StagePrelim,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, fig.width=12, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
# DimPlot(harmony_int_asex, reduction = "umap.unintegrated", group.by = c("StagePrelim", "orig.ident"), label = F, dims = c(1,2)) 
```


```{r, fig.width=12, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(harmony_int_asex, reduction = "umap.harmony", group.by = c("StagePrelim", "orig.ident"), label = F, dims = c(2,3)) 

```

```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
FeaturePlot(harmony_int_asex, reduction = "umap.harmony", features = "nFeature_RNA", label = F, dims = c(1,3))
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
# FeaturePlot(harmony_int_asex, reduction = "umap.unintegrated", features = "PF3D7-1243100")
FeaturePlot(harmony_int_asex, reduction = "umap.harmony", features = c("PF3D7-0502100", "PF3D7-0905800", "PF3D7-1323900"),order = T, pt.size = 0.5, dims = c(2,3), ncol = 4)
```



```{r}
##Read MCA V3 blood stage reference dataset
# EF_V3_ref <- readRDS("../sc_velo/data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.4.RDS")
EF_V3_ref <- readRDS("../Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")

# create seurat from raw counts
EF_V3_ref_asx_cnts <- CreateSeuratObject(counts = EF_V3_ref[, EF_V3_ref@meta.data$stagenew == "asexuals"][["RNA"]]$counts, meta.data = EF_V3_ref[, EF_V3_ref@meta.data$stagenew == "asexuals"]@meta.data)

# EF_V3_ref_asx_cnts[["RNA5"]] <- as(object = EF_V3_ref_asx_cnts[["RNA"]], Class = "Assay5")
# DefaultAssay(EF_V3_ref_asx_cnts) <- "RNA5"
# EF_V3_ref_asx_cnts[["RNA"]] <- NULL
# EF_V3_ref_asx_cnts <- RenameAssays(object = EF_V3_ref_asx_cnts, RNA5 = 'RNA')

```

```{r}
## get asexuals subset
m2_list_asex <- map(all_msc_filt_no_dblts_raw,  possibly(~.x[,.x@meta.data$StagePrelimC == "Asexual"], NULL)) %>% compact()

## get samples with more than 100 ascuals
m2_list_asex <- m2_list_asex[unlist(map(m2_list_asex, ~dim(.x)[2] > 100))]

# m2_list_asex <- map(m2_list_asex, ~{
#   obj <- .x
#   obj@reductions$pca <- NULL
#   obj@reductions$umap <- NULL
#   return(obj)
#   
# })

```

```{r}
## Making list of objects and renaming before integrating
v3_m2_list_asex <- merge(x = EF_V3_ref_asx_cnts, y = m2_list_asex, add.cell.ids = c("Lab", str_remove(names(m2_list_asex), "SC")))
```


```{r, fig.width=12, fig.height=4}
# add stage metadata
v3_m2_list_asex <- v3_m2_list_asex@meta.data %>% 
  rownames_to_column('bcode') %>%
  mutate(dset_labmrg = if_else(str_detect(orig.ident, "MSC"),  str_remove(orig.ident, ".mrna"), "Lab"),
         StageFL = coalesce(stagenewx, StagePrelim),
         StageFL_ND = str_to_sentence(StageFL),
         dset = if_else(str_detect(orig.ident, "MSC"),  "Field", "Lab")
         ) %>%
  select(bcode, dset_labmrg, StageFL, StageFL_ND, dset) %>%
  column_to_rownames('bcode') %>% 
  AddMetaData(v3_m2_list_asex,.)


```

```{r, message=FALSE, warning=FALSE}
## save for running stdseurat in farm
saveRDS(v3_m2_list_asex, "data/processed/Pf/m2_all/v3_m2_list_asex.RDS")

```


Write out annotated asexual lab+field in .h5ad 
```{r, message=FALSE, warning=FALSE, eval=T}
##Change factor metadata to character as these are not handled properly when converting to h5ad
v3_m2_list_asex_raw <- v3_m2_list_asex
v3_m2_list_asex_raw@meta.data <- v3_m2_list_asex_raw@meta.data %>% mutate(across(where(is.factor), as.character))

## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top
v3_m2_list_asex_raw[["RNA3"]] <- as(object = v3_m2_list_asex_raw[["RNA"]], Class = "Assay")
DefaultAssay(v3_m2_list_asex_raw) <- "RNA3"
v3_m2_list_asex_raw[["RNA"]] <- NULL
v3_m2_list_asex_raw <- RenameAssays(object = v3_m2_list_asex_raw, RNA3 = 'RNA')

# SaveH5Seurat(v3_m2_list_asex_raw, filename =  paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/v3_m2_list_asex_raw.h5Seurat"), overwrite = T)
SaveH5Seurat(v3_m2_list_asex_raw, filename =  paste0("data/processed/Pf/m2_all/v3_m2_list_asex_raw.h5Seurat"), overwrite = T)

# Convert(paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/v3_m2_list_asex_raw.h5Seurat"), dest = "h5ad", overwrite = T)
Convert(paste0("data/processed/Pf/m2_all/v3_m2_list_asex_raw.h5Seurat"), dest = "h5ad", overwrite = T)

```

!!!NB - FARM CODE - DONT DELETE
```{r, eval=T}
gogo()
## Run SNORM on farm like below
# nextflow run submsn_seu_norm.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/v3_m2_list_asex.RDS" --o_file_name "v3_m2_asex_seu_norm.RDS" --umap_name "UMAP_" --hvg 300 --cluster_resoln 1 --cl_nm "NULL" --umap_redctn_nm "umap.unintegrated" --redctn "pca" --integrtd "yes" --resume
```


```{r}
v3_m2_asex_seu_norm <-  readRDS("data/processed/Pf/m2_all/v3_m2_asex_seu_norm.RDS")
```

```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_seu_norm, reduction = "umap.unintegrated", group.by = "StagePrelimC", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_seu_norm, reduction = "umap.unintegrated", group.by = c("orig.ident", "StagePrelim"), label = F)
```


!!!NB - FARM CODE - DONT DELETE
```{r, eval=T}
gogo()
## Run SNORM on farm like below
# nextflow run submsn_intgrtn.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_integrtn.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/v3_m2_asex_seu_norm.RDS" --mthd "HarmonyIntegration" --o_file_name "v3_m2_asex_harmony.RDS" --cluster_resoln 1 --cl_nm "harmony_clusters_asex" --umap_redctn_nm "umap.harmony" --new_redctn "integrated.harmony" --resume
```


```{r}
v3_m2_asex_harmony <-  readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony.RDS")
```


```{r}
v3_m2_asex_harmony <- JoinLayers(v3_m2_asex_harmony)
```


```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "StagePrelimC", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = c("orig.ident", "StagePrelim"), label = F)
```


```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "harmony_clusters_asex", label = T, dims = c(1,2))
```




```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = c("StageFL"), label = F)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "pca", group.by = c("StageFL"), label = F)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "pca", group.by = c("dset"), label = F)
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_harmony@reductions[["pca"]]@cell.embeddings),
        x = ~PC_1, y = ~PC_2, z = ~PC_3,
        color= v3_m2_asex_harmony@meta.data$StageFL,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.2, 
        type = 'scatter3d'
)

```



```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_harmony@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= v3_m2_asex_harmony@meta.data$StageFL,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.2, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_harmony@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= v3_m2_asex_harmony@meta.data$StageFL,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.2, 
        type = 'scatter3d'
)

```


```{r}
strn_c_new_mdata_m22_min_qcd_mg <- strn_c_new_mdata_m22_min_qcd %>% 
  bind_rows(., .id = "donor") %>% 
  mutate(bcode = paste0(str_remove(donor, "SC"), "_", bcode), 
         StrainDon = paste0(str_remove(donor, "SC"), "_", StrainQCd)) %>%
  select(bcode, StrainQCd, StrainDon) %>% 
  column_to_rownames("bcode") 
```

```{r, fig.width=12, fig.height=4}
# add stage metadata
v3_m2_asex_harmony <- strn_c_new_mdata_m22_min_qcd_mg %>% select(StrainDon) %>%  AddMetaData(v3_m2_asex_harmony,.)

```

```{r, fig.width=12, fig.height=4}
# add stage metadata
v3_m2_asex_harmony <- v3_m2_asex_harmony@meta.data %>% 
  rownames_to_column('bcode') %>%
  mutate(donor = str_remove(orig.ident, ".mrna"), 
         # strain_don = case_when(str_detect(orig.ident, "MSC") ~ paste0(Strain_DN, "_", donor))
         ) %>%
  left_join(., irods_id_sk21_mdata_cln %>% select(sample_nm, Symptomatic, Hb_Type), by = c("donor" = "sample_nm")) %>%
  select(bcode, donor, Symptomatic, Hb_Type) %>%
  column_to_rownames('bcode') %>% 
  AddMetaData(v3_m2_asex_harmony,.)

```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
v3_m2_asex_harmony@meta.data %>% filter(str_detect(donor, "MSC")) %>% dplyr::count(donor, Symptomatic, Hb_Type)
```


```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
m2_asex_harmony_tr <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# pseudobulk the counts based on donor-condition-celltype
pseudo_m2_asex <- AggregateExpression(m2_asex_harmony_tr, assays = "RNA", return.seurat = T, group.by = c("Symptomatic", "donor", "StagePrelim"))

# each 'cell' is a donor-condition-celltype pseudobulk profile
tail(Cells(pseudo_m2_asex))
```

```{r, eval=T}
gogo()

```

```{r, fig.width=15, fig.height=5}
print(merge_dat[merge_dat$gene%in%c('SRGN','HLA-DRA'),c('gene','p_val.sc','p_val.bulk')])
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
SplitObject(harmony_int_asex, split.by = "orig.ident")
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
SplitObject(harmony_int_asex, split.by = "orig.ident")
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
SplitObject(harmony_int_asex, split.by = "orig.ident")
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
SplitObject(harmony_int_asex, split.by = "orig.ident")
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
SplitObject(harmony_int_asex, split.by = "orig.ident")
```

```{r}
## Making list of objects and renaming before integrating
m2_list <- merge(x = SplitObject(harmony_int_asex, split.by = "orig.ident"), y = EF_V3_ref[, EF_V3_ref@meta.data$stagenew == "asexuals"])
```

```{r, eval=FALSE}
m2_list <- IntegrateLayers(
  object = m2_list, method = CCAIntegration,
  orig.reduction = "pca", new.reduction = "integrated.cca",
  verbose = FALSE
)


```

```{r, eval=FALSE}
obj <- IntegrateLayers(
  object = obj, method = FastMNNIntegration,
  new.reduction = "integrated.mnn",
  verbose = FALSE
)
```

```{r, eval=FALSE}

m2_list <- FindNeighbors(m2_list, reduction = "integrated.cca", dims = 1:30)
m2_list <- FindClusters(m2_list, resolution = 2, cluster.name = "cca_clusters")
m2_list <- RunUMAP(m2_list, reduction = "integrated.cca", dims = 1:30, reduction.name = "umap.cca",n.components = 3L)
```

```{r, eval=FALSE}

DimPlot(m2_list, reduction = "umap.cca", group.by = c("StagePrelimAll", "cca_clusters"), combine = F, label = T)
```

```{r, eval=FALSE}
FeaturePlot(m2_list, reduction = "umap.cca", features = "scf_class_norm_dbt_prop")
```

```{r, eval=FALSE}
FeaturePlot(m2_list, reduction = "umap.cca", features = "nFeature_RNA")
```


```{r, eval=FALSE}
m2_list <- IntegrateLayers(
  object = m2_list, method = RPCAIntegration,
  orig.reduction = "pca", new.reduction = "integrated.rpca",
  verbose = FALSE
)
```


```{r, eval=FALSE}

m2_list <- FindNeighbors(m2_list, reduction = "integrated.rpca", dims = 1:30)
m2_list <- FindClusters(m2_list, resolution = 2, cluster.name = "rpca_clusters")
m2_list <- RunUMAP(m2_list, reduction = "integrated.rpca", dims = 1:30, reduction.name = "umap.rpca",n.components = 3L)
```

```{r, eval=FALSE}

DimPlot(m2_list, reduction = "umap.rpca", group.by = c("StagePrelimAll", "rpca_clusters", "orig.ident"), combine = F, label = T)
```


```{r}
m2_list <- IntegrateLayers(
  object = m2_list, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "integrated.harmony",
  verbose = T
)
```


```{r}

m2_list <- FindNeighbors(m2_list, reduction = "integrated.harmony", dims = 1:30)
m2_list <- FindClusters(m2_list, resolution = 2, cluster.name = "harmony_clusters_asex")
m2_list <- RunUMAP(m2_list, reduction = "integrated.harmony", dims = 1:30, reduction.name = "umap.harmony",n.components = 3L)
```

```{r}

DimPlot(m2_list, reduction = "umap.harmony", group.by = c("StagePrelimAll", "harmony_clusters_asex"), combine = F, label = T)
```


```{r}
m2_list <- JoinLayers(m2_list)
```

```{r}

DimPlot(m2_list, reduction = "umap.harmony", group.by = c("Strain_DN"), combine = F, label = F)
```


```{r}

FeaturePlot(m2_list, reduction = "umap.harmony", features = "PF3D7-1222600")
```


```{r}
FeaturePlot(m2_list, reduction = "umap.harmony", features = c("PF3D7-1222600", "PF3D7-0935400"),order = T, pt.size = 0.5, raster = F)
```


```{r}
#ap2g, gdv1, pfs16, etramp10.3, gexp5, pfg14-744, pfg14-748
DotPlot(m2_list, features = c("PF3D7-1222600", "PF3D7-0935400", "PF3D7-0406200", "PF3D7-1016900", "PF3D7-0936600", "PF3D7-1477300", "PF3D7-1477700"), group.by = "StagePrelimAll") + theme(axis.text.x = element_text(angle = 15)) + scale_x_discrete(labels = c("ap2g", "gdv1", "pfs16", "etramp10.3", "gexp5", "pfg14-744", "pfg14-748"))
```

```{r}
m2_list <- JoinLayers(m2_list)
```


```{r}
cells_ap2g_exp <- LayerData(m2_list, layer = "counts")["PF3D7-1222600",]
```


```{r}
cells_ap2g_exp
```

```{r}
table(m2_list$StagePrelimAll)

```

```{r}
m2_list@meta.data %>% filter(StagePrelimAll == "Field unlabelled") %>% select(contains("stage_lab"), nFeature_RNA)
```

```{r}
DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = names(cells_ap2g_exp[cells_ap2g_exp>1]), sizes.highlight = 2, raster=FALSE)
```


```{r}
DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = colnames(m2_list[,m2_list@meta.data$StagePrelimAll == "Developing"]), sizes.highlight = 2, raster=FALSE)
```


```{r}
DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = colnames(m2_list[,m2_list@meta.data$StagePrelimAll == "Field unlabelled"]), sizes.highlight = 2, raster=FALSE)
```

```{r}

m2_list@meta.data$str_stg <-  paste0(m2_list@meta.data$Strain_DN, "_", m2_list@meta.data$StagePrelimAll)

```


```{r}

m2_list@meta.data %>% glimpse
```

```{r}

DimPlot(m2_list, reduction = "umap.harmony", group.by = "str_stg") + theme(legend.position = "none")
```


```{r}

m2_list@meta.data %>% glimpse

```

```{r}

tbl <-m2_list@meta.data %>% mutate(oid = str_remove(orig.ident, ".mrna")) %>% left_join(., irods_id_sk21_mdata_cln, by = c("oid"="sample_nm")) %>% select(Symptomatic, Hb_Type)

```

```{r}

m2_list@meta.data$sympt <-  tbl$Symptomatic
m2_list@meta.data$hb <-  tbl$Hb_Type
```


```{r}

DimPlot(m2_list, reduction = "umap.harmony", group.by = "sympt") 
```


```{r}

DimPlot(m2_list, reduction = "umap.harmony", group.by = "hb")
```

```{r}

m2_list@meta.data$str_stg <-  paste0(m2_list@meta.data$Strain_DN, "_", m2_list@meta.data$StagePrelimAll)

```

```{r}

DimPlot(m2_list, reduction = "umap.harmony", group.by = c("StagePrelimAll", "harmony_clusters_asex"), combine = F, label = T, dims = c(1,3))
```


```{r}

FeaturePlot(m2_list, features = c("nFeature_RNA", "nCount_RNA"), combine = F, reduction = "umap.harmony", dims = c(1,3))
```


```{r}

DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = rownames(m2_list@meta.data[m2_list@meta.data$scf_class_norm_dbt_prop > 0.3, ]))
DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = rownames(m2_list@meta.data[m2_list@meta.data$scf_class_norm_dbt_prop > 0.3, ]), dims = c(1,3))
```


```{r}

DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = rownames(m2_list@meta.data[m2_list@meta.data$nFeature_RNA > 2000, ]))
DimPlot(m2_list, reduction = "umap.harmony", cells.highlight = rownames(m2_list@meta.data[m2_list@meta.data$nFeature_RNA > 2000, ]), dims = c(1,3))
```

```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
# donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4])))

DimPlot(msc_combi, group.by = c("stageHL", "orig.ident"),pt.size = 0.1) 
DimPlot(msc_combi, group.by = "stageHL", pt.size = 0.1, reduction = "pca", label = T) 
DimPlot(msc_combi,pt.size = 1, group.by = c("stageHL", "orig.ident"), dims = c(1,3)) 
```


```{r}

DimPlot(msc_combi,pt.size = 1, group.by = "orig.ident") + scale_color_manual(values = donor_cols_m)
DimPlot(msc_combi,pt.size = 1, group.by = "orig.ident", dims = c(1,3)) + scale_color_manual(values = donor_cols_m)


```


```{r}

DimPlot(msc_combi,pt.size = 1, group.by = "StageMCA") + scale_color_manual(values = sp_colors)
DimPlot(msc_combi,pt.size = 1, group.by = "StageMCA", dims = c(1,3)) + scale_color_manual(values = sp_colors)


```

```{r}
FeaturePlot(msc_combi, pt.size = 1,features  = "nFeature_RNA") 

DimPlot(msc_combi, group.by = "orig.ident",pt.size = 1, reduction = "pca") + scale_color_manual(values = donor_cols_m)
FeaturePlot(msc_combi, features  = "nFeature_RNA", pt.size = 1,reduction = "pca") 

```


```{r, results="asis"}
# cat("## Visualize field parasites {.tabset}   \n")
# DimPlot(all_msc_filt_no_dblts_stgstrqc[[1]], reduction = "umap", group.by = "stageHL", dims = c(1, 2),
#                 cols = sp_fld_colors ) 

```


```{r, results="asis"}

trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.5, "cm")
)

pmap(list(rep(c("umap","pca"),2), c(2,2,3,3)), ~{
  DimPlot(msc_combi, reduction = ..1, group.by = "stageHL", dims=c(1,..2), 
          cols = sp_fld_colors , pt.size = 0.1,  label = F, repel = T, label.size = 6 ,
  ) +
    theme(axis.text=element_blank(), 
          axis.title=element_blank(),
          plot.title = element_blank(),
          aspect.ratio = 1/1,
    ) + 
    guides(colour = F,
           x= trunc_axis,
           y= trunc_axis) +
    theme(axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed")),
          axis.title = element_text(hjust = 0,size=14, face = "bold")) +
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(breaks = NULL)
  
})


```



```{r, results="asis", fig.width=6, fig.height=3}

msc_combi@meta.data %>% dplyr::count(orig.ident, stageHL) %>% mutate(across(orig.ident, ~str_remove(., ".mrna"))) %>% ggplot(aes(x=orig.ident, y = n, fill = stageHL)) + geom_col(position = position_dodge(width = 0.8))+ scale_fill_manual(values = sp_fld_colors)+
  theme_classic()+
  theme(axis.text=element_text(size = 18),
        axis.title=element_blank(),
        plot.title = element_blank(),
        legend.position = "none"
  )

```


```{r}
plot_ly(data = as.data.frame(msc_combi@reductions[["umap"]]@cell.embeddings),
        x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3,
        color= msc_combi$stageHL,
        # colors = c("blue", "red"),
        opacity = 0.8,
        size = 1, 
        type = "scatter3d"
)
```


```{r}
saveRDS(msc_combi, "data/processed/Pf/msc_combi.RDS")
# saveRDS(msc_combi, "data/processed/Pf/share_w_sunil/msc_combi.RDS")
```

Write out annotated MCA reference in .h5ad 
```{r, message=FALSE, warning=FALSE}
##Save msc14 h5ad
msc_combi_diet <- DietSeurat(
  msc_combi,
  counts = TRUE,
  data = T,
  scale.data = FALSE,
  features = NULL,
  assays = NULL,
  dimreducs = c("pca", "umap", "umap_test"),
  graphs = NULL
)

##Change factor metadata to character as these are not handled properly when converting to h5ad
msc_combi_diet@meta.data <- msc_combi_diet@meta.data %>% mutate(across(where(is.factor), as.character))
## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top

SaveH5Seurat(msc_combi_diet, filename =  "/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_combi_diet.h5Seurat", overwrite = T)

Convert("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_combi_diet.h5Seurat", dest = "h5ad", overwrite = T)

```

NB-LOCAL CODE
Transfer MCA and MSC14 .h5ad to farm
```{r, eval=F}

system("rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_combi_diet.h5ad jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/"
)

```


Write out annotated MCA reference in .h5ad 
```{r, message=FALSE, warning=FALSE}
##Save msc14 h5ad

msc_combi_diet_raw <- CreateSeuratObject(counts = msc_combi[["RNA"]]$counts, meta.data = msc_combi@meta.data)

##Change factor metadata to character as these are not handled properly when converting to h5ad
msc_combi_diet_raw@meta.data <- msc_combi_diet_raw@meta.data %>% mutate(across(where(is.factor), as.character))
## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top

msc_combi_diet_raw@reductions$pca <- CreateDimReducObject(embeddings = msc_combi@reductions$pca@cell.embeddings, loadings = msc_combi@reductions$pca@feature.loadings, projected = msc_combi@reductions$pca@feature.loadings.projected, key = "PC_", assay = "RNA")

msc_combi_diet_raw@reductions$umap <- CreateDimReducObject(embeddings = msc_combi@reductions$umap@cell.embeddings, loadings = msc_combi@reductions$umap@feature.loadings, projected = msc_combi@reductions$umap@feature.loadings.projected, key = "UMAP_", assay = "RNA")

SaveH5Seurat(msc_combi_diet_raw, filename =  "/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_combi_diet_raw.h5Seurat", overwrite = T)

Convert("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_combi_diet_raw.h5Seurat", dest = "h5ad", overwrite = T)

```

Transfer MCA and MSC14 .h5ad to farm
```{r, eval=F}

system("rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_combi_diet_raw.h5ad jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/"
)

```
## MSC combi without asexuals

```{r}
## Making list of objects and renaming before integrating
msc_gam_list<-map(all_msc_filt_no_dblts, ~subset(.x, !(subset = StageMCA %in% c("Early trophozoite", "Late ring"))))
don_ids = str_remove(names(msc_gam_list), "sc")
msc_gam_list <- map2(msc_gam_list, don_ids, ~RenameCells( .x, add.cell.id = .y))
```


```{r}
## Run integration
msc_gam_combi <- std_seu_int(donor_obj=msc_gam_list) 

```



```{r}
## Run integration
# DefaultAssay(msc_gam_combi) <- "integrated"
# msc_gam_combi <- std_norm_seu_umap(seu_int_obj=msc_gam_combi) 

```


```{r}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the "RNA" assay
DefaultAssay(msc_gam_combi) <- "integrated"

msc_gam_combi <- msc_gam_combi %>%
  ScaleData(., features = rownames(.), verbose = FALSE) %>% 
  RunPCA(., features = VariableFeatures(object = .), verbose = FALSE) 
```

```{r}
ElbowPlot(msc_gam_combi)
```


```{r}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the "RNA" assay

msc_gam_combi <- msc_gam_combi %>%
  # RunUMAP(., reduction = "pca", dims = 1:6, seed.use = 123456, n.components = 3, min.dist = 0.3, n.neighbors = 30, spread = 1, local.connectivity = 1, a =60, b = 0.9, verbose = F) %>%
  RunUMAP(., reduction = "pca", dims = 1:7, seed.use = 116937, n.components = 3, n.neighbors =40, min.dist = 0.2) %>% #389
  FindNeighbors(k.param = 30, verbose = F)
```

```{r, fig.width= 10, fig.height=3.5}
## Add donor and stage labels appropriate for all samples
msc_gam_combi@meta.data[,"Donor"] <- factor(str_remove(msc_gam_combi@meta.data$orig.ident, ".mrna"), levels = don_ids)
DimPlot(msc_gam_combi, group.by = c("stageHL", "orig.ident"),pt.size = 0.1) 
# msc_gam_combi <- msc_gam_combi@meta.data %>% 
#   mutate(stageHL_ref = case_when(stageHL == "Gametocyte (female)" ~ "Female", 
#                                                            stageHL == "Gametocyte (male)" ~ "Male", 
#                                                            T ~ stageHL)) %>% 
#   select(stageHL_ref) %>% AddMetaData(msc_gam_combi, .)

```

```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
# donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4])))

DimPlot(msc_gam_combi, group.by = c("stageHL", "orig.ident"),pt.size = 0.1) 
DimPlot(msc_gam_combi, group.by = "stageHL", pt.size = 0.1, reduction = "pca", label = T) 
DimPlot(msc_gam_combi,pt.size = 1, group.by = c("stageHL", "orig.ident"), dims = c(1,3)) 
```

```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
# donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4])))

DimPlot(msc_gam_combi, group.by = c("stageHL", "orig.ident"),pt.size = 0.1) 
DimPlot(msc_gam_combi, group.by = "stageHL", pt.size = 0.1, reduction = "pca", label = T) 
DimPlot(msc_gam_combi,pt.size = 1, group.by = c("stageHL", "orig.ident"), dims = c(1,3)) 
```

```{r}
mgc_umap_mdta_df <- msc_gam_combi@reductions$umap@cell.embeddings[,c("UMAP_1", "UMAP_2")] %>% as.data.frame() %>% rownames_to_column("cell_bc") %>% left_join(., msc_gam_combi@meta.data[,c("Donor", "nCount_RNA", "nFeature_RNA")] %>% as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc") 

don_f_plt <- mgc_umap_mdta_df %>% 
  mutate(samp = sample(n())) %>%
  arrange(samp) %>%
  ggplot(aes(x = UMAP_1, y =UMAP_2, color = Donor)) + 
  geom_point(size=.1) + 
  scale_color_manual(name = "Donor", values = donor_cols) +
  theme_classic()+
  umap_theme+
  theme(legend.position = "bottom", 
        plot.title = element_blank()) +
  guides(colour = guide_legend(override.aes = list(size=4.5), title.position = "top", nrow = 2),
         x= trunc_axis,
         y= trunc_axis)+
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(breaks = NULL)

cnts_plt <- mgc_umap_mdta_df %>% 
  mutate(samp = sample(n())) %>%
  arrange(samp) %>%
  ggplot(aes(x = UMAP_1, y =UMAP_2, color = nFeature_RNA)) + 
  geom_point(size=.2) + 
  # scale_color_distiller(name="Transcript count (X1e3)", direction = -1, palette = "GnBu") + 
  scale_color_distiller(name="Gene count", direction = -1, palette = "GnBu") + 
  theme_classic()+  
  umap_theme+
  theme(legend.position = "bottom", 
        plot.title = element_blank()) +
  guides(colour = guide_colorbar(title.position = "top", barwidth = unit(5, "cm")),
         x= trunc_axis,
         y= trunc_axis)+
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(breaks = NULL)

don_f_plt
cnts_plt
```

```{r, fig.width= 12, fig.height=4}
##Put together the plots for main gam integration

gam_int_plt <- plot_grid(don_f_plt,
                         cnts_plt,
                         NULL,
                         NULL,
                         ncol = 4,
                         rel_widths = c(.2,.2, .22, .38),
                         labels = c("A", "B", "C", "D"),
                         label_size = 24,
                         vjust= 1.5)

gam_int_plt

```


```{r, fig.width= 12, fig.height=7}
##save svg for sk21
save_plot("plots/mali2021_all/fig4/gam_int_plt.svg",gam_int_plt, base_width = 12, base_height = 4)
```

```{r}

DimPlot(msc_gam_combi,pt.size = 1, group.by = "orig.ident") + scale_color_manual(values = donor_cols_m)
DimPlot(msc_gam_combi,pt.size = 1, group.by = "orig.ident", dims = c(1,3)) + scale_color_manual(values = donor_cols_m)

DimPlot(msc_gam_combi,pt.size = 1, group.by = "StageMCA") + scale_color_manual(values = sp_colors)
DimPlot(msc_gam_combi,pt.size = 1, group.by = "StageMCA", dims = c(1,3)) + scale_color_manual(values = sp_colors)


```

```{r}

stages_umap <- DimPlot(msc_gam_combi,pt.size = 1, group.by = "stageHL_ref") + 
  scale_color_manual(values = sp_fld_colors) +
  umap_theme+
  theme(legend.position = "bottom", 
        plot.title = element_blank(),
        legend.title = element_text(face = "bold", size = 14),
        legend.text = element_text( size = 14)) +
  guides(color = guide_legend(title = "Stage", override.aes = list(size = 3), title.position = "top", nrow = 2) ,
         x= trunc_axis,
         y= trunc_axis) +
  theme(axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed")),
        axis.title = element_text(hjust = 0,size=14, face = "bold")) +
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(breaks = NULL)

rna_cnt <- FeaturePlot(msc_gam_combi,pt.size = 1, features = "nCount_RNA") + 
  umap_theme+
  theme(legend.position = "bottom", 
        plot.title = element_blank(),
        legend.title = element_text(face = "bold"),
        legend.key.width = unit(1.2, "cm"),
        axis.line = element_blank(),
        axis.ticks = element_blank(),
        axis.text  = element_blank(),
        axis.title  = element_blank()) #+
# guides(x= trunc_axis,
#        y= trunc_axis) +
# theme(axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed")),
#       axis.title = element_text(hjust = 0,size=14, face = "bold")) +
# scale_x_continuous(breaks = NULL) +
# scale_y_continuous(breaks = NULL)

stages_umap
rna_cnt
```

```{r, fig.width= 10, fig.height=3.5}

umap_split_don <- DimPlot(msc_gam_combi,pt.size = 1, group.by = "Donor", split.by = "Donor") + scale_color_manual(values = donor_cols) + theme(legend.position = "none", plot.title = element_blank())

umap_split_don
```


```{r, fig.width= 12, fig.height=7.5, eval =T}
##Put together the plots for annotation

plot_grid(umap_split_don,
          plot_grid(stages_umap, 
                    rna_cnt,
                    NULL,
                    NULL,
                    ncol = 4,
                    rel_widths = c(.25,.25,.25,.25),
                    labels = c("B", "C", "D", "E"),
                    label_size = 24,
                    vjust= 0.3
          ),
          nrow = 2,
          rel_widths = c(.45,.55),
          labels = c("A", ""),
          label_size = 24,
          vjust= 1.5)


```

```{r}
saveRDS(msc_gam_combi, "data/processed/Pf/msc_gam_combi.RDS")
saveRDS(msc_gam_combi, "data/processed/Pf/share_w_sunil/msc_gam_combi.RDS")
```



Write out annotated MCA reference in .h5ad 
```{r, message=FALSE, warning=FALSE}
##Save msc14 h5ad

msc_gam_combi_diet_raw <- CreateSeuratObject(counts = msc_gam_combi[["RNA"]]$counts, meta.data = msc_gam_combi@meta.data)

##Change factor metadata to character as these are not handled properly when converting to h5ad
msc_gam_combi_diet_raw@meta.data <- msc_gam_combi_diet_raw@meta.data %>% mutate(across(where(is.factor), as.character))
## !!!NB - NEEDS SORTING IN FUTURE UPGRADING TO SEURAT 5 RESULTED IN ERROR AS DESCRIBED https://github.com/mojaveazure/seurat-disk/issues/147#top

msc_gam_combi_diet_raw@reductions$pca <- CreateDimReducObject(embeddings = msc_gam_combi@reductions$pca@cell.embeddings, loadings = msc_gam_combi@reductions$pca@feature.loadings, projected = msc_gam_combi@reductions$pca@feature.loadings.projected, key = "PC_", assay = "RNA")

msc_gam_combi_diet_raw@reductions$umap <- CreateDimReducObject(embeddings = msc_gam_combi@reductions$umap@cell.embeddings, loadings = msc_gam_combi@reductions$umap@feature.loadings, projected = msc_gam_combi@reductions$umap@feature.loadings.projected, key = "UMAP_", assay = "RNA")

SaveH5Seurat(msc_gam_combi_diet_raw, filename =  "/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_gam_combi_diet_raw.h5Seurat", overwrite = T)

Convert("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_gam_combi_diet_raw.h5Seurat", dest = "h5ad", overwrite = T)

```


Transfer MCA and MSC14 .h5ad to farm
```{r, eval=F}

system("rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/msc_gam_combi_diet_raw.h5ad jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/"
)

```


####integrate field gams without LE plus mca gams - all msc objects included separately

```{r}
# MCA_V3_ref_gam <- EF_V3_ref[, EF_V3_ref@meta.data$stagenewx %in% c("gametocyte (developing)", "gametocyte (male)", "gametocyte (female)")]
MCA_V3_ref_gam <- EF_V3_ref[, EF_V3_ref@meta.data$Stagex %in% c("early stalk","late stalk", "branching", "early female", "late female", "early male", "late male")]

# msc_y21_gams <-map(msc_list, ~.x[, !.x@meta.data$stageHL %in% c("Early trophozoite", "Late ring")])
msc_gam_list_noLE<-map(all_msc_filt_no_dblts, ~subset(.x, !(subset = stageHL %in% c("Early trophozoite", "Late ring", "Female (LE)", "Male (LE)"))))



DimPlot(MCA_V3_ref_gam, group.by = "orig.ident", dims = c(1,3))
map(msc_gam_list_noLE, ~DimPlot(., group.by = c("stageHL")))
```


```{r}
## Run integration
msc_gam_combi_noLE <- std_seu_int(donor_obj=msc_gam_list_noLE, lab_obj = MCA_V3_ref_gam) 

```

```{r}
## Run integration
DefaultAssay(msc_gam_combi_noLE) <- "integrated"
msc_gam_combi_noLE <- std_norm_seu_umap(seu_int_obj=msc_gam_combi_noLE) 

```


```{r}
## Function to Add metadata for visualization to object when integrating field with lab

add_mdata <- function(seu_obj) {
  seu_obj@meta.data %>% 
    mutate(stage_int = case_when(is.na(stageHL) ~ Stagex, T ~ StageMCA),
           dset_int = case_when(str_detect(orig.ident, "D") ~ Stagex, T ~ orig.ident),
           stageHL_fine_int = case_when(str_detect(orig.ident, "D") ~ Stagex, T ~ stageHL),
           stageHL_int = case_when(str_detect(orig.ident, "D") ~ stagenewx, T ~ stageHL),
           stagec_int = case_when(str_detect(orig.ident, "D") ~ "Lab", T ~ StageMCA)) %>% 
    select(stage_int, dset_int, stageHL_int, stagec_int, stageHL_fine_int) %>% 
    AddMetaData(seu_obj, .)
}


```


```{r}
## Add metadata for visualization to object

msc_gam_combi_noLE<- add_mdata(seu_obj= msc_gam_combi_noLE)
```

```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
# donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4])))

DimPlot(msc_gam_combi_noLE, group.by = c("stageHL_int", "orig.ident"),pt.size = 0.1) 
DimPlot(msc_gam_combi_noLE, group.by = "stageHL_int", pt.size = 0.1, reduction = "pca", label = T, order=c("Lab","Female", "Gametocyte (female)","Gametocyte (male)","Male")) 
DimPlot(msc_gam_combi_noLE,pt.size = 1, group.by = c("stageHL_int", "orig.ident"), dims = c(1,3)) 
```


```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
DimPlot(msc_gam_combi_noLE, group.by = "stageHL_int", pt.size = 0.1, reduction = "umap", order=c("Female (LE)","Male (LE)","Female","Male", "Gametocyte (female)","Gametocyte (male)","Lab"), cols = sp_fld_colors) 
```

```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
DimPlot(msc_gam_combi_noLE, group.by = "stagec_int", pt.size = 0.1, reduction = "pca", order=c("Female","Male", "Gametocyte (female)","Gametocyte (male)","Lab"), cols = sp_fld_colors) 
```


```{r, fig.width=8, fig.height=3.5}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the "RNA" assay

stg_order <- c("Gametocyte (female)", "Gametocyte (male)","Female", "Female (LE)", "early female", "late female",  "early male", "late male")

donor_order <- c(names(donor_cols_m), "merged_MCA", "D5", "D10a","D10b")
```


```{r, fig.width= 10, fig.height=5}
# donor_cols_m = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(names(all_msc_filt_no_dblts[1:4]), ".mrna"))
DimPlot(msc_gam_combi_noLE, group.by = "orig.ident", pt.size = 0.1, reduction = "pca", order=donor_order, cols = donor_cols_m) 
```


#### Slingshot


## Trajectory inference of integrated object
Adjust to get appropriate clustering so as to input into slingshot
```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"

msc_gam_combi_noLE <- FindClusters(msc_gam_combi_noLE, resolution = 0.32, verbose = F)

```

```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
DimPlot(msc_gam_combi_noLE, reduction = "pca", group.by = "seurat_clusters", pt.size = 0.5, dims = c(1,2), label = T)

```

```{r}
msc_gam_combi_noLE_sce <- as.SingleCellExperiment(msc_gam_combi_noLE)


msc_gam_combi_noLE@meta.data %>% dplyr::count(stage_int, seurat_clusters)
msc_gam_combi_noLE@meta.data %>% dplyr::count(seurat_clusters,stage_int )
```


Convert seurat object into single cell experiment filling in the slots appropriately
```{r}
##Raw counts - Preparing single cell experiment object for slingshot trajectory inference and tradeseq DE

#logcounts(msc14_sub_sce[[i]]) <- log2(counts(msc14_sub_sce[[i]]) + 1)
##This command gives error below when running the entire notebook or this whole chunk so run the command independently - coulb be and issue with rmd
##Error in x$.self$finalize() : attempt to apply non-function - need blank line after

sce_obj_creatn <- function(sce_obj, seu_obj){
  for(i in 1:length(sce_obj)) {
    sce_obj[[i]] <- SingleCellExperiment(assays = list(counts = as.matrix(seu_obj[[i]][["RNA"]]$counts)), colData = seu_obj[[i]]@meta.data)
    
    logcounts(sce_obj[[i]]) <- log2((counts(sce_obj[[i]]) + 1))
    normcounts(sce_obj[[i]]) <- as.matrix(seu_obj[[i]]@assays$RNA@data)
    rowData(sce_obj[[i]])$feature_symbol <- str_replace(rownames(sce_obj[[i]]), "-", "_")
    reducedDims(sce_obj[[i]]) <- SimpleList(PCA = seu_obj[[i]]@reductions[["pca"]]@cell.embeddings, UMAP = seu_obj[[i]]@reductions[["umap"]]@cell.embeddings)
    
  }
  return(sce_obj)
}

```


```{r}
msc_gam_combi_noLE_sce <- sce_obj_creatn(sce_obj = list(msc_gam_combi_noLE_sce), seu_obj = list(msc_gam_combi_noLE))
msc_gam_combi_noLE_sce <- msc_gam_combi_noLE_sce[[1]]
```

Run slingshot for trajectory inference
```{r}
##Function -slingshot

##Run slingshot to determine trajectories
# msc_y21_mca_fem_combi_sce$seurat_clusters <- as.character(msc_y21_mca_fem_combi_sce$seurat_clusters)
# msc_y21_mca_fem_combi_sce <- list(msc_y21_mca_fem_combi_sce, start.clus, end.clus) %>%
sling_start <- function(sce_obj, s_clstr, clstr ="seurat_clusters", e_clstr = NULL){ 
  slingshot(sce_obj, 
            clusterLabels = clstr, 
            reducedDim = "PCA", 
            start.clus = s_clstr,
            end.clus = e_clstr,
            shrink =1, 
            stretch = 1,
            allow.break = F)
  
}

```

```{r}
# msc_gam_combi_noLE_sce <- sling_start(sce_obj = msc_gam_combi_noLE_sce, s_clstr = "early stalk", clstr = "stage_int")
msc_gam_combi_noLE_sce <- sling_start(sce_obj = msc_gam_combi_noLE_sce, s_clstr = "1", clstr = "seurat_clusters")
```

```{r}
##Prepare slingshot dataset (SDS) object 
msc_gam_combi_noLE_sce_sds <- SlingshotDataSet(msc_gam_combi_noLE_sce)

plot(reducedDims(msc_gam_combi_noLE_sce)$PCA[,1:2], col = brewer.pal(9,"Set1")[as.factor(msc_gam_combi_noLE_sce$stage_int)], pch=16, asp = 0.7)
lines(msc_gam_combi_noLE_sce_sds, lwd=2, col="black",dims = 1:2)
```


```{r, results="asis"}

plot(reducedDims(msc_gam_combi_noLE_sce)$PCA[,1:2], col = brewer.pal(7,"Set1")[as.factor(msc_gam_combi_noLE_sce$seurat_clusters)], pch=16, asp = 0.7)
lines(msc_gam_combi_noLE_sce_sds, lwd=2, col="black",dims = 1:2)

plot(reducedDims(msc_gam_combi_noLE_sce)$PCA[,2:3], col = brewer.pal(7,"Set1")[as.factor(msc_gam_combi_noLE_sce$seurat_clusters)], pch=16, asp = 0.7)
lines(msc_gam_combi_noLE_sce_sds, lwd=2, col="black",dims = 2:3)

umap_meta_df <- reducedDims(msc_gam_combi_noLE_sce)$PCA[,c(1,2)] %>% as.data.frame() %>%rownames_to_column("cell_bc") %>% left_join(., msc_gam_combi_noLE_sce@colData[,c("stageHL", "slingPseudotime_1", "slingPseudotime_2", "orig.ident")] %>% as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc")  #%>% 
# mutate(across(strain_dset, ~factor(., levels = rev(strain_dset_lvls_don)))) %>% arrange(strain_dset

plt <- umap_meta_df %>% 
  ggplot(aes(x = PC_1, y =PC_2, color = slingPseudotime_1)) + geom_point(size=0.5) + 
  scale_color_distiller(name="Pseudotime", direction = 1, palette = "Blues") + 
  theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 22) ,
        legend.title = element_text(size = 24, face = "bold") )

print(plt)

plt <- umap_meta_df %>% 
  ggplot(aes(x = PC_1, y =PC_2, color = slingPseudotime_2)) + geom_point(size=0.5) + 
  scale_color_distiller(name="Pseudotime", direction = 1, palette = "Greens") + 
  theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 22) ,
        legend.title = element_text(size = 24, face = "bold") )

print(plt)

plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_1, col=stageHL, fill=stageHL)) + 
  scale_color_manual(name = "Stage", values = sp_fld_colors)+ 
  scale_fill_manual(name = "Stage", values = sp_fld_colors)+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)


plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_2, col=stageHL, fill=stageHL)) + 
  scale_color_manual(name = "Stage", values = sp_fld_colors)+ 
  scale_fill_manual(name = "Stage", values = sp_fld_colors)+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)

plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_1, col=orig.ident, fill=orig.ident)) + 
  scale_color_manual(name = "Stage", values = donor_cols_m)+
  scale_fill_manual(name = "Stage", values = donor_cols_m)+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)

plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_2, col=orig.ident, fill=orig.ident)) + 
  scale_color_manual(name = "Stage", values = donor_cols_m)+
  scale_fill_manual(name = "Stage", values = donor_cols_m)+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)


```



```{r, results="asis"}
msc_gam_combi_noLE_pt<- msc_gam_combi_noLE@meta.data  %>% 
  rownames_to_column("cell_bc") %>% 
  mutate(n_cb = str_remove(cell_bc, "_1")) %>% 
  left_join(., umap_meta_df %>% 
              mutate(n_cb = str_remove(cell_bc, "_1")), by=c("n_cb", "orig.ident") ) %>% 
  select(cell_bc.x, slingPseudotime_1, slingPseudotime_2) %>% 
  column_to_rownames("cell_bc.x") %>% 
  AddMetaData(msc_gam_combi_noLE, .)

```


```{r, results="asis"}
FeaturePlot(msc_gam_combi_noLE_pt, features = "slingPseudotime_1", reduction = "pca")
FeaturePlot(msc_gam_combi_noLE_pt, features = "slingPseudotime_2", reduction = "pca")
```



### Visualize trajectory{.tabset}


```{r, results="asis"}

# pt_plts <- function(sce_obj, sds_obj, pt){
#     plot(reducedDims(sce_obj)$PCA[,1:2], col = brewer.pal(7,"Set1")[as.factor(sce_obj$seurat_clusters)], pch=16, asp = 0.7)
#     lines(sds_obj, lwd=2, col="black",dims = 1:2)
#     
#     plot(reducedDims(sce_obj)$PCA[,2:3], col = brewer.pal(7,"Set1")[as.factor(sce_obj$seurat_clusters)], pch=16, asp = 0.7)
#     lines(sds_obj, lwd=2, col="black",dims = 2:3)
#     
#     umap_meta_df <- reducedDims(sce_obj)$PCA[,c(1,3)] %>% as.data.frame() %>%rownames_to_column("cell_bc") %>% left_join(., sce_obj@colData[,c("stageHL", pt)] %>% as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc")  #%>% 
#   # mutate(across(strain_dset, ~factor(., levels = rev(strain_dset_lvls_don)))) %>% arrange(strain_dset
#   
#   plt <- umap_meta_df %>% 
#     ggplot(aes(x = PC_1, y =PC_3, color = !!rlang::sym(pt))) + geom_point(size=2) + 
#     scale_color_distiller(name="Pseudotime", direction = 1, palette = "Blues") + 
#     theme_classic()+
#     theme(axis.text=element_blank(), 
#           axis.title=element_blank(),
#           legend.text = element_text(size = 22) ,
#           legend.title = element_text(size = 24, face = "bold") )
#   
#   print(plt)
#   
#   plt = umap_meta_df %>% 
#     ggplot(., aes(x = !! rlang::sym(pt), col=stageHL, fill=stageHL)) + 
#     scale_color_manual(name = "Stage", values = sp_fld_colors)+ 
#     scale_fill_manual(name = "Stage", values = sp_fld_colors)+
#     geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
#     # geom_freqpoly()+
#     theme_classic()+
#     theme(axis.text=element_text(size=26),
#           axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
#           axis.title=element_text(size=30,face="bold"), 
#           legend.text = element_text(size = 26) ,
#           legend.title = element_text(size = 28, face = "bold"),
#           # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
#           # legend.position="bottom",
#           # legend.direction="vertical"
#     ) +
#     labs(x = "Pseudotime", y="Counts") +
#     guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))
#   
#   print(plt)
#   
#   }
# 

```

```{r, results="asis"}
# pt_plts(sce_obj = msc_gam_combi_noLE_sce, sds_obj = msc_gam_combi_noLE_sce_sds, pt = c("slingPseudotime_1", "slingPseudotime_2"))
```


Save integrated dataset with pseudotimes
```{r}
saveRDS(msc_gam_combi_noLE, "data/processed/Pf/msc_gam_combi_noLE.RDS")
saveRDS(msc_gam_combi_noLE_sce, "data/processed/Pf/msc_gam_combi_noLE_sce.RDS")
saveRDS(msc_gam_combi_noLE_sce_sds, "data/processed/Pf/msc_gam_combi_noLE_sce_sds.RDS")
```

########################### END OF INTERGRATION FOR FIRST PART

