---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(scran)
  library(SingleR)
  library(scDblFinder)
  library(duckplyr)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(dplyr::desc)
  conflicts_prefer(dplyr::count)
})
```



# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}

knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Common functions

#### Source common functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")

```


initialize variable for the sample names

```{r}
# slct_sample_names <-c("MSC48", "MSC68", "MSC14", "MSC49", "MSC60", "MSC66", "MSC67", "MSC70", "MSC24", "MSC33", "MSC40", "MSC54", "MSC55") %>% str_sort(., numeric = T)

```

###################START- RAW READING#############

```{r, eval=T}
pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
```

## Genotype clustering for MSC

Read in clusters.tsv files with assignment of cells to strains and PCA plots metrics

Transfer clusters.tsv files with assignment of cells to strains and PCA plots metrics for the best K from farm to my local Mac

```{r}

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

```


```{r}
# Remove poor quality samples
# poorQC_samples <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39") 
# sample_name <- sample_name[!(sample_name %in% poorQC_samples)] %>% str_sort(., numeric = T)
```

Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read BPcells matrices
# sample_name_all <- c(sample_name, paste0(sample_name, "_150"), paste0(sample_name, "_edrops50"))

sample_names <- rep(sample_name, 2) 
# sample_name_all <- c(sample_name, paste0(sample_name, "_150"), paste0(sample_name, "_edrops50"))
sample_name_all <- c(paste0(sample_name, "_cr"), paste0(sample_name, "_edrops50"), paste0(sample_name, "_cbendr"))

```


##################################################  START - SingleR use markers to id cells ##################################################
NNew samples multisample


```{r}
msc_qcd_anotd <- readRDS("../Mali/data/processed/DB52e_star/msc_qcd_anotd.RDS") 

```


```{r}
lp_NF_postQC <- readRDS("../published_dsets/lasse_portugal/jcr_outs/lp_NF_postQC.RDS") 

```

```{r}
lp_NF_postQC@meta.data %>% filter(n() <30, .by = pseudotime.stages ) %>% count(pseudotime.stages)

```

```{r}
lp_NF_postQC_mdata <- lp_NF_postQC@meta.data %>% 
  mutate("stageHL" = case_when(pseudotime.stages == "EarlyRing" ~ 'Early ring',
                                pseudotime.stages == "LateRing" ~ 'Late ring',
                                pseudotime.stages == "EarlyTrophozoite" ~ 'Early trophozoite',
                                pseudotime.stages == "MidTrophozoite" ~ 'Mid trophozoite',
                                pseudotime.stages == "LateTrophozoite" ~ 'Late trophozoite',
                                pseudotime.stages == "DevelopingGametocyte" ~ 'Gametocyte (developing)',
                                pseudotime.stages == "FemaleGametocyte" ~ 'Female'))

lp_NF_postQC_mdata %>% count(pseudotime.stages)
```



```{r}
lp_NF_postQC_mdata_gt30 <- lp_NF_postQC_mdata %>% filter(n() >30, .by = pseudotime.stages ) 
lp_NF_postQC_mdata_gt30%>% count(pseudotime.stages)

```

```{r}
lp_NF_postQC_gt30 <- lp_NF_postQC[, rownames(lp_NF_postQC_mdata_gt30)] %>% 
  AddMetaData(., lp_NF_postQC_mdata_gt30[, "stageHL", drop = F])

```

```{r}
lp_NF_postQC_gt30_list <-  SplitObject(lp_NF_postQC_gt30, split.by = "orig.ident") %>% set_names(as.character(unique(lp_NF_postQC_gt30$orig.ident)))
```

```{r, message=F, warning=F}
# msc_qcd_anotd_sce <- map(msc_qcd_anotd, ~{
msc_lp_sce <- map(c(msc_qcd_anotd, lp_NF_postQC_gt30_list), ~{
  obj <- as.SingleCellExperiment(.x)
  logcounts(obj) <- NULL
  normcounts(obj) <- NULL
  obj <- logNormCounts(obj)
  
  })

```

```{r, message=F, warning=F}
# saveRDS(msc_qcd_anotd_sce[c("MSC13", "MSC14")], "data/processed/Pf/m2_all/annot_refs/msc_qcd_anotd_sce_13_14.RDS")

```

```{r, message=F, warning=F}
msc_lp_sce <- msc_lp_sce[!(names(msc_lp_sce) %in% c("MSC1", "MSC3"))]
saveRDS(msc_lp_sce, "data/processed/Pf/m2_all/annot_refs/msc_lp_sce.RDS")

```
#### Get singleR classic markers for multi-reference https://bioconductor.org/books/devel/SingleRBook/using-multiple-references.html

```{r, message=F, warning=F}
com_markers_localMac <- getClassicMarkers(
    ref = msc_lp_sce,
    labels = map(msc_lp_sce, ~.x$stageHL))

com_markers_localMac %>% str
```


```{r, message=F, warning=F}
## Save for farm
saveRDS(com_markers_localMac, "data/processed/Pf/m2_all/annot_refs/com_markers_localMac.RDS")

```


```{r}
# bsub < /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code_lnk/bash_scripts/submsn_seu_norm_CbEdCr_singleR.sh

```

```{r, message=F, warning=F}
# slct_sample_names <-c("MSC24", "MSC68", "MSC14", "MSC49", "MSC55") %>% str_sort(., numeric = T)
sample_name_ed_cr <- c(paste0(sample_name, "_edrops50"), paste0(sample_name, "_cbendr"))
sample_name_ed_cr
```

Read in cell annotation singleR processed in farm
```{r, message=F, warning=F}
##Read elbo plots
# all_seu_norm_elbo <- list.files("data/processed/Pf", pattern = "*seu_norm_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|bpc'))) %>% 

# all_ed_pref_elbo <- list.files("data/processed/Pf", pattern = "bpseu_ed_pref.*_sct_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
all_pref_pred <- list.files("data/processed/Pf", pattern = "bpc_seu_(cr|cbendr|edrops50)_pred_sR.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpc_seu|_pred.*'))) %>%
# all_pref_pred <- list.files("data/processed/Pf", pattern = "bpseu_.*pred_sR.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpseu|_pref.*'))) %>% 
  .[sample_name_all] %>%
  # .[sample_name_ed_cr] %>%
  map(readRDS)

# all_pref_pred
```


```{r}
map(all_pref_pred[1:6], ~table(.x$labels))
```



```{r, eval=FALSE, fig.width=16, fig.height=10}
map(all_pref_pred[1:2], ~plotScoreHeatmap(.x))
```

```{r, eval=FALSE}
map2(all_pref_pred, all_ed_pref_sce[3], ~plotScoreHeatmap(.x, annotation_col=as.data.frame(colData(.y)[,"Is_Cell_50",drop=FALSE])))

```


```{r, eval=FALSE}
map(all_pref_pred, ~{
  to.remove <- is.na(.x$pruned.labels)
  to.remove
  table(Label=.x$labels, Removed=to.remove)
})


```



```{r, eval=FALSE}
map(all_pref_pred[1], ~plotDeltaDistribution(.x))

```




```{r}
# config <- configureMarkerHeatmap(all_pref_pred[[1]], all_ed_pref_sce[[3]], "epsilon")
# mat <- assay(all_ed_pref_sce[[3]], "logcounts")[head(config$rows, 20), config$columns]
# aggregated <- scuttle::summarizeAssayByGroup(mat, config$predictions)
# pheatmap::pheatmap(assay(aggregated), cluster_col=FALSE)
```


```{r}
# map(all_pref_pred, all_ed_pref_sce[3], ~plotMarkerHeatmap(.x, .y, "Male"))

```

```{r}
# collected <- list()
# for (lab in unique(all_pref_pred[[1]]$labels)) {
#   print(lab)
#   collected[[lab]] <- SingleR::plotMarkerHeatmap(results = predictions, test = all_ed_pref_sce[[3]], label = lab)[[4]]
# }

```


```{r}
# config <- configureMarkerHeatmap(predictions, m13_qcd_anotd_sce, "epsilon")
# mat <- assay(m13_qcd_anotd_sce, "logcounts")[head(config$rows, 20), config$columns]
# aggregated <- scuttle::summarizeAssayByGroup(mat, config$predictions)
# pheatmap::pheatmap(assay(aggregated), cluster_col=FALSE)
```


```{r}
# tab <- table(cluster=m13_qcd_anotd_sce$seurat_clusters, label=predictions$labels) 
# pheatmap::pheatmap(log10(tab+10)) # using a larger pseudo-count for smoothing. 
```


```{r}
# scater::plotUMAP(m13_qcd_anotd_sce, colour_by = "seurat_clusters")
```

################## evaluate singleR MAnually

```{r, message=FALSE}
## get predictions for stages
all_pref_pred_df <- map(all_pref_pred, ~.x[, c("scores", "labels", "pruned.labels", "reference")] %>% 
                          as.data.frame() %>% 
                          rename_with(~str_replace_all(., c("\\.\\."="_", "\\."="_", "_$"="")), contains("score")) %>%
                          rownames_to_column("bcode"))
```


```{r, message=FALSE, warning=FALSE}
all_pref_pred_df <- map2(all_pref_pred_df, all_pref_pred, ~{
  df <- .x %>% 
    mutate(sr_score_ref = paste0("scores_", names(.y@listData$orig.results)[reference], "_scores"),
           sr_labl_ref = paste0("scores_", names(.y@listData$orig.results)[reference], "_labels")) 
    
  # df$sR_actl_score <- mapply(function(row, colname) df[row, colname], row = seq_len(nrow(df)), colname = df$sr_score_ref) ## baseR

  df <- df %>% 
    mutate(actl_score = purrr::map2_dbl(row_number(), sr_score_ref, ~df[.x, .y]),
           actl_labl = purrr::map2_chr(row_number(), sr_labl_ref, ~df[.x, .y]),
           MaleFem_consistency = case_when(if_any(ends_with("_labels"), ~str_detect(., "Female")) & if_any(ends_with("_labels"), ~str_detect(., "Male")) ~ "MaleFem",
                                              T ~ labels))
    

  return(df)
  
  
})

```

```{r, message=FALSE, warning=FALSE}

map(all_pref_pred_df, ~.x %>% count(labels, sR_actl_labl, sR_MaleFem_consistency, sR_MaleFem_consistencym1314) %>% filter(sR_MaleFem_consistency == "MaleFem"))
```


```{r, message=FALSE, warning=FALSE}

all_pref_pred_df_fnl <- map(all_pref_pred_df, ~.x %>% 
                                         # filter(!is.na(actl_score)) %>% 
                                         select("bcode", "labels", "pruned.labels", "reference", "actl_labl", "actl_score", "MaleFem_consistency") %>% 
                                         mutate(fine_stg = case_when(is.na(pruned.labels) ~ "weak_score",
                                                                     MaleFem_consistency == "MaleFem" ~ "discrepantFM",
                                                                     T ~ pruned.labels)) %>%
                                         rename_with(~paste0("sR_", .), .cols = -c(bcode)))
  
```

################## END evaluate singleR MAnually

```{r, message=F, warning=F}
##Read BPcells matrices
all_seu <- list.files("data/processed/Pf", pattern = "bpc_seu_(cr|cbendr|edrops50).RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpc_seu|.RDS'))) %>% 
# all_seu_norm <- list.files("data/processed/Pf", pattern = "bpc_seu.*_sct.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpc_seu|_sct.*'))) %>% 
  .[sample_name_all] %>%
  map(readRDS)

```


```{r, message=FALSE}
##Doublet removal
# all_seu_prelim_stg_norm <- list(all_ed_pref_norm, all_pref_pred_df_com_scores_fnl, all_pref_dblt_clst_df, str_remove(names(all_ed_pref), "_.*")) %>%
all_seu_prelim_stg <- list(all_seu, all_pref_pred_df_fnl) %>%
  pmap(~ {
    ..2 %>%
      column_to_rownames("bcode") %>%
      AddMetaData(..1, .)
       })

```


```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
imap(all_seu_prelim_stg, ~saveRDS(.x, paste0("data/processed/Pf/",str_remove(.y, "_[cr|ed|cb].*"), "/seu_obj/bpseu", str_extract(.y, "_[cr|ed|cb].*"), "_prelim_stg.RDS")))

```

##################################################  END - SingleR use markers to id cells ##################################################


!!!NB - FARM CODE - DONT DELETE

```{r, eval=T}
gogo()
# bsub < /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code_lnk/bash_scripts/submsn_seu_norm_CbEdCr_bsub.sh
# bsub < /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code_lnk/bash_scripts/submsn_seu_sct_CbEdCr_bsub.sh
```




```{r, message=F, warning=F}
##Read BPcells matrices
all_seu_prelim_stg_norm <- list.files("data/processed/Pf", pattern = "bpseu_(cr|cbendr|edrops50)_prelim_stg_norm.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpseu|_prelim.*'))) %>% 
# all_seu_norm <- list.files("data/processed/Pf", pattern = "bpc_seu.*_sct.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpc_seu|_sct.*'))) %>% 
  .[sample_name_ed_cr] %>%
  map(readRDS)

```


Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read elbo plots
all_seu_prelim_stg_norm_elbo <- list.files("data/processed/Pf", pattern = "bpseu_(cr|cbendr|edrops50)_prelim_stg_norm_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpseu|_prelim.*'))) %>% 
# all_seu_norm_elbo <- list.files("data/processed/Pf", pattern = "bpc_seu.*_sct_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|bpc_seu|_sct.*'))) %>% 
  .[sample_name_ed_cr] %>%
  map(readRDS)

all_seu_prelim_stg_norm_elbo
```


```{r}
map(all_seu_prelim_stg_norm, ~{table(.x@meta.data$nFeature > 100)})
```

```{r, warning=FALSE, message=FALSE}
imap(all_seu_prelim_stg_norm, ~DimPlot(.x, dims = c(1,2), label = T) + labs(title = .y))

```

```{r, warning=FALSE, message=FALSE}
imap(all_seu_prelim_stg_norm, ~DimPlot(.x, dims = c(1,2), label = T, group.by = "sR_fine_stg") + labs(title = .y))

```

```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_seu_prelim_stg_norm, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T) + labs(title = .y)}, "missing"))
imap(all_seu_prelim_stg_norm, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```


```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_seu_prelim_stg_norm, possibly(~{FeaturePlot(.x, features = "PF3D7-1215500", order = T) + labs(title = .y)}, "missing"))
imap(all_seu_prelim_stg_norm, possibly(~{FeaturePlot(.x, features = "PF3D7-1215500", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE}
## remove the _150
all_seu_prelim_stg_norm <- all_seu_prelim_stg_norm[!str_detect(names(all_seu_prelim_stg_norm), "_150")]

```


```{r, warning=FALSE, message=FALSE}
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "nFeature_RNA") + labs(title = .y))
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "nFeature_RNA", reduction = "pca") + labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}
## Pfs16
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-0406200") + labs(title = .y))
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-0406200", reduction = "pca") + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
## sbp1
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-0501300") + labs(title = .y))
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-0501300", reduction = "pca") + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
## mget
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-1469900") + labs(title = .y))
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-1469900", reduction = "pca") + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-0903800") + labs(title = .y))
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, features = "PF3D7-0903800", reduction = "pca") + labs(title = .y))
```



```{r}

# ncount_filt <- unlist(map(all_seu_prelim_stg_norm, ~quantile(.x@meta.data[, "nCount_RNA"], 0.999, na.rm = T)))
```


```{r, results='asis'}
std_theme <- theme(axis.text=element_text(size=18), 
            axis.title=element_text(size=18,face="bold"), 
            legend.text = element_text(size=22), 
            legend.title = element_text(size=22,face="bold"))


```

```{r, results='asis'}
# pmap(list(all_seu_prelim_stg_norm, ncount_filt, names(all_seu_prelim_stg_norm)), ~{
#     VlnPlot(..1, feature = "nCount_RNA", group.by = "seurat_clusters", pt.size = 1)+ 
#       # scale_color_manual(values = sp_fld_colors) +
#       geom_hline(yintercept = ..2)  +
#       labs(title = ..3) +
#       std_theme
# 
#   })

```

```{r, results='asis'}
gogo()
```

## !!!!!!NB - RECONSIDER FILTRATION PROPERLY LATER
```{r, eval=FALSE}
## !!!!!!NB -CONSIDER FORE REMOVAL LATER
#filter out cells with high ncountRNA 
# all_seu_prelim_stg_norm <- all_seu_prelim_stg_norm %>%
#   map2(., ncount_filt, ~.x  %>%
#          subset(., nCount_RNA < .y )
#   )

```

Get mt_filt based on automatic removal of quantiles
```{r}
all_seu_prelim_stg_norm <- all_seu_prelim_stg_norm%>% 
  map(., ~{
    
    mit_fts <- rownames(.x)[str_detect(rownames(.x), "MIT")]
    api_fts <- rownames(.x)[str_detect(rownames(.x), "API")]
    
    .x %>% PercentageFeatureSet(., features = mit_fts, col.name = 'percent.mt') %>%
    PercentageFeatureSet(., features = api_fts, col.name = 'percent.api') 
})

```

```{r}
mt_filt <- unlist(map(all_seu_prelim_stg_norm, ~quantile(.x@meta.data[, "percent.mt"], 0.995, na.rm = T)))
# mt_filt <- unlist(map(all_seu_prelim_stg_norm, ~quantile(.x@meta.data[, "percent.mt"], 0.99, na.rm = T)))
# api_filt <- unlist(map(all_seu_prelim_stg_norm, ~quantile(.x@meta.data[, "percent.api"], 0.995, na.rm = T)))

```


```{r, results='asis'}

pmap(list(all_seu_prelim_stg_norm, mt_filt, names(all_seu_prelim_stg_norm)), ~{
    FeatureScatter(..1, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "seurat_clusters", pt.size = 1)+ 
      scale_color_manual(values = sp_fld_colors) +
      geom_hline(yintercept = ..2)  +
      labs(title = ..3) +
      std_theme

  })

```

```{r}
map2(all_seu_prelim_stg_norm, mt_filt, ~.@meta.data %>% dplyr::count(percent.mt>.y ))
```

## !!!!!!NB - RECONSIDER FILTRATION PROPERLY LATER
```{r, eval=FALSE}
## !!!!!!NB -CONSIDER FORE REMOVAL LATER
#filter out cells with high mitochondrial count
all_seu_prelim_stg_norm <- all_seu_prelim_stg_norm %>%
  map2(., mt_filt, ~.x  %>%
         subset(., percent.mt < .y )
  )

```


```{r, results='asis'}
pmap(list(all_seu_prelim_stg_norm, mt_filt, names(all_seu_prelim_stg_norm)), ~{
    FeatureScatter(..1, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "seurat_clusters", pt.size = 1)+ 
      scale_color_manual(values = sp_fld_colors) +
      geom_hline(yintercept = ..2)  +
      labs(title = ..3) +
      std_theme

  })

```



```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
plts_seu <- imap(all_ed_pref_norm, possibly(~DimPlot(.x, label = T) +  theme_void() + theme(legend.position = "none") + labs(title = .y), "no cells"))

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
## in cellbender but not in edrops
plot_grid(plotlist = plts_seu[1:28], ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
## in cellbender but not in edrops
plot_grid(plotlist = plts_seu[29:56], ncol = 4)

```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
plts <- imap(all_ed_pref_norm, possibly(~DimPlot(.x, label = T, cells.highlight =  colnames(.x[,.x@meta.data$nFeature_RNA <100]), sizes.highlight = 0.05) + 
                                     theme(legend.position = "none",
                                           axis.text = element_blank(),
                                           axis.title = element_blank()) +
                                     labs(title = .y), "no cells"))

```


```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
## in cellbender but not in edrops
plot_grid(plotlist = plts[1:28], ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
plot_grid(plotlist = plts[29:56], ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
## in cellbender but not in edrops
plts[[29]]
```


```{r}
# !!!!NOTE - MOVE A NOTEBOOK UP
all_seu_prelim_stg_norm_cb_mdata <- map(all_seu_prelim_stg_norm, ~.x@meta.data)
# saveRDS(all_seu_prelim_stg_norm_cb_mdata, "data/processed/Pf/all_seu_prelim_stg_norm_cb_mdata.csv")
species <- "Pf"
sv_dir <- "pbulk_gtypes_preQC_resc"
system(paste0("mkdir -p data/processed/", species, "/" , sv_dir))

saveRDS(all_seu_prelim_stg_norm_cb_mdata, paste0("data/processed/", species, "/" , sv_dir, "/all_seu_prelim_stg_norm_cb_mdata.csv"))

```


```{r, warning=FALSE, message=FALSE, fig.width=16, fig.height=4}
# imap(all_ed_pref_norm, ~FeaturePlot(.x, label = T, features = c("nFeature_RNA", "nCount_RNA"), ncol = 4, dims = c(1,3)) + 
#        labs(title = .y) )

```

```{r, warning=FALSE, message=FALSE, fig.width=16, fig.height=4}
imap(all_seu_prelim_stg_norm, ~FeaturePlot(.x, label = T, features = c("nFeature_RNA", "nCount_RNA"), ncol = 4) + 
       labs(title = .y) )

```



```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in edrops but not in cellbender
ed_notin_cb <- pmap(list(
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "edrops")],
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "cbendr")],
  names(all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "edrops")])),~{
  DimPlot(..1, label = T, cells.highlight = colnames(..1)[!colnames(..1) %in% colnames(..2)], sizes.highlight = 0.1, label.size = 6) + 
      theme_void() +
      theme(legend.position = "none", text = element_text(size =12)) +labs(title = ..3)
})

```


```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
## in cellbender but not in edrops
plot_grid(plotlist = ed_notin_cb, ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
cb_notin_ed <- pmap(list(
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "edrops")],
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "cbendr")],
  names(all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "cbendr")])),~{
  DimPlot(..2, label = T, cells.highlight = colnames(..2)[!colnames(..2) %in% colnames(..1)], sizes.highlight = 0.1, label.size = 6) + 
      theme_void() +
      theme(legend.position = "none", text = element_text(size =12)) +labs(title = ..3)
})

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
plot_grid(plotlist = cb_notin_ed, ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in both edrops and cellbender
ed_in_cb <- pmap(list(
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "edrops")],
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "cbendr")],
  names(all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "edrops")])), ~{
  DimPlot(..1, label = T, cells.highlight = colnames(..1)[colnames(..1) %in% colnames(..2)], sizes.highlight = 0.01, label.size = 6) + 
      theme_void() +
      theme(legend.position = "none", text = element_text(size =12), 
            # axis.title = element_blank(), axis.text = element_blank()
            ) +labs(title = ..3)  
})

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
plot_grid(plotlist = ed_in_cb, ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=5}
## in both cellbender and edrops
cb_in_ed <- pmap(list(
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "cbendr")],
  all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "edrops")],
  names(all_seu_prelim_stg_norm[str_detect(names(all_seu_prelim_stg_norm), "cbendr")])), ~{
  DimPlot(..1, label = T, cells.highlight = colnames(..1)[colnames(..1) %in% colnames(..2)], sizes.highlight = 0.01, label.size = 6) + 
      theme_void() +
      theme(legend.position = "none", text = element_text(size =12), 
            # axis.title = element_blank(), axis.text = element_blank()
            ) +labs(title = ..3)  
})

```


```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
## in cellbender but not in edrops
plot_grid(plotlist = cb_in_ed, ncol = 4)

```



```{r, warning=FALSE, message=FALSE}
map(all_seu_prelim_stg_norm, ~DimPlot(.x, group.by = "mthds_union"))

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
plts_mthds <- imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = F, group.by = "mthds_union") +  theme_void() + theme(legend.position = "bottom") + labs(title = .y), "no cells"))

```

```{r, warning=FALSE, message=FALSE, fig.height= 15, fig.width= 12, eval=T}
plot_grid(plotlist = plts_mthds[1:28], ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.height= 15, fig.width= 12, eval=T}
plot_grid(plotlist = plts_mthds[29:56], ncol = 4)

```

```{r, warning=FALSE, message=FALSE, fig.height= 14, fig.width= 10, eval=T}
plot_grid(plotlist = plts_mthds[29:56], ncol = 4)
```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = T, cells.highlight =  colnames(.x[,.x@meta.data$nFeature_RNA <100]), sizes.highlight = 0.05) + 
                                     theme(legend.position = "none",
                                           text = element_text(size =20)) +
                                     labs(title = .y), "no cells"))

```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = T, cells.highlight =  colnames(.x[,.x@meta.data$status == "unassigned"]), sizes.highlight = 0.05) + 
                                     theme(legend.position = "none",
                                           text = element_text(size =20)) +
                                     labs(title = .y), "no cells"))

```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = F, group.by =  "fine_stg", label.size = 6) + 
                                     scale_color_manual(values = sp_col_new) +
                                     theme(legend.position = "none",
                                           text = element_text(size =20)) +
                                     labs(title = .y), "no cells"))

```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~FeaturePlot(.x, label = F, features =  "actl_score", label.size = 6, order = T) + 
                                     # scale_color_manual(values = sp_col_new) +
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(), 
                                           axis.text = element_blank()) +
                                     labs(title = .y), "no cells"))

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~FeaturePlot(.x, label = F, features =  "nCount_RNA", label.size = 6, order = T) + 
                                     # scale_color_manual(values = sp_col_new) +
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(), 
                                           axis.text = element_blank()) +
                                     labs(title = .y), "no cells"))

```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(nFeature_RNA <100 , status)
     )

```



```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
gogo()
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## in cellbender but not in edrops
soupc_status_cols <- c(singlet = "#f79256", doublet = "#048ba8", unassigned = "#a8dadc")
```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.8}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = F, group.by =  "status", label.size = 6, order = T) + 
                                     scale_color_manual(values = soupc_status_cols) +
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(), 
                                           axis.text = element_blank()) +
                                     labs(title = .y), "no cells"))

```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## in cellbender but not in edrops
soupc_status_barp <- imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(nFeature_RNA <100 , status) %>%
       rename("gn_less100" = `nFeature_RNA < 100`) %>%
       group_by(gn_less100) %>%
       mutate(prop_n = round(n/sum(n), 2))
     ) %>%
  bind_rows(., .id = "donor") %>%
  filter(str_detect(donor, "cbendr")) %>%
  ungroup() %>%
  ggplot(aes(x = gn_less100, y = prop_n, fill = status)) +
  geom_col(position = position_dodge()) +
  geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.15))) +
  scale_fill_manual(values = soupc_status_cols) +
  facet_grid(. ~donor)+
  theme_classic() + 
  theme(text = element_text(size =20))

soupc_status_barp
```


```{r, warning=FALSE, message=FALSE, fig.width=40, fig.height=5}
## in cellbender but not in edrops
soupc_status_barp+ 
  theme(legend.position = "bottom")
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## in cellbender but not in edrops
soupc_scdblf_barp <- imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(class , status) %>%
       rename("sdbfindr" = class) %>%
       group_by(sdbfindr) %>%
       mutate(prop_n = round(n/sum(n), 2))
     ) %>%
  bind_rows(., .id = "donor") %>%
  filter(str_detect(donor, "cbendr")) %>%
  ungroup() %>%
  ggplot(aes(x = sdbfindr, y = prop_n, fill = status)) +
  geom_col(position = position_dodge()) +
  geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.15))) +
  scale_fill_manual(values = soupc_status_cols) +
  facet_grid(. ~donor)+
  theme_classic() + 
  theme(text = element_text(size =20))

soupc_scdblf_barp
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## in cellbender but not in edrops
soupc_scdblf_barp <- imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count( status) %>%
       mutate(prop_n = round(n/sum(n), 2))
     ) %>%
  bind_rows(., .id = "donor") %>%
  filter(str_detect(donor, "cbendr")) %>%
  ungroup() %>%
  ggplot(aes(x = status, y = prop_n, fill = status)) +
  geom_col(position = position_dodge()) +
  geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 5, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.15))) +
  scale_fill_manual(values = soupc_status_cols) +
  facet_grid(. ~donor)+
  theme_classic() + 
  theme(text = element_text(size =20), 
        axis.text.x = element_blank())

soupc_scdblf_barp
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## in cellbender but not in edrops
fn_stg_barp <- imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(nFeature_RNA <100 , fine_stg) %>%
       rename("gn_less100" = `nFeature_RNA < 100`) %>%
       group_by(gn_less100) %>%
       mutate(prop_n = round(n/sum(n), 2))
     ) %>%
  bind_rows(., .id = "donor") %>%
  filter(str_detect(donor, "cbendr")) %>%
  ungroup() %>%
  ggplot(aes(x = gn_less100, y = prop_n, fill = fine_stg)) +
  geom_col(position = position_dodge()) +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 4.4, show.legend = F) +
  scale_fill_manual(values = sp_col_new) +
  facet_grid(. ~donor)+
  theme_classic() + 
  theme(text = element_text(size =20))

fn_stg_barp
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## in cellbender but not in edrops
fn_stg_barp+ 
  theme(legend.position = "bottom")
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=4}
## in cellbender but not in edrops
soupc_stg_barp <- imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(status, fine_stg) %>%
       group_by(status) %>%
       mutate(prop_n = round(n/sum(n), 2))
     ) %>%
  bind_rows(., .id = "donor") %>%
  filter(str_detect(donor, "cbendr")) %>%
  ungroup() %>%
  ggplot(aes(x = status, y = prop_n, fill = fine_stg)) +
  geom_col(position = position_dodge()) +
  # geom_text(aes(label=n), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 4.4, show.legend = F) +
  scale_fill_manual(values = sp_col_new) +
  facet_grid(. ~donor)+
  theme_classic() + 
  theme(text = element_text(size =20),
        axis.text.x = element_text(size =20, angle = 90))

soupc_stg_barp
```

```{r, fig.width = 14, fig.height = 3, message=FALSE, warning=FALSE}
imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(nFeature_RNA <100 , fine_stg) %>%
       rename("gn_less100" = `nFeature_RNA < 100`) %>%
       group_by(gn_less100) %>%
       mutate(prop_n = round(n/sum(n), 2))
     ) %>%
  bind_rows(., .id = "donor") %>%
  filter(str_detect(donor, "cbendr")) %>%
  ungroup() %>%
  ggplot(aes(x = gn_less100, y = prop_n, fill = fine_stg)) +
  geom_col(position= position_dodge(width = .9)) +
  scale_fill_manual(values =  sp_col_new, guide = guide_legend( direction = "horizontal"  ))+
  geom_text(aes(label=n), angle=90, hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 4.4, show.legend = F)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.2))) +
  scale_x_discrete(expand = expansion(add = c(1.2, 0.8))) +
  # facet_grid(.~Strain_DN) +
  ggh4x::facet_nested_wrap(vars(donor, gn_less100),  nrow = 1, scales = "free_x") +
  theme_classic() +
  labs( y = 'Counts (X100)')+
      theme(
        axis.text.y  = element_text(size = 14),
        axis.title.y  = element_text(size = 14),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 14),
        legend.title = element_text(size = 14, face = "bold"),
        legend.position="bottom",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        strip.text =element_text(size=15, colour = 'black'),
        panel.border=element_rect(colour="black",fill='transparent')#,
      ) +
    guides(fill = guide_legend( nrow = 1, title.position = "left"))

```


```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.8}
## in cellbender but not in edrops
dblt_cols <- c(Singlet = "#ffd9b4", Doublet ="#6A5ACD")
```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.8}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = F, group.by =  "class", label.size = 6, order = T) + 
                                     scale_color_manual(values = dblt_cols) +
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(), 
                                           axis.text = element_blank()) +
                                     labs(title = .y), "no cells"))

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.5}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, ~.x@meta.data %>% 
       count(nFeature_RNA <100 , fine_stg)
     )


```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=4}
## in both edrops and cellbender
imap(all_seu_prelim_stg_norm, ~{
  .x@meta.data %>% 
    count(labels, class, Strain)
})

```


```{r, warning=FALSE, message=FALSE, fig.width=16, fig.height=4}
## ccp4
imap(all_seu_prelim_stg_norm, ~DimPlot(.x, label = T, group.by = c("labels", "class", "status", "seurat_clusters"), ncol = 4) + 
       labs(title = .y) )

```



```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height= 4}

imap(all_seu_prelim_stg_norm, ~{
  .x@meta.data %>%
    ggplot(aes(x = seurat_clusters, y= actl_score, color = class)) + 
    geom_boxplot() + 
    geom_hline(yintercept = 0.85) + 
    facet_wrap(. ~ labels, nrow = 1, scales = "free_x") +
    labs(title = .y)
    })

```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height= 4}

imap(all_seu_prelim_stg_norm, ~{
  .x@meta.data %>%
    ggplot(aes(x = seurat_clusters, y= actl_score, color = class)) + 
    geom_boxplot() + 
    geom_hline(yintercept = 0.85) + 
    facet_wrap(. ~ labels, nrow = 1, scales = "free_x") +
    labs(title = .y)
    })

```



```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.8}

imap(all_seu_prelim_stg_norm, ~{
  .x@meta.data %>%
    # filter(nFeature_RNA < 300) %>%
    ggplot(aes(x = log10(nFeature_RNA), y= log10(nCount_RNA), color = labels)) + 
    geom_point(size = 0.1) + 
    scale_color_manual(values = sp_col_new) +
    geom_hline(yintercept = log10(100)) + 
    geom_vline(xintercept = log10(100)) +
    geom_vline(xintercept = log10(80)) +
    labs(title = .y) +
    theme_classic()+
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           # axis.title = element_blank(), 
                                           # axis.text = element_blank()
                                           ) 
    })

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.8}

imap(all_seu_prelim_stg_norm, ~{
  .x@meta.data %>%
    filter(nFeature_RNA < 300) %>%
    ggplot(aes(x = log10(nFeature_RNA), y= log10(nCount_RNA), color = labels)) + 
    geom_point(size = 1) + 
    scale_color_manual(values = sp_col_new) +
    geom_hline(yintercept = log10(100)) + 
    geom_vline(xintercept = log10(100)) +
    geom_vline(xintercept = log10(80)) +
    labs(title = .y) +
    theme_classic()+
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(),
                                           axis.text = element_blank()
                                           ) 
    })

```

```{r, warning=FALSE, message=FALSE, fig.width=4, fig.height=4.8}
## in cellbender but not in edrops
imap(all_seu_prelim_stg_norm, possibly(~DimPlot(.x, label = F, group.by =  "class", label.size = 6, order = T) + 
                                     scale_color_manual(values = dblt_cols) +
                                     theme(legend.position = "none",
                                           text = element_text(size =20), 
                                           axis.title = element_blank(), 
                                           axis.text = element_blank()) +
                                     labs(title = .y), "no cells"))

```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height= 4}

imap(all_seu_prelim_stg_norm, ~{
  .x@meta.data %>%
    ggplot(aes(x = nFeature_RNA > 100, y= actl_score, color = status)) + 
    geom_sina() + 
    geom_hline(yintercept = 0.85) + 
    facet_wrap(. ~ labels, nrow = 1, scales = "free_x") +
    labs(title = .y)
    })

```

```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_ed_pref, ~DimPlot(.x, label = T) + labs(title = .y))

```


```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_ed_pref_norm, ~DimPlot(.x, label = T) + labs(title = .y))

```


```{r, warning=FALSE, message=FALSE}
# table(all_ed_pref[[5]]@assays[["SCT"]]@counts["PF3D7-1222600",] > 0)
# table(all_ed_pref[[5]]@assays[["SCT"]]@data["PF3D7-1222600",] > 0)

``` 

```{r, warning=FALSE, message=FALSE, eval=FALSE}
dense_mtx <- subset(all_ed[[5]], subset = `PF3D7-1222600` > 0)
dense_mtx
colnames(dense_mtx)
``` 

```{r, warning=FALSE, message=FALSE, eval=FALSE}
a = subset(all_ed_pref[[5]], subset = `PF3D7-1222600` > 0)
a
colnames(a)
``` 

dense_mtx

```{r, warning=FALSE, message=FALSE, eval=FALSE}
colnames(a) %in% colnames(dense_mtx)
``` 

!!!!NOTE - VERY VERY STRANGE RESULTS - RERUNNING SCT SEEMS TO CHANGE THE SCT RAW COUNTS???????? SOME INFORMATION ON RECLUSTERING  https://github.com/satijalab/seurat/issues/2014 , https://github.com/satijalab/seurat/issues/4967 
implemented by SCTransform argument 'do.correct.umi -Place corrected UMI matrix in assay counts slot; default is TRUE'
This correction made some ap2g counts in some cells to be zero (is it 0.something??)

```{r, warning=FALSE, message=FALSE, eval=FALSE}
data.frame("reclst" = a[, "TCGACGGAGGCGACAT-1"]@assays$SCT$counts) %>% 
  rownames_to_column("bcode") %>% 
  left_join(., data.frame("unclst" = dense_mtx[, "TCGACGGAGGCGACAT-1"]@assays$SCT$counts) %>% rownames_to_column("bcode")) %>%
  ggplot(aes(x = reclst, y = unclst)) +
  geom_point()
``` 

```{r, warning=FALSE, message=FALSE, eval=FALSE}
crd_rclst_m5 <- all_ed_pref[[5]]
DefaultAssay(crd_rclst_m5) <- "RNA"
crd_rclst_m5[["SCT"]] <- NULL
crd_rclst_m5[["RNA"]]$counts <- as(object = crd_rclst_m5[["RNA"]]$counts, Class = "dgCMatrix")

crd_rclst_m5_v3 <- CreateAssayObject(counts = crd_rclst_m5[["RNA"]]$counts)

crd_rclst_m5[["RNA"]] <- crd_rclst_m5_v3

crd_rclst_m5
``` 

```{r, warning=FALSE, message=FALSE, eval=FALSE}
table(crd_rclst_m5[["RNA"]]$counts["PF3D7-1222600",] > 0)

crd_rclst_m5_ap2g <- crd_rclst_m5[,crd_rclst_m5[["RNA"]]$counts["PF3D7-1222600",] > 0]
crd_rclst_m5_ap2g
``` 


```{r, warning=FALSE, message=FALSE, eval=FALSE}
data.frame(crd_m5_tr_ap2g[, "TCGACGGAGGCGACAT-1"]@assays$RNA@counts) %>% 
  rownames_to_column("bcode") %>% 
  left_join(., data.frame(crd_rclst_m5_ap2g[, "TCGACGGAGGCGACAT-1"]@assays$RNA@counts) %>% rownames_to_column("bcode"), by = "bcode") %>%
  ggplot(aes(x = TCGACGGAGGCGACAT.1.x, y = TCGACGGAGGCGACAT.1.y)) +
  geom_point()
``` 


```{r, warning=FALSE, message=FALSE, eval=FALSE}
dense_mtx
``` 

```{r, warning=FALSE, message=FALSE}
## male le marker
imap(all_ed_pref, possibly(~{FeaturePlot(.x, features = "PF3D7-1215500", order = T) + labs(title = .y)}, "missing"))
imap(all_ed_pref, possibly(~{FeaturePlot(.x, features = "PF3D7-1215500", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE}
imap(all_ed_pref, ~DimPlot(.x, label = T)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
## ap2g
all_ed_pref_m70 <- all_ed_pref[[5]]
Idents(all_ed_pref_m70) <- "seurat_clusters"
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
DimPlot(all_ed_pref_m70, label = T)
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
DimPlot(all_ed_pref_m70, label = T)
```


```{r, warning=FALSE, message=FALSE, eval=FALSE}
# go()
de_gns_seu <- FindAllMarkers(all_ed_pref_m70, test.use = "MAST")
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
de_gns_seu <- de_gns_seu %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_seu  %>% filter(cluster == "0")
```

```{r, warning=FALSE, message=FALSE, eval=F, eval=FALSE}
de_gns_seu <- de_gns_seu %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_seu
```

```{r, warning=FALSE, message=FALSE, eval=TRUE, eval=FALSE}
map(as.character(c(0:3)), ~de_gns_seu %>% filter(avg_log2FC > 0, cluster == .x))


```


```{r}
## top markers calculated from qcd msc13 and msc14 objects in local mac computer in notebook m2_RBP_motif.Rmd
top_fLE_m13_m14 <- c("PF3D7-1146100", "PF3D7-1430100", "PF3D7-1429300", "PF3D7-1338800",  "PF3D7-0418800", "PF3D7-0508000", "PF3D7-1316300", "PF3D7-1331600", "PF3D7-1214800", "PF3D7-1207300", "PF3D7-0421900", "PF3D7-1322900",  "PF3D7-1452900", "PF3D7-1115100", "PF3D7-0630000")

# filter(p_val_adj < 0.05 & avg_log2FC > 0 & pct.1 > 0.7 & pct.2 < 0.2)
top_mLE_m13_m14 <- c("PF3D7-0322800", "PF3D7-1215500", "PF3D7-1325300", "PF3D7-1408800", "PF3D7-1245300", "PF3D7-1329200")

# # filter(p_val_adj < 0.05 & avg_log2FC > 0 & pct.1 > 0.7 & pct.2 < 0.25)
# top_mLE_m13_m14 <- c("PF3D7-0111000", "PF3D7-0627000", "PF3D7-0322800", "PF3D7-1215500",  "PF3D7-1325300", "PF3D7-1408800", "PF3D7-1355600", "PF3D7-1303900", "PF3D7-1006100", "PF3D7-1245300", "PF3D7-1329200", "PF3D7-0717400")


top_asex_m14 <- c("PF3D7-0936000", "PF3D7-1301700", "PF3D7-0935900", "PF3D7-0501300", "PF3D7-1102700", "PF3D7-1401400", "PF3D7-0202500", "PF3D7-0532300", "PF3D7-1252300")

asex_2_m70 <- c("PF3D7-0102200", "PF3D7-1149200", "PF3D7-1352900", "PF3D7-0935600", "PF3D7-0501400", "PF3D7-0108500", "PF3D7-1102800", "PF3D7-0702000")
```


```{r, warning=FALSE, message=FALSE}

all_ed_pref <- map(all_ed_pref, ~NormalizeData(., assay = "RNA"))
```

```{r}
makeSetScore_bp <- function(seur, gene.set, assay='RNA') {
  # Get mean expression of genes of interest per cell
  
  seur[[assay]]$data <- as(object = seur[[assay]]$data, Class = "dgCMatrix")
  mean.exp <- colMeans(x = as.matrix(GetAssayData(seur, slot = 'data', assay = assay))[gene.set, ], na.rm = TRUE)
  
  # Add mean expression values in 'object@meta.data$gene.set.score'
  if (all(names(x = mean.exp) == rownames(x = seur@meta.data))) {
    return(mean.exp)
  }
}

```


```{r, warning=FALSE, message=FALSE}

asex_avg_exp <- list(all_ed_pref)  %>%
  pmap(~makeSetScore(..1, top_asex_m14, assay = 'RNA') %>%  data.frame %>%  setNames('asx_avg')  )

mle_avg_exp <- list(all_ed_pref)  %>%
  pmap(~makeSetScore(..1, top_mLE_m13_m14, assay = 'RNA') %>%  data.frame %>%  setNames('mle_avg')  )

fle_avg_exp <- list(all_ed_pref)  %>%
  pmap(~makeSetScore(..1, top_fLE_m13_m14, assay = 'RNA') %>%  data.frame %>%  setNames('fle_avg')  )

asex2_avg_exp <- list(all_ed_pref)  %>%
  pmap(~makeSetScore(..1, asex_2_m70, assay = 'RNA') %>%  data.frame %>%  setNames('asx2_avg')  )


markers_avg_exp <- pmap(list(asex_avg_exp, mle_avg_exp, fle_avg_exp, asex2_avg_exp), ~ bind_cols(..1, ..2, ..3, ..4))
# markers_avg_exp[[1]]

```

```{r, warning=FALSE, message=FALSE}

# asex_avg_exp <- list(all_ed_pref)  %>%
#   pmap(~makeSetScore(..1, top_asex_m14, assay = 'SCT') %>%  data.frame %>%  setNames('asx_avg')  )
# 
# mle_avg_exp <- list(all_ed_pref)  %>%
#   pmap(~makeSetScore(..1, top_mLE_m13_m14, assay = 'SCT') %>%  data.frame %>%  setNames('mle_avg')  )
# 
# fle_avg_exp <- list(all_ed_pref)  %>%
#   pmap(~makeSetScore(..1, top_fLE_m13_m14, assay = 'SCT') %>%  data.frame %>%  setNames('fle_avg')  )
# 
# asex2_avg_exp <- list(all_ed_pref)  %>%
#   pmap(~makeSetScore(..1, asex_2_m70, assay = 'SCT') %>%  data.frame %>%  setNames('asx2_avg')  )
# 
# 
# markers_avg_exp <- pmap(list(asex_avg_exp, mle_avg_exp, fle_avg_exp, asex2_avg_exp), ~ bind_cols(..1, ..2, ..3, ..4))
# # markers_avg_exp[[1]]

```

```{r, warning=FALSE, message=FALSE}

all_ed_pref <- list(all_ed_pref, markers_avg_exp) %>% pmap(~AddMetaData(..1, ..2))
```

Visualization of clusters with markers

```{r, warning=FALSE, message=FALSE, eval=T}
## cells that contain 2 or more ap2g UMI
imap(all_ed_pref, ~DimPlot(.x, label = T)+ labs(title = .y))

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=3}
## cells that contain 2 or more ap2g UMI
map(all_ed_pref, ~FeaturePlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg"), ncol = 4))
map(all_ed_pref, ~VlnPlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg", "nFeature_RNA", "nCount_RNA"), ncol = 6))

```

```{r}
# gn_uc <- c(1900, 2600, 3300, 3800, 3200)
# gn_lc <- c(250, 250, 250, 300, 250)

all_ed_pref <- pmap(list(all_ed_pref, rep('asx_avg', length(all_ed_pref)), rep('mle_avg', length(all_ed_pref)), rep('fle_avg', length(all_ed_pref)), rep('asx2_avg', length(all_ed_pref))), ~{
  ..1@meta.data %>%
    rownames_to_column('bcode') %>%
    group_by(seurat_clusters)%>%
    mutate("asx_avg_ci_cut" = quantile(!!rlang::sym(..2), 0.50, na.rm = T),
            "mle_avg_ci_cut" = quantile(!!rlang::sym(..3), 0.50, na.rm = T),
            "fle_avg_ci_cut" = quantile(!!rlang::sym(..4),  0.50, na.rm = T),
            "asx2_avg_ci_cut" = quantile(!!rlang::sym(..5),  0.50, na.rm = T)
    ) %>%
    # mutate("asx_avg_ci_cut" = quantile(!!rlang::sym(..2), 0.10, na.rm = T),
    #         "mle_avg_ci_cut" = quantile(!!rlang::sym(..3), 0.10, na.rm = T),
    #         "fle_avg_ci_cut" = quantile(!!rlang::sym(..4),  0.10, na.rm = T),
    #         "asx2_avg_ci_cut" = quantile(!!rlang::sym(..5),  0.10, na.rm = T)
    # ) %>%
    ungroup()%>%
    select(bcode,contains("ci_cut")) %>%
    column_to_rownames('bcode')%>%
    AddMetaData(..1, .)
})
```

```{r}
# gn_uc 

all_ed_pref <- map(all_ed_pref, ~{
  .x@meta.data %>%
    mutate("likely_asex" = ifelse(asx_avg > max(asx_avg_ci_cut), "asex", NA),
           "likely_mle" = ifelse(mle_avg > max(mle_avg_ci_cut), "mle", NA),
           "likely_fle" = ifelse(fle_avg > max(fle_avg_ci_cut), "fle", NA),
           "likely_asex2" = ifelse(asx2_avg > max(asx2_avg_ci_cut), "asex2", NA)
    ) %>%
    unite(col = "likely_stage", likely_asex, likely_mle, likely_fle, likely_asex2, remove = F, na.rm = T) %>%
    mutate(across(likely_stage, ~ifelse(.x == "", "unassigned", .x))) %>%
    AddMetaData(..1, .)
})
    
    
```

```{r}
map(all_ed_pref, ~.x@meta.data %>% dplyr::count(seurat_clusters, likely_stage))
```

```{r}
map(all_ed_pref, ~.x@meta.data %>% dplyr::count(seurat_clusters, likely_stage))
```

```{r}
# gn_uc 

# all_ed_pref <- map(all_ed_pref, ~{
#   .x@meta.data %>%
#     mutate("likely_stage" = case_when(asx_avg > max(asx_avg_ci_cut) & mle_avg < max(mle_avg_ci_cut) & fle_avg < max(fle_avg_ci_cut) ~ "likely asex",
#                                       asx_avg < max(asx_avg_ci_cut) & mle_avg > max(mle_avg_ci_cut) & fle_avg < max(fle_avg_ci_cut) ~ "likely mle",
#                                       asx_avg < max(asx_avg_ci_cut) & mle_avg < max(mle_avg_ci_cut) & fle_avg > max(fle_avg_ci_cut) ~ "likely fle",
#                                       asx_avg > max(asx_avg_ci_cut) & mle_avg < max(mle_avg_ci_cut) & fle_avg > max(fle_avg_ci_cut) ~ "likely asex fle",
#                                       asx_avg > max(asx_avg_ci_cut) & mle_avg > max(mle_avg_ci_cut) & fle_avg < max(fle_avg_ci_cut) ~ "likely asex mle",
#                                       asx_avg < max(asx_avg_ci_cut) & mle_avg > max(mle_avg_ci_cut) & fle_avg > max(fle_avg_ci_cut) ~ "likely mle fle")
#            ) %>%
#     AddMetaData(..1, .)
# })
#     
#     
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
## cells that contain 2 or more ap2g UMI
map(all_ed_pref, ~FeaturePlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg"), ncol = 4))
map(all_ed_pref, ~VlnPlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg", "nFeature_RNA", "nCount_RNA"), group.by = "likely_stage", ncol = 6))

```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=4}
map(all_ed_pref, ~VlnPlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg", "nFeature_RNA"),  split.by = "likely_stage", ncol = 5))

```

```{r, fig.width=14, fig.height=3}
map(all_ed_pref, ~{
.x@meta.data %>%
    pivot_longer(cols = ends_with("_avg"), values_to = "avg_exprsn", names_to = "stage_rescue") %>%
  # rename(c('Counts' =n, 'Strain' = StrainDon))  %>%
  # ggplot(., aes(x = weave_factors(Strain, donor), y = Counts, fill = donor)) +
  ggplot(., aes(x = seurat_clusters, y = avg_exprsn, fill = likely_stage)) +
  geom_boxplot() +
  facet_wrap(stage_rescue ~ ., nrow = 1, drop = T, scales = "free") +
  theme_classic()
  
})
```


```{r}
imap(all_ed_pref, ~DimPlot(.x, label = T)+ labs(title = .y))

imap(all_ed_pref, ~DimPlot(.x, group.by = "likely_stage")+ labs(title = .y))

```


```{r, warning=FALSE, message=FALSE}
## ap2g
imap(all_ed_pref, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T) + labs(title = .y)}, "missing"))

```


```{r, warning=FALSE, message=FALSE}
## mle marker
imap(all_ed_pref, possibly(~{FeaturePlot(.x, features = "PF3D7-1215500", order = T) + labs(title = .y)}, "missing"))

```

```{r}
## Adding annotations to query cells
all_ed_pref <- map(all_ed_pref, ~{
  .x@meta.data %>% 
    mutate("likely_stage_asx_merged" = ifelse(likely_stage %in% c("asex", "asex2", "asex_asex2", "asex2_asex"), "asex_merged", likely_stage )) %>%
    select(likely_stage_asx_merged) %>%
    AddMetaData(.x, .) 
 } )

```

```{r}
# gogo()
map(all_ed_pref, ~.x@meta.data %>% dplyr::count(seurat_clusters, likely_stage_asx_merged) %>% group_by(seurat_clusters) %>% mutate(prop_n = round(n/sum(n), 2)))
```

```{r}

## remove clusters with unassigned cells > 70%
clstrs_assnd_marked <- map(all_ed_pref, ~.x@meta.data %>% 
                             dplyr::count(seurat_clusters, likely_stage_asx_merged) %>% 
                             group_by(seurat_clusters) %>% 
                             mutate(prop_n_unassigned = round(n/sum(n), 2)) %>% 
                             mutate(likely_unassigned = ifelse(likely_stage_asx_merged == "unassigned" & prop_n_unassigned > 0.7, "yes", "no"),
                                    low_n = ifelse(sum(n) <= 30, "yes", "no"))
                           ) #%>% dplyr::filter(!any(likely_unassigned == "yes" ))) #%>% dplyr::filter(likely_stage_asx_merged != "unassigned")
clstrs_assnd_marked
```


```{r}
unassnd_clstrs <- map(clstrs_assnd_marked, possibly(~{.x %>% dplyr::filter(any(likely_unassigned == "yes" )) %>% pull(seurat_clusters) %>% unique() }, "missing"))
low_n_clstrs <- map(clstrs_assnd_marked, possibly(~{.x %>% dplyr::filter(low_n == "yes") %>% pull(seurat_clusters)}, "missing"))

# possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE, eval=T, fig.width=12, fig.height=3}
clustr_filt_highlt_fn <- function(seu_obj, clusters_highlt) { 
    if (length(clusters_highlt) != 0) {
      plot_grid(
        plot_grid(
          plotlist = FeaturePlot(seu_obj, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg"), ncol = 4, combine = F) %>%
            map(., ~.x + 
                  theme(legend.position = "bottom",
                        axis.title = element_blank())),
          nrow = 1
          ),
        DimPlot(seu_obj, label = T) + 
          theme(legend.position = "none",
                axis.title = element_blank()),
        DimPlot(seu_obj, label = T, cells.highlight = colnames(seu_obj[, seu_obj$seurat_clusters %in% clusters_highlt]))+ 
          theme(legend.position = "bottom",
                axis.title = element_blank()),
        ncol = 3,
        rel_widths = c(.7, .15, .15)
      )
  }
      
}

```

```{r, warning=FALSE, message=FALSE, eval=T, fig.width=12, fig.height=3}
pmap(list(all_ed_pref, unassnd_clstrs), ~{
  clustr_filt_highlt_fn(..1, ..2)
  })

```



```{r, warning=FALSE, message=FALSE, eval=T}
## cells that contain 2 or more ap2g UMI
# map2(all_ed_pref, unassnd_clstrs, possibly(~{DimPlot(.x, label = T, cells.highlight = colnames(.x[, .x$seurat_clusters %in% .y]))}, "missing"))
```


```{r}

## remove clusters with largely unassigned cells (> 70%) and those with less than 30 cells
clstrs_assnd <- map(clstrs_assnd_marked, ~.x %>% dplyr::filter(!any(likely_unassigned == "yes" ) & low_n == "no")) #%>% dplyr::filter(likely_stage_asx_merged != "unassigned")
```


```{r}
## filter unassigned cells so as to calculate proportions for doublet removal
clstrs_assnd_cln <- map(clstrs_assnd, ~.x %>% dplyr::filter(likely_stage_asx_merged != "unassigned")) # %>% dplyr::filter(any(prop_n > 0.85))) 

```

```{r, warning=FALSE, message=FALSE, eval=T}
## cells that contain 2 or more ap2g UMI
imap(all_ed_pref, ~DimPlot(.x, label = T)+ labs(title = .y))
```

```{r}
## remove clusters with unassigned cells > 70%
clstrs_assnd_cln_dblt_marked <- map(clstrs_assnd_cln, ~.x  %>%  mutate(prop_n = round(n/sum(n), 2), "likely_doublet" = ifelse(any(prop_n >0.85), "no", "yes"))) 
clstrs_assnd_cln_dblt_marked
```

```{r}
dblt_clstrs <- map(clstrs_assnd_cln_dblt_marked, possibly(~{.x %>% dplyr::filter(likely_doublet == "yes" ) %>% pull(seurat_clusters)}, "missing"))

```

```{r, warning=FALSE, message=FALSE, eval=T, fig.width=12, fig.height=3}
pmap(list(all_ed_pref, dblt_clstrs), ~{
  clustr_filt_highlt_fn(..1, ..2)
  })

```


```{r}
clstrs_assnd_cln_no_dblt <- map(clstrs_assnd_cln_dblt_marked, ~.x %>% dplyr::filter(likely_doublet == "no")) 

```


```{r}
## remove clusters with unassigned cells > 70%
clstrs_cln <- map(clstrs_assnd_cln_no_dblt, ~.x %>% slice_max(prop_n) %>% pull(seurat_clusters) ) 
likely_stage <- map(clstrs_assnd_cln_no_dblt, ~.x %>% slice_max(prop_n) %>% select(seurat_clusters, likely_stage_asx_merged) %>% deframe ) 
likely_stage_fine <- map(clstrs_assnd_cln_no_dblt, ~.x %>% slice_max(prop_n) %>% select(seurat_clusters, likely_stage_asx_merged, ) %>% deframe )
```


```{r}
# pmap(list(unassnd_clstrs, low_n_clstrs, dblt_clstrs), possibly(~{ c("unassnd" = unique(..1), "low_n" = unique(..2), "dblt" = unique(..3))} ), "missing")
poorQC_clstrs <- pmap(list(unassnd_clstrs, low_n_clstrs, dblt_clstrs),
                      # possibly(
                        ~{
                        c("unassnd" = unique(..1), "low_n" = unique(..2), "dblt" = unique(..3)) %>%
                          as.data.frame(., col.names= c("grp", "clstr")) %>%
                          rownames_to_column("grp") %>% 
                          dplyr::select(".", "grp") %>% 
                          # mutate(across(grp, ~str_remove(., "[0-9]+"))) %>% 
                          deframe()
                        } #), "missing"
                      )

```

```{r}
# pmap(list(unassnd_clstrs, low_n_clstrs, dblt_clstrs), possibly(~{ c("unassnd" = unique(..1), "low_n" = unique(..2), "dblt" = unique(..3))} ), "missing")
all_clustrs <- pmap(list(likely_stage, poorQC_clstrs), possibly(~{ c(..1, ..2)} ), "missing")

```


```{r}
# rules <- Map(reformulate, shQuote(likely_stage), shQuote(names(likely_stage)))
(all_clustrs_rules <- map(all_clustrs, ~glue::glue('"{lbl}" ~ "{val}"', lbl = names(.x) , val = .x)))
```


```{r, warning=FALSE, message=FALSE}
## label clusters with likely stage from majority of cells in cluster

all_ed_pref <- map2(all_ed_pref, all_clustrs_rules, ~{
  .x@meta.data %>%
    mutate(marker_annot = case_match(seurat_clusters, !!!rlang::parse_exprs(.y))) %>%
    select(marker_annot) %>%
    AddMetaData(.x, .)
})


```


```{r}
imap(all_ed_pref, possibly({~DimPlot(.x, group.by = "marker_annot", label = T) + labs(title = .y)}, "missing")) 

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_ed_pref, ~VlnPlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg", "nFeature_RNA", "nCount_RNA"), group.by = "marker_annot", ncol = 6) + plot_annotation(title = .y))

```

```{r, warning=FALSE, message=FALSE}
gogo()
```

```{r, message=FALSE, warning=FALSE, eval=T}
## Save as h5ad 

a<-imap(all_ed_pref, ~{
  
  ##Convert to v3 object since v5 not supported by seuratdisk 
  obj <- .x
  DefaultAssay(obj) <- "RNA"
  # obj[["RNA"]]$counts <- as(object = obj[["RNA"]]$counts, Class = "dgCMatrix")
  # print(obj)
  # obj[["RNA3"]] <- as(object = obj[["RNA"]], Class = "Assay")
  # DefaultAssay(obj) <- "RNA3"
  # obj[["RNA"]] <- NULL
  # obj <- RenameAssays(object = obj, RNA3 = 'RNA')
  obj[["RNA"]] <- CreateAssayObject(counts = obj[["RNA"]]$counts) #Convert Assay v5 to Assay
  obj[["SCT"]] <- NULL
  print(obj)
  
  SeuratDisk::SaveH5Seurat(obj, filename =  paste0("data/processed/Pf/",str_remove(.y, "_.*"), "/ed", str_extract(.y, "_.*"), "_pref.h5Seurat"), overwrite = T)

  Convert(paste0("data/processed/Pf/",str_remove(.y, "_.*"), "/ed", str_extract(.y, "_.*"), "_pref.h5Seurat"), dest = "h5ad", overwrite = T)
  
})

```



```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
nms_all_ed_pref_no_m24 <- names(all_ed_pref)[!names(all_ed_pref) %in% "MSC24"]
```


```{r, message=F, warning=F}
##Read BPcells matrices
# all_seu_prelim_stg_norm <- list.files("data/processed/Pf", pattern = "*seu_norm.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|bpc'))) %>% 

# cellassign_predictions <- list.files("data/processed/Pf", pattern = "cellassign_predictions.csv$", full.names = T, recursive = T, include.dirs = T) %>% 
cellassign_predictions <- list.files("data/processed/Pf", pattern = "cellassign_predictions_.*csv$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|/cellassign_predictions|.csv'))) %>% 
  # .[sample_name_all] %>%
  # .[nms_all_ed_pref_no_m24] %>%
  .[names(all_ed_pref)] %>%
  map(read_csv)

```

```{r}
mdata_scanvi <- read_csv("data/processed/Pf/m2_all/ed_pref_all_query_mdata.csv")
```

```{r, message=F, warning=F}
mdata_scanvi_list <- mdata_scanvi %>% 
  rename("bcode" = `...1`) %>% 
  mutate(across(bcode, ~str_extract(.,"[A-Z]+-1"))) %>% 
  mutate(scanvi_stg = case_when(dominant_stg == "unassigned" ~ "unassigned",
                                predictions_scanvi == dominant_stg ~ dominant_stg,
                                predictions_scanvi != dominant_stg ~ "discordant"
                                )) %>% 
  select(contains("scanvi"), "donor", "dominant_stg", "bcode", "scanvi_stg") %>% 
  split(., f = factor(.$donor))

```



```{r, warning=FALSE, message=FALSE}
## label clusters with likely stage from majority of cells in cluster

# all_ed_pref <- pmap(list(all_ed_pref[nms_all_ed_pref_no_m24], cellassign_predictions, mdata_scanvi_list[nms_all_ed_pref_no_m24]), ~{
all_ed_pref <- pmap(list(all_ed_pref, cellassign_predictions), ~{
  ..2 %>% 
    mutate(cAssign_stage = case_when(asex_merged > 0.9 ~ "asex_merged",
                                     fLE > 0.9 ~ "fLE",
                                     mLE > 0.9 ~ "mLE",
                                     T ~ "Other")) %>%
    select(-c(`...1`)) %>%
    # left_join(., ..3 , by = "bcode") %>%
    column_to_rownames("bcode") %>%
    AddMetaData(..1, .)
  
})


```


```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_ed_pref, ~VlnPlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg", "nFeature_RNA", "nCount_RNA"), group.by = "cAssign_stage", ncol = 6) + plot_annotation(title = .y))

```


```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_ed_pref, ~.x@meta.data %>% count(seurat_clusters, dominant_stg, predictions_scanvi, scanvi_stg) )

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_ed_pref, ~VlnPlot(.x, features = c("asx_avg", "mle_avg", "fle_avg", "asx2_avg", "nFeature_RNA", "nCount_RNA"), group.by = "scanvi_stg", ncol = 6) + plot_annotation(title = .y))

```


```{r, warning=FALSE, message=FALSE, fig.width=6, fig.height=4}
imap(all_ed_pref, ~{
  seu <- .x
  nm <-.y
  VlnPlot(seu, features = c( "nFeature_RNA", "nCount_RNA"), group.by = "seurat_clusters", ncol = 6, combine = F) %>% imap(., ~.x + labs(title = nm) +geom_hline(yintercept = 100))
  })

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
gogo()
```

```{r, warning=FALSE, message=FALSE}
sample_rescd <- names(likely_stage[unlist(map(likely_stage, ~length(.x) > 0))])
```

```{r}
## remove clusters with unassigned cells > 70%
clstrs_cln <- clstrs_cln[sample_rescd] 
likely_stage <- likely_stage[sample_rescd] 
```

```{r}
## remove clusters with unassigned cells > 70%
all_ed_pref_resc <- all_ed_pref[sample_rescd] 
```

```{r}

# all_ed_pref_resc <- map2(all_ed_pref, clstrs_cln, ~subset(.x, subset = seurat_clusters %in% .y))
all_ed_pref_resc <- map2(all_ed_pref_resc, clstrs_cln, ~subset(.x, subset = seurat_clusters %in% .y)) 

```

```{r}
# rules <- Map(reformulate, shQuote(likely_stage), shQuote(names(likely_stage)))
(likely_stage_rules <- map(likely_stage, ~glue::glue('"{lbl}" ~ "{val}"', lbl = names(.x) , val = .x)))
```


```{r, warning=FALSE, message=FALSE}
all_ed_pref_resc <- all_ed_pref_resc[sample_rescd]
```

```{r, warning=FALSE, message=FALSE}
## label clusters with likely stage from majority of cells in cluster

all_ed_pref_resc <- map2(all_ed_pref_resc, likely_stage_rules, ~{
  .x@meta.data %>%
    mutate(marker_stage = case_match(seurat_clusters, !!!rlang::parse_exprs(.y))) %>%
    select(marker_stage) %>%
    AddMetaData(.x, .)
})


```

```{r}
map(all_ed_pref_resc, ~.x@meta.data %>% dplyr::count(marker_stage, likely_stage_asx_merged))
```

```{r}
imap(all_ed_pref_resc, possibly({~DimPlot(.x, group.by = "marker_stage", label = T) + labs(title = .y)}, "missing")) 

```

```{r}
imap(all_ed_pref_resc, possibly({~DimPlot(.x, label = T) + labs(title = .y)}, "missing"))

```


```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
### gn_edrops50
imap(all_ed_pref_resc, ~saveRDS(.x, paste0("data/processed/Pf/",str_remove(.y, "_edrops50"), "/seu_obj/ed_pref_resc.RDS")))
```

## SCT rerun for rescued cells subcluster

```{r}
## write only with more than 100 cells - low number of cells gives error in nextflow sct normalization run 
nms_all_ed_pref_resc_100 <- names(all_ed_pref_resc)[unlist(map(all_ed_pref_resc, ~dim(.x)[2] > 100))]
nms_all_ed_pref_resc_100
```


```{r}
## write only with more than 100 cells - low number of cells gives error in nextflow sct normalization run 
nms_all_ed_pref_resc_less100 <- names(all_ed_pref_resc)[unlist(map(all_ed_pref_resc, ~dim(.x)[2] <= 100))]
nms_all_ed_pref_resc_less100
```


```{r, eval=T}
## names to paste for nextflow run
paste(nms_all_ed_pref_resc_100, collapse = ",")
```


```{r, eval=T}
gogo()
## Run standard normalisation on farm like below
## !!NB - reducing the number of hypervariable genes results in scattered and disjointed clusters. higher hvg number better

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sctnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC40,MSC48,MSC49,MSC54,MSC55,MSC60,MSC67,MSC68,MSC70}/seu_obj/ed_pref_resc.RDS" --normlzn "sct" --umap_name "UMAP_" --hvg 200 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --resume

```

```{r, eval=T}
## names to paste for nextflow run
nms_all_ed_pref_resc_100 <- c("MSC40", "MSC48", "MSC49", "MSC54", "MSC55", "MSC60", "MSC67", "MSC68", "MSC70")
```

```{r, message=F, warning=F}
##Read BPcells matrices
# all_seu_prelim_stg_norm <- list.files("data/processed/Pf", pattern = "*seu_norm.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|bpc'))) %>% 

all_resc_sct <- list.files("data/processed/Pf", pattern = "ed_cr_dropd_resc.*_sct.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|ed_cr_dropd_resc|_sct.*'))) %>% 
  # .[sample_name_all] %>%
  .[nms_all_ed_pref_resc_100] %>%
  map(readRDS)

```


Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read elbo plots
# all_seu_norm_elbo <- list.files("data/processed/Pf", pattern = "*seu_norm_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|bpc'))) %>% 

all_resc_sct_sct_elbo <- list.files("data/processed/Pf", pattern = "ed_cr_dropd_resc.*_sct_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|ed_cr_dropd_resc|_sct.*'))) %>% 
  # .[sample_name_all] %>%
  .[nms_all_ed_pref_resc_100] %>%
  map(readRDS)

all_resc_sct_sct_elbo
```

```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_resc_sct, ~DimPlot(.x, label = T) + labs(title = .y))

```


```{r}
imap(all_resc_sct, possibly({~DimPlot(.x, group.by = "marker_stage", label = T) + labs(title = .y)}, "missing")) 

```

```{r, warning=FALSE, message=FALSE}
## subset cells that contain 2 or more sbp1 and cells that contain 2 or more ap2g UMI

all_resc_sct <- map(all_resc_sct, ~{
  .x@meta.data %>%
    mutate(stageHL_asex = case_when(marker_stage == "asex_merged" ~ "Asexual",
                                marker_stage == "mle" ~ "Male (LE)",
                                marker_stage == "fle" ~ "Female (LE)",
                                T ~ marker_stage),
           donor = str_remove_all(orig.ident, ".mrna")) %>%
    AddMetaData(.x, .)
})


```


```{r}
imap(all_resc_sct, possibly({~DimPlot(.x, group.by = "stageHL_asex", label = T) + labs(title = .y)}, "missing")) 

```


```{r, warning=FALSE, message=FALSE}
## ap2g
msc_qcd_anotd <- readRDS("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali1_thesis/data/processed/DB52e_star/msc_qcd_anotd.RDS") 

```


```{r, warning=FALSE, message=FALSE}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(all_resc_sct)) {
  DefaultAssay(all_resc_sct[[i]]) <- "RNA"
}

all_resc_sct
```

```{r, warning=FALSE, message=FALSE}
## ap2g
msc_qcd_anotd_comp_m22 <- map(msc_qcd_anotd[c("MSC13", "MSC14")], ~{
  .x@meta.data %>%
    mutate(stageHL_asex = case_when(stageHL %in% c("Early trophozoite", "Late ring") ~ "Asexual",
                                T ~ stageHL),
           donor = str_remove_all(orig.ident, ".mrna")) %>%
    AddMetaData(.x, .)
})

```

```{r, warning=FALSE, message=FALSE}
## ap2g
all_resc_sct_comp_m21_m22 <- c(all_resc_sct, msc_qcd_anotd_comp_m22)
```


```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle

# gn_exp <- map(all_resc_sct_comp_m21_m22[c("MSC13", "MSC14", "MSC48", "MSC49", "MSC54")], ~{
gn_exp <- map(all_resc_sct_comp_m21_m22, ~{
  seur <- .x
  seur[["RNA"]]$data <- as(object = seur[["RNA"]]$data, Class = "dgCMatrix")
  AverageExpression(seur, group.by = "stageHL_asex", assays = "RNA") %>% .[[1]] %>% as.data.frame() %>% rownames_to_column('gene')
}) %>% 
  bind_rows(., .id = "donor")

gn_exp
```



```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle
##female

gn_exp %>% pivot_longer(cols = c(contains("ale"), "Asexual"), names_to = "stages", values_to = "expression") %>% glimpse 

```



```{r}
library(GGally)

lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")
}
## correlation between genes by stage
gn_exp %>% filter(donor == 'MSC49') %>% select(-c(donor)) %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r, fig.width=10, fig.height=10}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Male (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```


```{r, fig.width=10, fig.height=10}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Female (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```

```{r, fig.width=10, fig.height=10}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Asexual')) %>% pivot_wider(names_from = 'donor', values_from = 'Asexual') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```



```{r, warning=FALSE, message=FALSE}

asex_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, top_asex_m14, assay = 'RNA') %>%  data.frame %>%  setNames('asx_avg_cln')  )

mle_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, top_mLE_m13_m14, assay = 'RNA') %>%  data.frame %>%  setNames('mle_avg_cln')  )

fle_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, top_fLE_m13_m14, assay = 'RNA') %>%  data.frame %>%  setNames('fle_avg_cln')  )

asex2_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, asex_2_m70, assay = 'RNA') %>%  data.frame %>%  setNames('asx2_avg_cln')  )


markers_avg_exp_cln <- pmap(list(asex_avg_exp_cln, mle_avg_exp_cln, fle_avg_exp_cln, asex2_avg_exp_cln), ~ bind_cols(..1, ..2, ..3, ..4))
# markers_avg_exp[[1]]

```


```{r, warning=FALSE, message=FALSE}

all_resc_sct_comp_m21_m22 <- list(all_resc_sct_comp_m21_m22, markers_avg_exp_cln) %>% pmap(~AddMetaData(..1, ..2))
```

Visualization of clusters with markers

```{r, warning=FALSE, message=FALSE, eval=T}
## cells that contain 2 or more ap2g UMI
imap(all_resc_sct_comp_m21_m22, ~DimPlot(.x, label = T)+ labs(title = .y))

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_resc_sct_comp_m21_m22, ~FeaturePlot(.x, features = c("asx_avg_cln", "mle_avg_cln", "fle_avg_cln", "asx2_avg_cln"),  ncol = 4) + plot_annotation(title = .y))

```
```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_resc_sct_comp_m21_m22, ~VlnPlot(.x, features = c("asx_avg_cln", "mle_avg_cln", "fle_avg_cln", "asx2_avg_cln", "nFeature_RNA", "nCount_RNA"), group.by = "stageHL_asex", ncol = 6) + plot_annotation(title = .y))

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=3}
imap(all_resc_sct_comp_m21_m22, ~VlnPlot(.x, features = c("asx_avg_cln", "mle_avg_cln", "fle_avg_cln", "asx2_avg_cln", "nFeature_RNA", "nCount_RNA"), ncol = 6) + plot_annotation(title = .y))

```




```{r, warning=FALSE, message=FALSE}

# ## correlation between genes by stage
# gn_exp %>% select(c('gene','donor', 'Female (LE)', 'Female')) %>% pivot_wider(names_from = 'donor', values_from = c('Female (LE)','Female')) %>% column_to_rownames('gene') %>% 
#   select(where(~!all(is.na(.x)))) %>%
#   ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
#           upper = list(continuous = wrap("cor", size = 5)),
#           progress = FALSE)

```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=10}

cor_fun <- function(data, mapping, method="pearson", use="pairwise", ndp=2, sz=5, stars=TRUE, ...){

# grab data
x <- eval_data_col(data, mapping$x)
y <- eval_data_col(data, mapping$y)

# calculate correlation: for significance stars
corr <- cor.test(x, y, method=method)
est <- corr$estimate
lb.size <- sz* abs(est)

# get significance stars
if(stars){
  stars <- c("***", "**", "*", "")[findInterval(corr$p.value, c(0, 0.001, 0.01, 0.05, 1))]
  lbl <- paste0(round(est, ndp), stars)
}else{
  lbl <- round(est, ndp)
}
# calculate correlation: for colored tiles
corr <- cor(x, y, method=method, use=use)

# calculate color based on correlation value
# corr = -1 => blue, 
# corr =  0 => white, 
# corr = +1 => red, 
colFn <- colorRampPalette(c("blue","white", "red"), interpolate ='spline')
fill <- colFn(100)[findInterval(corr, seq(0, 1, length=100))]

ggplot(data = data, mapping = mapping, ...) + 
  theme_void() +
  annotate("text",
           x=mean(x, na.rm=TRUE),
           y=mean(y, na.rm=TRUE),
           label=lbl,
           # size=lb.size,
           size=4,
           ...) +
  theme(panel.background = element_rect(fill=fill,  # to fill background of panel with color
                                        colour=NA), # to remove border of panel
        panel.grid.major = element_blank())
}


## correlation between genes by stage
# gn_exp %>% select(c('gene','donor', 'Early trophozoite', 'Late ring', 'Female (LE)', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = c('Female (LE)', 'Male (LE)', 'Early trophozoite', 'Late ring')) %>% 
gn_exp %>% select(c('gene','donor', 'Asexual', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = c('Male (LE)', 'Asexual')) %>% 
  select(where(~!all(is.na(.x)))) %>% 
  column_to_rownames('gene') %>% 
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          # upper = list(continuous = wrap("cor", size = 5)))
          upper = list(continuous = cor_fun),
          progress = FALSE)

```


```{r}

# Convert("data/processed/Pf/m2_all/ed_cr_dropd_all_query.h5ad", dest = "h5seurat", overwrite = T)

```


```{r}
# SeuratObject <- LoadH5Seurat("data/processed/Pf/m2_all/ed_cr_dropd_all_query.h5seurat") 
```

```{r}
mdata_scanvi <- read_csv("data/processed/Pf/m2_all/ed_cr_dropd_all_query_mdata.csv")
```


```{r}

## correlation between genes by stage

base_thm<-theme(axis.title=element_text(size=16,face="plain"), 
        axis.text = element_text(size=15), 
        legend.title = element_text(size=16,face="bold"),
        legend.text = element_text(size=15))
 
scattr_thm <- base_thm +
   theme(#aspect.ratio = 1, 
          legend.position = "none",
         plot.title = element_text(size=16, hjust = .5, face = "bold"),
        )

gam_corrs <- map(c("Female", "Female (LE)", "Male", "Male (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC13, y= MSC14, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  ggpubr::stat_cor(method = "pearson", label.x =150, label.y = 10)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(title = .x)+
    theme_classic()+
    scattr_thm
}) 

gam_corrs
```


```{r}

## correlation between genes by stage

gam_corrs3_13 <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC13/100, y= MSC3/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC13", y= "MSC3", title = .x)+
    theme_classic()+
    scattr_thm
}) 


gam_corrs3_14 <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC3/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC3", title = .x)+
    theme_classic()+
    scattr_thm
}) 

gam_corrs14_13 <- map(c("Female", "Female (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC13/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC13", title = .x)+
    theme_classic()+
    scattr_thm
}) 


gam_corrsm14_13 <- map(c("Male", "Male (LE)"), ~{gn_exp %>% 
  pivot_longer(cols = c(contains("ale"), "Early trophozoite", "Late ring"), names_to = "stages", values_to = "expression") %>% 
  pivot_wider(names_from = "donor", values_from = expression) %>% 
  filter(stages %in% .x)%>%
  select(where(~!all(is.na(.x)))) %>% 
  ggplot(aes(x= MSC14/100, y= MSC13/100, color = stages)) +
  geom_point() +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
  stat_cor(method = "pearson", label.x =150/100, label.y = 10/100)+ 
    # geom_vline( xintercept = 10)+
    # geom_hline( yintercept = log10(0.05))+
    labs(x="MSC14", y= "MSC13", title = .x)+
    theme_classic()+
    scattr_thm
}) 

gam_corrs3_14
gam_corrs3_13
gam_corrs14_13
gam_corrsm14_13
```


```{r}
lowerfun_le_corr <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")+
    stat_cor(method = "pearson", label.x =150, label.y = 10)+ 
    # labs(title = .x)+
    theme_classic()+
    theme(axis.text=element_text(size=20), 
          axis.title=element_text(size=18,face="bold"), 
          legend.position = "none",
          panel.border = element_rect(colour = "black",fill = NA))
}

## correlation between genes by stage
gam_corrsf_ga <- map(c("Female", "Female (E)"), ~{
  gn_exp %>% select(c('gene','donor', 'Female (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Female (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
GGally::ggduo(., 1:2, 3, types = list(continuous = lowerfun_le_corr)) +
  labs(title = 'Female (LE)')
})
gam_corrsf_ga

```


```{r}

gam_corrsm_ga <- map(c("Male", "Male (LE)"), ~{
  gn_exp %>% select(c('gene','donor', .x)) %>% pivot_wider(names_from = 'donor', values_from = .x) %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  # mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1, 2, types = list(continuous = lowerfun_le_corr))
}) 

gam_corrsm_ga
```

```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Male (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1, 2, types = list(continuous = lowerfun))

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Male')) %>% pivot_wider(names_from = 'donor', values_from = 'Male') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1, 2, types = list(continuous = lowerfun))
```

```{r}

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female (LE)')) %>% pivot_wider(names_from = 'donor', values_from = 'Female (LE)') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1:2, 3, types = list(continuous = lowerfun))

## correlation between genes by stage
gn_exp %>% select(c('gene','donor', 'Female')) %>% pivot_wider(names_from = 'donor', values_from = 'Female') %>% column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  mutate(across(everything(), log2)) %>%
GGally::ggduo(., 1:2, 3, types = list(continuous = lowerfun))
```




```{r, warning=FALSE, message=FALSE}
## ap2g
imap(all_resc_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T) + labs(title = .y)}, "missing"))
imap(all_resc_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

## SCT rerun for asexual subcluster

```{r, eval=T}
gogo()

```

```{r, eval=T}
## get asexuals for qp2g cluster search
all_ed_pref_resc_asex <- map(all_ed_pref_resc, ~subset(.x, subset = marker_stage == "asex_merged"))
```

```{r}
## write only with more than 100 cells - low number of cells gives error in nextflow sct normalization run 
all_ed_pref_resc_asex <- all_ed_pref_resc_asex[names(all_ed_pref_resc_asex)[unlist(map(all_ed_pref_resc_asex, ~dim(.x)[2] > 100))]]
```

```{r}
map(all_ed_pref_resc_asex, possibly({~DimPlot(.x, group.by = "marker_stage", label = T)}, "missing")) 
```

```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
### gn_edrops50
imap(all_ed_pref_resc_asex, ~saveRDS(.x, paste0("data/processed/Pf/",str_remove(.y, "_edrops50"), "/seu_obj/ed_cr_dropd_resc_asex.RDS")))
```

```{r, eval=T}
## names to paste for nextflow run
paste(names(all_ed_pref_resc_asex), collapse = ",")
```

```{r, eval=T}
gogo()
## Run standard normalisation on farm like below
## !!NB - reducing the number of hypervariable genes results in scattered and disjointed clusters. higher hvg number better

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sctnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC40,MSC48,MSC49,MSC54,MSC55,MSC60,MSC67,MSC68,MSC70}/seu_obj/ed_cr_dropd_resc_asex.RDS" --normlzn "sct" --umap_name "UMAP_" --hvg 100 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --resume
```


```{r, message=F, warning=F}
##Read BPcells matrices
# all_seu_prelim_stg_norm <- list.files("data/processed/Pf", pattern = "*seu_norm.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|bpc'))) %>% 

all_resc_asex_sct <- list.files("data/processed/Pf", pattern = "ed_cr_dropd_resc_asex.*_sct.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|ed_cr_dropd_resc_asex|_sct.*'))) %>% 
  # .[sample_name_all] %>%
  # .[names(all_ed_pref_resc_asex)] %>%
  map(readRDS)

```


Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read elbo plots
# all_seu_norm_elbo <- list.files("data/processed/Pf", pattern = "*seu_norm_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|_seu.*|bpc'))) %>% 

all_resc_asex_sct_elbo <- list.files("data/processed/Pf", pattern = "ed_cr_dropd_resc_asex.*_sct_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|ed_cr_dropd_resc_asex|_sct.*'))) %>% 
  # .[sample_name_all] %>%
  # .[names(all_ed_pref_resc_asex)] %>%
  map(readRDS)

all_resc_asex_sct_elbo
```

```{r, warning=FALSE, message=FALSE}
## ccp4
imap(all_resc_asex_sct, ~DimPlot(.x, label = T) + labs(title = .y))

```


```{r, warning=FALSE, message=FALSE, fig.width= 4, fig.height= 3}
## ap2g
imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T) + theme(text = element_text(size = 20)) + labs(title = .y)}, "missing"))
# imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE}
## ap2g
imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T) + theme(text = element_text(size = 20)) + labs(title = .y)}, "missing"))
# imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(all_resc_asex_sct)) {
  DefaultAssay(all_resc_asex_sct[[i]]) <- "RNA"
}

all_resc_asex_sct
```

```{r, warning=FALSE, message=FALSE}
## ap2g
imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", order = T, slot = "counts") + labs(title = .y)}, "missing"))
# imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, fig.width=14, fig.height=3.2}
# tbl <- sympt_fev_tbl1 %>%
#   count(condition, donor, StrainDon) %>% 
#   mutate(across(donor, ~str_remove(., "SC"))) %>%
#   group_by(condition, donor) %>%
#   add_count(name = "strn_n") %>% 
#   ungroup() %>%
#   mutate(prop_wdth =strn_n/n()) 
# 
# ## Get width proportions of the different facets based on number of strains per group
# p_width <- tbl%>% distinct(condition, donor, prop_wdth) %>% pull(prop_wdth) %>% round(., digits = 2)
# 
# tbl %>%
#   rename(c('Counts' =n, 'Strain' = StrainDon))  %>%
#   # ggplot(., aes(x = weave_factors(Strain, donor), y = Counts, fill = donor)) +
#   ggplot(., aes(x = Strain, y = Counts/1000, fill = donor)) +
#   geom_col(position = position_dodge2(width = 0.9, preserve = "single")) +
#   scale_fill_manual(values = donor_cols %>% set_names(str_remove(names(donor_cols), "_SLO"))) +
#   geom_text(aes(label=Counts), angle = 90, hjust=0, vjust= 0.5, position = position_dodge2(width = 0.9, preserve = "single"), size = 5, show.legend = F, fontface ='bold') +
#   scale_y_continuous(expand = expansion(mult = c(0.01, 0.62))) +
#   ggh4x::facet_nested_wrap(vars(condition, donor),  nrow = 1, scales = "free_x") +
#   theme_classic()+
#   theme(legend.position = "none", 
#         axis.title = element_text(size = 16),
#         axis.text.y = element_text(size = 14),
#         axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 15),
#         strip.text =element_text(size=16,face="bold", colour = 'black'
#                                  ),
#         panel.border=element_rect(colour="black",fill='transparent')
#         )+
#   # guides(x = "axis_nested") +
#   labs(x = "Strain", y = "Counts (X1000)") +
#   force_panelsizes(cols = p_width, respect = TRUE, total_width = unit(12.6, "inches"), total_height = unit(1.3, "inches"))
```



```{r, warning=FALSE, message=FALSE}
## subset cells that contain 2 or more sbp1 and cells that contain 2 or more ap2g UMI

ap2g_cells <- map(all_resc_asex_sct, possibly({~colnames(subset(.x, subset= `PF3D7-1222600` > 0, slot = 'counts'))}, "missing")) 

all_resc_asex_sct_ap2g <- map2(all_resc_asex_sct, ap2g_cells, ~{
  .x@meta.data %>%
    rownames_to_column("bcode") %>%
    mutate(ap2g_exp = case_when(marker_stage == "asex_merged" & bcode %in% .y ~ "asex_ap2g",
                                marker_stage == "asex_merged" & !bcode %in% .y ~ "asex_no_ap2g",
                                T ~ marker_stage)) %>%
    column_to_rownames("bcode") %>%
    select(ap2g_exp) %>%
    AddMetaData(.x, .)
})


```


```{r, warning=FALSE, message=FALSE}
## change default assay to RNA for getting cells with ap2g reads
saveRDS(all_resc_asex_sct_ap2g, "data/processed/Pf/m2_all/all_resc_asex_sct_ap2g.RDS")
```

```{r, warning=FALSE, message=FALSE}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(all_resc_asex_sct_ap2g)) {
  DefaultAssay(all_resc_asex_sct_ap2g[[i]]) <- "SCT"
}

all_resc_asex_sct_ap2g
```

```{r}
# # gogo()
all_resc_asex_sct_ap2g <- map(all_resc_asex_sct_ap2g, ~FindClusters(.x, resolution = 0.2, cluster.name = "seurat_clusters_ap2g"))

```

```{r, warning=FALSE, message=FALSE}
imap(all_resc_asex_sct_ap2g, ~DimPlot(.x, group.by = "ap2g_exp") + labs(title = .y))
imap(all_resc_asex_sct_ap2g, ~DimPlot(.x, group.by = "seurat_clusters_ap2g") + labs(title = .y))
```

```{r}

## remove clusters with unassigned cells > 70%
clstrs_ap2g <- map(all_resc_asex_sct_ap2g, ~.x@meta.data %>% 
                             dplyr::count(seurat_clusters_ap2g, ap2g_exp) %>% 
                             group_by(seurat_clusters_ap2g) %>% 
                             mutate(prop_n = round(n/sum(n), 2)) %>% 
                             mutate(likely_commited = ifelse(ap2g_exp == "asex_ap2g" & prop_n > 0.05, "yes", "no"))
                           ) #%>% dplyr::filter(!any(likely_unassigned == "yes" ))) #%>% dplyr::filter(likely_stage_asx_merged != "unassigned")
clstrs_ap2g
```
```{r}
ap2g_clstrs <- map(clstrs_ap2g, possibly(~{.x %>% dplyr::filter(likely_commited == "yes" ) %>% select(seurat_clusters_ap2g, likely_commited) %>% deframe}, "missing"))

```


```{r, warning=FALSE, message=FALSE}
## subset objects with ap2g like clusters
all_resc_asex_sct_ap2g_only <- all_resc_asex_sct_ap2g[unlist(map(ap2g_clstrs, ~length(.x) > 0))]

```


```{r}
# rules <- Map(reformulate, shQuote(likely_stage), shQuote(names(likely_stage)))
(ap2g_clstrs_rules <- map(ap2g_clstrs[unlist(map(ap2g_clstrs, ~length(.x) > 0))], ~glue::glue('"{lbl}" ~ "{val}"', lbl = names(.x) , val = .x)))
```

```{r, warning=FALSE, message=FALSE}
## label clusters with likely stage from majority of cells in cluster

all_resc_asex_sct_ap2g_only <- map2(all_resc_asex_sct_ap2g_only, ap2g_clstrs_rules, ~{
  .x@meta.data %>%
    mutate(ap2_clst = case_match(seurat_clusters_ap2g, !!!rlang::parse_exprs(.y),
                                 .default = "no"),
           ap2_clst_finer = ifelse(ap2_clst == "yes", "ap2g_clustr", paste0("clstr_", seurat_clusters_ap2g))) %>%
    select(ap2_clst, ap2_clst_finer) %>%
    AddMetaData(.x, .)
})


```

```{r, warning=FALSE, message=FALSE}
imap(all_resc_asex_sct_ap2g_only, ~DimPlot(.x, group.by = "ap2_clst") + labs(title = .y))
imap(all_resc_asex_sct_ap2g_only, ~DimPlot(.x, group.by = "ap2_clst_finer", label = T) + labs(title = .y))

```


```{r, warning=FALSE, message=FALSE, fig.width= 4, fig.height= 3}
## ap2g
imap(all_resc_asex_sct_ap2g_only, ~DimPlot(.x, group.by = "ap2_clst_finer", label = T, order = T, label.size = 7) + theme(text = element_text(size = 20), legend.position = "none", axis.text = element_blank(), axis.title = element_blank()) + labs(title = .y))
# imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE, fig.width= 4, fig.height= 3.5}
## ap2g
imap(all_resc_asex_sct_ap2g_only, ~VlnPlot(.x, features = "nCount_RNA", group.by = "ap2_clst_finer") + theme(text = element_text(size = 20), legend.position = "none", axis.text = element_blank(), axis.title = element_blank()) + labs(title = .y))
# imap(all_resc_asex_sct, possibly(~{FeaturePlot(.x, features = "PF3D7-1222600", reduction = "pca", order = T) + labs(title = .y)}, "missing"))
```

```{r, warning=FALSE, message=FALSE}
imap(all_resc_asex_sct_ap2g_only, ~VlnPlot(.x, features = "nCount_RNA", group.by = "ap2_clst_finer") + labs(title = .y))

```

```{r, warning=FALSE, message=FALSE}
## change ident to ap2g 
for (i in 1:length(all_resc_asex_sct_ap2g_only)) {
  Idents(all_resc_asex_sct_ap2g_only[[i]]) <- "ap2_clst"
}

```


```{r, warning=FALSE, message=FALSE}
## ap2g
de_gns <- map(all_resc_asex_sct_ap2g_only, ~FindMarkers(.x, test.use = "MAST", ident.1 = "yes", ident.2 = "no"))

```


```{r, warning=FALSE, message=FALSE}
de_gns_anot <- map(de_gns, ~.x %>% rownames_to_column("gene_id") %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")]))

de_gns_anot
```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more ap2g UMI
map(de_gns_anot, ~.x  %>% filter(p_val_adj < 0.05 & avg_log2FC > 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in ap2g positive cells

```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more ap2g UMI
imap(de_gns_anot, ~.x %>% filter(p_val_adj < 0.05 & avg_log2FC < 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in ap2g positive cells

```


```{r, warning=FALSE, message=FALSE}
# ## change ident to ap2g 
# for (i in 1:length(all_resc_asex_sct_ap2g_only)) {
#   Idents(all_resc_asex_sct_ap2g_only[[i]]) <- "ap2_clst_finer"
# }

```

```{r, warning=FALSE, message=FALSE}
# ## ap2g
# de_gns_fmall <- map(all_resc_asex_sct_ap2g_only, ~FindAllMarkers(.x, test.use = "MAST"))
# de_gns_fmall <- map(de_gns_fmall, ~.x %>% arrange(p_val_adj))

```

```{r, warning=FALSE, message=FALSE}
# de_gns_sig_anot <- map(de_gns_fmall, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))
# 
# de_gns_sig_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
# map(de_gns_sig_anot, ~{
#   mrks_tbl <- .x
#   map(unique(mrks_tbl$cluster), ~mrks_tbl %>% filter(avg_log2FC > 0, cluster == .x))
# })

```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
# map(de_gns_sig_anot, ~{
#   mrks_tbl <- .x
#   map(unique(mrks_tbl$cluster), ~mrks_tbl %>% filter(avg_log2FC < 0, cluster == .x))
# })

```


#### female high vs female low DE visualization
```{r, fig.height=7, fig.width=7}
map(de_gns_anot, ~{
  EnhancedVolcano(.x,
                  lab = .x$Gene,
                  selectLab =.x %>% 
                    # filter(p_val_adj < 0.05 & (avg_log2FC > 1 | avg_log2FC < -1) & !str_detect(Gene, 'PF3D7')) %>%
                    filter(p_val_adj < 0.05 & (avg_log2FC > 1 | avg_log2FC < -1) ) %>%
                    pull(Gene),
                  x = 'avg_log2FC',
                  y = 'p_val_adj',
                  pCutoff = 0.05,
                  FCcutoff = 1,
                  pointSize = 3.0,
                  labSize = 7.0,
                  title = '',
                  subtitle = "",
                  titleLabSize = 0,
                  subtitleLabSize = 0,
                  axisLabSize = 24,
                  drawConnectors = TRUE,
                  legendLabels =c(expression(Log[2] ~ FC), "p-value"),
                  gridlines.minor = FALSE)
})
```



```{r, warning=FALSE, message=FALSE}
## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in ap2g positive cells

## poran et al commitment
poran_commitment_ap2g <- c("PF3D7_1036000", "PF3D7_1335200", "PF3D7_0830800", "PF3D7_0402200", "PF3D7_1238500", "PF3D7_0424400", "PF3D7_1335000", "PF3D7_0617200", "PF3D7_0624600", "PF3D7_1002200", "PF3D7_0801900", "PF3D7_1139300", "PF3D7_1222400", "PF3D7_0201500", "PF3D7_1104200", "PF3D7_1472200", "PF3D7_1477800", "PF3D7_0406500", "PF3D7_1252600") %>% str_replace(., "_", "-")

```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more ap2g UMI
map(de_gns_anot, ~.x %>%  filter(gene_id %in% poran_commitment_ap2g) %>% filter(p_val_adj < 0.05 ))

```


```{r, fig.height=3.8, fig.width=3.8}
# M48, M49, M60
map(de_gns_anot, ~ggvenn(list("Poran_Up_Ap2g" = poran_commitment_ap2g,
            "Up_Ap2g" = .x %>% filter(p_val_adj < 0.05 & avg_log2FC < 0) %>% pull(gene_id),
            "Down_Ap2g" = .x %>% filter(p_val_adj < 0.05 & avg_log2FC > 0) %>% pull(gene_id)),
       show_percentage = F,
       text_size = 6)
)

```


########START TRY SCT ########

Integrated asexual to integrate with the rescued parasites
```{r}
## subset where field parasites are for slingshot
# v3_m2_asex_harmony_ss_res_fld_pt <- readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony_ss_res_fld_pt.RDS")
v3_m2_asex_harmony_ss_res_fld_pt <- readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony_res_fld_pt.RDS")
# %in% c("MSC25" ,"MSC41", "MSC51")

```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_harmony_ss_res_fld_pt@meta.data %>% dplyr::count(donor, dset_labmrg)
```

```{r}
## remove lab and M12
v3_m2_asex_cr_filt <- v3_m2_asex_harmony_ss_res_fld_pt[, !v3_m2_asex_harmony_ss_res_fld_pt@meta.data$dset_labmrg %in% c("Lab", "MSC12")]
```


```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt@meta.data %>% dplyr::count(donor, dset_labmrg)
```

```{r}
## remove weird 2 cells with >9000 ncountRNA
table(v3_m2_asex_cr_filt@meta.data$nCount_RNA >9000)
v3_m2_asex_cr_filt <- v3_m2_asex_cr_filt[, !v3_m2_asex_cr_filt@meta.data$nCount_RNA >9000]
```

```{r, warning=FALSE, message=FALSE}
# split layers
v3_m2_asex_cr_filt[["RNA"]] <- split(v3_m2_asex_cr_filt[["RNA"]], f = v3_m2_asex_cr_filt$donor)

```




```{r, warning=FALSE, message=FALSE}
# run sctransform
v3_m2_asex_cr_filt_sct <- SCTransform(v3_m2_asex_cr_filt, verbose = FALSE, variable.features.n = 300)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3_m2_asex_cr_filt_sct <- RunPCA(v3_m2_asex_cr_filt_sct, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(v3_m2_asex_cr_filt_sct, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}

v3_m2_asex_cr_filt_sct_int <- IntegrateLayers(
  object = v3_m2_asex_cr_filt_sct,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 90,
)


```

```{r, warning=FALSE, message=FALSE}

v3_m2_asex_cr_filt_sct_int <- FindNeighbors(v3_m2_asex_cr_filt_sct_int, dims = 1:6, reduction = "integrated.dr")
v3_m2_asex_cr_filt_sct_int <- FindClusters(v3_m2_asex_cr_filt_sct_int, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3_m2_asex_cr_filt_sct_int <- RunUMAP(v3_m2_asex_cr_filt_sct_int, dims = 1:6, verbose = FALSE, n.components = 3, seed.use = 867)
v3_m2_asex_cr_filt_sct_int <- FindNeighbors(v3_m2_asex_cr_filt_sct_int, dims = 1:6, verbose = FALSE)
v3_m2_asex_cr_filt_sct_int <- FindClusters(v3_m2_asex_cr_filt_sct_int, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}
saveRDS(v3_m2_asex_cr_filt_sct_int, "data/processed/Pf/m2_all/v3_m2_asex_cr_filt_sct_int.RDS")
```

```{r, warning=FALSE, message=FALSE}
# v3_m2_asex_cr_filt_sct_int <- readRDS("data/processed/Pf/m2_all/v3_m2_asex_cr_filt_sct_int.RDS")
```


```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, reduction = "pca")
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T)
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(v3_m2_asex_cr_filt_sct_int, group.by = "donor")
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "nFeature_RNA")
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "nCount_RNA")
VlnPlot(v3_m2_asex_cr_filt_sct_int, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(2,3))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(v3_m2_asex_cr_filt_sct_int, label = T, dims = c(1,3))

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", order = T) 
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", reduction = "pca", order = T) 
```
```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(v3_m2_asex_cr_filt_sct_int) <- "RNA"
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", order = T) 
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1222600", reduction = "pca", order = T) 
```



```{r, warning=FALSE, message=FALSE}
## ap2g
v3_m2_asex_cr_filt_sct_int_ap2g <- subset(v3_m2_asex_cr_filt_sct_int, subset = `PF3D7-1222600` > 0, slot = 'counts')

table(v3_m2_asex_cr_filt_sct_int_ap2g$donor)
```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1215500", order = T) 
FeaturePlot(v3_m2_asex_cr_filt_sct_int, features = "PF3D7-1215500", reduction = "pca", order = T) 
```


```{r, warning=FALSE, message=FALSE}

DefaultAssay(v3_m2_asex_cr_filt_sct_int) <- "SCT"
```

######################## START integrate cellranger filtered with rescued asexuals ########################
```{r, warning=FALSE, message=FALSE}
# run sctransform
all_resc_asex_rna <-  all_resc_asex_sct

```

```{r, warning=FALSE, message=FALSE}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(all_resc_asex_rna)) {
  DefaultAssay(all_resc_asex_rna[[i]]) <- "RNA"
  all_resc_asex_rna[[i]]@assays$SCT <- NULL
  all_resc_asex_rna[[i]]@project.name <- "SeuratProject"
}

all_resc_asex_rna
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
v3_m2_asex_cr_filt_list <-  SplitObject(v3_m2_asex_cr_filt, split.by = "donor")
```


```{r, warning=FALSE, message=FALSE}
## Donors with cells in the rescued object
resc_dons <- names(v3_m2_asex_cr_filt_list)[names(v3_m2_asex_cr_filt_list) %in% names(all_resc_asex_rna)]
non_resc_dons <- names(v3_m2_asex_cr_filt_list)[!names(v3_m2_asex_cr_filt_list) %in% names(all_resc_asex_rna)]

```

```{r, warning=FALSE, message=FALSE}
## Donors with cells in the rescued object
all_asex_rna <- list()
```


```{r, warning=FALSE, message=FALSE}
## for those with rescued cells merge these with cell ranger qcd cells
for (i in resc_dons) {
  all_asex_rna[i] <- merge(v3_m2_asex_cr_filt_list[[i]], all_resc_asex_rna[[i]])
  
}

all_asex_rna
```

```{r, warning=FALSE, message=FALSE}
## Add those without rescued cells to the object above created after merging
v3_m2_asex_cr_filt_no_resc <- v3_m2_asex_cr_filt_list[non_resc_dons]
all_asex_merged_list <- c(v3_m2_asex_cr_filt_no_resc, all_asex_rna)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- merge(all_asex_merged_list[[1]], all_asex_merged_list[2:length(all_asex_merged_list)])
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- JoinLayers(all_asex_merged)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged[["RNA"]] <- split(all_asex_merged[["RNA"]], f = all_asex_merged$donor)
all_asex_merged
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
all_asex_merged <- SCTransform(all_asex_merged, verbose = FALSE, variable.features.n = 200)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged <- RunPCA(all_asex_merged, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int <- IntegrateLayers(
  object = all_asex_merged,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int <- FindNeighbors(all_asex_merged_int, dims = 1:6, reduction = "integrated.dr")
all_asex_merged_int <- FindClusters(all_asex_merged_int, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int <- RunUMAP(all_asex_merged_int, dims = 1:6, verbose = FALSE, n.components = 3, seed.use = 867)
all_asex_merged_int <- FindNeighbors(all_asex_merged_int, dims = 1:6, verbose = FALSE)
all_asex_merged_int <- FindClusters(all_asex_merged_int, verbose = FALSE, resolution = 0.1)
```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_asex_merged_int, "data/processed/Pf/m2_all/all_asex_merged_int.RDS")
```

```{r, warning=FALSE, message=FALSE}
# all_asex_merged_int <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int.RDS")
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, reduction = "pca")
DimPlot(all_asex_merged_int, label = T)
DimPlot(all_asex_merged_int, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int, group.by = "donor")
FeaturePlot(all_asex_merged_int, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int, features = "nCount_RNA")
VlnPlot(all_asex_merged_int, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(2,3))
DimPlot(all_asex_merged_int, label = T, dims = c(2,3))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, dims = c(1,3))

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```
```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int) <- "RNA"
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int) <- "SCT"
```



```{r, warning=FALSE, message=FALSE}
## ap2g
all_asex_merged_int_ap2g <- subset(all_asex_merged_int, subset = `PF3D7-1222600` > 0, slot = 'counts')

table(all_asex_merged_int_ap2g$donor)
```

```{r, warning=FALSE, message=FALSE}
# ## change ident to ap2g
# for (i in 1:length(all_asex_merged_int)) {
#   Idents(all_asex_merged_int[[i]]) <- "ap2_clst_finer"
# }

```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int) <- "RNA"
all_asex_merged_int <- JoinLayers(all_asex_merged_int)

```



```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int)
de_gns_fmall_all <- de_gns_fmall_all %>% arrange(p_val_adj)

```


```{r, warning=FALSE, message=FALSE}
de_gns_sig_anot <- de_gns_fmall_all %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_anot
```

```{r, warning=FALSE, message=FALSE}
VlnPlot(all_asex_merged_int, features = c("nFeature_RNA", "nCount_RNA"))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T)
```




```{r, warning=FALSE, message=FALSE, eval=TRUE}
all_asex_merged_int@meta.data %>% filter(seurat_clusters == 4) %>% count(donor, seurat_clusters)

```


```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

# Cluster 4 highly expressess truRNA which is switched on when temperature is lowered from 37C to 26C activates!!! https://www.sciencedirect.com/science/article/pii/S0378111923003578
# https://pmc.ncbi.nlm.nih.gov/articles/PMC11597781/

```

```{r, warning=FALSE, message=FALSE}
gogo()
```

############ SCT2 -without activated asexuals
Remove the low temperature activate asexuals and cluster the rest
```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv <- all_asex_merged_int[, all_asex_merged_int$seurat_clusters != 4]
DimPlot(all_asex_merged_int_no_actv, label = T)
```


```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv[["RNA"]] <- split(all_asex_merged_int_no_actv[["RNA"]], f = all_asex_merged_int_no_actv$donor)
all_asex_merged_int_no_actv
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
all_asex_merged_int_no_actv <- SCTransform(all_asex_merged_int_no_actv, verbose = FALSE, variable.features.n = 100)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv <- RunPCA(all_asex_merged_int_no_actv, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged_int_no_actv, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int_no_actv <- IntegrateLayers(
  object = all_asex_merged_int_no_actv,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv <- FindNeighbors(all_asex_merged_int_no_actv, dims = 1:5, reduction = "integrated.dr")
all_asex_merged_int_no_actv <- FindClusters(all_asex_merged_int_no_actv, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv <- RunUMAP(all_asex_merged_int_no_actv, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867)
all_asex_merged_int_no_actv <- FindNeighbors(all_asex_merged_int_no_actv, dims = 1:5, verbose = FALSE)
all_asex_merged_int_no_actv <- FindClusters(all_asex_merged_int_no_actv, verbose = FALSE, resolution = 0.1)
```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_asex_merged_int_no_actv, "data/processed/Pf/m2_all/all_asex_merged_int_no_actv.RDS")
```

```{r, warning=FALSE, message=FALSE}
# all_asex_merged_int_no_actv <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int_no_actv.RDS")
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv, label = T, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv, label = T)
DimPlot(all_asex_merged_int_no_actv, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int_no_actv, group.by = "donor")
FeaturePlot(all_asex_merged_int_no_actv, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int_no_actv, features = "nCount_RNA")
VlnPlot(all_asex_merged_int_no_actv, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv) <- "RNA"
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int_no_actv) <- "SCT"
```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv) <- "RNA"
all_asex_merged_int_no_actv <- JoinLayers(all_asex_merged_int_no_actv)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_cln <- FindAllMarkers(all_asex_merged_int_no_actv)
de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv, label = T)
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

```

```{r, warning=FALSE, message=FALSE}
## Dormant (d5MDPs) genes after artemisinin - https://www-nature-com.ezp.lib.cam.ac.uk/articles/s41467-024-51846-0
d5MDP_up_gns <- c("PF3D7_1430900", "PF3D7_0928000", "PF3D7_0217100", "PF3D7_0412100", "PF3D7_1341100") %>% str_replace(., "_", "-")

d5MDP_down_gns<- c("PF3D7_0202500", "PF3D7_1102800", "PF3D7_1401400", "PF3D7_0202000") %>% str_replace(., "_", "-")


```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_up_gns)
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_down_gns)
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE}
VlnPlot(all_asex_merged_int, features = c("nFeature_RNA", "nCount_RNA"))

```

############ Harmony -without activated asexuals
Remove the low temperature activate asexuals and cluster the rest
```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony <- all_asex_merged_int[, all_asex_merged_int$seurat_clusters != 4]
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
```


```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony[["RNA"]] <- split(all_asex_merged_int_no_actv_harmony[["RNA"]], f = all_asex_merged_int_no_actv_harmony$donor)
all_asex_merged_int_no_actv_harmony
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
# all_asex_merged_int_no_actv_harmony <- SCTransform(all_asex_merged_int_no_actv_harmony, verbose = FALSE, variable.features.n = 100)

all_asex_merged_int_no_actv_harmony <- NormalizeData(all_asex_merged_int_no_actv_harmony)
all_asex_merged_int_no_actv_harmony <- FindVariableFeatures(all_asex_merged_int_no_actv_harmony, selection.method = "vst", nfeatures = 300)
all_asex_merged_int_no_actv_harmony <- ScaleData(all_asex_merged_int_no_actv_harmony, verbose = F)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv_harmony <- RunPCA(all_asex_merged_int_no_actv_harmony, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, reduction = "pca")
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.5)
```


```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "pca", group.by = "stage")


```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv@reductions[["pca"]]@cell.embeddings),
        x = ~PC_1, y = ~PC_2, z = ~PC_3,
        color= all_asex_merged_int_no_actv@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int_no_actv_harmony <- IntegrateLayers(
  object = all_asex_merged_int_no_actv_harmony,
  method = HarmonyIntegration,
  orig.reduction = "pca", 
  new.reduction = "harmony",
  # normalization.method = "SCT",
  verbose = F,
  # k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, reduction = "harmony")
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv_harmony <- RunUMAP(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, verbose = FALSE, resolution = 0.1)
```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_asex_merged_int_no_actv_harmony, "data/processed/Pf/m2_all/all_asex_merged_int_no_actv_harmony.RDS")
```

```{r, warning=FALSE, message=FALSE}
# all_asex_merged_int_no_actv_harmony <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int_no_actv_harmony.RDS")
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, label = T, reduction = "harmony")
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
DimPlot(all_asex_merged_int_no_actv_harmony, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int_no_actv_harmony, group.by = "donor")
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "nCount_RNA")
VlnPlot(all_asex_merged_int_no_actv_harmony, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "harmony", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "SCT"
```
```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "harmony", order = T, pt.size = 0.5) 

```



```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["harmony"]]@cell.embeddings),
        x = ~harmony_1, y = ~harmony_2, z = ~harmony_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
all_asex_merged_int_no_actv_harmony <- JoinLayers(all_asex_merged_int_no_actv_harmony)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_cln <- FindAllMarkers(all_asex_merged_int_no_actv_harmony)
de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```

Slingshot
```{r, warning=FALSE, message=FALSE}
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.05, cluster.name = "harmony_clusters_ss") 
```


```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "harmony", group.by = "harmony_clusters_ss", order = T, pt.size = 0.5) 

```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["harmony"]]@cell.embeddings),
        x = ~harmony_1, y = ~harmony_2, z = ~harmony_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, eval=T}
# gogo()
all_asex_merged_int_no_actv_harmony_sce <- as.SingleCellExperiment(all_asex_merged_int_no_actv_harmony)

```

```{r, eval=T}
rowData(all_asex_merged_int_no_actv_harmony_sce)$feature_symbol <- rownames(all_asex_merged_int_no_actv_harmony_sce)
```

!!!! NOTE -BEEN DOING SLINGSHOT WRONG. LOGCOUNTS SHOULD HAVE data values from seurat
```{r, eval=T}
# gogo()
# Seurat_object <- m2_asex_harmony_res

## Convert seurat object to SCE
all_asex_merged_int_no_actv_harmony_sce_ss <- slingshot(all_asex_merged_int_no_actv_harmony_sce,
            clusterLabels = "harmony_clusters_ss",
            reducedDim = "HARMONY",
            start.clus = 1,
            end.clus = 0,
            # shrink =1,
            # stretch = 1,
            # allow.break = F
            )


```


```{r}
##Prepare slingshot dataset (SDS) object
all_asex_merged_int_no_actv_harmony_sce_ss_sds <- SlingshotDataSet(all_asex_merged_int_no_actv_harmony_sce_ss)

# plot(reducedDims(all_asex_merged_int_no_actv_harmony_sce_ss)$UMAP[,1:2], col = brewer.pal(9,"Set1")[as.factor(all_asex_merged_int_no_actv_harmony_sce_ss$stage)], pch=16, asp = 0.7)
plot(reducedDims(all_asex_merged_int_no_actv_harmony_sce_ss)$HARMONY[,c(1,2)], col = brewer.pal(9,"Set1")[as.factor(all_asex_merged_int_no_actv_harmony_sce_ss$seurat_clusters)], pch=16, asp = 0.7, )

lines(all_asex_merged_int_no_actv_harmony_sce_ss_sds, lwd=2, col="black",dims = c(1,2))
```


```{r}
## Add pseudotime metadata to the object
all_asex_merged_int_no_actv_harmony_pt <- AddMetaData(all_asex_merged_int_no_actv_harmony, 
                                                      data.frame(all_asex_merged_int_no_actv_harmony_sce_ss@colData[, c("slingPseudotime_1"), drop = F]) #%>%
                                                        # rename("pseudotime" = slingPseudotime_1)
                                                      ) 

```

```{r}
## Save pseudotime objects
saveRDS(all_asex_merged_int_no_actv_harmony_pt, "data/processed/Pf/all_asex_merged_int_no_actv_harmony_pt.RDS")
saveRDS(all_asex_merged_int_no_actv_harmony_sce_ss, "data/processed/Pf/all_asex_merged_int_no_actv_harmony_sce_ss.RDS")
saveRDS(all_asex_merged_int_no_actv_harmony_sce_ss_sds, "data/processed/Pf/all_asex_merged_int_no_actv_harmony_sce_ss_sds.RDS")

```

```{r, warning=FALSE, message=FALSE}
##Generate pseudotime chunks categories with range of highest and lowest pseudotime values and breaks numbers to ensure balanced representation of cells and drop the cells without strain information

ptime_chunk_calc_fn <- function(sce_obj, ptime, chunk_num) {
  # v3_m2_asex_harmony_fld_sset_ptime1_fld@meta.data[,c('ptime_cat', 'ptime_bal_cat')] 
  sce_obj@colData[,c('ptime_cat', 'ptime_bal_cat')] <- sce_obj@colData[,c( ptime), drop=F] %>%
          data.frame() %>% 
          mutate(ptime_cat = cut(!!rlang::sym(ptime),  
                                 breaks = c(-Inf, seq(min(!!rlang::sym(ptime)),  max(!!rlang::sym(ptime)), round((max(!!rlang::sym(ptime)) - min(!!rlang::sym(ptime)))/chunk_num, 2)), Inf)),
                 pt_rank = dense_rank(!!rlang::sym(ptime)),
                 ptime_bal_cat = cut(pt_rank,  
                                 breaks = c(-Inf, seq(min(pt_rank),  max(pt_rank), round((max(pt_rank) - min(pt_rank))/chunk_num, 2)), Inf))
                 ) %>%
    mutate(ptime_grp_rnk = dense_rank(factor(ptime_cat)),
           ptime_bal_grp_rnk = dense_rank(factor(ptime_bal_cat)),
           ptime_cat = factor(paste0("ptime_", ptime_grp_rnk), levels = paste0("ptime_", 1:max(ptime_grp_rnk)) ),
           ptime_bal_cat = factor(paste0("ptime_", ptime_bal_grp_rnk), levels = paste0("ptime_", 1:max(ptime_bal_grp_rnk)) )
           ) %>%
          dplyr::select(ptime_cat, ptime_bal_cat)
  
  return(sce_obj)
}

```


```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
all_asex_merged_int_no_actv_harmony_sce_ss< - ptime_chunk_calc_fn(sce_obj = all_asex_merged_int_no_actv_harmony_sce_ss, ptime = "slingPseudotime_1", chunk_num = 10)
```



```{r, warning=FALSE, message=FALSE}
## subset cells that contain 2 or more sbp1 and cells that contain 2 or more ap2g UMI

ap2g_cells_all <- subset(all_asex_merged_int_no_actv_harmony_pt, subset= `PF3D7-1222600` > 0, slot = 'counts')

```


```{r}
all_asex_merged_int_no_actv_harmony_sce_ss@colData[,c('ap2g_exp_all')] <- ifelse(colnames(all_asex_merged_int_no_actv_harmony_sce_ss) %in% colnames(ap2g_cells_all), "yes", "no")
```



```{r, warning=FALSE, message=FALSE}
gogo()

```


```{r}
##Subsample the cells within pseudotime chunks calculated above to ensure balanced representation of field and lab cells along trajectories 

# ds_grp <- c("ap2g_exp_all")

balanced_cells_fn <- function(sce_obj, dsample_condtn, ptime_strategy) {
  
  set.seed(989778)
  
  ##Revert order in filter https://stackoverflow.com/questions/70951525/filter-dataframe-within-a-group-with-one-column-meeting-an-and-condition-in-r
  # balanced_cells_tbl_str <- list(msc14_sub_sce_unbal_ptchunks, grps, grps2)  %>%
  balanced_cells_tbl_str <- all_asex_merged_int_no_actv_harmony_sce_ss@colData[,c('ptime_cat', 'ptime_bal_cat', dsample_condtn)] %>% 
      as.data.frame() %>% 
      rownames_to_column('cell_bc') %>% 
      group_by(across(all_of(c(ptime_strategy, dsample_condtn)))) %>% 
      mutate(cnt  = sum(!is.na(!!rlang::sym(dsample_condtn)))) %>% 
               group_by(across(all_of(c(ptime_strategy)))) %>% 
               mutate(max_cnt = max(cnt)) %>% 
               mutate(max2_cnt = max(cnt[cnt != max_cnt], na.rm = TRUE)) %>%
               group_by(ptime_bal_cat, !!rlang::sym(dsample_condtn)) %>% 
               mutate(samp = sample(n())) %>%
               filter(samp <= max2_cnt) %>%
               ungroup() #%>%
               # mutate(across(ptime_bal_cat, droplevels))
  
  
  ##Subset the balanced cells for the 3 objects
  # m2_asex_ap2g_bal_sce <- all_asex_merged_int_no_actv_harmony_sce_ss[,balanced_cells_tbl_str$cell_bc]
  
  return(balanced_cells_tbl_str)

}
```


```{r, warning=FALSE, message=FALSE}
bal_cells_df <- balanced_cells_fn(sce_obj = all_asex_merged_int_no_actv_harmony_sce_ss, dsample_condtn = "ap2g_exp_all" , ptime_strategy = "ptime_bal_cat")

```


```{r, warning=FALSE, message=FALSE}
m2_asex_ap2g_bal_sce <- all_asex_merged_int_no_actv_harmony_sce_ss[, bal_cells_df$cell_bc]

```


```{r, warning=FALSE, message=FALSE}
# m2_asex_ap2g_bal_sce$donor <- str_remove_all(m2_asex_ap2g_bal_sce$orig.ident, "SC|.mrna")
m2_asex_ap2g_bal_sce$donor <- str_remove_all(m2_asex_ap2g_bal_sce$orig.ident, ".mrna")

```

```{r, results="asis", eval=T}
# balanced_cells_tbl_str %>% count(ptime_bal_cat, ap2g_exp_all)
```

### Retain cells with strains having more than 10 in each stage


```{r}
m2_asex_ap2g_bal_sce <- m2_asex_ap2g_bal_sce[,m2_asex_ap2g_bal_sce@colData[, c('ap2g_exp_all', 'ptime_bal_cat')] %>% data.frame()  %>% rownames_to_column('bcode') %>% group_by(ptime_bal_cat) %>% filter(n() >=30) %>% pull(bcode)]

table(m2_asex_ap2g_bal_sce@colData[, 'ap2g_exp_all'])

```

```{r, results="asis", eval=T}
m2_asex_ap2g_bal_sce@colData[, c('ap2g_exp_all', 'ptime_bal_cat')] %>% data.frame() %>% count(ptime_bal_cat, ap2g_exp_all)
```

```{r, results="asis", fig.width=4, fig.height=3}

tm <- theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 15) ,
        legend.title = element_text(size =15, face = "bold") )

gbox_theme<- theme(axis.text=element_text(size=15),
        axis.title=element_text(size=15,face="bold"), 
        legend.text = element_text(size = 15) ,
        legend.title = element_text(size = 15, face = "bold"),
        legend.position = "bottom")
        
dens_thme <- theme(axis.text=element_text(size=22),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=22,face="bold"), 
        legend.text = element_text(size = 15) ,
        legend.title = element_text(size = 15, face = "bold"),
        legend.position = "bottom")



```

```{r, results="asis", fig.width=5, fig.height=3.5}
library("ggpubr")

umap_meta_df <- reducedDims(all_asex_merged_int_no_actv_harmony_sce_ss)$UMAP[,c(1,2)] %>% as.data.frame() %>%rownames_to_column("cell_bc") %>% left_join(., all_asex_merged_int_no_actv_harmony_sce_ss@colData[,c("slingPseudotime_1", "orig.ident", "StrainDon", "ap2g_exp_all", "donor")] %>% as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc") 

tbl = umap_meta_df[umap_meta_df$cell_bc %in% colnames(m2_asex_ap2g_bal_sce),]

tbl %>%
  ggplot(., aes(x = slingPseudotime_1, col=ap2g_exp_all, fill=ap2g_exp_all)) + 
  geom_density(aes(y = ..count..*0.3), size=2, bw = 0.3, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  dens_thme +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))


tbl %>%
  ggboxplot(., y = "slingPseudotime_1",x="ap2g_exp_all", col="ap2g_exp_all") + 
  stat_compare_means(method = "wilcox.test") +gbox_theme


```


```{r}
## plot balanced symptomatic vs assymptomatic by pseudotime
all_asex_merged_int_no_actv_harmony_bal <- all_asex_merged_int_no_actv_harmony[,colnames(m2_asex_ap2g_bal_sce)]
all_asex_merged_int_no_actv_harmony_bal@meta.data$pseudotime <- m2_asex_ap2g_bal_sce$slingPseudotime_1
all_asex_merged_int_no_actv_harmony_bal@meta.data$ptime_cat <- m2_asex_ap2g_bal_sce$ptime_cat
all_asex_merged_int_no_actv_harmony_bal@meta.data$ap2g_exp_all <- m2_asex_ap2g_bal_sce$ap2g_exp_all
```

```{r}
all_asex_merged_int_no_actv_harmony_bal@meta.data$ap2g <- ifelse(all_asex_merged_int_no_actv_harmony_bal@meta.data$ap2g_exp_all == "yes", "ap2gPos", "ap2gNeg")
m2_asex_ap2g_bal_sce$ap2g <- ifelse(m2_asex_ap2g_bal_sce$ap2g_exp_all == "yes", "ap2gPos", "ap2gNeg")

```

```{r}
FeaturePlot(all_asex_merged_int_no_actv_harmony_bal, reduction = "harmony", features = "pseudotime", pt.size = 0.5, dims = c(1,2), split.by = "ap2g")

```

```{r}
## plot balanced symptomatic vs assymptomatic by pseudotime
# v3_m2_asex_harmony_ss_res_fld_pt_bal <- v3_m2_asex_harmony_ss_res_fld_pt[,colnames(m2_asex_sympto_bal_sce)]
# v3_m2_asex_harmony_ss_res_fld_pt_bal@meta.data$pseudotime <- m2_asex_sympto_bal_sce$slingPseudotime_1
# v3_m2_asex_harmony_ss_res_fld_pt_bal@meta.data$ptime_cat <- m2_asex_sympto_bal_sce$ptime_cat

DimPlot(all_asex_merged_int_no_actv_harmony_bal, reduction = "harmony", group.by = "ptime_cat", pt.size = 0.5, dims = c(1,2), label = T, split.by = "ap2g")

```


#### Put subseted objects into a list and calculate metrics need for plotting and tradeseq input
```{r}
str_cols = c("SC1" = "#dcbeff", "SC5" = "#008080")
lins = rep(1,length(m2_asex_ap2g_bal_sce))
# lins_m = rep(c(1,2),4)

m2_asex_ap2g_bal_sds <- SlingshotDataSet(m2_asex_ap2g_bal_sce)
m2_asex_ap2g_bal_ptime <- slingPseudotime(m2_asex_ap2g_bal_sce)[,1] 
m2_asex_ap2g_bal_cw <- slingCurveWeights(m2_asex_ap2g_bal_sce)[,1]

```



```{r}
## Batch design correction for tradeseq
m2_asex_ap2g_bal_u_don <- model.matrix(~  donor, m2_asex_ap2g_bal_sce@colData[, c("donor", "ap2g")] %>% data.frame())

# ## tradeseq manual example - Batch design correction for tradeseq
# counts <- as.matrix(countMatrix)
# 
# batch <- factor(rep(c("A", "B"), each = ncol(counts)/2))
# U <- model.matrix(~batch)
# sceBatch <- fitGAM(counts = counts,
#                    U = U,
#                    sds = crv, 
#                    nknots = 6, 
#                    verbose = FALSE)
```


```{r}
## create directory
system(paste0("mkdir -p /lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/"))

saveRDS(m2_asex_ap2g_bal_sce, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal_sce.RDS')

saveRDS(m2_asex_ap2g_bal_ptime, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal_ptime.RDS')

saveRDS(m2_asex_ap2g_bal_cw, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal_cw.RDS')

saveRDS(m2_asex_ap2g_bal_u_don, '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal_u_don.RDS')

```



```{r, results="asis", eval=T}
gogo()
## Tradeseq script

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_tseq_batch.nf --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal" --scrpt /nfs/users/nfs_j/jr35/multipurpose_scripts/tradeseq_batch_nf.R --condtn "ap2g" --knots "4" --gns_cut "5"  --resume

```


```{r, results="asis", eval=T}
gogo()
## Tradeseq script

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_tseq.nf --w_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal" --scrpt /nfs/users/nfs_j/jr35/multipurpose_scripts/tradeseq_nf.R --condtn "ap2g" --knots "4" --gns_cut "5"  --resume

```

```{r, results="asis", eval=T}
m2_asex_ap2g_bal_batch_ts <- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal_batch_ts.RDS")
m2_asex_ap2g_bal_batch_de <- readRDS("data/processed/Pf/pseudotime_tradeseq_ip/ap2g/m2_asex_ap2g_bal_batch_de.RDS")
```


```{r, results="asis", eval=T}
m2_asex_ap2g_bal_batch_de %>% arrange(desc(waldStat))
```

```{r}
##NOTE!!! replace the conditional test file with the newly generated one which applied 7 clusters
##Ajusting the global and pairwise pvalues - old p values discarded
m2_asex_ap2g_bal_batch_de <- m2_asex_ap2g_bal_batch_de %>%
                             mutate(across(starts_with('pvalue'), ~p.adjust(., "fdr"))) %>%
                             arrange(desc(waldStat)) %>%
                             rownames_to_column('gene_id')

```


```{r}

m2_asex_ap2g_bal_batch_de <- left_join(m2_asex_ap2g_bal_batch_de, Pf3D7_genes_plasmoDB %>% dplyr::select(gene_id, Gene, Ka_Ks, `Transcript Product Description`, TM_domains, `Computed GO Functions`), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene))

```


```{r, results="asis", eval=T}
m2_asex_ap2g_bal_batch_de %>% arrange(pvalue)
```

#### smoothers for select genes that are differentially expressed between strains {.tabset}

#### Plot smoothers function


```{r, results='asis'}
# cat('## Visualize field parasites {.tabset}   \n')
tde_smooth_plt <- function(tde_obj_lst = asex_bal_ts[1],
                           cond_obj_lst = cond_gns_de_asx[1],
                           pw_comp= 'waldStat_M33_SC5vM33_SC7', 
                           n_gns = c(1:5),
                           lege_pos = "bottom",
                           wstat = 20,
                           pval_cond = "pvalue",
                           pval = 0.05
)
{
  
  list(tde_obj_lst, cond_obj_lst) %>% 
    pmap(.,possibly(~{
      
      gns = ..2 %>% filter(!!rlang::sym(pw_comp) > wstat & !!rlang::sym(pval_cond)  < pval) %>% filter(!is.na(gene_id)) %>%  arrange(desc(!!rlang::sym(pw_comp)))  %>% pull(gene_id) %>% .[n_gns]
      
      gn_names = ..2 %>% filter(!!rlang::sym(pw_comp) > wstat & !!rlang::sym(pval_cond)  < pval) %>% filter(!is.na(gene_id))  %>% arrange(desc(!!rlang::sym(pw_comp))) %>% pull(Gene) %>% .[n_gns]
      print(gn_names)
      
      tde = ..1
      
      list(gns, gn_names) %>% 
        pmap(., possibly(~{
          plotSmoothers(tde, 
                        assays(tde)$counts,
                        gene = ..1,
                        alpha = 1, 
                        border = TRUE) +
            labs(title = ..2) +
            theme(text = element_text(size = 14),
                  legend.position = lege_pos) +
            guides(color = guide_legend(override.aes = list(size = 2), title.position = "top", ncol = 1))
        }, 'missing'))
    }, 'missing'))
}

```


```{r, results='asis', warning=F, message=F}

tde_smooth_plt(tde_obj_lst = list(m2_asex_ap2g_bal_batch_ts),
               cond_obj_lst = list(m2_asex_ap2g_bal_batch_de),
               pw_comp= 'waldStat',
               n_gns = c(1:20)
)

```

Heatmap plotting function
```{r}
### Annotation groups for the header ie column
tseq_de_hmap <- function(tseq = asex_bal_ts[[1]], de_tbl , pnts = 30, rnames = F, comp_grps, Pv = 0.05,  wstat =30){
  
  # gns = de_tbl %>% filter(if_any(matches("pvalue.*"), ~. < 0.05)) %>% pull(gene_id)
  gns = de_tbl %>% filter(pvalue < Pv & waldStat >= wstat) %>% pull(gene_id)

  # print(str(gns))
  ##Edited from the original version to allow the labelling of rows with gene names
  msc_yhatSmooth_w_annot <- predictSmooth(tseq, gene = gns, nPoints = pnts, tidy = FALSE) %>% 
        t() %>%scale() %>% t() %>% as.data.frame()%>% 
        rownames_to_column('gene_id') %>% 
        left_join(., Pf3D7_genes_plasmoDB  %>% dplyr::select("Gene", "gene_id", "Ka_Ks"), by = 'gene_id') %>% 
        mutate(Gene = case_when(is.na(Gene) ~ gene_id, T ~ Gene)) %>% 
        dplyr::select(-c(gene_id)) %>% 
        column_to_rownames('Gene')
  
  msc_yhatSmooth <- msc_yhatSmooth_w_annot %>% 
    dplyr::select(-c(Ka_Ks)) 
  
  # print(str(msc_yhatSmooth))
  msc_col_annot <- predictSmooth(tseq, gene = gns, nPoints = pnts, tidy = T)  %>% 
        mutate(point = paste0("point", rep(rep(1:pnts,length(unique(tseq@colData$tradeSeq$conditions))),length(gns))), 
               lin_id = paste0('lineage', lineage,'_condition', condition, '_',point)) %>%
        rownames_to_column('No')  %>% 
        dplyr::select(lin_id,time,condition)%>% 
        distinct()%>% 
        column_to_rownames('lin_id') %>%
        rename('Pseudotime' = time, 'Strain' = condition)
  # print(str(msc_col_annot))

  hmap<- ComplexHeatmap::Heatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix(),
                         cluster_columns = FALSE,
                         cluster_rows = FALSE,
                         show_row_names = {{rnames}}, 
                         show_column_names = F, 
                         # main = comp_grps,
                         show_heatmap_legend = T,
                         # silent = TRUE,
                         # left_annotation = rowAnnotation(df= msc_yhatSmooth_w_annot %>% select(Ka_Ks)),
                         top_annotation = HeatmapAnnotation(df= msc_col_annot %>% dplyr::select( Strain,Pseudotime) ,
                                                            # cols = list(Strain = msc_col_annot %>% mutate(cols = ifelse(Strain == 'M49_SC1') ~ '#e7a06b', '#9cb4cf') %>%  pull(cols) ), 
                                                            annotation_legend_param = list(Strain = list(title = "Strain"), 
                                                                                           direction = "horizontal", 
                                                                                           title_gp = gpar(fontsize = 14), labels_gp = gpar(fontsize = 14)),
                                                            show_legend = c(TRUE, FALSE) ##https://support.bioconductor.org/p/82447/
                                                            ),
                         # fontsize = 14,
                         # heatmap_legend_param = list(ncol = 1, nrow = 4, legend_direction = "vertical", by_row = T),
            # heatmap_legend_param = list(direction = "horizontal", legend_gp = gpar(fontsize = lgsize)),
            heatmap_legend_param = list(title = "Expression",
                                        title_gp = gpar(fontsize = 14),
                                        labels_gp = gpar(fontsize = 14),
                                        direction = "horizontal",
                                        labels_rot = 90#,
                                        # grid_width = unit(6, "mm"),
                                        # legend_width = unit(lg_width, "cm")
                                        ),
            row_names_gp = grid::gpar(fontsize = 14)
                         # 
      ) %>%
  ##Merging legends into 1 column. Could be set to F
  draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")
    
  return(hmap)
  # print(hmap)
  # print(pheatmap::pheatmap(msc_yhatSmooth %>% as.data.frame() %>% as.matrix()))
      # cat('\n')
      # cat("\n")
    
}
```

```{r, results='asis', warning=F, message=F, fig.width=6, fig.height=6}

tseq_de_hmap(tseq = m2_asex_ap2g_bal_batch_ts, de_tbl= m2_asex_ap2g_bal_batch_de, pnts = 30, rnames = T, comp_grps = "ap2g", Pv = 0.05)

```



## MAST DE adjusting for pseudotime

MAST DE on the entire dataset before subsampling. Subsampling is not necessary for MAST and also for tradeseq
### Prepare dataset
Normalizing and scaling the data in the subseted seurat objects - may have been done before on the larger parent object but just repeating since will not change anything as it starts from the raw counts.

Adding pseudotime metadata from the single cell experiment object 
Scaling is for visualization


```{r}
##REMEMBER TO REMOVE CLUSTERS WITH VERY FEW CELLS AND DOUBLETS SEURAT
# msc_gam_combi_noLE_de <- msc_gam_combi_noLE_de[, cells_de]
```


```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
all_asex_de_bal_nosct <- all_asex_merged_int_no_actv_harmony_bal
all_asex_de_bal_nosct[["SCT"]] <- NULL
DefaultAssay(all_asex_de_bal_nosct) <- "RNA"

```

```{r, message=FALSE, warning=FALSE}
all_gns = unique(rownames(all_asex_de_bal_nosct))

m2_asex_ap2g_bal_seu <- all_asex_de_bal_nosct %>%
         NormalizeData(., normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE) %>%
         ScaleData(., features = all_gns,  verbose = FALSE)

m2_asex_ap2g_bal_seu <- list(m2_asex_ap2g_bal_seu) %>% set_names("ap2g")

```

### Actual MAST DE between pairwise strains and fieldvslab 

MAST DE adjusting for pseudotime as a latent variable

```{r, message=FALSE, warning=FALSE}
de_idents <- c("ap2g")
m2_asex_ap2g_bal_mast<- map2(m2_asex_ap2g_bal_seu, de_idents, ~SetIdent(.x, value=.y))

```

```{r, message=FALSE, warning=FALSE}
saveRDS(m2_asex_ap2g_bal_mast, 'data/processed/Pf/m2_all/m2_asex_ap2g_bal_mast.RDS')

```


```{r, message=FALSE, warning=FALSE}
##shortcut to run relevant donors subsets only

```


```{r, message=FALSE, warning=FALSE, eval=F}
##!! NB Rerun only when the DE inputs change - takes looooonng

## Mast DE of dataset with male low (ml) included - shouldnt have any contribution as each stage handled differently
## Including lab and field strain DE
## !!!min.cells.group added for mh in the strain_dset where there's only balancing in lab VS field - we need to remove later

pts <- "pseudotime"
pts_all<- c(rep(pts,1))
pts_all<- pts

### !!! NOTE - RUNNING FOR NECESSARY DONORS, WILL REVERT LATER
ap2g_ub_pairwise_mkrs_pt_ml <- list(m2_asex_ap2g_bal_mast, pts, de_idents, names(m2_asex_ap2g_bal_mast)) %>%
  pmap(.,  ~{
    # ide = ..3
    # a = combn(levels(droplevels(..1@meta.data[,..3])), 2,simplify = F) 
    a = combn(unique(..1@meta.data[,..3]), 2,simplify = F) 
    obj = ..1
    pseudos = ..2
    
    print(class(obj))
    print(paste("running ...", ..4))
    map(a, ~print(paste(.x[1], "n ", .x[2])))
    
    map(a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], latent.vars = pseudos, logfc.threshold = 0, test.use = 'MAST', min.cells.group = 1) %>% rownames_to_column('gene_id')) %>%
      set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
      bind_rows(., .id = 'pw_comparison')
    
  }#, 'missing genes')
  )

saveRDS(ap2g_ub_pairwise_mkrs_pt_ml,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml.RDS')
# saveRDS(ap2g_ub_pairwise_mkrs_pt_ml_don,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml_don.RDS')
# c(ap2g_ub_pairwise_mkrs_pt_ml_lf, ap2g_ub_pairwise_mkrs_pt_ml)



```


```{r, message=FALSE, warning=FALSE, eval=F}
##!! NB Rerun only when the DE inputs change - takes looooonng
### Getting all genes for gsea - saved as RDS and analysis done in functional_GO_n_lit_comparison.RMD

# ap2g_ub_pairwise_mkrs_pt_ml_gsea_full <- list(m2_asex_ap2g_bal_mast, rep(pts,2), de_idents) %>%
#   pmap( ~{
#     # ide = ..3
#     a = combn(levels(..1@meta.data[,..3]), 2,simplify = F) 
#     obj = ..1
#     pseudos = ..2
#     map(a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], latent.vars = pseudos, logfc.threshold = 0, test.use = 'MAST', min.pct = -Inf, min.diff.pct = -Inf) %>% rownames_to_column('gene_id')) %>%
#       set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
#       bind_rows(., .id = 'pw_comparison')
#     
#   })
# saveRDS(ap2g_ub_pairwise_mkrs_pt_ml_gsea_full,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml_gsea_full.RDS')

```


```{r, message=FALSE, warning=FALSE}
ap2g_ub_pairwise_mkrs_pt_ml<- readRDS('data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml.RDS')

```


MAST DE not adjusting for pseudotime (no pseudotime - npt)
```{r, message=FALSE, warning=FALSE, eval=F}
##!! NB Rerun only when the DE inputs change - takes looooonng - also not necessary. Only used to assess effect of adjusting for pseudotime
ap2g_ub_pairwise_mkrs_npt_ml <- list(m2_asex_ap2g_bal_mast, rep(pts,1), de_idents) %>%
  pmap( ~{
    # a = combn(levels(..1@meta.data[,..3]), 2,simplify = F) 
    a = combn(unique(..1@meta.data[,..3]), 2,simplify = F) 
    obj = ..1
    pseudos = ..2
    map( a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], logfc.threshold = 0, test.use = 'MAST') %>% rownames_to_column('gene_id')) %>%
      set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
      bind_rows(., .id = 'pw_comparison')
    
  })

saveRDS(ap2g_ub_pairwise_mkrs_npt_ml,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_npt_ml.RDS')

```

```{r, message=FALSE, warning=FALSE, eval=FALSE}
ap2g_ub_pairwise_mkrs_npt_ml<- readRDS('data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_npt_ml.RDS')

```

#### Comparing MAST DE results when we adjust for pseudotime and when we dont

```{r, message=FALSE, warning=FALSE}
# ap2g_ub_pairwise_mkrs_npt_ml <- ap2g_ub_pairwise_mkrs_npt_ml %>% map(. %>% mutate(across(pw_comparison, ~str_replace_all(.,c('Field_vs_Lab' = 'Field_vs_Lab')))))

```

#### Comparing MAST DE results when we adjust for pseudotime and when we dont
```{r, fig.width= 8, fig.height=3, eval=F}
## !! NB - also not necessary. Only used to assess effect of adjusting for pseudotime
list(ap2g_ub_pairwise_mkrs_pt_ml, ap2g_ub_pairwise_mkrs_npt_ml) %>%
  pmap(
    ~..1 %>% 
      left_join(., ..2, by = c('gene_id', 'pw_comparison'), suffix =c(".pt", ".npt")) %>%
      ggplot(., aes(x=log10(p_val_adj.pt), y=log10(p_val_adj.npt))) + 
      geom_point() + 
      facet_grid(.~pw_comparison, scales = "free_x") +
      geom_vline( xintercept = -3)+
      geom_hline( yintercept = -3) +
      geom_abline()
  )

```


#### Comparing MAST DE results when we adjust for pseudotime and when we dont
```{r, fig.width= 8, fig.height=3, eval=F}
## !! NB - also not necessary. Only used to assess effect of adjusting for pseudotime
ap2g_ub_pairwise_mkrs_pt_ml

```


```{r, warning=FALSE, message=FALSE}
ap2g_ub_pairwise_mkrs_pt_ml_anot <- map(ap2g_ub_pairwise_mkrs_pt_ml, ~.x %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")]))

ap2g_ub_pairwise_mkrs_pt_ml_anot
```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more ap2g UMI
map(ap2g_ub_pairwise_mkrs_pt_ml_anot, ~.x  %>% filter(p_val_adj < 0.05 & avg_log2FC > 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in ap2g positive cells

```

```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more ap2g UMI
imap(ap2g_ub_pairwise_mkrs_pt_ml_anot, ~.x %>% filter(p_val_adj < 0.05 & avg_log2FC < 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in ap2g positive cells

```

```{r}
##!!!!!NOTE - remove the empty MSC1_fle
# ap2g_ub_pairwise_mkrs_pt_ml_sst <- ap2g_ub_pairwise_mkrs_pt_ml[c(1:5, 7:20)]
# m2_asex_ap2g_bal_seu_sst <- m2_asex_ap2g_bal_seu[c(1:5,7:20)]
ap2g_ub_pairwise_mkrs_pt_ml_sst <- ap2g_ub_pairwise_mkrs_pt_ml
m2_asex_ap2g_bal_seu_sst <- m2_asex_ap2g_bal_seu
cond_gns_005_ub_de_sst <- list(m2_asex_ap2g_bal_batch_de)
pts_sst <- pts

```

### MAST VS Tradeseq analsys
####  MAST gene QC
List genes expressed in less than 5 cells in each subset objects used as MAST input that were also removed when running tradeseq fitGAM

```{r}
# map2(ap2g_ub_pairwise_mkrs_pt_ml_sst, m2_asex_ap2g_bal_seu_sst, ~.x %>% filter(gene_id %in% rownames(.y)[!(rowSums(.y@assays$RNA@counts != 0) > 5)])) %>% bind_rows(., .id = 'Stage')

map2(ap2g_ub_pairwise_mkrs_pt_ml_sst, m2_asex_ap2g_bal_seu_sst, ~.x %>% filter(gene_id %in% rownames(.y)[!(rowSums(.y@assays$RNA@layers$counts != 0) > 5)])) %>% bind_rows(., .id = 'Stage')

```

Remove genes above ????NB NB - refer line 734
```{r}
ap2g_ub_pairwise_mkrs_pt_ml_sst <- map2(ap2g_ub_pairwise_mkrs_pt_ml_sst, m2_asex_ap2g_bal_seu_sst, ~.x %>% filter(gene_id %in% rownames(.y)[(rowSums(.y@assays$RNA@layers$counts != 0) > 5)]))

```

#### MAST & Tradeseq merge
Merge MAST results with tde results so as to use cut-off accounting for both methods (P<0.05 MAST & waldstat >10 TDE)
```{r, fig.width= 8, fig.height=3, warning=FALSE}
mast_tde_corr_tbls_w_na <- list(cond_gns_005_ub_de_sst[1], ap2g_ub_pairwise_mkrs_pt_ml_sst[1]) %>% 
    # pmap(~{ left_join(..1 %>% dplyr::select(gene_id, waldStat, pvalue), ..2, by = 'gene_id') %>%
    pmap(~{ full_join(..2, ..1 %>% dplyr::select(gene_id, waldStat, pvalue),  by = 'gene_id')})
  

```


##### MAST & Tradeseq NA genes 
Check the genes with missing values in either test - tradeseq/MAST
Thrown out probably because of not meeting the log fold threshold
```{r, fig.width= 8, fig.height=3, warning=FALSE}
##mast missing
map(mast_tde_corr_tbls_w_na, ~.x %>% dplyr::count(pw_comparison, is.na(waldStat), is.na(p_val_adj))) %>% bind_rows(.id = 'stage') %>% filter(`is.na(p_val_adj)` == T)

##tde missing
map(mast_tde_corr_tbls_w_na, ~.x %>% dplyr::count(pw_comparison, is.na(waldStat), is.na(p_val_adj))) %>% bind_rows(.id = 'stage') %>% filter(`is.na(waldStat)` == T)

```

Remove mast missing genes above
```{r}
map(mast_tde_corr_tbls_w_na,  ~.x %>% filter(is.na(waldStat)) %>% ggplot(., aes(x = pw_comparison, y =avg_log2FC)) +geom_violin())
```

```{r}
map(mast_tde_corr_tbls_w_na,  ~.x %>% filter(is.na(waldStat)) %>% ggplot(., aes(x = log10(p_val_adj), y =avg_log2FC)) +geom_point())
```



```{r}
map(mast_tde_corr_tbls_w_na,  ~.x %>% filter(is.na(p_val_adj)) %>% ggplot(., aes(x = waldStat, y =log10(pvalue))) +geom_point() + geom_hline(yintercept = 0.05) +geom_vline(xintercept = 10))
```

```{r}
map(mast_tde_corr_tbls_w_na,  ~.x %>% filter(is.na(waldStat) |is.na(p_val_adj)) %>% dim)

```

```{r}
mast_tde_corr_tbls <- map(mast_tde_corr_tbls_w_na,  ~.x %>% filter(!is.na(waldStat) & !is.na(p_val_adj)))

```

```{r}
##Number of DE genes
map(mast_tde_corr_tbls, ~.x |> filter(p_val_adj < 0.05 & waldStat>10) |> pull(gene_id) |> unique() |> length())

```

```{r}
saveRDS(mast_tde_corr_tbls, 'data/processed/Pf/m2_all/mast_tde_corr_tbls.RDS')

```


#### MAST & Tradeseq correlation
function to plot correlation between MAST and tradeseq and visualize disparities between the 2 methods for the different pairwise strain comparisons
```{r}
pw_corr_plot <- function(g_tbls = mast_tde_corr_tbls){
  for (i in 1:length(g_tbls)) {
    tryCatch({
      pt<-ggplot(g_tbls[[i]], aes(x=waldStat, y=log10(p_val_adj))) +
              # geom_point() + 
              # facet_grid(.~pw_comparison, scales = "free_x") +
              geom_point(aes(color = pw_comparison)) +
              stat_cor(method = "pearson", label.x = 50, label.y = -6)+ 
              geom_vline( xintercept = 10)+
              geom_hline( yintercept = log10(0.05))+
              theme_classic()+
              theme(axis.text=element_text(size=20), 
                    axis.title=element_text(size=18,face="bold"), 
                    legend.text = element_text(size = 16) ,
                    legend.title = element_text(size = 18, face = "bold"),
                    strip.text.x = element_text(size = 16, face = "bold"))
      return(pt)
    }, error = function(e) {
      message(as.character('ss')) # exclude this line to return NULL quietly
      return(NULL)
    })
  }
}
```

```{r}
pw_corr_plot <- function(g_tbls = mast_tde_corr_tbls){

      ggplot(g_tbls, aes(x=waldStat, y=-log10(p_val_adj))) +
              # geom_point() + 
              # facet_grid(.~pw_comparison, scales = "free_x") +
              geom_point(aes(color = pw_comparison)) +
              # scale_color_manual(values = fld_strns_nf54_7g8_comps_col, name = "Comparison") +
              stat_cor(method = "pearson", label.x = 40, label.y = 5)+ 
              geom_vline( xintercept = 10)+
              geom_hline( yintercept = -log10(0.05))+
              labs(x = "tradeSeq waldStat", y = "Adjusted\nMAST P(-log10)") +
              theme_classic()+
              theme(axis.text=element_text(size=20), 
                    axis.title=element_text(size=18,face="bold"), 
                    legend.text = element_text(size = 16) ,
                    legend.title = element_text(size = 18, face = "bold"),
                    strip.text.x = element_text(size = 16, face = "bold"))
}
```


Correlation between MAST and tradeseq 
```{r}
# pw_corr_plot(g_tbls = mast_tde_corr_tbls)
de_corr_plts <- map(mast_tde_corr_tbls, ~{
  pw_corr_plot(.x )
  } )

de_corr_plts
```






######################## END integrate cellranger filtered with rescued asexuals ########################

########END TRY SCT ########


```{r, warning=FALSE, message=FALSE}
## mget

imap(all_ed, ~FeaturePlot(.x, features = "PF3D7-1328800", order = T) + labs(title = .y))
imap(all_ed, ~FeaturePlot(.x, features = "PF3D7-1328800", reduction = "pca", order = T) + labs(title = .y))
```




##############START - Painter EArly asexual markers##############
```{r}
painter_raw = readxl::read_xlsx("../published_dsets/painter_29985403_48hr_dynamics/Expression_Peak_Timing_48h_IDC.xlsx", sheet = "Estimated Total Abundance", .name_repair = "universal") %>% mutate(across(Gene.ID, ~str_replace(., "_", "-"))) %>% rename_all(., ~str_remove(., "\\.\\.")) %>% rename_with(.fn = ~str_remove(paste0("hpi_", .x), ".hpi"), .cols = contains("hpi")) %>% rename("peak_hpi" = `hpi_Peak.Tim.`)
```

```{r}
painter_raw 
```


```{r}

# Get the column names that start with 'hpi'
hpi_cols <- grep("^hpi_[0-9]+$", names(painter_raw), value = TRUE)

# Dynamically split the columns into groups of 5
col_groups <- split(hpi_cols, ceiling(seq_along(hpi_cols) / 10))

# Compute the row-wise means for each group of 5 columns
means_list <- lapply(col_groups, function(cols) {
  rowMeans(painter_raw[, cols], na.rm = TRUE)
})

# Combine the results into a single data frame
means_df <- data.frame(means_list)
names(means_df)<- paste0("mean_", seq_along(means_df))

# Combine the original dataframe with the means
painter_raw_summary <- cbind(painter_raw[, c("Gene.ID", "Annotation")], means_df)

```


```{r}
painter_raw_summary  %>% arrange(desc(mean_1))
```

```{r}
painter_raw %>% filter(Gene.ID %in% c("PF3D7-1102800", "PF3D7-0202500", "PF3D7-1370300")) %>% 
  pivot_longer(cols = starts_with("hpi_"), names_to = "time", values_to = "expresion") %>% 
  mutate(across(time, ~as.numeric(str_extract(., "[0-9]+")))) %>% 
  ggplot(aes(x = time, y = expresion, color = Gene.ID)) + 
  geom_point()
```


```{r, warning=FALSE, message=FALSE}
## mget
imap(all_ed, ~FeaturePlot(.x, features = c("PF3D7-0501300", "PF3D7-1102800"), order = T) + labs(title = .y))
imap(all_ed, ~FeaturePlot(.x, features = c("PF3D7-0501300", "PF3D7-1102800"), reduction = "pca", order = T) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
## mget
FeaturePlot(all_msc_100_700, features = c("PF3D7-1370300", "PF3D7-1102800"), order = T, pt.size = 0.5)
FeaturePlot(all_msc_100_700, features = c("PF3D7-1370300", "PF3D7-1102800"), reduction = "pca", order = T, pt.size = 0.5) 
```

##############END - Painter EArly asexual markers##############

## scmap annotation

### Preprocessing references

#### msc_combi reference dataset
```{r, message=FALSE, warning=FALSE, eval = T}
## v3labmscs.ref.sce.stgasx.RDS generated in m2_QC_prelim_stage_anno 
v3labmscs.ref.sce.stgasx <- readRDS("data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS")
```


```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_edrops50_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd_edrops50.RDS" --stg_labl "stageHL_ref"

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```
SCT for raw
```{r, eval=T}
gogo()
## !! IMPORTANT - Include a new parameter - a stage column to contain the labels that we map to Collapse asexual stages into 1 "Asexual"

## Run scmap on farm like below

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_edrops50_sct.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd_edrops50.RDS" --stg_labl "stageHL_ref"

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC48,MSC68,MSC14}/seu_obj/bpc_seu_sct.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```

```{r, message=F, warning=F}
##Read the annotations
sample_name_ed <- sample_name_all[!str_detect(sample_name_all, "_150")]
sample_name_ed <- sample_name_all
```

```{r, message=F, warning=F}
##Read the annotations
m2_raw_anots <- list.files("data/processed/Pf", pattern = "*prelim_anotd.*RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/|/seu_obj/|prelim_anotd|.RDS'))) %>% 
  # .[sample_name_ed] %>%
  .[sample_name] %>%
  .[!is.na(.)] %>%
  map(readRDS)
```

```{r}
## Adding annotations to query cells
m2_raw_anotd_preQC <- list(m2_raw_anots, all_ed[sample_name_ed]) %>%
  pmap(~ ..1 %>%
         column_to_rownames("cell_bc") %>%
         mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
         # mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
         mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_replace_all(.x, c("gametocyte" = "Gametocyte", "late" = "Late", "early" = "Early")))) %>%
         mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
         AddMetaData(..2, .) 
  )

```

```{r}
## Adding annotations to query cells
m2_raw_anotd_preQC <- list(m2_raw_anotd_preQC) %>%
  pmap(~ ..1@meta.data %>%
         mutate(gene100 = ifelse(nFeature_RNA > 100, "yes", "no"),
                trancript100 = ifelse(nCount_RNA > 100, "yes", "no")) %>%
         AddMetaData(..1, .) 
  )

```

```{r}
# ## Adding annotations to query cells
# m2_raw_anotd_preQC <- list(m21_gn150_anots, all_ed) %>%
#   pmap(~ ..1 %>%
#          column_to_rownames("cell_bc") %>%
#          mutate(across(Stage_MSC, ~na_if(., "unassigned"))) %>%
#          mutate(across(c(starts_with("stage"), starts_with("Stage")), ~str_to_sentence(.x))) %>%
#          mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
#          AddMetaData(..2, .) 
#   )

```

```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'Stage_MSC', cols = sp_fld_colors) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'stage_lab1', cols = sp_fld_colors2)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}

imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'gene100')+ labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(Stage_MSC, stage_lab1))

```


```{r}
##Modify the stage labels for MSC14
# m2_raw_anotd_preQC[[4]] <- m2_raw_anotd_preQC[[4]]@meta.data %>%
m2_raw_anotd_preQC <- pmap(list(m2_raw_anotd_preQC, rep(0.4,length(sample_name_all)), c(rep('ring|trophozoite|developing', length(sample_name_all)))), ~{
  ..1@meta.data %>%
    mutate(across(Stage_MSC, as.character)) %>%
    # mutate(Stage_MSC = Stage_MSC) %>%
    mutate(StagePrelim_lf = case_when(
      is.na(Stage_MSC) & sim_lab1 > 0.35 & str_detect(stage_lab1, ..3) & str_detect(stage_lab2, ..3) & str_detect(stage_lab3, ..3) ~ stage_lab1, ## lower cos sim of 0.35 for asexuals instead of 0.4
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Male|male') & str_detect(stage_lab2, 'Male|male') & str_detect(stage_lab3, 'Male|male') ~ stage_lab1,
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Female|female') & str_detect(stage_lab2, 'Female|female') & str_detect(stage_lab3, 'Female|female') ~ stage_lab1,
      TRUE ~ Stage_MSC)#,
      # across(StagePrelim, str_to_sentence)
    )%>%
    mutate(StagePrelim = case_when(
      str_detect(StagePrelim_lf, "developing") ~ "Early trophozoite", ## !!NB - label developing gams as early trophs the closest stage
      str_detect(StagePrelim_lf, "schizont") & str_detect(stage_lab2, "ring|trophozoite")  & str_detect(stage_lab3, "ring|trophozoite") ~ str_to_sentence(stage_lab2),
      str_detect(StagePrelim_lf, "late|early") ~ str_to_sentence(StagePrelim_lf),
      str_detect(StagePrelim_lf, "gametocyte") ~ str_to_sentence(str_remove_all(StagePrelim_lf, "gametocyte \\(|\\)")),
      is.na(StagePrelim_lf) ~ "Unassigned",
      TRUE ~ StagePrelim_lf)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(StagePrelimC = case_when(
      str_detect(StagePrelim, "ring|trophozoite|developing|schizont") ~ "Asexual",
      TRUE ~ StagePrelim)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(across(c(StagePrelim, StagePrelimC), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    dplyr::select(StagePrelim, StagePrelim_lf, StagePrelimC) %>%
    # mutate(across(starts_with('Stage_V'), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    AddMetaData(.x, .)
})

```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~.x@meta.data %>% dplyr::count(Stage_MSC, StagePrelim, stage_lab1))
# map(m2_raw_anotd_preQC, ~.x@meta.data %>% dplyr::count(StagePrelim, StagePrelimC))
```

```{r, warning=FALSE, message=FALSE}
imap(m2_raw_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = F, group.by = 'StagePrelim', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}
# m21_gn150_anots[["MSC51"]]
```


```{r, warning=FALSE, message=FALSE}
imap(m2_raw_anotd_preQC, ~FeaturePlot(.x, dims = c(1,2), features = "nFeature_RNA")+ labs(title = .y))
```

```{r, results='asis',  fig.width = 8, fig.height = 4}
##adding clusters that are used to display mitochondrial count

plts <- list(m2_raw_anotd_preQC, names(m2_raw_anotd_preQC)) %>%
  pmap(., ~{
    
    cell_qc_plot <-  ..1 %>% DimPlot(., pt.size = 1, reduction = 'umap', dims = c(1,2), group.by = 'StagePrelim', cols = sp_fld_colors) +labs(title = ..2)#,
    
    cell_qc_plot$data <- cell_qc_plot$data %>% rownames_to_column('bcode') %>% left_join(., ..1@meta.data[,'seurat_clusters', drop = F] %>% rownames_to_column('bcode'), by='bcode') %>% column_to_rownames('bcode')
    
    # print(str(cell_qc_plot$data))
    
    cell_qc_plot<- LabelClusters(cell_qc_plot, id = "seurat_clusters", size = 9, repel = T)
    
    return(cell_qc_plot)

  })
```


```{r, results='asis',  fig.width = 8, fig.height = 4}
plts

```

compare filt and raw with gn count cut off of 150
```{r, message=F, warning=F}
# go()
##Read the annotations
m21_all_anots_mdata <- map2(m2_raw_anotd_preQC[1:3], m2_raw_anotd_preQC[4:6], ~{
  .x@meta.data %>% 
    rownames_to_column("bcode") %>%
    full_join(., .y@meta.data %>% rownames_to_column("bcode"), suffix = c("_filt", "_gn_150"), by = "bcode") %>%
    mutate(dset = case_when(!is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "both",
                            !is.na(nFeature_RNA_filt) & is.na(nFeature_RNA_gn_150)  ~ "filt",
                            is.na(nFeature_RNA_filt) & !is.na(nFeature_RNA_gn_150)  ~ "gn_150"))
})
```


```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% count(dset))
map(m21_all_anots_mdata, ~.x %>% count(dset, stage_lab1_gn_150))

```

```{r, message=F, warning=F}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% filter(dset == "gn_150") %>% count(StagePrelim_gn_150, Stage_MSC_gn_150, stage_lab1_gn_150, stage_lab2_gn_150, stage_lab3_gn_150))

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = sim_lab1_gn_150) ) +
      geom_boxplot() +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```


```{r, message=F, warning=F, fig.width=14, fig.height=3}
# go()
##Read the annotations
map(m21_all_anots_mdata, ~.x %>% 
      ggplot(aes(x = dset, y = nFeature_RNA_gn_150) ) +
      geom_boxplot() +
      geom_hline(yintercept = 100) +
      facet_wrap(stage_lab1_gn_150 ~., nrow = 1)
      )

```

Umap with stage labels in legend

```{r, fig.width = 7, fig.height = 5, eval=F}
trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.2, "cm")
)

# Assign the female High_low stage of the projected cell to the nearest cell in the reference dataset
# msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(msc_combi.sce))), length(m21_gn150_jc))
msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(v3labmscs.ref.sce.stgasx))), length(m21_gn150_jc))


for (i in 1:length(msc_qcd_anotd_mali_mapped14HL)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m2_raw_anotd_preQC[[i]]$StagePrelim
}


for (i in 1:length(m2_raw_anotd_preQC)){
  msc_qcd_anotd_mali_mapped14HL[[i]][m21_gn150_jc[[i]]$Stage_MSC$cells[1,]] <- m2_raw_anotd_preQC[[i]]$StagePrelim
  # msc_combi@meta.data[, paste0("stage_", names(m2_raw_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  v3labmscs.ref.sce.stgasx@colData[, paste0("stage_", names(m2_raw_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  
}

labs_order <- c('Late ring', 'Early trophozoite','Female (HE)',  'Male (HE)', 'Female (LE)', 'Male (LE)','Gametocyte (developing)', 'Gametocyte (male)','Gametocyte (female)',  'Atlas')
names(labs_order)<- labs_order

umap_theme <- theme(legend.position = 'right',
                    axis.title=element_text(size=12,face="bold",hjust = 0),
                    axis.text=element_text(size=18), 
                    plot.title = element_blank(),
                    legend.title=element_text(size=14,face="bold"),
                    legend.text=element_text(size=14),
                    legend.spacing.y = unit(0.25, "cm"),
                    axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed"))
)  


# Idents(msc_combi) <- 'MSC14_V3'
# m14_scmap_mca_umap <- map(names(m2_raw_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), 
#         order = labs_order) +
m14_scmap_mca_umap <- map(names(m2_raw_anotd_preQC), ~{
  cbind(reducedDims(v3labmscs.ref.sce.stgasx)$PCA, colData(v3labmscs.ref.sce.stgasx)[, c("stageHL_ref", paste0("stage_",.x)), drop=F]) %>% 
    data.frame() %>% 
    mutate(od_stg = droplevels(factor(!!rlang::sym(paste0("stage_",.x)), levels = rev(stg_refnd_lvls)))) %>%
    arrange(od_stg) %>%
    ggplot(aes(y = PC2, x=PC3, color = od_stg)) +
    geom_point() +
    # scale_color_manual(breaks=labs_order, values = sp_fld_colors, labels = str_wrap(labs_order, 10))+
    scale_color_manual(values = sp_fld_colors)+
    umap_theme+
    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
           x= trunc_axis,
           y= trunc_axis)+
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(breaks = NULL)
})

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap, 'plots/fig4/m14_scmap_mca_umap')

m14_scmap_mca_umap


```

Umap with stage labels inside
```{r, fig.width = 7, fig.height = 5, eval=F}

m14_scmap_mca_umap_labs_in <- map(names(m2_raw_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), label = T, order = labs_order, repel = T, label.size = 5) +
                                    scale_color_manual(values = sp_fld_colors)+ 
                                    umap_theme+
                                    theme(axis.title=element_text(size=18,face="bold"), 
                                          plot.title = element_blank(),
                                          legend.position='None'
                                    ) +
                                    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
                                           x= trunc_axis,
                                           y= trunc_axis)+
                                    scale_x_continuous(breaks = NULL) +
                                    scale_y_continuous(breaks = NULL)
)

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap_labs_in, 'plots/fig4/m14_scmap_mca_umap_labs_in.RDS')


m14_scmap_mca_umap_labs_in


```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim, Stage_MSC, stage_lab1, stage_lab2))

```


```{r, warning=FALSE, message=FALSE}
saveRDS(m2_raw_anotd_preQC, "data/processed/Pf/m2_raw_anotd_preQC.RDS")
```


```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data) %>% saveRDS(., "data/processed/Pf/m2_raw_anotd_preQC_mdata.RDS")

```

```{r, warning=FALSE, message=FALSE}

map(m2_raw_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim))

```




```{r, fig.width = 12, fig.height = 4}

##Plotting top two scmap stages among the top 10 with their similarities for MSC14
# 
# stage_fct_labs <- msc_qcd_anotd[[1]]@meta.data  %>% dplyr::group_by(stageHL, seurat_clusters) %>% dplyr::group_keys() %>% dplyr::mutate(across(c(stageHL, seurat_clusters), as.character), across(seurat_clusters, ~paste0("Cluster ",., ':\n',stageHL))) %>% deframe()
# 
# scm_barpt <- msc_qcd_anotd[[1]]@meta.data %>% 
#   dplyr::select(matches('s.*lab.'), stageHL) %>% 
#   rownames_to_column('cell_bc')%>%
#   pivot_longer(-c('cell_bc', 'stageHL'),
#                names_to = c( ".value", "Label"),
#                names_pattern = "(s.*)_(l.*)")%>% 
#   filter(!is.na(sim)) %>%
#   mutate(across(sim, as.numeric)) %>% 
#   mutate(across(Label, ~str_replace(., 'lab','Label '))) %>%
#   mutate(across(stage, ~droplevels(factor(., stg_refnd_lvls)))) %>%
#   ggplot(., aes(y = sim, x = stage, color = Label)) +
#   geom_boxplot(aes(fill = Label))+
#   scale_fill_manual(values =  c('Label 1' = '#007FFF', 'Label 2' = '#FFFF9F'), name = "Label",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 2)
#   )+
#   scale_color_manual(values =  c('Label 1' = 'black', 'Label 2' = 'black'), name = "Label")+ 
#   geom_point(aes(group=Label), size = 0.5, position = position_jitterdodge(jitter.width = 0.2))+
#   new_scale_fill()+
#   geom_tile(data = data.frame(x = c( 'Late ring', 'Early trophozoite', 'Gametocyte (developing)', 'Gametocyte (female)', 'Gametocyte (male)')), inherit.aes = F, aes(x = x, y = 0.37, fill = x, height = 0.03)) +
#   scale_fill_manual(values =  sp_fld_colors, name = "MCA stage",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 3)
#                     # guide = 'none'
#   )+
#   scale_y_continuous(expand = expansion(mult = c(0.01, 0.05))) +
#   facet_grid(.~stageHL, labeller = labeller(stageHL = stage_fct_labs)) +
#   geom_hline(yintercept = 0.4)+
#   labs(y = 'scmap \nsimilarity scores', x = 'MCA stage') +
#   theme_classic() +
#   theme(axis.text.x = element_blank(),
#         axis.title=element_text(size=16,face="plain"), 
#         axis.text = element_text(size=15), 
#         axis.ticks.x = element_blank(),
#         strip.text =element_text(size=16,face="plain", colour = 'black'),
#         panel.border=element_rect(colour="black",fill='transparent'),
#         legend.title = element_text(size=16,face="bold"),
#         legend.text = element_text(size=15),
#         legend.position = 'bottom')
# 
# scm_barpt
```


Run souporcell like below for the filtered barcodes of all infections
```{r}
gogo()

# nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/MSC*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```



Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
gogo()

#  nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373{63,72,89}/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/raw/Pf_mxd/MSC*/outs/sk21_bcodes_pp/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali/data/processed/Pf/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```
