---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  library(furrr)
  library(future.batchtools)
  library(parallel)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
})
```


```{r, setup}

# knitr::opts_knit$set(root.dir = '/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/Mali2')
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/')
```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("mali22_code_lnk/pf_common_vars.R")
```


initialize variable for the sample names

```{r}
irods_id_sk21_mdata_cln <- read_csv("data/raw/irods_id_sk21_mdata_cln.csv") %>%
  mutate(sample_nm = str_remove_all(sample_name, "_JC")) %>%
  arrange(sample_nm)

smpl_nms <-irods_id_sk21_mdata_cln$sample_nm
#c("MSC24", "MSC33", "MSC37", "MSC38", "MSC39", "MSC40", "MSC45", "MSC48", "MSC49", "MSC50_MACS", "MSC50_SLO", "MSC53", "MSC54", "MSC55", "MSC57", "MSC60", "MSC66", "MSC67", "MSC68", "MSC70")
# source_irods <- c("5736STDY11771545", "5736STDY11771544", "5736STDY11771536", "5736STDY11771546", "5736STDY11771535")
# sample_name <- c("MSC1272", "MSC13", "MSC3", "MSC14", "MSC1")
```


```{r}
mrc_sk21_summary <- read_csv("data/raw/MRC_Sample_Analysis_Summary_Summary.csv", skip = 1) 
```


```{r}
mrc_sk21_summary %>% filter(str_detect(mixed_infection, "pf") & (str_detect(shortids, "SetB|SetC") | Sample_Name %in% c("MSC1", "MSC3", "MSC13", "MSC14", "MSC12C72")) )
```

```{r}
## Getting only the jumpcode run where a sample was ran twice (non-jumpcode and jumpcode)
pf_22_jcode <- mrc_sk21_summary %>% 
  filter(str_detect(mixed_infection, "pf") & (str_detect(shortids, "SetB|SetC") | Sample_Name %in% c("MSC1", "MSC3", "MSC13", "MSC14", "MSC12C72") )) %>%
  mutate(sample_nm = str_remove_all(Sample_Name, "_JC|C72")) %>%
  group_by(sample_nm) %>%
  add_count(name = "runs") %>%
  filter(runs == 1 | runs == 2 & str_detect(Sample_Name, "JC")) 

```


```{r}
## Getting  pf solo infections
pf_solo_mdata <- pf_22_jcode %>% 
  filter(str_detect(mixed_infection2, "solo")) %>%
  group_by(sample_nm) %>%
  add_count(name = "runs") %>%
  filter(pf > 200) %>%
  filter(!(Sample_Name %in% c("MSC65", "MSC69-71"))) %>%
  filter(runs == 1 | runs ==2 & str_detect(Sample_Name, "JC")) 
```

```{r}
pf_solo_mdata
```


```{r}
## Getting both pf solo infections and those mixed with other species
pf_solo_mixed_mdata <- pf_22_jcode %>% 
  group_by(sample_nm) %>%
  add_count(name = "runs") %>%
  filter(pf > 200) %>%
  filter(!(Sample_Name %in% c("MSC65", "MSC69-71"))) %>%
  filter(runs == 1 | runs ==2 & str_detect(Sample_Name, "JC")) %>%
  rename('irods_id' = Sanger_SampleID)
```

```{r}
pf_solo_mixed_mdata
```


Copy the irods metadata file from local into farm
```{r, eval=T}
write_csv(pf_solo_mixed_mdata, "data/raw/pf_solo_mixed_mdata.csv")
```

```{r, eval=T}
pf_solo_mixed_mdata_decode <- pf_solo_mixed_mdata %>% select(sample_nm, irods_id, mixed_infection2)

write_csv(pf_solo_mixed_mdata_decode, "data/raw/pf_solo_mixed_mdata_decode.csv")
```

## Genotype clustering for MSC

Read in clusters.tsv files with assignment of cells to strains and PCA plots metrics

Transfer clusters.tsv files with assignment of cells to strains and PCA plots metrics for the best K from farm to my local Mac

```{r}

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```

NB-FARM_CODE
```{r, eval=T}
##Copy h5 files from cellranger and making relevant directories

for (i in 1:length(sample_name)) {
  
  ## Create new directories
  system(paste0('mkdir -p data/raw/Pf/', sample_name[i], '/outs/filtered_feature_bc_matrix'))
  system(paste0('mkdir -p data/processed/Pf/', sample_name[i], '/bpcells'))
  
  ##Transfers
  ##souporcell genotyping files
  system(
    paste0('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/', source_irods[i], '/outs/filtered_feature_bc_matrix.h5 data/raw/Pf/', sample_name[i], '/outs/')
  )
  system(
    paste0('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/', source_irods[i], '/outs/filtered_feature_bc_matrix/barcodes.tsv.gz data/raw/Pf/', sample_name[i], '/outs/filtered_feature_bc_matrix/')
  )
  
}


```


```{r}
## List filtered cellranger h5 matrices into R
list.files("data/raw/Pf/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T)
```


```{r, message=F, warning=F, eval=T}
##Write BPcells matrices
# list.files("data/raw/Pf/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/raw/Pf//|/outs.*'))) %>% 
##Write BPcells matrices
list.files("/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf//|/outs.*'))) %>% 
  .[source_irods] %>%
  set_names(names(source_irods)) %>%
  # .[sample_name] %>%
  map(~open_matrix_10x_hdf5(path = .)) %>% 
  imap(., ~write_matrix_dir(mat = .x, dir = paste0('data/processed/Pf/', .y, '/bpcells/counts_mtx'), overwrite = T))
```


```{r, message=F, warning=F, eval=T}
##Read BPcells matrices
all_msc_bp <- list.files("data/processed/", pattern = "counts_mtx$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed//Pf/|/bpcells.*'))) %>% 
  .[sample_name] %>%
  map(~open_matrix_dir(dir = .)) %>% 
  imap(., ~CreateSeuratObject(counts = .x, project = paste0(.y, ".mrna"))) 
```


```{r}
## Mixed species barcodes from sunil
sk21_mixed_sp_decode <- readRDS("data/processed/Part1_sp_barcodes_PfPmPoc.RDS")
```


```{r}
## Get the donor and irods ID for mixed infections
mxd_source_irods <- pf_solo_mixed_mdata[pf_solo_mixed_mdata$mixed_infection2 == "mixed",][,c("sample_nm", "irods_id")] %>% deframe()

```


```{r}
## Write Pf barcodes for mixed species donor for souporcell
imap(mxd_source_irods, ~{
    system(paste0("mkdir -p data/raw/Pf_mxd/", .y, "/outs/sk21_bcodes_pp"))
    write(sk21_mixed_sp_decode$sp_barcodes[[.x]]$pf_filt, paste0("data/raw/Pf_mxd/", .y, "/outs/sk21_bcodes_pp/barcodes.tsv"))
    system(paste0("gzip -f data/raw/Pf_mxd/", .y, "/outs/sk21_bcodes_pp/barcodes.tsv"))
  })

```

Remove non-Pf from the mixed infections
```{r, message=F, warning=F, eval=T}
## Subset 
for(i in 1:length(mxd_source_irods)) {
    all_msc_bp[names(mxd_source_irods)[i]] <- subset(all_msc_bp[[names(mxd_source_irods)[i]]], cells = sk21_mixed_sp_decode$sp_barcodes[[mxd_source_irods[i]]]$pf_filt)
  }

```

```{r, message=F, warning=F, eval=F}
# Convert to dense object - might be necessary in future instances
# all_msc <- all_msc_bp %>%
#   map(., ~{
#     .x[["RNA"]]$counts <- as(object = .x[["RNA"]]$counts, Class = "dgCMatrix")
#     return(.x)
#   })

```


```{r, eval=T}
##write out for normalization processing in farm

for (i in 1:length(sample_name)) {
  
  ## Create new directories
  system(paste0('mkdir -p data/processed/Pf/', sample_name[i], '/seu_obj'))
  
}


```

FARM CODE
```{r, eval=F}
## write out seurat object files for normalization processing in farm
# imap(all_msc, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/seu.RDS")))
```

```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
imap(all_msc_bp, ~saveRDS(.x, paste0("data/processed/Pf/",.y,"/seu_obj/bpc_seu.RDS")))
```

!!!NB - FARM CODE - DONT DELETE

```{r, eval=T}
gogo()
## Run standard normalisation on farm like below
## !!NB - reducing the number of hypervariable genes results in scattered and disjointed clusters. higher hvg number better

# nextflow run ../mali22_code_lnk/nf_scripts/submsn_seu_norm.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_stdnorm.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/seu_obj/bpc_seu.RDS" --normlzn "norm" --umap_name "UMAP_" --hvg 500 --cluster_resoln 0.1 --cl_nm "seurat_clusters" --umap_redctn_nm "umap" --redctn "pca" --integrtd_pc30 "no" --pca_redctn_nm "pca" --resume 
```


```{r, eval=T}
## write out seurat bpcells object files for normalization processing in farm
# imap(all_msc_bp, ~system(paste0("rm data/processed/Pf/",.y,"/seu_obj/seu_norm_elbo.RDS")))
```

Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read BPcells matrices
all_msc <- list.files("data/processed/", pattern = "^bpc_seu_norm.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed//Pf/|/seu_obj.*'))) %>% 
  .[sample_name] %>%
  map(readRDS)

```


Read in seurat objects processed in farm
```{r, message=F, warning=F}
##Read elbo plots
all_msc_elbo <- list.files("data/processed/", pattern = "^bpc_seu_norm_elbo.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed//Pf/|/seu_obj.*'))) %>% 
  .[sample_name] %>%
  map(readRDS)

all_msc_elbo
```

```{r, warning=FALSE, message=FALSE}
imap(all_msc, ~DimPlot(.x, dims = c(1,2), label = T) + labs(title = .y))
```


## scmap annotation

### Preprocessing references

#### msc_combi reference dataset
```{r, eval=FALSE}
##Read MCA V3 blood stage reference dataset
# msc_combi <- readRDS("../sc_velo/data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.4.RDS")
# msc_combi <- readRDS("../Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")
msc_combi <- readRDS("../Mali/data/processed/DB52e_star/msc_combi.RDS")
DefaultAssay(msc_combi) <- "RNA"

##Collapse asexual stages into 1 "Asexual"
msc_combi@meta.data$stageHL_ref_asx <- ifelse(msc_combi@meta.data$stageHL_ref %in% c("Early trophozoite", "Late ring"), "Asexual", msc_combi@meta.data$stageHL_ref)
```

```{r, message=FALSE, warning=FALSE, eval=FALSE}
##Convert to reference to sce
msc_combi.sce <- as.SingleCellExperiment(msc_combi)

##Place counts appropriately into single cell experiment object
# SingleCellExperiment::logcounts(msc_combi.sce) <- log2(SingleCellExperiment::counts(msc_combi.sce) + 1) !!NOTE This is done by as.SingleCellExperiment - putting lognormalized counts needed by scmap into logcounts https://github.com/hemberg-lab/scmap/issues/9
rowData(msc_combi.sce)$feature_symbol <- rownames(msc_combi.sce)
msc_combi.sce <- msc_combi.sce[!duplicated(rownames(msc_combi.sce)), ]

##scmap feature selection
msc_combi.sce <- selectFeatures(msc_combi.sce, suppress_plot = T)

##Cell indexing
set.seed(155)
msc_combi.sce <- indexCell(msc_combi.sce)
```

NB-FARM_CODE
```{r, eval=F}
## transfer integrated object from sunil farm to local
# system('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/for_scmap/v3labmscs.ref.seur.RDS data/raw/Pf/')
# system('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/for_scmap/v3labmscs.ref.sce.RDS data/raw/Pf/')

```

#### full msc_atlas reference dataset
```{r}
##Read MCA V3 + field blood stage reference dataset
# v3labmscs.ref.sce <- readRDS("../Mali2/data/raw/Pf/v3labmscs.ref.sce.RDS")
v3labmscs.ref.sce <- readRDS("data/raw/Pf/v3labmscs.ref.sce.RDS")
```



```{r, message=FALSE, warning=FALSE}
##Collapse asexual stages into 1 "Asexual"
v3labmscs.ref.sce@colData[,c("stageHL_ref", "stageHL_ref_asx")] <- colData(v3labmscs.ref.sce) %>% 
  data.frame() %>% 
  mutate(stageHL_ref = case_when(str_detect(orig.ident, "msc") & stageLR == "late ring" ~ "Late ring", 
                                 str_detect(orig.ident, "msc") & stageLR == "early trophozoite" ~ "Early trophozoite", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (female) LE" ~ "Female LE", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (female)" ~ "Female", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (male) LE" ~ "Male LE", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (male)" ~ "Male", T ~ stageLR)) %>% 
  mutate(stageHL_ref_asx = case_when(stageHL_ref %in% c("Early trophozoite", "Late ring") ~ "Asexual", T ~ stageHL_ref)) %>%
  select(stageHL_ref, stageHL_ref_asx)

```


```{r, message=FALSE, warning=FALSE}
scater::plotReducedDim(v3labmscs.ref.sce, dimred = "PCA", colour_by = "stageHL_ref_asx", ncomponents = c(2,3))
```


```{r, message=FALSE, warning=FALSE, eval = T}
saveRDS(v3labmscs.ref.sce, "data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS")
```

```{r, eval=T}
gogo()
## Run scmap on farm like below
# nextflow run ../mali22_code_lnk/nf_scripts/submsn_scmap.nf --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/scmap.R" --sce_ref "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/v3labmscs.ref.sce.stgasx.RDS" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/seu_obj/bpc_seu_norm.RDS" --w 3 --cos_sim 0.5 --o_file "prelim_anotd.RDS" --stg_labl "stageHL_ref"
```


```{r, message=F, warning=F}
##Read the annotations
all_msc_anots <- list.files("data/processed/", pattern = "prelim_anotd.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed//Pf/|/seu_obj.*'))) %>% 
  .[sample_name] %>% 
  .[!is.na(.)] %>%
  map(readRDS)
```

```{r}
## Adding annotations to query cells
all_msc_anotd_preQC <- list(all_msc_anots, all_msc) %>%
  pmap(~ ..1 %>%
         column_to_rownames("cell_bc") %>%
         mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
         AddMetaData(..2, .) 
  )

```

```{r, warning=FALSE, message=FALSE}

imap(all_msc_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'Stage_MSC', cols = sp_fld_colors) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}

imap(all_msc_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'stage_lab1', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data %>% dplyr::count(Stage_MSC, stage_lab1))

```


```{r}
##Modify the stage labels for MSC14
# all_msc_anotd_preQC[[4]] <- all_msc_anotd_preQC[[4]]@meta.data %>%
all_msc_anotd_preQC <- pmap(list(all_msc_anotd_preQC, rep(0.4,length(sample_name)), c(rep('ring|trophozoite|developing', length(sample_name)))), ~{
  ..1@meta.data %>%
    mutate(across(Stage_MSC, as.character)) %>%
    # mutate(Stage_MSC = Stage_MSC) %>%
    mutate(StagePrelim_lf = case_when(
      is.na(Stage_MSC) & sim_lab1 > 0.35 & str_detect(stage_lab1, ..3) & str_detect(stage_lab2, ..3) & str_detect(stage_lab3, ..3) ~ stage_lab1, ## lower cos sim of 0.35 for asexuals instead of 0.4
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Male|male') & str_detect(stage_lab2, 'Male|male') & str_detect(stage_lab3, 'Male|male') ~ stage_lab1,
      is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Female|female') & str_detect(stage_lab2, 'Female|female') & str_detect(stage_lab3, 'Female|female') ~ stage_lab1,
      TRUE ~ Stage_MSC)#,
      # across(StagePrelim, str_to_sentence)
    )%>%
    mutate(StagePrelim = case_when(
      str_detect(StagePrelim_lf, "developing") ~ "Early trophozoite", ## !!NB - label developing gams as early trophs the closest stage
      str_detect(StagePrelim_lf, "schizont") & str_detect(stage_lab2, "ring|trophozoite")  & str_detect(stage_lab3, "ring|trophozoite") ~ str_to_sentence(stage_lab2),
      str_detect(StagePrelim_lf, "late|early") ~ str_to_sentence(StagePrelim_lf),
      str_detect(StagePrelim_lf, "gametocyte") ~ str_to_sentence(str_remove_all(StagePrelim_lf, "gametocyte \\(|\\)")),
      is.na(StagePrelim_lf) ~ "Unassigned",
      TRUE ~ StagePrelim_lf)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(StagePrelimC = case_when(
      str_detect(StagePrelim, "ring|trophozoite|developing|schizont") ~ "Asexual",
      TRUE ~ StagePrelim)#,
      # across(StagePrelim, str_to_sentence)
    ) %>%
    mutate(across(c(StagePrelim, StagePrelimC), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    dplyr::select(StagePrelim, StagePrelim_lf, StagePrelimC) %>%
    # mutate(across(starts_with('Stage_V'), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
    AddMetaData(.x, .)
})

```


```{r, warning=FALSE, message=FALSE}

# map(all_msc_anotd_preQC, ~.x@meta.data %>% dplyr::count(Stage_MSC, StagePrelim, stage_lab1))
map(all_msc_anotd_preQC, ~.x@meta.data %>% dplyr::count(StagePrelim, StagePrelimC))
```

```{r, warning=FALSE, message=FALSE}
imap(all_msc_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = F, group.by = 'StagePrelim', cols = sp_fld_colors)+ labs(title = .y))
```

```{r, warning=FALSE, message=FALSE}
all_msc_anots[["MSC51"]]
```

```{r, warning=FALSE, message=FALSE}
## Check if the poorly annotated samples are from mixed infections
pf_solo_mixed_mdata %>% filter(mixed_infection2 == "mixed") 
```

```{r, warning=FALSE, message=FALSE}
imap(all_msc_anotd_preQC, ~FeaturePlot(.x, dims = c(1,2), features = "nFeature_RNA")+ labs(title = .y))
```

```{r, results='asis',  fig.width = 8, fig.height = 4}
##adding clusters that are used to display mitochondrial count

plts <- list(all_msc_anotd_preQC, names(all_msc_anotd_preQC)) %>%
  pmap(., ~{
    
    cell_qc_plot <-  ..1 %>% DimPlot(., pt.size = 1, reduction = 'umap', dims = c(1,2), group.by = 'StagePrelim', cols = sp_fld_colors) +labs(title = ..2)#,
    
    cell_qc_plot$data <- cell_qc_plot$data %>% rownames_to_column('bcode') %>% left_join(., ..1@meta.data[,'seurat_clusters', drop = F] %>% rownames_to_column('bcode'), by='bcode') %>% column_to_rownames('bcode')
    
    # print(str(cell_qc_plot$data))
    
    cell_qc_plot<- LabelClusters(cell_qc_plot, id = "seurat_clusters", size = 9, repel = T)
    
    return(cell_qc_plot)

  })
```


```{r, results='asis',  fig.width = 8, fig.height = 4}
plts

```

Umap with stage labels in legend

```{r, fig.width = 7, fig.height = 5, eval=F}
trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.2, "cm")
)

# Assign the female High_low stage of the projected cell to the nearest cell in the reference dataset
# msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(msc_combi.sce))), length(all_msc_jc))
msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(v3labmscs.ref.sce))), length(all_msc_jc))


for (i in 1:length(msc_qcd_anotd_mali_mapped14HL)){
  msc_qcd_anotd_mali_mapped14HL[[i]][all_msc_jc[[i]]$Stage_MSC$cells[1,]] <- all_msc_anotd_preQC[[i]]$StagePrelim
}


for (i in 1:length(all_msc_anotd_preQC)){
  msc_qcd_anotd_mali_mapped14HL[[i]][all_msc_jc[[i]]$Stage_MSC$cells[1,]] <- all_msc_anotd_preQC[[i]]$StagePrelim
  # msc_combi@meta.data[, paste0("stage_", names(all_msc_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  v3labmscs.ref.sce@colData[, paste0("stage_", names(all_msc_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  
}

labs_order <- c('Late ring', 'Early trophozoite','Female (HE)',  'Male (HE)', 'Female (LE)', 'Male (LE)','Gametocyte (developing)', 'Gametocyte (male)','Gametocyte (female)',  'Atlas')
names(labs_order)<- labs_order

umap_theme <- theme(legend.position = 'right',
                    axis.title=element_text(size=12,face="bold",hjust = 0),
                    axis.text=element_text(size=18), 
                    plot.title = element_blank(),
                    legend.title=element_text(size=14,face="bold"),
                    legend.text=element_text(size=14),
                    legend.spacing.y = unit(0.25, "cm"),
                    axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed"))
)  


# Idents(msc_combi) <- 'MSC14_V3'
# m14_scmap_mca_umap <- map(names(all_msc_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), 
#         order = labs_order) +
m14_scmap_mca_umap <- map(names(all_msc_anotd_preQC), ~{
  cbind(reducedDims(v3labmscs.ref.sce)$PCA, colData(v3labmscs.ref.sce)[, c("stageHL_ref", paste0("stage_",.x)), drop=F]) %>% 
    data.frame() %>% 
    mutate(od_stg = droplevels(factor(!!rlang::sym(paste0("stage_",.x)), levels = rev(stg_refnd_lvls)))) %>%
    arrange(od_stg) %>%
    ggplot(aes(y = PC2, x=PC3, color = od_stg)) +
    geom_point() +
    # scale_color_manual(breaks=labs_order, values = sp_fld_colors, labels = str_wrap(labs_order, 10))+
    scale_color_manual(values = sp_fld_colors)+
    umap_theme+
    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
           x= trunc_axis,
           y= trunc_axis)+
    scale_x_continuous(breaks = NULL) +
    scale_y_continuous(breaks = NULL)
})

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap, 'plots/fig4/m14_scmap_mca_umap')

m14_scmap_mca_umap


```

Umap with stage labels inside
```{r, fig.width = 7, fig.height = 5, eval=F}

m14_scmap_mca_umap_labs_in <- map(names(all_msc_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), label = T, order = labs_order, repel = T, label.size = 5) +
                                    scale_color_manual(values = sp_fld_colors)+ 
                                    umap_theme+
                                    theme(axis.title=element_text(size=18,face="bold"), 
                                          plot.title = element_blank(),
                                          legend.position='None'
                                    ) +
                                    guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
                                           x= trunc_axis,
                                           y= trunc_axis)+
                                    scale_x_continuous(breaks = NULL) +
                                    scale_y_continuous(breaks = NULL)
)

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap_labs_in, 'plots/fig4/m14_scmap_mca_umap_labs_in.RDS')


m14_scmap_mca_umap_labs_in


```


```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim, Stage_MSC, stage_lab1, stage_lab2))

```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_msc_anotd_preQC, "data/processed/Pf/all_msc_anotd_preQC.RDS")
```


```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data) %>% saveRDS(., "data/processed/Pf/all_msc_anotd_preQC_mdata.RDS")

```

```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim))

```



```{r, fig.width = 12, fig.height = 4}

##Plotting top two scmap stages among the top 10 with their similarities for MSC14
# 
# stage_fct_labs <- msc_qcd_anotd[[1]]@meta.data  %>% dplyr::group_by(stageHL, seurat_clusters) %>% dplyr::group_keys() %>% dplyr::mutate(across(c(stageHL, seurat_clusters), as.character), across(seurat_clusters, ~paste0("Cluster ",., ':\n',stageHL))) %>% deframe()
# 
# scm_barpt <- msc_qcd_anotd[[1]]@meta.data %>% 
#   dplyr::select(matches('s.*lab.'), stageHL) %>% 
#   rownames_to_column('cell_bc')%>%
#   pivot_longer(-c('cell_bc', 'stageHL'),
#                names_to = c( ".value", "Label"),
#                names_pattern = "(s.*)_(l.*)")%>% 
#   filter(!is.na(sim)) %>%
#   mutate(across(sim, as.numeric)) %>% 
#   mutate(across(Label, ~str_replace(., 'lab','Label '))) %>%
#   mutate(across(stage, ~droplevels(factor(., stg_refnd_lvls)))) %>%
#   ggplot(., aes(y = sim, x = stage, color = Label)) +
#   geom_boxplot(aes(fill = Label))+
#   scale_fill_manual(values =  c('Label 1' = '#007FFF', 'Label 2' = '#FFFF9F'), name = "Label",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 2)
#   )+
#   scale_color_manual(values =  c('Label 1' = 'black', 'Label 2' = 'black'), name = "Label")+ 
#   geom_point(aes(group=Label), size = 0.5, position = position_jitterdodge(jitter.width = 0.2))+
#   new_scale_fill()+
#   geom_tile(data = data.frame(x = c( 'Late ring', 'Early trophozoite', 'Gametocyte (developing)', 'Gametocyte (female)', 'Gametocyte (male)')), inherit.aes = F, aes(x = x, y = 0.37, fill = x, height = 0.03)) +
#   scale_fill_manual(values =  sp_fld_colors, name = "MCA stage",
#                     guide = guide_legend(direction = "horizontal", title.position = "top", ncol = 3)
#                     # guide = 'none'
#   )+
#   scale_y_continuous(expand = expansion(mult = c(0.01, 0.05))) +
#   facet_grid(.~stageHL, labeller = labeller(stageHL = stage_fct_labs)) +
#   geom_hline(yintercept = 0.4)+
#   labs(y = 'scmap \nsimilarity scores', x = 'MCA stage') +
#   theme_classic() +
#   theme(axis.text.x = element_blank(),
#         axis.title=element_text(size=16,face="plain"), 
#         axis.text = element_text(size=15), 
#         axis.ticks.x = element_blank(),
#         strip.text =element_text(size=16,face="plain", colour = 'black'),
#         panel.border=element_rect(colour="black",fill='transparent'),
#         legend.title = element_text(size=16,face="bold"),
#         legend.text = element_text(size=15),
#         legend.position = 'bottom')
# 
# scm_barpt
```

Put all the pf filt barcodes - including from mixed samples in one folder for souporcell

```{r}
## Get the donor and irods ID for pf only infections
pf_mix_solo <- pf_solo_mixed_mdata[,c("sample_nm", "mixed_infection2")] %>% deframe()
```


```{r}
## Write Pf barcodes for mixed species donor for souporcell
imap(pf_mix_solo, ~{
  system(paste0("mkdir -p data/raw/Pf/", .y, "/outs/pf_filt_bcodes"))
  if (.x == "solo"){
    system(paste0('cp data/raw/Pf/', .y, '/outs/filtered_feature_bc_matrix/barcodes.tsv.gz data/raw/Pf/', .y, '/outs/pf_filt_bcodes/'))
  }
  else if(.x == "mixed") {
    system(paste0('cp data/raw/Pf_mxd/', .y, '/outs/sk21_bcodes_pp/barcodes.tsv.gz data/raw/Pf/', .y, '/outs/pf_filt_bcodes/'))
  }
  })

```

Run souporcell like below for the filtered barcodes of all cellranger pf cells in all donor including mixed ones - combines the 2 below
```{r}
gogo()

# nextflow run ../mali_code_lnk/nf_scripts/soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/Pf/MSC*/outs/pf_filt_bcodes/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```


Run souporcell like below for the filtered barcodes of all infections
```{r}
gogo()

# nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/Pf/MSC*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```



Run souporcell like below for the filtered barcodes of pf in mixed infections
```{r}
gogo()

#  nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/pf_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/5736STDY134373{63,72,89}/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/Pf_mxd/MSC*/outs/sk21_bcodes_pp/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --soup_dir "soupc" --scrpt "/nfs/users/nfs_j/jr35/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_Pfalciparum3D7_Genome.fasta" --hsat_ref_file "Pf66" -resume
```
