---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  library(furrr)
  library(future.batchtools)
  library(parallel)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
})
```


```{r, setup}

# knitr::opts_knit$set(root.dir = '/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/Mali2')
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/')
```


# MSC preliminary stage annotation for strain assesment
### Variable declarations

```{r}

##Strain cluster colours
stg_lvls = c("Early ring", "Late ring", "Early trophozoite","Late trophozoite" , "Early schizont", "Late schizont",  "Gametocyte (developing)", "Female (HE)", "Female", "Female (LE)",  "Gametocyte (female)", "Male (HE)", "Male", "Male (LE)", "Gametocyte (male)", "Asexual", "Unassigned", "Failed QC", "Not assigned")

stg_refnd_lvls = c("Early ring", "Late ring", "Early trophozoite", "Late trophozoite", "Early schizont", "Late schizont", "Gametocyte (committed)", "Gametocyte (developing)", "Gametocyte (branching)", "Female (HE)", "Female", "Female (LE)", "Gametocyte (female)", "Gametocyte (early female)", "Gametocyte (late female)", "Gametocyte (male)", "Male (HE)", "Male", "Male (LE)", "Gametocyte (early male)", "Gametocyte (late male)", "Asexual", "early trophozoite", "late ring", "late trophozoite", "early schizont", "early ring", "late schizont", "gametocyte (developing)", "gametocyte (female)", "gametocyte (male)", "Unassigned", "Failed qc", "Not assigned", "Atlas")

stg_refnd_lvls2 = c('Early ring','Late ring','Early trophozoite','Late trophozoite','Early schizont','Late schizont', "Gametocyte (committed)","Gametocyte (developing)","Gametocyte (branching)", "Female (HE)", "Female", "Female (LE)", "Gametocyte (early female)","Gametocyte (late female)", "Male (HE)", "Male", "Male (LE)", "Gametocyte (early male)","Gametocyte (late male)", "Asexual", "Unassigned", "Failed QC", "Not assigned") #%>% str_to_sentence(.)

sp_colors <- c('Early ring'='#D1EC9F',
                     'Late ring'='#78C679',
                     'Early trophozoite'='#FEEEAA',
                     'Late trophozoite'='#FEB24C',
                     'Early schizont'='#C9E8F1',
                     'Late schizont'='#85B1D3',
                     'Gametocyte (developing)'='thistle',
                     'Gametocyte (female)'='#551A8B',
                     'Gametocyte (male)'='mediumpurple',
               'Unassigned'='red', 
               'Not assigned'='#D6CFC7',
               'Female (LE)'='#208cf7',
               'Female (HE)'='#551A8B',
               'Male (LE)'='#a894d1',
               'Male (HE)'='mediumpurple',
               'Failed QC' = 'grey',
               'Gametocyte (committed)' = '#C399A2', 
               'Gametocyte (developing)' = '#66444C', 
               'Gametocyte (branching)' = '#66444C', 
               'Gametocyte (early female)' = '#749E89', 
               'Gametocyte (early male)' = '#7D87B2', 
               'Gametocyte (late female)' = '#4E6D58', 
               'Gametocyte (late male)' = '#41507B'
)

sp_fld_colors <- c('Late ring'= '#78C679',#met.brewer('Java')[2],
                   'Early trophozoite'='#FEEEAA',#met.brewer('Java')[3],
                   'Female LE'= met.brewer('Ingres')[5],
                   'Female'= '#551A8B', #met.brewer('Ingres')[7],
                   # 'Female (HE)'= '#551A8B', #met.brewer('Ingres')[7],
                   # 'Male (LE)'=met.brewer('Nizami')[6],
                   # 'Male (HE)'=met.brewer('Nizami')[8],
                   # 'Gametocyte (male)'=met.brewer('Austria')[5],
                   'Male LE'=met.brewer('Ingres')[7],#'#a2e19e',
                   'Male'= 'mediumpurple',#'#039911',
                   ## 'Male (HE)'= 'mediumpurple',#'#039911',
                   'Gametocyte (male)'='#3DED97',
                   'Gametocyte (female)'=met.brewer('Ingres')[6],
                   'Lab' = '#D3D3D3',
                   'Unassigned'='#DCDCDC',
                   'Failed QC' = 'grey',
                   'Not assigned' = 'grey',
                   'Atlas' = 'grey',
                   'Early ring'='#D1EC9F',
                     'Late trophozoite'='#FEB24C',
                     'Early schizont'='#C9E8F1',
                     'Late schizont'='#85B1D3',
                   'Gametocyte (developing)'='thistle',
                   'early female' = '#749E89',
                   'late female'= '#4E6D58',
                    "Asexual" = "#FBC81D",
                   NULL = 'lightgrey',
                   'Field unlabelled'='#D2C1DC'
)

cluster_cols <- brewer.pal(7, name = 'Set2') %>% set_names(as.character(c(0:6)))



donor_cols_m = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" ), '.mrna'))
# donor_cols = met.brewer("Egypt")[c(2,4,1,3)] %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))
donor_cols = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" )))
# donor_cols_cap = c(met.brewer("Thomas")[c(1)],c('#F39EC8','#99ceff', '#b9e589')) %>% str_to_upper(set_names(paste0(c("MSC1", "MSC3", "MSC13", "MSC14" ))))


```


### Common functions

#### Seurat normalization function from Elias
```{r}

seurat_norm_stand = function(Seurat_object, nHVG = 500){
  Seurat_object = NormalizeData(Seurat_object, verbose = F)
  Seurat_object = FindVariableFeatures(Seurat_object, nfeatures = nHVG, verbose = F)
  Seurat_object = ScaleData(Seurat_object, verbose = F)
  Seurat_object = RunPCA(Seurat_object, verbose = F)
  return(Seurat_object)
}
```

initialize variable for the sample names

```{r}
irods_id_sk21_mdata_cln <- read_csv("data/raw/irods_id_sk21_mdata_cln.csv") %>%
  mutate(sample_nm = str_remove_all(sample_name, "_JC")) %>%
  arrange(sample_nm)

smpl_nms <-irods_id_sk21_mdata_cln$sample_nm
#c("MSC24", "MSC33", "MSC37", "MSC38", "MSC39", "MSC40", "MSC45", "MSC48", "MSC49", "MSC50_MACS", "MSC50_SLO", "MSC53", "MSC54", "MSC55", "MSC57", "MSC60", "MSC66", "MSC67", "MSC68", "MSC70")
```

```{r}
## Save for farm
write_csv(irods_id_sk21_mdata_cln, "data/raw/Pf/irods_id_sk21_mdata_cln.csv") 
```

NB-LOCAL_CODE
```{r, eval=FALSE}
system(
    paste0('rsync -av /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/raw/irods_id_sk21_mdata_cln.csv jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/irods_id_sk21_mdata_cln.csv')
    )
```
  
## Genotype clustering for MSC

Read in clusters.tsv files with assignment of cells to strains and PCA plots metrics

Transfer clusters.tsv files with assignment of cells to strains and PCA plots metrics for the best K from farm to my local Mac

```{r}
##Copy souporcell files for the optimum clusters

##Sample ID
sample_name <- irods_id_sk21_mdata_cln$sample_nm

##Sample IDs
source_irods <- irods_id_sk21_mdata_cln[,c('sample_nm', 'irods_id')] %>% deframe

```

NB-LOCAL_CODE

```{r, eval=F}
##Copy souporcell files for the optimum clusters

for (i in 1:length(sample_name)) {
# for (i in 32:33) {
  
  ## Create new directories
  system(paste0('mkdir -p /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/raw/Pf/', sample_name[i], '/outs'))
  system(paste0('mkdir -p /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/processed/Pf/', sample_name[i], '/bpcells'))
  
  ##Transfers
  ##souporcell genotyping files
  system(
    paste0('rsync --delete -av --exclude="molecule_info.h5" --exclude="raw_feature_bc_matrix.h5" --exclude="*bam*" --exclude="metrics_summary.csv" jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/', source_irods[i], '/outs/* /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/raw/Pf/', sample_name[i], '/outs/')
    )
  
  ##metadata and seurat object files
  # system(
  #   paste0('rsync -av --include="msc_filt_w_dblts_scmap_metadata.RDS" --include="msc_filt_w_dblts_scmap.RDS" --exclude="*" jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/Pf/', source_irods[i], '/* /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/raw/', sample_name[i], '/')
  # )
  
}


```

NB-FARM_CODE
```{r, eval=T}
##Copy souporcell files for the optimum clusters

for (i in 1:length(sample_name)) {
# for (i in 32:33) {
  
  ## Create new directories
  system(paste0('mkdir -p data/raw/Pf/', sample_name[i], '/outs'))
  system(paste0('mkdir -p data/processed/Pf/', sample_name[i], '/bpcells'))
  
  ##Transfers
  ##souporcell genotyping files
  system(
    paste0('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pf/', source_irods[i], '/outs/filtered_feature_bc_matrix.h5 data/raw/Pf/', sample_name[i], '/outs/')
    )
  
}


```

```{r, results='hide', eval=FALSE}
# list.files("data/processed/", pattern = "clusters.tsv", full.names = T, recursive = T, include.dirs = T)
list.files("data/raw/Pf/", pattern = "filtered_feature_bc_matrix$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/raw//|/Pf.*') %>% str_replace(., '/', '_')))
```


```{r}
##Read filtered cellranger matrices into R
list.files("data/raw/Pf/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T)
```

```{r}
##Read filtered cellranger matrices into R
# all_msc <- list.files("data/raw/", pattern = "filtered_feature_bc_matrix_h5", full.names = T, recursive = T, include.dirs = T)%>% 
#   set_names(nm = (str_remove_all(.,'data/raw//|/Pf.*') %>% str_replace(., '/', '_'))) %>% 
#   # str_sort(., numeric = T) %>%
#   .[sample_name] %>%
#   map(~Read10X(.,gene.column = 1)) %>% 
#   imap(., ~ CreateSeuratObject(.x, project = paste0(.y, ".mrna"))) 
```


```{r, message=F, warning=F}
##Write BPcells matrices
list.files("data/raw/Pf/", pattern = "filtered_feature_bc_matrix.h5$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/raw/Pf//|/outs.*'))) %>% 
  .[sample_name] %>%
  map(~open_matrix_10x_hdf5(path = .)) %>% 
  imap(., ~write_matrix_dir(mat = .x, dir = paste0('data/processed/Pf/', .y, '/bpcells/counts_mtx'), overwrite = T))
```


```{r, message=F, warning=F}
##Read BPcells matrices
all_msc <- list.files("data/processed/", pattern = "counts_mtx$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed//Pf/|/bpcells.*'))) %>% 
  .[sample_name] %>%
  map(~open_matrix_dir(dir = .)) %>% 
  imap(., ~CreateSeuratObject(counts = .x, project = paste0(.y, ".mrna"))) 
```


```{r, message=F, warning=F}
all_msc2 <- all_msc[1:2]
all_msc2[[1]][["RNA"]]$counts <- as(object = all_msc2[[1]][["RNA"]]$counts, Class = "dgCMatrix")
all_msc2[[2]][["RNA"]]$counts <- as(object = all_msc2[[2]][["RNA"]]$counts, Class = "dgCMatrix")
```


```{r, message=F, warning=F}
# all_msc2 <- all_msc[1:2]
all_msc2m <- all_msc[1:2] %>%
  map(., ~{
    .x[["RNA"]]$counts <- as(object = .x[["RNA"]]$counts, Class = "dgCMatrix")
    return(.x)
    })

```


Check clustering of cells before filtering out by mt.percent
```{r, warning=FALSE, message=FALSE}
##Seurat standard transformation
# plan(cluster, workers = availableWorkers())
# all_msc_tr <- mclapply(all_msc2, ~seurat_norm_stand(.x))
start.time <- Sys.time()
all_msc_tr <- mclapply(all_msc2m, function(x) {
  seurat_norm_stand(x)
})
end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
# plan(sequential)
```


Check clustering of cells before filtering out by mt.percent
```{r, warning=FALSE, message=FALSE}
##Seurat standard transformation
# plan(cluster, workers = availableWorkers())
# all_msc_tr <- mclapply(all_msc2, ~seurat_norm_stand(.x))
start.time <- Sys.time()
all_msc_tr <- mclapply(all_msc[1:2], function(x) {
  seurat_norm_stand(x)
})
end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
# plan(sequential)
```

```{r, message=F, warning=F}
# file.dir <- "/brahms/hartmana/vignette_data/h5ad_files/"
# files.set <- c("ahern_pbmc.h5ad", "jin_pbmc.h5ad", "yoshida_pbmc.h5ad")
# 
# # Loop through h5ad files and output BPCells matrices on-disk
# data.list <- c()
# metadata.list <- c()
# 
# for (i in 1:length(files.set)) {
#   path <- paste0(file.dir, files.set[i])
#   data <- open_matrix_anndata_hdf5(path)
#    write_matrix_dir(
#      mat = data,
#      dir = paste0(gsub(".h5ad", "", path), "_BP")
#    )
#   # Load in BP matrices
#   mat <- open_matrix_dir(dir = paste0(gsub(".h5ad", "", path), "_BP"))
#   mat <- Azimuth:::ConvertEnsembleToSymbol(mat = mat, species = "human")
#   # Get metadata
#   metadata.list[[i]] <- LoadH5ADobs(path = path)
#   data.list[[i]] <- mat
# }
# # Name layers
# names(data.list) <- c("ahern", "jin", "yoshida")
# 
# # Add Metadata
# for (i in 1:length(metadata.list)) {
#   metadata.list[[i]]$publication <- names(data.list)[i]
# }
# metadata.list <- lapply(metadata.list, function(x) {
#   x <- x[, c("publication", "sex", "cell_type", "donor_id", "disease")]
#   return(x)
# })
# metadata <- Reduce(rbind, metadata.list)
```

```{r, warning=FALSE, message=FALSE}
# # map(all_msc, ElbowPlot)
# plan(batchtools_lsf(template = ".batchtools.lsf.tmpl", resources = list(queue = "normal", memory = "70000")))
# 
# # This does run in parallel!
# sta=Sys.time()
# future_map(c("hello", "world","hello", "world","hello"), ~.x)
# sto=Sys.time()
# sto-sta
# plan(sequential)
```

```{r, warning=FALSE, message=FALSE}
# # map(all_msc, ElbowPlot)
# plan(cluster, workers = availableWorkers())
# 
# # This does run in parallel!
# sta=Sys.time()
# future_map(c("hello", "world","hello", "world","hello"), ~.x)
# sto=Sys.time()
# sto-sta
```


```{r, warning=FALSE, message=FALSE}
# # map(all_msc, ElbowPlot)
# plan(sequential)
# 
# # This does run in parallel!
# sta=Sys.time()
# future_map(c("hello", "world","hello", "world","hello"), ~.x)
# sto=Sys.time()
# sto-sta
```

NB-FARM_CODE
```{r, eval=T}
plan(sequential)
start.time <- Sys.time()
nothingness <- future_map(c(2, 2, 2, 2, 2), ~Sys.sleep(.x))
end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
```

```{r, eval=T}
plan(multisession, worker)
start.time <- Sys.time()
nothingness <- future_map(c(2, 2, 2, 2, 2), ~Sys.sleep(.x))
end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
```

```{r, eval=T}
# plan(batchtools_lsf,resources = list(nodes = "1:ppn=12", vmem = "5gb"))
# plan(batchtools_lsf, resources = list(u = 'jr35', q = 'normal', n= 6, M= 50000, R= "span[hosts=1] select[mem>50000] rusage[mem=50000]", output = "data/"))
plan(batchtools_lsf, 
     resources = list(queue="normal",
                      walltime=60,#minutes
                      memory=8192,
                      ncpus=2)
)

x %<-% { Sys.sleep(5); 3.14 }
y %<-% { Sys.sleep(5); 2.71 }

x + y

```


```{r, eval=T}
batchtools::makeClusterFunctionsLSF(
  template = "~/.batchtools.lsf.tmpl",
  scheduler.latency = 1,
  fs.latency = 65
)

reg = batchtools::makeRegistry(file.dir = NA, seed = 1)
reg$cluster.functions = batchtools::makeClusterFunctionsLSF(
  template = "~/.batchtools.lsf.tmpl",
  scheduler.latency = 1,
  fs.latency = 65
)
cluster.functions = batchtools::makeClusterFunctionsInteractive()
```



```{r, eval=T}
piApprox = function(n) {
  nums = matrix(runif(2 * n), ncol = 2)
  d = sqrt(nums[, 1]^2 + nums[, 2]^2)
  4 * mean(d <= 1)
}
set.seed(42)
piApprox(1000)
```


```{r, eval=T}
batchMap(fun = piApprox, n = rep(1e5, 10))
```


```{r, eval=T}
names(getJobTable())
```


```{r, eval=T}
submitJobs(resources = list(walltime = 3600, memory = 1024))
```


```{r, eval=T}
getStatus()
```


NB-FARM_CODE
```{r, eval=T}
# plan(batchtools_lsf,resources = list(nodes = "1:ppn=12", vmem = "5gb"))
# plan(batchtools_lsf, resources = list(u = 'jr35', q = 'normal', n= 6, M= 50000, R= "span[hosts=1] select[mem>50000] rusage[mem=50000]", output = "data/"))
plan(batchtools_lsf, 
     resources = list(queue="normal",
                      walltime=60,#minutes
                      memory=8192,
                      ncpus=2)
)

start.time <- Sys.time()
all_msc_tr_md <- all_msc[1:5] %>%
  map(., ~.x@meta.data)

end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken
```

#### full msc_atlas reference dataset
```{r}
plan(multisession, workers = 2)
start.time <- Sys.time()
all_msc_tr_md <- all_msc[1:2] %>%
  future_map(., ~seurat_norm_stand(.x))

end.time <- Sys.time()
time.taken <- end.time - start.time
time.taken


```

Check clustering of cells before filtering out by mt.percent
```{r, warning=FALSE, message=FALSE}
##Seurat standard transformation
# plan(cluster, workers = availableWorkers())
all_msc_tr_md <- all_msc[1:5] %>%
  map(., ~.x@meta.data)

# plan(sequential)
```

Check clustering of cells before filtering out by mt.percent
```{r, warning=FALSE, message=FALSE}
##Seurat standard transformation
# plan(cluster, workers = availableWorkers())
all_msc <- all_msc %>%
  map(., ~seurat_norm_stand(.x))

# plan(sequential)
```

```{r, warning=FALSE, message=FALSE}
map(all_msc, ElbowPlot)

```

```{r, warning=FALSE, message=FALSE}
# sdev<-prcomp(t(data),scale. = T)$sdev[1:30]
# sdev <- sort(sdev, decreasing = TRUE)
# npc <- findPC(sdev = sdev)
```

```{r, warning=FALSE, message=FALSE}
##Seurat UMAP
pc_list <-list(6L,6L,6L, 7L, 7L, 6L)
pc_list <-list(rep(6L,length(sample_name)))


all_msc <- all_msc %>%
  map2(., pc_list, ~RunUMAP(.x, dims = 1:.y, n.components = 3L, verbose=F, seed.use = 949) %>%
         FindNeighbors(reduction = "pca", dims = 1:.y, verbose = FALSE) 
       )

```


```{r, warning=FALSE, message=FALSE}
##Seurat UMAP
res_list <-list(rep(0.04, length(sample_name)))


all_msc <- all_msc %>%
  map2(., res_list, ~FindClusters(.x, resolution = .y, verbose = FALSE, k.param = 30))

# old_seurat_clusters <-  all_msc %>% map(., ~.x@meta.data %>% dplyr::select(seurat_clusters) %>% rename('old_seu_cluster' = seurat_clusters))

map(all_msc, ~DimPlot(.x, dims = c(1,2), label = T))
```

```{r, warning=FALSE, message=FALSE}

# saveRDS(all_msc, "data/processed/all_msc.RDS")
```

## scmap annotation

### Preprocessing references

#### msc_combi reference dataset
```{r, eval=FALSE}
##Read MCA V3 blood stage reference dataset
# msc_combi <- readRDS("../sc_velo/data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.4.RDS")
# msc_combi <- readRDS("../Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")
msc_combi <- readRDS("../Mali/data/processed/DB52e_star/msc_combi.RDS")
DefaultAssay(msc_combi) <- "RNA"

##Collapse asexual stages into 1 "Asexual"
msc_combi@meta.data$stageHL_ref_asx <- ifelse(msc_combi@meta.data$stageHL_ref %in% c("Early trophozoite", "Late ring"), "Asexual", msc_combi@meta.data$stageHL_ref)
```

```{r, message=FALSE, warning=FALSE, eval=FALSE}
##Convert to reference to sce
msc_combi.sce <- as.SingleCellExperiment(msc_combi)

##Place counts appropriately into single cell experiment object
SingleCellExperiment::logcounts(msc_combi.sce) <- log2(SingleCellExperiment::counts(msc_combi.sce) + 1)
rowData(msc_combi.sce)$feature_symbol <- rownames(msc_combi.sce)
msc_combi.sce <- msc_combi.sce[!duplicated(rownames(msc_combi.sce)), ]

##scmap feature selection
msc_combi.sce <- selectFeatures(msc_combi.sce, suppress_plot = T)

##Cell indexing
set.seed(155)
msc_combi.sce <- indexCell(msc_combi.sce)
```

NB-LOCAL_CODE
```{r, eval=F}
## transfer integrated object from sunil farm to local
system('rsync -av jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/for_scmap/v3labmscs.ref.seur.RDS /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/raw/Pf/')
system('rsync -av jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/for_scmap/v3labmscs.ref.sce.RDS /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/raw/Pf/')

```

NB-FARM_CODE
```{r, eval=T}
## transfer integrated object from sunil farm to local
# system('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/for_scmap/v3labmscs.ref.seur.RDS data/raw/Pf/')
system('cp /lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/rdata/for_scmap/v3labmscs.ref.sce.RDS data/raw/Pf/')

```

#### full msc_atlas reference dataset
```{r}
##Read MCA V3 + field blood stage reference dataset
# v3labmscs.ref.sce <- readRDS("../Mali2/data/raw/Pf/v3labmscs.ref.sce.RDS")
v3labmscs.ref.sce <- readRDS("data/raw/Pf/v3labmscs.ref.sce.RDS")
```



```{r, message=FALSE, warning=FALSE}
##Collapse asexual stages into 1 "Asexual"
v3labmscs.ref.sce@colData[,c("stageHL_ref", "stageHL_ref_asx")] <- colData(v3labmscs.ref.sce) %>% 
  data.frame() %>% 
  mutate(stageHL_ref = case_when(str_detect(orig.ident, "msc") & stageLR == "late ring" ~ "Late ring", 
                                 str_detect(orig.ident, "msc") & stageLR == "early trophozoite" ~ "Early trophozoite", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (female) LE" ~ "Female LE", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (female)" ~ "Female", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (male) LE" ~ "Male LE", 
                                 str_detect(orig.ident, "msc") & stageLR == "gametocyte (male)" ~ "Male", T ~ stageLR)) %>% 
  mutate(stageHL_ref_asx = case_when(stageHL_ref %in% c("Early trophozoite", "Late ring") ~ "Asexual", T ~ stageHL_ref)) %>%
  select(stageHL_ref, stageHL_ref_asx)

```


```{r, message=FALSE, warning=FALSE}
scater::plotReducedDim(v3labmscs.ref.sce, dimred = "PCA", colour_by = "stageHL_ref_asx", ncomponents = c(2,3))
```

```{r, message=FALSE, warning=FALSE}
##scmap feature selection
v3labmscs.ref.sce <- selectFeatures(v3labmscs.ref.sce, suppress_plot = T)

##Cell indexing
set.seed(155)
v3labmscs.ref.sce <- indexCell(v3labmscs.ref.sce)
```


### Preprocessing MSC query datasets

```{r}

#Convert to SCE and subset
# all_msc_sce <- all_msc %>%
#   map(. %>% as.SingleCellExperiment(.))
```

Convert seurat object into single cell experiment filling in the slots appropriately
```{r}

##This command gives error below when running the entire notebook or this whole chunk so run the command independently - coulb be and issue with rmd
##Error in x$.self$finalize() : attempt to apply non-function - need blank line after
# 
# sce_obj_creatn <- function(sce_obj, seu_obj){
#   for(i in 1:length(seu_obj)) {
#     sce_obj[[i]] <- SingleCellExperiment(assays = list(counts = as(object = seu_obj[[i]][["RNA"]]$counts, Class = "dgCMatrix")), colData = seu_obj[[i]]@meta.data)
#     
#     logcounts(sce_obj[[i]]) <- log2((SingleCellExperiment::counts(sce_obj[[i]]) + 1))
#     # normcounts(sce_obj[[i]]) <- as.matrix(seu_obj[[i]]@assays$RNA@data)
#     rowData(sce_obj[[i]])$feature_symbol <- rownames(sce_obj[[i]])
#     sce_obj[[i]] <- sce_obj[[i]][!duplicated(rownames(sce_obj[[i]])), ]
#     # rowData(sce_obj[[i]])$feature_symbol <- str_replace(rownames(sce_obj[[i]]), "-", "_")
#     # reducedDims(sce_obj[[i]]) <- SimpleList(PCA = seu_obj[[i]]@reductions[["pca"]]@cell.embeddings, UMAP = seu_obj[[i]]@reductions[["umap"]]@cell.embeddings)
# 
#   }
#   
#   return(sce_obj)
#   
# }


```

```{r}

##This command gives error below when running the entire notebook or this whole chunk so run the command independently - coulb be and issue with rmd
##Error in x$.self$finalize() : attempt to apply non-function - need blank line after
# all_msc_sce <- list()
sce_obj_creatn <- function(sce_obj, seu_obj){
  for(i in 1:length(seu_obj)) {
    obj1 <- SingleCellExperiment(assays = list(counts = as(object = seu_obj[[i]][["RNA"]]$counts, Class = "dgCMatrix")), colData = seu_obj[[i]]@meta.data)
    
    logcounts(obj1) <- log2((SingleCellExperiment::counts(obj1) + 1))
    # normcounts(obj1) <- as.matrix(seu_obj[[i]]@assays$RNA@data)
    rowData(obj1)$feature_symbol <- rownames(obj1)
    obj1 <- obj1[!duplicated(rownames(obj1)), ]
    # rowData(obj1)$feature_symbol <- str_replace(rownames(obj1), "-", "_")
    # reducedDims(obj1) <- SimpleList(PCA = seu_obj[[i]]@reductions[["pca"]]@cell.embeddings, UMAP = seu_obj[[i]]@reductions[["umap"]]@cell.embeddings)
    
    sce_obj[[i]] <- obj1
  }
  
  return(sce_obj)
  
}


```

```{r}

all_msc_sce <- sce_obj_creatn(sce_obj = list(), seu_obj = all_msc)

```

### Stage label assignment (projection)
transferring labels from  reference datasets
```{r}
##Projection
all_msc_jc <- all_msc_sce %>%
  map(. %>% scmapCell(.,list(
    # Stage_MSC = metadata(msc_combi.sce)$scmap_cell_index
    Stage_MSC = metadata(v3labmscs.ref.sce)$scmap_cell_index
  )
  )
  )
```

```{r}
##Projection
rm(all_msc_sce)
```

Assigning stage labels only when the top 3 most closest reference cells to a query cell all have a similar stage label and a cosine similarity greater than 0.4
```{r}
##Assigning stage labels using thresholds
all_msc_lbls <- all_msc_jc %>%
  map(. %>% scmapCell2Cluster(.,list(
    # as.character(SummarizedExperiment::colData(msc_combi.sce)$stageHL_ref)
    as.character(SummarizedExperiment::colData(v3labmscs.ref.sce)$stageHL_ref)
  ),w = 3, threshold = 0.5)
  )
```

Obtaining the reference cell that is closest to each query cell in neighbourhood space regardless of cosine similarity. For those where the top 10 most closest cells constitute several stages, we also obtain the second most similar stage.


```{r, eval=T}
##Get the top V3 cell and top 3 similarity scores that is most similar to each MSC14 cell 
sc_topv3_cells_sims_all <- list(all_msc_jc) %>%
    pmap(~{
      ..1$Stage_MSC$similarities %>% 
        .[1:3,] %>% 
        t() %>% 
        data.frame() %>% 
        rename_all(., ~str_replace(., 'X', 'sim_lab')) %>% 
        rownames_to_column('bcode') %>%
        left_join(., ..1$Stage_MSC$cells %>% 
                    .[1:3,] %>% t() %>% 
                    data.frame() %>% 
                    rename_all(., ~str_replace(., 'X', 'msc_ref_cells')) %>% 
                    rownames_to_column('bcode'),
                  by = 'bcode') %>% 
        # mutate(across(starts_with('msc_ref_cells'), ~colnames(msc_combi.sce)[.]))
        mutate(across(starts_with('msc_ref_cells'), ~colnames(v3labmscs.ref.sce)[.]))
    })

# saveRDS(sc_topv3_cells_sims_all, 'data/processed/DB52e_star/sc_topv3_cells_sims_all.RDS')
```

```{r}
##Add the empty element into the position with list lacking
# scmap_cells_sim_nw <- readRDS('data/processed/MSC14/5736STDY11771546_DB52e_star/scmap_cells_sim_nw.RDS')

##Get similarities for cells that have similar stage assignment by scmap and row bind
# scmap_cs <- readRDS('data/processed/MSC14/5736STDY11771546_DB52e_star/scmap_cs.RDS')

##Get similarities for cells that have similar stage assignment by scmap and row bind
# sc_topv3_cells_sims_all <- readRDS('data/processed/DB52e_star/sc_topv3_cells_sims_all.RDS')

```


Adding annotations to MSC query datasets
```{r}
## Adding annotations to query cells
# stg_anots <- msc_combi@meta.data[,c('stageHL_ref'), drop=F]
stg_anots <- v3labmscs.ref.sce@colData[,c('stageHL_ref'), drop=F] %>% data.frame()

## scmap merge the top labels and similarity scores
top_stg_sim <- map(sc_topv3_cells_sims_all,
~left_join(.x, stg_anots %>% rownames_to_column('msc_ref_cells1') %>% rename('stage_lab1' = stageHL_ref), by = 'msc_ref_cells1') %>%
                       left_join(stg_anots %>% rownames_to_column('msc_ref_cells2') %>% rename('stage_lab2' = stageHL_ref), by = 'msc_ref_cells2') %>%
                       left_join(stg_anots %>% rownames_to_column('msc_ref_cells3') %>% rename('stage_lab3' = stageHL_ref), by = 'msc_ref_cells3') %>% column_to_rownames('bcode')
)


# all_msc_anotd_preQC <- list(all_msc_lbls, all_msc_jc, all_msc, scmap_cs, sc_topv3_cell_id) %>%
all_msc_anotd_preQC <- list(all_msc_lbls, all_msc_jc, all_msc, top_stg_sim) %>%
  pmap(~ ..1$scmap_cluster_labs %>% 
         data.frame()  %>% 
         mutate(cell_bc = colnames(..2$Stage_MSC$cells))%>%
         column_to_rownames("cell_bc") %>%
         # dplyr::rename("Stage_MSC" = Stage_MSC)%>% 
         mutate(across(Stage_MSC, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
         AddMetaData(..3, .) %>%
         AddMetaData(., ..4) #%>%
         # AddMetaData(., ..5)
       )

```


```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'Stage_MSC', cols = sp_fld_colors))
```


```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = T, group.by = 'stage_lab1', cols = sp_fld_colors))
```

```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data %>% dplyr::count(Stage_MSC, stage_lab1))

```


```{r}
##Modify the stage labels for MSC14
# all_msc_anotd_preQC[[4]] <- all_msc_anotd_preQC[[4]]@meta.data %>%
all_msc_anotd_preQC <- pmap(list(all_msc_anotd_preQC, rep(0.4,length(sample_name)), c(rep('Late ring|Early trophozoite|late ring|early trophozoite', length(sample_name)))), ~{
  ..1@meta.data %>%
  mutate(across(Stage_MSC, as.character)) %>%
  # mutate(Stage_MSC = Stage_MSC) %>%
  mutate(StagePrelim_lf = case_when(
    is.na(Stage_MSC) & sim_lab1 > 0.35 & str_detect(stage_lab1, ..3) & str_detect(stage_lab2, ..3) & str_detect(stage_lab3, ..3) ~ stage_lab1,
    is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Male|male') & str_detect(stage_lab2, 'Male|male') & str_detect(stage_lab3, 'Male|male') ~ stage_lab1,
    is.na(Stage_MSC) & sim_lab1 > ..2 & str_detect(stage_lab1, 'Female|female') & str_detect(stage_lab2, 'Female|female') & str_detect(stage_lab3, 'Female|female') ~ stage_lab1,
    TRUE ~ Stage_MSC)#,
    # across(StagePrelim, str_to_sentence)
  )%>%
  mutate(StagePrelim = case_when(
    str_detect(StagePrelim_lf, "gametocyte (developing)|schizont") ~ "Field unlabelled",
    str_detect(StagePrelim_lf, "late|early") ~ str_to_sentence(StagePrelim_lf),
    str_detect(StagePrelim_lf, "gametocyte") ~ str_to_sentence(str_remove_all(StagePrelim_lf, "gametocyte \\(|\\)")),
    is.na(StagePrelim_lf) ~ "Field unlabelled",
    TRUE ~ StagePrelim_lf)#,
    # across(StagePrelim, str_to_sentence)
  ) %>%
  dplyr::select(StagePrelim, StagePrelim_lf) %>%
  # mutate(across(starts_with('Stage_V'), ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
  AddMetaData(.x, .)
})

```


```{r, warning=FALSE, message=FALSE}

# map(all_msc_anotd_preQC, ~.x@meta.data %>% dplyr::count(Stage_MSC, StagePrelim, stage_lab1))
map(all_msc_anotd_preQC, ~.x@meta.data %>% dplyr::count(StagePrelim))
```

```{r, warning=FALSE, message=FALSE}
map(all_msc_anotd_preQC, ~DimPlot(.x, dims = c(1,2), label = F, group.by = 'StagePrelim', cols = sp_fld_colors))
```

```{r, warning=FALSE, message=FALSE}
map(all_msc_anotd_preQC, ~FeaturePlot(.x, dims = c(1,2), features = "nFeature_RNA"))
```

```{r, results='asis',  fig.width = 8, fig.height = 4}
##adding clusters that are used to display mitochondrial count
# all_msc_nosct_corrected_dblt_db[[4]] <- AddMetaData(all_msc_nosct_corrected_dblt_db[[4]], old_seurat_clusters)

list(all_msc_anotd_preQC) %>%
  pmap(., ~{
  
  cell_qc_plot <-  ..1 %>% DimPlot(., pt.size = 1, reduction = 'umap', dims = c(1,2), group.by = 'StagePrelim', cols = sp_fld_colors)#,

cell_qc_plot$data <- cell_qc_plot$data %>% rownames_to_column('bcode') %>% left_join(., ..1@meta.data[,'seurat_clusters', drop = F] %>% rownames_to_column('bcode'), by='bcode') %>% column_to_rownames('bcode')

cell_qc_plot<- LabelClusters(cell_qc_plot, id = "seurat_clusters", size = 9, repel = T)

# return(list(cell_qc_filt_obj, cell_qc_plot))
print(cell_qc_plot)

})
```


Umap with stage labels in legend

```{r, fig.width = 7, fig.height = 5, eval=T}
trunc_axis <- ggh4x::guide_axis_truncated(
  trunc_lower = unit(0, "npc"),
  trunc_upper = unit(2.2, "cm")
)

# Assign the female High_low stage of the projected cell to the nearest cell in the reference dataset
# msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(msc_combi.sce))), length(all_msc_jc))
msc_qcd_anotd_mali_mapped14HL = rep(list(rep("Not assigned", ncol(v3labmscs.ref.sce))), length(all_msc_jc))


for (i in 1:length(msc_qcd_anotd_mali_mapped14HL)){
  msc_qcd_anotd_mali_mapped14HL[[i]][all_msc_jc[[i]]$Stage_MSC$cells[1,]] <- all_msc_anotd_preQC[[i]]$StagePrelim
}


for (i in 1:length(all_msc_anotd_preQC)){
  msc_qcd_anotd_mali_mapped14HL[[i]][all_msc_jc[[i]]$Stage_MSC$cells[1,]] <- all_msc_anotd_preQC[[i]]$StagePrelim
  # msc_combi@meta.data[, paste0("stage_", names(all_msc_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))
  v3labmscs.ref.sce@colData[, paste0("stage_", names(all_msc_anotd_preQC[i]))] <-  droplevels(factor(msc_qcd_anotd_mali_mapped14HL[[i]], levels = stg_refnd_lvls))

}

labs_order <- c('Late ring', 'Early trophozoite','Female (HE)',  'Male (HE)', 'Female (LE)', 'Male (LE)','Gametocyte (developing)', 'Gametocyte (male)','Gametocyte (female)',  'Atlas')
names(labs_order)<- labs_order

umap_theme <- theme(legend.position = 'right',
        axis.title=element_text(size=12,face="bold",hjust = 0),
        axis.text=element_text(size=18), 
        plot.title = element_blank(),
        legend.title=element_text(size=14,face="bold"),
        legend.text=element_text(size=14),
        legend.spacing.y = unit(0.25, "cm"),
        axis.line = element_line(arrow = arrow(angle = 20, length = unit(0.15, "inches"), ends = "last", type = "closed"))
        )  


# Idents(msc_combi) <- 'MSC14_V3'
# m14_scmap_mca_umap <- map(names(all_msc_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), 
#         order = labs_order) +
m14_scmap_mca_umap <- map(names(all_msc_anotd_preQC), ~{
  cbind(reducedDims(v3labmscs.ref.sce)$UMAP, colData(v3labmscs.ref.sce)[, c("stageHL_ref", paste0("stage_",.x)), drop=F]) %>% 
    data.frame() %>% 
    mutate(od_stg = factor(!!rlang::sym(paste0("stage_",.x)), levels = rev(stg_refnd_lvls))) %>%
    arrange(od_stg) %>%
    ggplot(aes(y = UMAP_2, x=UMAP_3, color = od_stg)) +
  geom_point() +
  # scale_color_manual(breaks=labs_order, values = sp_fld_colors, labels = str_wrap(labs_order, 10))+
  scale_color_manual(values = sp_fld_colors)+
    umap_theme+
  guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
         x= trunc_axis,
         y= trunc_axis)+
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(breaks = NULL)
})

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap, 'plots/fig4/m14_scmap_mca_umap')

m14_scmap_mca_umap


```

Umap with stage labels inside
```{r, fig.width = 7, fig.height = 5, eval=F}

m14_scmap_mca_umap_labs_in <- map(names(all_msc_anotd_preQC), ~DimPlot(msc_combi, reduction = "umap",  group.by = paste0("stage_",.x), pt.size = 1, dims = c(2,3), label = T, order = labs_order, repel = T, label.size = 5) +
  scale_color_manual(values = sp_fld_colors)+ 
  umap_theme+
  theme(axis.title=element_text(size=18,face="bold"), 
        plot.title = element_blank(),
        legend.position='None'
        ) +
  guides(color = guide_legend(title = 'Stage',override.aes = list(size = 4), ncol = 1, byrow = TRUE),
         x= trunc_axis,
         y= trunc_axis)+
  scale_x_continuous(breaks = NULL) +
  scale_y_continuous(breaks = NULL)
)

##Save image for manuscript fig 4
# saveRDS(m14_scmap_mca_umap_labs_in, 'plots/fig4/m14_scmap_mca_umap_labs_in.RDS')


m14_scmap_mca_umap_labs_in


```


```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim, Stage_MSC, stage_lab1, stage_lab2))

```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_msc_anotd_preQC, "data/processed/Pf/all_msc_anotd_preQC.RDS")
```


```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data) %>% saveRDS(., "data/processed/Pf/all_msc_anotd_preQC_mdata.RDS")

```

```{r, warning=FALSE, message=FALSE}

map(all_msc_anotd_preQC, ~..1@meta.data %>% dplyr::count(StagePrelim))

```