---
title: "strain deconvolution for pre and post QC each donor singly"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---


```{r, message=FALSE}
suppressPackageStartupMessages({
  # library(Seurat)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  # library(cowplot)
  library(pheatmap)
  # library(rmarkdown)
  # library('vcfR')
  library(conflicted)
  library(RColorBrewer)
  # library(MetBrewer)
  library(ComplexHeatmap)
  library(SeqArray)
  library(ggrepel)
  library(ggvenn)
  # library(magick)
  # library(VariantAnnotation)
  library(gridExtra)
  library(Matrix)
  # library(UpSetR)
  library(SNPRelate)
  # library(openxlsx)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
})
```


# MSC parasites strain deconvolution and VCF visualization 

## Setup and sourcing 

```{r setup}
knitr::opts_knit$set(root.dir = paste0('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/'))
```

#### Source variables specific to this run - this pseudobulk strain analysis script is ran several times with different inputs
```{r}
## Get the necessary variable for the pseudobulking workflow
# source("mali22_code_lnk/local_rmd_nbooks/m2_pbulk_gtypes_GE_postQC_header.R")
# source("mali22_code_lnk/local_rmd_nbooks/m2_pbulk_gtypes_preQC_header.R")
source("mali22_code_lnk/local_rmd_nbooks/m2_pbulk_gtypes_preQC_cln_header.R")
# source("mali22_code_lnk/local_rmd_nbooks/m2_pbulk_gtypes_preQC_header_m55_m60.R")
# source("mali22_code_lnk/local_rmd_nbooks/m2_pm_pbulk_gtypes_preQC_header.R")
# source("mali22_code_lnk/local_rmd_nbooks/m2_pbulk_gtypes_postQC_cln_header.R")
sample_name_nms
```

#### Source colours, plotting functions and commonly used tables etc
```{r}
## Get the necessary variables for common housekeeping
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("mali22_code_lnk/pf_common_vars.R")
```

VCF to GDS conversion
```{r, warning=FALSE, message=FALSE, eval=T}
##Read in the genotypes and annotate

gt_ss_afm_p1_bi <-pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  gp <- ..2
  
  map(grps, ~seqVCF2GDS(paste0("data/processed/",species, "/", sm ,"/", gtyp_step, "/", .x,"_k", gp,"_", algn_nm,"/p1/fb_biallelic.vcf"), paste0("data/processed/",species, "/", sm ,"/", gtyp_step, "/", .x,"_k", gp,"_", algn_nm,"/p1/fb_biallelic.gds"),  storage.option="ZIP_RA"))
  
})

```

```{r, warning=FALSE, message=FALSE, eval=T}
##Read in the genotypes and annotate

pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  gp <- ..2
  map(grps, ~paste0("data/processed/",species, "/", sm ,"/", gtyp_step, "/", .x,"_k", gp,"_", algn_nm,"/p1/fb_biallelic.vcf"), paste0("data/processed/",species, "/", sm ,"/", gtyp_step, "/", .x,"_k", gp,"_", algn_nm,"/p1/fb_biallelic.gds"),  storage.option="ZIP_RA")

})

```


```{r, results="hide"}
##Merge SNP information above with annotations from plasmoDB to get gene positions etc for plotting
# dp  %>% filter_if(any(!is.na()))
gene_n_pos <- Pf3D7_genes_plasmoDB %>% 
  mutate(
    chr = str_extract(`Genomic Location (Gene)`, '.*v3'), 
    gn_start = str_remove_all(`Genomic Location (Gene)`, '.*v3:|,|\\.\\..*'), 
    gn_end = str_remove_all(`Genomic Location (Gene)`, '.*\\.\\.|,|\\(.*')
  ) %>%
  mutate(across(c(gn_start, gn_end), ~as.numeric(.))) %>% 
  dplyr::select(gene_id, chr, gn_start, gn_end, Gene) 

# write_csv(strain_top_snps, '/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/Mali2/data/processed/Pf/strain_top_snps.csv')

```



filter pf7 variantss
```{r}

pf7_waf <- read_csv("../../Pf3D7_genomes_gtfs_indexs/pf7_data/pf7_variants/AF-W-freqs.csv")

pf7_waf %>% 
  filter(is_invariant == F &  is_biallelic == T & minor_af > 0.3 & major_af < 0.7) %>% dim

pf7_waf_all <- data.table::fread("../../Pf3D7_genomes_gtfs_indexs/pf7_data/waf/Pf3D7_v3_g.txt")

pf7_waf_all_snp_qc <- pf7_waf_all %>% dplyr::select(V1, V2,V3,V4, paste0('V', c(11:17))) %>% rename(c('chrom' = V1, 'pos'=V2, 'ref'=V3, 'alt'=V4, 'filter'=V11,'type'=V12, 'VQSLOD'=V13,   'CDS'=V14, 'AC'=V15, 'AN'=V16, 'AF'=V17)) %>%  mutate(pf_all = 'yes')

```

filter pf7 variants
```{r}

pf7_mali <- read_csv("../../Pf3D7_genomes_gtfs_indexs/pf7_data/pf7_variants/Mali-freqs.csv")

pf7_mali_ibd_whitelst <- pf7_mali %>% 
  filter(is_invariant == F &  is_biallelic == T & minor_af > 0.1)

```


filter pf7 all variants
```{r}

pf7_mali_only <- data.table::fread("../../Pf3D7_genomes_gtfs_indexs/pf7_data/pf7_variants/Pf3D7_v3_g_bisplit.txt", na.strings='.')
pf7_mali_only_snp_qc <- pf7_mali_only %>% dplyr::select(V1, V2,V3,V4, paste0('V', c(11:17))) %>% rename(c('chrom' = V1, 'pos'=V2, 'ref'=V3, 'alt'=V4, 'filter'=V11,'type'=V12, 'VQSLOD'=V13,   'CDS'=V14, 'AC'=V15, 'AN'=V16, 'AF'=V17)) %>%  mutate(pf_all = 'yes')

# pf7_mali_only_nmajor001 <- data.table::fread("../../Pf3D7_genomes_gtfs_indexs/pf7_data/pf7_variants/Pf3D7_v3_g_pass_bisplit_snp_001.txt", na.strings='.')
# pf7_mali_only_nmajor001_snp_qc <- pf7_mali_only_nmajor001 %>% dplyr::select(V1, V2,V3,V4, paste0('V', c(11:17))) %>% rename(c('chrom' = V1, 'pos'=V2, 'ref'=V3, 'alt'=V4, 'filter'=V11,'type'=V12, 'VQSLOD'=V13,   'CDS'=V14, 'AC'=V15, 'AN'=V16, 'AF'=V17)) %>%  mutate(pf_all = 'yes')

```

read in variants from sunil for the pf high molecular weight ULI pacbio sequences - vcf generated by syri
```{r}

msc28_ULI_variants <- data.table::fread("../../Pf3D7_genomes_gtfs_indexs/sk21_variants/msc28_variants.txt", na.strings='.')

msc28_ULI_only_snp_qc <- msc28_ULI_variants %>% dplyr::select(V1, V2,V3,V4,V5,V6 ) %>% 
  rename(c('chrom' = V1, 'pos'=V2, 'ref'=V3, 'alt'=V4, 'filter'=V5,'type'=V6)) %>% 
  filter(filter == "PASS" & type == "SNP") %>% 
  mutate(msc28_ULI = 'yes',
         across(chrom, ~paste0(str_replace(., "chr", "Pf3D7_"), "_v3")))

msc61_ULI_variants <- data.table::fread("../../Pf3D7_genomes_gtfs_indexs/sk21_variants/msc61_variants.txt", na.strings='.')

msc61_ULI_only_snp_qc <- msc61_ULI_variants %>% dplyr::select(V1, V2,V3,V4,V5,V6 ) %>% 
  rename(c('chrom' = V1, 'pos'=V2, 'ref'=V3, 'alt'=V4, 'filter'=V5,'type'=V6)) %>% 
  filter(filter == "PASS" & type == "SNP") %>% 
  mutate(msc61_ULI = 'yes',
         across(chrom, ~paste0(str_replace(., "chr", "Pf3D7_"), "_v3")))

```



###########k8_raw
```{r}
## Function to read in the genotypes and annotate

gds_rd <- function(sample_id, grp, opt_k){  
  
  p1_bi <- seqOpen(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k,"_", algn_nm,"/p1/fb_biallelic.gds"), allow.duplicate=T)
  print(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k,"_", algn_nm,"/p1/fb_biallelic.gds"))
  
  print(head(seqGetData(p1_bi, "allele")))
  # print(head(seqGetData(p1_bi, "$dosage")))
  # print(head(seqGetData(p1_bi, "annotation/format/QR")))

  gtypes_p1_bi <- base::rbind(seqGetData(p1_bi, "allele"), seqGetData(p1_bi, "annotation/qual"),seqGetData(p1_bi, "chromosome"), seqGetData(p1_bi, "position"), seqGetData(p1_bi, "annotation/info/DP"), seqGetData(p1_bi, "annotation/filter"), seqGetData(p1_bi, "$dosage"), seqGetData(p1_bi, "annotation/format/DP"), seqGetData(p1_bi, "annotation/format/QR"), seqGetData(p1_bi, "annotation/format/QA")) %>% t()
  
  seqClose(p1_bi)
  print(gtypes_p1_bi[1:5,1:5])
  
  strn_nms <- seqVCF_SampID(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k,"_", algn_nm,"/p1/fb_biallelic.vcf")) %>% str_replace_all(c("PoorQC"="P", "Low_n"="Lo", "Female"="F", "Male" = "M", "Asexual" = "A", "Gametocyte" = "G"))
  
  colnames(gtypes_p1_bi) <- c('alleles', 'quality', 'chrom', 'pos', 'depth', 'filt', strn_nms, paste0(strn_nms, rep(c('_DP','_RefN', '_AltN'), each=length(strn_nms))))

  return(gtypes_p1_bi)
  
}
```


###########k8_raw stage_afm_strain
```{r, warning=FALSE, message=FALSE}
##Read in the genotypes and annotate

gt_ss_afm_p1_bi <-pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  k <- ..2
  map(grps, ~gds_rd(sample_id = sm, grp = .x, opt_k=k)) %>% set_names(paste0(..1 ,'.', grps, '.k', ..2))
  
})
```

Vartrix matrix to get number of molecules supporting either reference or alternate allele - all

```{r}
##Function to read vartrix output and annotate

vtx_rd <- function(sample_id, grp, opt_k ){
  
  ##Read in the alternate allele matrix for all the stg/strn combinations and combine into a list - also shortening the names of the matrices (list items)
  alt_mtx <- list.files(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k, "_", algn_nm,"/p1/vartrix_biallelic/"), pattern = "^[D|SC|G|P|N|M].*.mtx", full.names = T, recursive = T, include.dirs = T) %>% 
    set_names(nm = (str_remove_all(.,paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k, "_", algn_nm,"/p1/vartrix_biallelic/", '/|.mtx')) %>% str_replace_all(c("PoorQC"="P", "Low_n"="Lo", "Female"="F", "Male" = "M", "Asexual" = "A", "Gametocyte" = "G", "_a_" = "_A_"))  %>%  tools::file_path_sans_ext())) %>%
    map(readMM) 
  
  # print(str(alt_mtx))
  
  ##Count, for each SNP, the total alternate allele count over all cells and proportion of cell that have alternate allele - Using base R because it is faster
  alt_mtx_stats = lapply(alt_mtx, function(x) base::cbind(x, rowSums(x) , rowSums(x != 0)/ncol(x))[,c(ncol(x)+1, ncol(x)+2)])
  
  print(c("1", sample_id, grp))
  
  ##Name the matrix columns with the name of the stg/strn and suffix of snpNum (total snp count over all cell) and celProp (proportion of cell that have alternate allele)
  nmd_alt_mt<- lapply(names(alt_mtx_stats), function(x) {
       colnames(alt_mtx_stats[[x]]) <- paste0(x, c('_sNum', '_cProp'))
       as.matrix(alt_mtx_stats[[x]])
   })
  
  # print(str(nmd_alt_mt))
  print(c("2", sample_id, grp))
  
  nmd_alt_mt <- do.call(base::cbind, nmd_alt_mt) 
  print(c("finito", sample_id, grp))
  return(nmd_alt_mt)
  
}

```


```{r}
## For troubleshooting above script
##Read in the alternate allele matrix for all the stg/strn combinations and combine into a list - also shortening the names of the matrices (list items)
# sf <- list.files(paste0("data/processed/",species, "/", "MSC33" ,"/", gtyp_step, "/", "stage_afm_strain", "_k", "8", "_", algn_nm,"/p1/vartrix_biallelic/"), pattern = "^[D|SC|G|P|N].*.mtx", full.names = T, recursive = T, include.dirs = T) %>% 
#     set_names(nm = (str_remove_all(.,paste0("data/processed/",species, "/", "MSC33" ,"/", gtyp_step, "/", "stage_afm_strain", "_k", "8", "_", algn_nm,"/p1/vartrix_biallelic/", '/|.mtx')) %>% str_replace_all(c("PoorQC"="P", "Low_n"="Lo", "Female"="F", "Male" = "M", "Asexual" = "A", "Gametocyte" = "G", "_a_" = "_A_"))  %>%  tools::file_path_sans_ext())) %>%
#     map(readMM) 


```


###########k8_raw stage_afm_strain
```{r}
##Read in the genotypes and annotate
# NOTE - if we get error then check the header of the raw .mtx and .ref files from vartrix for consistency in number of records "number of rows of matrices must match (see arg 15)"

ss_afm_vtx_mtx <-pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  k <- ..2
  map(grps, ~vtx_rd(sample_id = sm, grp = .x, opt_k=k)) %>% set_names(paste0(..1 ,'.', grps))
})

```

###########k8_raw -TO BE DELETED IF RAW NOT PREFERED


Combine vartix snp count per cell with freebayes calls
```{r}
##remove doublets as well and low quality SNPS - maybe not include doublets in variant calling with freebayes. Adds no value? or could be used to demonstrate high number of non-distinct and rule out small clusters as doublets and be part of QC
fb_vt_bind <- function(fb_gtype , vtx_mtx, rm_grps="None") {
  gt_fb_vtx_p1_bi <- base::cbind(fb_gtype, vtx_mtx) %>% 
    as.data.frame() %>% 
    mutate(across(c(pos, depth, quality, matches('.*_DP$|.*_RefN$|.*_AltN$|.*sNum$|.*cProp$')), as.numeric)) %>%
    dplyr::select(-c(matches(rm_grps))) %>%
    filter(if_any(starts_with('SC'), ~!is.na(.))) %>%
    filter(if_any(starts_with('MSC'), ~!is.na(.))) %>%
    # filter(quality>75 & depth>40)  %>%
    mutate(filt_reason1 = case_when(quality<=75 | depth<=40 ~ 'Low quality',
                                    str_detect(chrom, mt_chrom_nm) ~ 'Mt',
                                    T ~ "Pass_1"))
  
  print(dim(gt_fb_vtx_p1_bi))
  print(table(gt_fb_vtx_p1_bi$filt_reason1, useNA = 'ifany'))
  
  ##remove  mitochondrial calls
  # gt_fb_vtx_p1_bi <- gt_fb_vtx_p1_bi %>%
    # filter(!str_detect(chrom, 'Pf_M'))%>%
    # mutate(filt_reason1 = case_when(str_detect(chrom, 'Pf_M') ~ 'Mt'))
  
  print(dim(gt_fb_vtx_p1_bi))
  
  return(gt_fb_vtx_p1_bi)
}
```


```{r}
## raw freebayes vcf + vatrix high quality no mitochondrial DNA

ss_afm_fb_vtx_p1bi <- pmap(list(unlist(gt_ss_afm_p1_bi, recursive = F), unlist(ss_afm_vtx_mtx, recursive = F)), ~fb_vt_bind(fb_gtype = ..1, vtx_mtx = ..2))
```


## SNPS from the pf7 paper without filtering

```{r}
##soupo with pf7 all variants including poor quality

gt_pf7_merge <- function(gt_tbl){  
  gt_pf7_mali <- pf7_mali_only_snp_qc %>% 
  left_join(gt_tbl,., by = c("chrom", "pos"), multiple = "all")
  
  grpd_tbl <- gt_pf7_mali %>% group_by(chrom,pos) %>% mutate(across(type, ~factor(., levels = c('SNP', 'INDEL', 'OVERLAP')))) %>% arrange(type, .by_group = T) %>% arrange(dplyr::desc(AF), .by_group = T) 
    
  grpd_tbl %>% dplyr::select(c('chrom', 'pos','alleles','ref', 'alt',  'type', 'AF', 'AC')) %>% dim() %>% print()
  
  return(gt_pf7_mali)
}

```

```{r}
## raw genotypes merge with pf Mali

ss_afm_pf7_mali <- map(ss_afm_fb_vtx_p1bi, ~gt_pf7_merge(gt_tbl=.x))

```


```{r}
##Visualize the distribution of SNPs, INDELs and overlaps within the split multi-alleles

ss_afm_pf7_mali[[1]] %>%  group_by(chrom,pos) %>% mutate(across(type, ~factor(., levels = c('SNP', 'INDEL', 'OVERLAP')))) %>% arrange(type, .by_group = T) %>% arrange(dplyr::desc(AF), .by_group = T) #%>% View()


```

```{r}
##Select only the SNP record from the split alleles and if more than one SNP in position then keep the most frequent alternate allele - distinct keeps the first row. No big difference when we dont first sort by type and just choose the most frequent alternate allele - SNP still top except for around 6 where indel swaps with SNP
## In vcf drop non-mali alleles, split after pass filter then select SNPs from this and maybe filter by allele frequency if too many SNPs
# merge with SNPs only after spliting

pf7_snps_qc_multitop <- function(gt_pf7_mali) {
  
  pf7_mali_nqc_snptop <- gt_pf7_mali %>% group_by(chrom,pos) %>% mutate(across(type, ~factor(., levels = c('SNP', 'INDEL', 'OVERLAP')))) %>% arrange(type, .by_group = T) %>% arrange(dplyr::desc(AF), .by_group = T) %>% distinct(., pick(chrom, pos), .keep_all = T) %>% ungroup()
  
}


```

```{r}
## raw genotypes merge with pf7M (pf7Mali) retaining only the top SNPs from the Mali dataset where there are complex alleles with overlaps with indel etc

ss_afm_pf7M_snptop <- map(ss_afm_pf7_mali, ~pf7_snps_qc_multitop(gt_pf7_mali=.x))
```


```{r}
##Jacob filtered SNPs - some missing in MSC14
# m14_pf7_prep_viz <- function(m14_gt, pf7_base = pf7_mali) {
#   
#   m14_pf7_all_jcb <- pf7_mali %>% 
#     mutate(pf7_jcb = 'yes') %>%
#     left_join(m14_gt,., by = c("chrom" = "chromosome", "pos" = "position"))  %>% 
#     dplyr::select(c(chrom, pos, depth, type, filter,VQSLOD,CDS,AC,AN,AF, pf_all, quality, ends_with('_qc'), pf7_jcb,filt_reason1))%>% 
#     mutate(p_dset = case_when(pf_all == 'yes' & pf7_jcb == 'yes' ~ 'm14_pAll_pJcb', 
#                               pf_all == 'yes' & is.na(pf7_jcb) ~ 'm14_pAll', 
#                               is.na(pf_all) & pf7_jcb == 'yes' ~ 'm14_pJcb', 
#                               is.na(pf_all) & is.na(pf7_jcb)  ~ 'm14')) 
#   
#   m14_pf7_all_jcb %>% dplyr::count(p_dset, type, filter) %>%
#     print()
#   
#   plt<- m14_pf7_all_jcb  %>% dplyr::count(p_dset, type, filter) %>% 
#     filter(n>1) %>%
#     mutate(across(filter, ~str_replace_all(., c(';'= '\n','Hyp'= '\nHyp')))) %>% 
#     ggplot(., aes(x =p_dset, y=n, fill = type)) + 
#     geom_col(position = position_dodge(width = 1)) + 
#     facet_grid(.~filter) + 
#     theme(axis.text.x = element_text(angle=90), text = element_text(size = 14)) 
#     
#   print(plt)
#   
#   return(m14_pf7_all_jcb)
#   
# }
```


```{r, fig.height=5, fig.width=10}
##strain stage k8 raw genotypes merge with pf Mali

# ss_afm_pf7M_jcb <- map(ss_afm_pf7M_snptop, ~m14_pf7_prep_viz(m14_gt=.x))
```


```{r, fig.height=5, fig.width=10}
msc_ULI_snps <- msc28_ULI_only_snp_qc %>% bind_rows(., msc61_ULI_only_snp_qc) %>% distinct(chrom, pos, .keep_all = T) %>% mutate(ULI_snp = coalesce(msc28_ULI, msc61_ULI)) %>% select(chrom, pos,, ULI_snp)

```

```{r}
##Jacob filtered SNPs - some missing in MSC14
m14_pf7_prep_viz <- function(m14_gt, pf7_base = pf7_mali) {
  
  m14_pf7_all_jcb <- pf7_base %>% 
    mutate(pf7_jcb = 'yes') %>%
    left_join(m14_gt,., by = c("chrom" = "chromosome", "pos" = "position"))  %>% 
    left_join(., msc_ULI_snps) %>%
    mutate(across(filter, ~case_when(is.na(.) & ULI_snp == 'yes' ~ 'PASS', 
                                       T ~ as.character(.))),
             across(type, ~case_when(is.na(.) & ULI_snp == 'yes' ~ 'SNP', 
                                       T ~ as.character(.)))
           ) %>%
    # dplyr::select(c(chrom, pos, depth, type, filter,VQSLOD,CDS,AC,AN,AF, pf_all, quality, ends_with('_qc'), pf7_jcb,filt_reason1, ULI_snp))%>% 
    mutate(p_dset = case_when(pf_all == 'yes' & pf7_jcb == 'yes' ~ 'm14_pAll_pJcb', 
                              pf_all == 'yes' & is.na(pf7_jcb) ~ 'm14_pAll', 
                              is.na(pf_all) & pf7_jcb == 'yes' ~ 'm14_pJcb', 
                              is.na(pf_all) & ULI_snp == 'yes' ~ 'm14_ULI', 
                              is.na(pf_all) & is.na(pf7_jcb)  ~ 'm14')
           ) 
  
  m14_pf7_all_jcb %>% dplyr::count(p_dset, type, filter) %>%
    print()
  
  plt<- m14_pf7_all_jcb  %>% dplyr::count(p_dset, type, filter) %>% 
    filter(n>1) %>%
    mutate(across(filter, ~str_replace_all(., c(';'= '\n','Hyp'= '\nHyp')))) %>% 
    ggplot(., aes(x =p_dset, y=n, fill = type)) + 
    geom_col(position = position_dodge(width = 1)) + 
    facet_grid(.~filter) + 
    theme(axis.text.x = element_text(angle=90), text = element_text(size = 14)) 
    
  print(plt)
  
  return(m14_pf7_all_jcb)
  
}
```


```{r, fig.height=5, fig.width=10}
##strain stage k8 raw genotypes merge with pf Mali

ss_afm_pf7M_jcb <- map(ss_afm_pf7M_snptop, ~m14_pf7_prep_viz(m14_gt=.x))
```


```{r}
kas_togo_pf <- readxl::read_xlsx("../Mali/data/raw/published_lit/kassegne_togo_pf_wgs/Table 1.XLSX", sheet = 'Table 1', skip = 1) %>% 
  rename(c('chrom'='CHROMOSOME', 'pos'='POS')) %>% 
  mutate(kas_togo = 'yes')
```


```{r, fig.height=5, fig.width=10}
ss_afm_pf7M_jcb[[1]] %>% 
  left_join(., kas_togo_pf, by = c('chrom', 'pos'))  %>% dplyr::count(pf_all, ULI_snp, kas_togo)

# ss_ag_k6r_pf7M_jcb %>% 
#   left_join(., kas_togo_pf, by = c('chrom', 'pos')) %>% filter(filt_reason1 == "Pass_1") %>% dplyr::count(pf_all, kas_togo)
# 
# s_k6r_pf7M_jcb %>% 
#   left_join(., kas_togo_pf, by = c('chrom', 'pos')) %>% filter(filt_reason1 == "Pass_1") %>% dplyr::count(pf_all, kas_togo)
```


```{r, fig.height=5, fig.width=10}
# gogo()
```



Add the ULI SNPs and change filter column to contain PASS if it is NA and the SNPs is in the ULI snp
```{r, fig.height=5, fig.width=10}
# ss_afm_pf7M_jcb <- map(ss_afm_pf7M_jcb2, ~.x %>% 
#       # left_join(., msc_ULI_snps, by = c('chrom', 'pos'))  %>% 
#       mutate(across(filter, ~case_when(is.na(.) & ULI_snp == 'yes' ~ 'PASS', 
#                                        T ~ as.character(.))),
#              across(type, ~case_when(is.na(.) & ULI_snp == 'yes' ~ 'SNP', 
#                                        T ~ as.character(.)))
#            ) 
#   )

```


```{r}
##Jacob filtered SNPs - some missing in MSC14
# map(ss_afm_pf7M_jcb, ~.x  %>% dplyr::count(p_dset, type, filter) %>% 
#     filter(n>1) %>%
#     mutate(across(filter, ~str_replace_all(., c(';'= '\n','Hyp'= '\nHyp')))) %>% 
#     ggplot(., aes(x =p_dset, y=n, fill = type)) + 
#     geom_col(position = position_dodge(width = 1)) + 
#     facet_grid(.~filter) + 
#     theme(axis.text.x = element_text(angle=90), text = element_text(size = 14)) 
# )
```

```{r, fig.height=5, fig.width=10}
regions_hyp <- read_delim("../../Pf3D7_genomes_gtfs_indexs/pf7_data/regions-20130225.onebased.txt", col_names = F)
regions_hyp <- regions_hyp %>% mutate(chrom = as.numeric(str_remove_all(X1, 'Pf3D7_|_v3')))
```

mark snps not in pf7 dataset
```{r}
##remove doublets as well and low quality SNPS - maybe not include doublets in variant calling with freebayes. Adds no value? or could be used to demonstrate high number of non-distinct and rule out small clusters as doublets and be part of QC
pf7_hlght <- function(m14_gt ) {
  pf7_h <-  m14_gt %>%
    mutate(across(filt_reason1, ~case_when(. == "Pass_1" & (filter != 'PASS' | is.na(filter)) ~ 'Absent Pf7_Pass & ULI', ## absent in pf 7 means also non - PASS 
                                           . == "Pass_1" ~ 'Pass',
                                           T ~ as.character(.)))
           ) 
  
  pf7_h %>% dplyr::count(filt_reason1, ULI_snp, filter)%>% print()
  
  return(pf7_h)
}
```


```{r}
ss_afm_pf7M_pf7qc <- map(ss_afm_pf7M_jcb, ~pf7_hlght(m14_gt=.x))
```

```{r}
# ss_afm_pf7M_pf7qc <- map(ss_afm_pf7M_snptop, ~pf7_hlght(m14_gt=.x))
```

```{r, fig.height=4, fig.width=15}
##Function for cleaner plot of SNPs falling in the different filter categories for each stage +/ strain cluster

broadqc_lvls <- c('Pass', 'Absent Pf7_Pass & ULI', 'Mt', 'Low quality')
broadqc_cols <- met.brewer('Isfahan2')[c(2:4,1)] %>% set_names(broadqc_lvls)


cln_plt_snp_filt_categ_pf7qc <- function(annot_snp_qc, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm=''){
  
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
annot_snp_qc %>% 
  filter(!(!!rlang::sym(filt_c) %in% filt_rm)) %>%
  mutate(across(!!rlang::sym(filt_c), ~droplevels(factor(., levels = qc_lvls)))) %>%
  # dplyr::count(groups, !!rlang::sym(filt_c), .drop=F) %>%
    dplyr::count(!!rlang::sym(filt_c), .drop=F) %>%
  # group_by(groups) %>% mutate(perc = round((n/sum(n))*100, digits = 1)) %>%
    mutate(perc = round((n/sum(n))*100, digits = 1)) %>%
  ggplot(aes(x=!!rlang::sym(filt_c),fill=!!rlang::sym(filt_c), y = n/1000)) + 
  geom_col(position = position_dodge(width = 1)) + 
  geom_text(aes(label=paste0(n, "(", perc,")"), color = !!rlang::sym(filt_c)), vjust = 0.5, hjust = -0.1, position = position_dodge(width = 1), size = 6.5, angle = 90, show.legend = F, fontface= 'bold') + 
  scale_color_manual(values = qc_cols) +
  scale_fill_manual(values = qc_cols) +
  scale_y_continuous(expand = expansion(mult =c(0, 0.8))) + 
  # facet_wrap(.~groups, nrow = 1,scales = "free_x", drop = T) +
  labs(x = xlb, y = 'Number of SNPs (X 1e3)', fill = 'SNP category') +
  theme_classic()+
  theme(axis.text.y = element_text(size = 14),
        axis.title.x = element_blank(),
        strip.text.x.top = element_text(size = 15, face='bold'),
        axis.text.x = element_blank(),
        axis.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 18, face = 'bold'),
        legend.position="bottom",
        panel.border=element_rect(colour="grey",fill='transparent'),#,
        # plot.margin = unit(c(1,2,1,1),"cm")
        # legend.margin=margin()
      ) +
    # coord_fixed()+
    # coord_flip() +
    guides(fill = guide_legend( nrow = 1, title.position = "top"))
}
```


```{r}
##!!!CHEATING - SORT OUT
ss_afm_broadqc_plt <- map(ss_afm_pf7M_pf7qc, ~cln_plt_snp_filt_categ_pf7qc(annot_snp_qc = .x, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm=''))

ss_afm_broadqc_plt

```

```{r}
##!!!CHEATING - SORT OUT
ss_afm_broadqc_plt <- map(ss_afm_pf7M_pf7qc, ~cln_plt_snp_filt_categ_pf7qc(annot_snp_qc = .x, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm=''))

ss_afm_broadqc_plt

```

```{r}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
# ss_ag_broadqc_plt <- cln_plt_snp_filt_categ_pf7qc(annot_snp_qc = ss_ag_k6r_pf7M_pf7qc, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm='')
# ss_ag_broadqc_plt
# 
# s_k6r_broadqc_plt <- cln_plt_snp_filt_categ_pf7qc(annot_snp_qc = s_k6r_pf7M_pf7qc, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm='')
# s_k6r_broadqc_plt

```

##Assess filtered SNPs

```{r}
##!!!! NOTE :Adding S to [G|D] [G|D|S] so as to detect the SC labeles in the k6 raw
##!!!! NOTE :Using instead of gt_fb_vtx_p1_bi_qc_pf7 since we want to only look at SNPs in pf 7 Mali
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
# snp_qc_tbl <- gt_fb_vtx_p1_bi_qc_pf7 %>% rename_with( ~str_replace(., '$','_raw'), matches("^[G|D|S].*_[A-Z]$")) %>% dplyr::select(chrom, pos,

snps_qc_filts <- function(m14_gt){
  
  snp_qc_tbl <- m14_gt %>% filter(filt_reason1 == 'Pass') %>% rename_with( ~str_replace(., '$','_raw'), matches("^(G|D|SC|P_|N|M).*(_A|_F|_M|_G|_Lo|_G|blet|tive|Unassigned|fail|[0-9])$")) %>% dplyr::select(chrom, pos, matches('.*_raw$|.*_qc$|.*_sNum$|.*_cProp$|.*_DP$|.*_RefN$|.*_AltN$'), type, filter,VQSLOD,CDS,AC,AN,AF, alleles)   %>% 
      pivot_longer(.,cols = matches('^(G|D|SC|P_|N|M).*'), 
                   names_pattern = "(^[G|D|SC|P_|N|M].*[A|M|F|G|Lo|t|tive|Unassigned|stgQC_fail|(0-9)])_(alt_sNum|alt_cProp|ref_sNum|ref_cProp|raw|qc|DP|RefN|AltN)", 
                   names_to = c( "groups" ,".value")) %>%
      mutate(filt_reason = case_when(filter != 'PASS' | is.na(filter) ~ 'Absent Pf7_Pass & ULI', ##is.na(type) ~ 'Absent Pf7_Pass & ULI',### !!NB - Marking those missing in the pf7 dataset first. May need to be put last below if we want to include those absent if pf7 as some of these have been seen to be of good quality
                                     alt_sNum < 8 & ref_sNum < 8 ~  "Low UMI support",
                                     alt_sNum >= 15 & ref_sNum >= 15 ~  "Non distinct",
                                     abs(alt_sNum - ref_sNum) < 8 ~ 'Noisy' ,
                                     raw == '1' & alt_sNum > ref_sNum ~ 'Ref inverted' ,
                                     raw == '0' & ref_sNum > alt_sNum ~ 'Alt inverted',
                                     is.na(raw)&alt_sNum!=0&ref_sNum!=0  ~ "No fbayes call",
                                     str_detect(filter, 'Hyper') ~ 'Hypervariable',
                                     # is.na(type) ~ 'Absent Pf7_Pass & ULI',
                                     T ~ 'Clean'
                                     # !is.na(qc) ~ "Clean"
                                     )
             ) %>%
    mutate(qc2 = case_when(filt_reason == 'Clean' ~ raw))
  
  return(snp_qc_tbl)

  }
```

```{r}
ss_afm_pf7M_qc <- map(ss_afm_pf7M_pf7qc, ~snps_qc_filts(m14_gt=.x))
```


```{r}
## make directory for strain summary output
system(paste0("mkdir -p data/processed/",species, "/" , sv_dir))

```

```{r, fig.height=5, fig.width=14}
saveRDS(ss_afm_pf7M_qc, paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_qc.RDS"))
```

```{r, eval=FALSE}
## Make snp_qc directory for storing qc files
imap(ss_afm_pf7M_qc, ~system(paste0('mkdir data/processed/',species,'/', str_remove(.y, '\\..*'),'/', sv_dir, '/',str_remove(.y, 'M.*\\.'),'_k', optimum_k[str_remove(.y, '\\..*')],'_', algn_nm,'/p1/snp_qc')))

imap(ss_afm_pf7M_qc, ~write_csv(.x, paste0('data/processed/',species,'/', str_remove(.y, '\\..*'),'/', sv_dir, '/',str_remove(.y, 'M.*\\.'),'_k', optimum_k[str_remove(.y, '\\..*')],'_', algn_nm,'/p1/snp_qc/ss_afm_pf7M_qc.csv')))


```

Minimal QC (lessQC [lqc]) retaining the "Noisy", "Non distinct", 'Ref inverted' and 'Alt inverted' categories
```{r}
##!!!! NOTE :Same script as above just commenting out the lines to remove "Noisy", "Non distinct", 'Ref inverted' and 'Alt inverted' categories - Need to find a clever way to do this to avoid repetition
##!!!! NOTE :Adding S to [G|D] [G|D|S] so as to detect the SC labels in the k6 raw
##!!!! NOTE :Using instead of gt_fb_vtx_p1_bi_qc_pf7 since we want to only look at SNPs in pf 7 Mali
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
# snp_qc_tbl <- gt_fb_vtx_p1_bi_qc_pf7 %>% rename_with( ~str_replace(., '$','_raw'), matches("^[G|D|S].*_[A-Z]$")) %>% dplyr::select(chrom, pos,

snps_qc_filts_lqc <- function(m14_gt){
    
    snp_qc_tbl <- m14_gt %>% filter(filt_reason1 == 'Pass') %>% rename_with( ~str_replace(., '$','_raw'), matches("^(G|D|SC|P_|N|M).*(_A|_F|_M|_G|_Lo|_G|blet|tive|Unassigned|fail|[0-9])$")) %>% dplyr::select(chrom, pos, matches('.*_raw$|.*_qc$|.*_sNum$|.*_cProp$|.*_DP$|.*_RefN$|.*_AltN$'), type, filter,VQSLOD,CDS,AC,AN,AF, alleles)   %>% 
        pivot_longer(.,cols = matches('^(G|D|SC|P_|N|M).*'), 
                     # names_pattern = "(^[G|D|S].*[A|M|F|blet])_([alt|ref|raw|qc|DP|RefN|AltN].*)",
                     # names_pattern = "(^[G|D|S].*[A|M|F|t])_(alt_sNum|alt_cProp|ref_sNum|ref_cProp|raw|qc|DP|RefN|AltN)", 
                     names_pattern = "(^[G|D|SC|P_|N|M].*[A|M|F|G|Lo|t|tive|Unassigned|stgQC_fail|(0-9)])_(alt_sNum|alt_cProp|ref_sNum|ref_cProp|raw|qc|DP|RefN|AltN)", 
                     names_to = c( "groups" ,".value")) %>%
        mutate(filt_reason = case_when(filter != 'PASS' | is.na(filter) ~ 'Absent Pf7_Pass & ULI', ##is.na(type) ~ 'Absent Pf7_Pass & ULI',### !!NB - Marking those missing in the pf7 dataset first. May need to be put last below if we want to include those absent if pf7 as some of these have been seen to be of good quality
                                       alt_sNum < 5 & ref_sNum < 5 ~  "Low UMI support",
                                       # alt_sNum >= 15 & ref_sNum >= 15 ~  "Non distinct",
                                       # abs(alt_sNum - ref_sNum) < 8 ~ 'Noisy' ,
                                       # raw == '1' & alt_sNum > ref_sNum ~ 'Ref inverted' ,
                                       # raw == '0' & ref_sNum > alt_sNum ~ 'Alt inverted',
                                       # is.na(raw)&alt_sNum!=0&ref_sNum!=0  ~ "No fbayes call",
                                       # str_detect(filter, 'Hyper') ~ 'Hypervariable',
                                       # is.na(type) ~ 'Absent Pf7_Pass & ULI',
                                       T ~ 'Clean'
                                       # !is.na(qc) ~ "Clean"
        )
        ) %>%
        mutate(qc2 = case_when(filt_reason == 'Clean' ~ raw))
    
    return(snp_qc_tbl)
    
}
```


```{r}
# s_lqc_k6r_pf7M_qc <- snps_qc_filts_lqc(m14_gt=s_k6r_pf7M_pf7qc)
ss_afm_pf7M_lqc <- map(ss_afm_pf7M_pf7qc, ~snps_qc_filts_lqc(m14_gt = .x))
# 
# s_lqc_k6r_pf7M_qc %>% write_csv(., '/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/',sample_set,'/data/processed/msc48/', gtyp_step, '/strain_qcd_k8_minimp/p1/snp_qc/s_lqc_k6r_pf7M_qc.csv')

```

```{r}
# ss_afm_pf7M_qc <- map(ss_afm_pf7M_pf7qc, ~snps_qc_filts(m14_gt=.x))
```

```{r, fig.height=4, fig.width=20}
##Function to plot snps per the number of SNPs falling in the different filter categories for each stage + strain cluster
plt_snp_filt_categ <- function(annot_snp_qc, xlb = 'Strain'){
  
  annot_snp_qc %>% dplyr::count(groups, filt_reason) %>% 
  ggplot(aes(x=groups,fill=filt_reason, y = n)) + 
  geom_col(position = position_dodge(width = 1)) + 
  geom_text(aes(label=n, color = filt_reason), vjust = 0.5, hjust = -0.1, position = position_dodge(width = 1), size = 5, label.padding = unit(0.1, "lines"), angle = 90, show.legend = F) + 
  scale_y_continuous(expand = expansion(mult =c(0, 0.25))) + 
  labs(x = xlb, y = 'Number of SNPs', fill = 'SNP category') +
  theme_classic()+
  theme(axis.text.x = element_text(size = 16, angle = 90),
        axis.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 18, face = 'bold')) 
  
}


```


```{r, fig.height=4, fig.width=20}
##Function to plot snps per the number of SNPs falling in the different filter categories for each stage + strain cluster
map(ss_afm_pf7M_qc, ~plt_snp_filt_categ(annot_snp_qc = .x))

```


```{r, fig.height=4, fig.width=20}
##Function to plot snps per the number of SNPs falling in the different filter categories for each stage + strain cluster
# plt_snp_filt_categ(annot_snp_qc = ss_ag_k6r_pf7M_qc, xlb = 'Strain stage combinations')

```


```{r, fig.height=4, fig.width=20}
##Function to plot snps per the number of SNPs falling in the different filter categories for each stage + strain cluster
# plt_snp_filt_categ(annot_snp_qc = s_k6r_pf7M_qc)
# plt_snp_filt_categ(annot_snp_qc = s_lqc_k6r_pf7M_qc)
```

Count those absent in pf7

```{r, fig.height=5, fig.width=14}
##SNPs in pf7
# ss_ag_k6r_pf7M_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups) %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% distinct(pick(chrom,pos)) %>% dim
# 
# ##SNPs where there's less than 8 UMI
# ss_ag_k6r_pf7M_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups) %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% filter(filt_reason == 'Low UMI support') %>% distinct(chrom,pos) %>% dim
# 
# ##Genotypes where there's less than 8 UMI
# ss_ag_k6r_pf7M_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups) %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% filter(filt_reason == 'Low UMI support') %>% dim
#
# ##Counts of genotypes where there's less than 8 UMI across groups (range medium)
# ss_ag_k6r_pf7M_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups) %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% filter(filt_reason == 'Low UMI support') %>% dplyr::count(groups, filt_reason)
# 
# ##Summary of genotypes where there's less than 8 UMI across groups (range medium)
# ss_ag_k6r_pf7M_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups) %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% filter(filt_reason == 'Low UMI support') %>% dplyr::count(groups, filt_reason) %>% pull(n) %>% summary

```

```{r, fig.height=4, fig.width=15}
##Function for cleaner plot of SNPs falling in the different filter categories for each stage +/ strain cluster

snp_qc_lvls <- c('Clean','Low UMI support', 'Non distinct', 'Noisy', 'Ref inverted', 'Alt inverted', 'No fbayes call')
snp_qc_cols <- c(met.brewer('NewKingdom')[c(2,3,1,4,5)],met.brewer('Navajo')[c(2,3)])  %>% set_names(snp_qc_lvls)


cln_plt_snp_filt_categ <- function(annot_snp_qc, xlb = 'Strain', ttl){
  
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
annot_snp_qc %>% 
  # filter(!filt_reason %in% c('Absent Pf7_Pass & ULI', 'Low UMI support')) %>% 
  mutate(across(filt_reason, ~droplevels(factor(., levels = snp_qc_lvls)))) %>%
  dplyr::count(groups, filt_reason, .drop=F) %>%
    mutate(n_imp = as.numeric(case_when(filt_reason == 'Low UMI support' ~ NA, T ~ n)))%>%
  mutate(across(groups, ~str_remove_all(., "MSC[0-9]+_"))) %>% 
    group_by(groups) %>% 
    mutate(perc = round((n_imp/sum(n_imp, na.rm = T))*100, digits = 1)) %>%
  ggplot(aes(x=filt_reason,fill=filt_reason, y = n/1000)) + 
  geom_col(position = position_dodge(width = 1)) + 
  geom_text(aes(label=ifelse(is.na(n_imp), "", paste0(n_imp, "(", perc,")")), color = filt_reason), vjust = 0.5, hjust = -0.1, position = position_dodge(width = 1), size = 6.5, angle = 90, show.legend = F, fontface= 'bold') + 
  scale_color_manual(values = snp_qc_cols) +
  scale_fill_manual(values = snp_qc_cols) +
  scale_y_continuous(expand = expansion(mult =c(0, 0.05))) + 
  facet_wrap(.~groups, nrow = 1,scales = "free_x", drop = T) +
  labs(x = xlb, y = 'Number of SNPs (X 1e3)', fill = 'SNP category', title = ttl) +
  theme_classic()+
  theme(axis.text.y = element_text(size = 14),
        axis.title.x = element_blank(),
        strip.text.x.top = element_text(size = 15, face='bold'),
        axis.text.x = element_blank(),
        axis.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 18, face = 'bold'),
        legend.position="bottom",
        panel.border=element_rect(colour="grey",fill='transparent'),
        plot.title = element_text(size = 20)#,
        # plot.margin = unit(c(1,2,1,1),"cm")
        # legend.margin=margin()
      ) +
    # coord_fixed()+
    # coord_flip() +
    guides(fill = guide_legend( nrow = 1, title.position = "top"))
}
```


```{r, fig.height=6, fig.width=30}
map(ss_afm_pf7M_qc, ~.x %>% count(filt_reason ))

```

```{r, fig.height=6, fig.width=30}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
##!!!CHEATING - SORT OUT
ss_afm_k6r_qc_plt <- imap(ss_afm_pf7M_qc, ~cln_plt_snp_filt_categ(annot_snp_qc = .x, ttl = .y))
ss_afm_k6r_qc_plt

```



```{r, fig.height=6, fig.width=24, eval=FALSE}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
##!!!CHEATING - SORT OUT
ss_afm_k6r_lqc_plt <- imap(ss_afm_pf7M_lqc, ~cln_plt_snp_filt_categ(annot_snp_qc = .x, ttl = .y))
ss_afm_k6r_lqc_plt

```


```{r, fig.width= 12, fig.height=15.4}
##Put together the plots for annotation
sag_bqc <- ss_afm_broadqc_plt[str_detect(names(ss_afm_broadqc_plt), "\\.stage_ag_strain_qcd")]
s_bqc <- ss_afm_broadqc_plt[str_detect(names(ss_afm_broadqc_plt), "\\.strain_qcd")]
sag_nqc <- ss_afm_k6r_qc_plt[str_detect(names(ss_afm_k6r_qc_plt), "\\.stage_ag_strain_qcd")]
s_nqc <- ss_afm_k6r_qc_plt[str_detect(names(ss_afm_k6r_qc_plt), "\\.strain_qcd")]


pmap(list(sag_bqc, s_bqc, sag_nqc, s_nqc), ~{
        plot_grid(plot_grid(..1 +
            theme(plot.margin = margin(t=1.1, b=0.3, l=0.1,r=0.3, unit = "cm")),
                    ..2 +
            theme(plot.margin = margin(t=1.1, b=0.3,l=0.1,r=0.3, unit = "cm")), nrow = 1,
              labels = c('A', 'B'),
              label_size = 24,
              vjust= 1) ,
          ..3 +
            theme(plot.margin = margin(t=1.1, b=0.3, unit = "cm")),
          ..4 +
              theme(plot.margin = margin(t=1.1, b=0.3, unit = "cm")),
              nrow = 3,
              labels = c('', 'C', 'D'),
              label_size = 24,
              vjust= 1)
      }
)

``` 



```{r, fig.height=5, fig.width=14}
##Visualize and count the different classes of SNPs removed in QC
snp_filt_bias_viz <- function(annot_snp_qc, filt_cat){
  
  snps_filt_bias <- annot_snp_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups)  %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% group_by(chrom,pos) %>% filter(!all(filt_reason == 'Low UMI support')) 
  
  snps_filt_bias_fc <- snps_filt_bias %>% filter(any(filt_reason == filt_cat)) %>% group_by(chrom, pos) %>% ungroup() 
  # snps_filt_bias_fc %>%  pivot_wider(names_from = groups, values_from = filt_reason) %>% View()
  snps_filt_bias_fc %>%  pivot_wider(names_from = groups, values_from = filt_reason) %>% dim() %>% print()
  snps_filt_bias_fc %>% filter(filt_reason == filt_cat) %>% dplyr::count(groups) %>% print()
  snps_filt_bias_fc %>%  dplyr::count(filt_reason, groups) %>% print()
  snps_filt_bias %>% filter(any(filt_reason == filt_cat)) %>% filter(any(filt_reason == 'Clean'))  %>% group_by(chrom, pos) %>% filter(sum(filt_reason == filt_cat) > 2) %>% ungroup() %>%  pivot_wider(names_from = groups, values_from = filt_reason) %>% dim %>% print()

}
```


```{r, fig.height=5, fig.width=14}
##Visualize and count the different classes of SNPs removed in QC
for (fc in c('Non distinct', 'Noisy', 'Alt inverted', 'Ref inverted')){
  map(ss_afm_pf7M_qc, ~snp_filt_bias_viz(annot_snp_qc = .x, filt_cat = fc))
}
```


```{r, fig.height=5, fig.width=14}
##Visualize and count SNPs where some stage strain groups support 'Non distinct' or 'Noisy' and other groups support 'Clean'
# ss_afm_pf7M_qc[["MSC14.stage_ag_strain_qcd"]] %>% dplyr::distinct(chrom, pos, filt_reason, groups) %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% filter(filt_reason != 'Absent Pf7_Pass & ULI') %>% group_by(chrom,pos) %>% filter(!all(filt_reason == 'Low UMI support')) %>% filter(any(filt_reason == 'Non distinct' | filt_reason == 'Noisy')) %>% filter(any(filt_reason == 'Clean'))  %>% group_by(chrom, pos) %>% ungroup %>% dplyr::count(filt_reason, groups) 
# #%>% View


```


```{r}
##Viz and Remove doublets and get informative SNPs
snp_rm_dblt_infomtv <- function(annot_snp_qc, clst2rmv = "Doublet"){
  annot_snp_qc %>% filter(str_detect(groups, clst2rmv)) %>% distinct(., pick(chrom,pos), .keep_all = T) %>% mutate(id =paste0(chrom,'_',pos)) %>% pull(id) %>% length() %>% print()

  annot_snp_qc_nd <- annot_snp_qc %>% filter(!str_detect(groups, clst2rmv))%>% filter(!is.na(qc2))
  
  annot_snp_qc_nd %>% group_by(chrom, pos) %>% filter(n_distinct(qc2) == 1 ) %>% distinct(., pick(chrom,pos), .keep_all = T) %>% dim() %>% print()
  
  annot_snp_qc_infomtv <- annot_snp_qc_nd  %>% group_by(chrom, pos) %>% filter(n_distinct(qc2) ==2 ) %>% ungroup
  
  annot_snp_qc_infomtv %>% distinct(., pick(chrom,pos), .keep_all = T) %>% dim() %>% print()


return(list(annot_snp_qc_nd, annot_snp_qc_infomtv))

}
```

```{r, fig.height=5, fig.width=14}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
ss_afm_pf7M_qc_infomtv <- map(ss_afm_pf7M_qc, ~snp_rm_dblt_infomtv(annot_snp_qc = .x, clst2rmv = "None"))
```

```{r, fig.height=5, fig.width=14}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
ss_afm_pf7M_lqc_infomtv <- map(ss_afm_pf7M_lqc, ~snp_rm_dblt_infomtv(annot_snp_qc = .x))

```


```{r}
# pf_chr_size = read_delim('../../Pf3D7_genomes_gtfs_indexs/PlasmoDB-52_Pfalciparum3D7_Genome.size', col_names = c('chrom', 'size')) %>% mutate(chr = str_remove_all(chrom,'Pf3D7_0|Pf3D7_|_v3')) 
```

```{r}
## Get genome size like this cut -f1,2 fasta.fai > genome.size
pf_chr_size = read_delim(chrm_sz, col_names = c('chrom', 'size')) %>% mutate(chr = str_remove_all(chrom,chrm_prfx)) 
```

IBS megacells from Mi

#------------------------ 2. compute IBS for megaCells -----------------------------------------#####

#------------------------------------ #
# Compute ibs for pseudo 'mega' cells #
#------------------------------------ #


```{r}

gtypes_chr_cln<- function(snp_qc_fnl ){
  gtypes_chr_pos_pos <- snp_qc_fnl %>% dplyr::select(chrom, pos, groups, qc2, alleles) %>% pivot_wider(names_from = groups, values_from = qc2) %>% unite(., col="id", c(chrom, pos), sep = "_", remove = F) %>% mutate(chr = str_remove_all(chrom, chrm_prfx))
}
```


```{r}

ss_afm_pf7M_cln_gt<- map(ss_afm_pf7M_qc_infomtv, ~gtypes_chr_cln(snp_qc_fnl = .x[[1]]))
```


```{r}

saveRDS(ss_afm_pf7M_cln_gt, paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_cln_gt.RDS"))
```


```{r}
ss_afm_pf7M_cln_gt_lqc <- map(ss_afm_pf7M_lqc_infomtv, ~gtypes_chr_cln(snp_qc_fnl = .x[[1]]))
```


```{r}

saveRDS(ss_afm_pf7M_cln_gt_lqc, paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_cln_gt_lqc.RDS"))
```



For the verification of souporcell K assignment among several Ks proceed to notebook 'm2_strain_deconv_K_verification_all.Rmd' after running this
For plotting of final qcd K and set of genotypes proceed to notebook 'm2_strain_postQC_decon_IBD.Rmd' after running this


