---
title: "scripts for final QCd cells strain deconvolution"
output: html_document
date: "2024-03-12"
---


```{r, message=FALSE}
suppressPackageStartupMessages({
  # library(Seurat)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  library(pheatmap)
  # library(rmarkdown)
  # library('vcfR')
  library(conflicted)
  # library(RColorBrewer)
  # library(MetBrewer)
  library(ComplexHeatmap)
  library(SNPRelate)
  library(SeqArray)
  # library(ggrepel)
  # library(ggvenn)
  # library(magick)
  # library(svglite)
  # library(plotly)
  # library(ggsankey)
  # library(gganimate)
  # library(av)
  library(dcifer)
  
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("count", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("saveRDS", "base")
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::union)
})

```


# MSC parasites strain deconvolution and VCF visualization 

## setup

```{r setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/')
```

## Variable declarations

### Common functions

#### Source commong functions, varriables and tables
```{r}
## Get the necessary variables for common housekeeping
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```


```{r}
## Get the necessary variable for the pseudobulking workflow
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_postQC_cln_header.R")
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_preQC_cln_header.R")
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_1vcf_header.R")
source("mali22_code_lnk/local_rmd_nbooks/m2_pbulk_gtypes_headers/m2_pbulk_gtypes_preQC_1VCF_final_header.R")

```

```{r}
# sample_set = 'Mali2'
# algn_nm = 'minmap'

# # soup_dir = "soupc"
# # sv_dir = "pbulk_gtypes_preQC"
# 
# # soup_dir = "soupc_GE_postQC"
# sv_dir = "pbulk_gtypes_postQC_cln"
# 
# strain_levels_univrs <- c(paste0("SC", 1:20), "Doublet", "Negative")
```

## Declare variables

```{r, eval=FALSE}
##Copy souporcell files for the optimum clusters
# 
# ##Sample ID
# # sample_name_nms <- c("MSC33", "MSC48","MSC49", "MSC55", "MSC60", "MSC66") %>% sort()
# ## All sample ids removing "MSC38" which seems to have lots of doublets
# sample_name_nms <- c("MSC33", "MSC48","MSC49", "MSC55", "MSC60", "MSC66", "MSC38", "MSC40", "MSC50_MACS", "MSC53", "MSC57", "MSC24", "MSC37", "MSC39", "MSC45", "MSC50_SLO", "MSC54", "MSC67", "MSC68", "MSC70")
# 
# # sample_name_nms <- c("MSC33", "MSC48","MSC49", "MSC55", "MSC60") %>% sort()
# # algn_nms = c('minmap', 'hsat', 'lr_ref')
# algn_nms = c('minmap', 'hsat')
# 
# sample_name_nms_al <- rep(sample_name_nms, length(algn_nms))
# algn_nms_al <- rep(algn_nms, each= length(sample_name_nms))
# 
# sm_aln_nms_al <- paste(rep(sample_name_nms, length(algn_nms)), rep(algn_nms, each= length(sample_name_nms)), sep ='_')
# 
# ## optimum QCd ks
# opt_clusters_nms <- list("MSC33" = c(6:10),"MSC48" = c(3:7),"MSC49" = c(5:9),"MSC55" = c(4:8),"MSC60" = c(5:9), "MSC66" = c(7:11), "MSC38" = c(2:6), "MSC40" = c(6:10), "MSC50_MACS" = c(2:6), "MSC53" = c(1:3), "MSC57" = c(1:3), "MSC24" = c(1:3), "MSC37" = c(5:9), "MSC39" = c(3:7), "MSC45" = c(1:3), "MSC50_SLO" = c(2:6), "MSC54" = c(1:3), "MSC67" = c(1:3), "MSC68" = c(2:6), "MSC70" = c(1:3))
# 
# 
# # optimum_k <- c("MSC24" = 1, "MSC33" = 8, "MSC37" = 5, "MSC38" = 4, "MSC39" = 4, "MSC40" = 8, "MSC45" = 1, "MSC48" = 4, 
# # "MSC49" = 6, "MSC50_MACS" = 1, "MSC50_SLO" = 1, "MSC53" = 1, "MSC54" = 1, "MSC55" = 7, "MSC57" = 1, 
# # "MSC60" = 7, "MSC66" = 8, "MSC67" = 1, "MSC68" = 3, "MSC70" = 1)
# 
# optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)
# 
# optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]
# 
# # opt_clusters_nms <- list("MSC33" = c(6:10),"MSC48" = c(3:7),"MSC49" = c(5:9),"MSC55" = c(4:8),"MSC60" = c(5:9))
# # opt_clusters_nms <- list("MSC33" = c(3:9),"MSC48" = c(2:3),"MSC49" = c(2:6),"MSC55" = c(4:10),"MSC60" = c(2:7), "MSC66")
# 
# opt_clusters_len <- map(opt_clusters_nms, length)
# 
# ## number of optimum clusters to investigate per sample factoring the 2 alignments
# opt_clusters_nms_aln <- rep(opt_clusters_nms, length(algn_nms)) %>% set_names(sm_aln_nms_al)
# opt_clusters_len_aln <- map(opt_clusters_nms_aln, length)
# 
# ##Sample IDs
# # source_irods_nms <- list("5736STDY13782211","5736STDY13782212","5736STDY13782213","5736STDY13782214","5736STDY13782215", "5736STDY13437393") %>% set_names(sample_name_nms)
# # source_irods_nms <- list("5736STDY13782211","5736STDY13782212","5736STDY13782213","5736STDY13782214","5736STDY13782215") %>% set_names(sample_name_nms)
# 
# # source_irods <- map2(source_irods_nms, opt_clusters_len[names(source_irods_nms)], ~rep(.x, each = .y))
# 
# # source_irods <- unlist(flatten(source_irods))
# 
# ##Samples
# sample_name <- map2(sample_name_nms, opt_clusters_len[sample_name_nms], ~rep(.x, each = .y))
# 
# opt_clusters <- unlist(flatten(opt_clusters_nms[sample_name_nms]))
# 
# sample_name <- unlist(flatten(sample_name))
# 
# ## sorted donor_opt_cluster names
# sorted_names <- paste0(sample_name, "_", opt_clusters)
# 
# ## sorted factoring alignment
# sorted_names_al <-paste0(sample_name, "_",rep(algn_nms,each=length(sample_name)),"_", opt_clusters)
# 
# 
# iter = opt_clusters_len


```



Section to write the genotypes from the QCd cells into a files

```{r}
# ## Retain qcd donors troublesome samples - removing MSC38, MSC55 which are giving challenges in assigning strains
# sample_name_nms <- c("MSC33","MSC37","MSC38","MSC39","MSC40", "MSC48", "MSC49", "MSC55", "MSC60", "MSC66", "MSC68")
# # sample_challenges <- c("MSC37","MSC38","MSC39", "MSC55")
# sample_name_nms <- c("MSC33", "MSC48", "MSC49")
# 
# sample_name_nms <- sample_name_nms[!sample_name_nms %in% sample_challenges]

```

```{r}

## Read in necessary files
ss_afm_pf7M_qc <- readRDS(paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_qc.RDS")) %>% .[str_detect(names(.), paste(paste0(sample_name_nms_dir, ".st"), collapse = "|"))]
ss_afm_pf7M_cln_gt <- readRDS(paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_cln_gt.RDS")) %>% .[str_detect(names(.), paste(paste0(sample_name_nms_dir, ".st"), collapse = "|"))]
# saveRDS(ss_afm_pf7M_cln_gt, paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_cln_gt.RDS"))

```



```{r}
## !!!NOTE - For QCD genotypes this should not be necessary for negatives unless they are genuine negative that can't be collapsed with strain manually
## Remove doublets and negatives
# ss_afm_pf7M_qc <- map(ss_afm_pf7M_qc, ~.x %>% filter(!str_detect(groups, paste(c("Doublet","Negative", don_rm), collapse = "|"))) %>% mutate(across(groups, ~str_replace(., "MSC", "M"))))
ss_afm_pf7M_qc <- map(ss_afm_pf7M_qc, ~.x %>% filter(!str_detect(groups, paste(c("Doublet","Negative"), collapse = "|"))) %>% mutate(across(groups, ~str_replace(., "MSC", "M"))))

                      
                      
# ss_afm_pf7M_cln_gt <- map(ss_afm_pf7M_cln_gt, ~.x %>% select(-c(contains("Doublet"), contains("Negative"))))
# ss_afm_pf7M_cln_gt <- map(ss_afm_pf7M_cln_gt, ~.x[,!(str_detect(colnames(.x), paste(c("Doublet","Negative", don_rm), collapse = "|")))] %>% rename_all(., ~str_replace(., "MSC", "M")) )
ss_afm_pf7M_cln_gt <- map(ss_afm_pf7M_cln_gt, ~.x[,!(str_detect(colnames(.x), paste(c("Doublet","Negative"), collapse = "|")))] %>% rename_all(., ~str_replace(., "MSC", "M")) )


```



```{r}
## !!!NOTE - REMOVE MSC55 - WE SHOULD REMOVE IT AT THE STRAIN ASSIGNMENT LEVEL OR 

# ss_afm_pf7M_qc <- map(ss_afm_pf7M_qc, ~.x %>% filter(!str_detect(groups, paste(c("M55"), collapse = "|")))) 
# 
# ss_afm_pf7M_cln_gt <- map(ss_afm_pf7M_cln_gt, ~.x[,!(str_detect(colnames(.x), paste(c("M55"), collapse = "|")))])


```


```{r}
# don_k1 <- opt_narrow_clusters_nms[opt_narrow_clusters_nms == 1]
```


```{r}
# map(ss_afm_pf7M_cln_gt, dim) %>% data.frame() %>% select(contains(".strain")) %>% rename_all(., ~str_remove(., ".s.*")) %>% .[1,] %>% bind_rows(., data.frame(opt_narrow_clusters_nms)) %>% t() %>% data.frame() %>% rownames_to_column("donor") %>% rename("SNP_No"=X1, "K"=X2) %>% filter(!is.na(K)) %>% mutate(monoclonal = case_when(K == 1 ~ "Yes", K>1 ~ "No")) %>% ggplot(aes(x = monoclonal, y = SNP_No)) + geom_point()

```

```{r}
# map(ss_afm_pf7M_cln_gt[str_detect(names(ss_afm_pf7M_cln_gt), paste0(names(don_k1), collapse = "|"))], dim)

```

Needs to go to common scripts
```{r}

##Function to save tables in excel as sheets

createWorksheet <- function(x, y){
  wb <- createWorkbook()
  for(i in seq_along(x)){
    addWorksheet(wb, names(x)[i])
    writeData(wb, sheet = names(x)[i], x[[i]])
    saveWorkbook(wb, file = y, overwrite = TRUE)
  }
}

```



Export the SNPs with QC metrics and only clean SNPs for stage&strain combinations and just strain
```{r, eval=F}
##!!!! NOTE - MANUSCRIPT SNPS table

## subset SNP tables to write and rename lists/sheets 
### QC
snps_export_qc <- ss_afm_pf7M_qc[str_detect(names(ss_afm_pf7M_qc), "stage_ag|\\.strain_qcd")] %>% map(., ~rename_with(., ~paste0('Pf7_',.), c("filter","type","VQSLOD","CDS","AC","AN","AF")))
names(snps_export_qc) <- paste0(10:(length(snps_export_qc)+9), ".", str_replace_all(names(snps_export_qc), c("\\.stage_ag_strain_qcd" = " stage_strain QC", "\\.strain_qcd" = " strain QC", "MSC" = "MSC")))

### Clean
snps_export_cln <- ss_afm_pf7M_cln_gt[str_detect(names(ss_afm_pf7M_cln_gt), "stage_ag|\\.strain_qcd")]  %>% map(., ~dplyr::select(.x, -c(chr))) 
names(snps_export_cln) <- paste0(2:(length(snps_export_cln)+1), ".", str_replace_all(names(snps_export_cln), c("\\.stage_ag_strain_qcd" = " stage_strain clean", "\\.strain_qcd" = " strain clean", "MSC" = "MSC")))

createWorksheet(c(snps_export_cln, snps_export_qc), paste0("../",sample_set,"/v3_science_manuscript/tables/data S9 - Strain & stage-strain SNPs including QC categories.xlsx"))
```


Sections containing IBD and other stuff for QCd strains
!!! NOTE MAKE DYNAMIC
```{r, fig.height=4.1, fig.width=4.5, eval=T}
## All genotypes done separately since when done together the error for DUPLICATE colnames appears
# ss_cln_gt_tbl_afm <- ss_afm_pf7M_cln_gt[c(seq(1,12,3))] %>%
#     imap(~ rename_with(.x, function(x) paste(.y, x, sep = "_"), -c(id,chrom, pos, alleles, chr))) %>%
#     purrr::reduce(left_join, by = c('id','chrom', 'pos', 'alleles', 'chr')) %>% 
#   rename_with(~str_remove(., '\\.st.*ain'), starts_with(c("MSC")) )

# ss_cln_gt_tbl_afm<-ss_cln_gt_tbl_afm[,sort(colnames(ss_cln_gt_tbl_afm))]

ss_cln_gt_tbl_ag <- ss_afm_pf7M_cln_gt[[2]] %>%
#     imap(~ rename_with(.x, function(x) paste(.y, x, sep = "_"), -c(id,chrom, pos, alleles, chr))) %>%
#     purrr::reduce(left_join, by = c('id','chrom', 'pos', 'alleles', 'chr')) %>%
#   rename_with(~str_remove_all(., '\\.st.*ain|k[0-9]+_'), starts_with(c("MSC")) )%>%
  rename_with(~str_replace_all(., c('MSC'='M', '\\.'='_')), starts_with(c("MSC")) )
# 
# ss_cln_gt_tbl_ag <- ss_cln_gt_tbl_ag[,sort(colnames(ss_cln_gt_tbl_ag))]

ss_cln_gt_tbl_s <- ss_afm_pf7M_cln_gt[[1]] %>%
    # imap(~ rename_with(.x, function(x) paste(.y, x, sep = "_"), -c(id,chrom, pos, alleles, chr))) %>%
    # purrr::reduce(left_join, by = c('id','chrom', 'pos', 'alleles', 'chr')) %>% 
  # rename_with(~str_remove_all(., '\\.st.*ain|k[0-9]+_'), starts_with(c("MSC")) )%>% 
  rename_with(~str_replace_all(., c('MSC'='M', '\\.'='_')), starts_with(c("MSC")) )

ss_cln_gt_tbl_s <- ss_cln_gt_tbl_s[,sort(colnames(ss_cln_gt_tbl_s))]

## 
# ss_cln_all_mtx <- list(ss_cln_gt_tbl_afm, ss_cln_gt_tbl_ag, ss_cln_gt_tbl_s) %>% set_names(c("all.stage_afm_strain_qcd", "all.stage_afm_strain_qcd", "all.strain_qcd"))
# ss_cln_all_mtx <- list(ss_cln_gt_tbl_ag, ss_cln_gt_tbl_s) %>% set_names(c("all.stage_ag_strain_qcd", "all.strain_qcd"))
# ss_cln_all_mtx <- list( ss_cln_gt_tbl_s) %>% set_names(c( "all.strain_qcd"))


```


Subset the asexuals and gametocytes into separate object
```{r, fig.height=18, fig.width=18}
## get standard columns needed for all tables
std_cols <- colnames(ss_cln_gt_tbl_ag)[!str_detect(colnames(ss_cln_gt_tbl_ag), "_A|_G")]

## get asexuals and put "chr" at the end
c(std_cols[!std_cols == "chr"], colnames(ss_cln_gt_tbl_ag)[str_detect(colnames(ss_cln_gt_tbl_ag), "_A")], "chr")
ss_cln_gt_tbl_a <- ss_cln_gt_tbl_ag[,c(std_cols[!std_cols == "chr"], colnames(ss_cln_gt_tbl_ag)[str_detect(colnames(ss_cln_gt_tbl_ag), "_A")], "chr")]

## get gams
ss_cln_gt_tbl_g <- ss_cln_gt_tbl_ag[,c(std_cols[!std_cols == "chr"], colnames(ss_cln_gt_tbl_ag)[str_detect(colnames(ss_cln_gt_tbl_ag), "_G")], "chr")]

```

Add the above asexual and gametocytes to the list
```{r, fig.height=18, fig.width=18}
## mtx
ss_cln_all_mtx <- list(ss_cln_gt_tbl_ag, ss_cln_gt_tbl_s, ss_cln_gt_tbl_a, ss_cln_gt_tbl_g) %>% set_names(c("all.stage_ag_strain_qcd", "all.strain_qcd", "all.stage_a_strain_qcd", "all.stage_g_strain_qcd"))

```

```{r, fig.height=4.1, fig.width=4.5}
## gtype for clusters with more than 100
# strn_c_new_mdata <- readRDS("data/processed/Pf/strn_c_new_mdata_m21.RDS")
# # strn_c_new_mdata <- readRDS("data/processed/Pf/strn_c_new_mdata_m21.RDS")
# 
# clstr_200 <- map(strn_c_new_mdata, . %>% 
#       filter(cluster_ag_n >= 15) %>%
#       dplyr::count(Strain_qc, stage_ag) %>% 
#       filter(stage_ag == 'Gametocyte') %>%
#       filter(Strain_qc != 'PoorQC') %>%
#       unite(col = 'cluster',Strain_qc, stage_ag, sep = '_') %>%
#       pull(cluster) %>%
#       str_remove_all(., 'ametocyte|sexual')
# ) %>% imap(., ~paste(.x, .y, sep='_')) %>% unlist() %>% as.vector()
# 
# clstr_200_no55 <- map(strn_c_new_mdata, . %>% 
#       filter(cluster_ag_n >= 15) %>%
#       dplyr::count(Strain_qc, stage_ag) %>% 
#       filter(stage_ag == 'Gametocyte') %>%
#       filter(Strain_qc != 'PoorQC') %>%
#       unite(col = 'cluster',Strain_qc, stage_ag, sep = '_') %>%
#       pull(cluster) %>%
#       str_remove_all(., 'ametocyte|sexual')
# ) %>% .[names(.) != 'MSC55'] %>% imap(., ~paste(.x, .y, sep='_')) %>% unlist() %>% as.vector()
# 
# ss_cln_gt_tbl_100 <- ss_cln_gt_tbl[c("id","chrom" ,"pos","alleles","chr",clstr_200_no55)]

```



```{r, warning=FALSE, eval=T}
# prev_sv_dir = 
# strn_c_new_mdata_m21 <- readRDS(paste0("../Mali/data/processed/DB52e_star/strn_c_new_mdata_m21.RDS"))
```

```{r, warning=FALSE, eval=T}
# prev_sv_dir = 
# strn_c_new_mdata_m22_min_qcd <- readRDS(paste0("data/processed/",species, "/" , gtyp_step, "/strn_c_new_mdata_m22_min_qcd.RDS"))
# msc_w_cln_strns_mdata <- readRDS("data/processed/Pf/m2_all/msc_w_cln_strns_mdata.RDS")
msc_w_cln_strns_mdata <- readRDS("data/processed/Pf/m2_all/msc_w_cln_strns_mdata.RDS")

msc_w_cln_strns_mdata_strains <- msc_w_cln_strns_mdata %>% 
  # mutate(stage_ag = ifelse(stage_c == "Asexual", "Asexual", "Gametocytes"),
  #        StrainQCd_Man = strain)  %>% 
  mutate(stage_ag = ifelse(stage %in% c("Early ring", "Late ring", "Early trophozoite"), "Asexual", "Gametocytes"),
         StrainQCd_Man = Strain_DN,
         StrainDon = paste0(str_remove(donor, "SC"),"_",Strain_DN))  %>% 
  select(StrainQCd_Man, StrainDon, stage_ag, donor)
```


```{r, warning=FALSE, eval=T}
# get symptomatic data
# m21_pf_status <- irods_id_sk21_mdata_cln %>% mutate(Status = ifelse(Symptomatic == "Yes", "Symptomatic", "Asymptomatic")) %>% rename("donor"=sample_nm) %>% select(donor, Status) 
# all_pf_status <- m21_pf_status %>%
#   bind_rows(., data.frame("donor" = c("MSC1", "MSC3", "MSC13", "MSC14"), "Status" = rep("Asymptomatic", 4))) %>%
#   bind_rows(., data.frame("donor" = c("MSC25", "MSC41", "MSC51"), "Status" = rep("Symptomatic", 3))) %>%
#   mutate(across(donor, ~str_remove(., "SC")))
```


```{r, fig.width=15, fig.height=5}
# get cleaned symptomatic data
donor_mdata_slct <- readRDS("data/processed/Pf/m2_all/donor_mdata_slct.RDS")

## Create annotation DF for donor and status and add the SLO and MACs information for MSC50 since this is the same donor an only represented for one entry in the metadata file
all_pf_status <- donor_mdata_slct %>%
  mutate("Status" = condition,
         "donor" = str_remove(short_name, "SC"),
         across(donor, ~case_when(. == "M50" ~ paste0(., "_SLO"),
                                  T ~ .))) %>%
  select(donor, short_name, Status)

## Add MSC50 MACs column
all_pf_status <- bind_rows(all_pf_status,
          c("M50_MACS", "MSC50", "Symptomatic") %>% 
            set_names(colnames(all_pf_status))) %>%
  select(donor, Status)

```

```{r}
## Read in msc annotations 

# strn_c_new_mdata_m22_min <- readRDS('data/processed/Pf/strn_c_new_mdata_m22_min_qcd.RDS')
```




```{r}
# ## Read in msc annotations 
# # clustr_numbers_afm <- map(msc_qcd_anotd_stgstrqc[1:4], ~.x@meta.data %>% 
# #         filter(strain_ag_n >= 30 & !is.na(strain_qcd)) %>%
# #         distinct(strain_qcd,stage_afm,strain_afm_n) %>% 
# #         mutate(across(stage_afm, ~str_remove_all(., 'emale|ale|sexual|ametocyte'))) %>%
# #         unite(col = 'cluster', strain_qcd, stage_afm, sep = '_'))  %>% 
# #     bind_rows(., .id='sample') %>%
# #   mutate(across(sample, ~str_remove_all(., 'sc'))) %>%
# #     unite(col = 'sample_cluster', sample, cluster, sep = '_')%>%
# #   mutate(across(sample_cluster, ~str_replace(., "MSC", "M")))
# # 
# # rownames(clustr_numbers_afm) <- NULL
# 
# clustr_numbers_ag <- map(strn_c_new_mdata_m22_min_qcd, ~.x %>% 
#                            add_count(stage_ag, StrainQCd_Man, name = "strain_ag_n") %>%
#         filter(strain_ag_n >= 20 & !is.na(StrainQCd_Man)) %>%
#         distinct(StrainQCd_Man,stage_ag,strain_ag_n) %>% 
#         mutate(across(stage_ag, ~str_remove_all(., 'emale|ale|sexual|ametocyte'))) %>%
#         unite(col = 'cluster', StrainQCd_Man, stage_ag, sep = '_'))  %>% 
#     bind_rows(., .id='sample') %>%
#   mutate(across(sample, ~str_remove_all(., 'SC'))) %>%
#     unite(col = 'sample_cluster', sample, cluster, sep = '_')#%>%
#   #mutate(across(sample_cluster, ~str_replace(., "MSC", "M")))
# 
# rownames(clustr_numbers_ag) <- NULL

## This will replace the above once the donor labels msc are capitalized
# clustr_numbers_s <- map(strn_c_new_mdata_m22_min_qcd, ~.x %>% 
#                            add_count(stage_ag, StrainQCd_Man, name = "strain_ag_n") %>%
#         filter(strain_ag_n >= 20 & !is.na(StrainQCd_Man)) %>%
#           add_count(StrainQCd_Man, name = "strain_n") %>% 
#         distinct(StrainQCd_Man, strain_n) %>% 
#         unite(col = 'cluster', StrainQCd_Man, sep = '_'))  %>% 
#     bind_rows(., .id='sample') %>%
#   mutate(across(sample, ~str_remove_all(., 'SC')))%>% 
#     unite(col = 'sample_cluster', sample, cluster, sep = '_')

clustr_numbers_s <- msc_w_cln_strns_mdata_strains %>% 
                        group_by(donor) %>%
                           add_count(stage_ag, StrainQCd_Man, name = "strain_ag_n") %>%
        filter(strain_ag_n >= 20 & !is.na(StrainQCd_Man)) %>%
          add_count(StrainQCd_Man, name = "strain_n") %>% 
        distinct(donor, StrainQCd_Man, strain_n) %>% 
        unite(col = 'cluster', StrainQCd_Man, sep = '_')  %>% 
    mutate('sample'=donor) %>%
  ungroup() %>%
  select(-c(donor)) %>%
  mutate(across(sample, ~str_remove_all(., 'SC')))%>% 
    unite(col = 'sample_cluster', sample, cluster, sep = '_')

clustr_numbers_ag <- msc_w_cln_strns_mdata_strains %>% 
                        group_by(donor) %>%
                        add_count(stage_ag, StrainQCd_Man, name = "strain_ag_n") %>%
        filter(strain_ag_n >= 20 & !is.na(StrainQCd_Man)) %>%
        distinct(donor, StrainQCd_Man,stage_ag,strain_ag_n) %>%
        mutate(across(stage_ag, ~str_remove_all(., 'emale|ale|sexual|ametocytes'))) %>%
        unite(col = 'cluster', StrainQCd_Man, stage_ag, sep = '_')  %>%
    mutate('sample'=donor) %>%
  ungroup() %>%
  select(-c(donor)) %>%
  mutate(across(sample, ~str_remove_all(., 'SC'))) %>%
    unite(col = 'sample_cluster', sample, cluster, sep = '_')

# clustr_numbers_ag <- map(strn_c_new_mdata_m22_min_qcd, ~.x %>%
#                            add_count(stage_ag, StrainQCd_Man, name = "strain_ag_n") %>%
#         filter(strain_ag_n >= 20 & !is.na(StrainQCd_Man)) %>%
#         distinct(StrainQCd_Man,stage_ag,strain_ag_n) %>%
#         mutate(across(stage_ag, ~str_remove_all(., 'emale|ale|sexual|ametocyte'))) %>%
#         unite(col = 'cluster', StrainQCd_Man, stage_ag, sep = '_'))  %>%
#     bind_rows(., .id='sample') %>%
#   mutate(across(sample, ~str_remove_all(., 'SC'))) %>%
#     unite(col = 'sample_cluster', sample, cluster, sep = '_')#%>%
#   #mutate(across(sample_cluster, ~str_replace(., "MSC", "M")))
# 
# # rownames(clustr_numbers_ag) <- NULL
```


```{r}
## Read in msc annotations 

# msc_qcd_anotd_stgstrqc <- readRDS('../Mali/data/processed/DB52e_star/msc_qcd_anotd.RDS')
```


```{r}
# ## Read in msc annotations 
# clustr_numbers_ag_m1 <- map(msc_qcd_anotd_stgstrqc[1:4], ~.x@meta.data %>% 
#         filter(strain_ag_n >= 20 & !is.na(strain_qcd)) %>%
#         distinct(strain_qcd,stage_ag,strain_ag_n) %>% 
#         mutate(across(stage_ag, ~str_remove_all(., 'emale|ale|sexual|ametocyte'))) %>%
#         unite(col = 'cluster', strain_qcd, stage_ag, sep = '_'))  %>% 
#     bind_rows(., .id='sample') %>%
#   mutate(across(sample, ~str_remove_all(., 'sc'))) %>%
#     unite(col = 'sample_cluster', sample, cluster, sep = '_')%>%
#   mutate(across(sample_cluster, ~str_replace(., "MSC", "M")))
# 
# rownames(clustr_numbers_ag_m1) <- NULL

## This will replace the above once the donor labels msc are capitalized
# clustr_numbers_s_cap_m1 <- map(msc_qcd_anotd_stgstrqc[1:4], ~.x@meta.data %>% 
#         filter(strain_ag_n >= 20 & !is.na(strain_qcd)) %>%
#           add_count(strain_qcd, name = "strain_n") %>% 
#         distinct(strain_qcd, strain_n) %>% 
#         unite(col = 'cluster', strain_qcd, sep = '_'))  %>% 
#     bind_rows(., .id='sample') %>%
#   mutate(across(sample, ~str_remove_all(., 'sc')))%>% 
#     unite(col = 'sample_cluster', sample, cluster, sep = '_')%>%
#   mutate(across(sample_cluster, ~str_replace(., "MSC", "M")))
# 
# rownames(clustr_numbers_s_cap_m1) <- NULL
```

```{r}
## Read in msc annotations 
# clustr_numbers_s_m1_m2 <- bind_rows(clustr_numbers_s, clustr_numbers_s_cap_m1)
clustr_numbers_s_m1_m2 <- clustr_numbers_s

# clustr_numbers_ag_m1_m2 <- bind_rows(clustr_numbers_ag, clustr_numbers_ag_m1)
clustr_numbers_ag_m1_m2 <- clustr_numbers_ag
clustr_numbers_a_m1_m2 <- clustr_numbers_ag %>% filter(str_detect(sample_cluster, "_A" ))
clustr_numbers_g_m1_m2 <- clustr_numbers_ag %>% filter(str_detect(sample_cluster, "_G" ))
```

Remove strain(/stage) groups less than 50
```{r}
## Read in msc annotations 
# clustr_low_n <- map2(list(clustr_numbers_ag_m1_m2, clustr_numbers_s_m1_m2), list("strain_ag_n", "strain_n"), ~.x %>% filter(str_detect(sample_cluster, "Doublet|Negative") | !!rlang::sym(.y) < 60) %>% pull(sample_cluster))
# clustr_low_n <- pmap(list(list(clustr_numbers_ag_m1_m2, clustr_numbers_s_m1_m2, clustr_numbers_ag_m1_m2, clustr_numbers_ag_m1_m2), 
clustr_low_n <- pmap(list(list(clustr_numbers_ag_m1_m2, clustr_numbers_s_m1_m2, clustr_numbers_a_m1_m2, clustr_numbers_g_m1_m2), 
                          c("strain_ag_n", "strain_n", "strain_ag_n", "strain_ag_n"),
                          # c(40, 40, 40, 21)), ~{
                          c(100, 100, 100, 100)), ~{
  ..1 %>% 
    filter(str_detect(sample_cluster, "Doublet|Negative") | !!rlang::sym(..2) < ..3) %>% 
    pull(sample_cluster)
  })

```

```{r}
clustr_low_n

```


```{r, fig.height=18, fig.width=18}
## !!!NOTE Caused by error in `str_detect()`
ss_cln_all_high_n_mtx <- map2(ss_cln_all_mtx, clustr_low_n, ~.x[,!(str_detect(colnames(.x), paste(.y, collapse = "|")))])
```


```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_chr_cln_ibs_mtx <- function(gtypes_chr_pos){
  
  # ##Remove SC8 and SC1_F
  # gtypes_chr_pos <- snp_qc_tbl %>% filter(!(groups %in% c("SC8_F", "SC1_F")))%>% dplyr::select(chrom, pos, groups, qc2, alleles) %>% pivot_wider(names_from = groups, values_from = qc2) %>% unite(., col="id", c(chrom, pos), sep = "_", remove = F) %>% mutate(chr = str_remove_all(chrom, 'Pf3D7_0|Pf3D7_|_v3'))
  
  gtMat = gtypes_chr_pos %>% dplyr::select(-c(alleles,chrom, pos, chr))  %>% column_to_rownames("id") %>% mutate(across(everything(), ~as.numeric(str_replace(., '1', '2'))))%>% as.matrix() 
  gtMat = gtMat[,sort(colnames(gtMat))]
  
  
  gdsFile = tempfile()
    
  snpgdsCreateGeno(gdsFile, genmat = gtMat, sample.id = colnames(gtMat), 
                     snp.id = rownames(gtMat), snp.chromosome = gtypes_chr_pos$chr, 
                     snp.position = gtypes_chr_pos$pos, snp.allele = as.character(gtypes_chr_pos$alleles), snpfirstdim = TRUE)  
  
  gds = snpgdsOpen(gdsFile)
  
  gdsSub = snpgdsLDpruning(gds, ld.threshold = 0.2)
  
  ibs = snpgdsIBS(gds, num.thread = 4)
  
  rownames(ibs$ibs) = colnames(ibs$ibs) = colnames(gtMat)
    
  mat<-ibs$ibs
   
}
```

Multisample heatmap

```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_ibs_heatmap <- function(mtx, dendo = F){
  
  colFun = suppressWarnings(brewer.pal(100, 'Greens'))
  colFun = circlize::colorRamp2(seq(0.5, 1, length.out = length(colFun)), 
                                  colFun)
   print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
  fa = str_extract(colnames(mtx), 'MSC.*')
  print(str(fa))
  
  # c_d = cluster_between_groups(mtx, fa)
  # dendog = if(dendo == T) c_d else F
  dendog = if(dendo == T) cluster_between_groups(mtx, fa) else F
  print(str(fa))
  print(str(dendog))
  
  # fa = str_extract(colnames(mtx), 'MSC.*')
  # # fa_col = c("a" = 2, "b" = 3, "c" = 4)
  # dend1 = cluster_between_groups(mtx, fa)
  

  hm = Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog)
  
  
  hm2<-Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), border='black',   heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
                  layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(mtx, i, j)
                      l = v >0.98
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "#a4c2f4", lwd = 2, fill = "transparent"))
})
  
  print(hm)
  print(hm2)
  # print(hm3)
}
```


```{r, fig.height=15, fig.width=20}

# gt_all_ibs_mtx<- gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=ss_cln_gt_tbl)
# gt_all_ibs_hmap<- gtypes_ibs_heatmap(mtx=gt_all_ibs_mtx)

gt_all_ibs_mtx<- map(ss_cln_all_high_n_mtx, ~gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=.x))
gt_all_ibs_hmap<- map(gt_all_ibs_mtx, ~gtypes_ibs_heatmap(mtx=.x))


# gt_all_ibs_100_mtx<- gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=ss_cln_gt_tbl_100)
# gt_all_ibs_100_hmap<- gtypes_ibs_heatmap(mtx=gt_all_ibs_100_mtx, dendo = T)

```


```{r, fig.height=8, fig.width=10.5}
##Merge strain numbers to IBD results
ibs_don_hm <- function(ibsmtx, clustr_numbers, strn_cat, start_no = 0.5, test_nm = "IBS"){
colFun = suppressWarnings(brewer.pal(100, 'Greens'))
colFun = circlize::colorRamp2(seq(start_no, 1, length.out = length(colFun)), 
                                  colFun)
  
sibd_pht_mtx <- ibsmtx
print(colnames(sibd_pht_mtx))
sibd_pht_mtx <- sibd_pht_mtx[,row.names(sibd_pht_mtx)]

# fa = str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]')
# fa = str_remove(colnames(sibd_pht_mtx), '_qcd.*')
fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
c_d = cluster_within_group(sibd_pht_mtx, fa)

print(colnames(sibd_pht_mtx))
print(fa)
# fr = str_extract(rownames(sibd_pht_mtx), 'ms.*[0-9]')
# fr = str_remove(rownames(sibd_pht_mtx), '_qcd.*')
fr = str_remove(rownames(sibd_pht_mtx), '_SC.*')
# print(fr)
c_r = cluster_within_group(sibd_pht_mtx, fr)

fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove(colnames(sibd_pht_mtx), '_SC.*')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")


# nms = unique(str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]'))
nms = unique(str_remove(colnames(sibd_pht_mtx), '_qcd.*'))
# sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
sample_c_cols <- donor_cols

# sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% 
  #mutate(across(sid, ~str_replace_all(., c('MSC' = 'M', '_qcd'='')))) %>% rename_with(~str_replace_all(., c('MSC' = 'M', '_qcd'='')), starts_with('MSC')) %>% 
  left_join(., clustr_numbers, by = c('sid'='sample_cluster')) %>% 
  mutate(across(sid, ~paste0(., ' [', !!rlang::sym(strn_cat), ']'))) %>% select(-c(!!rlang::sym(strn_cat))) %>%
  column_to_rownames('sid') %>% 
  as.matrix()

print(head(sibd_pht_mtx_cln))
  
sibd_pht<- Heatmap(sibd_pht_mtx_cln, name=test_nm, 
                   col = colFun, 
                   show_row_names = TRUE, 
                   row_names_side = c("left"), 
                   show_column_names = TRUE,  
                   show_row_dend = T,  
                   show_column_dend = F, 
                   row_title_rot = 0, 
                   cluster_rows = c_r, 
                   clustering_method_rows = "complete",
                   cluster_columns = c_d, 
                   column_names_gp = gpar(fontsize = 15), 
                   row_names_gp = gpar(fontsize = 15), 
                   heatmap_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15)),
                   bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                 col = list(donor = sample_c_cols, Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ), 
                                                         annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                         annotation_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15))),
                   column_split = length(nms), 
                   border = TRUE, 
                   left_annotation=rowAnnotation(df = fa_df,
                                                 col = list(donor = sample_c_cols, Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ), 
                                                 annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                 show_legend = F), 
                   row_split = length(nms), 
                   column_title = NULL,
                   column_title_gp = gpar(fontsize = 16), 
                   row_title = NULL,
                   row_title_gp = gpar(fontsize = 16)
              # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
              ) %>%
  draw(., merge_legend = TRUE)

sibd_pht
}
```


```{r, fig.height=18, fig.width=18}

# hms<- pmap(list(gt_all_ibs_mtx, list(clustr_numbers_afm, clustr_numbers_ag, clustr_numbers_s), c("strain_afm_n", "strain_ag_n", "strain_n")), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))
hms<- pmap(list(gt_all_ibs_mtx, list( clustr_numbers_ag_m1_m2, clustr_numbers_s_m1_m2, clustr_numbers_ag_m1_m2, clustr_numbers_ag_m1_m2), c("strain_ag_n", "strain_n", "strain_ag_n", "strain_ag_n" )), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))
# hms<- pmap(list(gt_all_ibs_mtx, list( clustr_numbers_s_m1_m2), c("strain_n" )), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))


# hms<- pmap(list(gt_all_ibs_mtx, list(clustr_numbers_s_m1_m2), c("strain_n" )), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))

```

######## IBS for polyclonal with asexuals
```{r}
# sample_name_nms_polyc <- sample_name_nms[map(opt_narrow_clusters_nms, ~.x > 1) %>% unlist()] 
# sample_name_nms_monoc <- sample_name_nms[map(opt_narrow_clusters_nms, ~.x == 1) %>% unlist()] 

## strains with asexuals
# sample_name_nms_polyc_asx <- clustr_numbers_ag_m1_m2 %>% filter(!str_detect(sample_cluster, "Doublet|Negative") & strain_ag_n >= 60 & str_detect(sample_cluster, "_A") & !str_detect(sample_cluster, "SC0")) %>% pull(sample_cluster) %>% str_remove(., "_S.*") %>% unique()
sample_name_nms_polyc_asx <- clustr_numbers_ag_m1_m2 %>% filter(!str_detect(sample_cluster, "Doublet|Negative") & strain_ag_n >= 100 & str_detect(sample_cluster, "_A") & !str_detect(sample_cluster, "SC0")) %>% pull(sample_cluster) %>% str_remove(., "_S.*") %>% unique()

ss_cln_all_mtx_polyc_asx <- imap(ss_cln_all_high_n_mtx, ~.x %>% .[str_detect(names(.), paste(c(sample_name_nms_polyc_asx, "alleles", "chr", "chrom", "id", "pos") , collapse = "|"))])

## Write out for IBD
# imap(ss_cln_all_mtx_polyc_asx[2], ~.x %>% dplyr::select(-c(id, chrom, alleles))%>% dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% filter(!if_all(starts_with('M'), ~is.na(.))) %>% mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% write_tsv(., paste0("data/processed/Pf/ss_ibd_cln_", .y,"_mtx_polyc_asx.tsv")))
```


```{r, fig.height=18, fig.width=18}

gt_all_ibs_mtx_polyc_asx<- map(ss_cln_all_mtx_polyc_asx, ~gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=.x))

hms<- pmap(list(gt_all_ibs_mtx_polyc_asx, list( clustr_numbers_ag_m1_m2, clustr_numbers_s_m1_m2, clustr_numbers_ag_m1_m2, clustr_numbers_ag_m1_m2), c("strain_ag_n", "strain_n", "strain_ag_n", "strain_ag_n" )), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))
```



```{r, fig.height=4.5, fig.width=6}

```

```{r, fig.height=4.5, fig.width=6}

# hms<- pmap(list(gt_all_ibs_mtx[3], list(clustr_numbers_s), c("strain_n")), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))

```


```{r, fig.height=4.5, fig.width=6}
## Capitalize the donor labels - will need to sort out later for entire notebook
# cap_gt_all_ibs_high_n_mtx <- gt_all_ibs_mtx[[3]] %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('MSC' = 'MSC', '_qcd'='')))) %>% rename_with(~str_replace_all(., c('MSC' = 'MSC', '_qcd'='')), starts_with('MSC')) %>% column_to_rownames('sid') %>% as.matrix()

```

```{r, fig.height=10, fig.width=10, eval=FALSE}
##Merge strain numbers to IBD results

colFun = suppressWarnings(brewer.pal(100, 'Greens'))
colFun = circlize::colorRamp2(seq(0.5, 1, length.out = length(colFun)), 
                                  colFun)
  
sibd_pht_mtx <- cap_gt_all_ibs_mtx
print(colnames(sibd_pht_mtx))
sibd_pht_mtx <- sibd_pht_mtx[,row.names(sibd_pht_mtx)]

# fa = str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]')
fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
c_d = cluster_within_group(sibd_pht_mtx, fa)

# fr = str_extract(rownames(sibd_pht_mtx), 'ms.*[0-9]')
fr = str_remove(rownames(sibd_pht_mtx), '_SC.*')
print(fr)
c_r = cluster_within_group(sibd_pht_mtx, fr)

# nms = unique(str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]'))
nms = unique(str_remove(colnames(sibd_pht_mtx), '_SC.*'))
# sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
sample_c_cols <- donor_cols

row_labs <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(r_labs =str_remove(sid, "MS.*_")) %>% left_join(., clustr_numbers_s_cap, by = c('sid'='sample_cluster')) %>% mutate(across(r_labs, ~paste0(., ' [', !!rlang::sym("strain_n"), ']'))) %>% select(sid, r_labs) %>% deframe

col_labs <- str_remove(colnames(sibd_pht_mtx), "MS.*_") %>% set_names(colnames(sibd_pht_mtx))

  
sibd_pht<- Heatmap(sibd_pht_mtx, name="IBS", 
                   col = colFun, 
                   show_row_names = TRUE, 
                   row_labels = row_labs,
                   column_labels = col_labs,
                   row_names_side = c("left"), 
                   show_column_names = TRUE,  
                   show_row_dend = F,  
                   show_column_dend = F, 
                   row_title_rot = 0, 
                   cluster_rows = c_r, 
                   clustering_method_rows = "complete",
                   cluster_columns = c_d, 
                   column_names_gp = gpar(fontsize = 15), 
                   row_names_gp = gpar(fontsize = 15), 
                   heatmap_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15)),
                   bottom_annotation = HeatmapAnnotation(Donor = fa, 
                                                         col = list(Donor = sample_c_cols), 
                                                         annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                         annotation_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15))),
                   column_split = length(nms), 
                   border = TRUE, 
                   left_annotation=rowAnnotation(Donor = fa, col = list(Donor = sample_c_cols), annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), show_legend = F), row_split = length(nms), 
                   column_title = NULL,
                   column_title_gp = gpar(fontsize = 16), 
                   row_title = NULL,
                   row_title_gp = gpar(fontsize = 16)
              # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
              ) %>%
  draw(., merge_legend = TRUE)

sibd_pht

```

```{r, fig.width= 12, fig.height=7, eval=FALSE}
##save svg for sk21
save_plot('plots/mali2021_all/fig4/sibd_pht.svg', plot_grid(ggdraw(sibd_pht)), base_width = 4, base_height = 5)
```


```{r, fig.height=8, fig.width=10.5, eval=FALSE}
# ##Merge strain numbers to IBD results - needs IBD!!!!!
# colFun = suppressWarnings(brewer.pal(100, 'Greens'))
# colFun = circlize::colorRamp2(seq(0.5, 1, length.out = length(colFun)), 
#                                   colFun)
#   
# sibd_pht_mtx <- gt_all_ibs_mtx
# 
# sibd_pht_mtx <- sibd_pht_mtx[,row.names(sibd_pht_mtx)]
# 
# # fa = str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]')
# fa = str_remove(colnames(sibd_pht_mtx), '_qcd.*')
# c_d = cluster_within_group(sibd_pht_mtx, fa)
# 
# # fr = str_extract(rownames(sibd_pht_mtx), 'ms.*[0-9]')
# fr = str_remove(rownames(sibd_pht_mtx), '_qcd.*')
# print(fr)
# c_r = cluster_within_group(sibd_pht_mtx, fr)
# 
# # nms = unique(str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]'))
# nms = unique(str_remove(colnames(sibd_pht_mtx), '_qcd.*'))
# # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
# sample_c_cols <- donor_cols
# 
# # sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
# sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('MSC' = 'M', '_qcd'='')))) %>% rename_with(~str_replace_all(., c('MSC' = 'M', '_qcd'='')), starts_with('MSC')) %>% 
#   left_join(., clustr_numbers, by = c('sid'='sample_cluster')) %>% mutate(across(sid, ~paste0(., ' [', strain_afm_n, ']'))) %>% select(-c(strain_afm_n)) %>%
#   column_to_rownames('sid') %>% as.matrix()
# 
#   
# sibd_pht<- Heatmap(sibd_pht_mtx_cln, name='IBS', 
#                    col = colFun, 
#                    show_row_names = TRUE, 
#                    row_names_side = c("left"), 
#                    show_column_names = TRUE,  
#                    show_row_dend = T,  
#                    show_column_dend = F, 
#                    row_title_rot = 0, 
#                    cluster_rows = c_r, 
#                    clustering_method_rows = "complete",
#                    cluster_columns = c_d, 
#                    column_names_gp = gpar(fontsize = 15), 
#                    row_names_gp = gpar(fontsize = 15), 
#                    heatmap_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15)),
#                    bottom_annotation = HeatmapAnnotation(Donor = fa, 
#                                                          col = list(Donor = sample_c_cols), 
#                                                          annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
#                                                          annotation_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15))),
#                    column_split = length(nms), 
#                    border = TRUE, 
#                    left_annotation=rowAnnotation(Donor = fa, col = list(Donor = sample_c_cols), annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), show_legend = F), row_split = length(nms), 
#                    column_title = NULL,
#                    column_title_gp = gpar(fontsize = 16), 
#                    row_title = NULL,
#                    row_title_gp = gpar(fontsize = 16)
#               # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
#               ) %>%
#   draw(., merge_legend = TRUE)
# 
# sibd_pht

# Legend(labels = month.name[1:10], legend_gp = gpar(fill = 1:10), 
#     title = "foo", nrow = 3)
```



Read in MSC14 genotypes
```{r, fig.height=5, fig.width=6}
# m14_ss_ag_k8r_pf7M_cln_gt <- read_tsv("/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_ag_k8r_pf7M_cln_gt.tsv") %>% rename_with(~paste0(., "_msc14"), starts_with("SC"))
```


```{r}
##!!NOTE - gfv_p1_bi_strn_pf7_c is created somewhere downstream so this is likely to give an error

# ss_cln_gt_tbl_100 %>% dplyr::select(-c(id, chrom, alleles))%>% dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% filter(!if_all(starts_with('SC'), ~is.na(.))) %>% mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% write_tsv(., paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_cln_gt_tbl_100_m21.tsv"))
# 
# ##merge m14 with other donors
# ss_cln_gt_100_plus_m14 <-m14_ss_ag_k8r_pf7M_cln_gt %>% mutate(across(-pos, as.character)) %>% mutate(across(starts_with('SC'), ~na_if(.,"-1"))) %>% select(-c(contains("_A_"))) %>% full_join(.,ss_cln_gt_tbl_100, by=c('chr', 'pos')) %>% mutate(id= paste0(chr, "_", pos))
# 
# ss_cln_gt_100_plus_m14 %>% dplyr::select(-c(id, chrom, alleles))%>% dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% filter(!if_all(starts_with('SC'), ~is.na(.))) %>% mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% arrange(chr, pos) %>% write_tsv(., paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_cln_gt_tbl_100_m21_plus_m14.tsv"))

```


```{r}
#dcifer
## once in local send output back to farm
# library(dcifer)

# pardef <- par(no.readonly = TRUE)
# 
# sfile <- system.file("extdata", "MozParagon.csv", package = "dcifer")
# dset <- read_csv(sfile)
# dsmp <- readDat(sfile, svar = "sampleID", lvar = "locus", avar = "allele")
# str(dsmp, list.len = 2)
# 
# dsmp <- formatDat(dset, svar = "sampleID", lvar = "locus", avar = "allele")
# 
# lrank <- 2
# coi   <- getCOI(dsmp, lrank = lrank)
```


```{r}
# optionally, extract location information
# meta <- unique(read.csv(sfile)[c("sampleID", "province")])
# meta <- meta[match(names(dsmp), meta$sampleID), ]  # order samples as in dsmp
```


```{r}
# dsmp_pm <- readRDS( "data/processed/Pf/dsmp_pm.rds")
# afreq_pm <- readRDS( "data/processed/Pf/afreq_pm.rds")
# dcf_pm <- readRDS( "data/processed/Pf/dcf_pm.rds")
```

```{r, eval=T}
#dcifer
dcf_pm <- ss_cln_all_high_n_mtx[["all.strain_qcd"]] %>%
  pivot_longer(cols = starts_with("M"), values_to = "allele", names_to = "sampleID",  values_drop_na = T) %>%
  mutate(locus = paste0(chrom, "_", id)) %>%
  select(sampleID, locus, allele )

saveRDS(dcf_pm, "data/processed/Pf/dcf_pm.rds")
  
```

Run below in farm
```{r, eval=T}
gogogo()
## NOTE - Takes long so request an interactive session and run there

# dcifer
# module load HGI/softpack/groups/team222/Pf_scRNAseq/25
# bsub -Is -q normal -n 10 -M 15000 -R "span[hosts=1] select[mem>15000] rusage[mem=15000]" R


## bsub -Is -q normal -n 25 -M 150000 -R "span[hosts=1] select[mem>150000] rusage[mem=150000]" R


```

```{r, eval=F}

# library(dplyr)
# library(dcifer)
# 
# # setwd("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/")
# dcf_pm <- readRDS("data/processed/Pf/dcf_pm.rds")
# #
# dsmp_pm <- formatDat(dcf_pm, svar = "sampleID", lvar = "locus", avar = "allele")
# saveRDS(dsmp_pm, "data/processed/Pf/dsmp_pm.rds")
# #
# coi_pm  <- rep(1, dim(dcf_pm)[1]) %>% set_names(paste0(dcf_pm$sampleID,".", dcf_pm$locus))
# saveRDS(coi_pm, "data/processed/Pf/coi_pm.rds")
# #
# afreq_pm <- calcAfreq(dsmp_pm, coi_pm, tol = 1e-5)
# str(afreq_pm, list.len = 2)
# saveRDS(afreq_pm, "data/processed/Pf/afreq_pm.rds")
# #
# dres0_pm <- ibdDat(dsmp_pm, coi_pm, afreq_pm, pval = TRUE, confint = TRUE, rnull = 0,
#                 alpha = 0.05, nr = 1e3)
# saveRDS(dres0_pm, "data/processed/Pf/dres0_pm.rds")
```


```{r, eval=T}
#dcifer
dcf_pm_ag <- ss_cln_all_high_n_mtx[["all.stage_ag_strain_qcd"]] %>%
  pivot_longer(cols = starts_with("M"), values_to = "allele", names_to = "sampleID",  values_drop_na = T) %>%
  mutate(locus = paste0(chrom, "_", id)) %>%
  select(sampleID, locus, allele )

saveRDS(dcf_pm_ag, "data/processed/Pf/dcf_pm_ag.rds")
  
```

Run below in farm
```{r, eval=T}
gogogo()
## NOTE - Takes long so request an interactive session and run there

# dcifer
# module load HGI/softpack/groups/team222/Pf_scRNAseq/25
# bsub -Is -q normal -n 25 -M 150000 -R "span[hosts=1] select[mem>150000] rusage[mem=150000]" R


```

```{r, eval=F}

# library(dplyr)
# library(dcifer)
# library(purrr)
# 
# setwd("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/")
# dcf_pm_ag <- readRDS("data/processed/Pf/dcf_pm_ag.rds")
# 
# dsmp_pm_ag <- formatDat(dcf_pm_ag, svar = "sampleID", lvar = "locus", avar = "allele")
# saveRDS(dsmp_pm_ag, "data/processed/Pf/dsmp_pm_ag.rds")
# 
# coi_pm_ag  <- rep(1, dim(dcf_pm_ag)[1]) %>% set_names(paste0(dcf_pm_ag$sampleID,".", dcf_pm_ag$locus))
# saveRDS(coi_pm_ag, "data/processed/Pf/coi_pm_ag.rds")
# 
# afreq_pm_ag <- calcAfreq(dsmp_pm_ag, coi_pm_ag, tol = 1e-5)
# str(afreq_pm_ag, list.len = 2)
# saveRDS(afreq_pm_ag, "data/processed/Pf/afreq_pm_ag.rds")
# 
# dres0_pm_ag <- ibdDat(dsmp_pm_ag, coi_pm_ag, afreq_pm_ag, pval = TRUE, confint = TRUE, rnull = 0,
#                 alpha = 0.05, nr = 1e3)
# saveRDS(dres0_pm_ag, "data/processed/Pf/dres0_pm_ag.rds")
```


```{r, eval = F}
# optionally, extract location information
# meta <- data.frame("sampleID" = names(dsmp), "donor" = str_remove(names(dsmp), '_SC.*')) %>% 
#   left_join(all_pf_status, by = "donor")# %>% 
  # column_to_rownames("rname")

# meta <- meta[match(names(dsmp), meta$sampleID), ]  # order samples as in dsmp
```

```{r}
#dcifer
# dcf_pm <- readRDS("data/processed/Pf/dcf_pm.rds")
  
dsmp_pm <- readRDS("data/processed/Pf/dsmp_pm.rds")

# coi_pm <- readRDS("data/processed/Pf/coi_pm.rds")

# afreq_pm <- readRDS("data/processed/Pf/afreq_pm.rds")

dres0_pm <- readRDS("data/processed/Pf/dres0_pm.rds")
```


```{r}
# optionally, extract location information
meta <- data.frame("sampleID" = names(dsmp_pm), "donor" = str_remove(names(dsmp_pm), '_SC.*')) %>% 
  left_join(all_pf_status, by = "donor")# %>% 
  # column_to_rownames("rname")

# meta <- meta[match(names(dsmp_pm), meta$sampleID), ]  # order samples as in dsmp_pm
```


```{r}
pardef <- par(no.readonly = TRUE)
```

```{r, fig.height=6, fig.width = 6}
par(mar = c(3, 3, 1, 1))
alpha <- 0.05                          # significance level                    
dmat <- dres0_pm[, , "estimate"]
# create symmetric matrix
dmat[upper.tri(dmat)] <- t(dmat)[upper.tri(t(dmat))]  
# determine significant, reverse columns for upper triangle
isig <- which(dres0_pm[, , "p_value"] <= alpha, arr.ind = TRUE)[, 2:1] 
plotRel(dmat, isig = isig, draw_diag = TRUE, lwd_diag = 0.5, idlab = TRUE, 
        col_id = donor_cols[meta$donor]) 
```


```{r}
par(pardef)

```


```{r, fig.height=6, fig.width = 6}
nsmp  <- length(dsmp_pm)
```


```{r, fig.height=5, fig.width = 10}
par(mfrow = c(1, 2), mar = c(1, 0, 1, 0.2))
plotRel(dres0_pm, draw_diag = TRUE, alpha = alpha)
mtext("p-values", 3, 0.2)
# p-values for upper triangle
pmat <- matrix(NA, length(dsmp_pm), length(dsmp_pm))
pmat[upper.tri(pmat)] <- t(log(dres0_pm[, , "p_value"]))[upper.tri(pmat)]
pmat[pmat == -Inf] <- min(pmat[is.finite(pmat)])
plotRel(pmat, rlim = NULL, draw_diag = TRUE, col = hcl.colors(101, "Red-Purple"), 
        sig = FALSE, add = TRUE, col_diag = "white", border_diag = "gray45")
# abline(v = atsep, h = atsep, col = "gray45", lty = 5)

# number of non-missing loci for upper triangle
dmiss <- lapply(dsmp_pm, function(lst) sapply(lst, function(v) all(!v)))
nmat <- matrix(NA, nsmp, nsmp) 
for (jsmp in 2:nsmp) {
  for (ismp in 1:(jsmp - 1)) {
    nmat[ismp, jsmp] <- sum(!dmiss[[ismp]] & !dmiss[[jsmp]])
  }
}
nrng <- range(nmat, na.rm = TRUE)    
par(mar = c(1, 0.2, 1, 0))
plotRel(dres0_pm, draw_diag = TRUE, alpha = alpha)
mtext("number of loci", 3, 0.2)
coln <- hcl.colors(diff(nrng)*2.4, "Purple-Blue", rev = TRUE)[1:(diff(nrng) + 1)]
plotRel(nmat, rlim = NA, col = coln, add = TRUE, 
        draw_diag = TRUE, col_diag = "gray45", border_diag = "white")
```


```{r}
par(pardef)

```


```{r, fig.height=6, fig.width = 6}
# horizontal colorbar 
par(mar = c(1, 1, 1, 3))
border_sig <- "darkviolet"
plotRel(dres0_pm, draw_diag = TRUE, border_diag = border_sig, alpha = alpha, 
        border_sig = border_sig, lwd_sig = 2)
legend(32, 20, pch = 0, col = border_sig, pt.lwd = 2, pt.cex = 1.4, 
       box.col = "gray", legend = expression("significant at" ~~ alpha == 0.05))
text(1:nsmp + 0.3, 1:nsmp - 0.5, labels = names(dsmp_pm), col = donor_cols[meta$donor], adj = 0,
     cex = 0.55, xpd = TRUE)
par(fig = c(0.25, 1, 0.81, 0.92), mar = c(1, 1, 1, 1), new = TRUE)
plotColorbar(at = c(0.2, 0.4, 0.6, 0.8), horiz = TRUE)
ncol <- 301
lines(c(0, ncol, ncol, 0, 0), c(0, 0, 1, 1, 0), col = "gray")
```


```{r}
par(pardef)

```


```{r, fig.height=8, fig.width=8}
##Heatmap function

# gtypes_ibs_heatmap <- function(mtx, dendo = F){


  colFun = suppressWarnings(brewer.pal(100, 'Greens'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), 
                                  colFun)
   
  mtx <- dres0_pm[, , "estimate"]  
  p_mtx <- dres0_pm[, , "p_value"] 
  dendo = "no_clustering"

  print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
   p_mtx <- p_mtx[rownames(mtx), colnames(mtx)]
   
  # fa = str_extract(colnames(mtx), 'MSC.*')
  fa = str_extract(colnames(mtx), 'M.*')
  print(str(fa))
  
  # c_d = cluster_between_groups(mtx, fa)
  # dendog = if(dendo == T) c_d else F
  dendog = if(dendo == T) cluster_between_groups(mtx, fa) else F
  print(str(fa))
  print(str(dendog))
  
  # fa = str_extract(colnames(mtx), 'MSC.*')
  # # fa_col = c("a" = 2, "b" = 3, "c" = 4)
  # dend1 = cluster_between_groups(mtx, fa)
  

  hm = Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog)
  
  
  hm2<-Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), border='black',   heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
                  layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(p_mtx, i, j)
                      l = v <0.05
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "#a4c2f4", lwd = 2, fill = "transparent"))
})
  
  print(hm)
  print(hm2)
  # print(hm3)
# }
```

remove M50 MACS
```{r}

est <- dres0_pm[, , "estimate"] 
est <- est[!str_detect(rownames(est), "MACS"), !str_detect(colnames(est), "MACS")]
# colnames(est) <- str_remove(colnames(est), "_SLO")
# rownames(est) <- str_remove(rownames(est), "_SLO")

pval <- dres0_pm[, , "p_value"] 
pval <- pval[!str_detect(rownames(pval), "MACS"), !str_detect(colnames(pval), "MACS")]
# colnames(pval) <- str_remove(colnames(pval), "_SLO")
# rownames(pval) <- str_remove(rownames(pval), "_SLO")

```


```{r, fig.height=18.5, fig.width=15, eval=T}
  
  # colFun = suppressWarnings(brewer.pal(100, 'YlGnBu'))
  colFun = suppressWarnings(brewer.pal(100, 'Blues'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), 
                                  colFun)

  dendo = "no_clustering"

  strn_numbers = clustr_numbers_s_m1_m2 
  strn_number_col = "strain_n" 
  str2remove_rowcolnms = "M[0-9]+_|_.$|SLO_"
  
  
  full_samples <- union(rownames(est), colnames(est))  # All unique sample names
  mtx_full <- matrix(NA, nrow = length(full_samples), ncol = length(full_samples),
                 dimnames = list(full_samples, full_samples))

# Fill in the existing data
mtx <- mtx_full
mtx[rownames(est), colnames(est)] <- est
diag(mtx) <- 1

# Fill in the existing data
p_mtx <- mtx_full
p_mtx[rownames(pval), colnames(pval)] <- pval
diag(p_mtx) <- 1

  print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
   p_mtx <- p_mtx[rownames(mtx), colnames(mtx)]
   
   # Mirror the matrix
   m_mtx <- mtx
   m_mtx[upper.tri(m_mtx)] <- t(m_mtx)[upper.tri(m_mtx)]
   diag(m_mtx) <- 0
   
   m_p_mtx <- p_mtx
   m_p_mtx[upper.tri(m_p_mtx)] <- t(m_p_mtx)[upper.tri(m_p_mtx)]

   
  sibd_pht_mtx <- m_mtx  
  
  # fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
  # fa = str_replace_all(str_remove(colnames(sibd_pht_mtx), '_SC.*'), c("_MACS" = "M", "_SLO" = "S"))
  # fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
  fa = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')
  c_d = cluster_within_group(sibd_pht_mtx, fa)
  # c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))
  c_names = str_remove_all(colnames(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(colnames(sibd_pht_mtx))
  
  ## Add total cell numbers in bracket to the column names
  c_names_w_n <- c_names %>% data.frame() %>% rownames_to_column("sample_cluster") %>% left_join(., strn_numbers, by = 'sample_cluster') %>% mutate(nm = paste0(., " (", !!rlang::sym(strn_number_col),")")) %>% select(sample_cluster, nm) %>% deframe
  
  
  # fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
  fr = str_remove_all(rownames(sibd_pht_mtx), '_SC.*|_SLO')
  print(fr)
  c_r = cluster_within_group(sibd_pht_mtx, fr)
  # r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))
  r_names = str_remove_all(row.names(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(row.names(sibd_pht_mtx))
  
  # create annotation dataframe
  fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")
  
  # nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
  nms = unique(str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO'))
  # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
  # sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
  sample_c_cols <- donor_cols
  
  # sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
  
  split_df <- fa_df[, c("Status", "donor")] %>% mutate(across(Status, ~factor(., levels = c("Asymptomatic", "Symptomatic"))))
  
  split_gap_lens <- split_df %>% distinct() %>% arrange(Status) %>% count(Status) %>% pull(n)
  split_gap_key <- c(rep(1.2, split_gap_lens[1]-1),5, rep(1.2, split_gap_lens[2]))
  split_donor_names <- split_df %>% arrange(Status, donor) %>% distinct() %>% pull(donor) #%>% paste0(., "| ")
  
  # sibd_pht_mtx[upper.tri(sibd_pht_mtx)] <- NA
  diag(sibd_pht_mtx) <- NA
  
hm<-  Heatmap(sibd_pht_mtx, name='IBD', 
                     col = colFun, 
                     show_row_names = TRUE, 
                     row_labels = r_names,
                     row_names_side = c("left"), 
                     show_column_names = TRUE,  
                     column_labels = c_names_w_n,
                     show_row_dend = F,  
                     show_column_dend = F, 
                     row_title_rot = 0, 
                     # cluster_rows = c_r, 
                     # cluster_columns = c_d, 
                     cluster_rows = T,
                     cluster_columns = T,
                     column_names_gp = gpar(fontsize = 16), 
                     row_names_gp = gpar(fontsize = 17), 
                     heatmap_legend_param = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                 labels_gp = gpar(fontsize = 17),
                                                 direction = "horizontal"),
                     bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                           col = list(donor = sample_c_cols, 
                                                                      Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                           annotation_name_gp = gpar(fontsize = 0, fontface = 'bold'), 
                                                           annotation_legend_param = list(
                                                             donor = list(title = "Donor",
                                                                          title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 17),
                                                                          direction = "horizontal",
                                                                          nrow = 3),
                                                             Status = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                           labels_gp = gpar(fontsize = 17),
                                                                           direction = "horizontal",
                                                                           nrow = 2)
                                                           )),
                     # column_split = length(nms), 
                     border = TRUE, 
                     # border_gp = gpar(col = "white"),
                     left_annotation=rowAnnotation(df = fa_df[, c("Status","donor")],
                                                   col = list(donor = sample_c_cols, 
                                                              Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                   annotation_name_gp = gpar(fontsize = 17, fontface = 'bold'), 
                                                   show_legend = F), 
                     # row_split = length(nms), 
                     row_split = split_df,
                     cluster_column_slices =F,
                     cluster_row_slices =F,
                     row_gap = unit(split_gap_key, "mm"),
                     column_split = split_df,
                     column_gap = unit(split_gap_key, "mm"),
                     column_title = split_donor_names,
                     column_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                     column_title_side = c("bottom"),
                     column_title_rot = 90,
                     row_title = split_donor_names,
                     row_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                      na_col = "white",
                     layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(m_p_mtx, i, j)
                      l = v < 0.05
                      if (any(l)) {  # Only execute if there are TRUE values in l
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                      # grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                     # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
  ) %>%
    # draw(., merge_legend = TRUE)
    draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")

hm
```


```{r}
#dcifer
# dcf_pm_ag <- readRDS("data/processed/Pf/dcf_pm_ag.rds")
  
dsmp_pm_ag <- readRDS("data/processed/Pf/dsmp_pm_ag.rds")

# coi_pm_ag <- readRDS("data/processed/Pf/coi_pm_ag.rds")

# afreq_pm_ag <- readRDS("data/processed/Pf/afreq_pm_ag.rds")

dres0_pm_ag <- readRDS("data/processed/Pf/dres0_pm_ag.rds")
```


```{r}
# optionally, extract location information
meta_ag <- data.frame("sampleID" = names(dsmp_pm_ag ), "donor" = str_remove(names(dsmp_pm_ag ), '_SC.*')) %>% 
  left_join(all_pf_status, by = "donor")# %>% 
  # column_to_rownames("rname")

# meta <- meta[match(names(dsmp_pm), meta$sampleID), ]  # order samples as in dsmp_pm
```


retain only asexuals with >100
```{r}
est_ag <- dres0_pm_ag[, , "estimate"] 
est_ag <- est_ag[!str_detect(rownames(est_ag), "MACS"), !str_detect(colnames(est_ag), "MACS")]
est_a <- est_ag[str_detect(rownames(est_ag), "_A"), str_detect(colnames(est_ag), "_A")]
est_g <- est_ag[str_detect(rownames(est_ag), "_G"), str_detect(colnames(est_ag), "_G")]
# colnames(est) <- str_remove(colnames(est), "_SLO")
# rownames(est) <- str_remove(rownames(est), "_SLO")

pval_ag <- dres0_pm_ag[, , "p_value"] 
pval_ag <- pval_ag[!str_detect(rownames(pval_ag), "MACS"), !str_detect(colnames(pval_ag), "MACS")]
pval_a <- pval_ag[str_detect(rownames(pval_ag), "_A"), str_detect(colnames(pval_ag), "_A")]
pval_g <- pval_ag[str_detect(rownames(pval_ag), "_G"), str_detect(colnames(pval_ag), "_G")]
# colnames(pval) <- str_remove(colnames(pval), "_SLO")
# rownames(pval) <- str_remove(rownames(pval), "_SLO")

```


```{r, fig.height=12.5, fig.width=11, eval=T}
  est <- est_a
  pval <- pval_a


  # colFun = suppressWarnings(brewer.pal(100, 'YlGnBu'))
  colFun = suppressWarnings(brewer.pal(100, 'Blues'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), 
                                  colFun)

  dendo = "no_clustering"

  strn_numbers = clustr_numbers_ag_m1_m2 
  strn_number_col = "strain_ag_n" 
  str2remove_rowcolnms = "M[0-9]+_|SLO_"
  
  
  full_samples <- union(rownames(est), colnames(est))  # All unique sample names
  mtx_full <- matrix(NA, nrow = length(full_samples), ncol = length(full_samples),
                 dimnames = list(full_samples, full_samples))

# Fill in the existing data
mtx <- mtx_full
mtx[rownames(est), colnames(est)] <- est
diag(mtx) <- 1

# Fill in the existing data
p_mtx <- mtx_full
p_mtx[rownames(pval), colnames(pval)] <- pval
diag(p_mtx) <- 1

  print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
   p_mtx <- p_mtx[rownames(mtx), colnames(mtx)]
   
   # Mirror the matrix
   m_mtx <- mtx
   m_mtx[upper.tri(m_mtx)] <- t(m_mtx)[upper.tri(m_mtx)]
   diag(m_mtx) <- 0
   
   m_p_mtx <- p_mtx
   m_p_mtx[upper.tri(m_p_mtx)] <- t(m_p_mtx)[upper.tri(m_p_mtx)]

   
  sibd_pht_mtx <- m_mtx  
  
  # fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
  # fa = str_replace_all(str_remove(colnames(sibd_pht_mtx), '_SC.*'), c("_MACS" = "M", "_SLO" = "S"))
  # fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
  fa = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')
  c_d = cluster_within_group(sibd_pht_mtx, fa)
  # c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))
  c_names = str_remove_all(colnames(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(colnames(sibd_pht_mtx))
  
  ## Add total cell numbers in bracket to the column names
  c_names_w_n <- c_names %>% data.frame() %>% rownames_to_column("sample_cluster") %>% left_join(., strn_numbers, by = 'sample_cluster') %>% mutate(nm = paste0(., " (", !!rlang::sym(strn_number_col),")")) %>% select(sample_cluster, nm) %>% deframe
  
  
  # fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
  fr = str_remove_all(rownames(sibd_pht_mtx), '_SC.*|_SLO')
  print(fr)
  c_r = cluster_within_group(sibd_pht_mtx, fr)
  # r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))
  r_names = str_remove_all(row.names(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(row.names(sibd_pht_mtx))
  
  # create annotation dataframe
  fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")
  
  # nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
  nms = unique(str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO'))
  # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
  # sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
  sample_c_cols <- donor_cols
  
  # sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
  
  split_df <- fa_df[, c("Status", "donor")] %>% mutate(across(Status, ~factor(., levels = c("Asymptomatic", "Symptomatic"))))
  
  split_gap_lens <- split_df %>% distinct() %>% arrange(Status) %>% count(Status) %>% pull(n)
  split_gap_key <- c(rep(1.2, split_gap_lens[1]-1),5, rep(1.2, split_gap_lens[2]))
  split_donor_names <- split_df %>% arrange(Status, donor) %>% distinct() %>% pull(donor) #%>% paste0(., "| ")
  
  # sibd_pht_mtx[upper.tri(sibd_pht_mtx)] <- NA
  diag(sibd_pht_mtx) <- NA
  
hm<-  Heatmap(sibd_pht_mtx, name='IBD', 
                     col = colFun, 
                     show_row_names = TRUE, 
                     row_labels = r_names,
                     row_names_side = c("left"), 
                     show_column_names = TRUE,  
                     column_labels = c_names_w_n,
                     show_row_dend = F,  
                     show_column_dend = F, 
                     row_title_rot = 0, 
                     # cluster_rows = c_r, 
                     # cluster_columns = c_d, 
                     cluster_rows = T,
                     cluster_columns = T,
                     column_names_gp = gpar(fontsize = 16), 
                     row_names_gp = gpar(fontsize = 17), 
                     heatmap_legend_param = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                 labels_gp = gpar(fontsize = 17),
                                                 direction = "horizontal"),
                     bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                           col = list(donor = sample_c_cols, 
                                                                      Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                           annotation_name_gp = gpar(fontsize = 0, fontface = 'bold'), 
                                                           annotation_legend_param = list(
                                                             donor = list(title = "Donor",
                                                                          title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 17),
                                                                          direction = "horizontal",
                                                                          nrow = 3),
                                                             Status = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                           labels_gp = gpar(fontsize = 17),
                                                                           direction = "horizontal",
                                                                           nrow = 2)
                                                           )),
                     # column_split = length(nms), 
                     border = TRUE, 
                     # border_gp = gpar(col = "white"),
                     left_annotation=rowAnnotation(df = fa_df[, c("Status","donor")],
                                                   col = list(donor = sample_c_cols, 
                                                              Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                   annotation_name_gp = gpar(fontsize = 17, fontface = 'bold'), 
                                                   show_legend = F), 
                     # row_split = length(nms), 
                     row_split = split_df,
                     cluster_column_slices =F,
                     cluster_row_slices =F,
                     row_gap = unit(split_gap_key, "mm"),
                     column_split = split_df,
                     column_gap = unit(split_gap_key, "mm"),
                     column_title = split_donor_names,
                     column_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                     column_title_side = c("bottom"),
                     column_title_rot = 90,
                     row_title = split_donor_names,
                     row_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                      na_col = "white",
                     layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(m_p_mtx, i, j)
                      l = v < 0.05
                      if (any(l)) {  # Only execute if there are TRUE values in l
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                      # grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                     # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
  ) %>%
    # draw(., merge_legend = TRUE)
    draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")

hm
```

```{r, fig.height=12.5, fig.width=11, eval=T}

  est <- est_a
  pval <- pval_a


  # colFun = suppressWarnings(brewer.pal(100, 'YlGnBu'))
  colFun = suppressWarnings(brewer.pal(100, 'Blues'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), 
                                  colFun)

  dendo = "no_clustering"

  strn_numbers = clustr_numbers_ag_m1_m2 
  strn_number_col = "strain_ag_n" 
  str2remove_rowcolnms = "M[0-9]+_|SLO_"
  
  
  full_samples <- union(rownames(est), colnames(est))  # All unique sample names
  mtx_full <- matrix(NA, nrow = length(full_samples), ncol = length(full_samples),
                 dimnames = list(full_samples, full_samples))

# Fill in the existing data
mtx <- mtx_full
mtx[rownames(est), colnames(est)] <- est
diag(mtx) <- 1

# Fill in the existing data
p_mtx <- mtx_full
p_mtx[rownames(pval), colnames(pval)] <- pval
diag(p_mtx) <- 1

  print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
   p_mtx <- p_mtx[rownames(mtx), colnames(mtx)]
   
   # Mirror the matrix
   m_mtx <- mtx
   m_mtx[upper.tri(m_mtx)] <- t(m_mtx)[upper.tri(m_mtx)]
   diag(m_mtx) <- 0
   
   m_p_mtx <- p_mtx
   m_p_mtx[upper.tri(m_p_mtx)] <- t(m_p_mtx)[upper.tri(m_p_mtx)]

   
  sibd_pht_mtx <- m_mtx  
  
  # fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
  # fa = str_replace_all(str_remove(colnames(sibd_pht_mtx), '_SC.*'), c("_MACS" = "M", "_SLO" = "S"))
  # fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
  fa = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')
  c_d = cluster_within_group(sibd_pht_mtx, fa)
  # c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))
  c_names = str_remove_all(colnames(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(colnames(sibd_pht_mtx))
  
  ## Add total cell numbers in bracket to the column names
  c_names_w_n <- c_names %>% data.frame() %>% rownames_to_column("sample_cluster") %>% left_join(., strn_numbers, by = 'sample_cluster') %>% mutate(nm = paste0(., " (", !!rlang::sym(strn_number_col),")  ")) %>% select(sample_cluster, nm) %>% deframe
  
  
  # fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
  fr = str_remove_all(rownames(sibd_pht_mtx), '_SC.*|_SLO')
  print(fr)
  c_r = cluster_within_group(sibd_pht_mtx, fr)
  # r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))
  r_names = str_remove_all(row.names(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(row.names(sibd_pht_mtx))
  
  # create annotation dataframe
  fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")
  
  # nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
  nms = unique(str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO'))
  # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
  # sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
  sample_c_cols <- donor_cols
  
  # sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
  
  split_df <- fa_df[, c("Status", "donor")] %>% mutate(across(Status, ~factor(., levels = c("Asymptomatic", "Symptomatic"))))
  
  split_gap_lens <- split_df %>% distinct() %>% arrange(Status) %>% count(Status) %>% pull(n)
  split_gap_key <- c(rep(1.2, split_gap_lens[1]-1),5, rep(1.2, split_gap_lens[2]))
  split_donor_names <- split_df %>% arrange(Status, donor) %>% distinct() %>% pull(donor) #%>% paste0(., "| ")
  
  # sibd_pht_mtx[upper.tri(sibd_pht_mtx)] <- NA
  diag(sibd_pht_mtx) <- NA
  
# by defalt, there no space on the top of the title, here we add 4pt to the top.
# it can be reset by `ht_opt(RESET = TRUE)`
ht_opt$TITLE_PADDING = unit(c(2, 2), "mm")
# ht_opt$COLUMN_ANNO_PADDING = unit(20, "mm")
# ht_opt@layout$padding[3] <- unit(2, "cm")

hm<-  Heatmap(sibd_pht_mtx, name='IBD', 
                     col = colFun, 
                     show_row_names = TRUE, 
                     row_labels = r_names,
                     row_names_side = c("right"), 
                     show_column_names = TRUE,  
                     column_labels = c_names_w_n,
                     column_names_side = c("top"), 
                     show_row_dend = F,  
                     show_column_dend = F, 
                     row_title_rot = 0, 
                     # cluster_rows = c_r, 
                     # cluster_columns = c_d, 
                     cluster_rows = T,
                     cluster_columns = T,
                     column_names_gp = gpar(fontsize = 16), 
                     row_names_gp = gpar(fontsize = 17), 
                     heatmap_legend_param = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                 labels_gp = gpar(fontsize = 17),
                                                 direction = "horizontal"),
                     top_annotation = HeatmapAnnotation(df = fa_df,
                                                           col = list(donor = sample_c_cols, 
                                                                      Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                           annotation_name_gp = gpar(fontsize = 0, fontface = 'bold'), 
                                                           annotation_legend_param = list(
                                                             donor = list(title = "Donor",
                                                                          title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 17),
                                                                          direction = "horizontal",
                                                                          nrow = 3),
                                                             Status = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                           labels_gp = gpar(fontsize = 17),
                                                                           direction = "horizontal",
                                                                           nrow = 2)
                                                           )),
                     # column_split = length(nms), 
                     border = TRUE, 
                     # border_gp = gpar(col = "white"),
                     right_annotation=rowAnnotation(df = fa_df[, c("Status","donor")],
                                                   col = list(donor = sample_c_cols, 
                                                              Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                   annotation_name_gp = gpar(fontsize = 17, fontface = 'bold'), 
                                                   show_legend = F,
                                                   annotation_name_side = "top"), 
                     # row_split = length(nms), 
                     row_split = split_df,
                     cluster_column_slices =F,
                     cluster_row_slices =F,
                     row_gap = unit(split_gap_key, "mm"),
                     column_split = split_df,
                     column_gap = unit(split_gap_key, "mm"),
                     column_title = split_donor_names,
                     column_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                     column_title_side = c("top"),
                     column_title_rot = 90,
                     row_title = split_donor_names,
                     row_title_side = c("right"),
                     row_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                      na_col = "white",
                     layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(m_p_mtx, i, j)
                      l = v < 0.05
                      if (any(l)) {  # Only execute if there are TRUE values in l
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                      # grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                     # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
  ) %>%
    # draw(., merge_legend = TRUE)
    draw(., newpage = F, merge_legends = T, heatmap_legend_side = "top", annotation_legend_side = "top")

hm

ht_opt(RESET = TRUE)
```

```{r}
gogo()
```

```{r, fig.height=18.5, fig.width=15, eval=T}
  est <- est_g
  pval <- pval_g


  # colFun = suppressWarnings(brewer.pal(100, 'YlGnBu'))
  colFun = suppressWarnings(brewer.pal(100, 'Blues'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), 
                                  colFun)

  dendo = "no_clustering"

  strn_numbers = clustr_numbers_ag_m1_m2 
  strn_number_col = "strain_ag_n" 
  str2remove_rowcolnms = "M[0-9]+_|SLO_"
  
  
  full_samples <- union(rownames(est), colnames(est))  # All unique sample names
  mtx_full <- matrix(NA, nrow = length(full_samples), ncol = length(full_samples),
                 dimnames = list(full_samples, full_samples))

# Fill in the existing data
mtx <- mtx_full
mtx[rownames(est), colnames(est)] <- est
diag(mtx) <- 1

# Fill in the existing data
p_mtx <- mtx_full
p_mtx[rownames(pval), colnames(pval)] <- pval
diag(p_mtx) <- 1

  print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
   p_mtx <- p_mtx[rownames(mtx), colnames(mtx)]
   
   # Mirror the matrix
   m_mtx <- mtx
   m_mtx[upper.tri(m_mtx)] <- t(m_mtx)[upper.tri(m_mtx)]
   diag(m_mtx) <- 0
   
   m_p_mtx <- p_mtx
   m_p_mtx[upper.tri(m_p_mtx)] <- t(m_p_mtx)[upper.tri(m_p_mtx)]

   
  sibd_pht_mtx <- m_mtx  
  
  # fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
  # fa = str_replace_all(str_remove(colnames(sibd_pht_mtx), '_SC.*'), c("_MACS" = "M", "_SLO" = "S"))
  # fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
  fa = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')
  c_d = cluster_within_group(sibd_pht_mtx, fa)
  # c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))
  c_names = str_remove_all(colnames(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(colnames(sibd_pht_mtx))
  
  ## Add total cell numbers in bracket to the column names
  c_names_w_n <- c_names %>% data.frame() %>% rownames_to_column("sample_cluster") %>% left_join(., strn_numbers, by = 'sample_cluster') %>% mutate(nm = paste0(., " (", !!rlang::sym(strn_number_col),")")) %>% select(sample_cluster, nm) %>% deframe
  
  
  # fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
  fr = str_remove_all(rownames(sibd_pht_mtx), '_SC.*|_SLO')
  print(fr)
  c_r = cluster_within_group(sibd_pht_mtx, fr)
  # r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))
  r_names = str_remove_all(row.names(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(row.names(sibd_pht_mtx))
  
  # create annotation dataframe
  fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")
  
  # nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
  nms = unique(str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO'))
  # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
  # sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
  sample_c_cols <- donor_cols
  
  # sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
  
  split_df <- fa_df[, c("Status", "donor")] %>% mutate(across(Status, ~factor(., levels = c("Asymptomatic", "Symptomatic"))))
  
  split_gap_lens <- split_df %>% distinct() %>% arrange(Status) %>% count(Status) %>% pull(n)
  split_gap_key <- c(rep(1.2, split_gap_lens[1]-1),5, rep(1.2, split_gap_lens[2]))
  split_donor_names <- split_df %>% arrange(Status, donor) %>% distinct() %>% pull(donor) #%>% paste0(., "| ")
  
  # sibd_pht_mtx[upper.tri(sibd_pht_mtx)] <- NA
  diag(sibd_pht_mtx) <- NA
  
hm<-  Heatmap(sibd_pht_mtx, name='IBD', 
                     col = colFun, 
                     show_row_names = TRUE, 
                     row_labels = r_names,
                     row_names_side = c("left"), 
                     show_column_names = TRUE,  
                     column_labels = c_names_w_n,
                     show_row_dend = F,  
                     show_column_dend = F, 
                     row_title_rot = 0, 
                     # cluster_rows = c_r, 
                     # cluster_columns = c_d, 
                     cluster_rows = T,
                     cluster_columns = T,
                     column_names_gp = gpar(fontsize = 16), 
                     row_names_gp = gpar(fontsize = 17), 
                     heatmap_legend_param = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                 labels_gp = gpar(fontsize = 17),
                                                 direction = "horizontal"),
                     bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                           col = list(donor = sample_c_cols, 
                                                                      Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                           annotation_name_gp = gpar(fontsize = 0, fontface = 'bold'), 
                                                           annotation_legend_param = list(
                                                             donor = list(title = "Donor",
                                                                          title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 17),
                                                                          direction = "horizontal",
                                                                          nrow = 3),
                                                             Status = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                           labels_gp = gpar(fontsize = 17),
                                                                           direction = "horizontal",
                                                                           nrow = 2)
                                                           )),
                     # column_split = length(nms), 
                     border = TRUE, 
                     # border_gp = gpar(col = "white"),
                     left_annotation=rowAnnotation(df = fa_df[, c("Status","donor")],
                                                   col = list(donor = sample_c_cols, 
                                                              Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                   annotation_name_gp = gpar(fontsize = 17, fontface = 'bold'), 
                                                   show_legend = F), 
                     # row_split = length(nms), 
                     row_split = split_df,
                     cluster_column_slices =F,
                     cluster_row_slices =F,
                     row_gap = unit(split_gap_key, "mm"),
                     column_split = split_df,
                     column_gap = unit(split_gap_key, "mm"),
                     column_title = split_donor_names,
                     column_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                     column_title_side = c("bottom"),
                     column_title_rot = 90,
                     row_title = split_donor_names,
                     row_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                      na_col = "white",
                     layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(m_p_mtx, i, j)
                      l = v < 0.05
                      if (any(l)) {  # Only execute if there are TRUE values in l
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                      # grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                     # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
  ) %>%
    # draw(., merge_legend = TRUE)
    draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")

hm
```

```{r}
est_ag <- dres0_pm_ag[, , "estimate"] 
est_ag <- est_ag[!str_detect(rownames(est_ag), "MACS"), !str_detect(colnames(est_ag), "MACS")]
est_m49m66 <- est_ag[str_detect(rownames(est_ag), "M49|M66"), str_detect(colnames(est_ag), "M49|M66")]

pval_ag <- dres0_pm_ag[, , "p_value"] 
pval_ag <- pval_ag[!str_detect(rownames(pval_ag), "MACS"), !str_detect(colnames(pval_ag), "MACS")]
pval_m49m66 <- pval_ag[str_detect(rownames(pval_ag), "M49|M66"), str_detect(colnames(pval_ag), "M49|M66")]


```


```{r, fig.height=6.5, fig.width=8.5, eval=T}
  est <- est_m49m66
  pval <- pval_m49m66


  # colFun = suppressWarnings(brewer.pal(100, 'YlGnBu'))
  colFun = suppressWarnings(brewer.pal(100, 'Blues'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), 
                                  colFun)

  dendo = "no_clustering"

  strn_numbers = clustr_numbers_ag_m1_m2 
  strn_number_col = "strain_ag_n" 
  str2remove_rowcolnms = "M[0-9]+_|SLO_"
  
  
  full_samples <- union(rownames(est), colnames(est))  # All unique sample names
  mtx_full <- matrix(NA, nrow = length(full_samples), ncol = length(full_samples),
                 dimnames = list(full_samples, full_samples))

# Fill in the existing data
mtx <- mtx_full
mtx[rownames(est), colnames(est)] <- est
diag(mtx) <- 1

# Fill in the existing data
p_mtx <- mtx_full
p_mtx[rownames(pval), colnames(pval)] <- pval
diag(p_mtx) <- 1

  print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
   
   p_mtx <- p_mtx[rownames(mtx), colnames(mtx)]
   
   # Mirror the matrix
   m_mtx <- mtx
   m_mtx[upper.tri(m_mtx)] <- t(m_mtx)[upper.tri(m_mtx)]
   diag(m_mtx) <- 0
   
   m_p_mtx <- p_mtx
   m_p_mtx[upper.tri(m_p_mtx)] <- t(m_p_mtx)[upper.tri(m_p_mtx)]

   
  sibd_pht_mtx <- m_mtx  
  
  # fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
  # fa = str_replace_all(str_remove(colnames(sibd_pht_mtx), '_SC.*'), c("_MACS" = "M", "_SLO" = "S"))
  # fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
  fa = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')
  c_d = cluster_within_group(sibd_pht_mtx, fa)
  # c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))
  c_names = str_remove_all(colnames(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(colnames(sibd_pht_mtx))
  
  ## Add total cell numbers in bracket to the column names
  c_names_w_n <- c_names %>% data.frame() %>% rownames_to_column("sample_cluster") %>% left_join(., strn_numbers, by = 'sample_cluster') %>% mutate(nm = paste0(., " (", !!rlang::sym(strn_number_col),")")) %>% select(sample_cluster, nm) %>% deframe
  
  
  # fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
  fr = str_remove_all(rownames(sibd_pht_mtx), '_SC.*|_SLO')
  print(fr)
  c_r = cluster_within_group(sibd_pht_mtx, fr)
  # r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))
  r_names = str_remove_all(row.names(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(row.names(sibd_pht_mtx))
  
  # create annotation dataframe
  fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")
  
  # nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
  nms = unique(str_remove_all(colnames(sibd_pht_mtx), '_SC.*|_SLO'))
  # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
  # sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
  sample_c_cols <- donor_cols
  
  # sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
  
  split_df <- fa_df[, c("Status", "donor")] %>% mutate(across(Status, ~factor(., levels = c("Asymptomatic", "Symptomatic"))))
  
  split_gap_lens <- split_df %>% distinct() %>% arrange(Status) %>% count(Status) %>% pull(n)
  split_gap_key <- c(rep(1.2, split_gap_lens[1]-1),5, rep(1.2, split_gap_lens[2]))
  split_donor_names <- split_df %>% arrange(Status, donor) %>% distinct() %>% pull(donor) #%>% paste0(., "| ")
  
  # sibd_pht_mtx[upper.tri(sibd_pht_mtx)] <- NA
  diag(sibd_pht_mtx) <- NA
  
hm<-  Heatmap(sibd_pht_mtx, name='IBD', 
                     col = colFun, 
                     show_row_names = TRUE, 
                     row_labels = r_names,
                     row_names_side = c("left"), 
                     show_column_names = TRUE,  
                     column_labels = c_names_w_n,
                     show_row_dend = F,  
                     show_column_dend = F, 
                     row_title_rot = 0, 
                     # cluster_rows = c_r, 
                     # cluster_columns = c_d, 
                     cluster_rows = T,
                     cluster_columns = T,
                     column_names_gp = gpar(fontsize = 16), 
                     row_names_gp = gpar(fontsize = 17), 
                     heatmap_legend_param = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                 labels_gp = gpar(fontsize = 17),
                                                 direction = "horizontal"),
                     bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                           col = list(donor = sample_c_cols, 
                                                                      Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                           annotation_name_gp = gpar(fontsize = 0, fontface = 'bold'), 
                                                           annotation_legend_param = list(
                                                             donor = list(title = "Donor",
                                                                          title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 17),
                                                                          direction = "horizontal",
                                                                          nrow = 3),
                                                             Status = list(title_gp = gpar(fontsize = 17, fontface = "bold"), 
                                                                           labels_gp = gpar(fontsize = 17),
                                                                           direction = "horizontal",
                                                                           nrow = 2)
                                                           )),
                     # column_split = length(nms), 
                     border = TRUE, 
                     # border_gp = gpar(col = "white"),
                     left_annotation=rowAnnotation(df = fa_df[, c("Status","donor")],
                                                   col = list(donor = sample_c_cols, 
                                                              Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                   annotation_name_gp = gpar(fontsize = 17, fontface = 'bold'), 
                                                   show_legend = F), 
                     # row_split = length(nms), 
                     row_split = split_df,
                     cluster_column_slices =F,
                     cluster_row_slices =F,
                     row_gap = unit(split_gap_key, "mm"),
                     column_split = split_df,
                     column_gap = unit(split_gap_key, "mm"),
                     column_title = split_donor_names,
                     column_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                     column_title_side = c("bottom"),
                     column_title_rot = 90,
                     row_title = split_donor_names,
                     row_title_gp = gpar(fontsize = 17, fill = c("white"), fontface = "bold"),
                      na_col = "white",
                     layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(m_p_mtx, i, j)
                      l = v < 0.05
                      if (any(l)) {  # Only execute if there are TRUE values in l
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 1, fill = "transparent"))
                      }
                      # grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "red", lwd = 2, fill = "transparent"))
                      }
                     # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
  ) %>%
    # draw(., merge_legend = TRUE)
    draw(., newpage = F, merge_legends = T, heatmap_legend_side = "right", annotation_legend_side = "right")

hm
```


### Write out for IBD analysis
```{r}
##!!NOTE - gfv_p1_bi_strn_pf7_c is created somewhere downstream so this is likely to give an error

imap(ss_cln_all_high_n_mtx, ~.x %>% 
       dplyr::select(-c(id, chrom, alleles))%>% 
       dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% 
       filter(!if_all(starts_with('M'), ~is.na(.))) %>% 
       mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% 
       mutate(across(chr, as.integer)) %>%
       write_tsv(., paste0("data/processed/Pf/", sv_dir, "/ss_ibd_cln_", .y,"_mtx.tsv")))
```

```{r}
##!!NOTE - gfv_p1_bi_strn_pf7_c is created somewhere downstream so this is likely to give an error

imap(ss_cln_all_high_n_mtx, ~.x %>% 
       dplyr::select(-c(id, chrom, alleles))%>% 
       dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% 
       filter(!if_all(starts_with('M'), ~is.na(.))) %>% 
       mutate(across(chr, as.integer)) %>%head)
```

Remove the monoclonal as they need more invesitagation later

```{r}
sample_name_nms_polyc <- sample_name_nms[map(opt_narrow_clusters_nms, ~.x > 1) %>% unlist()] 
sample_name_nms_monoc <- sample_name_nms[map(opt_narrow_clusters_nms, ~.x == 1) %>% unlist()] 

ss_cln_all_mtx_polyc <- imap(ss_cln_all_high_n_mtx, ~.x %>% .[!(str_detect(names(.), paste(sample_name_nms_monoc %>% str_remove(., "SC") , collapse = "|")))])

imap(ss_cln_all_mtx_polyc, ~.x %>% 
       dplyr::select(-c(id, chrom, alleles))%>% 
       dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% 
       filter(!if_all(starts_with('M'), ~is.na(.))) %>% 
       mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% 
       write_tsv(., paste0("data/processed/Pf/", sv_dir, "/ss_ibd_cln_", .y,"_mtx_polyc.tsv")))
```

```{r, eval=F}
##Copy souporcell files for the optimum clusters


# system(paste0('rsync -av jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/ss_ibd_cln* /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/processed/Pf/'))

```




```{r}
##Run IBD 
# wdir<- '/software/team222/jr35/hmmIBD'
# 
# map(names(ss_cln_all_high_n_mtx[1]), ~system(paste0(wdir,'/hmmIBD -i data/processed/Pf/m2_all/', sv_dir, '/ss_ibd_cln_', .x,'_mtx.tsv -o data/processed/Pf/m2_all', sv_dir, '/ss_ibd_cln_', .x,' -m 100')))
# map(names(ss_cln_all_high_n_mtx[1]), ~system(paste0(wdir,'/hmmIBD -i data/processed/Pf/m2_all', sv_dir, '/ss_ibd_cln_', .x,'_mtx_polyc.tsv -o data/processed/Pf/m2_all', sv_dir, '/ss_ibd_cln_polyc_', .x,' -m 100')))
```


```{r}
##Run IBD 
# source_dir<- 'scp jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/'
# local_dir<- '/Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/Mali2/data/processed/Pf/m2_all/'
# 
# map(names(ss_cln_all_high_n_mtx[1]), ~paste0(source_dir, sv_dir,'/ss_ibd_cln_', .x,'_mtx.tsv ', local_dir, sv_dir, '/'))
# map(names(ss_cln_all_high_n_mtx[1]), ~paste0(source_dir, sv_dir,'/ss_ibd_cln_', .x,'_mtx_polyc.tsv ', local_dir, sv_dir, '/'))


```


Commands to run IBD not working on farm - not sure why so copying files to local to run there and copying back here.
```{r, eval = F}
##Run IBD 
wdir<- '/software/team222/jr35/hmmIBD'

map(names(ss_cln_all_high_n_mtx[1]), ~paste0(wdir,'/hmmIBD -i data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,'_mtx.tsv -o data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,' -m 100'))
map(names(ss_cln_all_high_n_mtx[1]), ~paste0(wdir,'/hmmIBD -i data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,'_mtx_polyc.tsv -o data/processed/Pf/', sv_dir, '/ss_ibd_cln_polyc_', .x,' -m 100'))
```


```{r}
stg_strn_nms <- names(ss_cln_all_high_n_mtx)

```

```{r}
gogo()
##paste copying directories
source_dir<- 'scp jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/'
local_dir<- '/Users/jr35/Google\ Drive/My\ Drive/PHD_documents/malaria_phd/Mali2/data/processed/Pf/'

map(stg_strn_nms, ~paste0('cd ', local_dir, sv_dir, '/'))
map(stg_strn_nms, ~paste0(source_dir, sv_dir,'/ss_ibd_cln_', .x,'_mtx.tsv .'))
map(stg_strn_nms, ~paste0(source_dir, sv_dir,'/ss_ibd_cln_', .x,'_mtx_polyc.tsv .'))

```


```{r}
gogo()
## once in local mac paste hmmibd commands

map(stg_strn_nms, ~paste0('hmmIBD -i ss_ibd_cln_', .x,'_mtx.tsv -o ss_ibd_cln_', .x,'_mtx -m 100'))
map(stg_strn_nms, ~paste0('hmmIBD -i ss_ibd_cln_', .x,'_mtx_polyc.tsv -o ss_ibd_cln_', .x,'_mtx_polyc -m 100'))


```


```{r}
gogo()
## once in local send output back to farm
dest_dir<- 'jr35@farm22:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/'

map(stg_strn_nms, ~paste0('scp ss_ibd_cln_', .x,'_mtx.hmm* ', dest_dir, sv_dir, '/'))
map(stg_strn_nms, ~paste0('scp ss_ibd_cln_', .x,'_mtx_polyc.hmm* ', dest_dir, sv_dir, '/'))

```


```{r}


# read in ibd results
ss_ibd_cln_all_strain_hmmibd <- map(stg_strn_nms, ~read_delim(paste0('data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,'_mtx.hmm.txt'))) %>% set_names(stg_strn_nms)

ss_ibd_cln_all_strain_hmmibd_f <- map(stg_strn_nms, ~read_delim(paste0('data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,'_mtx.hmm_fract.txt')))%>% set_names(stg_strn_nms)

ss_ibd_cln_all_strain_hmmibd_1 <- ss_ibd_cln_all_strain_hmmibd[[1]]
ss_ibd_cln_all_strain_hmmibd_f_1 <- ss_ibd_cln_all_strain_hmmibd_f[[1]]

ss_ibd_cln_all_strain_hmmibd_2 <- ss_ibd_cln_all_strain_hmmibd[[2]]
ss_ibd_cln_all_strain_hmmibd_f_2 <- ss_ibd_cln_all_strain_hmmibd_f[[2]]

```


Only poloclonal

```{r}
# read in ibd results
ss_ibd_cln_all_strain_hmmibd_polyc <- map(stg_strn_nms, ~read_delim(paste0('data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,'_mtx_polyc.hmm.txt'))) %>% set_names(stg_strn_nms)
ss_ibd_cln_all_strain_hmmibd_f_polyc <- map(stg_strn_nms, ~read_delim(paste0('data/processed/Pf/', sv_dir, '/ss_ibd_cln_', .x,'_mtx_polyc.hmm_fract.txt'))) %>% set_names(stg_strn_nms)

ss_ibd_cln_all_strain_hmmibd_polyc_1 <- ss_ibd_cln_all_strain_hmmibd_polyc[[1]]
ss_ibd_cln_all_strain_hmmibd_f_polyc_1 <- ss_ibd_cln_all_strain_hmmibd_f_polyc[[1]]

ss_ibd_cln_all_strain_hmmibd_polyc_2 <- ss_ibd_cln_all_strain_hmmibd_polyc[[2]]
ss_ibd_cln_all_strain_hmmibd_f_polyc_2 <- ss_ibd_cln_all_strain_hmmibd_f_polyc[[2]]

```

```{r}
# wdir<- '/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master'
# 
# ss_ibd_cln_all_strain_hmmibd <- read_delim('/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_cln_gt_tbl_100_m21.hmm.txt')
# ss_cln_gt_tbl_100_hmmibd_f <- read_delim('/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_cln_gt_tbl_100_m21.hmm_fract.txt')
# 
# ss_cln_gt_tbl_100_plus_m14_hmmibd <- read_delim('/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_cln_gt_tbl_100_m21_plus_m14.hmm.txt')
# ss_cln_gt_tbl_100_plus_m14_hmmibd_f <- read_delim('/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/ss_cln_gt_tbl_100_m21_plus_m14.hmm_fract.txt')


```


```{r, fig.height=13, fig.width=15, eval=T}
ss_ibd_cln_all_strain_hmmibd_f[["all.stage_ag_strain_qcd"]] %>% filter(sample1 == "M66_SC5_A" & sample2 == "M66_SC5_G")
```

```{r, fig.height=13, fig.width=15, eval=T}
ss_ibd_cln_all_strain_hmmibd_f[["all.stage_ag_strain_qcd"]] %>% filter(sample1 == "M55_SC6_A" & sample2 == "M55_SC6_G")
```


```{r, fig.height=5, fig.width=6}
colFun = suppressWarnings(brewer.pal(100, 'Blues'))
colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), colFun)
  
# cols_rand <- c("#fb9a99", "#cab2d6", "#b2df8a", "#e31a1c", "#ffff99", "#33a02c", "#ff7f00", "#b15928", "#fdbf6f")
```




ASEXUALS

ALL STAGES AND STRAINS
```{r, fig.height=10, fig.width=12, eval=T}

hmap_ibd_fn <- function(input_gtyps, strn_numbers, strn_number_col, str2remove_rowcolnms = "M[0-9]+_|_.$" ) {
  
  colFun = suppressWarnings(brewer.pal(100, 'Blues'))
  colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), colFun)
  ##Merge strain numbers to IBD results
  # ss_n_ibd_tbl100 = ss_cln_gt_tbl_100_hmmibd_f %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% left_join(., clustr_numbers, by = c('sample1'='sample_cluster')) %>% left_join(., clustr_numbers, by = c('sample2'='sample_cluster'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1, '\n[', n_1, ']'), id2 = paste0(sample2, '\n[', n_2, ']'))  
  
  # ss_n_ibd_tbl100 = ss_ibd_cln_all_strain_hmmibd_f_2 %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% 
  ss_n_ibd_tbl100 = input_gtyps %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% 
    # left_join(., clustr_numbers_plusm14, by = c('sample1'='sample_cluster')) %>% 
    left_join(., strn_numbers, by = c('sample2'='sample_cluster'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1), id2 = paste0(sample2))  
  
  ##Get the IBD and the number of sites for each pairwise comparison to annotate heatmap
  s_ibdf_nsites_anno <- ss_n_ibd_tbl100 %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., ss_n_ibd_tbl100) %>% mutate(ibdf_nsites = paste0(round(fract_sites_IBD,2), '\n(',N_informative_sites,')'))%>% dplyr::select(starts_with('id'), ibdf_nsites) %>% pivot_wider(names_from = id2, values_from = ibdf_nsites) %>% arrange(id1) %>% column_to_rownames('id1')
  
  
  ##Create mirrored matrix and plot
  sibd_pht_mtx <- ss_n_ibd_tbl100 %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., ss_n_ibd_tbl100) %>% dplyr::select(starts_with('id'), fract_sites_IBD) %>%  pivot_wider(names_from = id2, values_from = fract_sites_IBD) %>% arrange(id1) %>% column_to_rownames('id1')  
  
  
  sibd_pht_mtx <- sibd_pht_mtx[,row.names(sibd_pht_mtx)]
  
  # fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
  fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
  c_d = cluster_within_group(sibd_pht_mtx, fa)
  # c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))
  c_names = str_remove_all(colnames(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(colnames(sibd_pht_mtx))
  
  ## Add total cell numbers in bracket to the column names
  c_names_w_n <- c_names %>% data.frame() %>% rownames_to_column("sample_cluster") %>% left_join(., strn_numbers, by = 'sample_cluster') %>% mutate(nm = paste0(., " (", !!rlang::sym(strn_number_col),")")) %>% select(sample_cluster, nm) %>% deframe
  
  
  # fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
  fr = str_remove(rownames(sibd_pht_mtx), '_SC.*')
  print(fr)
  c_r = cluster_within_group(sibd_pht_mtx, fr)
  # r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))
  r_names = str_remove_all(row.names(sibd_pht_mtx), str2remove_rowcolnms) %>% set_names(row.names(sibd_pht_mtx))
  
  # create annotation dataframe
  fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove(colnames(sibd_pht_mtx), '_SC.*')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")
  
  # nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
  nms = unique(str_remove(colnames(sibd_pht_mtx), '_SC.*'))
  # sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
  # sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
  sample_c_cols <- donor_cols
  
  sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
  
  split_df <- fa_df[, c("Status", "donor")] %>% mutate(across(Status, ~factor(., levels = c("Asymptomatic", "Symptomatic"))))
  
  split_gap_lens <- split_df %>% distinct() %>% arrange(Status) %>% count(Status) %>% pull(n)
  split_gap_key <- c(rep(0, split_gap_lens[1]-1),5, rep(0, split_gap_lens[2]))
  split_donor_names <- split_df %>% arrange(Status, donor) %>% distinct() %>% pull(donor) %>% paste0(., "| ")
  
  sibd_pht<- Heatmap(sibd_pht_mtx_cln, name='IBD', 
                     col = colFun, 
                     show_row_names = TRUE, 
                     row_labels = r_names,
                     row_names_side = c("left"), 
                     show_column_names = TRUE,  
                     column_labels = c_names_w_n,
                     show_row_dend = F,  
                     show_column_dend = F, 
                     row_title_rot = 0, 
                     # cluster_rows = c_r, 
                     # cluster_columns = c_d, 
                     cluster_rows = T,
                     cluster_columns = T,
                     column_names_gp = gpar(fontsize = 15), 
                     row_names_gp = gpar(fontsize = 15), 
                     heatmap_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), 
                                                 labels_gp = gpar(fontsize = 15),
                                                 direction = "horizontal"),
                     bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                           col = list(donor = sample_c_cols, 
                                                                      Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                           annotation_name_gp = gpar(fontsize = 0, fontface = 'bold'), 
                                                           annotation_legend_param = list(
                                                             donor = list(title = "Donor",
                                                                          title_gp = gpar(fontsize = 15, fontface = "bold"), 
                                                                          labels_gp = gpar(fontsize = 15),
                                                                          direction = "horizontal",
                                                                          nrow = 3),
                                                             Status = list(title_gp = gpar(fontsize = 15, fontface = "bold"), 
                                                                           labels_gp = gpar(fontsize = 15),
                                                                           direction = "horizontal",
                                                                           nrow = 2)
                                                           )),
                     # column_split = length(nms), 
                     border = TRUE, 
                     left_annotation=rowAnnotation(df = fa_df[, c("Status","donor")],
                                                   col = list(donor = sample_c_cols, 
                                                              Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                   annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                   show_legend = F), 
                     # row_split = length(nms), 
                     row_split = split_df, 
                     cluster_column_slices =F,
                     cluster_row_slices =F,
                     row_gap = unit(split_gap_key, "mm"),
                     column_split = split_df, 
                     column_gap = unit(split_gap_key, "mm"),
                     column_title = split_donor_names,
                     column_title_gp = gpar(fontsize = 16),
                     column_title_side = c("bottom"),
                     column_title_rot = 90,
                     row_title = split_donor_names,
                     row_title_gp = gpar(fontsize = 16)
                     # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
  ) %>%
    # draw(., merge_legend = TRUE)
    draw(., newpage = F, merge_legends = T, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")
  
  return(sibd_pht)
}

```


```{r, fig.height=13, fig.width=15, eval=T}
hmap_ibd_a <- map(c(ss_ibd_cln_all_strain_hmmibd_f[c("all.stage_a_strain_qcd", "all.stage_g_strain_qcd")]), ~hmap_ibd_fn(input_gtyps = .x, strn_numbers = clustr_numbers_ag_m1_m2, strn_number_col = "strain_ag_n", str2remove_rowcolnms = "M[0-9]+_|_.$"))
```


```{r, fig.height=9.5, fig.width=7.5, eval=T}
hmap_ibd_a[[1]]
```


```{r, fig.height=16.5, fig.width=14, eval=T}
hmap_ibd_a[[2]]
```


```{r, fig.height=13, fig.width=15, eval=T}
hmap_ibd_ag <- map(c(ss_ibd_cln_all_strain_hmmibd_f[c("all.stage_ag_strain_qcd")]), ~hmap_ibd_fn(input_gtyps = .x, strn_numbers = clustr_numbers_ag_m1_m2, strn_number_col = "strain_ag_n", str2remove_rowcolnms = "M[0-9]+_"))
```


```{r, fig.height=19.5, fig.width=20, eval=T}
hmap_ibd_ag
```


```{r, fig.height=13, fig.width=15, eval=T}
hmap_ibd_s <- map(c(ss_ibd_cln_all_strain_hmmibd_f[c("all.strain_qcd")]), ~hmap_ibd_fn(input_gtyps = .x, strn_numbers = clustr_numbers_s_m1_m2, strn_number_col = "strain_n", str2remove_rowcolnms = "M[0-9]+_"))
```


```{r, fig.height=16, fig.width=15, eval=T}
hmap_ibd_s
```



```{r, fig.height=13, fig.width=15, eval=T}
ss_ibd_cln_all_strain_hmmibd_f_2 <- ss_ibd_cln_all_strain_hmmibd_f[["all.stage_ag_strain_qcd"]] %>% filter(str_detect(sample1, "M33|M49|M55|M60|M66")) %>% filter(str_detect(sample2, "M33|M49|M55|M60|M66"))

hmap_ibd_fn(input_gtyps = ss_ibd_cln_all_strain_hmmibd_f_2, strn_numbers = clustr_numbers_ag_m1_m2, strn_number_col = "strain_ag_n", str2remove_rowcolnms = "M[0-9]+_")
```


```{r, fig.height=13, fig.width=15, eval=T}
ss_ibd_cln_all_strain_hmmibd_f_polyc_asx <- ss_ibd_cln_all_strain_hmmibd_f[[c("all.stage_ag_strain_qcd")]] %>% filter(str_detect(sample1, "M33|M49|M60|M66") & str_detect(sample2, "M33|M49|M60|M66"))
```

```{r, fig.height=9.5, fig.width=8.5, eval=T}
hmap_ibd_fn(input_gtyps = ss_ibd_cln_all_strain_hmmibd_f_polyc_asx, strn_numbers = clustr_numbers_ag_m1_m2, strn_number_col = "strain_ag_n", str2remove_rowcolnms = "M[0-9]+_")
```


!! NOTE - UNNECESSARY BUT KEEP FOR TROUBLESHOOTING AND ALSO FOR CLUSTERING BETWEEN DONOR GROUPS - FOR ALL STAGES AND STRAINS
```{r, fig.height=10, fig.width=12, eval=T}
colFun = suppressWarnings(brewer.pal(100, 'Blues'))
colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), colFun)
##Merge strain numbers to IBD results
# ss_n_ibd_tbl100 = ss_cln_gt_tbl_100_hmmibd_f %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% left_join(., clustr_numbers, by = c('sample1'='sample_cluster')) %>% left_join(., clustr_numbers, by = c('sample2'='sample_cluster'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1, '\n[', n_1, ']'), id2 = paste0(sample2, '\n[', n_2, ']'))  

# ss_n_ibd_tbl100 = ss_ibd_cln_all_strain_hmmibd_f_2 %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% 
ss_n_ibd_tbl100 = ss_ibd_cln_all_strain_hmmibd_f_2 %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% 
  # left_join(., clustr_numbers_plusm14, by = c('sample1'='sample_cluster')) %>% 
  left_join(., clustr_numbers_s, by = c('sample2'='sample_cluster'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1), id2 = paste0(sample2))  

##Get the IBD and the number of sites for each pairwise comparison to annotate heatmap
s_ibdf_nsites_anno <- ss_n_ibd_tbl100 %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., ss_n_ibd_tbl100) %>% mutate(ibdf_nsites = paste0(round(fract_sites_IBD,2), '\n(',N_informative_sites,')'))%>% dplyr::select(starts_with('id'), ibdf_nsites) %>% pivot_wider(names_from = id2, values_from = ibdf_nsites) %>% arrange(id1) %>% column_to_rownames('id1')


##Create mirrored matrix and plot
sibd_pht_mtx <- ss_n_ibd_tbl100 %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., ss_n_ibd_tbl100) %>% dplyr::select(starts_with('id'), fract_sites_IBD) %>%  pivot_wider(names_from = id2, values_from = fract_sites_IBD) %>% arrange(id1) %>% column_to_rownames('id1')  
  

sibd_pht_mtx <- sibd_pht_mtx[,row.names(sibd_pht_mtx)]

# fa = str_extract(colnames(sibd_pht_mtx), 'M[0-9]+')
fa = str_remove(colnames(sibd_pht_mtx), '_SC.*')
c_d = cluster_within_group(sibd_pht_mtx, fa)
c_names = str_remove(colnames(sibd_pht_mtx), "M.*_") %>% set_names(colnames(sibd_pht_mtx))

# fr = str_extract(rownames(sibd_pht_mtx), 'M[0-9]+')
fr = str_remove(rownames(sibd_pht_mtx), '_SC.*')
print(fr)
c_r = cluster_within_group(sibd_pht_mtx, fr)
r_names = str_remove(row.names(sibd_pht_mtx), "M.*_") %>% set_names(row.names(sibd_pht_mtx))

# create annotation dataframe
fa_df <- data.frame("rname" = colnames(sibd_pht_mtx), "donor" = str_remove(colnames(sibd_pht_mtx), '_SC.*')) %>% left_join(all_pf_status, by = "donor") %>% column_to_rownames("rname")

# nms = unique(str_extract(colnames(sibd_pht_mtx), 'M[0-9]+'))
nms = unique(str_remove(colnames(sibd_pht_mtx), '_SC.*'))
# sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
# sample_c_cols <- cols_rand[1:length(nms)] %>% set_names(nms)
sample_c_cols <- donor_cols

sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()

  
sibd_pht<- Heatmap(sibd_pht_mtx_cln, name='IBD', 
                   col = colFun, 
                   show_row_names = TRUE, 
                   row_labels = r_names,
                   row_names_side = c("left"), 
                   show_column_names = TRUE,  
                   column_labels = c_names,
                   show_row_dend = T,  
                   show_column_dend = F, 
                   row_title_rot = 0, 
                   cluster_rows = c_r, 
                   cluster_columns = c_d, 
                   column_names_gp = gpar(fontsize = 15), 
                   row_names_gp = gpar(fontsize = 15), 
                   heatmap_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15)),
                   bottom_annotation = HeatmapAnnotation(df = fa_df,
                                                 col = list(donor = sample_c_cols, 
                                                            Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                         annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                         annotation_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15))),
                   column_split = length(nms), 
                   border = TRUE, 
                   left_annotation=rowAnnotation(df = fa_df,
                                                 col = list(donor = sample_c_cols, 
                                                            Status = c("Asymptomatic" = "lightblue", "Symptomatic" = "lightgreen") ),
                                                 annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                 show_legend = F), 
                   row_split = length(nms), 
                   column_title = NULL,
                   column_title_gp = gpar(fontsize = 16), 
                   row_title = NULL,
                   row_title_gp = gpar(fontsize = 16)
              # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
              ) %>%
  draw(., merge_legend = TRUE)


```


Compare hmmibd with dcifer
```{r}
# dcifer p val and estimate for ibd
dres0_pm[, , "estimate"] %>% data.frame() %>% rownames_to_column("nm1") %>% pivot_longer(cols = starts_with("M"), names_to = "nm2", values_to = "estimate", values_drop_na = T) %>%
  left_join(., dres0_pm[, , "p_value"] %>% data.frame() %>% rownames_to_column("nm1") %>% pivot_longer(cols = starts_with("M"), names_to = "nm2", values_to = "p_value", values_drop_na = T))

```

```{r}
# hmmibd p val and estimate for ibd

dcf_ibd_p <- dres0_pm[, , "estimate"] %>% data.frame() %>% rownames_to_column("sample2") %>% pivot_longer(cols = starts_with("M"), names_to = "sample1", values_to = "estimate", values_drop_na = T) %>%
  left_join(., dres0_pm[, , "p_value"] %>% data.frame() %>% rownames_to_column("sample2") %>% pivot_longer(cols = starts_with("M"), names_to = "sample1", values_to = "p_value", values_drop_na = T), by = c("sample1", "sample2"))

```


```{r}
# hmmibd p val and estimate for ibd
ss_ibd_cln_all_strain_hmmibd_f[["all.strain_qcd"]] %>% select(sample1, sample2, log_p, fract_sites_IBD) %>% left_join(dcf_ibd_p, by = c("sample1", "sample2")) %>% ggplot(aes(x= fract_sites_IBD, y = estimate)) + geom_point()

```

```{r}
# hmmibd p val and estimate for ibd
ss_ibd_cln_all_strain_hmmibd_f[["all.strain_qcd"]] %>% select(sample1, sample2, log_p, fract_sites_IBD) %>% left_join(dcf_ibd_p, by = c("sample1", "sample2")) %>% filter(p_value < 0.05) %>% ggplot(aes(x= log_p, y = log(p_value))) + geom_point()

```

```{r, fig.width=3, fig.height=3}
# hmmibd p val and estimate for ibd
ss_ibd_cln_all_strain_hmmibd_f[["all.strain_qcd"]] %>% select(sample1, sample2, log_p, fract_sites_IBD) %>% left_join(dcf_ibd_p, by = c("sample1", "sample2")) %>% filter(p_value < 0.05) %>% ggplot(aes(x= fract_sites_IBD, y = estimate)) + geom_point() + theme_classic()

```


```{r}
# hmmibd p val and estimate for ibd
ss_ibd_cln_all_strain_hmmibd_f[["all.strain_qcd"]] %>% select(sample1, sample2, log_p, fract_sites_IBD) %>% left_join(dcf_ibd_p, by = c("sample1", "sample2")) %>% ggplot(aes(x= fract_sites_IBD, y = estimate)) + geom_point()

```


```{r, fig.height=10, fig.width=12.5, eval=FALSE}
gogo()
```



```{r, fig.height=10, fig.width=12.5, eval=FALSE}
sibd_pht

Legend(labels = month.name[1:10], legend_gp = gpar(fill = 1:10), 
    title = "foo", nrow = 3)
```


```{r, fig.height=10, fig.width=12.5, eval=FALSE}


sibd_pht

Legend(labels = month.name[1:10], legend_gp = gpar(fill = 1:10), 
    title = "foo", nrow = 3)
```


```{r, fig.height=5, fig.width=6, eval=FALSE}
##Merge strain numbers to IBD results
s_strn_n_ibd_tbl = s_k6r_pf7M_cln_gt_hmmibd_f %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% left_join(., strn_qcd_n, by = c('sample1'='strain_qcd')) %>% left_join(., strn_qcd_n, by = c('sample2'='strain_qcd'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1, '\n[', n_1, ']'), id2 = paste0(sample2, '\n[', n_2, ']'))  

##Get the IBD and the number of sites for each pairwise comparison to annotate heatmap
s_combi_ibdf_nsites_anno <- s_strn_n_ibd_tbl %>% 
  rename('id2' = id1,  'id1' = id2) %>% 
  bind_rows(., s_lqc_strn_n_ibd_tbl) %>% 
  mutate(ibdf_nsites = paste0(round(fract_sites_IBD,2), '\n(',N_informative_sites,')'))%>% 
  dplyr::select(starts_with('id'), ibdf_nsites) %>% 
  pivot_wider(names_from = id2, values_from = ibdf_nsites) %>% 
  arrange(id1) %>%
  mutate(across(starts_with('SC'), ~replace_na(., '1')))%>%
  column_to_rownames('id1')


##Create mirrored matrix and plot
##!!! TIP - Refer to this if you want more complex arrangements https://github.com/jokergoo/ComplexHeatmap/issues/712 - eg different colours for the different half matrices
s_combi_ibd_pht <- s_strn_n_ibd_tbl %>% 
  rename('id2' = id1,  'id1' = id2) %>% 
  bind_rows(., s_lqc_strn_n_ibd_tbl) %>% 
  dplyr::select(starts_with('id'), fract_sites_IBD) %>%  
  pivot_wider(names_from = id2, values_from = fract_sites_IBD) %>% 
  arrange(id1) %>% 
  mutate(across(starts_with('SC'), ~replace_na(., 1)))%>%
  column_to_rownames('id1') %>% 
  as.matrix()%>%
  Heatmap(., name='IBD', col = colFun, show_row_names = TRUE, row_names_side = c("left"),
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = F, cluster_columns = F, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), 
              heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),border='black',
              cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%s", s_combi_ibdf_nsites_anno[i, j]), x, y, gp = gpar(fontsize = 13))
})

s_combi_ibd_pht
  
```

#########€#€######€€€€€€€

```{r}
# m = matrix(rnorm(120), nc = 12)
# colnames(m) = 1:12
# fa = rep(c("a", "b", "c"), times = c(2, 4, 6))
# # fa_col = c("a" = 2, "b" = 3, "c" = 4)
# dend1 = cluster_between_groups(m, fa)
# 
# Heatmap(m, cluster_columns = dend1, 
#     row_title = "cluster_between_groups")
# 
# m
# fa
# fa_col
```


```{r, fig.height=4.1, fig.width=4.5}

# ss_afm_hmap_ibs<- map(ss_afm_pf7M_cln_gt, ~gtypes_chr_cln_ibs_hm(gtypes_chr_pos=.x))
```

```{r, fig.height=4.1, fig.width=4.5}
## Remove SC8
# ss_afm_pf7M_cln_gt_m14_noSC8 <- ss_afm_pf7M_cln_gt[c("MSC14.stage_ag_strain_qcd", "MSC14.strain_qcd")] %>% map(~.x %>% select(!contains("SC8")))
# ss_afm_hmap_ibs_m14_noSC8 <- map(ss_afm_pf7M_cln_gt_m14_noSC8, ~gtypes_chr_cln_ibs_hm(gtypes_chr_pos=.x))
# 
# ss_afm_hmap_ibs_m14_noSC8
```


```{r, fig.height=4.1, fig.width=4.5}
## Loci with a genotype in every sample
ss_afm_pf7M_cln_gt[[1]] %>% filter(if_all(starts_with("M"), ~!is.na(.)))

## Loci where all samples have missing genotype or the 0 allele
ss_afm_pf7M_cln_gt[[1]] %>% filter(if_all(starts_with("M"), ~is.na(.)| . == "0"))

## Loci where all samples have missing genotype or the 1 allele
ss_afm_pf7M_cln_gt[[1]] %>% filter(if_all(starts_with("M"), ~is.na(.)| . == "1"))

```


Manual chromosome paints

```{r}
##Function to creat pairwise comparison groups which are the used for plotting chromosome paints
strn_pw_comps <- function(gtypes_chr_pos, strn_regex){
  
  snp_pf7_afm <- gtypes_chr_pos %>% dplyr::select(id, chr, pos, starts_with(strn_regex)) #%>%  dplyr::select(!matches('^SC[0-9]_.$')) 
  
  col_pairs_strn <- snp_pf7_afm %>% dplyr::select(starts_with(strn_regex)) %>% colnames() %>% sort() %>% combn(., 2, simplify = F)
  
  strn_afm_comps <- map_dfc(col_pairs_strn, function(pair) {
    snp_pf7_afm %>%
      transmute(!!paste0(pair[1], "_vs_", pair[2]) := case_when(!is.na(.data[[pair[1]]]) & !is.na(.data[[pair[2]]]) & .data[[pair[1]]] == .data[[pair[2]]] ~ TRUE, 
                                                                !is.na(.data[[pair[1]]]) & !is.na(.data[[pair[2]]]) & .data[[pair[1]]] != .data[[pair[2]]] ~ FALSE))
  })
  
  strn_afm_comps <- base::cbind(snp_pf7_afm %>% dplyr::select(chr, pos), 
                                         strn_afm_comps)  %>%
    filter(!if_all(starts_with(strn_regex), ~is.na(.)))
  
  return(strn_afm_comps)
}
```

```{r}
##stage strain pairwise comparison groups
ss_afm_pf7M_afm_comps <- map(ss_afm_pf7M_cln_gt, possibly(~strn_pw_comps(gtypes_chr_pos = .x, strn_regex = c('M','N')), "only one comparison group"))

```

```{r}
##Function to create binned windows across genome and compare pairwise strains and generate the plotting details for chromosome painting
b_strn_c_wndw_fn <- function(strn_c_tbl, wndw = 100000, strn_regex = c('SC','N')) {
  strn_c_tbl %>% 
    group_by(chr) %>% 
    filter(!is.na(chr)) %>%
    mutate(pos_range = cut(pos, breaks = c(-Inf,seq(0, max(pos), by = wndw),Inf))) %>% 
    pivot_longer(., cols = starts_with(strn_regex), names_to = 'comps', values_to = 'sim') %>% 
    dplyr::count(comps, chr, pos_range, sim) %>% 
    group_by(chr, comps, pos_range) %>% 
    mutate(overlap_TF = ifelse(any(is.na(sim)), sum(!is.na(sim)), n_distinct(sim)))%>%
    ungroup() %>% 
    filter(!(is.na(sim) & overlap_TF>0))  %>% 
    mutate(rltd = case_when(overlap_TF == 0 ~ 'No SNP', overlap_TF == 2 ~ 'Both', overlap_TF==1& sim==TRUE ~ 'Same', overlap_TF==1&sim==FALSE ~ 'Different')) %>% 
    group_by(chr, comps, pos_range, rltd) %>% 
    arrange(dplyr::desc(sim), .by_group=T) %>%
    summarise(overlap_snps_No = paste(n, collapse = ","), 
              overlap_snps_prop = log10(n[1]/n[2]))  %>% 
    mutate(across(overlap_snps_No, ~ifelse(rltd == 'No SNP', NA, .))) %>%
    ungroup()  %>%
    mutate(normalized_snps_prop = case_when(rltd == "Same" ~ max(overlap_snps_prop, na.rm = TRUE), 
                                            rltd == "Different" ~ min(overlap_snps_prop, na.rm = TRUE), 
                                            # rltd == 'No SNP' ~ NA, 
                                            T ~ overlap_snps_prop)) %>%
    mutate(start = as.numeric(str_extract(pos_range, "\\d+.*(?=,)")), end = as.numeric(str_extract(pos_range, "(?<=,)\\d+.*(?=])"))) %>% 
    left_join(.,pf_chr_size, by =c('chr')) %>%
    mutate(across(chr, ~factor(., levels = c(1:14)))) 
}
```


```{r}
## Get genome size like this cut -f1,2 fasta.fai > genome.size
pf_chr_size = read_delim(chrm_sz, col_names = c('chrom', 'size')) %>% mutate(chr = str_remove_all(chrom,chrm_prfx)) 
```

```{r}
##stage strain pairwise comparison chromosome binning of SNPs and creation of windows for plotting
ss_afm_wndws <- map(ss_afm_pf7M_afm_comps, possibly(~b_strn_c_wndw_fn(strn_c_tbl = .x, wndw = 30000, strn_regex = c('M','N')), "only one group"))

```

IBD and plots generation


##Simpler chromosome plots(no continous gradient) for SS_AG and S

```{r}
## Function for chromosome painting

chrom_paintr <- function(gtypes_comps_tbl = b_strn_c, ttl, expn=10000, x_font_sz = 13) {

      gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+
      geom_segment(
    aes(x = 0-(2000/expn), xend = size+(2000/expn), y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.8, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = rltd
    ),
    linewidth = 2.8
  ) +
  # scale_color_manual(values = c('Same'= '#D4AF37', 'Different'= '#2C3E50','Both'='lightgrey', 'No SNP'='white'))+      
  scale_color_manual(values = c('Same'= '#DACB7D', 'Different'= '#2C3E50','Both'='#A9B7C5', 'No SNP'='white'))+ ##prev colours "black" ,"#ffd94d" 
  # facet_wrap(comps~., ncol = 5) +
      labs(title = ttl) +
      labs(colour = "Allele \ncomparison", x = 'Position (bases)', y ="Chromosome") +
  theme_classic() +
  theme(#strip.text = element_text(size = 9, face = 'bold'),
        axis.text.y = element_text(size =  13),
        axis.text.x = element_text(size =  x_font_sz),
        # axis.title = element_text(size = 9),
        axis.title = element_blank(),
        # legend.title = element_text(size = 9, face = "bold"),
        # legend.text = element_text(size = 9),
        title = element_text(size = 12, face = "bold"),
        legend.position = "none") 
    
    pie_bar_tbl <- gtypes_comps_tbl %>% 
      separate(.,overlap_snps_No, into=(c('same2', 'diff2')), sep=',' ) %>% 
      pivot_longer(cols = c('same2','diff2'), names_to = 'shared2', values_to = 'shared_n') %>% 
      filter(!is.na(shared_n)) %>%
      mutate('Shared' = case_when(rltd == 'Both' & shared2 == 'same2' ~ 'Same', 
                                  rltd == 'Both' & shared2 == 'diff2' ~ 'Different', 
                                  rltd == 'Different' & shared2 == 'same2' ~ 'Different', 
                                  T ~ rltd)) %>%
      mutate(across(shared_n, as.numeric))
    
    pp <- pie_bar_tbl %>% 
      group_by(Shared) %>% 
      summarise('snpsNo' = sum(shared_n))%>%
    arrange(dplyr::desc(snpsNo), .by_group = T) %>%
    #mutate(lab.ypos = cumsum(prop) - 0.5*prop) %>%
    ggplot(., aes(x="", y=snpsNo, fill=Shared))+
    geom_bar(width = 1, stat = "identity") + 
    coord_polar("y", start=0)+
    # geom_text(aes(y = snpsNo/2 + c(0, cumsum(snpsNo)[-length(snpsNo)]), 
    #               label = snpsNo), size=3, colour = "#90c7ff") +
      geom_label(aes(y = snpsNo/2 + c(0, cumsum(snpsNo)[-length(snpsNo)]), 
                  label = snpsNo), size=5, label.padding = unit(0.07, "lines"), label.size = 0, color = 'black', fill = 'white') +
      scale_fill_manual(values = c('Same'= '#DACB7D', 'Different'= '#2C3E50'))+ #'#D4AF37'
    # labs(title = .y)+
    theme_void()+
      theme(legend.position = "none")
  #theme_minimal_hgrid(11)

    
    ggdraw(gs ) +
                draw_plot(pp, x=.5, y=.1, width = .50, height= .50)

  }

```

comparisons with ibd >0.1
```{r, warning=F}
grp_ibd_1 <- ss_ibd_cln_all_strain_hmmibd_f_2 %>% filter(fract_sites_IBD>0.1) %>% mutate(grp = paste0(sample1, "_vs_" , sample2)) %>% pull(grp)

```
strains split by stage Without colour gradients
```{r, warning=F}
# strain stage (afm) chromosome painting
# ss_afm_chromp <- map(ss_afm_wndws, possibly(~{
#   smp_tbl <- .x
#   smp_tbl %>% split(., ~factor(comps)) %>% .[grp_ibd_1] %>% imap(.,possibly(~chrom_paintr(gtypes_comps_tbl =.x, ttl=.y), 'No overlapping SNPs between the 2 groups'))
# }, 'No overlapping SNPs between the 2 groups'))

ss_afm_chromp <- ss_afm_wndws[[1]] %>% split(., ~factor(comps)) %>% .[grp_ibd_1]  %>% imap(.,possibly(~chrom_paintr(gtypes_comps_tbl =.x, ttl=.y), 'No overlapping SNPs between the 2 groups'))

```



```{r}
# strain stage (afm) chromosome painting
ss_afm_chromp
```

```{r}

# plot_grid(plotlist = ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC1_A_vs_SC2_A", "SC1_A_vs_SC2_A", "SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G", "SC6_A_vs_SC7_G", "SC6_G_vs_SC7_G")], ncol = 3)
# plot_grid(plotlist = plts_strn_ss[c("SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G")], ncol = 2)
```


```{r}

# plot_grid(plotlist = ss_k6r_ag_wndws[c("SC1_A_vs_SC2A", "SC1_A_vs_SC2_G", "SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G", "SC6_A_vs_SC7_G", "SC6_G_vs_SC7_G")], ncol = 3)
# plot_grid(plotlist = ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G")], ncol = 1)
```

```{r, fig.height=6, fig.width=12}
# plot_grid(plotlist = ss_afm_chromp[["MSC14.strain_qcd"]], ncol = 7)
```

```{r, warning=FALSE, eval=FALSE}
# Check M55 comparisons
# m55_comps <- ss_afm_wndws[[2]] %>% filter(str_detect(comps, "M55_SC._._vs_M55_SC._.")) 
m55_comps <- ss_afm_wndws[[2]] %>% filter(str_detect(comps, "M55_SC._._vs_M60_SC._.|M60_SC._._vs_M55_SC._.")) 

m55_comps %>% count(comps)
```

```{r, warning=FALSE, eval=FALSE}
# strain stage (afm) chromosome painting

ssmth_afm_chromp_m55 <- ss_afm_wndws[[2]] %>% filter(comps %in% unique(m55_comps$comps)) %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr(gtypes_comps_tbl =.x, ttl=.y), 'No overlapping SNPs between the 2 groups'))

```


```{r, warning=FALSE, eval=FALSE}
ssmth_afm_chromp_m55[1:5]
```



```{r, warning=FALSE, eval=FALSE}

ssmth_afm_chromp_m55
```

strains split by stage Without colour gradients
```{r, eval=FALSE}
# strain stage (afm) chromosome painting
s_chromp_xlab_mod <- map(ss_afm_wndws[str_detect(names(ss_afm_wndws), "\\.strain_qcd")], ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr(gtypes_comps_tbl =.x, ttl=.y, x_font_sz = 0), 'No overlapping SNPs between the 2 groups'))
})

```


```{r, eval=FALSE}
## Function for chromosome painting

chrom_paintr_smth_gradient <- function(gtypes_comps_tbl = b_strn, ttl, expn=100000, x_font_sz = 13, lg = "right") {

      # gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+
        geom_segment(
    aes(x = 0-2000/expn, xend = size+2000/expn, y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.5, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = normalized_snps_prop#,
      # colour = rltd
    ),
    linewidth = 2.5
  ) +
  # scale_color_manual(values = c('Same'='black', 'Different'='#ffd94d', 'Both'='lightgrey', 'No SNP'='white'))+
  scale_color_gradient2(low="black" , mid = "#F5F5F5", high ="#ffd94d" ,
                        limits = c(min(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), max(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE)),
                        breaks = seq(min(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), max(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), length.out = 30),
                        oob = scales::squish,
                        na.value = "white",
                        midpoint = 0,
                        guide = lg) +
  # geom_label_repel(aes(label=new_column, x=(start + end)/2, y = comps), size=3,label.padding = unit(0.01, "lines"),alpha = 1,label.size = 0.01, color = alpha('black', .2))+
  # facet_wrap(comps~., ncol = 5) ++
      labs(title = ttl) +
  theme_classic() +
  theme(strip.text = element_text(size = 12, face = 'bold'),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size =  x_font_sz)) 
    
    pp <- gtypes_comps_tbl %>% 
      separate(.,overlap_snps_No, into=(c('same2', 'diff2')), sep=',' ) %>% 
      pivot_longer(cols = c('same2','diff2'), names_to = 'shared2', values_to = 'shared_n') %>% 
      filter(!is.na(shared_n)) %>%
      mutate('Shared' = case_when(rltd == 'Both' & shared2 == 'same2' ~ 'Same', 
                                  rltd == 'Both' & shared2 == 'diff2' ~ 'Different', 
                                  rltd == 'Different' & shared2 == 'same2' ~ 'Different', 
                                  T ~ rltd)) %>%
      mutate(across(shared_n, as.numeric)) %>%
      group_by(Shared) %>% 
      summarise('snpsNo' = sum(shared_n))%>%
      arrange(dplyr::desc(snpsNo), .by_group = T) %>%
    ggplot(., aes(x="", y=snpsNo, fill=Shared))+
    geom_bar(width = 1, stat = "identity") + 
    coord_polar("y", start=0)+
    geom_text(aes(y = snpsNo/2 + c(0, cumsum(snpsNo)[-length(snpsNo)]), 
                  label = snpsNo), size=5) +
    theme_void() +
      theme(legend.position = lg)

    ggdraw(gs + theme_half_open(12)) +
      draw_plot(pp, .5, .11, .36, .36) 
}


```


strains split by stage Without colour gradients
```{r, warning=FALSE, eval=FALSE}
# strain stage (afm) chromosome painting

ssmth_afm_chromp <- ss_afm_wndws[[1]] %>% split(., ~factor(comps)) %>% .[grp_ibd_1]  %>% imap(.,possibly(~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))

```



```{r}

# k1_ibd_sites_f <- list(gt_fb_vtx_p1_bi_strn_pf7_asx_gam_ibd, gfv_p1_bi_pf7_k7_all_ibd) %>%
# k8_ibd_sites_f <- list(ss_ag_k6r_pf7M_cln_gt_hmmibd, s_k6r_pf7M_cln_gt_hmmibd) %>%
# k8_ibd_sites_f <- map(ss_afm_pf7M_cln_gt_hmmibd[7:8], ~{
k8_ibd_sites_f <- ss_ibd_cln_all_strain_hmmibd_2%>% 
  mutate(across(chr,as.character)) %>% 
  left_join(.,pf_chr_size, by ='chr') %>% 
  mutate(across(chr, ~factor(., levels = c(1:14)))) %>%
  unite(., 'comps',sample1, sample2, sep='vs', remove = F) %>%  
  group_by(chr) %>% 
  mutate(chr_s = 0, chr_e = size) %>% 
  ungroup() %>% 
  mutate(across(different, as.character))

```



```{r}
## Function for chromosome painting

chrom_paintr_ibd <- function(gtypes_comps_tbl, ttl, expn= 100000, x_font_sz = 13) {
  gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+ 
      geom_segment(
    aes(x = 0-(2000/expn), xend = size+(2000/expn), y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.8, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = rltd
    ),
    linewidth = 2.8
  ) +
  # scale_color_manual(values = c('Same'= '#D4AF37', 'Different'= '#2C3E50','Both'='lightgrey', 'No SNP'='white'))+ ##prev colours "black" ,"#ffd94d" 
  scale_color_manual(values = c('Same'= '#DACB7D', 'Different'= '#2C3E50','Both'='#A9B7C5', 'No SNP'='white'))+
  # facet_wrap(comps~., ncol = 5) +
      labs(title = ttl) +
      labs(colour = "Allele \ncomparison", x = 'Position (bases)', y ="Chromosome") +
  theme_classic() +
  theme(axis.text.y = element_text(size =  13),
        axis.text.x = element_text(size =  x_font_sz),
        # axis.ticks.x = element_blank(),
        # axis.title = element_text(size = 9),
        axis.title = element_blank(),
        # legend.title = element_text(size = 9, face = "bold"),
        # legend.text = element_text(size = 9),
        title = element_text(size = 12, face = "bold"),
        legend.position = "none",
        plot.background = element_rect(fill = "#bfe5fc")) 

  }


```


```{r, fig.height=14, fig.width=24}
ss_afm_k6r_pf7M_ibd <- k8_ibd_sites_f %>% mutate(across(comps, ~str_replace(., 'vs', '_vs_')), rltd = case_when(different == '0' ~ 'Same', different == '1' ~ 'Different')) 

ss_afm_k6r_pf7M_ibd_p <- ss_afm_k6r_pf7M_ibd %>% split(., ~factor(comps)) 
```


```{r, fig.height=14, fig.width=24}
ss_afm_k6r_pf7M_ibd_plts <- ss_afm_k6r_pf7M_ibd_p %>% .[grp_ibd_1] %>% imap(., ~chrom_paintr_ibd(.x, .y))
```

```{r, fig.height=4, fig.width=6}
ss_afm_k6r_pf7M_ibd_plts[1]
```



```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[1:10]
# plot_grid(plotlist = c(ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC3_A_vs_SC4_G")], ss_ag_k6r_pf7M_ibd_p[c("SC3_A_vs_SC4_G")]), ncol = 2)
```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[str_detect(names(ss_afm_k6r_pf7M_ibd_p), "M60.*M68")]
```




```{r}


# b_ovrl <-map(ss_afm_pf7M_afm_comps[str_detect(names(ss_afm_pf7M_afm_comps), "\\.strain_qcd")], ~.x %>% 
b_ovrl <-map(ss_afm_pf7M_afm_comps, ~.x %>% 
    group_by(chr) %>% 
    filter(!is.na(chr)) %>%
    mutate(pos_range = cut(pos, breaks = c(-Inf,seq(0, max(pos), by = 100000),Inf)))  %>% 
    dplyr::group_by(chr, pos_range) %>%
    summarise(SNPs_sum = n()) %>%
    ungroup() %>%
    left_join(.,pf_chr_size, by =c('chr')) %>%
    mutate(across(chr, ~factor(., levels = c(1:14)))) %>%
    mutate(start = as.numeric(str_extract(pos_range, "\\d+.*(?=,)")), end = as.numeric(str_extract(pos_range, "(?<=,)\\d+.*(?=])"))) %>%
    mutate(across(c(end), ~case_when(is.na(.) ~ size, T ~ .)))
)

```


```{r, fig.width=14, fig.height=6.5}

plt_ovrl <- b_ovrl %>%
  map(~{ .x %>% ggplot() + 
      geom_segment(
    aes(x = 0-2000, xend = size+2000, y = chr, yend = chr),
    linewidth = 7.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 6.5, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr
    ),
      colour = "#C9E0F1",
    linewidth = 6.5
  ) +
  geom_label(aes(label=SNPs_sum, x=(start + end)/2, y = chr), size=4, label.padding = unit(0.05, "lines"), fontface = 'bold', label.size = 0, color = alpha('black', .9))+
  # facet_wrap(comps~., ncol = 5) +
      labs(y = 'Chromosome', x = 'Position (bases)') +
  theme_classic() +
  theme(strip.text = element_text(size = 12, face = 'bold'),
        axis.text = element_text(size = 18),
        axis.title = element_text(size = 20, face = 'bold'),
        title = element_text(size = 15, face = "bold"))
})

plt_ovrl


```



######Moved IBD codes from above to below here - may create complications
hmmIBD START

```{r}
##!!NOTE - gfv_p1_bi_strn_pf7_c is created somewhere downstream so this is likely to give an error

system(paste0('mkdir data/processed/ibd/'))
system(paste0('rm data/processed/ibd/*'))

imap(ss_afm_pf7M_cln_gt[!str_detect(names(ss_afm_pf7M_cln_gt), "\\.stage_afm_strain_qcd")], ~{
  .x %>% 
    dplyr::select(-c(id, chrom, alleles))%>% 
    dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% 
    filter(!if_all(starts_with('SC'), ~is.na(.))) %>% 
    mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% 
    write_tsv(., paste0("data/processed/ibd/", .y, ".tsv"))
})

imap(ss_afm_pf7M_cln_gt_lqc[!str_detect(names(ss_afm_pf7M_cln_gt_lqc), "\\.stage_afm_strain_qcd")], ~{
  .x %>% 
    dplyr::select(-c(id, chrom, alleles))%>% 
    dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% 
    filter(!if_all(starts_with('SC'), ~is.na(.))) %>% 
    mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% 
    write_tsv(., paste0("data/processed/ibd/", .y, "_lqc.tsv"))
})

## Write out the combined snps across all samples for ibd
imap(ss_cln_all_high_n_mtx[3], ~{
  .x %>% 
    dplyr::select(-c(id, chrom, alleles))%>% 
    dplyr::select(chr, pos, any_of(sort(colnames(.)))) %>% 
    filter(!if_all(starts_with('MSC14_qcd_SC'), ~is.na(.))) %>% 
    mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% 
    write_tsv(., paste0("data/processed/ibd/", .y, ".tsv"))
})

# ##!!!NOTE - Remove SC8 first - use this to include SC8 dplyr::select(chr, pos, paste0('SC', 1:8))
# s_k6r_pf7M_cln_gt %>% dplyr::select(-c(id, chrom, alleles))%>% dplyr::select(chr, pos, any_of(sort(colnames(.))))  %>% filter(!if_all(starts_with('SC'), ~is.na(.))) %>% mutate(across(everything(), ~as.numeric(replace_na(., '-1')))) %>% write_tsv(., "/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master/s_k6r_pf7M_cln_gt.tsv")

```


```{r}
##Run IBD 
##!!!!!!NOTE - sort out the output file -o argument - failure to save in the same directory. Goes to working dir

# imap(ss_afm_pf7M_cln_gt, ~system(paste0('hmmIBD -i data/processed/Pf/', str_remove(.y, "\\.st.*"), '/ibd/', .y, '.tsv -o data/processed/Pf/', str_remove(.y, "\\.st.*"), '/ibd/', .y, ' -m 100000')))

imap(ss_afm_pf7M_cln_gt[!str_detect(names(ss_afm_pf7M_cln_gt), "\\.stage_afm_strain_qcd")], ~system(paste0('hmmIBD -i data/processed/ibd/', .y, '.tsv -o data/processed/ibd/', .y, ' -m 100000')))
imap(ss_afm_pf7M_cln_gt_lqc[!str_detect(names(ss_afm_pf7M_cln_gt_lqc), "\\.stage_afm_strain_qcd")], ~system(paste0('hmmIBD -i data/processed/ibd/', .y, '_lqc.tsv -o data/processed/ibd/', .y, '_lqc -m 100000')))

## IBD of genotypes combined snps across all samples
imap(ss_cln_all_high_n_mtx[3], ~system(paste0('hmmIBD -i data/processed/ibd/', .y, '.tsv -o data/processed/ibd/', .y, ' -m 100000')))

# system('/Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/sware/hmmIBD-master/hmmIBD -i /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/sware/hmmIBD-master/ss_afm_pf7M_cln_gt.tsv -o ss_afm_pf7M_cln_gt -m 100000')
  
```


```{r}
# wdir<- '/Users/jr35/Google Drive/My Drive/PHD_documents/sware/hmmIBD-master'

ss_afm_pf7M_cln_gt_hmmibd <- map(names(ss_afm_pf7M_cln_gt[!str_detect(names(ss_afm_pf7M_cln_gt), "\\.stage_afm_strain_qcd")]), ~read_delim(paste0('data/processed/ibd/', .x, '.hmm.txt'))) %>% set_names(names(ss_afm_pf7M_cln_gt[!str_detect(names(ss_afm_pf7M_cln_gt), "\\.stage_afm_strain_qcd")]))
ss_afm_pf7M_cln_gt_hmmibd_f <- map(names(ss_afm_pf7M_cln_gt[!str_detect(names(ss_afm_pf7M_cln_gt), "\\.stage_afm_strain_qcd")]), ~read_delim(paste0('data/processed/ibd/', .x, '.hmm_fract.txt'))) %>% set_names(names(ss_afm_pf7M_cln_gt[!str_detect(names(ss_afm_pf7M_cln_gt), "\\.stage_afm_strain_qcd")]))

ss_afm_pf7M_cln_gt_hmmibd_lqc <- map(names(ss_afm_pf7M_cln_gt_lqc[!str_detect(names(ss_afm_pf7M_cln_gt_lqc), "\\.stage_afm_strain_qcd")]), ~read_delim(paste0('data/processed/ibd/', .x, '_lqc.hmm.txt'))) %>% set_names(names(ss_afm_pf7M_cln_gt_lqc[!str_detect(names(ss_afm_pf7M_cln_gt_lqc), "\\.stage_afm_strain_qcd")]))
ss_afm_pf7M_cln_gt_hmmibd_f_lqc <- map(names(ss_afm_pf7M_cln_gt_lqc[!str_detect(names(ss_afm_pf7M_cln_gt_lqc), "\\.stage_afm_strain_qcd")]), ~read_delim(paste0('data/processed/ibd/', .x, '_lqc.hmm_fract.txt'))) %>% set_names(names(ss_afm_pf7M_cln_gt_lqc[!str_detect(names(ss_afm_pf7M_cln_gt_lqc), "\\.stage_afm_strain_qcd")]))

## IBD of genotypes combined snps across all samples
all_donors_hmmibd <- map(names(ss_cln_all_high_n_mtx[3]), ~read_delim(paste0('data/processed/ibd/', .x, '.hmm.txt'))) %>% set_names(names(ss_cln_all_high_n_mtx[3]))
all_donors_hmmibd_f <- map(names(ss_cln_all_high_n_mtx[3]), ~read_delim(paste0('data/processed/ibd/', .x, '.hmm_fract.txt'))) %>% set_names(names(ss_cln_all_high_n_mtx[3]))


```


```{r}
# ## Remove SC8 from the msc14 results
# ss_afm_pf7M_cln_gt_hmmibd_m14 <- ss_afm_pf7M_cln_gt_hmmibd[[8]] %>% filter(sample1 != "SC8" & sample2 != "SC8" )
# ss_afm_pf7M_cln_gt_hmmibd_f_m14 <- ss_afm_pf7M_cln_gt_hmmibd_f[[8]] %>% filter(sample1 != "SC8" & sample2 != "SC8" )
# 
# ss_afm_pf7M_cln_gt_hmmibd_lqc_m14 <- ss_afm_pf7M_cln_gt_hmmibd_lqc[[8]] %>% filter(sample1 != "SC8" & sample2 != "SC8" )
# ss_afm_pf7M_cln_gt_hmmibd_f_lqc_m14 <- ss_afm_pf7M_cln_gt_hmmibd_f_lqc[[8]] %>% filter(sample1 != "SC8" & sample2 != "SC8" )

```


Export the SNPs with QC metrics and only clean SNPs for stage&strain combinations and just strain
```{r, eval=T}
##!!!! NOTE -MANUSCRIPT IBD table

## subset IBD tables to write and rename lists/sheets 
### IBD results
ibd_export <- ss_afm_pf7M_cln_gt_hmmibd_f[str_detect(names(ss_afm_pf7M_cln_gt_hmmibd_f), "\\.strain_qcd")] %>% map(., ~rename_with(., ~str_replace(.,'sample', 'Strain'), contains("sample")))
names(ibd_export) <- paste0(2:(length(ibd_export)+1), ".", str_replace_all(names(ibd_export), c("\\.strain_qcd" = " IBD", "MSC" = "MSC")))

### IBD summary
ibd_f_export <- ss_afm_pf7M_cln_gt_hmmibd[str_detect(names(ss_afm_pf7M_cln_gt_hmmibd), "\\.strain_qcd")] %>% map(., ~rename_with(., ~str_replace(.,'sample', 'Strain'), contains("sample")))
names(ibd_f_export) <- paste0(6:(length(ibd_f_export)+5), ".", str_replace_all(names(ibd_f_export), c("\\.strain_qcd" = " IBD sites summary", "MSC" = "MSC")))


createWorksheet(c(ibd_export, ibd_f_export), paste0("../",sample_set,"/v3_science_manuscript/tables/data S10 - Strain pairwise IBD.xlsx"))
```


```{r}
## Read in msc annotations 

# msc14_qcd_anotd <- readRDS('data/processed/msc14/msc_qcd_anotd.RDS') %>% .[[1]]

```


```{r, fig.height=5, fig.width=6}
colFun = suppressWarnings(brewer.pal(100, 'Blues'))
colFun = circlize::colorRamp2(seq(0, 1, length.out = length(colFun)), colFun)
  
##Strain relatedness counts
strn_qcd_n <- msc_qcd_anotd_stgstrqc[[4]]@meta.data %>% dplyr::count(strain_qcd) %>% filter(!is.na(strain_qcd))

##Merge strain numbers to IBD results
s_lqc_strn_n_ibd_tbl = ss_afm_pf7M_cln_gt_hmmibd_f_lqc[["MSC14.strain_qcd"]] %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% left_join(., strn_qcd_n, by = c('sample1'='strain_qcd')) %>% left_join(., strn_qcd_n, by = c('sample2'='strain_qcd'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1, '\n[', n_1, ']'), id2 = paste0(sample2, '\n[', n_2, ']'))

##Get the IBD and the number of sites for each pairwise comparison to annotate heatmap
s_lqc_ibdf_nsites_anno <- s_lqc_strn_n_ibd_tbl %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., s_lqc_strn_n_ibd_tbl)  %>% mutate(ibdf_nsites = paste0(round(fract_sites_IBD,2), '\n(',N_informative_sites,')'))%>% dplyr::select(starts_with('id'), ibdf_nsites) %>% pivot_wider(names_from = id2, values_from = ibdf_nsites) %>% arrange(id1) %>% mutate(across(starts_with('SC'), ~replace_na(., '1')))%>% column_to_rownames('id1')

##Create mirrored matrix and plot
##!!! TIP - Refer to this if you want more complex arrangements https://github.com/jokergoo/ComplexHeatmap/issues/712 
s_lqc_sibd_pht_mtx <- s_lqc_strn_n_ibd_tbl %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., s_lqc_strn_n_ibd_tbl) %>% dplyr::select(starts_with('id'), fract_sites_IBD) %>%  pivot_wider(names_from = id2, values_from = fract_sites_IBD) %>% arrange(id1) %>% mutate(across(starts_with('SC'), ~replace_na(., 1)))%>% column_to_rownames('id1') %>% as.matrix() 
  
s_lqc_ibd_pht<- Heatmap(s_lqc_sibd_pht_mtx, rect_gp = gpar(type = "none"), name='IBD', col = colFun, show_row_names = TRUE, row_names_side = c("left"),
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = F, cluster_columns = F, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), 
              heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
              # border='black',
              cell_fun = function(j, i, x, y, width, height, fill) {
                if(i >= j) { ## diagonal(i==j), lower triangle (i>j), upper triangle (j>i)
                  grid.rect(x, y, width, height, gp = gpar(fill = fill, col = fill))
                  grid.text(sprintf("%s", s_lqc_ibdf_nsites_anno[i, j]), x, y, gp = gpar(fontsize = 13))
        }
})



s_lqc_ibd_pht

```

```{r, fig.height=5, fig.width=6}
##Merge strain numbers to IBD results
s_strn_n_ibd_tbl = ss_afm_pf7M_cln_gt_hmmibd_f[["MSC14.strain_qcd"]] %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% left_join(., strn_qcd_n, by = c('sample1'='strain_qcd')) %>% left_join(., strn_qcd_n, by = c('sample2'='strain_qcd'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1, '\n[', n_1, ']'), id2 = paste0(sample2, '\n[', n_2, ']'))  

##Get the IBD and the number of sites for each pairwise comparison to annotate heatmap
s_ibdf_nsites_anno <- s_strn_n_ibd_tbl %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., s_strn_n_ibd_tbl) %>% mutate(ibdf_nsites = paste0(round(fract_sites_IBD,2), '\n(',N_informative_sites,')'))%>% dplyr::select(starts_with('id'), ibdf_nsites) %>% pivot_wider(names_from = id2, values_from = ibdf_nsites) %>% arrange(id1) %>% mutate(across(starts_with('SC'), ~replace_na(., '1')))%>% column_to_rownames('id1')


##Create mirrored matrix and plot
sibd_pht_mtx <- s_strn_n_ibd_tbl %>% rename('id2' = id1,  'id1' = id2) %>% bind_rows(., s_strn_n_ibd_tbl) %>% dplyr::select(starts_with('id'), fract_sites_IBD) %>%  pivot_wider(names_from = id2, values_from = fract_sites_IBD) %>% arrange(id1) %>% mutate(across(starts_with('SC'), ~replace_na(., 1)))%>% column_to_rownames('id1')  
  
sibd_pht<- Heatmap(sibd_pht_mtx,rect_gp = gpar(type = "none"), name='IBD', col = colFun, show_row_names = TRUE, row_names_side = c("left"),
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = F, cluster_columns = F, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), 
              heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
              # border='black',
              cell_fun = function(j, i, x, y, width, height, fill) {
                if(i >= j) {
                  grid.rect(x, y, width, height, gp = gpar(fill = fill, col = fill))
                  grid.text(sprintf("%s", s_ibdf_nsites_anno[i, j]), x, y, gp = gpar(fontsize = 13))
        }
})


sibd_pht
  
```

```{r, fig.height=5, fig.width=6}
##Merge strain numbers to IBD results
s_strn_n_ibd_tbl = ss_afm_pf7M_cln_gt_hmmibd_f[["MSC14.strain_qcd"]] %>% dplyr::select(sample1, sample2, fract_sites_IBD,N_informative_sites) %>% left_join(., strn_qcd_n, by = c('sample1'='strain_qcd')) %>% left_join(., strn_qcd_n, by = c('sample2'='strain_qcd'), suffix = c('_1', '_2')) %>% mutate(id1 = paste0(sample1, '\n[', n_1, ']'), id2 = paste0(sample2, '\n[', n_2, ']'))  

##Get the IBD and the number of sites for each pairwise comparison to annotate heatmap
s_combi_ibdf_nsites_anno <- s_strn_n_ibd_tbl %>% 
  rename('id2' = id1,  'id1' = id2) %>% 
  bind_rows(., s_lqc_strn_n_ibd_tbl) %>% 
  mutate(ibdf_nsites = paste0(round(fract_sites_IBD,2), '\n(',N_informative_sites,')'))%>% 
  dplyr::select(starts_with('id'), ibdf_nsites) %>% 
  pivot_wider(names_from = id2, values_from = ibdf_nsites) %>% 
  arrange(id1) %>%
  mutate(across(starts_with('SC'), ~replace_na(., '1')))%>%
  column_to_rownames('id1')


##Create mirrored matrix and plot
##!!! TIP - Refer to this if you want more complex arrangements https://github.com/jokergoo/ComplexHeatmap/issues/712 - eg different colours for the different half matrices
s_combi_ibd_pht <- s_strn_n_ibd_tbl %>% 
  rename('id2' = id1,  'id1' = id2) %>% 
  bind_rows(., s_lqc_strn_n_ibd_tbl) %>% 
  dplyr::select(starts_with('id'), fract_sites_IBD) %>%  
  pivot_wider(names_from = id2, values_from = fract_sites_IBD) %>% 
  arrange(id1) %>% 
  mutate(across(starts_with('SC'), ~replace_na(., 1)))%>%
  column_to_rownames('id1') %>% 
  as.matrix()%>%
  Heatmap(., name='IBD', col = colFun, show_row_names = TRUE, row_names_side = c("left"),
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = F, cluster_columns = F, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), 
              heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),border='black',
              cell_fun = function(j, i, x, y, width, height, fill) {
        grid.text(sprintf("%s", s_combi_ibdf_nsites_anno[i, j]), x, y, gp = gpar(fontsize = 13))
})

s_combi_ibd_pht
  
```



```{r}
pf_chr_size = read_delim('../../Pf3D7_genomes_gtfs_indexs/PlasmoDB-52_Pfalciparum3D7_Genome.size', col_names = c('chrom', 'size')) %>% mutate(chr = str_remove_all(chrom,'Pf3D7_0|Pf3D7_|_v3')) 
```


```{r}

# k1_ibd_sites_f <- list(gt_fb_vtx_p1_bi_strn_pf7_asx_gam_ibd, gfv_p1_bi_pf7_k7_all_ibd) %>%
# k8_ibd_sites_f <- list(ss_ag_k6r_pf7M_cln_gt_hmmibd, s_k6r_pf7M_cln_gt_hmmibd) %>%
# k8_ibd_sites_f <- map(ss_afm_pf7M_cln_gt_hmmibd[7:8], ~{
k8_ibd_sites_f <- map(list(ss_ibd_cln_all_strain_hmmibd_2), ~{
  tb1=.x
  list(tb1) %>%
  map(~{
    .x%>% 
      mutate(across(chr,as.character)) %>% 
  left_join(.,pf_chr_size, by ='chr') %>% 
  mutate(across(chr, ~factor(., levels = c(1:14)))) %>%
  unite(., 'comps',sample1, sample2, sep='vs', remove = F) %>%  
  group_by(chr) %>% 
  mutate(chr_s = 0, chr_e = size) %>% 
  ungroup() %>% 
  mutate(across(different, as.character))

    })

})
```


```{r}
## Function for chromosome painting

chrom_paintr_ibd <- function(gtypes_comps_tbl, ttl, expn= 100000, x_font_sz = 13) {
  gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+ 
      geom_segment(
    aes(x = 0-(2000/expn), xend = size+(2000/expn), y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.8, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = rltd
    ),
    linewidth = 2.8
  ) +
  # scale_color_manual(values = c('Same'= '#D4AF37', 'Different'= '#2C3E50','Both'='lightgrey', 'No SNP'='white'))+ ##prev colours "black" ,"#ffd94d" 
  scale_color_manual(values = c('Same'= '#DACB7D', 'Different'= '#2C3E50','Both'='#A9B7C5', 'No SNP'='white'))+
  # facet_wrap(comps~., ncol = 5) +
      labs(title = ttl) +
      labs(colour = "Allele \ncomparison", x = 'Position (bases)', y ="Chromosome") +
  theme_classic() +
  theme(axis.text.y = element_text(size =  13),
        axis.text.x = element_text(size =  x_font_sz),
        # axis.ticks.x = element_blank(),
        # axis.title = element_text(size = 9),
        axis.title = element_blank(),
        # legend.title = element_text(size = 9, face = "bold"),
        # legend.text = element_text(size = 9),
        title = element_text(size = 12, face = "bold"),
        legend.position = "none",
        plot.background = element_rect(fill = "#bfe5fc")) 

  }


```


```{r, fig.height=14, fig.width=24}
ss_afm_k6r_pf7M_ibd <- k8_ibd_sites_f[[1]][[1]] %>% mutate(across(comps, ~str_replace(., 'vs', '_vs_')), rltd = case_when(different == '0' ~ 'Same', different == '1' ~ 'Different')) 

ss_afm_k6r_pf7M_ibd_p <- ss_afm_k6r_pf7M_ibd %>% split(., ~factor(comps)) %>% imap(., ~chrom_paintr_ibd(.x, .y))
```


```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[1:10]
# plot_grid(plotlist = c(ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC3_A_vs_SC4_G")], ss_ag_k6r_pf7M_ibd_p[c("SC3_A_vs_SC4_G")]), ncol = 2)
```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[str_detect(names(ss_afm_k6r_pf7M_ibd_p), "M60.*M68")]
```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[str_detect(names(ss_afm_k6r_pf7M_ibd_p), "M49.*M60")]
```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[str_detect(names(ss_afm_k6r_pf7M_ibd_p), "M48.*M49")]
```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p["M48_SC3_vs_M48_SC4"]
```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[str_detect(names(ss_afm_k6r_pf7M_ibd_p), "M49.*M49.*")]

```

```{r, fig.height=2.7, fig.width=6}
ss_afm_k6r_pf7M_ibd_p[str_detect(names(ss_afm_k6r_pf7M_ibd_p), "M68.*M68.*")]
```


```{r, fig.height=14, fig.width=24}
ss_ag_k6r_pf7M_ibd <- k8_ibd_sites_f[[1]][[1]] %>% mutate(across(comps, ~str_replace(., 'vs', '_vs_')), rltd = case_when(different == '0' ~ 'Same', different == '1' ~ 'Different')) 

ss_ag_k6r_pf7M_ibd_p <- ss_ag_k6r_pf7M_ibd %>% split(., ~factor(comps)) %>% imap(., ~chrom_paintr_ibd(.x, .y))
```

```{r, fig.height=2.7, fig.width=6}

# plot_grid(plotlist = c(ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC1_A_vs_SC8A")], ss_ag_k6r_pf7M_ibd_p[c("SC1_A_vs_SC8A")]), ncol = 2)
```

```{r, fig.height=2.7, fig.width=6}
ss_ag_k6r_pf7M_ibd_p
# plot_grid(plotlist = c(ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC3_A_vs_SC4_G")], ss_ag_k6r_pf7M_ibd_p[c("SC3_A_vs_SC4_G")]), ncol = 2)
```

```{r, fig.height=2.7, fig.width=12}

plot_grid(plotlist = c(ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC2_A_vs_SC2_G")], ss_ag_k6r_pf7M_ibd_p[c("SC2_A_vs_SC2_G")],ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC6_A_vs_SC6_G")],ss_ag_k6r_pf7M_ibd_p[c("SC6_A_vs_SC6_G")]), ncol = 4)
```

```{r, fig.height=5.5, fig.width=6}
plot_grid(
plot_grid(plotlist = ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G")], ncol = 1),
plot_grid(plotlist =ss_ag_k6r_pf7M_ibd_p[c("SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G")], ncol = 1) , nrow = 1
)
```

```{r, fig.height=14, fig.width=24}
s_k6r_ibd <- k8_ibd_sites_f[[2]][[1]]  %>% mutate(across(comps, ~str_replace(., 'vs', '_vs_')), rltd = case_when(different == '0' ~ 'Same', different == '1' ~ 'Different')) 

s_k6r_ibd_comps = s_k6r_ibd %>% split(., ~factor(comps)) %>% names
s_k6r_ibd_sc7_comps <- c("SC1_vs_SC7", "SC2_vs_SC7", "SC3_vs_SC7", "SC4_vs_SC7", "SC5_vs_SC7", "SC6_vs_SC7")
s_k6r_xlab_sc7 <- ifelse(s_k6r_ibd_comps %in% s_k6r_ibd_sc7_comps, 13, 0)

s_k6r_ibd_plts <- s_k6r_ibd %>% split(., ~factor(comps)) %>% imap(., ~chrom_paintr_ibd(.x, .y))
```

```{r}
s_k6r_ibd_plts_xlab_mod = s_k6r_ibd %>% split(., ~factor(comps)) %>% list(., names(.), s_k6r_xlab_sc7) %>% pmap(., ~chrom_paintr_ibd(..1, ..2, x_font_sz = ..3))

```

```{r}
s_k6r_ibd_plts[1:3]
```


```{r, fig.height=6, fig.width=8}

plot_grid(plotlist = c(ss_afm_chromp[["MSC14.strain_qcd"]][c("SC1_vs_SC2")],ss_afm_chromp[["MSC14.strain_qcd"]][c("SC2_vs_SC5")], ss_afm_chromp[["MSC14.strain_qcd"]][c("SC2_vs_SC6")],ss_afm_chromp[["MSC14.strain_qcd"]][c("SC6_vs_SC7")],ss_afm_chromp[["MSC14.strain_qcd"]][c("SC2_vs_SC7")]), ncol = 3)
```


```{r, fig.height=16, fig.width=12}
## Without position labels in all plots but bottom row
plot_grid(
  plot_grid(plotlist = s_chromp_xlab_mod[["MSC14.strain_qcd"]][1:5], nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod[["MSC14.strain_qcd"]][6:9], s_k6r_ibd_plts_xlab_mod[1]), nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod[["MSC14.strain_qcd"]][10:12], s_k6r_ibd_plts_xlab_mod[c(6,2)]), nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod[["MSC14.strain_qcd"]][13:14], s_k6r_ibd_plts_xlab_mod[c(10,7,3)]), nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod[["MSC14.strain_qcd"]][15], s_k6r_ibd_plts_xlab_mod[c(13,11,8,4)]),nrow = 1),
  plot_grid(plotlist = s_k6r_ibd_plts_xlab_mod[c(15,14,12,9,5)], nrow = 1),
  ncol = 1
)
```

```{r, fig.width= 12, fig.height=15}
##Put together the plots for annotation


plot_grid(plt_ovrl[["MSC14.strain_qcd"]] +
            theme(plot.margin = margin(t=0.7, b=0.3, unit = "cm")),
          plot_grid(grid.grabExpr(draw(ss_afm_hmap_ibs_m14_noSC8[["MSC14.stage_ag_strain_qcd"]])),
                    plot_grid(
                      plot_grid(plotlist =ss_afm_chromp[["MSC14.stage_ag_strain_qcd"]][c("SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G")], ncol = 1),
                      plot_grid(plotlist =ss_ag_k6r_pf7M_ibd_p[c("SC2_A_vs_SC2_G", "SC6_A_vs_SC6_G")], ncol = 1) , 
                      nrow = 1
                    ),
                    ncol = 2,
                    labels = c('B', 'C'),
                    label_size = 24,
                    vjust= 0.1
          )+
              theme(plot.margin = margin(t=0.7, b=0.3, unit = "cm")),
              plot_grid(grid.grabExpr(draw(ss_afm_hmap_ibs_m14_noSC8[["MSC14.strain_qcd"]])), 
              # grid.grabExpr(draw(sibd_pht)),
              grid.grabExpr(draw(s_combi_ibd_pht)),
              nrow = 1,
              rel_widths = c(.45, .55),
              # labels = c('D', 'E', 'F'),
              labels = c('D', 'E'),
              label_size = 24,
              vjust= 0.1) +
            theme(plot.margin = margin(t=0.7, b=0.3, unit = "cm")),
          nrow = 3,
          rel_heights = c(.3, .4, .3),
          labels = c('A', ''),
          label_size = 24,
          vjust= 1
)

``` 

```{r, fig.height=16, fig.width=12}
## Without position labels in all plots but bottom row
s_chromp_xlab_mod_m14_noSC8 <- s_chromp_xlab_mod[["MSC14.strain_qcd"]][!str_detect(names(s_chromp_xlab_mod[["MSC14.strain_qcd"]]), "_SC8")]
s_k6r_ibd_plts_xlab_mod_m14_noSC8 <- s_k6r_ibd_plts_xlab_mod[!str_detect(names(s_k6r_ibd_plts_xlab_mod), "_SC8")]

plot_grid(
  plot_grid(plotlist = s_chromp_xlab_mod_m14_noSC8[1:5], nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod_m14_noSC8[6:9], s_k6r_ibd_plts_xlab_mod_m14_noSC8[1]), nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod_m14_noSC8[10:12], s_k6r_ibd_plts_xlab_mod_m14_noSC8[c(6,2)]), nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod_m14_noSC8[13:14], s_k6r_ibd_plts_xlab_mod_m14_noSC8[c(10,7,3)]), nrow = 1),
  plot_grid(plotlist = c(s_chromp_xlab_mod_m14_noSC8[15], s_k6r_ibd_plts_xlab_mod_m14_noSC8[c(13,11,8,4)]),nrow = 1),
  plot_grid(plotlist = s_k6r_ibd_plts_xlab_mod_m14_noSC8[c(15,14,12,9,5)], nrow = 1),
  ncol = 1
)
```


```{r, fig.height=8, fig.width=20}
# k8_ibd_sites_f %>%
#   map(~{
#     .x%>%
# ggplot() +
#   geom_segment(
#     aes(x = 0, xend = chr_e, 
#         y = comps, yend = comps),
#     linewidth = 2.5, lineend = "round", colour = "grey75"
#   ) +
#   geom_segment(
#     aes(
#       x = start, xend = end,
#       y =  comps, yend =  comps,
#       colour = different
#     ),
#     linewidth = 2.5
#   ) +
#   scale_color_manual(values = c('0'='#D4AF37', '1'='#2C3E50'))+
#   facet_wrap(chr~., ncol = 5)+
#   theme_classic() +
#   theme(strip.text = element_text(size = 12, face = 'bold'),
#         axis.text = element_text(size = 12))
#     
#     
#   })

```

```{r, fig.height=14, fig.width=24}
# g8_ibd_chrom_ps <- k8_ibd_sites_f %>%
#   map(~{
#     
# a_plt<- .x %>%
# ggplot() +
#   geom_segment(
#     aes(x = 0, xend = chr_e, y = chr, yend = chr),
#     linewidth = 2, lineend = "round", colour = "grey75"
#   ) +
#   geom_segment(
#     aes(
#       x = start, xend = end,
#       y =  chr, yend =  chr,
#       colour = different
#     ),
#     linewidth = 2
#   ) +
#   scale_color_manual(values = c('0'='black', '1'='#ffd94d'))+
#   facet_wrap(comps~., ncol = 5) +
#   theme_classic() +
#   theme(strip.text = element_text(size = 12, face = 'bold'),
#         axis.text = element_text(size = 12))
# a_plt
# })
```


```{r, fig.height=6, fig.width=20}
# k8_ibd_sites_f %>%
#   map(~{
#     
# b_plt<-.x %>% 
#   ggplot(aes(x=end,y=Nsnp, color = different, fill = different)) +
#   geom_col() + 
#   ggplot2::facet_wrap(comps ~ ., ncol = 5)+theme_classic() + 
#   scale_color_manual(values = c('0'='black', '1'='#ffd94d'))+ 
#   scale_fill_manual(values = c('0'='black', '1'='#ffd94d'))+
#   theme(strip.text = element_text(size = 14, face = 'bold'),
#         axis.text = element_text(size = 13),
#         axis.text.x = element_text(angle = 45, hjust = 1))
# b_plt
# })
```

```{r, fig.height=12, fig.width=14}
# cowplot::plot_grid(a_plt+ 
#                         theme(axis.text.x = element_blank(),
#                               axis.ticks.x = element_blank(),
#                               axis.title.x = element_blank() ), 
#                    b_plt ,
#                    nrow = 2,
#                    ncol = 1,
#                    labels = "auto",
#                    align = "h"
#                    )
```


```{r, fig.height=6, fig.width=20}
k8_ibd_sites_f[[1]] %>%
  map(~{
    .x%>% 
      mutate(across(Nsnp, as.character)) %>%
ggplot() +
  geom_segment(
    aes(x = 0, xend = chr_e, 
        y = comps, yend = comps),
    linewidth =  1.5, lineend = "round", colour = "grey75"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  comps, yend =  comps,
      colour = different
    ),
    linewidth = 1.5
  ) +
  scale_color_manual(values = c('0'='black', '1'='#ffd94d'))+
  geom_label_repel(aes(label=Nsnp, x=(start + end)/2, y = comps), size=4,label.padding = unit(0.05, "lines"),alpha = 1,label.size = 0.01, color = alpha('black', .2)) +
  facet_wrap(chr~., ncol = 7)+
  theme_classic() +
  theme(strip.text = element_text(size = 14, face = 'bold'),
        axis.text = element_text(size = 12),
        axis.text.x = element_text(angle = 45, hjust = 1)) 
  })
```

```{r, fig.height=6, fig.width=20}
k8_ibd_sites_f[[2]][[1]] %>% group_by(chr) %>% group_split() %>%
  map(~{
    .x%>% 
      mutate(across(Nsnp, as.character)) %>%
ggplot() +
  geom_segment(
    aes(x = 0, xend = chr_e, 
        y = comps, yend = comps),
    linewidth =  1.5, lineend = "round", colour = "grey75"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  comps, yend =  comps,
      colour = different
    ),
    linewidth = 1.5
  ) +
  scale_color_manual(values = c('0'='black', '1'='#ffd94d'))+
  geom_label_repel(aes(label=Nsnp, x=(start + end)/2, y = comps), size=4,label.padding = unit(0.05, "lines"),alpha = 1,label.size = 0.01, color = alpha('black', .2)) +
  facet_wrap(chr~., ncol = 7)+
  theme_classic() +
  theme(strip.text = element_text(size = 14, face = 'bold'),
        axis.text = element_text(size = 12),
        axis.text.x = element_text(angle = 45, hjust = 1)) 
  })
```


```{r, fig.height=7, fig.width=20}
k8_ibd_sites_f[[1]] %>%
  map(~{
    .x%>% 
ggplot() +
  geom_segment(
    aes(x = 0, xend = chr_e, 
        y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "grey75"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = Nsnp#,
      # fill = Nsnp,
    ),
    linewidth = 3
  ) +
  # scale_color_manual(values = c('0'='black', '1'='#ffd94d'))+
  facet_wrap(comps~., ncol = 6) +
  theme_classic() +
  theme(strip.text = element_text(size = 14, face = 'bold'),
        axis.text = element_text(size = 12),
        axis.text.x = element_text(angle = 45, hjust = 1))
  })
```

HMMIBD END!!!!!!!!!!!!!

igraph
```{r}
# #==========================================================
# # Function to create an adjacency matrix
# #=========================================================
# construct_adj_matrix = function(Result, Entry){
#   
#   sample_names = unique(c(as.character(Result$individual1), as.character(Result$individual2)))
#   sample_count = length(sample_names)
#   adj_matrix = array(data = NA, dim = c(sample_count, sample_count), 
#                      dimnames = list(sample_names, sample_names))
#   for(i in 1:nrow(Result)){
#     indi = as.character(Result$individual1[i])
#     indj = as.character(Result$individual2[i])
#     adj_matrix[indi, indj] = Result[i, Entry]
#   }
#   return(adj_matrix)
# }
# 
# 
# #==========================================================
# # Function to filter highly related samples within sites
# #==========================================================
# rm_highly_related_within = function(Result, Cities, Edge){
#   
#   if(Edge){
#     # Keep edges statistically diff. from one
#     Result_filtered = Result[Result$`r97.5.` < 1-eps,] 
#   } else {
#     # First construct an adjaceny matrix of 97.5% CI IBD
#     adj_matrix = construct_adj_matrix(Result, Entry = 'r97.5.')
#     sample_names = rownames(adj_matrix)
#     
#     # Second, modify s.t. only edges NOT statistically diff. from one
#     adj_matrix_high = adj_matrix
#     adj_matrix_high[adj_matrix < 1-eps] = 0
#     
#     # Third, create a graph and list its components
#     G = graph_from_adjacency_matrix(adjmatrix = adj_matrix_high, 
#                                     mode='upper', 
#                                     diag=F, weighted = 'w', 
#                                     add.colnames = T)
#     S = components(graph = G) # Lists components
#     
#     # Forth, draw a single sample per site within a component 
#     rep_samples = unlist(lapply(1:S$no, function(s){
#       samples_ids = which(S$membership==s)
#       if(length(samples_ids) == 1){
#         rep_samples = sample_names[samples_ids] # If only one sample per component, keep
#       } else { # Keep one per city
#         cities = Cities[sample_names[samples_ids]]
#         # Pick sample with earliest data (deterministically for comparison across plots)
#         rep_samples = sapply(unique(cities), function(city){
#           samples_to_choose_from = names(cities)[cities == city]
#           rep_sample = samples_to_choose_from[which.min(SNPData[samples_to_choose_from, 'COLLECTION.DATE'])]
#         })
#       }
#       return(rep_samples)
#     }))
#     
#     # Fifth, restructure Result s.t. only one sample per city per component
#     inds = Result$individual1 %in% rep_samples & Result$individual2 %in% rep_samples
#     Result_filtered = Result[inds,]
#   }
#   return(Result_filtered)
# }
# 
# #==========================================================
# # Function to return city subgraph
# #==========================================================
# extract_city_graph = function(x, city1, city2, col_edge = 'white', 
#                               a = 0, b = 1, c = 2){
#   inds = V(x)$site == city1|V(x)$site == city2 # indices for cities
#   z = induced_subgraph(x, vids = which(inds)) # extract subgraph
#   Jitter_sign = rep(1, length(V(z)$site ))
#   Jitter_sign[V(z)$site == city2] = -1
#   # vertical layout
#   Y = as.numeric(V(z)$site == city1) + (Jitter[V(z)$name] * Jitter_sign)
#   attr(z, 'layout') = cbind(X = V(z)$date, Y) # Full layout
#   V(z)$color = M_cols[V(z)$name] # Colour by clonal component
#   E(z)$weight[is.na(E(z)$weight)] = 0 # NAs introduced when filter edges 
#   
#   # Scale edge widths (not weights) and transparancy to range a, b, c
#   weights_rescaled = a + (b-a) * (E(z)$weight - min(E(z)$weight)) / (max(E(z)$weight) - min(E(z)$weight))
#   E(z)$width = weights_rescaled * c
#   E(z)$color = sapply(weights_rescaled, function(x) adjustcolor(col_edge, alpha.f = x)) # Make transparency a function of rhat
#   v_names = do.call(rbind, strsplit(attributes(E(z))$vnames, split = "\\|"))
#   edge_col_ind = which(M_cols[v_names[,1]] == M_cols[v_names[,2]] & M_cols[v_names[,1]] != "#FFFFFF")
#   if(any(edge_col_ind)){
#     E(z)$color[edge_col_ind] = M_cols[v_names[,1]][edge_col_ind]
#   }
#   return(z)}

```

```{r}
# #################################################################################
# #' Script to generate and plot a graph of relatedness between samples
# #################################################################################
# rm(list = ls())
# library(igraph) 
# library(RColorBrewer)
# source('../igraph_functions.R') # for construct_adj_matrix
# eps <- 0.01 # threshold below which LCI is considered zero in Taylor et al. 2019
# a = 0; b = 0.5; c = 0.25 # Scaling parameters for plotting edge width and transparancy
# PDF <- TRUE
# 
# # Load relatedness results and metadata
# freqs_used <- "Taylor2020"
# load(sprintf('../../RData/mles_CIs_extended_freqs%s_meta.RData', freqs_used)) 
# load(sprintf('../../RData/metadata_extended.RData'))
# load("../../RData/sids_to_remove_from_graphs.RData")
# 
# # Define city cols
# cities <- unique(metadata$City)
# cols_cities <- array(c(rev(brewer.pal(5, 'Spectral')), 
#                        brewer.pal(length(cities)-5, 'Dark2')), 
#                      dimnames = list(cities))
# 
# # Remove samples with one or more NA estimate
# sids_to_keep <- setdiff(metadata$SAMPLE.CODE, sids_remv) 
# inds_to_keep <- mle_CIs$individual1 %in% sids_to_keep & mle_CIs$individual2 %in% sids_to_keep
# mles_to_keep <- mle_CIs[inds_to_keep,]
# 
# A_est <- construct_adj_matrix(mles_to_keep, Entry = 'rhat')
# 
# if(PDF) pdf("../../Plots/Relatedness_graphs.pdf")
# 
# for(high_relatedness_threshold in c(0,0.25)){
#   A_est[A_est < high_relatedness_threshold] <- 0
#   fields::image.plot(A_est)
#   
#   # Create graph 
#   G_est <- graph_from_adjacency_matrix(A_est, mode='upper', diag=F, weighted = T) 
#   art_pnts <- names(articulation.points(G_est))
#   ccompnts <- components(G_est)
#   
#   # Add metadata
#   V(G_est)$marker_count <- metadata[V(G_est)$name,  "snp_count"]
#   V(G_est)$city <- metadata[V(G_est)$name, "City"]
#   V(G_est)$color <- cols_cities[V(G_est)$city] 
#   
#   weights_rescaled = a + (b-a) * (E(G_est)$weight - min(E(G_est)$weight)) / (max(E(G_est)$weight) - min(E(G_est)$weight))
#   
#   # Plot graph 
#   plot(G_est, 
#        main = sprintf("High relatedness threshold: %s", high_relatedness_threshold), 
#        cex.main = 0.75, 
#        vertex.label.color = "black", 
#        vertex.label.cex = 1, 
#        vertex.frame.color = NA,
#        vertex.label = NA, 
#        vertex.shape = c("square","circle")[metadata[V(G_est)$name, "PloSGen2020"] + 1],
#        vertex.size = 0.5*log(metadata[V(G_est)$name, "snp_count"]),
#        edge.width = weights_rescaled * c, 
#        edge.color = sapply(weights_rescaled, function(x)adjustcolor('black', alpha.f = x)))
#   
#   # Legend
#   cities_ <- unique(V(G_est)$city)
#   legend('bottomleft', pch = 16, bty = 'n', cex = 0.5, pt.cex = 1, 
#          inset = -0.1, col = cols_cities[cities_], legend = cities_)
#   
#   # R estimate range
#   range(A_est, na.rm = T)
# }
# 
# if(PDF) dev.off()
```

