---
title: "strain deconvolution for pre and post QC each donor singly"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---


```{r, message=FALSE}
suppressPackageStartupMessages({
  # library(Seurat)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  # library(cowplot)
  library(pheatmap)
  # library(rmarkdown)
  # library('vcfR')
  library(conflicted)
  library(RColorBrewer)
  # library(MetBrewer)
  library(ComplexHeatmap)
  library(SeqArray)
  library(ggrepel)
  library(ggvenn)
  # library(magick)
  # library(VariantAnnotation)
  library(gridExtra)
  library(Matrix)
  # library(UpSetR)
  library(SNPRelate)
  # library(openxlsx)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
})
```


# MSC parasites strain deconvolution and VCF visualization 

## Setup and sourcing 

```{r setup}
knitr::opts_knit$set(root.dir = paste0('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/'))
```

#### Source variables specific to this run - this pseudobulk strain analysis script is ran several times with different inputs
```{r}
## Get the necessary variable for the pseudobulking workflow
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_GE_postQC_header.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_preQC_header.R")
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_preQC_cln_header.R")
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pm_pbulk_gtypes_preQC_header.R")
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_postQC_cln_header.R")
sample_name_nms
```

#### Source colours, plotting functions and commonly used tables etc
```{r}
## Get the necessary variables for common housekeeping
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/plotting_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/pf_common_vars.R")
```

```{r}

## Read in necessary files
ss_afm_pf7M_qc <- readRDS(paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_qc.RDS")) %>% .[str_detect(names(.), paste(sample_name_nms, collapse = "|"))]
ss_afm_pf7M_cln_gt <- readRDS(paste0("data/processed/",species, "/" , sv_dir, "/ss_afm_pf7M_cln_gt.RDS")) %>% .[str_detect(names(.), paste(sample_name_nms, collapse = "|"))]

```

```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_chr_cln_ibs_hm <- function(gtypes_chr_pos){
  
  # ##Remove SC8 and SC1_F
  # gtypes_chr_pos <- snp_qc_tbl %>% filter(!(groups %in% c("SC8_F", "SC1_F")))%>% dplyr::select(chrom, pos, groups, qc2, alleles) %>% pivot_wider(names_from = groups, values_from = qc2) %>% unite(., col="id", c(chrom, pos), sep = "_", remove = F) %>% mutate(chr = str_remove_all(chrom, 'Pf3D7_0|Pf3D7_|_v3'))
  
  gtMat = gtypes_chr_pos %>% dplyr::select(-c(alleles,chrom, pos, chr))  %>% column_to_rownames("id") %>% mutate(across(everything(), ~as.numeric(str_replace(., '1', '2'))))%>% as.matrix() 
  
  
  gdsFile = tempfile()
    
  snpgdsCreateGeno(gdsFile, genmat = gtMat, sample.id = colnames(gtMat), 
                     snp.id = rownames(gtMat), snp.chromosome = gtypes_chr_pos$chr, 
                     snp.position = gtypes_chr_pos$pos, snp.allele = as.character(gtypes_chr_pos$alleles), snpfirstdim = TRUE)  
  
  gds = snpgdsOpen(gdsFile)
  
  gdsSub = snpgdsLDpruning(gds, ld.threshold = 0.2)
  
  ibs = snpgdsIBS(gds, num.thread = 4)
  
  rownames(ibs$ibs) = colnames(ibs$ibs) = colnames(gtMat)
    
    
  colFun = suppressWarnings(brewer.pal(100, 'Greens'))
  colFun = circlize::colorRamp2(seq(0.5, 1, length.out = length(colFun)), 
                                  colFun)

  hm = Heatmap(ibs$ibs, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = T, cluster_columns = T)
  
  mat<-ibs$ibs

  hm2<-Heatmap(mat, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = T, cluster_columns = T, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), border='black',   heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
                  layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(mat, i, j)
                      l = v >0.98
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "#a4c2f4", lwd = 2, fill = "transparent"))
})
  ##Make rectangle heatmap https://jokergoo.github.io/2021/07/22/make-triangle-heatmap/
  
  od =  hclust(dist(mat))$order
  hm2_od = mat[od, od]
  
  hm3<- Heatmap(hm2_od, rect_gp = gpar(type = "none"), name='IBS', col = colFun, show_row_names = TRUE, row_names_side = c("right"),
                 # show_column_names = TRUE, 
                show_row_dend = T, 
                 show_column_dend = F, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), 
              heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
              border='black',
              cell_fun = function(j, i, x, y, w, h, fill) {
        if(as.numeric(x) <= 1 - as.numeric(y) + 1e-6) {
            grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
        }
    })
  
  print(hm)
  print(hm2)
  print(hm3)
}
```

```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_chr_cln_ibs_mtx <- function(gtypes_chr_pos){
  
  # ##Remove SC8 and SC1_F
  # gtypes_chr_pos <- snp_qc_tbl %>% filter(!(groups %in% c("SC8_F", "SC1_F")))%>% dplyr::select(chrom, pos, groups, qc2, alleles) %>% pivot_wider(names_from = groups, values_from = qc2) %>% unite(., col="id", c(chrom, pos), sep = "_", remove = F) %>% mutate(chr = str_remove_all(chrom, 'Pf3D7_0|Pf3D7_|_v3'))
  
  gtMat = gtypes_chr_pos %>% dplyr::select(-c(alleles,chrom, pos, chr))  %>% column_to_rownames("id") %>% mutate(across(everything(), ~as.numeric(str_replace(., '1', '2'))))%>% as.matrix() 
  gtMat = gtMat[,sort(colnames(gtMat))]
  
  
  gdsFile = tempfile()
    
  snpgdsCreateGeno(gdsFile, genmat = gtMat, sample.id = colnames(gtMat), 
                     snp.id = rownames(gtMat), snp.chromosome = gtypes_chr_pos$chr, 
                     snp.position = gtypes_chr_pos$pos, snp.allele = as.character(gtypes_chr_pos$alleles), snpfirstdim = TRUE)  
  
  gds = snpgdsOpen(gdsFile)
  
  gdsSub = snpgdsLDpruning(gds, ld.threshold = 0.2)
  
  ibs = snpgdsIBS(gds, num.thread = 4)
  
  rownames(ibs$ibs) = colnames(ibs$ibs) = colnames(gtMat)
    
  mat<-ibs$ibs
   
}
```

```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_ibs_heatmap_one_don <- function(mtx, dendog = F){
# gtypes_ibs_heatmap <- function(mtx, dendo = F){
  
  colFun = suppressWarnings(brewer.pal(100, 'Greens'))
  colFun = circlize::colorRamp2(seq(0.5, 1, length.out = length(colFun)), 
                                  colFun)
   print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
  # fa = str_extract(colnames(mtx), 'MSC.*')
  # print(str(fa))
  
  # c_d = cluster_between_groups(mtx, fa)
  # dendog = if(dendo == T) c_d else F
  # dendog = if(dendo == T) cluster_between_groups(mtx, fa) else F
  # print(str(fa))
  # print(str(dendog))
  
  # fa = str_extract(colnames(mtx), 'MSC.*')
  # # fa_col = c("a" = 2, "b" = 3, "c" = 4)
  # dend1 = cluster_between_groups(mtx, fa)
  

  hm = Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog)
  
  
  hm2<-Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), border='black',   heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
                  layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(mtx, i, j)
                      l = v >0.98
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "#a4c2f4", lwd = 2, fill = "transparent"))
})
  
  print(hm)
  print(hm2)
  # print(hm3)
}
```

```{r, fig.height=4.1, fig.width=4.5}

ss_afm_ibs_mtx<- map(ss_afm_pf7M_cln_gt, possibly(~gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=.x), "only one comparison group"))
ss_afm_ibs_hmap<- map(ss_afm_ibs_mtx, possibly(~gtypes_ibs_heatmap_one_don(mtx=.x, dendo = F), "only one comparison group"))
```

```{r, fig.height=5.1, fig.width=5.5}

ss_afm_ibs_hmap_clust <- map(ss_afm_ibs_mtx, possibly(~gtypes_ibs_heatmap_one_don(mtx=.x, dendo = T), "NAs cant be clustered"))
# imap(ss_afm_wndws, possibly(~p_chart_shared_snps(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))
ss_afm_ibs_hmap_clust
                      
```



```{r}
##Function to creat pairwise comparison groups which are the used for plotting chromosome paints
strn_pw_comps <- function(gtypes_chr_pos, strn_regex){
  
  snp_pf7_afm <- gtypes_chr_pos %>% dplyr::select(id, chr, pos, starts_with(strn_regex)) #%>%  dplyr::select(!matches('^SC[0-9]_.$')) 
  
  col_pairs_strn <- snp_pf7_afm %>% dplyr::select(starts_with(strn_regex)) %>% colnames() %>% sort() %>% combn(., 2, simplify = F)
  
  strn_afm_comps <- map_dfc(col_pairs_strn, function(pair) {
    snp_pf7_afm %>%
      transmute(!!paste0(pair[1], "_vs_", pair[2]) := case_when(!is.na(.data[[pair[1]]]) & !is.na(.data[[pair[2]]]) & .data[[pair[1]]] == .data[[pair[2]]] ~ TRUE, 
                                                                !is.na(.data[[pair[1]]]) & !is.na(.data[[pair[2]]]) & .data[[pair[1]]] != .data[[pair[2]]] ~ FALSE))
  })
  
  strn_afm_comps <- base::cbind(snp_pf7_afm %>% dplyr::select(chr, pos), 
                                         strn_afm_comps)  %>%
    filter(!if_all(starts_with(strn_regex), ~is.na(.)))
  
  return(strn_afm_comps)
}
```


```{r}
##stage strain pairwise comparison groups
ss_afm_pf7M_afm_comps <- map(ss_afm_pf7M_cln_gt, possibly(~strn_pw_comps(gtypes_chr_pos = .x, strn_regex = c('SC','N')), "only one comparison group"))

```

```{r}
##Function to create binned windows across genome and compare pairwise strains and generate the plotting details for chromosome painting
b_strn_c_wndw_fn <- function(strn_c_tbl, wndw = 100000) {
  strn_c_tbl %>% 
    group_by(chr) %>% 
    filter(!is.na(chr)) %>%
    mutate(pos_range = cut(pos, breaks = c(-Inf,seq(0, max(pos), by = wndw),Inf))) %>% 
    pivot_longer(., cols = starts_with(c('SC','N')), names_to = 'comps', values_to = 'sim') %>% 
    dplyr::count(comps, chr, pos_range, sim) %>% 
    group_by(chr, comps, pos_range) %>% 
    mutate(overlap_TF = ifelse(any(is.na(sim)), sum(!is.na(sim)), n_distinct(sim)))%>%
    ungroup() %>% 
    filter(!(is.na(sim) & overlap_TF>0))  %>% 
    mutate(rltd = case_when(overlap_TF == 0 ~ 'No SNP', overlap_TF == 2 ~ 'Both', overlap_TF==1& sim==TRUE ~ 'Same', overlap_TF==1&sim==FALSE ~ 'Different')) %>% 
    group_by(chr, comps, pos_range, rltd) %>% 
    arrange(dplyr::desc(sim), .by_group=T) %>%
    summarise(overlap_snps_No = paste(n, collapse = ","), 
              overlap_snps_prop = log10(n[1]/n[2]))  %>% 
    mutate(across(overlap_snps_No, ~ifelse(rltd == 'No SNP', NA, .))) %>%
    ungroup()  %>%
    mutate(normalized_snps_prop = case_when(rltd == "Same" ~ max(overlap_snps_prop, na.rm = TRUE), 
                                            rltd == "Different" ~ min(overlap_snps_prop, na.rm = TRUE), 
                                            # rltd == 'No SNP' ~ NA, 
                                            T ~ overlap_snps_prop)) %>%
    mutate(start = as.numeric(str_extract(pos_range, "\\d+.*(?=,)")), end = as.numeric(str_extract(pos_range, "(?<=,)\\d+.*(?=])"))) %>% 
    left_join(.,pf_chr_size, by =c('chr')) %>%
    mutate(across(chr, ~factor(., levels = c(1:14)))) 
}
```


```{r}
## Get genome size like this cut -f1,2 fasta.fai > genome.size
pf_chr_size = read_delim(chrm_sz, col_names = c('chrom', 'size')) %>% mutate(chr = str_remove_all(chrom,chrm_prfx)) 
```

```{r}
##stage strain pairwise comparison chromosome binning of SNPs and creation of windows for plotting
ss_afm_wndws <- map(ss_afm_pf7M_afm_comps, possibly(~b_strn_c_wndw_fn(strn_c_tbl = .x), "only one group"))

```


```{r}
## Function for chromosome painting

p_chart_shared_snps <- function(gtypes_comps_tbl = b_strn, ttl, lg = "right") {

      
    gtypes_comps_tbl %>% 
      separate(.,overlap_snps_No, into=(c('same2', 'diff2')), sep=',' ) %>% 
      pivot_longer(cols = c('same2','diff2'), names_to = 'shared2', values_to = 'shared_n') %>% 
      filter(!is.na(shared_n)) %>%
      mutate('Shared' = case_when(rltd == 'Both' & shared2 == 'same2' ~ 'Same', 
                                  rltd == 'Both' & shared2 == 'diff2' ~ 'Different', 
                                  rltd == 'Different' & shared2 == 'same2' ~ 'Different', 
                                  T ~ rltd)) %>%
      mutate(across(shared_n, as.numeric)) %>%
      group_by(comps, Shared) %>% 
      summarise('snpsNo' = sum(shared_n))%>%
      mutate(snpsProp = snpsNo/sum(snpsNo)) %>%
      arrange(dplyr::desc(snpsProp), .by_group = T) %>%
  separate(comps, into=(c('pb_grp1', 'pb_grp2')), sep='_vs_' ) %>% 
    ggplot(., aes(x="", y=snpsProp, fill=Shared))+
    geom_bar(width = 1, stat = "identity") +
    geom_text(aes(y = ifelse(Shared == 'Same', 0.2, 0.8),
                  label = snpsNo), size=3.5) +
  facet_grid(pb_grp1~pb_grp2) +
    labs(title = ttl)+
    theme_classic() +
      theme(legend.position = lg)


}


```

Pie of snps compared
```{r, warning=FALSE}
# strain stage (afm) chromosome painting
pp_plts <- imap(ss_afm_wndws, possibly(~p_chart_shared_snps(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))


```

```{r, warning=FALSE, fig.width=8, fig.height=8}
# strain stage (afm) chromosome painting
pp_plts
```
33=8, 48=4, 49=6, 55=hsat7???, 38=4/6??,40=8,39=4
37= 5, 60=7hsat/8minmap,66=8,68=3

list("MSC33" = c(7:9),"MSC48" = c(3:5),"MSC49" = c(5:7),"MSC55" = c(5:7),"MSC60" = c(6:8), "MSC66" = c(7:9), "MSC38" = c(2:4), "MSC40" = c(7:9), "MSC50_MACS" = c(2:4), "MSC53" = c(2), "MSC57" = c(2), "MSC24" = c(2), "MSC37" = c(5:7), "MSC39" = c(4:6), "MSC45" = c(2), "MSC50_SLO" = c(2:4), "MSC54" = c(2), "MSC67" = c(2), "MSC68" = c(2:4), "MSC70" = c(2)) 
Turning the below off since they take long and I only need the information relevant to find the right K - chromosome painting is necessary for relatedness exploration after I arrive on right K

MSC25 = 2; but could be 3 or more
MSC41 = 3;
MSC51 = 1;

MSC38 best guess = 4
```{r, warning=FALSE, eval=T}
## optimum k from above
# optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1, "MSC41"= 3, "MSC51"= 1, "MSC25"= 2)
optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =6, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =2, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1, "MSC41"= 3, "MSC51"= 1, "MSC25"= 2)
optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]

##updated strain metadata
strn_c_new_mdata_m22 <- readRDS(paste0("data/processed/", species, "/" , gtyp_step, "/strn_c_new_mdata_m22.RDS"))
strn_c_new_mdata_m22_min <- strn_c_new_mdata_m22[str_detect(names(strn_c_new_mdata_m22), "_minmap")]
names(strn_c_new_mdata_m22_min) <-  str_remove(names(strn_c_new_mdata_m22_min), "_minmap")
```


##Simpler chromosome plots(no continous gradient) for SS_AG and S

```{r}
## Function for chromosome painting

chrom_paintr <- function(gtypes_comps_tbl = b_strn_c, ttl, expn=100000, x_font_sz = 13) {

      gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+
      geom_segment(
    aes(x = 0-(2000/expn), xend = size+(2000/expn), y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.8, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = rltd
    ),
    linewidth = 2.8
  ) +
  # scale_color_manual(values = c('Same'= '#D4AF37', 'Different'= '#2C3E50','Both'='lightgrey', 'No SNP'='white'))+      
  scale_color_manual(values = c('Same'= '#DACB7D', 'Different'= '#2C3E50','Both'='#A9B7C5', 'No SNP'='white'))+ ##prev colours "black" ,"#ffd94d" 
  # facet_wrap(comps~., ncol = 5) +
      labs(title = ttl) +
      labs(colour = "Allele \ncomparison", x = 'Position (bases)', y ="Chromosome") +
  theme_classic() +
  theme(axis.text.y = element_text(size =  13),
        axis.text.x = element_text(size =  x_font_sz),
        axis.title = element_blank(),
        title = element_text(size = 12, face = "bold"),
        legend.position = "none") 
    

  }

```


strains split by stage Without colour gradients
```{r, warning=F}
# strain stage (afm) chromosome painting
ss_afm_chromp <- map(ss_afm_wndws, ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr(gtypes_comps_tbl =.x, ttl=.y), 'No overlapping SNPs between the 2 groups'))
})

```
CONFIRMATION OF STRAINS
33=8, 
```{r, warning=F}
# strain stage (afm) chromosome painting
names(ss_afm_chromp)

```

```{r, warning=F}
# strain stage (afm) chromosome painting
ss_afm_chromp[str_detect(names(ss_afm_chromp), "MSC50_SLO.stage_afm_strain.k2")]

```
Confirmed

msc40 k=8 low count clusters - SC1A, SC5A, SC8M

MSC48 k=4 - NegativeF = SC3, NEGATIVEA=SC2

MSC49 - SC4M-LOW SNPS NEGATIVE_A=SC1A


```{r, warning=FALSE, eval=T}

for (i in names(strn_c_new_mdata_m22_min)) {
        if (optimum_k_qcd[[i]] == 1) {
          strn_c_new_mdata_m22_min[[i]][,"StrainQCd"] = "SC0"
        } else if (optimum_k_qcd[[i]] > 1) {
          strn_c_new_mdata_m22_min[[i]][,"StrainQCd"] = strn_c_new_mdata_m22_min[[i]][,paste0('Strain_hmo_', str_remove(i, 'SC'), "_minmap_",optimum_k_qcd[[i]])]
        } 
    
  }

```

Manually assign negative for which we assessed to the relevant Ks
MSC48 = k4 : NegF==SC3F,  NegA == SC2A
MSC49 = k6 : NegA==SC1A
MSC60 = k7 : NegA==SC3A
MSC66 = k8 : NegA==SC5A
MSC68 = k4 : NegF==SC2F


```{r, warning=FALSE, eval=T}
man_neg_K  <- list(
c("Negative", "Female", "SC3"),  
c("Negative", "Asexual", "SC2"),
c("Negative", "Asexual", "SC1"),
c("Negative", "Asexual", "SC3"),
c("Negative", "Asexual", "SC5"),
c("Negative", "Female", "SC2")
) %>% set_names(c("MSC48", "MSC48", "MSC49", "MSC60", "MSC66", "MSC68"))
```


```{r, warning=FALSE, eval=T}
# 
# for (i in names(man_neg_K)) {
#         if (i %in% names(strn_c_new_mdata_m22_min)) {
#             print(i)
#             if (man_neg_K[[i]][2] == "Asexual") {
#             print(man_neg_K[[i]][1])
#             print(man_neg_K[[i]][3])
#           } else if (man_neg_K[[i]][2] == "Female") {
#             print(man_neg_K[[i]][1])
#             print(man_neg_K[[i]][3])
#           } 
#         } else if (names(strn_c_new_mdata_m22_min)) {
#           print(i)
#         } 
#     
#   }
# 
# for (i in names(strn_c_new_mdata_m22_min)) {
#         if (i %in% names(man_neg_K)) {
#           strn_c_new_mdata_m22_min[[i]][,"StrainQCd_Man"] = strn_c_new_mdata_m22_min[[i]] %>% 
#             mutate(StrainQCd_Man = case_when(stage_afm == "Asexual" & StrainQCd == man_neg_K[[i]][2] ~ "SC2", 
#                                              stage_afm == "Female" & StrainQCd == man_neg_K[[i]][2] ~ "SC3",
#                                              T ~ StrainQCd)) %>%
#             select(StrainQCd_Man)
#         } else if (optimum_k_qcd[[i]] > 1) {
#           strn_c_new_mdata_m22_min[[i]][,"StrainQCd_Man"] = strn_c_new_mdata_m22_min[[i]]$StrainQCd
#         } 
#     
#   }

```

Manually assign negative for which we assessed to the relevant Ks
```{r, warning=FALSE, eval=T}


strn_c_new_mdata_m22_min[["MSC48"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC48"]] %>% 
  mutate(StrainQCd_Man = case_when(stage_afm == "Asexual" & StrainQCd == "Negative" ~ "SC2", 
                                   stage_afm == "Female" & StrainQCd == "Negative" ~ "SC3",
                                   T ~ StrainQCd)) %>% select(StrainQCd_Man)
strn_c_new_mdata_m22_min[["MSC49"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC49"]] %>% 
  mutate(StrainQCd_Man = case_when(stage_afm == "Asexual" & StrainQCd == "Negative" ~ "SC1",
                                   T ~ StrainQCd))%>% select(StrainQCd_Man)
strn_c_new_mdata_m22_min[["MSC60"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC60"]] %>% 
  mutate(StrainQCd_Man = case_when(stage_afm == "Asexual" & StrainQCd == "Negative" ~ "SC3", 
                                   T ~ StrainQCd))%>% select(StrainQCd_Man)
strn_c_new_mdata_m22_min[["MSC66"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC66"]] %>% 
  mutate(StrainQCd_Man = case_when(stage_afm == "Asexual" & StrainQCd == "Negative" ~ "SC5", 
                                   T ~ StrainQCd))%>% select(StrainQCd_Man)
strn_c_new_mdata_m22_min[["MSC68"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC68"]] %>% 
  mutate(StrainQCd_Man = case_when(stage_afm == "Female" & StrainQCd == "Negative" ~ "SC2",
                                   T ~ StrainQCd))%>% select(StrainQCd_Man)
strn_c_new_mdata_m22_min[["MSC50_SLO"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC50_SLO"]] %>% 
  mutate(StrainQCd_Man = case_when(StrainQCd == "Negative" ~ "SC2",
                                   T ~ StrainQCd))%>% select(StrainQCd_Man)
strn_c_new_mdata_m22_min[["MSC55"]][,"StrainQCd_Man"] <- strn_c_new_mdata_m22_min[["MSC55"]] %>% 
  mutate(StrainQCd_Man = case_when(StrainQCd == "Negative" ~ "SC2",
                                   T ~ StrainQCd)) %>% select(StrainQCd_Man)



```

```{r, warning=FALSE, eval=T}

donors_manual_strain_asn <- c("MSC48", "MSC49", "MSC60", "MSC66", "MSC68", "MSC50_SLO", "MSC55")

for (i in names(strn_c_new_mdata_m22_min)) {
        if (!(i %in% donors_manual_strain_asn)) {
          print(i)
          strn_c_new_mdata_m22_min[[i]][,"StrainQCd_Man"] = strn_c_new_mdata_m22_min[[i]]$StrainQCd
        }

  }

```



```{r, warning=FALSE, eval=T}
saveRDS(strn_c_new_mdata_m22_min, paste0("data/processed/",species, "/" , sv_dir, "/strn_c_new_mdata_m22_min_qcd.RDS"))
```


```{r}
# gogo()
```

```{r, warning=FALSE, eval=T}
glimpse(strn_c_new_mdata_m22_min$MSC48)
```

## Write out for genotyping

Write out strain_qcd + stage_ag for pseudobulk genotyping

```{r, warning=FALSE, eval=T}
## Get donors with more than 1 strain
optimum_k_polyk <- optimum_k_qcd[optimum_k_qcd > 1]
# optimum_k_polyk <- optimum_k_polyk[!names(optimum_k_polyk) %in% c("MSC37", "MSC38")]

## Genotype also for monoclonal
optimum_k_polyk <- optimum_k_qcd
```


```{r, eval=T}
## DOUBLETS WRITTEN INTO THE SAME FOLDER IN MSC_PP.rmd - SHOULD BE COMBINED INTO ONE NOTEBOOK
# sample_ids = map2(sample_name_nms, source_irods_nms, ~paste0(.x, '_', .y))
# sv_dir = "pbulk_gtypes_postQC_cln"

# for (s in names(strn_c_new_mdata_m22_min)) {
for (s in names(optimum_k_polyk)) {
  # narrow_k <- opt_narrow_clusters_nms[[str_remove(s, "_[a-z]+")]]
  narrow_k <- optimum_k_polyk[s]
  ## Assign minimap/hsat mapper directory
  # algn_nm = str_extract(s, "hsat|minmap|lr_ref")
  algn_nm = "minmap"
  # s_nm = str_remove(s, "_[a-z]+")
  s_nm = s
  
  print(algn_nm)
  print(s_nm)
  
  ## !!!NOTE - Dangerous command - activate only when barcodes are modified
  system(paste0('rm -r data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/st*'))
  
  for (k in narrow_k) {
    print(narrow_k)
    # for (str_q in c('StrainQCd_Man')){
      
      ss_qcd_ano <- map(strn_c_new_mdata_m22_min[s], ~.x %>% rename("str" = StrainQCd_Man) %>% group_by(str) %>% filter(n() >= 20) %>% ungroup() %>% mutate(across(c(str), ~droplevels(factor(.)))) %>% split(f = ~ str, drop = T))
      
      names(ss_qcd_ano[[1]]) <- names(ss_qcd_ano[[1]]) %>% str_replace(., '\\.', '_')
      # print(str(ss_qcd_ano[[1]]))

      system(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/', algn_nm,'/strain_k',k,'/'))
      print(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/', algn_nm,'/strain_k',k,'/'))

      imap(ss_qcd_ano[[1]], ~write(.x$bcode, paste0('data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/strain_k',k,'/',.y, '.tsv')))

      # for(stg_c in c('stage_afm', 'stage_ag')){
      # for(stg_c in c('stage_ag')){
        
        ss_qcd_ano <- map(strn_c_new_mdata_m22_min[s], ~.x %>% rename("str" = StrainQCd_Man, "stg" = stage_ag) %>% group_by(str, stg) %>% filter(n() >= 20) %>% ungroup() %>% mutate(across(c(str, stg), ~droplevels(factor(.)))) %>% split(f = ~ str+stg, drop = T))
        names(ss_qcd_ano[[1]]) <- names(ss_qcd_ano[[1]]) %>% str_replace(., '\\.', '_')
        print(names(ss_qcd_ano[[1]]))

        system(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/stage_ag_strain_k',k,'/'))
        print(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/stage_ag_strain_k',k,'/'))

        imap(ss_qcd_ano[[1]], ~write(.x$bcode, paste0('data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/stage_ag_strain_k',k,'/',.y, '.tsv')))

    #   }
    # }
  
}

}
```


```{r, warning=FALSE, eval=T}
gogo()
# nextflow pbulk_gtyping.nf --sv_dir "pbulk_gtypes_postQC_cln" --bam "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/MSC*/soupc_GE_postQC/*/parent/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/{MSC55,MSC50_SLO}/pbulk_gtypes_postQC_cln/bcodes/minmap/st*/*tsv" --odir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/" --mapper "minmap" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/cr_subset_bam_linux.sh" --ref  "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/references_for_souporcell/PlasmoDB-66_Pfalciparum3D7_Genome.fasta"
```

```{r, eval=FALSE}
## Function for chromosome painting

chrom_paintr_smth_gradient <- function(gtypes_comps_tbl = b_strn, ttl, expn=100000, x_font_sz = 13, lg = "right") {

      # gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+
        geom_segment(
    aes(x = 0-2000/expn, xend = size+2000/expn, y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.5, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = normalized_snps_prop#,
      # colour = rltd
    ),
    linewidth = 2.5
  ) +
  # scale_color_manual(values = c('Same'='black', 'Different'='#ffd94d', 'Both'='lightgrey', 'No SNP'='white'))+
  scale_color_gradient2(low="black" , mid = "#F5F5F5", high ="#ffd94d" ,
                        limits = c(min(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), max(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE)),
                        breaks = seq(min(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), max(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), length.out = 30),
                        oob = scales::squish,
                        na.value = "white",
                        midpoint = 0,
                        guide = lg) +
  # geom_label_repel(aes(label=new_column, x=(start + end)/2, y = comps), size=3,label.padding = unit(0.01, "lines"),alpha = 1,label.size = 0.01, color = alpha('black', .2))+
  # facet_wrap(comps~., ncol = 5) ++
      labs(title = ttl) +
  theme_classic() +
  theme(strip.text = element_text(size = 12, face = 'bold'),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size =  x_font_sz)) 
    
    pp <- gtypes_comps_tbl %>% 
      separate(.,overlap_snps_No, into=(c('same2', 'diff2')), sep=',' ) %>% 
      pivot_longer(cols = c('same2','diff2'), names_to = 'shared2', values_to = 'shared_n') %>% 
      filter(!is.na(shared_n)) %>%
      mutate('Shared' = case_when(rltd == 'Both' & shared2 == 'same2' ~ 'Same', 
                                  rltd == 'Both' & shared2 == 'diff2' ~ 'Different', 
                                  rltd == 'Different' & shared2 == 'same2' ~ 'Different', 
                                  T ~ rltd)) %>%
      mutate(across(shared_n, as.numeric)) %>%
      group_by(Shared) %>% 
      summarise('snpsNo' = sum(shared_n))%>%
      arrange(dplyr::desc(snpsNo), .by_group = T) %>%
    ggplot(., aes(x="", y=snpsNo, fill=Shared))+
    geom_bar(width = 1, stat = "identity") + 
    coord_polar("y", start=0)+
    geom_text(aes(y = snpsNo/2 + c(0, cumsum(snpsNo)[-length(snpsNo)]), 
                  label = snpsNo), size=5) +
    theme_void() +
      theme(legend.position = lg)

    ggdraw(gs + theme_half_open(12)) +
      draw_plot(pp, .5, .11, .36, .36) 
}


```


strains split by stage Without colour gradients
```{r, warning=FALSE, eval=FALSE}
# strain stage (afm) chromosome painting
ss_afm_chromp_smth <- map(ss_afm_wndws, ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))
})

# ss_afm_chromp_smth[1:3]
```

```{r, warning=FALSE, eval=FALSE}
# strain stage (afm) chromosome painting
ss_afm_chromp_smth <- map(ss_afm_wndws, ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% imap(.,~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y, lg = "none"))
})

# ss_afm_chromp_smth[1:3]
```

```{r, warning=FALSE, eval=FALSE}
# strain stage (afm) chromosome painting
map(ss_afm_wndws[1], ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% .[1] %>% imap(.,possibly(~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))
})

# ss_afm_chromp_smth[1:3]
```

strains split by stage Without colour gradients
```{r, warning=FALSE, eval=FALSE}
# strain stage (afm) chromosome painting


tr_tbl <- ss_afm_wndws[[4]] %>% split(., ~factor(comps)) %>% .["SC4_F_vs_SC5_F"]

tr <- imap(tr_tbl, ~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y, lg = "none"))

# ss_afm_chromp_smth[1:3]
```

```{r, warning=FALSE, eval=FALSE}
tr
```

```{r, warning=FALSE, eval=FALSE}
map(ss_afm_chromp_smth[[1]], class)
```

```{r, eval=FALSE}

names(ss_afm_chromp_smth)
```



```{r, eval=FALSE}
## Remove empty slots where there are no overlapping SNPs between the 2 groups being compared
pts <- ss_afm_chromp_smth[[1]][unlist(map(ss_afm_chromp_smth[[1]], ~length(class(.x)) == 2))]

chrom_pdf_save <- function(plt_list = pts, No_plots_p_page =24, No_cols = 4, don = "MSC33", grp = "stage_afm_strain"){
  print(length(plt_list))
  plt_list <- plt_list[unlist(map(plt_list, ~length(class(.x)) == 2))]
  print(length(plt_list))
  
  plot_lists_sep <- split(plt_list, (seq_along(plt_list) - 1) %/% No_plots_p_page)
  
  print(don)
  
  system(paste0("mkdir -p plots/pbulk_gtypes_preQC/", don))
  
  pdf(paste0("plots/pbulk_gtypes_preQC/", don, "/", grp, ".pdf"),  onefile = TRUE, width=12, height = 16.91567) ## A4 aspect ratio (1.409639) but bigger
  for (i in seq_along(plot_lists_sep)) {
    print(i)
    grid.arrange(grobs = plot_lists_sep[[i]], ncol = No_cols)
  }
  dev.off()
}
```


```{r, eval=FALSE}
## Remove empty slots where there are no overlapping SNPs between the 2 groups being compared
imap(ss_afm_chromp_smth, ~chrom_pdf_save(plt_list = .x, don = str_remove(.y, "\\..*"), grp = paste0(str_replace(.y, ".", "_"),"k")))
```


```{r, eval=FALSE}

ss_afm_chromp_smth[1]
```

m55= 3=4,3=6, 4=6

strains low QC (lqc) split by stage Without colour gradients

```{r, eval=FALSE}
# strain stage (afm) chromosome painting
s_lqc_k6r_chromp <- s_lqc_k6r_wndws %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y), 'No overlapping SNPs between the 2 groups'))

s_lqc_k6r_chromp[1:4]
```
