---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(dreamlet)
  library(muscat)
  library(ExperimentHub)
  library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(BiocParallel)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  
})
```


#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/plotting_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/pf_common_vars.R")
```

```{r setup}
sample_set = "Mali2"

optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)

optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]

# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```


NB-LOCAL CODE
```{r, eval=F}

system(paste0("rsync -av jr35@farm5:/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/",sample_set,"/data/processed/Pf/all_msc_w_strns_don.RDS /Users/jr35/Google\\ Drive/My\\ Drive/PHD_documents/malaria_phd/",sample_set,"/data/processed/Pf/"
))

```

initialize variable for the sample names

```{r}
all_msc_w_strns_don <- readRDS("data/processed/Pf/all_msc_w_strns_don.RDS")
```


```{r}
strn_c_new_mdata_m22_min_qcd <- readRDS("data/processed/Pf/strn_c_new_mdata_m22_min_qcd.RDS")
```


```{r, fig.width=12, fig.height=4}
# add stage metadata
# all_msc_w_strns_don <- pmap(list(all_msc_w_strns_don[names(strn_c_new_mdata_m22_min_qcd)], strn_c_new_mdata_m22_min_qcd, names(strn_c_new_mdata_m22_min_qcd)), ~{
#   ..2 %>% 
#   mutate(StrainDon = paste0(str_remove(..3, "SC"), "_", StrainQCd)) %>%
#   select(bcode, StrainQCd, StrainDon) %>% 
#     column_to_rownames("bcode") %>%
#   AddMetaData(..1,.)
# })

```


```{r, eval=T}

imap(all_msc_w_strns_don, ~DimPlot(.x, group.by = "StrainDon")  + labs(title = .y))
```


```{r}
msc_qcd_anotd <-readRDS("../Mali/data/processed/DB52e_star/msc_qcd_anotd.RDS")

```

```{r, eval=T}

msc_qcd_anotd_mdata_14 <- msc_qcd_anotd[["MSC14"]]@meta.data %>% mutate(StrainQCd= strain_qcd) 
```


```{r, fig.width = 18, fig.height = 4, message=FALSE, warning=FALSE}

strn_c_new_mdata_m22_min_qcd <-  strn_c_new_mdata_m22_min_qcd[names(optimum_k_qcd[optimum_k_qcd>1])]

ss_bar_plt <-list( c(strn_c_new_mdata_m22_min_qcd, list(msc_qcd_anotd_mdata_14 )), c(names(strn_c_new_mdata_m22_min_qcd), "MSC14"))  %>% 
  purrr::pmap(.,~{
  a_tbl <- ..1 %>%
      mutate(across(stage_afm, ~factor(., levels=c('Asexual', 'Female', 'Male', 'Developing', 'Unassigned')))) %>%
      dplyr::count(StrainQCd, stage_afm, .drop = FALSE) 
  
  ##get stage_ag labels and counts to add to the StrainQCd+stage_afm counts generated above
  stage_ag_cnt <- ..1 %>% 
    dplyr::select(StrainQCd, stage_ag, stage_afm) %>% 
    add_count(StrainQCd, stage_ag, name = 'stage_ag_n') %>% 
    distinct(pick(StrainQCd, stage_afm), .keep_all = T) %>% 
    left_join(a_tbl, ., by =c('StrainQCd', 'stage_afm')) %>% 
    rename(c('Stage'= stage_afm, 'Counts' =n, 'Strain' = StrainQCd))  
  
  print(stage_ag_cnt)
  
    b_plt <- stage_ag_cnt %>%
      # ggplot(., aes(x = Stage, y = Counts, fill = Stage, color=stage_ag_n<30)) +
      ggplot(., aes(x = Stage, y = Counts, fill = Stage)) +
      geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
      geom_text(aes(label=Counts), hjust=0.5, vjust= -0.3, position = position_dodge(width = .9), size = 7, show.legend = F, fontface ='bold', colour='black')+
      scale_fill_manual(values = c_stg_cols) + #'Asexual' = "#FFE5B4", 'Male' = "#BFD7B5", 'Female'="#8ab6d6"
      # scale_color_manual(values = c('FALSE' = "white", 'TRUE' = "blue")) +
  geom_tile(aes(x = Stage, y = -10, fill = Stage, height = 14)) +
  facet_wrap( Strain ~ ., nrow = 1) +
      scale_y_continuous(expand = expansion(mult = c(0.01, 0.15))) +
      scale_x_discrete(expand = expansion(add = c(1, 1))) +
      theme_classic() +
  labs( y = 'Counts', title = ..2)+
      theme(
        axis.text.y  = element_text(size = 22),
        axis.title.y  = element_text(size = 22, face = "bold"),
        axis.title.x = element_blank(),
        # axis.title = element_text(size = 18, face = "bold"),
        legend.text = element_text(size = 22),
        legend.title = element_text(size = 22, face = "bold"),
        legend.position="bottom",
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        strip.text =element_text(size=20,face="bold", colour = 'black'),
        panel.border=element_rect(colour="black",fill='transparent')
      )+ 
      guides(color = F)
    print(b_plt)
    return(b_plt)
})


```

```{r, fig.width=12, fig.height=4}
# add stage metadata
# all_msc_w_strns_don <- map(all_msc_w_strns_don, ~{
#   .x@meta.data %>% 
#   rownames_to_column('bcode') %>%
#   mutate(donor = str_remove(orig.ident, ".mrna"), 
#          # strain_don = case_when(str_detect(orig.ident, "MSC") ~ paste0(Strain_DN, "_", donor))
#          ) %>%
#   left_join(., irods_id_sk21_mdata_cln %>% select(sample_nm, Symptomatic, Hb_Type), by = c("donor" = "sample_nm")) %>%
#   select(bcode, donor, Symptomatic, Hb_Type) %>%
#   column_to_rownames('bcode') %>% 
#   AddMetaData(.x,.)
# })

```



```{r, eval=T}

imap(all_msc_w_strns_don, ~DimPlot(.x, group.by = "StagePrelimC") + scale_color_manual(values = sp_fld_colors) + labs(title = .y))
```

```{r, eval=T}

imap(all_msc_w_strns_don, ~DimPlot(.x, group.by = "StagePrelimC", reduction = "pca") + scale_color_manual(values = sp_fld_colors) + labs(title = .y))
```

```{r}
all_msc_w_strns_don <- all_msc_w_strns_don[!names(all_msc_w_strns_don) %in% c("MSC38", "MSC39")]
```



```{r}
##Read MCA V3 blood stage reference dataset
# EF_V3_ref <- readRDS("../sc_velo/data/raw/RoadtoUMAP/pseudotime/raw_combined_cut_0.4.RDS")
EF_V3_ref <- readRDS("../Mali/data/raw/mca_v3_ref/raw_combined_cut_0.4_nodoublets_stagenew.RDS")
```

```{r}
##make refined metadata column that includes the gametocyte substages for finer resolution mapping
EF_V3_ref <- EF_V3_ref@meta.data %>%
  mutate(stage_fine = case_when(optionB_Stage == "no assignment" ~ as.character(Stage),
                                TRUE ~ as.character(optionB_Stage))
  ) %>% 
  mutate(across(stage_fine, ~str_to_sentence(str_replace_all(., c("Gametocytes"="Gametocyte", "comitted" = "committed"))))) %>%
  mutate(across(stage_fine, ~droplevels(factor(., levels = stg_refnd_lvls)))) %>%
  mutate(across(stagenewx, ~droplevels(factor(str_to_sentence(.), levels = stg_refnd_lvls)))) %>% 
  dplyr::select(stage_fine, stagenewx) %>%
  AddMetaData(EF_V3_ref, .)
```




```{r}
harmony_int <-  readRDS("data/processed/Pf/m2_all/harmony_int.RDS")
```


```{r}
harmony_int <- JoinLayers(harmony_int)
```


```{r}

DimPlot(harmony_int, reduction = "umap.harmony", group.by = c("StagePrelimAll", "harmony_clusters"), combine = F, label = T)
```

```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(harmony_int, reduction = "umap.harmony", group.by = "StagePrelim", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```




```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(harmony_int@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= harmony_int@meta.data$StagePrelimC,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r}
v3_m2_asex_harmony <-  readRDS("data/processed/Pf/m2_all/v3_m2_asex_harmony.RDS")
```


```{r}
v3_m2_asex_harmony <- JoinLayers(v3_m2_asex_harmony)
```


```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "StagePrelimC", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = c("orig.ident", "StagePrelim"), label = F)
```


```{r, fig.width=6, fig.height=5}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "harmony_clusters_asex", label = T, dims = c(1,2))
```




```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = c("StageFL"), label = F)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "pca", group.by = c("StageFL"), label = F)
```

```{r, fig.width=12, fig.height=4}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
DimPlot(v3_m2_asex_harmony, reduction = "pca", group.by = c("dset"), label = F)
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_harmony@reductions[["pca"]]@cell.embeddings),
        x = ~PC_1, y = ~PC_2, z = ~PC_3,
        color= v3_m2_asex_harmony@meta.data$StageFL,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.2, 
        type = 'scatter3d'
)

```



```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_harmony@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= v3_m2_asex_harmony@meta.data$StageFL,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.2, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE, eval=FALSE}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_harmony@reductions[["umap.harmony"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= v3_m2_asex_harmony@meta.data$StageFL,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.2, 
        type = 'scatter3d'
)

```


```{r}
# strn_c_new_mdata_m22_min_qcd_mg <- strn_c_new_mdata_m22_min_qcd %>% 
#   bind_rows(., .id = "donor") %>% 
#   mutate(bcode = paste0(str_remove(donor, "SC"), "_", bcode), 
#          StrainDon = paste0(str_remove(donor, "SC"), "_", StrainQCd)) %>%
#   select(bcode, StrainQCd, StrainDon) %>% 
#   column_to_rownames("bcode") 
```

```{r, fig.width=12, fig.height=4}
# add stage metadata
# v3_m2_asex_harmony <- strn_c_new_mdata_m22_min_qcd_mg %>% select(StrainDon) %>%  AddMetaData(v3_m2_asex_harmony,.)

```

```{r, fig.width=12, fig.height=4}
# add stage metadata
# v3_m2_asex_harmony <- v3_m2_asex_harmony@meta.data %>% 
#   rownames_to_column('bcode') %>%
#   mutate(donor = str_remove(orig.ident, ".mrna"), 
#          # strain_don = case_when(str_detect(orig.ident, "MSC") ~ paste0(Strain_DN, "_", donor))
#          ) %>%
#   left_join(., irods_id_sk21_mdata_cln %>% select(sample_nm, Symptomatic, Hb_Type), by = c("donor" = "sample_nm")) %>%
#   select(bcode, donor, Symptomatic, Hb_Type) %>%
#   column_to_rownames('bcode') %>% 
#   AddMetaData(v3_m2_asex_harmony,.)

```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
v3_m2_asex_harmony@meta.data %>% filter(str_detect(donor, "MSC")) %>% dplyr::count(donor, Symptomatic, Hb_Type)
```


```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
m2_asex_harmony_tr <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# pseudobulk the counts based on donor-condition-celltype
pseudo_m2_asex <- AggregateExpression(m2_asex_harmony_tr, assays = "RNA", return.seurat = T, group.by = c("Symptomatic", "donor", "StagePrelim"))

# each 'cell' is a donor-condition-celltype pseudobulk profile
tail(Cells(pseudo_m2_asex))
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
pseudo_m2_asex$stage.sympt <- paste(pseudo_m2_asex$StagePrelim, pseudo_m2_asex$Symptomatic, sep = "_")
```


```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2_asex) <- "stage.sympt"

bulk.LR.de <- FindMarkers(object = pseudo_m2_asex, 
                         ident.1 = "Late ring_Yes", 
                         ident.2 = "Late ring_No",
                         test.use = "DESeq2")
head(bulk.LR.de, n = 15)
```



```{r, fig.width=15, fig.height=5}

head(bulk.LR.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6650087/ - PF3D7-1240400
#https://journals.asm.org/doi/10.1128/mbio.01636-21 https://www.frontiersin.org/articles/10.3389/fmala.2023.1075755/full  - hyp10
#http://dr.iiserpune.ac.in:8080/xmlui/bitstream/handle/123456789/7531/20153385_Abhishek_Kanyal_PhD_Thesis.pdf?sequence=1&isAllowed=y - PF3D7-1401300
```

```{r, fig.width=15, fig.height=5}
VlnPlot(pseudo_m2_asex, features = "PF3D7-0114500", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 
```



```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
# ifnb$donor_id.stim <- paste0(ifnb$stim, "-", ifnb$donor_id)
m2_asex_harmony_tr$donor_symptomatic <- paste0(m2_asex_harmony_tr$donor, m2_asex_harmony_tr$Symptomatic, sep = "_")

# generate violin plot 
Idents(m2_asex_harmony_tr) <- "stage.sympt"
# print(merge_dat[merge_dat$gene%in%common[1:2],c('gene','p_val.sc','p_val.bulk')])
```

```{r, fig.width=15, fig.height=5}
m2_asex_harmony_tr$stage.sympt <- paste(m2_asex_harmony_tr$StagePrelim, m2_asex_harmony_tr$Symptomatic, sep = "_")
Idents(m2_asex_harmony_tr) <- "stage.sympt"

```

```{r, fig.width=7, fig.height=3.5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-0114500", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1240400", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1233300", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-0819500", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1401300", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
```


```{r, fig.width=6, fig.height=3}
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Late ring"],  reduction = "umap.harmony", features = "PF3D7-0114500", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Late ring"],  reduction = "umap.harmony", features = "PF3D7-1240400", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Late ring"],  reduction = "umap.harmony", features = "PF3D7-1233300", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Late ring"],  reduction = "umap.harmony", features = "PF3D7-0819500", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Late ring"],  reduction = "umap.harmony", features = "PF3D7-1401300", split.by = c("Symptomatic"), ncol = 2, order = T)


```


```{r, fig.width=15, fig.height=5}
VlnPlot(pseudo_m2_asex, features = "PF3D7-0114500", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 
VlnPlot(pseudo_m2_asex, features = "PF3D7-1240400", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 
VlnPlot(pseudo_m2_asex, features = "PF3D7-1233300", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 
VlnPlot(pseudo_m2_asex, features = "PF3D7-0819500", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 
VlnPlot(pseudo_m2_asex, features = "PF3D7-1401300", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 

```

```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2_asex) <- "stage.sympt"

bulk.ET.de <- FindMarkers(object = pseudo_m2_asex, 
                         ident.1 = "Early trophozoite_Yes", 
                         ident.2 = "Early trophozoite_No",
                         test.use = "DESeq2")

head(bulk.ET.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))

```

```{r, fig.width=7, fig.height=3.5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1401300", idents = c("Early trophozoite_Yes", "Early trophozoite_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1240400", idents = c("Early trophozoite_Yes", "Early trophozoite_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1469400", idents = c("Early trophozoite_Yes", "Early trophozoite_No"), group.by = "donor_symptomatic")  + theme(legend.position = "none")
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-0321600", idents = c("Early trophozoite_Yes", "Early trophozoite_No"), group.by = "donor_symptomatic") + theme(legend.position = "none")
```


```{r, fig.width=6, fig.height=3}
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Early trophozoite"],  reduction = "umap.harmony", features = "PF3D7-1401300", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Early trophozoite"],  reduction = "umap.harmony", features = "PF3D7-1240400", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Early trophozoite"],  reduction = "umap.harmony", features = "PF3D7-1469400", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Early trophozoite"],  reduction = "umap.harmony", features = "PF3D7-0321600", split.by = c("Symptomatic"), ncol = 2, order = T)


```


```{r, fig.width=10, fig.height=5}
DimPlot(m2_asex_harmony_tr, reduction = "umap.harmony", group.by = c("Symptomatic", "StagePrelim"), ncol = 2)
```

```{r, fig.width=10, fig.height=5}
DimPlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Late ring"], reduction = "umap.harmony", split.by = c("Symptomatic"), ncol = 2)
```

```{r, fig.width=10, fig.height=5}
DimPlot(m2_asex_harmony_tr[,m2_asex_harmony_tr$StagePrelim == "Early trophozoite"], reduction = "umap.harmony", split.by = c("Symptomatic"), ncol = 2)
```

```{r, fig.width=10, fig.height=5}
FeaturePlot(m2_asex_harmony_tr,  reduction = "umap.harmony", features = "PF3D7-0114500", split.by = c("Symptomatic"), ncol = 2, order = T)
```


```{r, fig.width=10, fig.height=5}
FeaturePlot(m2_asex_harmony_tr,  reduction = "umap.harmony", features = "PF3D7-1240400", split.by = c("Symptomatic"), ncol = 2, order = T)
```

```{r, fig.width=30, fig.height=10}
FeaturePlot(m2_asex_harmony_tr, features = "PF3D7-1240400", split.by = "donor_symptomatic", ncol = 5)
```
```{r, fig.width=30, fig.height=10}
FeaturePlot(m2_asex_harmony_tr, features = "PF3D7-0114500", split.by = "donor_symptomatic", ncol = 5)
```

```{r, fig.width=10, fig.height=5}
FeaturePlot(m2_asex_harmony_tr,  reduction = "umap.harmony", features = "PF3D7-1240400", split.by = c("Symptomatic"), ncol = 2, order = T)
FeaturePlot(m2_asex_harmony_tr,  reduction = "umap.harmony", features = "PF3D7-0114500", split.by = c("Symptomatic"), ncol = 2, order = T)
```



```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(m2_asex_harmony_tr, features = "PF3D7-1240400", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic")
```


```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# m2_asex_harmony_tr <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]
```


```{r, fig.width=12, fig.height=4}
# add stage metadata
# harmony_int <- harmony_int@meta.data %>% 
#   rownames_to_column('bcode') %>%
#   mutate(donor = str_remove(orig.ident, ".mrna"), 
#          strain_don = case_when(str_detect(orig.ident, "MSC") ~ paste0(Strain_DN, "_", donor))
#          ) 
#   select(bcode, donor, strain_don, donor, Symptomatic, Hb_Type) %>%
#   column_to_rownames('bcode') %>% 
#   AddMetaData(harmony_int,.)


```


```{r}
harmony_int <-  readRDS("data/processed/Pf/m2_all/harmony_int.RDS")
```


```{r}
harmony_int <- JoinLayers(harmony_int)
```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot
harmony_int@meta.data %>% dplyr::count(Symptomatic, donor) 

```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot
harmony_int@meta.data %>% dplyr::count(Symptomatic, donor, StagePrelimC)  %>% pivot_wider(names_from = StagePrelimC, values_from = n)

```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot
harmony_int@meta.data %>% dplyr::count(Symptomatic, donor)

```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# pseudobulk the counts based on donor-condition-celltype
pseudo_m2 <- AggregateExpression(harmony_int, assays = "RNA", return.seurat = T, group.by = c("Symptomatic", "donor", "StagePrelim"))

# each 'cell' is a donor-condition-celltype pseudobulk profile
tail(Cells(pseudo_m2))
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
pseudo_m2$stage.sympt <- paste(pseudo_m2$StagePrelim, pseudo_m2$Symptomatic, sep = "_")
```


```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2) <- "stage.sympt"

m2_bulk.Fem.de <- FindMarkers(object = pseudo_m2, 
                         ident.1 = "Female_Yes", 
                         ident.2 = "Female_No",
                         test.use = "DESeq2")

head(m2_bulk.Fem.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))
```

```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
# ifnb$donor_id.stim <- paste0(ifnb$stim, "-", ifnb$donor_id)
harmony_int$donor_symptomatic <- paste0(harmony_int$donor, harmony_int$Symptomatic, sep = "_")

# generate violin plot 
Idents(harmony_int) <- "stage.sympt"
# print(merge_dat[merge_dat$gene%in%common[1:2],c('gene','p_val.sc','p_val.bulk')])
```

```{r, fig.width=15, fig.height=5}
harmony_int$stage.sympt <- paste(harmony_int$StagePrelim, harmony_int$Symptomatic, sep = "_")
Idents(harmony_int) <- "stage.sympt"

```

```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(harmony_int, features = "PF3D7-1101300", idents = c("Female_Yes", "Female_No"), group.by = "donor_symptomatic") 
```


```{r, fig.width=10, fig.height=5}
FeaturePlot(harmony_int[,harmony_int$StagePrelim == "Female"],  reduction = "umap.harmony", features = "PF3D7-1101300", split.by = c("Symptomatic"), ncol = 2, order = T)

```

```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2) <- "stage.sympt"

m2_bulk.FLE.de <- FindMarkers(object = pseudo_m2, 
                         ident.1 = "Female LE_Yes", 
                         ident.2 = "Female LE_No",
                         test.use = "DESeq2")
head(m2_bulk.FLE.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))
```

```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2) <- "stage.sympt"

m2_bulk.M.de <- FindMarkers(object = pseudo_m2, 
                         ident.1 = "Male_Yes", 
                         ident.2 = "Male_No",
                         test.use = "DESeq2")
head(m2_bulk.M.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))
```


```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2) <- "stage.sympt"

m2_bulk.ET.de <- FindMarkers(object = pseudo_m2, 
                         ident.1 = "Early trophozoite_Yes", 
                         ident.2 = "Early trophozoite_No",
                         test.use = "DESeq2")
head(m2_bulk.ET.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))
```

```{r, fig.width=15, fig.height=5}
Idents(pseudo_m2) <- "stage.sympt"

m2_bulk.LR.de <- FindMarkers(object = pseudo_m2, 
                         ident.1 = "Late ring_Yes", 
                         ident.2 = "Late ring_No",
                         test.use = "DESeq2")
head(m2_bulk.LR.de, n = 15) %>% rownames_to_column('gene_id')  %>% left_join(Pf3D7_genes_plasmoDB%>% select(gene_id, Gene))
```



```{r, fig.width=15, fig.height=5}
VlnPlot(pseudo_m2, features = "PF3D7-0114500", idents = c("Late ring_Yes", "Late ring_No"), group.by = "stage.sympt", ncol = 1) 
```


```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
# ifnb$donor_id.stim <- paste0(ifnb$stim, "-", ifnb$donor_id)
harmony_int$donor_symptomatic <- paste0(harmony_int$donor, harmony_int$Symptomatic, sep = "_")

# generate violin plot 
Idents(harmony_int) <- "stage.sympt"
# print(merge_dat[merge_dat$gene%in%common[1:2],c('gene','p_val.sc','p_val.bulk')])
```

```{r, fig.width=15, fig.height=5}
harmony_int$stage.sympt <- paste(harmony_int$StagePrelim, harmony_int$Symptomatic, sep = "_")
Idents(harmony_int) <- "stage.sympt"

```

```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(harmony_int, features = "PF3D7-1401300", idents = c("Late ring_Yes", "Late ring_No"), group.by = "donor_symptomatic") 
```


```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(harmony_int, features = "PF3D7-1101300", idents = c("Female_Yes", "Female_No"), group.by = "donor_symptomatic") 
```


```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(harmony_int, features = "PF3D7-1401300", idents = c("Early trophozoite_Yes", "Early trophozoite_No"), group.by = "donor_symptomatic") 
```

```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(harmony_int, features = "PF3D7-1240400", idents = c("Early trophozoite_Yes", "Early trophozoite_No"), group.by = "donor_symptomatic") 
```

```{r, fig.width=15, fig.height=5}
# create a new column to annotate sample-condition-celltype in the single-cell dataset
VlnPlot(harmony_int, features = "PF3D7-1101300", idents = c("Female_Yes", "Female_No"), group.by = "donor_symptomatic") 
```


#### Slingshot


```{r}
## slingshot
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "harmony_clusters_asex")
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "dset")
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", split.by = "dset", label = T)
```


```{r}
## slingshot
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "harmony_clusters_asex")
DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "dset")
```


```{r}
v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]@meta.data %>% count(harmony_clusters_asex)
```

```{r}
fld_clusters <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]@meta.data %>% count(harmony_clusters_asex) %>% pull(harmony_clusters_asex) 

fld_clusters_cutoff <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]@meta.data %>% count(harmony_clusters_asex) %>% filter(n>200) %>% pull(harmony_clusters_asex) 
```

```{r}
## Retained
DimPlot(v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$harmony_clusters_asex %in% fld_clusters_cutoff], reduction = "umap.harmony", split.by = "dset", label = T)
```

```{r}
## Removed
DimPlot(v3_m2_asex_harmony[,!(v3_m2_asex_harmony@meta.data$harmony_clusters_asex %in% fld_clusters_cutoff)], reduction = "umap.harmony", split.by = "dset", label = T)
```


```{r}
## subset where field parasites are for slingshot
v3_m2_asex_harmony_fld_sset <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$harmony_clusters_asex %in% fld_clusters_cutoff]
```


```{r, fig.width=8, fig.height=5}
DimPlot(v3_m2_asex_harmony_fld_sset,  reduction = "umap.harmony", split.by = c("Symptomatic"))
```

## Trajectory inference of integrated object
Adjust to get appropriate clustering so as to input into slingshot
```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"

DimPlot(v3_m2_asex_harmony_fld_sset, reduction = "umap.harmony", label = T)
# msc_gam_combi_noLE <- FindClusters(msc_gam_combi_noLE, resolution = 0.32, verbose = F)

```

```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
DimPlot(v3_m2_asex_harmony_fld_sset, reduction = "umap.harmony", group.by = "harmony_clusters_asex", pt.size = 0.5, dims = c(1,2), label = T)

```

```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
# cl_markrs <- FindMarkers(v3_m2_asex_harmony_fld_sset, ide)

```


```{r}
## 
v3_m2_asex_harmony_fld_sset <- FindClusters(v3_m2_asex_harmony_fld_sset, resolution = 0.05, cluster.name = "harmony_clusters_asex")


```

```{r}
## Adjusting resolution to find the optimum for slingshot trajectory inference to avoid error  "values must be length 150, but FUN(X[[1]]) result is length 38689"
DimPlot(v3_m2_asex_harmony_fld_sset, reduction = "umap.harmony", group.by = "harmony_clusters_asex", pt.size = 0.5, dims = c(1,2), label = T)

```

```{r}
## subset where field parasites are for slingshot
saveRDS(v3_m2_asex_harmony_fld_sset, "data/processed/Pf/m2_all/v3_m2_asex_harmony_fld_sset.RDS")
```

```{r, eval=T}
# gogo()
## Run SNORM on farm like below
# nextflow run submsn_slingshot.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/slingshot.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/v3_m2_asex_harmony_fld_sset.RDS" --o_file_name "v3_m2_asex_harmony_fld_sshot.RDS" --umap_nm "umap.harmony" --clstr "harmony_clusters_asex" --s_clstr "0" --e_clstr "2" --resume
```


```{r, eval=T}
# gogo()
## Run SNORM on farm like below
# nextflow run submsn_slingshot.nf --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/slingshot.R" --input_seu "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/v3_m2_asex_harmony_fld_sset.RDS" --o_file_name "v3_m2_asex_harmony_fld_sshot.RDS" --umap_nm "umap.harmony" --clstr "StageFL" --s_clstr "late ring" --e_clstr "early trophozoite" --resume
```

```{r, eval=T}
v3_m2_asex_harmony_sce_ss <- readRDS("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pf/m2_all/v3_m2_asex_harmony_fld_sshot.RDS")
```

```{r, eval =F}
##Raw counts - Preparing single cell experiment object for slingshot trajectory inference and tradeseq DE

#logcounts(msc14_sub_sce[[i]]) <- log2(counts(msc14_sub_sce[[i]]) + 1)
##This command gives error below when running the entire notebook or this whole chunk so run the command independently - coulb be and issue with rmd
##Error in x$.self$finalize() : attempt to apply non-function - need blank line after

# sce_obj_creatn <- function( seu_obj, umap = "umap"){
# 
#     sce_obj <- SingleCellExperiment(assays = list(counts = as(seu_obj[["RNA"]]$counts, Class = "dgCMatrix")), colData = seu_obj@meta.data)
#     
#     logcounts(sce_obj) <- log2((counts(sce_obj) + 1))
#     normcounts(sce_obj) <- as(seu_obj[["RNA"]]$data, Class = "dgCMatrix")
#     rowData(sce_obj)$feature_symbol <- str_replace(rownames(sce_obj), "-", "_")
#     reducedDims(sce_obj) <- SimpleList(PCA = seu_obj@reductions[["pca"]]@cell.embeddings, UMAP = seu_obj@reductions[[umap]]@cell.embeddings)
#     
#     return(sce_obj)
# 
# }

```


```{r, eval =F}
# v3_m2_asex_harmony_sce <- sce_obj_creatn(seu_obj = v3_m2_asex_harmony, umap = "umap.harmony")
```

Run slingshot for trajectory inference
```{r, eval =F}
##Function -slingshot

##Run slingshot to determine trajectories
# msc_y21_mca_fem_combi_sce$seurat_clusters <- as.character(msc_y21_mca_fem_combi_sce$seurat_clusters)
# msc_y21_mca_fem_combi_sce <- list(msc_y21_mca_fem_combi_sce, start.clus, end.clus) %>%
# sling_start <- function(sce_obj, s_clstr, clstr ="seurat_clusters", e_clstr = NULL){ 
#   slingshot(sce_obj, 
#             clusterLabels = clstr, 
#             reducedDim = "PCA", 
#             start.clus = s_clstr,
#             end.clus = e_clstr,
#             shrink =1, 
#             stretch = 1,
#             allow.break = F)
#   
# }

```


```{r, fig.width=6, fig.height=5, eval =F}
# visualize by batch and cell type annotation
# cell type annotations were previously added by Azimuth
# DimPlot(v3_m2_asex_harmony, reduction = "umap.harmony", group.by = "StagePrelimC", label = F, dims = c(1,2)) + scale_color_manual(values = sp_fld_colors)
```

```{r, eval =F}
# msc_gam_combi_noLE_sce <- sling_start(sce_obj = msc_gam_combi_noLE_sce, s_clstr = "early stalk", clstr = "stage_int")
# v3_m2_asex_harmony_sce_ss <- sling_start(sce_obj = v3_m2_asex_harmony_sce, s_clstr = "11", clstr = "harmony_clusters_asex")

```

```{r}
##Prepare slingshot dataset (SDS) object 
v3_m2_asex_harmony_sce_ss_sds <- SlingshotDataSet(v3_m2_asex_harmony_sce_ss)

plot(reducedDims(v3_m2_asex_harmony_sce_ss)$PCA[,1:2], col = brewer.pal(9,"Set1")[as.factor(v3_m2_asex_harmony_sce_ss$StagePrelim)], pch=16, asp = 0.7)
lines(v3_m2_asex_harmony_sce_ss_sds, lwd=2, col="black",dims = 1:2)
```




```{r, results="asis"}

umap_meta_df <- reducedDims(v3_m2_asex_harmony_sce_ss)$UMAP[,c(1,2,3)] %>% as.data.frame() %>%rownames_to_column("cell_bc") %>% left_join(., v3_m2_asex_harmony_sce_ss@colData[,c("StagePrelim", "slingPseudotime_1", "slingPseudotime_2", "orig.ident")] %>% as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc")  #%>% 
# mutate(across(strain_dset, ~factor(., levels = rev(strain_dset_lvls_don)))) %>% arrange(strain_dset

tm <- theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 22) ,
        legend.title = element_text(size = 24, face = "bold") )

umap_meta_df %>% 
  ggplot(aes(x = umapharmony_1, y =umapharmony_2, color = slingPseudotime_1)) + geom_point(size=0.5) + 
  scale_color_distiller(name="Pseudotime", direction = 1, palette = "Blues") + 
  tm

umap_meta_df %>% 
  ggplot(aes(x = umapharmony_2, y =umapharmony_3, color = slingPseudotime_1)) + geom_point(size=0.5) + 
  scale_color_distiller(name="Pseudotime", direction = 1, palette = "Blues") + 
  tm


```

```{r, results="asis"}

# plot(reducedDims(v3_m2_asex_harmony_sce_ss)$UMAP[,1:2], col = brewer.pal(7,"Set1")[as.factor(v3_m2_asex_harmony_sce_ss$harmony_clusters_asex)], pch=16, asp = 0.7)
# lines(v3_m2_asex_harmony_sce_ss_sds, lwd=2, col="black",dims = 1:2)
# 
# plot(reducedDims(v3_m2_asex_harmony_sce_ss)$UMAP[,2:3], col = brewer.pal(7,"Set1")[as.factor(v3_m2_asex_harmony_sce_ss$harmony_clusters_asex)], pch=16, asp = 0.7)
# lines(v3_m2_asex_harmony_sce_ss_sds, lwd=2, col="black",dims = 2:3)

umap_meta_df <- reducedDims(v3_m2_asex_harmony_sce_ss)$UMAP[,c(1,2)] %>% as.data.frame() %>%rownames_to_column("cell_bc") %>% left_join(., v3_m2_asex_harmony_sce_ss@colData[,c("StagePrelim", "slingPseudotime_1", "slingPseudotime_2", "orig.ident", "StrainDon", "Symptomatic")] %>% as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc")  #%>% 
# mutate(across(strain_dset, ~factor(., levels = rev(strain_dset_lvls_don)))) %>% arrange(strain_dset

plt <- umap_meta_df %>% 
  ggplot(aes(x = umapharmony_1, y =umapharmony_2, color = slingPseudotime_1)) + geom_point(size=0.5) + 
  scale_color_distiller(name="Pseudotime", direction = 1, palette = "Blues") + 
  theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 22) ,
        legend.title = element_text(size = 24, face = "bold") )

print(plt)

plt <- umap_meta_df %>% 
  ggplot(aes(x = umapharmony_1, y =umapharmony_2, color = slingPseudotime_2)) + geom_point(size=0.5) + 
  scale_color_distiller(name="Pseudotime", direction = 1, palette = "Greens") + 
  theme_classic()+
  theme(axis.text=element_blank(), 
        axis.title=element_blank(),
        legend.text = element_text(size = 22) ,
        legend.title = element_text(size = 24, face = "bold") )

print(plt)

plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_1, col=StagePrelim, fill=StagePrelim)) + 
  scale_color_manual(name = "Stage", values = sp_fld_colors)+ 
  scale_fill_manual(name = "Stage", values = sp_fld_colors)+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)


plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_2, col=StagePrelim, fill=StagePrelim)) + 
  scale_color_manual(name = "Stage", values = sp_fld_colors)+ 
  scale_fill_manual(name = "Stage", values = sp_fld_colors)+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)

plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_1, col=orig.ident, fill=orig.ident)) + 
  # scale_color_manual(name = "Stage")+
  # scale_fill_manual(name = "Stage")+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)

plt = umap_meta_df %>% 
  ggplot(., aes(x = slingPseudotime_2, col=orig.ident, fill=orig.ident)) + 
  # scale_color_manual(name = "Stage")+
  # scale_fill_manual(name = "Stage")+
  geom_density(aes(y = ..count..*0.18), size=2, bw = 0.18, alpha=.2)+
  # geom_freqpoly()+
  theme_classic()+
  theme(axis.text=element_text(size=26),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        axis.title=element_text(size=30,face="bold"), 
        legend.text = element_text(size = 26) ,
        legend.title = element_text(size = 28, face = "bold"),
        # legend.box.margin = margin(t = 0, r = 4.1, b = 0, l = 0, unit = "cm"),#######
        # legend.position="bottom",
        # legend.direction="vertical"
  ) +
  labs(x = "Pseudotime", y="Counts") +
  guides(color = guide_legend(override.aes = list(size = 4), title.position = "top"))

print(plt)


```

## Placing the more sophisticated DE at the top to save on time since MAST findmarkers run takes long

## DREAMLET
## Proper DE accounting for factors
```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
v3_m2_asex_harmony_fld_sset@meta.data %>% filter(!str_detect(StrainDon, "Doublet|Negative") & StagePrelim %in% c("Early trophozoite", "Late ring"))  %>% group_by(StrainDon,StagePrelim) %>% filter(n() > 10) %>% dplyr::count(StrainDon, StagePrelim)
```


```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
v3_m2_asex_harmony_fld_sset@meta.data %>% filter(!str_detect(StrainDon, "Doublet|Negative") & StagePrelim %in% c("Early trophozoite", "Late ring"))  %>% group_by(StrainDon, StagePrelim) %>% filter(n() > 10) %>% ungroup() %>% add_count(donor, name = "don_count") %>% add_count(StrainDon, name = "strn_count") %>% mutate(strn_prop = strn_count/don_count) %>% dplyr::count(StrainDon, StagePrelim,don_count,strn_count,strn_prop)
```


```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
irods_id_sk21_mdata_cln
```

```{r, fig.width=15, fig.height=5}
donor_mdata <- read_csv("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/MRC_Sample_Manifest_METADATA_raw.csv") 
donor_mdata <- donor_mdata %>% rename_all(., ~str_to_lower(str_replace_all(.,c("\\["="_", "\\]"="", " "="", "-"="",  "\\/"="_", "º"=""))))
```




```{r, fig.width=15, fig.height=5}
library(lubridate)
donor_mdata <- donor_mdata %>%
  filter(!is.na(short_name)) %>% 
  mutate(across(c(parasitaemia_j0, parasitaemia_j1), ~as.numeric(str_remove(., " Pf.*|Tf.*"))), 
         parasitaemia = coalesce(parasitaemia_j0, parasitaemia_j1),
         sampling_time_cln = lubridate::hms(paste0(str_remove(time_of_sampling, c(" AM| PM")), ":00")),
         processing_time_cln = lubridate::hms(time_of_sample_processing))  %>%
  mutate(smpl_wait_time = as.numeric(as.duration(processing_time_cln - sampling_time_cln),"minutes")) 

donor_mdata_slct <- donor_mdata %>% 
  select(short_name, parasitaemia, smpl_wait_time, age_years, gender_m_f, weight_kg)
```

```{r}
donor_mdata   %>% 
  mutate(across(c(parasitaemia_j0, parasitaemia_j1), ~as.numeric(str_remove(., " Pf.*|Tf.*"))), parasitaemia = coalesce(parasitaemia_j0, parasitaemia_j1))  %>% 
  # filter(donor %in% ) %>%
  ggplot(aes(x = symptomatic, y = parasitaemia)) +
  geom_boxplot()
```

```{r, fig.width=15, fig.height=5}
donor_mdata  %>% select(age_years, symptomatic) %>%  ggplot(aes(x = symptomatic, y = age_years)) + geom_boxplot()
```

add metadata
```{r, fig.width=15, fig.height=5}
v3_m2_asex_harmony_fld_sset <- v3_m2_asex_harmony_fld_sset@meta.data  %>%
  rownames_to_column("bcode") %>% 
  select(bcode, donor) %>%
  left_join(., donor_mdata_slct, by=c("donor"="short_name"), relationship = "many-to-one") %>%
  select(bcode, parasitaemia, smpl_wait_time, age_years, gender_m_f, weight_kg) %>% 
  column_to_rownames("bcode") %>% 
  AddMetaData(v3_m2_asex_harmony_fld_sset, .)
```

Add pseudotime
```{r, fig.width=15, fig.height=5}
v3_m2_asex_harmony_fld_sset <- v3_m2_asex_harmony_fld_sset@meta.data %>% 
  rownames_to_column("cell_bc") %>% 
  select(cell_bc, donor) %>%
  left_join(., v3_m2_asex_harmony_sce_ss@colData[,c("slingPseudotime_1", "slingPseudotime_2")] %>% 
              as.data.frame() %>% rownames_to_column("cell_bc"), by = "cell_bc")%>%
  select(cell_bc, slingPseudotime_1, slingPseudotime_2) %>% 
  column_to_rownames("cell_bc") %>%
  AddMetaData(v3_m2_asex_harmony_fld_sset, .)
```

```{r, fig.width=15, fig.height=5}
donor_sk21_summary <- read_csv("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/MRC_Sample_Analysis_Summary_Summary.csv", skip = 1) 
donor_sk21_summary <- donor_sk21_summary %>% rename_all(., ~str_replace_all(.,c("\\["="_", "\\]"="", " "="", "-"="",  "\\/"="_", "º"=""))) %>% filter(!is.na(Sample_Name))
```



```{r, fig.width=15, fig.height=5}
donor_sk21_summary %>% filter(mixed_infection== "pf", mixed_infection2=="solo") %>%  select(Sample_Name, pf, pm, po, mixed_infection, mixed_infection2) 
```

```{r, fig.width=15, fig.height=5}
donor_sk21_summary %>% filter(mixed_infection== "pf" & mixed_infection2=="solo") %>%  select(Sample_Name, pf, pm, po, mixed_infection, mixed_infection2) 
```



```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# v3_m2_asex_harmony_fld_sset@meta.data %>% filter(!str_detect(StrainDon, "Doublet|Negative") & StagePrelim %in% c("Early trophozoite", "Late ring"))  %>% group_by(StrainDon,StagePrelim) %>% filter(n() > 10) %>% dplyr::count(StrainDon) %>% group_by(donor) %>% mutate(prop = n/sum(n))
```

## check the strains going into the analysis
```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# v3_m2_asex_harmony_fld_sset@meta.data  %>% filter(count(donor, StagePrelim) > 20) %>% filter(!str_detect(StrainDon, "Doublet|Negative") & StagePrelim %in% c("Early trophozoite", "Late ring")) 
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
v3_m2_asex_harmony@meta.data %>% dplyr::count(StrainDon, StagePrelim)
```



```{r}
## Weird cluster 2 markers
v3_m2_asex_harmony_fld_sset@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, Symptomatic, StrainDon) %>% filter(n>20) %>% group_by(donor) %>% filter(n() > 1 )

```


```{r}
## Weird cluster 2 markers
v3_m2_asex_harmony_fld_sset@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, Symptomatic, StrainDon) %>% filter(n>20) 

```

```{r}
st_dist <- v3_m2_asex_harmony_fld_sset@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, Symptomatic, StrainDon) %>% filter(n>20) %>% group_by(donor) 

```

```{r}
st_dist %>% mutate(across(Symptomatic, ~str_replace_all(., c("No" = "Asymptomatic", "Yes" = "Symptomatic")))) %>% ggplot(aes(x = donor, fill = StrainDon, y = n)) + geom_col(position = position_dodge2()) + facet_wrap(Symptomatic~., scales = "free_x")
```


```{r, fig.width= 10, fig.height= 4}
v3_m2_asex_harmony_fld_sset@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, Symptomatic) %>% filter(n>20)  %>% mutate(across(Symptomatic, ~str_replace_all(., c("No" = "Asymptomatic", "Yes" = "Symptomatic")))) %>% ggplot(aes(x = donor, fill = donor, y = n)) + geom_col(position = position_dodge2()) + facet_wrap(Symptomatic~., scales = "free_x")
```


```{r, fig.width= 10, fig.height= 4}
v3_m2_asex_harmony_fld_sset@meta.data %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative"))) %>% count(donor, Symptomatic, StrainDon) %>% filter(n>20) %>% mutate(across(Symptomatic, ~str_replace_all(., c("No" = "Asymptomatic", "Yes" = "Symptomatic")))) %>% ggplot(aes(x = donor, fill = StrainDon, y = n)) + geom_col(position = position_dodge2()) + facet_wrap(Symptomatic~., scales = "free_x") + theme(legend.position = "right")
```

```{r, results="asis"}
# plotReducedDim(m2_asex_tr_sce, dimred = "UMAP.HARMONY")
```




Subset to retain only field asexual strains with great than 20 strains per straindon group. Remove stages with few cells - early rings and schizonts which are a minority and maybe poorly labelled
```{r, results="asis"}
qcd_asex <- v3_m2_asex_harmony_fld_sset@meta.data %>% rownames_to_column("bcode") %>% filter(dset != "Lab" & !(str_detect(StrainDon, "Doublet|Negative")) & StagePrelim %in% c("Early trophozoite", "Late ring"))  %>% add_count(donor, Symptomatic, StrainDon) %>% filter(n>20) %>% pull(bcode)
  
```

```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
m2_asex_harmony_tr <- v3_m2_asex_harmony_fld_sset[,qcd_asex]
```

```{r, fig.width=15, fig.height=5}
m2_asex_tr_sce <- as.SingleCellExperiment(m2_asex_harmony_tr)
```

## check the strains going into the analysis
```{r, fig.width=15, fig.height=5}
# visualize by batch and cell type annotation
# m2_asex_harmony_tr <- v3_m2_asex_harmony[,v3_m2_asex_harmony@meta.data$dset == "Field"]
# m2_asex_tr_sce <- as.SingleCellExperiment(m2_asex_harmony_tr)
```


```{r, results="asis"}
# only keep cells with sufficient reads
m2_asex_tr_sce <- m2_asex_tr_sce[rowSums(counts(m2_asex_tr_sce) > 0) > 0, ]

## 

# compute QC metrics
m2_asex_qc <- perCellQCMetrics(m2_asex_tr_sce)

# remove cells with few or many detected genes
ol <- isOutlier(metric = m2_asex_qc$detected, nmads = 2, log = TRUE)

table(ol)

```


```{r, results="asis"}
# only keep cells with sufficient reads
m2_asex_tr_sce <- m2_asex_tr_sce[, !ol]
plotReducedDim(m2_asex_tr_sce, dimred = "UMAP.HARMONY")
```
## pseudotime chunks
### Calculate pseudotime chunks
```{r}
##Generate pseudotime chunks categories with range of highest and lowest pseudotime values and breaks numbers to ensure balanced representation of cells and drop the cells without strain information

##Remove those without pseudotime
m2_asex_tr_sce_pt_chunks <- m2_asex_tr_sce[,!is.na(m2_asex_tr_sce@colData[,'slingPseudotime_1'])]

      # print(sce_obj)
m2_asex_tr_sce_pt_chunks@colData[,c('ptime_cat', 'ptime_bal_cat')] <- m2_asex_tr_sce_pt_chunks@colData[,c( "slingPseudotime_1"), drop=F] %>%
        data.frame() %>% 
        mutate(ptime_cat = cut(!!rlang::sym("slingPseudotime_1"),  
                               breaks = c(-Inf, seq(min(!!rlang::sym("slingPseudotime_1")),  max(!!rlang::sym("slingPseudotime_1")), round((max(!!rlang::sym("slingPseudotime_1")) - min(!!rlang::sym("slingPseudotime_1")))/7, 2)), Inf)),
               pt_rank = dense_rank(!!rlang::sym("slingPseudotime_1")),
               ptime_bal_cat = cut(pt_rank,  
                               breaks = c(-Inf, seq(min(pt_rank),  max(pt_rank), round((max(pt_rank) - min(pt_rank))/7, 2)), Inf))
               ) %>%
  mutate(ptime_grp_rnk = dense_rank(factor(ptime_cat)),
         ptime_bal_grp_rnk = dense_rank(factor(ptime_bal_cat)),
         ptime_cat = paste0("ptime_", ptime_grp_rnk),
         ptime_bal_cat = paste0("ptime_", ptime_bal_grp_rnk)
         ) %>%
        dplyr::select(ptime_cat, ptime_bal_cat)

```


```{r}
plotReducedDim(m2_asex_tr_sce_pt_chunks, dimred = "UMAP.HARMONY", colour_by = "ptime_bal_cat")
```

```{r}
plotReducedDim(m2_asex_tr_sce_pt_chunks, dimred = "UMAP.HARMONY", colour_by = "ptime_cat")
```

```{r}
m2_asex_tr_sce_pt_chunks@colData[,c("ptime_cat", "donor")] %>% as.data.frame() %>% dplyr::count(ptime_cat)
```

```{r}
m2_asex_tr_sce_pt_chunks_sset <- m2_asex_tr_sce_pt_chunks[,m2_asex_tr_sce_pt_chunks$ptime_cat %in% c("ptime_3", "ptime_4", "ptime_5")]
```

```{r}
m2_asex_tr_sce_pt_chunks_sset@colData[,c("ptime_cat", "donor")] %>% as.data.frame() %>% dplyr::count(ptime_cat)
```

```{r}
# ##Generate pseudotime chunks categories with range of highest and lowest pseudotime values and breaks numbers to ensure balanced representation of cells and drop the cells without strain information
# cut_pt_denom <- rep(c(6,6,6,6), 2)
# # pts <- rep(c("female_pt", "female_pt", "male_pt", "male_pt" ),4)
# pts <- rep(c("female_pt", "male_pt" ),4)
# 
# 
# msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci_ptchunks <- list(msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci, pts, cut_pt_denom) %>%
#   pmap(
#     # possibly(
#     ~{
#       sce_obj <- ..1[,!is.na(..1@colData[,'dset'])]
#       # print(sce_obj)
#       sce_obj@colData[,'ptime_cat'] <- sce_obj@colData[,c( ..2), drop=F] %>%
#         data.frame() %>% 
#         mutate(ptime_cat = cut(!!rlang::sym(..2),  breaks = c(-Inf, seq(min(!!rlang::sym(..2)),  max(!!rlang::sym(..2)), round((max(!!rlang::sym(..2)) - min(!!rlang::sym(..2)))/..3, 2)), Inf))) %>%
#         dplyr::select(ptime_cat)
#       
#       # print(glimpse(a))
#       return(sce_obj)
#     }
#     # , 'e')
#   )
# 
# ##Retain only for MSC14 as we dont want to compare field vs lab for the other samples
# # msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci_ptchunks <- msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci_ptchunks[str_detect(names(msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci_ptchunks), "msc14")]
# msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci_ptchunks

```

### Calculate pseudotime chunks donor strains
```{r}
# ##Generate pseudotime chunks categories with range of highest and lowest pseudotime values and breaks numbers to ensure balanced representation of cells and drop the cells without strain information
# cut_pt_denom <- rep(c(6,6), 9)
# pts <- rep(c("female_pt",  "male_pt" ), 9)
# 
# 
# msc_gcombi_noLE_sub_sce_unbal_ptchunks <- list(msc_gcombi_noLE_sub_mf_sce_unbal_auto_ci_strainsAll, pts, cut_pt_denom) %>%
#   pmap(
#     # possibly(
#     ~{
#       sce_obj <- ..1[,!is.na(..1@colData[,'lf_strain'])]
#       # print(sce_obj)
#       sce_obj@colData[,'ptime_cat'] <- sce_obj@colData[,c( ..2), drop=F] %>%
#         data.frame() %>% 
#         mutate(ptime_cat = cut(!!rlang::sym(..2),  breaks = c(-Inf, seq(min(!!rlang::sym(..2)),  max(!!rlang::sym(..2)), round((max(!!rlang::sym(..2)) - min(!!rlang::sym(..2)))/..3, 2)), Inf))) %>%
#         dplyr::select(ptime_cat)
#       
#       # print(glimpse(a))
#       return(sce_obj)
#     }
#     # , 'e')
#   )
#   

```


```{r, results="asis"}

```


```{r, results="asis"}
traj_milo <- Milo(m2_asex_tr_sce)
```


```{r, results="asis"}
traj_milo <- buildGraph(traj_milo, k = 30, d = 30, reduced.dim = "PCA")
```

```{r, results="asis"}
traj_milo <- makeNhoods(traj_milo, prop = 0.1, k = 30, d=30, refined = TRUE, reduced_dims = "PCA")

```

```{r, results="asis"}
plotNhoodSizeHist(traj_milo)
```

```{r, results="asis"}
# traj_milo <- countCells(traj_milo, meta.data = data.frame(colData(traj_milo)), samples="StrainDon")
traj_milo <- countCells(traj_milo, meta.data = data.frame(colData(traj_milo)), sample="donor")
```

```{r, results="asis"}
head(nhoodCounts(traj_milo))
```

```{r, results="asis"}
traj_design <- data.frame(colData(traj_milo))[,c("donor", "Symptomatic")]
# traj_design <- data.frame(colData(traj_milo))[,c("StrainDon", "Symptomatic", "donor")]
traj_design <- distinct(traj_design)
# rownames(traj_design) <- traj_design$StrainDon
rownames(traj_design) <- traj_design$donor
## Reorder rownames to match columns of nhoodCounts(milo)
# traj_design <- traj_design[colnames(nhoodCounts(traj_milo)), , drop=FALSE]

traj_design

# embryo_design <- data.frame(colData(embryo_milo))[,c("sample", "stage", "sequencing.batch")]
# 
# ## Convert batch info from integer to factor
# embryo_design$sequencing.batch <- as.factor(embryo_design$sequencing.batch) 
# embryo_design <- distinct(embryo_design)
# rownames(embryo_design) <- embryo_design$sample
# 
# embryo_design
```

```{r, results="asis"}
traj_milo <- calcNhoodDistance(traj_milo, d=30)
```

```{r, results="asis"}
da_results <- testNhoods(traj_milo, design = ~ Symptomatic, design.df = traj_design)
# da_results <- testNhoods(traj_milo, design = ~ donor + Symptomatic, design.df = traj_design)
```

```{r, results="asis"}
da_results %>%
  arrange(SpatialFDR) %>%
  head() 
```


```{r, results="asis"}
ggplot(da_results, aes(PValue)) + geom_histogram(bins=50)
```

```{r, results="asis"}
ggplot(da_results, aes(logFC, -log10(SpatialFDR))) + 
  geom_point() +
  geom_hline(yintercept = 1) ## Mark significance threshold (10% FDR)
```


```{r, results="asis"}
traj_milo <- buildNhoodGraph(traj_milo)

```

```{r, results="asis"}
conflicts_prefer(graphics::layout)
## Plot single-cell UMAP
plotReducedDim(traj_milo, dimred = "UMAP.HARMONY", colour_by="Symptomatic", text_by = "seurat_clusters", 
                          text_size = 3, point_size=0.5)+
  guides(fill="none") + 
  plotNhoodGraphDA(traj_milo, da_results, alpha=0.05, layout = "UMAP.HARMONY") #+
  # plot_layout(guides="collect")



# ## Plot neighbourhood graph
# nh_graph_pl <- plotNhoodGraphDA(embryo_milo, da_results, layout="umap",alpha=0.1) 
#   
# umap_pl + nh_graph_pl +
#   plot_layout(guides="collect")
```

```{r, results="asis"}
# da_results <- annotateNhoods(traj_milo, da_results, coldata_col = "ptime_cat")
da_results <- annotateNhoods(traj_milo, da_results, coldata_col = "seurat_clusters")
head(da_results)
```


```{r, results="asis"}
ggplot(da_results, aes(seurat_clusters_fraction)) + geom_histogram(bins=50)
# ggplot(da_results, aes(ptime_cat_fraction)) + geom_histogram(bins=50)
```

```{r, results="asis"}
da_results$seurat_clusters <- ifelse(da_results$seurat_clusters_fraction < 0.7, "Mixed", da_results$seurat_clusters)
```


```{r, results="asis"}
plotDAbeeswarm(da_results, group.by = "seurat_clusters")
```

```{r, results="asis"}
## Add log normalized count to Milo object
traj_milo <- logNormCounts(traj_milo)

da_results$NhoodGroup <- as.numeric(da_results$SpatialFDR < 0.1 & da_results$logFC < 0)
# da_nhood_markers <- findNhoodGroupMarkers(traj_milo, da_results, subset.row = rownames(traj_milo)[1:10])

```

```{r, results="asis"}
da_nhood_markers <- findNhoodGroupMarkers(traj_milo, da_results, subset.row = rownames(traj_milo), 
                                          aggregate.samples = TRUE, sample_col = "donor")

```

```{r, results="asis"}
head(da_nhood_markers)
```

Automatic grouping of neighbourhoods

```{r, results="asis"}
## Run buildNhoodGraph to store nhood adjacency matrix
traj_milo <- buildNhoodGraph(traj_milo)

## Find groups
da_results <- groupNhoods(traj_milo, da_results, max.lfc.delta = 2)
head(da_results)
```

```{r, results="asis"}
plotNhoodGroups(traj_milo, da_results, layout="UMAP.HARMONY") 
```



```{r, results="asis"}
plotDAbeeswarm(da_results, group.by = "NhoodGroup")
```

```{r, results="asis"}
# trial with parameters adjusted
plotDAbeeswarm(groupNhoods(traj_milo, da_results, max.lfc.delta = 1) , group.by = "NhoodGroup") + ggtitle("max LFC delta=1")
plotDAbeeswarm(groupNhoods(traj_milo, da_results, max.lfc.delta = 2)   , group.by = "NhoodGroup") + ggtitle("max LFC delta=2")
plotDAbeeswarm(groupNhoods(traj_milo, da_results, max.lfc.delta = 3)   , group.by = "NhoodGroup") + ggtitle("max LFC delta=3")
```


```{r, results="asis"}
# trial with parameters adjusted
plotDAbeeswarm(groupNhoods(traj_milo, da_results, max.lfc.delta = 5, overlap=1) , group.by = "NhoodGroup") + ggtitle("overlap=5")
plotDAbeeswarm(groupNhoods(traj_milo, da_results, max.lfc.delta = 5, overlap=5)   , group.by = "NhoodGroup") + ggtitle("overlap=10")
plotDAbeeswarm(groupNhoods(traj_milo, da_results, max.lfc.delta = 5, overlap=10)   , group.by = "NhoodGroup") + ggtitle("overlap=20")
```


```{r, results="asis"}
set.seed(42)
da_results <- groupNhoods(traj_milo, da_results, max.lfc.delta = 10, overlap=1)
plotNhoodGroups(traj_milo, da_results, layout="UMAP.HARMONY")
```


```{r, results="asis"}
plotDAbeeswarm(da_results, group.by = "NhoodGroup")
```



Finding gene signatures for neighbourhoods

```{r, results="asis"}
## Exclude zero counts genes
keep.rows <- rowSums(logcounts(traj_milo)) != 0
traj_milo <- traj_milo[keep.rows, ]

## Find HVGs
dec <- modelGeneVar(traj_milo)
hvgs <- getTopHVGs(dec, n=200)
head(hvgs)
```

one-vs-all differential gene expression for each neighbourhood group
```{r, results="asis"}
nhood_markers <- findNhoodGroupMarkers(traj_milo, da_results, subset.row = hvgs, 
                                       aggregate.samples = TRUE, sample_col = "donor")

head(nhood_markers)
```

```{r, results="asis"}
gr2_markers <- nhood_markers[c("GeneID", "logFC_6", "adj.P.Val_6")] 
colnames(gr2_markers) <- c("GeneID", "logFC", "adj.P.Val")

head(gr2_markers[order(gr2_markers$adj.P.Val), ])
```
subset.groups tests 8-VS-all markers
```{r, results="asis"}
nhood_markers8 <- findNhoodGroupMarkers(traj_milo, da_results, subset.row = hvgs, 
                                       aggregate.samples = TRUE, sample_col = "donor",
                                       subset.groups = c("8")
                                       )

head(nhood_markers8)
```

Compare neighbourhoods - seems like it is each neighbour hood versus all the rest otherwise there should be only one p value
```{r, results="asis"}
nhood_markers2 <- findNhoodGroupMarkers(traj_milo, da_results, subset.row = hvgs,
                                       subset.nhoods = da_results$NhoodGroup %in% c('1','2'),
                                       aggregate.samples = TRUE, sample_col = "donor")

head(nhood_markers2)
```

```{r, results="asis"}
ggplot(nhood_markers2, aes(logFC_2,-log10(adj.P.Val_2 ))) + 
  geom_point(alpha=0.5, size=0.5) +
  geom_hline(yintercept = 2)
```

```{r, results="asis"}
markers <- nhood_markers2$GeneID[nhood_markers$adj.P.Val_2 < 0.3 & nhood_markers2$logFC_2 > 0]
```

```{r, results="asis"}
plotNhoodExpressionGroups(traj_milo, da_results, features=markers[1:20],
                          subset.nhoods = da_results$NhoodGroup %in% c('3','2'), scale=TRUE,
                          grid.space = "fixed")
```

DGE testing within a group
```{r, results="asis"}
dge_9 <- testDiffExp(traj_milo, da_results, design = ~ Symptomatic, meta.data = data.frame(colData(traj_milo)), subset.nhoods=da_results$NhoodGroup=="3")
```

```{r, results="asis"}
dge_9
```

Milo GLLM

```{r, results="asis"}
# traj_milo <- Milo(m2_asex_tr_sce_pt_chunks)
```

```{r, results="asis"}
traj_sce <- m2_asex_tr_sce_pt_chunks
```

```{r, results="asis"}
# uncomment the row below to allow multi-processing and comment out the SerialParam line.
# bpparam <- MulticoreParam(workers=4)
bpparam <- SerialParam()
register(bpparam)
```

```{r, results="asis"}
# set.seed(42)
# # remove sparser cells
# drop.cells <- colSums(counts(traj_sce)) < 600
# traj_sce <- computePooledFactors(traj_sce[, !drop.cells], BPPARAM=bpparam)
# traj_sce <- logNormCounts(traj_sce)
# 
# pbmc.hvg <- modelGeneVar(traj_sce)
# pbmc.hvg$FDR[is.na(pbmc.hvg$FDR)] <- 1
# 
# traj_sce <- runPCA(traj_sce, subset_row=rownames(traj_sce)[pbmc.hvg$FDR < 0.5], scale=TRUE, ncomponents=30, assay.type="logcounts", name="PCA", BPPARAM=bpparam)
# traj_sce
```

```{r, results="asis"}
# set.seed(42)
# traj_sce <- runUMAP(traj_sce, n_neighbors=30, pca=20, BPPARAM=bpparam) # add a UMAP for plotting results later
# 
# traj_milo_gllm <- Milo(traj_sce) # from the SCE object
# reducedDim(traj_milo_gllm, "UMAP") <- reducedDim(traj_sce, "UMAP")
# 
# plotUMAP(traj_milo_gllm, colour_by="StagePrelim") + plotUMAP(traj_milo_gllm, colour_by="donor")
```



```{r, results="asis"}
# set.seed(42)
# traj_milo_gllm <- Milo(m2_asex_tr_sce_pt_chunks) # from the SCE object
traj_milo_gllm <- Milo(m2_asex_tr_sce) # from the SCE object
reducedDim(traj_milo_gllm, "UMAP") <- reducedDim(m2_asex_tr_sce, "UMAP.HARMONY")

plotUMAP(traj_milo_gllm, colour_by="StagePrelim") + plotUMAP(traj_milo_gllm, colour_by="donor")
```

```{r, results="asis"}
set.seed(42)
# we build KNN graph
traj_milo_gllm <- buildGraph(traj_milo_gllm, k = 60, d = 30)
```

```{r, results="asis"}
traj_milo_gllm <- makeNhoods(traj_milo_gllm, prop = 0.2, k = 60, d=30, refined = TRUE, refinement_scheme="graph") # make nhoods using graph-only as this is faster
```

```{r, results="asis"}
# colData(traj_milo_gllm)$ObsID <- paste(colData(traj_milo_gllm)$tenx_lane, colData(traj_milo_gllm)$sampleid, sep="_")
traj_milo_gllm <- countCells(traj_milo_gllm, meta.data = data.frame(colData(traj_milo_gllm)), samples="StrainDon") # make the nhood X sample counts matrix
```

```{r, results="asis"}
head(nhoodCounts(traj_milo_gllm))
```

```{r, results="asis"}
plotNhoodSizeHist(traj_milo_gllm)
```

```{r, results="asis"}
symp.design <- data.frame(colData(traj_milo_gllm))[,c("donor", "StrainDon", "Symptomatic")]
symp.design <- distinct(symp.design)
rownames(symp.design) <- symp.design$StrainDon
## Reorder rownames to match columns of nhoodCounts(milo)
symp.design <- symp.design[colnames(nhoodCounts(traj_milo_gllm)), , drop=FALSE]
table(symp.design$Symptomatic)
```


```{r, results="asis"}
symp.design
```

```{r, results="asis", eval=FALSE}
set.seed(42)
rownames(symp.design) <- symp.design$StrainDon

da_results <- testNhoods(traj_milo_gllm, design = ~ Symptomatic + (1|donor), design.df = symp.design, fdr.weighting="graph-overlap", glmm.solver="HE-NNLS", REML=TRUE, norm.method="TMM", BPPARAM = bpparam, fail.on.error=FALSE, max.iters = 1000)
```

```{r, results="asis", eval=FALSE}
table(da_results$SpatialFDR < 0.99)
```

```{r, results="asis", eval=FALSE}
which(is.na(da_results$logFC))
```

```{r, results="asis", eval=FALSE}
which(checkSeparation(traj_milo_gllm, design.df=symp.design, condition="Symptomatic", min.val=5))
```

```{r, results="asis", eval=FALSE}
length(which(checkSeparation(traj_milo_gllm, design.df=symp.design, condition="Symptomatic", min.val=5)))
```
```{r, results="asis", fig.width=30, fig.height=20, eval=FALSE}
plotNhoodCounts(traj_milo_gllm, which(checkSeparation(traj_milo_gllm, design.df=symp.design, condition="Symptomatic", min.val=5)), design.df=symp.design, condition="Symptomatic")
```

```{r, results="asis", eval=FALSE}
head(da_results[!is.na(da_results$logFC), ])
```

```{r, results="asis", eval=FALSE}
da_results[!is.na(da_results$logFC) & da_results$Converged == T, ] %>% arrange(PValue)
```

```{r, results="asis", eval=FALSE}
traj_milo_gllm <- buildNhoodGraph(traj_milo_gllm, overlap=25)

# we need to subset the plotting results as it can't handle the NAs internally.
plotUMAP(traj_milo_gllm, colour_by="Symptomatic") + plotNhoodGraphDA(traj_milo_gllm, da_results[!is.na(da_results$logFC), ], subset.nhoods=!is.na(da_results$logFC), alpha=0.1) +  plot_layout(guides="auto" )
```


```{r, results="asis", eval=FALSE}
set.seed(42)
# we need to use place the test variable at the end of the formula
glm_results <- testNhoods(traj_milo_gllm, design = ~Symptomatic, design.df = symp.design, fdr.weighting="graph-overlap",
                          REML=TRUE, norm.method="TMM", BPPARAM = bpparam)
```

```{r, results="asis", eval=FALSE}
table(glm_results$SpatialFDR < 0.1)
```

```{r, results="asis", eval=FALSE}
plotUMAP(traj_milo_gllm, colour_by="Symptomatic") + plotNhoodGraphDA(traj_milo_gllm, glm_results, alpha=0.1) +
  plot_layout(guides="auto" )
```

```{r, results="asis", fig.width=12, fig.height=4, eval=FALSE}
comp.da <- merge(da_results, glm_results, by='Nhood')
comp.da$Sig <- "none"
comp.da$Sig[comp.da$SpatialFDR.x < 0.1 & comp.da$SpatialFDR.y < 0.1] <- "Both"
comp.da$Sig[comp.da$SpatialFDR.x >= 0.1 & comp.da$SpatialFDR.y < 0.1] <- "GLM"
comp.da$Sig[comp.da$SpatialFDR.x < 0.1 & comp.da$SpatialFDR.y >= 0.1] <- "GLMM"

ggplot(comp.da, aes(x=logFC.x, y=logFC.y)) +
    geom_point(data=comp.da[, c("logFC.x", "logFC.y")], aes(x=logFC.x, y=logFC.y),
               colour='grey80', size=1) +
    geom_point(aes(colour=Sig)) +
    labs(x="GLMM LFC", y="GLM LFC") +
    facet_wrap(~Sig) +
    NULL
```


```{r, results="asis", eval=FALSE}
gogo()
```

```{r, results="asis", eval=FALSE}
library(scRNAseq)
# uncomment the row below to allow multi-processing and comment out the SerialParam line.
# bpparam <- MulticoreParam(workers=4)
bpparam <- SerialParam()
register(bpparam)

pbmc.sce <- KotliarovPBMCData(mode="rna", ensembl=TRUE) # download the PBMC data from Kotliarov _et al._
```

```{r, results="asis", eval=FALSE}
# downsample cells to reduce compute time
prop.cells <- 0.75
n.cells <- floor(ncol(pbmc.sce) * prop.cells)
set.seed(42)
keep.cells <- sample(colnames(pbmc.sce), size=n.cells)
pbmc.sce <- pbmc.sce[, colnames(pbmc.sce) %in% keep.cells]


# downsample the number of samples
colData(pbmc.sce)$ObsID <- paste(colData(pbmc.sce)$tenx_lane, colData(pbmc.sce)$sampleid, sep="_")
n.samps <- 80
keep.samps <- sample(unique(colData(pbmc.sce)$ObsID), size=n.samps)
keep.cells <- rownames(colData(pbmc.sce))[colData(pbmc.sce)$ObsID %in% keep.samps]
pbmc.sce <- pbmc.sce[, colnames(pbmc.sce) %in% keep.cells]

pbmc.sce

```

```{r, results="asis", eval=FALSE}
pbmc.milo <- Milo(pbmc.sce) # from the SCE object
```

```{r, results="asis", eval=FALSE}
pbmc.design <- data.frame(colData(pbmc.milo))[,c("tenx_lane", "adjmfc.time", "sample", "sampleid", "timepoint", "ObsID")]
pbmc.design <- distinct(pbmc.design)
rownames(pbmc.design) <- pbmc.design$ObsID        
```


```{r, results="asis", eval=FALSE}
pbmc.design      
```


```{r, results="asis", eval=FALSE}
pbmc.design %>% count(tenx_lane, ObsID, adjmfc.time)   
```


```{r, results="asis", eval=FALSE}
pbmc.design %>%
  group_by(tenx_lane) %>%
  summarise(count = n_distinct(adjmfc.time)) 
```

```{r, results="asis", eval=FALSE}
symp.design %>% count(tenx_lane, ObsID)   
```

```{r, results="asis"}

```

```{r, results="asis"}

```


```{r, results="asis"}
# gogo()
```
## MUSCAT


```{r, results="asis"}
## solution to error https://github.com/HelenaLC/muscat/issues/88

m2_asex_tr_sce_musct <- prepSCE(m2_asex_tr_sce_pt_chunks, 
               kid = "ptime_cat", # subpopulation assignments
               gid = "Symptomatic",  # group IDs (ctrl/stim)
               sid = "donor",   # sample IDs (ctrl/stim.1234)
               drop = TRUE) #
```

```{r, results="asis"}
pb <- aggregateData(m2_asex_tr_sce_musct,
    assay = "counts", 
    fun = "sum")

# one sheet per subpopulation
assayNames(pb)
```

```{r, results="asis"}
t(head(assay(pb)))
```



```{r, results="asis"}
pb_mds <- pbMDS(pb)
# use very distinctive shaping of groups & change cluster colors
pb_mds <- pb_mds + 
  scale_shape_manual(values = c(17, 4)) +
  scale_color_manual(values = RColorBrewer::brewer.pal(8, "Set2"))
# change point size & alpha
pb_mds$layers[[1]]$aes_params$size <- 5
pb_mds$layers[[1]]$aes_params$alpha <- 0.6
# pb_mds
```

```{r, results="asis"}
# run DS analysis
res <- pbDS(pb, verbose = FALSE)
# access results table for 1st comparison
tbl <- res$table[[1]]
# one data.frame per cluster
names(tbl)
```

```{r, results="asis"}
# view results for 1st cluster
k1 <- tbl[[2]]
head(format(k1[, -ncol(k1)], digits = 2))
```

```{r, results="asis"}
# construct design & contrast matrix
ei <- metadata(m2_asex_tr_sce_musct)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("Yes-No", levels = mm)

# run DS analysis
pbDS(pb, design = mm, contrast = contrast)
```


```{r, results="asis"}
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil <- lapply(tbl, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.3, abs(logFC) > .01)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil, nrow, numeric(1))
p_de <- format(n_de / nrow(m2_asex_tr_sce_musct) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)
```



```{r, results="asis"}
# view top 2 hits in each cluster
top2 <- bind_rows(lapply(tbl_fil, top_n, 2, p_adj.loc))
format(top2[, -ncol(top2)], digits = 2)
```



```{r, results="asis"}
frq <- calcExprFreqs(m2_asex_tr_sce_musct, assay = "counts", th = 0)
# one sheet per cluster
assayNames(frq)
```


```{r, results="asis"}
# expression frequencies in each
# sample & group; 1st cluster
t(head(assay(frq), 5))
```



```{r, results="asis"}
gids <- levels(m2_asex_tr_sce_musct$group_id)
frq10 <- vapply(as.list(assays(frq)), 
  function(u) apply(u[, gids] > 0.00001, 1, any), 
  logical(nrow(m2_asex_tr_sce_musct)))
t(head(frq10))
```

```{r, results="asis"}
nk <- length(kids <- levels(m2_asex_tr_sce_musct$cluster_id))
ns <- length(sids <- levels(m2_asex_tr_sce_musct$sample_id))
names(kids) <- kids; names(sids) <- sids
```



```{r, results="asis"}
tbl_fil2 <- lapply(kids, function(k)
  tryCatch(
    expr = { dplyr::filter(tbl_fil[[k]], 
    gene %in% names(which(frq10[, k])))}, error = function(e) NULL))

tbl_fil2<-tbl_fil2[!sapply(tbl_fil2,is.null)]

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil2, nrow, numeric(1))
p_de <- format(n_de / nrow(m2_asex_tr_sce_musct) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)
```


```{r, results="asis", eval=FALSE}
# tidy format; attach pre-computed expression frequencies
resDS(m2_asex_tr_sce_musct, res, bind = "row", frq = frq)

# big-table (wide) format; attach CPMs
resDS(m2_asex_tr_sce_musct, res, bind = "col", cpm = TRUE)
```


```{r, results="asis"}
# compute expression frequencies on the fly
resDS(m2_asex_tr_sce_musct, res, frq = TRUE)
```


```{r, results="asis", eval=FALSE}
library(UpSetR)
de_gs_by_k <- map(tbl_fil, "gene")
upset(fromList(de_gs_by_k))
```


```{r, results="asis", eval=FALSE}
# pull top-8 DS genes across all clusters
top8 <- bind_rows(tbl_fil) %>% 
  slice_min(p_adj.loc, n = 8, 
    with_ties = FALSE) %>% 
  pull("gene")

# for ea. gene in 'top8', plot t-SNE colored by its expression 
ps <- lapply(top8, function(g)
  .plot_dr(m2_asex_tr_sce_musct[, cs100], "TSNE", g) + 
    ggtitle(g) + theme(legend.position = "none"))

# arrange plots
plot_grid(plotlist = ps, ncol = 4, align = "vh")
```


```{r, results="asis"}
plotExpression(m2_asex_tr_sce_musct[, m2_asex_tr_sce_musct$cluster_id == "ptime_4"],
  features = tbl_fil$ptime_3$gene[seq_len(1)],
  x = "sample_id", colour_by = "group_id", ncol = 3) +
  guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

```{r, results="asis"}
plotExpression(m2_asex_tr_sce_musct[, m2_asex_tr_sce_musct$cluster_id == "ptime_4"],
  features = tbl_fil$ptime_4$gene[seq_len(2)],
  x = "sample_id", colour_by = "group_id", ncol = 3) +
  guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r, results="asis", eval=FALSE}
# top-5 DS genes per cluster
pbHeatmap(m2_asex_tr_sce_musct, res[[1]], k = "ptime_4")
```


```{r, results="asis", eval=FALSE}
# top-5 DS genes per cluster
pbHeatmap(m2_asex_tr_sce_musct, res, top_n = 1)
```

## MUSCAT mixed

```{r, results="asis", eval=F}
## solution to error https://github.com/HelenaLC/muscat/issues/88

m2_asex_tr_sce_musct_sd <- prepSCE(m2_asex_tr_sce_pt_chunks, 
               kid = "ptime_cat", # subpopulation assignments
               gid = "Symptomatic",  # group IDs (ctrl/stim)
               sid = "StrainDon",   # sample IDs (ctrl/stim.1234)
               # drop = TRUE
               ) #
```

```{r, results="asis", eval=F}
# 1st approach
mm <- mmDS(m2_asex_tr_sce_musct_sd, method = "nbinom",
  n_cells = 10, n_samples = 2,
  min_count = 1, min_cells = 20)

# 2nd & 3rd approach
# mm <- mmDS(m2_asex_tr_sce_musct, method = "vst", vst = "sctransform")
# mm <- mmDS(m2_asex_tr_sce_musct, method = "nbinom")
```

## Dreamlet

```{r, results="asis"}
# Since 'ind' is the individual and 'StimStatus' is the stimulus status,
# create unique identifier for each sample
# m2_asex_tr_sce$id <- paste0(m2_asex_tr_sce$Symptomatic, m2_asex_tr_sce$donor, m2_asex_tr_sce$StrainDon)
m2_asex_tr_sce_pt_chunks$id <- paste0(m2_asex_tr_sce_pt_chunks$Symptomatic, m2_asex_tr_sce_pt_chunks$donor)


# Create pseudobulk data by specifying cluster_id and sample_id
# Count data for each cell type is then stored in the `assay` field
# assay: entry in assayNames(sce) storing raw counts
# cluster_id: variable in colData(sce) indicating cell clusters
# sample_id: variable in colData(sce) indicating sample id for aggregating cells
m2_asex_tr_pb <- aggregateToPseudoBulk(m2_asex_tr_sce_pt_chunks,
  assay = "counts",
  # cluster_id = "StagePrelim",
  cluster_id = "ptime_cat",
  sample_id = "id",
  verbose = FALSE
)

# one 'assay' per cell type
assayNames(m2_asex_tr_pb)
```



```{r, results="asis"}
# one 'assay' per cell type
assayNames(m2_asex_tr_pb)

metadata(m2_asex_tr_pb)$aggr_means
```


```{r, results="asis"}

# form <- ~ (1|donor) + Symptomatic + slingPseudotime_1
form <- ~ Symptomatic + slingPseudotime_1

# Normalize and apply voom/voomWithDreamWeights
res.proc <- processAssays(m2_asex_tr_pb, form, min.count = 2)

# the resulting object of class dreamletProcessedData stores
# normalized data and other information
res.proc
```

```{r, results="asis", eval=FALSE}
gogo()
```

```{r, results="asis"}
# Normalize and apply voom/voomWithDreamWeights
details(res.proc)
```

```{r, results="asis"}
# show voom plot for each cell clusters
plotVoom(res.proc)
```

```{r, results="asis"}

# run variance partitioning analysis
vp.lst <- fitVarPart(res.proc, form)
```

```{r, results="asis"}
# Show variance fractions at the gene-level for each cell type
genes <- vp.lst$gene[1:2]
plotPercentBars(vp.lst[vp.lst$gene %in% genes, ])
```


```{r, results="asis", fig.width=10, fig.height=10}
# Summarize variance fractions genome-wide for each cell type
plotVarPart(vp.lst, label.angle = 60)
```

```{r, results="asis"}
# Differential expression analysis within each assay,
# evaluated on the voom normalized data
res.dl <- dreamlet(res.proc, form, method = "random")

# names of estimated coefficients
coefNames(res.dl)
```

```{r, results="asis"}
# the resulting object of class dreamletResult
# stores results and other information
res.dl
```

```{r, results="asis"}
plotVolcano(res.dl, coef = "SymptomaticYes")
```

```{r, results="asis"}
# genes <- c("PF3D7-1240400", "PF3D7-1240900")
# plotGeneHeatmap(res.dl, coef = "SymptomaticYes", genes = genes)
```

```{r, results="asis"}
conflicts_prefer(dreamlet::topTable)
# results from full analysis
topTable(res.dl, coef = "SymptomaticYes")
```

```{r, results="asis"}
# only B cells
topTable(res.dl[["ptime_3"]], coef = "SymptomaticYes")
```

```{r, results="asis"}
plotForest(res.dl, coef = "SymptomaticYes", gene = c("PF3D7-1240400"))
```

```{r, results="asis"}
# get data
df <- extractData(res.proc, "ptime_3", genes = "PF3D7-1240400")

# set theme
thm <- theme_classic() +
  theme(aspect.ratio = 1, plot.title = element_text(hjust = 0.5))

# make plot
ggplot(df, aes(Symptomatic, `PF3D7-1240400`)) +
  geom_boxplot() +
  thm +
  ylab(bquote(Expression ~ (log[2] ~ CPM))) +
  ggtitle("PF3D7-1240400")
```

```{r, results="asis"}
# create a contrasts called 'Diff' that is the difference between expression
# in the stimulated and controls.
# More than one can be specified
contrasts <- c(Diff = "SymptomaticYes - SymptomaticNo")

# Evalaute the regression model without an intercept term.
# Instead estimate the mean expression in stimulated, controls and then
# set Diff to the difference between the two
res.dl2 <- dreamlet(res.proc, ~ 0 + Symptomatic, contrasts = contrasts)

# see estimated coefficients
coefNames(res.dl2)
```

```{r, results="asis"}
# Volcano plot of Diff
plotVolcano(res.dl2[1:2], coef = "Diff")
```

```{r, results="asis"}
# results from full analysis
topTable(res.dl, coef = "SymptomaticYes")

```


```{r, results="asis"}
# results from full analysis
# topTable(res.dl, coef = "SymptomaticNo")

```


```{r, results="asis", eval=FALSE}
# results from full analysis
topTable(res.dl, coef = "SymptomaticNo")

```

```{r, results="asis", eval=FALSE}
gogo()
```
