---
title: "MSC QC cbender_comparisons"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(scran)
  library(SingleR)
  library(scDblFinder)
  library(ggsankey)
  library(scales)
  library(readxl)
  library(openxlsx)
  library(ComplexHeatmap)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(dplyr::desc)
  conflicts_prefer(dplyr::count)
  conflicts_prefer(purrr::reduce)
})
```



# MSC Preprocessing

## setup, read input files and exploratory visualizations and general functions

```{r, setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/project_plotting_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/cell_calling/plotting_fns_n_variables.R")


```

###################START- RAW READING#############

```{r, eval=T}
pf_all_id_decode <- read_csv("data/raw/pf_all_id_decode.csv")
pf_all_id_decode$sample_nm
```


```{r, eval=T}
# pf_solo_mixed_mdata_decode <- read_csv("data/raw/pf_solo_mixed_mdata_decode.csv")
pf_solo_mixed_mdata_decode <- pf_all_id_decode 

```

## Genotype clustering for MSC

initialize variable for the sample names



```{r}

slct_sample_names <- pf_solo_mixed_mdata_decode$sample_nm
# slct_sample_names <-c("MSC68", "MSC14", "MSC49", "MSC55") %>% str_sort(., numeric = T)


# ## poor quality samples
# sample_rm <- c("MSC1", "MSC24", "MSC37", "MSC38", "MSC39", "MSC50_SLO", "MSC33")
sample_rm <- c("MSC1", "MSC24", "MSC36", "MSC37", "MSC38", "MSC39")

sample_mxd <- pf_solo_mixed_mdata_decode %>% filter(mixed_infection2 == "mixed") %>% pull(sample_nm)
# 
## slct_sample_names[-c(sample_rm)]
# slct_sample_names <- slct_sample_names[!slct_sample_names %in% c(sample_rm, sample_mxd)]

all_samples2rm <- c(sample_rm, sample_mxd)
all_samples2rm <- c(all_samples2rm, paste0(all_samples2rm, "C"), paste0(all_samples2rm, "noJC"))
slct_sample_names <- slct_sample_names[!slct_sample_names %in% all_samples2rm]

## get relevant sample names
pf_solo_mixed_mdata_decode_slct <- pf_solo_mixed_mdata_decode %>% filter(sample_nm %in% slct_sample_names)

##Sample ID
sample_name <- str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)
smpl_nms <-str_sort(pf_solo_mixed_mdata_decode_slct$sample_nm)

##Sample IDs
source_irods <- pf_solo_mixed_mdata_decode_slct[,c('sample_nm', 'irods_id')] %>% deframe() %>%.[sample_name]

```

## Samples to be included in the paper
```{r, message=F, warning=F, eval=T}
key_samples <- c("MSC3", "MSC13", "MSC14", "MSC33C", "MSC40C", "MSC45", "MSC48C", "MSC49C", "MSC50_MACS", "MSC50_SLOC", "MSC53", "MSC54", "MSC57", "MSC60C", "MSC66C", "MSC67", "MSC68C", "MSC70C")

don_order <- str_remove(key_samples, "C$")
don_order_m <- str_remove_all(key_samples, "SC|C$")
```



## Seurat object for main integrated asexual object

```{r, message=F, warning=F}
## Generated in 'm2_integration_asx_clustr_gn_explrtn.Rmd'
asex_noM55_jcC_harmony <- readRDS('data/processed/Pf/m2_all/v3_m2_asex_cr_filt_noM55_jcC_norm_harmony_jnd_fine_clustrs3.RDS')

```

```{r, message=F, warning=F}
asex_noM55_jcC_harmony_light <- CreateSeuratObject(counts = LayerData(asex_noM55_jcC_harmony, layer = "counts"), data = LayerData(asex_noM55_jcC_harmony, layer = "data"),  meta.data = asex_noM55_jcC_harmony@meta.data)

asex_noM55_jcC_harmony_light@reductions$umap.harmony_PC_man = asex_noM55_jcC_harmony@reductions$umap.harmony_PC_man

asex_noM55_jcC_harmony_light

```

```{r, message=F, warning=F}
rm(asex_noM55_jcC_harmony)
```

```{r, warning=FALSE, message=FALSE}
seu_cols <- brewer.pal(12, name = "Set3") %>% set_names(c(0:11))
```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=3}
DimPlot(asex_noM55_jcC_harmony_light, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_fine_erings", label.size = 6, label.box = F) +
  scale_color_manual(values = seu_cols) +
  scale_fill_manual(values = seu_cols) +
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 15),
        legend.position = "none",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) 

```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=3}
DimPlot(asex_noM55_jcC_harmony_light, label = T, reduction = "umap.harmony_PC_man", dims = c(1,3), group.by = "harmony_clusters_fine_erings", label.size = 6, label.box = F) +
  scale_color_manual(values = seu_cols) +
  scale_fill_manual(values = seu_cols) +
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 15),
        legend.position = "none",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        # axis.title = element_blank()
        ) 

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
asex_noM55_jcC_harmony_light_figs <- asex_noM55_jcC_harmony_light@meta.data %>%
  mutate(across(donor, ~factor(str_remove_all(., "SC|C$"), levels = don_order_m)),
         # across(donor, ~factor(str_remove(., "C$"), levels = don_order)),
         stage_fld = str_replace_all(stage_fld, c("Mid trophozoite" = "Early trophozoite", ".*schizont" = "Schizont", "Gametocyte \\(developing\\)" = "Early ring")),
         across(stage_fld, ~factor(., levels = stg_refnd_lvls)),
         condition = factor(if_else(orig.ident == "MSC50_SLOC.mrna", "Symptomatic", condition), levels = c("Asymptomatic", "Symptomatic"))) %>%
  select(donor, stage_fld, condition) %>% 
  AddMetaData(asex_noM55_jcC_harmony_light, .)

```

```{r, warning=FALSE, message=FALSE, fig.width=4.5, fig.height=4.4}
DimPlot(asex_noM55_jcC_harmony_light_figs, reduction = "umap.harmony_PC_man", group.by = "stage_fld") + 
  scale_color_manual(values = sp_fld_colors2) +
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 19),
        axis.title = element_text(size = 15),
        legend.position = "bottom",
        axis.text = element_blank(),
        # axis.title = element_blank(),
        axis.ticks = element_blank()
        ) +
  guides(color = guide_legend(nrow = 3, override.aes = list(size = 4)))

```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
asex_noM55_jcC_harmony_light_figs_umap_mdata <- asex_noM55_jcC_harmony_light_figs@meta.data %>% 
  rownames_to_column('bcode') %>% 
  left_join(., data.frame(asex_noM55_jcC_harmony_light_figs@reductions$umap.harmony_PC_man@cell.embeddings) %>% 
              rownames_to_column('bcode')) 

asex_noM55_jcC_harmony_light_figs_umap_mdata
```


```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
p <- asex_noM55_jcC_harmony_light_figs_umap_mdata %>%
  ggplot(aes(x = umapharmonyPCman_1, y = umapharmonyPCman_2, colour = harmony_clusters_fine_erings)) +
  geom_point(size = 0.1) +
  scale_color_manual(values = seu_cols) +
  ggh4x::facet_nested_wrap(vars(condition, donor),  nrow = 1, scales = "free_x")+
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 18),
        legend.position = "bottom",
        axis.text = element_blank(),
        # axis.title = element_blank(),
        axis.ticks = element_blank()
        ) +
  guides(color = guide_legend(title = "Seurat clusters" ,nrow = 1, override.aes = list(size = 4)))

p 
```


```{r, warning=FALSE, message=FALSE, fig.width=2.5, fig.height=4}
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-1222600"), reduction = "umap.harmony_PC_man", order = T, ncol = 1)+
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 22),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.title = element_blank(),
        axis.ticks = element_blank()) 

```


```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
# ap2g, surf1.2, gexp05
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-1222600", "PF3D7-0113600", "PF3D7-1102500"), reduction = "umap.harmony_PC_man", order = T, ncol = 3)

```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
# antisense gdv1 , gdv1
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-0935390", "PF3D7-0935400"), reduction = "umap.harmony_PC_man", order = T, ncol = 2, pt.size = 0.5)
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-0935390", "PF3D7-0935400"), reduction = "umap.harmony_PC_man", order = T, ncol = 2, pt.size = 0.5, dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
# antisense gdv1 , gdv1
gdv1_asgdv1 <- FetchData(object = asex_noM55_jcC_harmony_light, layer = "data", vars = c("PF3D7-0935400", "PF3D7-0935390"))
gdv1_asgdv1_mdata_umap2 <- cbind(asex_noM55_jcC_harmony_light@meta.data, asex_noM55_jcC_harmony_light@reductions$umap@cell.embeddings, gdv1_asgdv1)
gdv1_asgdv1_mdata_umap2 %>% glimpse
```


```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
gdv1_asgdv1 %>% count(`PF3D7-0935400` > 2, `PF3D7-0935390` > 2)
```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
# antisense gdv1 , gdv1
gdv1_asgdv1 %>% head
```


```{r, warning=FALSE, message=FALSE}
# CBP2
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-1301700"), reduction = "umap.harmony_PC_man")
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-1301700"), reduction = "umap.harmony_PC_man", dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE}
# PTP4, PF3D7-0406500
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-0730900"), reduction = "umap.harmony_PC_man", dims = c(1,3))
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-0406500"), reduction = "umap.harmony_PC_man", dims = c(1,3))

```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
# CBP2
VlnPlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-1301700"), group.by = "harmony_clusters_fine_erings", pt.size = 0)
```
```{r}
asex_noM55_jcC_harmony_light@meta.data %>% ggplot(aes(y = nFeature_RNA, x = seurat_clusters)) + geom_violin() + geom_hline(yintercept = 300)
asex_noM55_jcC_harmony_light@meta.data %>% ggplot(aes(y = nCount_RNA, x = seurat_clusters)) + geom_violin() + geom_hline(yintercept = 1000)

```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=3}
FeaturePlot(asex_noM55_jcC_harmony_light, features = c("PF3D7-1222600", "PF3D7-0113600", "PF3D7-1102500"), reduction = "umap.harmony_PC_man", order = T, ncol = 3, dims = c(1,3))

```

## Seurat object for committed rings object

```{r, message=F, warning=F, eval=T}
##Generate in 'm2_integration_asx_clustr_gn_explrtn.Rmd'
# v3_early_asex <- readRDS('data/processed/Pf/m2_all/v3_early_asex3.RDS') %>% .[["noM55_jcC"]]
v3_early_asex_jnd_noM55_jcC <- readRDS("data/processed/Pf/m2_all/v3_early_asex_jnd_noM55_jcC_3.RDS")
```


```{r, message=F, warning=F}
# v3_early_asex_jnd_noM55_jcC[["RNA"]]$counts <- as(object = v3_early_asex_jnd_noM55_jcC[["RNA"]]$counts, Class = "dgCMatrix")
```

```{r, message=F, warning=F}
# v3_early_asex_jnd_noM55_jcC <- CreateSeuratObject(counts = LayerData(v3_early_asex_jnd_noM55_jcC, layer = "counts"), data = LayerData(v3_early_asex_jnd_noM55_jcC, layer = "data"),  meta.data = v3_early_asex_jnd_noM55_jcC@meta.data)
# 
# v3_early_asex_jnd_noM55_jcC@reductions$umap = v3_early_asex_jnd_noM55_jcC@reductions$umap
# v3_early_asex_jnd_noM55_jcC@reductions$harmony = v3_early_asex_jnd_noM55_jcC@reductions$harmony
# 
# v3_early_asex_jnd_noM55_jcC

```

```{r, message=F, warning=F}
# rm(v3_early_asex_jnd_noM55_jcC)
```

```{r, warning=FALSE, message=FALSE}
v3_early_asex_jnd_noM55_jcC$commited_lores_clusters <- ifelse(v3_early_asex_jnd_noM55_jcC$merged_clusters == "committed", "GCR", "EAR")
```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=2.5}
# cmtd_clstr_cols <-  c("0" = "#e05f85", "1" = "#5fc5e0")
# cmtd_clstr_cols <-  c("0" = "#5fc5e0", "1" = "#FC8D62")
cmtd_clstr_cols <-  c("EAR" = "#5fc5e0", "GCR" = "#FC8D62")

```

```{r, warning=FALSE, message=FALSE, fig.width=2.5, fig.height=3.5}
DimPlot(v3_early_asex_jnd_noM55_jcC, label = T, reduction = "umap", group.by = "commited_lores_clusters", label.size = 7, label.box = T) +
  theme_classic() +
  scale_color_manual(values = cmtd_clstr_cols) +
  scale_fill_manual(values = cmtd_clstr_cols) +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 17),
        legend.position = "none",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) 


```

```{r, warning=FALSE, message=FALSE, fig.width=2.5, fig.height=3.5}
FeaturePlot(v3_early_asex_jnd_noM55_jcC, features = "PF3D7-0113600", reduction = "umap", ncol = 1, label = F, label.size = 8, order = T) +
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 24),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) 

```

```{r, warning=FALSE, message=FALSE, fig.width=2.5, fig.height=3.5}
FeaturePlot(v3_early_asex_jnd_noM55_jcC, features = "PF3D7-1102500", reduction = "umap", ncol = 1, label = F, label.size = 8, order = T) +
  theme_classic() +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 24),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) 

```


```{r, message=F, warning=F, eval=T}
##Generate in 'm2_integration_asx_clustr_gn_explrtn.Rmd'
commited_de_mrgd_anot_noM55_jcC <- readRDS("data/processed/Pf/m2_all/commited_de_mrgd_anot_noM55_jcC_3.RDS")

```


```{r, fig.height=6, fig.width=6}
vlcano_mrg <- map(c("committed"), ~{
  EnhancedVolcano(commited_de_mrgd_anot_noM55_jcC %>% filter(cluster == .x),
                lab = commited_de_mrgd_anot_noM55_jcC %>% filter(cluster == .x) %>% pull(Gene),
                # selectLab = .y,
                x = 'avg_log2FC',
                y = 'p_val_adj',
                pCutoff = 0.0001,
                FCcutoff = 2,
                pointSize = 1.5,
                labSize = 4.0,
                title = '',
                subtitle = "",
                titleLabSize = 0,
                subtitleLabSize = 0,
                axisLabSize = 22,
                drawConnectors = TRUE,
                # boxedLabels = TRUE,
                # legendLabels =c( "p_val_adj"),
                gridlines.minor = FALSE
                ) +
    theme(legend.position = "none")
})
vlcano_mrg

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```


```{r, fig.height=6, fig.width=5.5}
vlcano_mrg[[1]] +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16, angle = 90, hjust = 0.5),
        axis.title = element_text(size = 16))
```

```{r, warning=FALSE, message=FALSE}
# v3_early_asex_jnd_noM55_jcC$commited_lores_clusters <- v3_early_asex_jnd_noM55_jcC$seurat_clusters
# v3_early_asex_jnd_noM55_jcC <- FindClusters(v3_early_asex_jnd_noM55_jcC, resolution = 0.26, cluster.name = "seurat_clusters")
```

```{r, warning=FALSE, message=FALSE, fig.width=3, fig.height=2.5}
seu_clst_cols <- c("0" = "#9966CC",
"1" = "#81B29A",
"2" = "#E07A5F",
"3" = "#F2CC8F",
"4" = "#994F00",
"5" = "#005AB5",
"6" = "#3D405B")
```

```{r, warning=FALSE, message=FALSE, fig.width=2.5, fig.height=3.5}
DimPlot(v3_early_asex_jnd_noM55_jcC, label = T, reduction = "umap",  label.size = 7, label.box = T, group.by = "seurat_clusters") +
  theme_classic() +
  scale_color_manual(values = seu_clst_cols) +
  scale_fill_manual(values = seu_clst_cols) +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 17),
        legend.position = "none",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) 


```

```{r, warning=FALSE, message=FALSE}
##Generate in 'm2_integration_asx_clustr_gn_explrtn.Rmd'
commited_de_anot <- readRDS("data/processed/Pf/m2_all/commited_de_anot3.RDS")  %>% .[["noM55_jcC"]]
```

```{r, fig.height=6, fig.width=6}
vlcano <- map(c(0:1), ~{
  EnhancedVolcano(commited_de_anot %>% filter(cluster == .x),
                lab = commited_de_anot %>% filter(cluster == .x) %>% pull(Gene),
                # selectLab = .y,
                x = 'avg_log2FC',
                y = 'p_val_adj',
                pCutoff = 0.05,
                FCcutoff = 2,
                pointSize = 1.5,
                labSize = 4.0,
                title = '',
                subtitle = "",
                titleLabSize = 0,
                subtitleLabSize = 0,
                axisLabSize = 22,
                drawConnectors = TRUE,
                # boxedLabels = TRUE,
                # legendLabels =c( "p_val_adj"),
                gridlines.minor = FALSE
                ) +
    theme(legend.position = "none")
})
vlcano

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```


```{r, fig.height=6, fig.width=5.5}
vlcano[[2]] +
  theme(axis.text.x = element_text(size = 16),
        axis.text.y = element_text(size = 16, angle = 90, hjust = 0.5),
        axis.title = element_text(size = 16))
```


```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
# cmtd_rings <- list(c('2','1'), c('3', '0'), c('1', '4'), c('4', '6'))
# cmtd_rings_devlpmnt <- list(c('0','5'), c('2', '5'))
cmtd_rings_devlpmnt <- list(c('0','4'), c('2', '5'))

```

```{r, warning=FALSE, message=FALSE}
## Generated in 'm2_integration_asx_clustr_gn_explrtn.Rmd'
josling_ap2g_DE_long <- readRDS("data/processed/Pf/m2_all/asex_integration_comtmnt_DE_figs/josling_ap2g_DE_long.RDS")

cmtd_rings_de_gns_vs_all_anot <- readRDS("data/processed/Pf/m2_all/asex_integration_comtmnt_DE_figs/cmtd_rings_de_gns_vs_all_anot.RDS")
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
cmtd_rings_de_gns_vs_all_anot_tbl <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% 
                                        mutate(spec_index = pct.1 - pct.2))

```

```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_gns_vs_all_anot_tbl, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 5))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
clustr_gexp2_low_gns_tbl <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[1]] & p_val_adj < 0.05 & avg_log2FC > 0))
clustr_gexp2_low_gns <- map(clustr_gexp2_low_gns_tbl, ~.x %>% pull(gene))

clustr_gexp2_up_gns_tbl <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[2]] & p_val_adj < 0.05 & avg_log2FC > 0))
clustr_gexp2_up_gns <- map(clustr_gexp2_up_gns_tbl, ~.x %>% pull(gene))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
clustr_gexp2_low_gns <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[1]] & p_val_adj < 0.05 & avg_log2FC > 0) %>% pull(gene))
clustr_gexp2_up_gns <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[2]] & p_val_adj < 0.05 & avg_log2FC > 0) %>% pull(gene))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
clustr_surf_up_gns_tbl <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% 
                                 filter(cluster == .y[[1]] & p_val_adj < 0.05 & avg_log2FC > 0) %>%
                                 mutate(spec_index = pct.1 - pct.2))
clustr_gexp2_up_gns_tbl <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% 
                                  filter(cluster == .y[[2]] & p_val_adj < 0.05 & avg_log2FC > 0) %>%
                                 mutate(spec_index = pct.1 - pct.2))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
clustr_surf_up_gns_stringent <- map2(clustr_surf_up_gns_tbl, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[1]] & p_val_adj < 0.0001 & avg_log2FC > 1 & spec_index > 0.1 & pct.2 < 0.4))
clustr_gexp2_up_gns_stringent <- map2(clustr_gexp2_up_gns_tbl, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[2]] & p_val_adj < 0.0001 & avg_log2FC > 1 & spec_index > 0.1 & pct.2 < 0.4))
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4.5}
timepoint_levels <- str_remove_all(levels(josling_ap2g_DE_long$timepoint), "pi|P")
ap2g_cols <- c("On" = "#FFC20A", "Off" =  "#005AB5")
  
```

```{r, warning=FALSE, message=FALSE, fig.width=4.7, fig.height=4.2}

# imap(commited_uphigh_gns, ~josling_ap2g_DE_long %>%
map(list(clustr_gexp2_low_gns[["noM55_jcC"]], clustr_gexp2_up_gns[["noM55_jcC"]]), ~{
  josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  mutate(across(timepoint, ~factor(str_remove_all(., "pi|P"), levels = timepoint_levels))) %>%
  mutate(across(ap2g_cond, ~str_to_sentence(str_remove(., "ap2g")))) %>%
  dplyr::rename("ap2g" = ap2g_cond) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g)) +
    scale_color_manual(values = ap2g_cols) +
  labs(x = "Timepoint (hpi)", y= "Expression") +
  theme_classic() +
    theme(axis.text.x = element_text(angle=90, hjust = 1),
          legend.position = "bottom",
          text = element_text(size = 20),
        axis.text = element_text(size = 17),
        # axis.ticks = element_blank(),
        # axis.title = element_blank()
        ) +
  guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
  })


```


```{r, warning=FALSE, message=FALSE, fig.width=8, fig.height=4.2}
list(clustr_gexp2_low_gns[["noM55_jcC"]], clustr_gexp2_up_gns[["noM55_jcC"]]) %>% 
  set_names(c("Cluster 0 upregulated genes", "Cluster 4 upregulated genes")) %>% 
  stack() %>%
  rename_with(~str_replace_all(., c("values" = "gene_id", "ind" = "Cluster")), everything()) %>%
  left_join(josling_ap2g_DE_long, by = c("gene_id" ="gene_id")) %>%
  mutate(across(timepoint, ~factor(str_remove_all(., "pi|P"), levels = timepoint_levels))) %>%
  mutate(across(ap2g_cond, ~str_to_sentence(str_remove(., "ap2g")))) %>%
  dplyr::rename("ap2g" = ap2g_cond) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g)) +
    scale_color_manual(values = ap2g_cols) +
  labs(x = "Timepoint (hpi)", y= "Expression") +
  facet_wrap(.~Cluster) +
  theme_classic() +
    theme(axis.text.x = element_text(angle=90, hjust = 1),
          legend.position = "bottom",
          text = element_text(size = 20),
        axis.text = element_text(size = 17),
        # axis.ticks = element_blank(),
        # axis.title = element_blank()
        ) +
  guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))



```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
# Genes upregulated in gexp02 cluster that are in jcO but not in jcC
jcO_gns_missing_in_jcC <- clustr_gexp2_up_gns[[2]][!clustr_gexp2_up_gns[[2]] %in% clustr_gexp2_up_gns[[1]]]
jcO_gns_missing_in_jcC
jcO_jcC_intersect <- intersect(clustr_gexp2_up_gns[[2]], clustr_gexp2_up_gns[[1]])
jcO_jcC_intersect
```


```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
# ETRAMP9, 1354900, Pfg14-744, 1309100
FeaturePlot(v3_early_asex_jnd_noM55_jcC, reduction = "umap",  label = T, features = clustr_gexp2_up_gns[[1]], order = T, ncol = 7)
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 4 & p_val_adj < 0.0001 & avg_log2FC > 0) %>% mutate(spec_index = pct.1 - pct.2) %>% arrange(desc(spec_index)) %>% filter(spec_index > 0.2 & pct.2 <0.1)
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 4 & p_val_adj < 0.0001 & avg_log2FC > 0 & pct.2 <0.01) 
```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=6}
# ETRAMP9, 1354900, Pfg14-744, 1309100
FeaturePlot(v3_early_asex_jnd_noM55_jcC, reduction = "umap",  label = T, features = cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 4 & p_val_adj < 0.05 & avg_log2FC > 0 & pct.2 <0.01) %>% pull(gene), order = T, ncol = 4)
```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=6}
# markers
FeaturePlot(v3_early_asex_jnd_noM55_jcC, reduction = "harmony", dims = c(2,3), label = T, features = cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 4 & p_val_adj < 0.05 & avg_log2FC > 0 & pct.2 <0.01) %>% pull(gene), order = T, ncol = 4)
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 4 & p_val_adj < 0.05 & avg_log2FC > 0) 
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
cmtd_rings_de_gns_vs_all_anot[["noM55_jcO"]] %>% filter(cluster == 4 & p_val_adj < 0.05 & avg_log2FC > 0& pct.2 <0.01) 
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
top_early_gam_markers <- cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 0 & p_val_adj < 0.0001 & avg_log2FC > 0) %>% mutate(spec_index = pct.1 - pct.2) %>% arrange(desc(spec_index)) %>% filter(spec_index > 0.2 & pct.2 <0.1)
top_early_gam_markers
```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=6}
# ETRAMP9, 1354900, Pfg14-744, 1309100
FeaturePlot(v3_early_asex_jnd_noM55_jcC, reduction = "umap",  label = T, features = top_early_gam_markers %>% pull(gene), order = T, ncol = 4)
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=9}
cmtd_rings_de_gns_vs_all_anot[["noM55_jcC"]] %>% filter(cluster == 0 & p_val_adj < 0.0001 & avg_log2FC > 0) %>% mutate(spec_index = pct.1 - pct.2) %>% arrange(desc(spec_index)) %>% filter(spec_index > 0.2)

```

## Assess distribution per donor


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
v3_early_asex_jnd_noM55_jcC <- v3_early_asex_jnd_noM55_jcC@meta.data %>%
  mutate(across(donor, ~factor(str_remove_all(., "SC|C$"), levels = don_order_m)),
         # across(donor, ~factor(str_remove(., "C$"), levels = don_order)),
         stage_fld = str_replace_all(stage_fld, c("Mid trophozoite" = "Early trophozoite", ".*schizont" = "Schizont", "Gametocyte \\(developing\\)" = "Early ring")),
         across(stage_fld, ~factor(., levels = stg_refnd_lvls)),
         condition = factor(if_else(orig.ident == "MSC50_SLOC.mrna", "Symptomatic", condition), levels = c("Asymptomatic", "Symptomatic"))) %>%
  select(donor, stage_fld, condition) %>%
  AddMetaData(v3_early_asex_jnd_noM55_jcC, .)

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

v3_early_asex_jnd_noM55_jcC@meta.data %>% count(seurat_clusters, donor)

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

v3_early_asex_jnd_noM55_jcC@meta.data %>% count(seurat_clusters, donor, Strain_DN) %>% filter(seurat_clusters %in% cmtd_rings_devlpmnt[[1]]) %>% filter(n >10)

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

v3_early_asex_jnd_noM55_jcC@meta.data %>% count(commited_lores_clusters, donor, Strain_DN) %>% filter(n >10)

```


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=4}

don_clst_barplot <- function(dset, count_var, facets, label_var, y_lab = 'Counts', manual_cols = strain_cols_n, x_lab = "Strain") {
  ggplot(dset, aes(x = commited_lores_clusters, y = !!rlang::sym(count_var), fill = commited_lores_clusters)) +
  geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=!!rlang::sym(label_var)), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90) +
  scale_fill_manual(values = manual_cols) +
  ggh4x::facet_nested_wrap(vars(!!!syms(facets)),  nrow = 1, scales = "free_x")+
  scale_y_continuous(expand = expansion(mult = c(0.01, 1))) +
  scale_x_discrete(expand = expansion(add = c(1, 1))) +
  theme_classic() +
  labs( y = y_lab, x = x_lab)+
  theme(
    axis.text.y  = element_text(size = 14),
    axis.title.y  = element_text(size = 17, face = "bold"),
    axis.text.x  = element_text(size = 14, face = "bold", angle = 0),
    # axis.title.x = element_blank(),
    axis.title.x = element_text(size = 17, face = "bold"),
    legend.text = element_text(size = 17),
    legend.title = element_text(size = 17, face = "bold"),
    legend.position="bottom",
    strip.text =element_text(size=16,face="bold", colour = 'black'),
    panel.border=element_rect(colour="black",fill='transparent')
  ) + 
  guides(fill = guide_legend(title.position = 'left', nrow = 1, title = x_lab))
}
```

```{r, warning=FALSE, message=FALSE, fig.width=6, fig.height=4.2}

tbl_s <- v3_early_asex_jnd_noM55_jcC@meta.data %>% count(condition, commited_lores_clusters, donor) %>% 
  filter(n >10) %>%
  group_by(donor) %>%
  mutate(prop = round(n/sum(n), 2),
         prop_n = paste0(prop, " (", n, ")")) 

don_clst_barplot(tbl_s, count_var = "prop", label_var= "prop_n", facets =c("condition", "donor"), y_lab = "Proportion", x_lab = "Cluster", manual_cols = cmtd_clstr_cols)

```


```{r, warning=FALSE, message=FALSE, fig.width=4.6, fig.height=3.8}

don_clst_barplot(tbl_s, count_var = "prop", label_var= "prop_n", facets =c("condition", "donor"), y_lab = "Proportion", x_lab = "Cluster", manual_cols = cmtd_clstr_cols) +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=4}

don_clst_strn_barplot <- function(dset, count_var, facets, label_var, y_lab = 'Counts' ) {
  ggplot(dset, aes(x = Strain_DN, y = !!rlang::sym(count_var), fill = Strain_DN)) +
  geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=!!rlang::sym(label_var)), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90) +
  scale_fill_manual(values = strain_cols_n) +
  ggh4x::facet_nested_wrap(vars(!!!syms(facets)),  nrow = 1, scales = "free_x")+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.8))) +
  scale_x_discrete(expand = expansion(add = c(1, 1))) +
  theme_classic() +
  labs( y = y_lab)+
  theme(
    axis.text.y  = element_text(size = 14),
    axis.title.y  = element_text(size = 17, face = "bold"),
    axis.text.x  = element_text(size = 14, face = "bold", angle = 90),
    # axis.title.x = element_blank(),
    axis.title.x = element_text(size = 17, face = "bold"),
    legend.text = element_text(size = 17),
    legend.title = element_text(size = 17, face = "bold"),
    legend.position="bottom",
    strip.text =element_text(size=16,face="bold", colour = 'black'),
    panel.border=element_rect(colour="black",fill='transparent')
  ) + 
  guides(fill = guide_legend(title.position = 'left', nrow = 1, name = 'Strain (K8)'))
}
```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}

v3_early_asex_jnd_noM55_jcC@meta.data %>% count(condition, commited_lores_clusters, donor, Strain_DN) %>% filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  don_clst_strn_barplot(., count_var = "n", label_var= "n", facets =c("condition", "donor", "commited_lores_clusters"))
```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}

v3_early_asex_jnd_noM55_jcC@meta.data %>% count(condition, commited_lores_clusters, donor, Strain_DN) %>% filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  don_clst_strn_barplot(., count_var = "n", label_var= "n", facets =c("condition", "donor", "commited_lores_clusters"))
```


```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=4}

tbl_props <- v3_early_asex_jnd_noM55_jcC@meta.data %>% count(condition, commited_lores_clusters, donor, Strain_DN) %>% filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  group_by(commited_lores_clusters, donor) %>%
  mutate(prop = round(n/sum(n), 2),
         prop_n = paste0(prop, " (", n, ")")) 

tbl_props
```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=4.3}
tbl_props %>% 
  don_clst_strn_barplot(., count_var = "prop", label_var= "prop_n", facets =c("condition", "donor", "commited_lores_clusters"), y_lab = "Proportion")
```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=4.3}
tbl_props %>% 
  don_clst_strn_barplot(., count_var = "prop", label_var= "prop_n", facets =c("donor", "commited_lores_clusters"), y_lab = "Proportion")
```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=4}

v3_early_asex_jnd_noM55_jcC@meta.data %>%
  count(seurat_clusters, donor, Strain_DN) %>% 
  filter(seurat_clusters %in% cmtd_rings_devlpmnt[[1]]) %>% 
  filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  don_clst_strn_barplot(., count_var = "n", label_var = "n", facets =c("donor", "seurat_clusters"))

```
```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=4}

v3_early_asex_jnd_noM55_jcC@meta.data %>%
  count(seurat_clusters, donor, Strain_DN) %>% 
  filter(seurat_clusters %in% cmtd_rings_devlpmnt[[1]]) %>% 
  filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  don_clst_strn_barplot(., count_var = "n", label_var = "n", facets =c("donor", "seurat_clusters"))

```


```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=4}
v3_early_asex_jnd_noM55_jcC@meta.data %>% 
  count(condition, seurat_clusters, donor, Strain_DN) %>% 
  filter(seurat_clusters %in% cmtd_rings_devlpmnt[[1]]) %>% 
  filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  group_by(seurat_clusters, donor) %>%
  mutate(prop = round(n/sum(n), 2),
         prop_n = paste0(prop, " (", n, ")")) %>% 
  don_clst_strn_barplot(., count_var = "prop", label_var = "prop_n", facets =c("condition", "donor", "seurat_clusters"), y_lab = "Proportion")

```


```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=4}
v3_early_asex_jnd_noM55_jcC@meta.data %>% 
  count(condition, seurat_clusters, donor, Strain_DN) %>% 
  filter(seurat_clusters %in% cmtd_rings_devlpmnt[[1]]) %>% 
  filter(n >10) %>%
  mutate(across(Strain_DN, ~str_remove(., "ative"))) %>% 
  group_by(seurat_clusters, donor) %>%
  mutate(prop = round(n/sum(n), 2),
         prop_n = paste0(prop, " (", n, ")")) %>% 
  don_clst_strn_barplot(., count_var = "prop", label_var = "prop_n", facets =c("condition", "donor", "seurat_clusters"), y_lab = "Proportion")

```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
# ETRAMP9, 1354900, Pfg14-744, 1309100
FeaturePlot(v3_early_asex_jnd_noM55_jcC, reduction = "umap",  label = T, features = jcO_gns_missing_in_jcC, order = T, ncol = 4)
```


```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3}
# ap2-g pfs16 gene_count
DimPlot(v3_early_asex_jnd_noM55_jcC, split.by =  "donor", reduction = "umap", pt.size = 0.01)

```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=3}
# get gene from seurat object and merge to metadata
asx_gns_df <- FetchData(object = v3_early_asex_jnd_noM55_jcC, layer = "data", vars = c("PF3D7-1222600", "PF3D7-1102500", "PF3D7-0113600"))
colnames(asx_gns_df) <- str_replace(colnames(asx_gns_df), "-", "_")
asx_gns_mdata_umap <- cbind(v3_early_asex_jnd_noM55_jcC@meta.data, v3_early_asex_jnd_noM55_jcC@reductions$umap@cell.embeddings, asx_gns_df)
# glimpse(asx_gns_mdata_umap)
```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
asx_gns_mdata_umap %>%
  ggplot(aes(x = umap_1, y = umap_2, colour = commited_lores_clusters)) +
  geom_point(size = 0.2) +
  scale_color_manual(values = cmtd_clstr_cols) +
  facet_nested_wrap(vars(condition, donor), nrow = 1) +
  theme_classic() +
  theme(strip.text = element_text(size = 13, face = "bold"),
        text = element_text(size = 13),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank()) +
  guides(colour = guide_legend(title = "Cluster", nrow = 1, override.aes = list(size = 4)))

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=11, fig.height=3.2}
umap_feature_plot_fn <- function(dset = asx_gns_mdata_umap, gene = "PF3D7_1222600", gn_nm = "ap2g", clstr = "commited_lores_clusters", dim1 = "umap_1", dim2 = "umap_2", grp_var = c("condition", "donor"), row_num = 1, pt_size = 1) {
  
cluster_centers <- dset %>%
  group_by(!!rlang::sym(clstr)) %>%
  dplyr::summarize(umap_cntr_1 = median(!!rlang::sym(dim1)), umap_cntr_2 = median(!!rlang::sym(dim2)))

dset %>%
  group_by(!!!rlang::syms(grp_var)) %>%
  arrange(!!rlang::sym(gene)) %>%
  ggplot(aes(x = !!rlang::sym(dim1), y = !!rlang::sym(dim2), colour = !!rlang::sym(gene))) +
  geom_point(size = pt_size) +
  geom_text(data = cluster_centers,      # Add labels at centers
            aes(x = umap_cntr_1, y = umap_cntr_2, label = !!rlang::sym(clstr)), 
            color = "black", 
            fontface = "bold") +
  scale_color_distiller(direction = 1) +
  facet_nested_wrap(vars(!!!rlang::syms(grp_var)), nrow = row_num) +
  theme_classic() +
  theme(strip.text = element_text(size = 13, face = "bold"),
        text = element_text(size = 13),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank()) +
  guides(colour = guide_colorbar(title = gn_nm, direction = "horizontal"))
}
```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
umap_feature_plot_fn(dset = asx_gns_mdata_umap, gene = "PF3D7_1222600", gn_nm = "ap2g") 
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
umap_feature_plot_fn(dset = asx_gns_mdata_umap, gene = "PF3D7_0113600", gn_nm = "surf1.2") 
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
umap_feature_plot_fn(dset = asx_gns_mdata_umap, gene = "PF3D7_1102500", gn_nm = "gexp02") 
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=11, fig.height=3.2}
# umap_feature_plot_fn(dset = asx_gns_mdata_umap, gene = "commited_lores_clusters", gn_nm = "Seurat clusters") 
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=11, fig.height=3.2}
# 
# asx_gns_mdata %>%
#   group_by(donor) %>%
#   mutate(across(donor, ~factor(str_remove(., "C$"), levels = don_order)),
#          stage_fld = str_replace_all(stage_fld, c("Mid trophozoite" = "Early trophozoite", ".*schizont" = "Schizont", "Gametocyte \\(developing\\)" = "Early ring")),
#          across(stage_fld, ~factor(., levels = stg_refnd_lvls)),
#          batch = ifelse(str_detect(orig.ident, "C.mrna"), "JC+nonJC", "JC")) %>% 
#   ggplot(aes(x = UMAP_1, y = UMAP_2, colour = batch)) +
#   scale_color_manual(values = jc_batch_cols) +
#   geom_point(size = 0.001) +
#   facet_nested_wrap(vars(donor), ncol = 8, scales = "free") +
#   theme_classic() +
#   theme(strip.text = element_text(size = 13, face = "bold"),
#         text = element_text(size = 13),
#         legend.position = "bottom",
#         axis.text = element_blank(),
#         axis.ticks = element_blank()) +
#   guides(colour = guide_legend(title = "Stage", nrow = 1, override.aes = list(size = 4)))
#   
```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=3}
# merge with genes with metadata

```
##########START# COMPARISON WITH AP2G ACTIVATED SEXUAL RING GENES from lora batle and poran ##############

```{r, eval=T}
# josling_ap2g_DE = readxl::read_xlsx("../published_dsets/painter_29985403_48hr_dynamics/Expression_Peak_Timing_48h_IDC.xlsx", sheet = "Estimated Total Abundance", .name_repair = "universal") %>% mutate(across(Gene.ID, ~str_replace(., "_", "-"))) %>% rename_all(., ~str_remove(., "\\.\\.")) %>% rename_with(.fn = ~str_remove(paste0("hpi_", .x), ".hpi"), .cols = contains("hpi")) %>% rename("peak_hpi" = `hpi_Peak.Tim.`)
```

```{r, eval=T}
# # josling_ap2g_DE = readxl::read_xlsx("../published_dsets/josling_ap2g_chipseq_2020/apg_activation_DE_gns_across_timepoints.xlsx", skip = 1, .name_repair = "universal", col_names = T)
# josling_ap2g_DE = read_csv("../published_dsets/josling_ap2g_chipseq_2020/apg_activation_DE_gns_across_timepoints.csv", skip = 1, col_names = T, name_repair = "universal")
# 
# josling_ap2g_DE <- josling_ap2g_DE %>% 
#   rename("gene_id" = `Gene.ID`) %>% 
#   mutate(across(gene_id, ~str_replace(., "_", "-"))) %>% 
#   # rename_all(., ~str_remove_all(., "\\.+\\.|TP\\.[0-9]+\\.+|TP[0-9]\\.+"))%>% 
#   rename_all(., ~str_replace_all(str_remove_all(.,"\\.\\.\\.\\.|"), c("\\.\\.\\."="_", "\\.\\."="_", "TP\\."="TP", "\\."="")))%>% 
#   # rename_with(.fn = ~str_remove(paste0("ap2gON_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi|",c(3:13)))) %>%
#   rename_with(.fn = ~str_remove(paste0("ap2gON_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi",c(3:13), collapse = "|"))) %>%
#   rename_with(.fn = ~str_remove(paste0("ap2gOFF_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi",c(14:24), collapse = "|"))) %>%
#   left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene_id" = "gene_id")) # %>% 
#  # glimpse
# 
# glimpse(josling_ap2g_DE)
```




```{r, warning=FALSE, message=FALSE}
## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in ap2g positive cells

## poran et al commitment
poran_commitment_ap2g <- c("PF3D7_1036000", "PF3D7_1335200", "PF3D7_0830800", "PF3D7_0402200", "PF3D7_1238500", "PF3D7_0424400", "PF3D7_1335000", "PF3D7_0617200", "PF3D7_0624600", "PF3D7_1002200", "PF3D7_0801900", "PF3D7_1139300", "PF3D7_1222400", "PF3D7_0201500", "PF3D7_1104200", "PF3D7_1472200", "PF3D7_1477800", "PF3D7_0406500", "PF3D7_1252600") %>% str_replace(., "_", "-")

```



```{r, fig.height=3.8, fig.width=3.8}
# M48, M49, M60
ggvenn(list("Poran_Up" = poran_commitment_ap2g,
            # "Up_early" = clustr_gexp2_low_gns[["noM55_jcC"]],
            # "Up_later" = clustr_gexp2_up_gns[["noM55_jcC"]]),
            "Up_early" = clustr_surf_up_gns_stringent[["noM55_jcC"]] %>% pull(gene),
            "Up_later" = clustr_gexp2_up_gns_stringent[["noM55_jcC"]] %>% pull(gene)),
       show_percentage = F,
       text_size = 6)


```


```{r, eval=T}
# josling_ap2g_DE = readxl::read_xlsx("../published_dsets/josling_ap2g_chipseq_2020/apg_activation_DE_gns_across_timepoints.xlsx", skip = 1, .name_repair = "universal", col_names = T)
early_lbatle_ap2g = read_xlsx("../published_dsets/lora_batille_2020/de_genes_file_s1.xlsx", skip = 3, sheet = "40-45 hpi log2(FC)>1", .name_repair = "universal") %>% 
  rename_all(., ~str_remove_all(.,"^\\.+"))

all_lbatle_ap2g = read_xlsx("../published_dsets/lora_batille_2020/de_genes_file_s1.xlsx", skip = 3, sheet = "log2(FC)>3", .name_repair = "universal") %>% 
  rename_all(., ~str_remove_all(.,"^\\.+"))


# josling_ap2g_DE <- josling_ap2g_DE %>% 
#   rename("gene_id" = `Gene.ID`) %>% 
#   mutate(across(gene_id, ~str_replace(., "_", "-"))) %>% 
#   # rename_all(., ~str_remove_all(., "\\.+\\.|TP\\.[0-9]+\\.+|TP[0-9]\\.+"))%>% 
#   rename_all(., ~str_replace_all(str_remove_all(.,"\\.\\.\\.\\.|"), c("\\.\\.\\."="_", "\\.\\."="_", "TP\\."="TP", "\\."="")))%>% 
#   # rename_with(.fn = ~str_remove(paste0("ap2gON_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi|",c(3:13)))) %>%
#   rename_with(.fn = ~str_remove(paste0("ap2gON_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi",c(3:13), collapse = "|"))) %>%
#   rename_with(.fn = ~str_remove(paste0("ap2gOFF_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi",c(14:24), collapse = "|"))) %>%
#   left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene_id" = "gene_id")) # %>% 
#  # glimpse
# 
# glimpse(josling_ap2g_DE)
```


```{r, fig.height=3.8, fig.width=3.8}
ggvenn(list("lb_early" = early_lbatle_ap2g %>% filter(if_any(c("40.45h"), ~.x>=1)) %>% pull(Gene_id) %>% str_replace(., "_", "-"),
            # "Up_early" = clustr_gexp2_low_gns[["noM55_jcC"]],
            # "Up_later" = clustr_gexp2_up_gns[["noM55_jcC"]]),
            "Up_early" = clustr_surf_up_gns_stringent[["noM55_jcC"]] %>% pull(gene),
            "Up_later" = clustr_gexp2_up_gns_stringent[["noM55_jcC"]] %>% pull(gene)),
       show_percentage = F,
       text_size = 6)


```


```{r, fig.height=3.8, fig.width=3.8}
# M48, M49, M60
ggvenn(list("lb_all" = all_lbatle_ap2g %>% filter(if_any(c("40.45h", "5.10h", "15.20h"), ~.x >= 3)) %>% pull(Gene_id) %>% str_replace(., "_", "-"),
            # "Up_early" = clustr_gexp2_low_gns[["noM55_jcC"]],
            # "Up_later" = clustr_gexp2_up_gns[["noM55_jcC"]]),
            "Up_early" = clustr_surf_up_gns_stringent[["noM55_jcC"]] %>% pull(gene),
            "Up_later" = clustr_gexp2_up_gns_stringent[["noM55_jcC"]] %>% pull(gene)),
       show_percentage = F,
       text_size = 6)


```


```{r, fig.height=3.8, fig.width=3.8}
ggvenn(list("lb_later" = all_lbatle_ap2g %>% filter(if_any(c("5.10h", "15.20h"), ~.x >= 1)) %>% pull(Gene_id) %>% str_replace(., "_", "-"),
            # "Up_early" = clustr_gexp2_low_gns[["noM55_jcC"]],
            # "Up_later" = clustr_gexp2_up_gns[["noM55_jcC"]]),
            "Up_early" = clustr_surf_up_gns_stringent[["noM55_jcC"]] %>% pull(gene),
            "Up_later" = clustr_gexp2_up_gns_stringent[["noM55_jcC"]] %>% pull(gene)),
       show_percentage = F,
       text_size = 6)


```



```{r, fig.height=3.8, fig.width=3.8}
praj_fld_gam_rings = read_xlsx("../published_dsets/prajapati_bulkRNA_natural_infection_commitment_2020/bulkRNA_field_commitment.xlsx", skip = 2, sheet = "Fig1", .name_repair = "universal") %>% 
  .[,7:dim(.)[2]] %>%
  filter(if_any(everything(), ~ !is.na(.x))) %>%
  rename_all(., ~str_remove_all(.,"^\\.+")) %>%
  rename_all(., ~str_replace(.,"\\.+", "_")) %>%
  rename("gene" = `7`) %>%
  separate_wider_delim(., gene, " (", names = c("old_gene_id", "gene_name")) %>%
  mutate(across(gene_name, ~str_remove(., "\\)")))
```


```{r, fig.height=3.8, fig.width=3.8}
praj_fld_gam_rings 
```



```{r, fig.height=3.8, fig.width=3.8}
# plasmodb_old_ids <- Pf3D7_genes_plasmoDB %>% 
#   mutate(prev_id = `Previous ID(s)`,
#          old_pf_id = str_extract(prev_id, "PF.*"),
#          old_mal_id = str_extract(prev_id, "MAL.*?(?=;)")) %>%
#   select(old_gene_id, Gene, old_pf_id, old_mal_id, prev_id)
```


```{r, fig.height=3.8, fig.width=3.8}
# plasmodb_old_ids %>% glimpse
```

```{r, fig.height=3.8, fig.width=3.8}
## Read the gene aliases table downloaded from plasmoDB https://plasmodb.org/common/downloads/Current_Release/Pfalciparum3D7/txt/PlasmoDB-68_Pfalciparum3D7_GeneAliases.txt
max_cols <- max(count.fields("../../Pf3D7_genomes_gtfs_indexs/PlasmoDB-68_Pfalciparum3D7_GeneAliases.txt", sep = "\t"))

plasmodb_gn_aliases <- read.table("../../Pf3D7_genomes_gtfs_indexs/PlasmoDB-68_Pfalciparum3D7_GeneAliases.txt", sep = "\t", header = FALSE, fill = TRUE, comment.char = "", stringsAsFactors = FALSE, col.names = paste0("V", seq_len(max_cols)))   

## Convert to long to have all the aliases in one column since some genes have more than one alias
plasmodb_gn_aliases_long <- plasmodb_gn_aliases %>% rename("gene_id" = V1) %>% pivot_longer(., cols = -c(gene_id),  names_to = "version", values_to = "old_gene_id", values_drop_na = T) %>% filter(old_gene_id != "")

## Capitalize the aliases with lower case to create a secon instance of each since the some of the ids mismatch because of case
plasmodb_gn_aliases_all <- rbind(plasmodb_gn_aliases_long,
                                 plasmodb_gn_aliases_long %>% mutate(across(old_gene_id, str_to_upper))) %>%
  distinct(old_gene_id, .keep_all = T)
```

```{r, fig.height=3.8, fig.width=3.8}
plasmodb_gn_aliases_all %>% filter(gene_id =="PF3D7_0406200")

```

```{r, fig.height=3.8, fig.width=3.8}
plasmodb_gn_aliases_all %>% filter(str_detect(gene_id, "MAL"))

```

```{r, fig.height=3.8, fig.width=3.8}
praj_fld_gam_rings_cln <- praj_fld_gam_rings %>% 
  left_join(., plasmodb_gn_aliases_all, by = c("old_gene_id" = "old_gene_id")) %>% 
  mutate(across(gene_id, ~str_replace(., "_", "-")),
         class = case_when(old_gene_id %in% c("PF14_0350", "PFE0065W", "PFB0100C") ~ "Non-commited",
                           old_gene_id %in% c("PFD0310W", "PF13_0011", "PF14_0748", "PFL2550W", "PF14_0745", "PF14_0735", "PF14_0744", "PF10_0303") ~ "Previous gam",
                           T ~ "Committed high"
                           ))

```

```{r, fig.height=3.8, fig.width=3.8}
ggvenn(list("praj_fld" = praj_fld_gam_rings_cln %>% filter(class == "Committed high") %>% pull(gene_id),
            # "Up_early" = clustr_gexp2_low_gns[["noM55_jcC"]],
            # "Up_later" = clustr_gexp2_up_gns[["noM55_jcC"]]),
            "Up_early" = clustr_surf_up_gns_stringent[["noM55_jcC"]] %>% pull(gene),
            "Up_later" = clustr_gexp2_up_gns_stringent[["noM55_jcC"]] %>% pull(gene)),
       show_percentage = F,
       text_size = 6)


```

```{r, fig.height=3.8, fig.width=3.8}
# M48, M49, M60
ggvenn(list("Up_early" = clustr_gexp2_low_gns[["noM55_jcC"]],
            "Up_later" = clustr_gexp2_up_gns[["noM55_jcC"]],
            "praj_fld" = praj_fld_gam_rings_cln %>% filter(class == "Committed high") %>% pull(gene_id)),
       show_percentage = F,
       text_size = 6)


```

```{r, warning=FALSE, message=FALSE}
##Generate in 'lasse_stages_marker_genes_explrtn.Rmd'
lasse_clst_mrkrs_rings_anot <- readRDS("data/processed/Pf/m2_all/plot_tables_data/m2_lasse_integration_figs/lasse_clst_mrkrs_rings_anot.RDS")
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
lasse_cmtd_up_gns_tbl <- lasse_clst_mrkrs_rings_anot %>% 
  filter(cluster == "7" & p_val_adj < 0.05 & avg_log2FC > 0) %>%
  mutate(spec_index = pct.1 - pct.2)

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
lasse_cmtd_up_gns_tbl_stringent <- lasse_cmtd_up_gns_tbl %>% filter(p_val_adj < 0.0001 & avg_log2FC > 1 & spec_index > 0.1 & pct.2 < 0.4)
lasse_cmtd_up_gns_tbl_stringent
```

##########START Josling genes prep - May need to be moved ############### 
```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
df_pairs <- josling_ap2g_DE_long %>%
  pivot_longer(cols = matches("^(FDR|Foldchange).*"), names_to = c("metric", "pair"), names_pattern = "(FDR|Foldchange)_(.*)", values_to = "value") %>%
  pivot_wider(names_from = metric, values_from = value) %>%
  mutate(signf_DE = if_else(FDR <= 0.1 & Foldchange >= 1.5,"DE", "noDE")) %>%
  pivot_wider(id_cols = -c(Foldchange, FDR), names_from = pair, values_from = signf_DE, names_prefix = "signf_DE_"  )


josling_ap2g_DE_long_signfDE <- josling_ap2g_DE_long %>%
  left_join(., df_pairs, by = base::setdiff(names(df), names(df_pairs)))
```

```{r, warning=FALSE, message=FALSE, fig.width=8.2, fig.height=5}
## !!!!! NOTE - Josling et al mention 42 DE gene based on FDR <= 0.1 & Foldchange >= 1.5 while I find 54 genes based on the same metrics. 
## I therefore get the names of the 42 genes in their publication https://pmc.ncbi.nlm.nih.gov/articles/PMC7083873/#sec23 figure 3c 
## i do this by taking a screenshot of the names and asking chatgpt to put in a vector resulting in the below josling_42_gns vector
## I then look for these genes in my 54 genes to see if included and whats special about the missing ones.

josling_42_gns <- c(
  "PF3D7-1477400", "Pfg14.744", "PF3D7-1476900", "Pfg14.748", "Pfg27/25",
  "ETRAMP10.3", "ncRNA (Pfa_npcR_6753a)", "Pfs16", "ETRAMP4", "ncRNA (PF10TR007)",
  "GEXP05", "REX4", "PF3D7-1476600", "GEXP02", "ELO3", "PF3D7-0422000",
  "PF3D7-0937200", "GEXP14", "PF3D7-1427900", "PF3D7-1476500", "PF3D7-0315600",
  "PF3D7-1410100", "ETRAMP9", "PF3D7-0109300", "NUP116/NSP116", "VAP1",
  "PF3D7-1350900", "PF3D7-1467600", "PF3D7-1252500", "HDA1", "PF3D7-0206300",
  "LSD2", "AP2-G", "MSRP1", "PF3D7-1467700", "PF3D7-1473200", "PF3D7-1038700",
  "PF3D7-1134600", "PF3D7-0113400", "PF3D7-1038600", "PF3D7-1466200", "ACS9"
)
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_ap2g_DE_long_signfDE %>% filter(signf_DE_1_11 == "DE") %>% pull(Gene) %>% unique()
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_jr35_gn_ids <-josling_ap2g_DE_long_signfDE %>% filter(if_any(starts_with("signf_DE"), ~.x == "DE")) %>% pull(gene_id) %>% unique()
josling_jr35_gn_ids
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_jr35_gns <- josling_ap2g_DE_long_signfDE %>% filter(if_any(starts_with("signf_DE"), ~.x == "DE")) %>% pull(Gene) %>% unique()
josling_jr35_gns
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_42_gns_hmnz <- str_replace_all(josling_42_gns, c("Pfg14." = "Pfg14-", "Pfg27/25" = "G27/25", "ncRNA \\(" ="", "\\)" ="", "NUP116/NSP116" = "NUP116", "Pfa_" = "Pfa-", "GEXP05"= "GEXP5"))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_jr35_gns_uniq <- unique(c(josling_jr35_gns, josling_jr35_gn_ids))
josling_42_gns_hmnz[!josling_42_gns_hmnz %in% josling_jr35_gns_uniq]
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_manQC_common_gns <- josling_42_gns_hmnz[josling_42_gns_hmnz %in% josling_jr35_gns_uniq]
josling_manQC_common_gns

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_manQC_common_gn_ids <- Pf3D7_genes_plasmoDB %>% filter(gene_id %in% josling_manQC_common_gns | Gene %in% josling_manQC_common_gns) %>% pull(gene_id) %>% unique()
josling_manQC_common_gn_ids
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
gns_missing_in_pfdb <- josling_manQC_common_gns[!josling_manQC_common_gns %in% c(Pf3D7_genes_plasmoDB$Gene,Pf3D7_genes_plasmoDB$gene_id)]
all_josling_manQC_common_gn_ids <- c(josling_manQC_common_gn_ids, gns_missing_in_pfdb)
```

##########END Josling genes prep - May need to be moved ############### 



```{r, fig.height=3.8, fig.width=3.8}
cmtd_gns_overlap <- reduce(
  list(
  data.frame("gene_id"= clustr_surf_up_gns_stringent[["noM55_jcC"]] %>% pull(gene), "Up_early" = "yes"),
  data.frame("Up_later" = "yes", "gene_id" = clustr_gexp2_up_gns_stringent[["noM55_jcC"]] %>% pull(gene)),
  data.frame("LP_cmtd" = "yes", "gene_id" = lasse_cmtd_up_gns_tbl_stringent %>% pull(gene)),
  data.frame("praj_fld" = "yes", "gene_id" = praj_fld_gam_rings_cln %>% filter(class == "Committed high") %>% pull(gene_id)),
  data.frame("Jos35_Up" = "yes", "gene_id" = all_josling_manQC_common_gn_ids), 
  data.frame("lb_early" = "yes", "gene_id" = early_lbatle_ap2g %>% filter(if_any(c("40.45h"), ~.x>=1)) %>% pull(Gene_id) %>% str_replace(., "_", "-")),
  data.frame("lb_later" = "yes", "gene_id" = all_lbatle_ap2g %>% filter(if_any(c("5.10h", "15.20h"), ~.x >= 1)) %>% pull(Gene_id) %>% str_replace(., "_", "-")),
  # data.frame("lb_all" = "yes", "gene_id" = all_lbatle_ap2g %>% filter(if_any(c("40.45h", "5.10h", "15.20h"), ~.x >= 3)) %>% pull(Gene_id) %>% str_replace(., "_", "-")),
  data.frame("Poran_Up" = "yes", "gene_id" = poran_commitment_ap2g)
  ), full_join
  ) %>%
  left_join(Pf3D7_genes_plasmoDB[, c("gene_id", "Gene")], by = "gene_id") %>%
  mutate(Gene = coalesce(Gene, gene_id))

```


```{r, fig.height=3.8, fig.width=3.8}
cmtd_gns_overlap 
```


```{r, fig.height=3.8, fig.width=3.8}
cmtd_gns_overlap_bin <- cmtd_gns_overlap %>%
  mutate(across(-c(gene_id, Gene), ~ ifelse(is.na(.x), 0, 1)))  %>%
  select(-c(gene_id)) %>%
  # distinct(Gene, .keep_all = T) %>%
  column_to_rownames("Gene")


```

```{r, fig.height=14, fig.width=4.3}
# Plot heatmap
Heatmap(
  as.matrix(cmtd_gns_overlap_bin),
  name = "Presence",
  col = c("0" = "#9dd7ef", "1" = "brown"),
  show_row_names = T,
  show_column_names = TRUE,
  cluster_rows = F,
  cluster_columns = F
)
```

```{r, fig.height=12, fig.width=9}
# Plot heatmap
Heatmap(
  as.matrix(cmtd_gns_overlap_bin),
  name = "Presence",
  col = c("0" = "lightgreen", "1" = "brown"),
  show_row_names = T,
  show_column_names = TRUE,
  cluster_rows = TRUE,
  cluster_columns = F
)
```




```{r, fig.height=3.8, fig.width=3.8}
clustr_surf_up_gns_stringent[["noM55_jcC"]]
clustr_gexp2_up_gns_stringent[["noM55_jcC"]] 
```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=3}
# get genes that are in our dataset either in the early or late commited set and not in any of the other datasets
our_uniq_gnset <- cmtd_gns_overlap %>% 
  # filter(Up_early == "yes" | Up_later == "yes") %>% 
  filter((Up_early == "yes" | Up_later == "yes") & (LP_cmtd == "yes" | praj_fld == "yes")) %>%
  # filter(if_all(c('praj_fld', 'lb_early', 'lb_later', 'Poran_Up'), ~is.na(.x))) %>% 
  filter(if_all(c('lb_early', 'lb_later', 'Poran_Up'), ~is.na(.x))) %>% 
  select(gene_id, Gene)

our_uniq_gnset
```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=3}
asx_gns_df2 <- FetchData(object = v3_early_asex_jnd_noM55_jcC, layer = "data", vars = our_uniq_gnset$gene_id)
# colnames(asx_gns_df2) <- str_replace(colnames(asx_gns_df2), "-", "_")
asx_gns_mdata_umap2 <- cbind(v3_early_asex_jnd_noM55_jcC@meta.data, v3_early_asex_jnd_noM55_jcC@reductions$umap@cell.embeddings, asx_gns_df2)

```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
map2(our_uniq_gnset$gene_id, our_uniq_gnset$Gene, ~umap_feature_plot_fn(dset = asx_gns_mdata_umap2, gene = .x, gn_nm = .y, pt_size = 0.1)) 
```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=3.2}
asx_gns_mdata_umap2_gn_exprsn <- asx_gns_mdata_umap2 %>% pivot_longer(cols = starts_with("PF3D7"), names_to = "gene", values_to = "exprsn") %>% left_join(., Pf3D7_genes_plasmoDB[, c("gene_id", "Gene")], by =  c("gene" = "gene_id"))

```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=5.5, fig.height=3.5}
umap_feature_plot_fn(dset = asx_gns_mdata_umap2_gn_exprsn, gene = "exprsn", gn_nm = "Gene", grp_var = "Gene", row_num = 2, pt_size = 0.05) 
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=5.5, fig.height=6}
asx_gns_mdata_umap2_gn_exprsn %>%
  # group_by(gene) %>%
  ggplot(aes(x = commited_lores_clusters, y = exprsn, colour = commited_lores_clusters)) +
  scale_color_manual(values = cmtd_clstr_cols) +
  geom_boxplot() +
  facet_nested_wrap(vars(Gene), nrow = 2) +
  theme_classic() +
  theme(strip.text = element_text(size = 13, face = "bold"),
        text = element_text(size = 13),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank()) 
```

```{r, message=FALSE, warning=FALSE, eval=T, fig.width=5.5, fig.height=3.5}
asx_gns_mdata_umap2_gn_exprsn %>%
  # group_by(gene) %>%
  ggplot(aes(x = commited_lores_clusters, y = log2(exprsn), colour = commited_lores_clusters)) +
  scale_color_manual(values = cmtd_clstr_cols) +
  geom_boxplot() +
  facet_nested_wrap(vars(Gene), nrow = 2) +
  theme_classic() +
  theme(strip.text = element_text(size = 13, face = "bold"),
        text = element_text(size = 13),
        legend.position = "bottom",
        axis.text = element_blank(),
        axis.ticks = element_blank()) 
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_ap2g_DE_long %>%
  filter(gene_id %in% our_uniq_gnset$gene_id) 

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_ap2g_DE_long_signfDE %>% filter(if_any(starts_with("signf_DE"), ~.x == "DE"))%>% select(Gene, contains("Foldchange"), contains("FDR"), contains("signf_DE")) %>% distinct(Gene, .keep_all = T)

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_ap2g_DE_long_signfDE %>% filter(if_any(starts_with("signf_DE"), ~.x == "DE"))%>% filter(Gene %in% josling_42_gns) %>% select(Gene, contains("Foldchange"), contains("FDR"), contains("signf_DE")) %>% distinct(Gene, .keep_all = T)

```

```{r, warning=FALSE, message=FALSE, fig.width=9.9, fig.height=5}
#LSD(PF3D7-1472200), HDA

josling_ap2g_DE_long_signfDE_tbl <- josling_ap2g_DE_long_signfDE %>%
  filter(gene_id %in% c(our_uniq_gnset$gene_id, "PF3D7-1102500")) %>%
  mutate(Gene_fdr = paste0(Gene, " FDR:", FDR_1_11, "")) 

jos_gn_levels <- c(sort(unique(josling_ap2g_DE_long_signfDE_tbl$Gene_fdr)[2:6]), unique(josling_ap2g_DE_long_signfDE_tbl$Gene_fdr)[1])

josling_ap2g_DE_long_signfDE_tbl %>%
  mutate(across(Gene_fdr, ~factor(.x, levels = jos_gn_levels) )) %>%
  mutate(across(timepoint, ~factor(str_remove_all(., "pi|P"), levels = timepoint_levels))) %>%
  mutate(across(ap2g_cond, ~str_to_sentence(str_remove(., "ap2g")))) %>%
  dplyr::rename("ap2g" = ap2g_cond) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g), se = F) +
  scale_color_manual(values = ap2g_cols) +
  # geom_text(aes(x = 1.5, y = 4, label = FDR_1_11), colour = "black", size = 3, fontface = "plain") +
  geom_text(aes(x = 2, y = 4, label = FDR_1_3), colour = "black", size = 5, fontface = "plain") +
  geom_text(aes(x = 5.5, y = 4, label = FDR_4_7), colour = "black", size = 5, fontface = "plain") +
  geom_text(aes(x = 9.5, y = 4, label = FDR_8_11), colour = "black", size = 5, fontface = "plain") +
  geom_vline(aes(xintercept = 1), colour = 'lightgrey') +
  geom_vline(aes(xintercept = 3), colour = 'lightgrey') +
  geom_vline(aes(xintercept = 4), colour = 'lightgrey') +
  geom_vline(aes(xintercept = 7), colour = 'lightgrey') +
  geom_vline(aes(xintercept = 8), colour = 'lightgrey') +
  geom_vline(aes(xintercept = 11), colour = 'lightgrey') +
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.1))) +
  facet_wrap(vars(Gene_fdr)) +
  labs(x = "Timepoint (hpi)", y= "Expression") +
  theme_classic() +
    theme(axis.text.x = element_text(angle=90, hjust = 1),
          legend.position = "bottom",
          text = element_text(size = 17),
        axis.text = element_text(size = 17),
        strip.text = element_text(size = 17, colour = "black", face = "bold"),
        # axis.title = element_blank()
        ) +
  guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))


```

```{r, warning=FALSE, message=FALSE, fig.width=8.2, fig.height=5}
#LSD(PF3D7-1472200), HDA

josling_ap2g_DE_long_signfDE %>%
  filter(gene_id %in% c(our_uniq_gnset$gene_id, "PF3D7-1467600")) %>%
  mutate(Gene_fdr = paste0(Gene, " (", FDR_1_11, ")")) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond), se = F) +
  # geom_text(aes(x = 1.5, y = 4, label = FDR_1_11), colour = "black", size = 3, fontface = "plain") +
  geom_text(aes(x = 2, y = 4, label = FDR_1_3), colour = "black", size = 5, fontface = "plain") +
  geom_text(aes(x = 5.5, y = 4, label = FDR_4_7), colour = "black", size = 5, fontface = "plain") +
  geom_text(aes(x = 9.5, y = 4, label = FDR_8_11), colour = "black", size = 5, fontface = "plain") +
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.1))) +
  facet_wrap(vars(Gene_fdr)) +
    # labs(title = Gene) +
  theme_classic() +
    theme(axis.text.x = element_text(angle=90, hjust = 1),
          legend.position = "bottom",
          text = element_text(size = 20),
        axis.text = element_text(size = 17),
        # axis.ticks = element_blank(),
        # axis.title = element_blank()
        ) +
  guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))


```



```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
# AP2G, SURF1.2, GEXP02
map2(c("PF3D7-1222600", "PF3D7-0113600", "PF3D7-1102500"), c("ap2g", "surf1.2", "gexp02"), ~{
  josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    labs(title = .y) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom")
})


```


```{r, message=FALSE, warning=FALSE, eval=T, fig.width=6, fig.height=9}
asx_gns_mdata_umap2_gn_exprsn 

```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=6}
asx_gns_mdata_umap2
```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=6}
FeaturePlot(v3_early_asex_jnd_noM55_jcC, features = our_uniq_gnset$gene_id, reduction = "umap", order = T, ncol = 3)

```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=6}
VlnPlot(v3_early_asex_jnd_noM55_jcC, features = our_uniq_gnset$gene_id, group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE, fig.width=9, fig.height=6}
VlnPlot(v3_early_asex_jnd_noM55_jcC, features = our_uniq_gnset$gene_id, group.by = "seurat_clusters")

```



#################START Write metadata to excel sheet for manuscript ###############################

Write the GO terms from above into file
```{r}
commited_de_mrgd_anot_noM55_jcC_dataS <- commited_de_mrgd_anot_noM55_jcC %>% filter(cluster == "committed") %>% select(c(gene, Gene, everything()))

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

josling_ap2g_DE_long %>%
  pivot_wider(names_from = timepoint, values_from = expression) 
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

josling_ap2g_DE_wide <- josling_ap2g_DE_long %>%
  pivot_longer(cols = matches("^(FDR|Foldchange).*"), names_to = c("metric", "pair"), names_pattern = "(FDR|Foldchange)_(.*)", values_to = "value") %>%
  pivot_wider(names_from = metric, values_from = value) %>%
  mutate(signf_DE = if_else(FDR <= 0.1 & Foldchange >= 1.5,"DE", "noDE")) %>%
  pivot_wider( names_from = pair, values_from = c(Foldchange, FDR, signf_DE), names_glue = "{.value}_{pair}") %>%
  pivot_wider(names_from = timepoint, values_from = expression)

josling_ap2g_DE_wide
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
early_late_gns_gexp02 <- map(list(clustr_gexp2_low_gns_tbl[["noM55_jcC"]], clustr_gexp2_up_gns_tbl[["noM55_jcC"]]), 
                             ~.x %>% 
                               select(c(gene, Gene, everything())) %>%
                               left_join(., josling_ap2g_DE_wide %>% select(-c(Gene)), by = c("gene" ="gene_id")) %>%
                               mutate(across(ap2g_cond, ~str_to_sentence(str_remove(., "ap2g")))) %>%
                               dplyr::rename("ap2g" = ap2g_cond)
)


```


```{r, fig.height=3.8, fig.width=3.8}
cmtd_gns_overlap_dataS <- cmtd_gns_overlap %>%
  select(gene_id, Gene, everything()) %>%
  rename("Upregulated in early GCR" = Up_early,
         "Upregulated in late GCR" = Up_later,
         "Upregulated in Votborg-Novél et al 2025 GCR" = LP_cmtd,
         "Upregulated in Prajapati et al 2020 GCR" = praj_fld,
         "Upregulated in Josling et al ap2gOn" = Jos35_Up,
         "Upregulated in Llorà-Batlle et al 2020 early (40-45hpi) commitment" = lb_early,
         "Upregulated in Llorà-Batlle et al 2020 late (5-20hpi) commitment" = lb_later,
         "Upregulated in Poran et al 2017 GCR" = Poran_Up)
```


```{r, fig.height=3.8, fig.width=3.8}
cmtd_gns_overlap_natural_infection_uniques_dataS <- cmtd_gns_overlap_dataS %>% 
  filter(gene_id %in% our_uniq_gnset$gene_id) %>%
  left_join(., josling_ap2g_DE_wide %>% select(-c(Gene)), by = "gene_id") 


```


```{r}
## Join the datasets in a list and name the lists for friendly tab names
commitment_dataS_list <- list(commited_de_mrgd_anot_noM55_jcC_dataS, early_late_gns_gexp02[[1]], early_late_gns_gexp02[[2]], cmtd_gns_overlap_dataS, cmtd_gns_overlap_natural_infection_uniques_dataS) %>% 
  set_names(c("GCR upregulated genes", "early GCR upregulated genes", "late GCR upregulated genes", "commitment genes overlap", "field unique GCR genes"))

```



```{r}
##Export the list of field strain_strain DE dfs with appropriate names to workbook with each stage as a sheet
createWorksheet(commitment_dataS_list, "data/processed/Pf/m2_all/asex_integration_comtmnt_DE_figs/data S4 - Genes upregulated in committed rings and comparison to other datasets.xlsx")

```

#################END Write metadata to excel sheet for manuscript ###############################



