---
title: "MSC QC and preprocessing"
date: '2022-07-11'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(sctransform)
  # library(rgl)
  library(patchwork)
  # library(DoubletFinder)
  library(ggrepel)
  library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  library(plotly)
  library(Hmisc)
  library(rmarkdown)
  library(RColorBrewer)
  # library(Platypus)
  library(ggvenn)
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  library(ggh4x)
  library(SingleCellExperiment)
  # library(scmap)
  # library(ggsankey)
  # library(findPC)
  library(BPCells)
  # library(furrr)
  # library(future.batchtools)
  library(parallel)
  library(EnhancedVolcano)
  library(DropletUtils)
  library(ggforce)
  library(slingshot)
  library(GGally)
  library(gtools)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("map", "purrr")
  conflict_prefer("count", "dplyr")
  conflicts_prefer(base::saveRDS)
  conflicts_prefer(SingleCellExperiment::colData)
  conflicts_prefer(base::as.data.frame)
})
```







# MSC Preprocessing


## setup, read input files and exploratory visualizations and general functions


```{r, setup}
# library(reticulate)
# # use_condaenv("/Users/jr35/envs/scv_jupcon/", required = TRUE)
# # matplotlib <- reticulate::import("matplotlib")
# # matplotlib$use("Agg", force = TRUE)
# 
# knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali')
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2')

```

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/nfs/users/nfs_j/jr35/multipurpose_scripts/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```


```{r, message=F, warning=F}
##Read BPcells matrices
# v3_m2_asex_cr_filt_norm_harmony_jnd <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_(all|noM55)_jc._norm_harmony_jnd.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
# v3_m2_asex_cr_filt_norm_harmony_jnd <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_noM55_jc._norm_harmony_jnd.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
# v3_m2_asex_cr_filt_norm_harmony_jnd <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_noM55_jc._norm_harmony_jnd3.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
# v3_m2_asex_cr_filt_norm_harmony_jnd <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_noM55_jc._norm_harmony_jnd_slim3.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
v3_m2_asex_cr_filt_norm_harmony_jnd <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_(noM55|all)_jcC_norm_harmony_jnd_slim3.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/m2_all/v3_m2_asex_cr_filt_|_norm.*'))) %>% 
  # .[sample_name] %>%
  map(readRDS)

```

```{r, warning=FALSE, message=FALSE}
# imap(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_PC_man_clusters") + labs(title = .y))

```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x, label = F, reduction = "umap.harmony_PC_man", group.by = "stage_fld") + scale_color_manual(values = sp_fld_colors2))
```


```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x, label = F, reduction = "umap.harmony_PC_man", group.by = "stage_fld", dims = c(1, 3)) + scale_color_manual(values = sp_col_new))
```

```{r, warning=FALSE, message=FALSE}
# imap(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x, reduction = "umap.harmony_PC_man", cells.highlight = colnames(.x[,.x@meta.data$stage_fld == "Unassigned"])) + labs(title = .y))

```

```{r, warning=FALSE, message=FALSE}
# for (nm in names(v3_m2_asex_cr_filt_norm_harmony_jnd)) {
#   v3_m2_asex_cr_filt_norm_harmony_jnd[[nm]][["RNA"]]$data <- as(v3_m2_asex_cr_filt_norm_harmony_jnd[[nm]][["RNA"]]$data, Class = "dgCMatrix")
# }
```


```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~SetIdent(.x, value = "harmony_PC_man_clusters"))
```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x))
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd[[1]]
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd[[1]]@meta.data %>% count(orig.ident)
```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~VlnPlot(.x, features = "pf_gn_count", group.by = "donor"))
```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~VlnPlot(.x, features = "nFeature_RNA", group.by = "donor"))
```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~VlnPlot(.x, features = "nFeature_RNA", group.by = "donor"))
```

```{r, warning=FALSE, message=FALSE}
cb_asex_de <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data"))

```


```{r, warning=FALSE, message=FALSE}
cb_asex_de_anot <- map(cb_asex_de, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

cb_asex_de_anot
```


```{r, warning=FALSE, message=FALSE}

# map2(cb_asex_de_anot, c("9", "9", "3", "8"), ~.x %>% filter(cluster == .y & avg_log2FC >0))
map2(cb_asex_de_anot, c("3", "8"), ~.x %>% filter(cluster == .y & avg_log2FC >0))

```

```{r, warning=FALSE, message=FALSE}
map(cb_asex_de_anot, ~.x %>% filter(cluster %in% c("5", "13") & avg_log2FC >0) %>% arrange(desc(avg_log2FC)))
```

```{r, warning=FALSE, message=FALSE}
# ap2-g
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", label = T, features = "PF3D7-1222600", order = T))

```

```{r, warning=FALSE, message=FALSE}
# ap2-g
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", label = T, features = "PF3D7-0113600", order = T))

```

```{r, warning=FALSE, message=FALSE}
# pfs16 gexp02
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", label = T, features = "PF3D7-0406200",  order = T))
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", label = T, features = "PF3D7-1102500", order = T))

```

```{r, warning=FALSE, message=FALSE}
# gene_count
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~VlnPlot(.x, features = c( "nCount_RNA"), pt.size = 0))
```

```{r, warning=FALSE, message=FALSE}
# pfs16 gexp02
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", label = T, features = "nFeature_RNA"))
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FeaturePlot(subset(.x, subset = nCount_RNA <2500), reduction = "umap.harmony_PC_man", label = T, features = "nCount_RNA"))

```

```{r, warning=FALSE, message=FALSE}
map(cb_asex_de_anot, ~.x %>% filter(cluster %in% c("13") & avg_log2FC >0))
```

```{r, warning=FALSE, message=FALSE}
map(cb_asex_de_anot, ~.x %>%
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% filter(p_val_adj< 0.05) %>% group_by(Gene) %>% slice_max(order_by = avg_log2FC))

```


```{r, warning=FALSE, message=FALSE}
# cb_asex_de_anot %>% filter(gene %in% ap2g_gns_down) %>% arrange(avg_log2FC) %>% filter(p_val_adj< 0.05) %>% group_by(Gene) %>% slice_min(order_by = avg_log2FC)
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FindNeighbors(.x, dims = 1:8, reduction = "integrated.harmony_PC_man"))
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FindClusters(.x, resolution = 0.15, cluster.name = "harmony_clusters_coarse", random.seed = 949))
```


```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_coarse"))

```


```{r, warning=FALSE, message=FALSE}
# Idents(v3_m2_asex_cr_filt_norm_harmony_jnd) <- "harmony_clusters_coarse"
```

```{r, warning=FALSE, message=FALSE, eval=T}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(v3_m2_asex_cr_filt_norm_harmony_jnd)) {
  Idents(v3_m2_asex_cr_filt_norm_harmony_jnd[[i]]) <- "harmony_clusters_coarse"
  
}

```

```{r, warning=FALSE, message=FALSE}
cb_asex_de_coarse <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data"))
```

```{r, warning=FALSE, message=FALSE}
cb_asex_de_coarse_anot <- map(cb_asex_de_coarse, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

```

```{r, warning=FALSE, message=FALSE}
# ap2-g
map(cb_asex_de_coarse_anot, ~.x %>% 
      filter(p_val_adj == 0 & avg_log2FC > 0) %>% 
      arrange(desc(avg_log2FC)) %>% 
      group_by(cluster) %>% 
      slice_min(order_by = p_val_adj, n = 10) %>% 
      slice_max(avg_log2FC, n =10))

```

```{r, warning=FALSE, message=FALSE, fig.width=16, fig.height=3}
map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_coarse", split.by = "donor"))
```

```{r, warning=FALSE, message=FALSE, eval=F}
cb_asex_de_coarse_mast <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))
```



```{r, warning=FALSE, message=FALSE, eval=F}
# saveRDS(cb_asex_de_coarse_mast, "data/processed/Pf/cb_asex_de_coarse_mast3.RDS")

```

```{r}
# imap(v3_m2_asex_cr_filt_all, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/v3_m2_asex_cr_filt_all_", .y, ".RDS")))
# imap(cb_asex_de_coarse_mast, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/cb_asex_de_coarse_mast_", .y, "_3.RDS")))

```

```{r, warning=FALSE, message=FALSE, eval=F}
# cb_asex_de_coarse_mast <- readRDS("data/processed/Pf/cb_asex_de_coarse_mast3.RDS")

```


```{r, warning=FALSE, message=FALSE, eval=F}
cb_asex_de_coarse_mast_anot <- map(cb_asex_de_coarse_mast, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

```



```{r, warning=FALSE, message=FALSE, eval=F}
# ap2-g
map(cb_asex_de_coarse_mast_anot, ~.x %>% 
      filter(p_val_adj == 0 & avg_log2FC > 0) %>% 
      arrange(desc(avg_log2FC)) %>% 
      group_by(cluster) %>% 
      slice_min(order_by = p_val_adj, n = 10) %>% 
      slice_max(avg_log2FC, n =10))

```

```{r, warning=FALSE, message=FALSE, eval=F}
cb_asex_de_coarse_anot %>% filter(gene == "PF3D7-0930300")
```


```{r}
# library(GGally)

lowerfun <- function(data, mapping) {
  ggplot(data = data, mapping = mapping) +
    geom_point(size = 0.3, alpha = 0.3) +
    geom_abline(slope = 1, intercept = 0, color = "blue")
}

```

```{r, eval=F}
FetchData(object = v3_m2_asex_cr_filt_norm_harmony_jnd[, v3_m2_asex_cr_filt_norm_harmony_jnd$harmony_pc_man_clusters %in% c("13", "5", "12", "2")], 
          vars = c("PF3D7-1467600", "PF3D7-1472200", "PF3D7-1473200", "PF3D7-1222600", "PF3D7-0406500", "PF3D7-0424400", "PF3D7-1102500"),
          slot = "data") %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)
  
```


```{r, eval=F}
FetchData(object = v3_m2_asex_cr_filt_norm_harmony_jnd[, v3_m2_asex_cr_filt_norm_harmony_jnd$harmony_pc_man_clusters %in% c("13", "5", "12", "2")], 
          vars = c("PF3D7-1467600", "PF3D7-1472200", "PF3D7-1473200", "PF3D7-1222600", "PF3D7-0406500", "PF3D7-0424400"),
          slot = "data") %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)
  
```


############# END find all markers to test for merozoites #############
############# START Recluster to make the early and late commited gametocytes form distint clusters #############

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- map(v3_m2_asex_cr_filt_norm_harmony_jnd, ~FindNeighbors(.x, dims = 1:10, reduction = "integrated.harmony_PC_man"))

v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~FindClusters(.x, resolution =1.3, cluster.name = "harmony_clusters_fine_erings", random.seed = 949))

v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~RunUMAP(.x, reduction = "integrated.harmony_PC_man", dims = 1:10, seed.use = 949, reduction.name = "umap.harmony_PC_man", n.components = 3L, min.dist = 0.3, n.neighbors = 30))

```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_fine_erings"))

map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_fine_erings", dims = c(1,3)))

map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, c("16", "18"), ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", cells.highlight = colnames(.x[,.x$harmony_clusters_fine_erings %in% .y]), dims = c(1,3), sizes.highlight = 0.01))
map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, c("16", "18"), ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", cells.highlight = colnames(.x[,.x$harmony_clusters_fine_erings %in% .y]), dims = c(1,2), sizes.highlight = 0.01))

```



```{r, warning=FALSE, message=FALSE, eval=F}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs[["noM55_jcC"]]@reductions[["umap.harmony_PC_man"]]@cell.embeddings),
        x = ~umapharmonyPCman_1, y = ~umapharmonyPCman_2, z = ~umapharmonyPCman_3,
        color= v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs[["noM55_jcC"]]@meta.data$harmony_clusters_fine_erings,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE}
map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, c("16", "18"), ~DimPlot(.x, label = T, reduction = "integrated.harmony_PC_man", cells.highlight = colnames(.x[,.x$harmony_clusters_fine_erings %in% .y]), dims = c(1,3), sizes.highlight = 0.01))

```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
#SURF1.2 gexp02 Pfg14-744 ap2g
map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man",  label = T, features = c("PF3D7-0113600", "PF3D7-1102500", "PF3D7-1477300", "PF3D7-1222600"), order = T, dims = c(1,3), ncol = 4))
```

############# START Findmarkers in the commited cluster in the overal asexual integrated object to avoid subclustering figure #############


```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~SetIdent(.x, value = "harmony_clusters_fine_erings"))

```


```{r, warning=FALSE, message=FALSE}
object_nms <- c("all_jcC", "noM55_jcC")

```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
commited_de_hi_res <- map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~FindAllMarkers(.x, assay = "RNA", slot = "data"))
# commited_de <- map(v3_early_asex_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data"))
```


```{r, warning=FALSE, message=FALSE, eval=FALSE}
commited_de_hi_res_anot <- map(commited_de_hi_res, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

# commited_de_anot
```

```{r, warning=FALSE, message=FALSE}
# saveRDS(commited_de_anot, "data/processed/Pf/m2_all/commited_de_anot3.RDS")
```

```{r, eval=FALSE}
imap(commited_de_hi_res_anot, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/integrated_asex_exploration/commited_de_hi_res_anot_", .y, ".RDS")))

```

```{r}
commited_de_hi_res_anot <- map(object_nms, ~readRDS(paste0("data/processed/Pf/m2_all/integrated_asex_exploration/commited_de_hi_res_anot_", .x, ".RDS")))

```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_fine_erings", dims = c(1,3)))

```

```{r}
early_ring_clstr <- list(list(c("10","5")), list(c("13", "4")))
early_comtd_clstr <- list(list(c("12","13")), list("15"))
late_comtd_clstr <- list(list("16"), list("18"))

merged_clstrs <- Map(c, early_ring_clstr, early_comtd_clstr, late_comtd_clstr)
```



```{r, warning=FALSE, message=FALSE}
map2(commited_de_hi_res_anot, merged_clstrs,  ~.x %>% 
  filter(cluster %in% .y) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0 & pct.2 < 0.1) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 10))
```



```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_asex_ring_v_commited_gns <- map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, merged_clstrs, ~FindMarkers(.x, ident.1 = c(.y[[2]], .y[[3]]), ident.2 = .y[[1]], assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_asex_ring_v_commited_gns_anot <- map(cmtd_rings_de_asex_ring_v_commited_gns, ~.x %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% rownames_to_column("gene") %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

cmtd_rings_de_asex_ring_v_commited_gns_anot
```

```{r, eval=FALSE}
imap(cmtd_rings_de_asex_ring_v_commited_gns_anot, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/integrated_asex_exploration/cmtd_rings_de_asex_ring_v_commited_gns_anot_", .y, ".RDS")))

```

```{r}
cmtd_rings_de_asex_ring_v_commited_gns_anot <- map(object_nms, ~readRDS(paste0("data/processed/Pf/m2_all/integrated_asex_exploration/cmtd_rings_de_asex_ring_v_commited_gns_anot_", .x, ".RDS")))

```

```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_asex_ring_v_commited_gns_anot, ~.x %>% 
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```


```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_asex_ring_v_commited_gns_anot, ~.x %>% 
  filter(avg_log2FC < 0) %>%
  arrange(desc(abs(avg_log2FC))) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```


```{r}

merged_clstrs <- Map(c, early_ring_clstr, early_comtd_clstr, late_comtd_clstr)
```


```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_late_v_all_gns <- map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, merged_clstrs, ~FindMarkers(.x, ident.1 = .y[[3]], ident.2 = c(.y[[1]], .y[[2]]), assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_late_v_all_gns_anot <- map(cmtd_rings_de_late_v_all_gns, ~.x %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% rownames_to_column("gene") %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

cmtd_rings_de_late_v_all_gns_anot
```

```{r, eval=FALSE}
imap(cmtd_rings_de_late_v_all_gns_anot, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/integrated_asex_exploration/cmtd_rings_de_late_v_all_gns_anot_", .y, ".RDS")))

```

```{r}
cmtd_rings_de_late_v_all_gns_anot <- map(object_nms, ~readRDS(paste0("data/processed/Pf/m2_all/integrated_asex_exploration/cmtd_rings_de_late_v_all_gns_anot_", .x, ".RDS")))

```

```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_late_v_all_gns_anot, ~.x %>% 
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```


```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_late_v_all_gns_anot, ~.x %>% 
  filter(avg_log2FC < 0) %>%
  arrange(desc(abs(avg_log2FC))) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```


```{r}
# gogo()
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_early_v_all_gns <- map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, merged_clstrs, ~FindMarkers(.x, ident.1 = .y[[2]], ident.2 = c(.y[[1]], .y[[3]]), assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))
```

```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_early_v_all_gns_anot <- map(cmtd_rings_de_early_v_all_gns, ~.x %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% rownames_to_column("gene") %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

cmtd_rings_de_early_v_all_gns_anot
```

```{r, eval=FALSE}
imap(cmtd_rings_de_early_v_all_gns_anot, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/integrated_asex_exploration/cmtd_rings_de_early_v_all_gns_anot_", .y, ".RDS")))

```

```{r}
cmtd_rings_de_early_v_all_gns_anot <- map(object_nms, ~readRDS(paste0("data/processed/Pf/m2_all/integrated_asex_exploration/cmtd_rings_de_early_v_all_gns_anot_", .x, ".RDS")))

```

```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_early_v_all_gns_anot, ~.x %>% 
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```


```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_early_v_all_gns_anot, ~.x %>% 
  filter(avg_log2FC < 0) %>%
  arrange(desc(abs(avg_log2FC))) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```

Based on clusters markers above we can now specify the commited clusters


```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, merged_clstrs, ~.x@meta.data %>%
                                                          rownames_to_column("bcode") %>%
                                                          mutate(commitment_status = 
                                                                   case_when(harmony_clusters_fine_erings %in% .y[[1]] ~ "early asexual rings",
                                                                             harmony_clusters_fine_erings %in% .y[[2]] ~ "early committed rings",
                                                                             harmony_clusters_fine_erings %in% .y[[3]] ~ "late committed rings",
                                                                             T ~ stage_fld)) %>%
                                                          column_to_rownames("bcode") %>%
                                                          select(commitment_status) %>%
                                                          AddMetaData(.x, .)
                                                        )

```


```{r, warning=FALSE, message=FALSE}
# gogo()
# imap(v3_m2_asex_cr_filt_norm_harmony_jnd, ~saveRDS(., paste0("data/processed/Pf/m2_all/v3_m2_asex_cr_filt_", .y,"_norm_harmony_jnd.RDS")))
# imap(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~saveRDS(., paste0("data/processed/Pf/m2_all/v3_m2_asex_cr_filt_", .y,"_norm_harmony_jnd_fine_clustrs3.RDS")))

imap(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~saveRDS(., paste0("data/processed/Pf/m2_all/integrated_asex_exploration/v3_m2_asex_cr_filt_", .y,"_norm_harmony_jnd_fine_clustrs.RDS")))

```


############# END Findmarkers in the commited cluster in the overal asexual integrated object to avoid subclustering figure #############


```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~DimPlot(.x, label = F, reduction = "umap.harmony_PC_man", group.by = "stage_fld"))
```

```{r, warning=FALSE, message=FALSE}
map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~FeaturePlot(.x, label = T, reduction = "umap.harmony_PC_man", features = "scf_score_rand", order = T, dims = c(1,3)))
```

############# START Findmarkers in the commited cluster in the main figure #############


############# END Findmarkers in the commited cluster in the main figure #############

############# START SK21 gene likely commited vs non-commited #############

```{r, warning=FALSE, message=FALSE}
# pfs16
map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~FeaturePlot(.x[, .x$harmony_clusters_coarse %in% c("2")], reduction = "umap.harmony_PC_man", label = T, features = "nCount_RNA"))

```


```{r}
# v3_early_asex <- map(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, ~.x[, .x$harmony_clusters_coarse %in% c("2")])
# Clusters to subset
# er_clstr_set <- list(c("6", "4"), "5")
# er_clstr_set <- list(c("1", "10"), c("6", "11"))
# er_clstr_set <- list(c("4", "10"), c("6", "12"))
# er_clstr_set <- list(c("4", "10"), c("7", "12"))

# er_clstr_set <- list(c("5","10","12","13","16"), c("4","13","15","18"))
# 
# v3_early_asex <- map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, er_clstr_set, ~.x[, .x$harmony_clusters_fine_erings %in% .y])

v3_early_asex <- map2(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs, er_clstr_set, ~subset(.x, subset = commitment_status %in% c("early asexual rings", "early committed rings", "late committed rings")))


```


```{r, warning=FALSE, message=FALSE}
# gene_count
map(v3_early_asex, ~VlnPlot(.x, features = "nFeature_RNA", group.by = "harmony_clusters_fine_erings") + geom_hline(yintercept = 600))
# map(v3_early_asex, ~VlnPlot(.x, features = "nCount_RNA") + geom_hline(yintercept = 2000))
map(v3_early_asex, ~VlnPlot(.x, features = "nCount_RNA", group.by = "harmony_clusters_fine_erings") + geom_hline(yintercept = 5000))

```

```{r, warning=FALSE, message=FALSE}
# gene_count
map(v3_early_asex, ~VlnPlot(.x, features = "nFeature_RNA", group.by = "harmony_clusters_fine_erings") + geom_hline(yintercept = 500))
# map(v3_early_asex, ~VlnPlot(.x, features = "nCount_RNA") + geom_hline(yintercept = 2000))
map(v3_early_asex, ~VlnPlot(.x, features = "nCount_RNA", group.by = "harmony_clusters_fine_erings") + geom_hline(yintercept = 1000))

```


```{r, warning=FALSE, message=FALSE}
# gene_count
v3_early_asex <- map(v3_early_asex, ~subset(.x, subset = nFeature_RNA < 500))
# v3_early_asex <- map(v3_early_asex, ~subset(.x, subset = nCount_RNA < 2500))
v3_early_asex <- map(v3_early_asex, ~subset(.x, subset = nCount_RNA < 1000))

```

```{r, warning=FALSE, message=FALSE}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_fine_erings"))
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "integrated.harmony_PC_man", group.by = "harmony_clusters_fine_erings"))

```



```{r, warning=FALSE, message=FALSE, eval=T}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(v3_early_asex)) {
  DefaultAssay(v3_early_asex[[i]]) <- "RNA"
  v3_early_asex[[i]][["RNA"]]$data <- NULL 
  # v3_early_asex[[i]][["RNA"]]$scale.data <- NULL 
}
```

```{r}

## Donors with more than 50 cells
map(v3_early_asex, ~table(.x$donor))
```

```{r}

## Donors with more than 50 cells
don_50 <- map(v3_early_asex, ~table(.x$donor) %>% data.frame() %>% filter(Freq >= 125) %>% pull(Var1) %>% as.character)
```

```{r}
## Retain only donors with more than 50 cells
v3_early_asex <- map2(v3_early_asex, don_50, ~subset(.x, subset = donor %in% .y))
```

```{r, warning=FALSE, message=FALSE}
# for (nm in names(v3_early_asex)) {
#   v3_early_asex[[nm]][["RNA"]]$counts <- as(v3_early_asex[[nm]][["RNA"]]$counts, Class = "dgCMatrix")
# }
```

```{r}
v3_early_asex <- map(v3_early_asex, ~split(.x, f = .x$donor, layers = "counts"))
```

```{r}
# v3_early_asex <- map(v3_early_asex, ~split(.x, f = .x$donor, layers = "counts"))
```



```{r, warning=FALSE, message=FALSE}
v3_early_asex <- map(v3_early_asex, ~NormalizeData(.x))
v3_early_asex <- map(v3_early_asex, ~FindVariableFeatures(.x, selection.method = "vst", nfeatures =  100))
v3_early_asex <- map(v3_early_asex, ~ScaleData(.x, verbose = F))
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3_early_asex <- map(v3_early_asex, ~RunPCA(.x, verbose = FALSE, npcs = 10))

```

```{r, warning=FALSE, message=FALSE}

v3_early_asex <- map(v3_early_asex, ~FindNeighbors(.x, dims = 1:5, reduction = "pca"))
# v3_early_asex <- map(v3_early_asex, ~FindClusters(.x, resolution = 0.5))
v3_early_asex <- map(v3_early_asex, ~FindClusters(.x, resolution = 0.25))

```


```{r, warning=FALSE, message=FALSE}
# run sctransform
map(v3_early_asex, ~DimPlot(.x, reduction = "pca"))

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
map(v3_early_asex, ~DimPlot(.x, reduction = "pca"))

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)
set.seed(949)

v3_early_asex <- map(v3_early_asex, ~IntegrateLayers(
  object = .x,
  method = HarmonyIntegration,
  orig.reduction = "pca", 
  new.reduction = "harmony",
  # normalization.method = "SCT",
  verbose = F,
  # k.weight = 98,
))



# obj <- IntegrateLayers(
#   object = obj, method = HarmonyIntegration,
#   orig.reduction = "pca", new.reduction = "harmony",
#   verbose = FALSE
# )
```


```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
map(v3_early_asex, ~ElbowPlot(.x, reduction = "pca"))
```

```{r, warning=FALSE, message=FALSE}

v3_early_asex <- map(v3_early_asex, ~FindNeighbors(.x, dims = 1:4, reduction = "harmony"))
v3_early_asex <- map(v3_early_asex, ~FindClusters(.x, resolution = 0.3))
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3_early_asex <- map(v3_early_asex, ~RunUMAP(.x, dims = 1:4, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "harmony"))
# all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}

# v3sex <- FindNeighbors(v3sex, dims = 1:10, reduction = "integrated.harmony_pc_man")

```

```{r, warning=FALSE, message=FALSE}
# # These are now standard steps in the Seurat workflow for visualization and clustering
# v3sex <- RunUMAP(v3sex, dims = 1:10, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "integrated.harmony_pc_man")
# # v3sex <- FindNeighbors(v3sex, dims = 1:6, verbose = FALSE)
# # v3sex <- FindClusters(v3sex, verbose = FALSE, resolution = 0.1)
```

```{r, warning=FALSE, message=FALSE}
# v3sex <- FindClusters(v3sex, resolution = 0.2)
```


```{r, warning=FALSE, message=FALSE}
v3_early_asex <- map(v3_early_asex, ~FindClusters(.x, resolution = 0.05))
```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters", dims = c(1,3)))
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "harmony_clusters_fine_erings", dims = c(1,2)))


```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.5}
#SURF1.2 gexp02
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = c("PF3D7-0113600", "PF3D7-1102500"), order = T))
```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.5}
#SURF1.2 gexp02
map(v3_early_asex, ~FeaturePlot(.x, reduction = "harmony",  label = T, features = c("PF3D7-0113600", "PF3D7-1102500"), order = T))
```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "harmony", group.by = "seurat_clusters", dims = c(1,2)))
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "harmony", group.by = "seurat_clusters", dims = c(2,3)))
```

```{r, warning=FALSE, message=FALSE, eval=T}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(v3_early_asex)) {
  DefaultAssay(v3_early_asex[[i]]) <- "RNA"
  v3_early_asex[[i]]@assays$SCT <- NULL
  
}

v3_early_asex
```

```{r}
# # Write new BPCells backing files
# 
# data_dir <- paste0("data/processed/Pf/m2_all/bpc_matrices/v3_early_asex_tmp_3")
# # data_dir <- tempfile()
# 
# converted_assay <- CreateAssay5Object(
#   counts = write_matrix_dir(v3_early_asex[[1]][["RNA"]]$counts, file.path(data_dir, "counts")),
#   data = write_matrix_dir(v3_early_asex[[1]][["RNA"]]$data, file.path(data_dir, "data"))
# )
# 
# pbmc[["SCT"]] <- converted_assay

```


```{r}
# # Write new BPCells backing files
# 
# data_dir <- paste0("data/processed/Pf/m2_all/bpc_matrices/v3_early_asex_", names(v3_early_asex[i]), "_3")
# # data_dir <- tempfile()
# 
# converted_assay <- CreateAssay5Object(
#   counts = write_matrix_dir(pbmc[["SCT"]]$counts, file.path(data_dir, "counts")),
#   data = write_matrix_dir(pbmc[["SCT"]]$data, file.path(data_dir, "data"))
# )
# converted_assay$scale.data <- write_matrix_dir(
#   as(pbmc[["SCT"]]$scale.data, "dgCMatrix"), 
#   file.path(data_dir, "scale.data")
# )
# pbmc[["SCT"]] <- converted_assay

```

```{r}
# imap(v3_m2_asex_cr_filt_all, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/v3_m2_asex_cr_filt_all_", .y, ".RDS")))
imap(v3_early_asex, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/v3_early_asex_", .y, "_3.RDS")))


```



```{r, warning=FALSE, message=FALSE, eval=F}
# Display the combined datasets as an interactive 3D plot

map(v3_early_asex, ~plot_ly(data = data.frame(.x@reductions[["umap"]]@cell.embeddings),
        x = ~harmony_1, y = ~harmony_2, z = ~harmony_3,
        color= .x@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
))

```

```{r, warning=FALSE, message=FALSE}
# gogo()anot <- map(cb_asex_de, ~.x %>% filter(p_val_adj < 0.0
```

```{r, message=F, warning=F}
##Read BPcells matrices
bp_CbendrCr_prelim_stg_sct_m68C <- readRDS("data/processed/Pf/MSC68C/seu_obj/bp_CbendrCr_prelim_stg_sct.RDS")
bp_CbendrCr_prelim_stg_sct_m68 <- readRDS("data/processed/Pf/MSC68/seu_obj/bp_CbendrCr_prelim_stg_sct.RDS")
raw_seu_m68 <- list(bp_CbendrCr_prelim_stg_sct_m68C, bp_CbendrCr_prelim_stg_sct_m68) %>% set_names(c("m68C", "m68"))
```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.5}
#SURF1.2 gexp02
map(raw_seu_m68, ~FeaturePlot(.x, reduction = "umap",  label = T, features = c("PF3D7-0113600", "PF3D7-1102500"), order = T, pt.size = 0.01))
```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters", dims = c(1,3)))

```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=F}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters", dims = c(1,3)))

```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.65, eval=T}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "stage_fld"))
# map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", cells.highlight = colnames(.x[,.x$stage_fld == "Unassigned"])))

```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.65, eval=T}
map(v3_early_asex, ~.x@meta.data %>% count(stage_fld, seurat_clusters))

```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.65, eval=T}
map(v3_early_asex, ~.x@meta.data %>% count(seurat_clusters, stage_fld))

```

```{r, warning=FALSE, message=FALSE, eval=F}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_early_asex@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= v3_early_asex@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
# gene_count
map(v3_early_asex, ~VlnPlot(.x, features = c("nFeature_RNA", "nCount_RNA"), pt.size = 0) )
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# ap2-g pfs16 gexp02 gene_count
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1222600", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0406200", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1102500", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "nFeature_RNA", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "nCount_RNA", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "pf_umi_count", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "hs_umi_count", order = T))


```

```{r}
map(v3_early_asex, ~FetchData(object = .x[, .x$seurat_clusters %in% c("1", "3")], 
          vars = c("PF3D7-1467600", "PF3D7-1472200", "PF3D7-1473200", "PF3D7-1222600", "PF3D7-0406500", "PF3D7-0424400", "PF3D7-1102500"),
          slot = "data") %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = GGally::wrap("cor", size = 5)),
          progress = FALSE))
  
```


```{r, warning=FALSE, message=FALSE}
# ap2-g pfs16 gene_count
map(v3_early_asex, ~FeaturePlot(subset(.x, subset = nCount_RNA < 600), reduction = "umap",  label = T, features = "nCount_RNA", order = T))

```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3}
# ap2-g pfs16 gene_count
map(v3_early_asex, ~DimPlot(.x, split.by =  "donor", reduction = "umap", pt.size = 0.01))

```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3}
# ap2-g pfs16 gene_count
map(v3_early_asex, ~DimPlot(.x, split.by =  "donor", reduction = "umap", group.by = "harmony_clusters_fine_erings", pt.size = 0.01))

```

```{r, warning=FALSE, message=FALSE, fig.width=15, fig.height=3}
# ap2-g pfs16 gexp02 gene_count
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap", split.by =  "donor", label = T, features = "PF3D7-1222600", order = T, pt.size = 0.01))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap", split.by =  "donor", label = T, features = "PF3D7-0406200", order = T, pt.size = 0.01))
# map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap", split.by =  "donor", label = T, features = "PF3D7-1102500", order = T)
# map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap", split.by =  "donor", label = T, features = "nFeature_RNA", order = T)
# map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap", split.by =  "donor", label = T, features = "nCount_RNA", order = T)


```

```{r, warning=FALSE, message=FALSE}
v3_early_asex_jnd <- map(v3_early_asex, ~JoinLayers(.x))
```


```{r, warning=FALSE, message=FALSE}
# for (nm in names(v3_early_asex_jnd)) {
#   v3_early_asex_jnd[[nm]][["RNA"]]$data <- as(v3_early_asex_jnd[[nm]][["RNA"]]$data, Class = "dgCMatrix")
# }
```

```{r, warning=FALSE, message=FALSE}
# saveRDS(v3_early_asex_jnd, "data/processed/Pf/m2_all/v3_early_asex_jnd3.RDS")
```

```{r}
# imap(v3_early_asex_jnd, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/v3_early_asex_jnd_", .y, "_3.RDS")))

```


```{r, message=F, warning=F, eval=F}
# ##Read BPcells matrices
# # v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_(all|noM55)_jc._norm_harmony_jnd.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
# # v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs <- list.files("data/processed/Pf", pattern = "v3_m2_asex_cr_filt_noM55_jc._norm_harmony_jnd.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
# v3_early_asex_jnd <- list.files("data/processed/Pf", pattern = "v3_early_asex_jnd.*_3.RDS$", full.names = T, recursive = T, include.dirs = T) %>% 
#   set_names(nm = (str_remove_all(.,'data/processed/Pf/m2_all/v3_early_asex_jnd|_3.*'))) %>% 
#   # .[sample_name] %>%
#   map(readRDS)

```

```{r, warning=FALSE, message=FALSE}
commited_de <- map(v3_early_asex_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))
# commited_de <- map(v3_early_asex_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data"))
```


```{r, warning=FALSE, message=FALSE}
commited_de_anot <- map(commited_de, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

# commited_de_anot
```

```{r, warning=FALSE, message=FALSE}
# saveRDS(commited_de_anot, "data/processed/Pf/m2_all/commited_de_anot3.RDS")
```

```{r}
imap(commited_de_anot, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/commited_de_anot_", .y, "_3.RDS")))

```

```{r, warning=FALSE, message=FALSE}
map(commited_de_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 10))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% filter(p_val_adj< 0.05 & avg_log2FC>0) %>% group_by(cluster) %>% slice_min(order_by = p_val_adj, n = 10))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# HRPIII AROM RESA SURF1.2
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1372200", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0206300", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0102200", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0113600", order = T))
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# ap2-g pfs16 gexp02 gene_count
# map(v3_early_asex_jnd, ~FeaturePlot(.x@meta.data, reduction = "umap",  label = T, features = "PF3D7-1372200", order = T))

```

```{r, warning=FALSE, message=FALSE}
map(commited_de_anot, ~.x %>%  filter(cluster %in% c("0") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE}
## LSD, HDA1
map(commited_de_anot, ~.x %>%  filter(Gene %in% c("HDA1", "LSD2") ))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_anot, ~.x %>%  filter(cluster %in% c("1") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_anot, ~.x %>% filter(cluster %in% c("4") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_anot, ~.x %>%  filter(cluster %in% c("0") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# HRPIII AROM RESA SURF1.2
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0102200", order = T))
```

#### committed VS non-committed
```{r, fig.height=6, fig.width=6}
vlcano <- map(c(0:1), ~{
  clstr = .x
  map(commited_de_anot, ~EnhancedVolcano(.x %>% filter(cluster == clstr),
                lab = .x %>% filter(cluster == clstr) %>% pull(Gene),
                # selectLab = .y,
                x = 'avg_log2FC',
                y = 'p_val_adj',
                pCutoff = 0.05,
                FCcutoff = 2,
                pointSize = 1.5,
                labSize = 4.0,
                title = '',
                subtitle = "",
                titleLabSize = 0,
                subtitleLabSize = 0,
                axisLabSize = 22,
                drawConnectors = TRUE,
                # boxedLabels = TRUE,
                # legendLabels =c( "p_val_adj"),
                gridlines.minor = FALSE
                ) +
    theme(legend.position = "none"))

}) 
vlcano

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```


```{r, warning=FALSE, message=FALSE}
map(v3_early_asex_jnd, ~DimPlot(.x, label = T, reduction = "umap", group.by = "harmony_PC_man_clusters"))

```

```{r}
# ap2g_gns_up
```


##########START# COMPARISON WITH AP2G ACTIVATED SEXUAL RING GENES ##############

```{r, eval=T}
# josling_ap2g_DE = readxl::read_xlsx("../published_dsets/painter_29985403_48hr_dynamics/Expression_Peak_Timing_48h_IDC.xlsx", sheet = "Estimated Total Abundance", .name_repair = "universal") %>% mutate(across(Gene.ID, ~str_replace(., "_", "-"))) %>% rename_all(., ~str_remove(., "\\.\\.")) %>% rename_with(.fn = ~str_remove(paste0("hpi_", .x), ".hpi"), .cols = contains("hpi")) %>% rename("peak_hpi" = `hpi_Peak.Tim.`)
```

```{r, eval=T}
# josling_ap2g_DE = readxl::read_xlsx("../published_dsets/josling_ap2g_chipseq_2020/apg_activation_DE_gns_across_timepoints.xlsx", skip = 1, .name_repair = "universal", col_names = T)
josling_ap2g_DE = read_csv("../published_dsets/josling_ap2g_chipseq_2020/apg_activation_DE_gns_across_timepoints.csv", skip = 1, col_names = T, name_repair = "universal")

josling_ap2g_DE <- josling_ap2g_DE %>% 
  rename("gene_id" = `Gene.ID`) %>% 
  mutate(across(gene_id, ~str_replace(., "_", "-"))) %>% 
  # rename_all(., ~str_remove_all(., "\\.+\\.|TP\\.[0-9]+\\.+|TP[0-9]\\.+"))%>% 
  rename_all(., ~str_replace_all(str_remove_all(.,"\\.\\.\\.\\.|"), c("\\.\\.\\."="_", "\\.\\."="_", "TP\\."="TP", "\\."="")))%>% 
  # rename_with(.fn = ~str_remove(paste0("ap2gON_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi|",c(3:13)))) %>%
  rename_with(.fn = ~str_remove(paste0("ap2gON_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi",c(3:13), collapse = "|"))) %>%
  rename_with(.fn = ~str_remove(paste0("ap2gOFF_", .x), "[0-9]+$"), .cols = matches(paste0(".*hpi",c(14:24), collapse = "|"))) %>%
  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene_id" = "gene_id"))  %>% 
  mutate(across(Gene, ~ifelse(is.na(.x), gene_id, .x)))
 # glimpse

glimpse(josling_ap2g_DE)
```

```{r, eval=T}
# Rename the fold change and FDR columns since they were identically named and read_csv has automatically assigned the number of the columns. Make it more sensible by appending the right timepoint information

josling_ap2g_DE <- josling_ap2g_DE %>% 
  rename("Foldchange_1_3" = Foldchange_25) %>% 
  rename("FDR_1_3" = FDR_26) %>% 
  rename("Foldchange_4_7" = Foldchange_27) %>% 
  rename("FDR_4_7" = FDR_28) %>% 
  rename("Foldchange_8_11" = Foldchange_29) %>% 
  rename("FDR_8_11" = FDR_30) %>% 
  rename("Foldchange_1_11" = Foldchange_31) %>% 
  rename("FDR_1_11" = FDR_32) 

glimpse(josling_ap2g_DE)
```

```{r, warning=FALSE, message=FALSE}
commited_top_cluster_DE <- map(commited_de_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 50) %>% 
    slice_max(order_by = avg_log2FC, n = 50) )

commited_top_cluster_DE_gns <- map(commited_top_cluster_DE, ~.x %>% 
                                     select(cluster, gene) %>% 
                                     ungroup() %>% 
                                     mutate(across(cluster, ~paste0("cluster_",.))) 
                                   )

```


```{r, warning=FALSE, message=FALSE}
gns_list <- map(commited_top_cluster_DE_gns, ~tapply(.x$gene, .x$cluster, c))
```

```{r, warning=FALSE, message=FALSE}
josling_ap2g_DE_long <- josling_ap2g_DE %>% pivot_longer(cols = contains("hpi"), names_pattern = "(ap2g.*)_(TP.*)", names_to = c("ap2g_cond", "timepoint"), values_to = "expression") %>% mutate(across(timepoint, ~factor(., levels = mixedsort(unique(.)))))

josling_ap2g_DE_long %>% glimpse()
```


```{r, warning=FALSE, message=FALSE}
## save for plotting
saveRDS(josling_ap2g_DE_long, "data/processed/Pf/m2_all/asex_integration_comtmnt_DE_figs/josling_ap2g_DE_long.RDS")
```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
map(v3_early_asex_jnd, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1102500", order = T))

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
imap(gns_list[[2]], ~josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, group = gene_id, colour = Gene)) +
    geom_point() +
    geom_line() +
    facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "none"))


```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
imap(gns_list, ~{josling_ap2g_DE_long
       gn_list <- .x
       dset <- .y
       
       imap(gn_list, ~{
        josling_ap2g_DE_long %>%
        filter(gene_id %in% .x) %>%
        ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
        geom_point() +
        # geom_line(aes(colour = ap2g_cond, group = Gene)) +
        # facet_grid(ap2g_cond ~ .) +
        geom_smooth(aes(group = ap2g_cond)) +
        labs(title = paste(dset, .y, sep = "_")) +
        theme(axis.text.x = element_text(angle=45),
              legend.position = "bottom")
        
          })
})


```

##save the previous seurat clusters for separation of commited and non-commited 
```{r, fig.width=15, fig.height=5}
v3_early_asex_jnd <-  map2(v3_early_asex, v3_early_asex_jnd, ~.x@meta.data %>%
  mutate(commited_lores_clusters = seurat_clusters) %>%
  select(commited_lores_clusters) %>% 
  AddMetaData(.y, .)
)
```

```{r, warning=FALSE, message=FALSE}
# v3_early_asex_jnd <- map(v3_early_asex_jnd, ~FindClusters(.x, resolution = 0.2, random.seed = 489))
# v3_early_asex_jnd <- map(v3_early_asex_jnd, ~FindClusters(.x, resolution = 0.26))
v3_early_asex_jnd <- map2(v3_early_asex_jnd, c(0.18, 0.2), ~FindClusters(.x, resolution = .y, random.seed = 489))

map(v3_early_asex_jnd, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
# cmtd_rings <- list(c('2','1'), c('3', '0'), c('1', '4'), c('4', '6'))
# cmtd_rings_devlpmnt <- list(c('0','4'), c('2', '5'))
cmtd_rings_devlpmnt <- list(c('2','0'), c('0', '4'))
# cmtd_rings_devlpmnt <- list(c('1','3'), c('1', '3'))

```


```{r}
# imap(v3_early_asex_jnd, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/v3_early_asex_jnd_", .y, "_3.RDS")))
```

#############################START - CHECK IF THE HUMAN GENE PROFILE OF THE COMMITED PARASITES IS DIFFERENT FROM THAT OF THE NONCOMMITED ##########

```{r}
imap(v3_early_asex_jnd, ~VlnPlot(.x, features = "hs_gn_count", group.by = "seurat_clusters") + scale_y_log10())

```

```{r}
imap(v3_early_asex_jnd, ~VlnPlot(.x, features = "hs_gn_count", group.by = "seurat_clusters"))

```

```{r}
imap(v3_early_asex_jnd, ~VlnPlot(.x[,.x@meta.data$hs_gn_count < 100], features = "hs_gn_count", group.by = "seurat_clusters"))

```


```{r}
imap(v3_early_asex_jnd, ~VlnPlot(.x, features = "hs_umi_count", group.by = "seurat_clusters"))
imap(v3_early_asex_jnd, ~VlnPlot(.x[,.x@meta.data$hs_umi_count < 150], features = "hs_umi_count", group.by = "seurat_clusters"))

```


```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.5}
# Get the barcodes for the investigation of rbc content of the commited VS non-commited cells
map(v3_early_asex_jnd, ~.x@meta.data %>% 
      filter(pf_umi_count <200 & hs_umi_count <200) %>% 
      ggplot(aes(x = pf_umi_count, y = hs_umi_count)) +
      geom_point(size = 0.01) +
      scale_x_log10() +
      scale_y_log10() +
      facet_wrap(.~seurat_clusters, nrow = 1))
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
# Get the barcodes for the investigation of rbc content of the commited VS non-commited cells
map(v3_early_asex_jnd, ~.x@meta.data %>% 
      filter(pf_gn_count <200 & hs_gn_count <200) %>% 
      ggplot(aes(x = pf_gn_count, y = hs_gn_count)) +
      geom_point(size = 0.01) +
      geom_smooth()+
      scale_x_log10() +
      scale_y_log10() +
      facet_wrap(.~seurat_clusters, nrow = 1))
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
# Get the barcodes for the investigation of rbc content of the commited VS non-commited cells
v3_early_asex_bcodes_clustrs <- map2(v3_early_asex_jnd, cmtd_rings_devlpmnt, ~{
  .x@meta.data %>% 
    select(seurat_clusters, donor) %>% 
    rownames_to_column("bcode") %>%
    mutate(cluster_stage = case_when(seurat_clusters == .y[1] ~ "early_committed",
                                     seurat_clusters == .y[2] ~ "gexp02_late_committed",
                                     T ~ "non_committed")) 
}) %>% bind_rows(., .id = "dataset")  %>%
    mutate(dataset = str_remove(dataset, "^.")) 
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
# write the barcodes for the investigation of rbc content of the commited VS non-commited cells
saveRDS(v3_early_asex_bcodes_clustrs, "data/processed/Pf/m2_all/hs_commitment_signal/v3_early_asex_bcodes_clustrs.RDS")

```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3}
# write the barcodes for the investigation of rbc content of the commited VS non-commited cells
saveRDS(v3_early_asex_bcodes_clustrs, "data/processed/Pf/m2_all/hs_commitment_signal/v3_early_asex_bcodes_clustrs.RDS")

```


#############################END - CHECK IF THE HUMAN GENE PROFILE OF THE COMMITED PARASITES IS DIFFERENT FROM THAT OF THE NONCOMMITED ##########

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
map(v3_early_asex_jnd, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1222600", order = T))
# map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0406200", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1102500", order = T))
```

```{r, warning=FALSE, message=FALSE, fig.width=7, fig.height=3.5}
#SURF1.2 gexp02
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = c("PF3D7-0113600", "PF3D7-1102500"), order = T))
```


```{r, warning=FALSE, message=FALSE, eval=F}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_early_asex_jnd[[1]]@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= v3_early_asex_jnd[[1]]@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE}
cmtd_rings_de_gns_vs_all <- map(v3_early_asex_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor")) 
```

```{r, warning=FALSE, message=FALSE}
cmtd_rings_de_gns_vs_all_anot <- map(cmtd_rings_de_gns_vs_all, ~.x %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj)  %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

cmtd_rings_de_gns_vs_all_anot
```


```{r, warning=FALSE, message=FALSE}
## save for plotting
saveRDS(cmtd_rings_de_gns_vs_all_anot, "data/processed/Pf/m2_all/asex_integration_comtmnt_DE_figs/cmtd_rings_de_gns_vs_all_anot.RDS")
```


```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_gns_vs_all_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 5))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
clustr_gexp2_low_gns <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[1]] & p_val_adj < 0.05 & avg_log2FC > 0) %>% pull(gene))
clustr_gexp2_up_gns <- map2(cmtd_rings_de_gns_vs_all_anot, cmtd_rings_devlpmnt, ~.x %>% filter(cluster == .y[[2]] & p_val_adj < 0.05 & avg_log2FC > 0) %>% pull(gene))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
# imap(commited_uphigh_gns, ~josling_ap2g_DE_long %>%
imap(clustr_gexp2_low_gns, ~josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom"))


```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
# imap(commited_uphigh_gns, ~josling_ap2g_DE_long %>%
imap(clustr_gexp2_up_gns, ~josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom"))


```




```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
# Genes upregulated in gexp02 cluster that are in jcO but not in jcC
jcO_gns_missing_in_jcC <- clustr_gexp2_up_gns[[2]][!clustr_gexp2_up_gns[[2]] %in% clustr_gexp2_up_gns[[1]]]
jcO_gns_missing_in_jcC
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
josling_ap2g_DE_long %>%
  filter(gene_id %in% jcO_gns_missing_in_jcC) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom")


```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

imap(v3_early_asex_jnd, ~.x@meta.data %>% count(seurat_clusters, donor))

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}

map2(v3_early_asex_jnd, cmtd_rings_devlpmnt, ~.x@meta.data %>% count(seurat_clusters, donor, Strain_DN) %>% filter(seurat_clusters %in% .y))

```

```{r, warning=FALSE, message=FALSE, fig.width=12, fig.height=3.5}
# ETRAMP9, 1354900, Pfg14-744, 1309100
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = jcO_gns_missing_in_jcC, order = T, ncol = 4))
```



```{r, warning=FALSE, message=FALSE, eval=FALSE}
cmtd_rings_de_gns <- map2(v3_early_asex_jnd, cmtd_rings_devlpmnt, ~FindMarkers(.x, ident.1 = .y[1], ident.2 = .y[2], assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))
```

```{r, warning=FALSE, message=FALSE}
cmtd_rings_de_gns_anot <- map(cmtd_rings_de_gns, ~.x %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% rownames_to_column("gene") %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

cmtd_rings_de_gns_anot
```


```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_gns_anot, ~.x %>% 
  filter(avg_log2FC > 0) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20)) 
```

```{r, warning=FALSE, message=FALSE}
map(cmtd_rings_de_gns_anot, ~.x %>% 
  filter(avg_log2FC < 0) %>%
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05) %>%
    slice_min(order_by = p_val_adj, n = 20) )
```

```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
map(v3_early_asex, ~DimPlot(.x, label = T, reduction = "umap", group.by = "seurat_clusters"))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0113600", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1467600", order = T))
map(v3_early_asex, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1102500", order = T))
```


```{r, warning=FALSE, message=FALSE}
## Cluster comtd_lo and comtd_hi markers
# 
# commited_uphigh_DE_gns <- map2(commited_de_anot, cmtd_rings, ~.x %>%
#     filter(p_val_adj< 0.05 & avg_log2FC>0 & cluster == '1') %>%
#       pull(gene)
#     )
# 
# commited_uplow_DE_gns <- map2(commited_de_anot, cmtd_rings, ~.x %>%
#     filter(p_val_adj< 0.05 & avg_log2FC>0 & cluster == '0') %>%
#       pull(gene)
#     )


```

```{r, warning=FALSE, message=FALSE}
commited_top_cluster_DE_uphigh <- map(cmtd_rings_de_gns_anot, ~.x %>% 
  filter(avg_log2FC < 0 & p_val_adj< 0.05) %>% 
    pull(gene)
  )
commited_top_cluster_DE_uplow <- map(cmtd_rings_de_gns_anot, ~.x %>% 
  filter(avg_log2FC > 0 & p_val_adj< 0.05) %>% 
    pull(gene)
  )

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
# imap(commited_uplow_gns, ~josling_ap2g_DE_long %>%
imap(commited_top_cluster_DE_uplow, ~josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom"))


```

```{r, warning=FALSE, message=FALSE}
## Intersection between the markers of the cluster versus all cells and markers versus the other commited rings cluster
# commited_uphigh_gns <- map2(commited_uphigh_DE_gns, commited_top_cluster_DE_uphigh,  ~base::intersect(.x, .y ))
# commited_uphigh_gns
# commited_uplow_gns <- map2(commited_uplow_DE_gns, commited_top_cluster_DE_uplow,  ~base::intersect(.x, .y ))
# commited_uplow_gns
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
# imap(commited_uphigh_gns, ~josling_ap2g_DE_long %>%
imap(commited_top_cluster_DE_uphigh, ~josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom"))


```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
# imap(commited_uplow_gns, ~josling_ap2g_DE_long %>%
imap(commited_top_cluster_DE_uplow, ~josling_ap2g_DE_long %>%
  filter(gene_id %in% .x) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom"))


```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=4}
#LSD(PF3D7-1472200), HDA

josling_ap2g_DE_long %>%
  filter(gene_id %in% c("PF3D7-1472200")) %>%
  ggplot(aes(x=timepoint, y=expression, colour = ap2g_cond)) +
    geom_point() +
    # geom_line(aes(colour = ap2g_cond, group = Gene)) +
    # facet_grid(ap2g_cond ~ .) +
    geom_smooth(aes(group = ap2g_cond)) +
    theme(axis.text.x = element_text(angle=45),
          legend.position = "bottom")


```
###############################START ## DE between the merged 2 committed clusters and the rest of the clusters merged ######################
```{r, warning=FALSE, message=FALSE}
v3_early_asex_jnd <- map2(v3_early_asex_jnd, cmtd_rings_devlpmnt, ~.x@meta.data %>%
      mutate(merged_clusters = ifelse(seurat_clusters %in% .y, "committed", "non_committed")) %>%
       select(merged_clusters) %>%
       AddMetaData(.x, .))

```


```{r}
imap(v3_early_asex_jnd, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/v3_early_asex_jnd_", .y, "_3.RDS")))
```


```{r, warning=FALSE, message=FALSE}
v3_early_asex_jnd <- map(v3_early_asex_jnd, ~SetIdent(.x, value = "merged_clusters"))
```


```{r, warning=FALSE, message=FALSE, fig.width=5.3, fig.height=3.65, eval=T}
map(v3_early_asex_jnd, ~DimPlot(.x, label = T, reduction = "umap"))
```

```{r, warning=FALSE, message=FALSE}
commited_de_mrgd <- map(v3_early_asex_jnd, ~FindAllMarkers(.x, assay = "RNA", slot = "data", test.use = "MAST", latent.vars = "donor"))

```


```{r, warning=FALSE, message=FALSE}
commited_de_mrgd_anot <- map(commited_de_mrgd, ~.x %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id")))

# commited_de_mrgd_anot
```

```{r, warning=FALSE, message=FALSE}
# saveRDS(commited_de_mrgd_anot, "data/processed/Pf/m2_all/commited_de_mrgd_anot3.RDS")
```

```{r}
imap(commited_de_mrgd_anot, ~saveRDS(.x, paste0("data/processed/Pf/m2_all/commited_de_mrgd_anot_", .y, "_3.RDS")))

```

```{r, warning=FALSE, message=FALSE}
map(commited_de_mrgd_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% 
    filter(p_val_adj< 0.05 & avg_log2FC>0) %>%
    group_by(cluster) %>% 
    slice_min(order_by = p_val_adj, n = 10) %>% 
    slice_max(order_by = avg_log2FC, n = 10))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_mrgd_anot, ~.x %>% 
  # filter(gene %in% gns) %>% 
  arrange(desc(avg_log2FC)) %>% filter(p_val_adj< 0.05 & avg_log2FC>0) %>% group_by(cluster) %>% slice_min(order_by = p_val_adj, n = 10))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# HRPIII AROM RESA SURF1.2
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-1372200", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0206300", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0102200", order = T))
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0113600", order = T))
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# ap2-g pfs16 gexp02 gene_count
# map(v3_early_asex_jnd, ~FeaturePlot(.x@meta.data, reduction = "umap",  label = T, features = "PF3D7-1372200", order = T))

```

```{r, warning=FALSE, message=FALSE}
map(commited_de_mrgd_anot, ~.x %>%  filter(cluster %in% c("0") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE}
## LSD, HDA1
map(commited_de_mrgd_anot, ~.x %>%  filter(Gene %in% c("HDA1", "LSD2") ))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_mrgd_anot, ~.x %>%  filter(cluster %in% c("1") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_mrgd_anot, ~.x %>% filter(cluster %in% c("4") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE}
map(commited_de_mrgd_anot, ~.x %>%  filter(cluster %in% c("0") & avg_log2FC >0) %>% arrange(p_val_adj))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3.5}
# HRPIII AROM RESA SURF1.2
map(v3_early_asex_jnd, ~FeaturePlot(.x, reduction = "umap",  label = T, features = "PF3D7-0102200", order = T))
```

#### committed VS non-committed
```{r, fig.height=6, fig.width=6}
vlcano <- map(c("committed"), ~{
  clstr = .x
  map(commited_de_mrgd_anot, ~EnhancedVolcano(.x %>% filter(cluster == clstr),
                lab = .x %>% filter(cluster == clstr) %>% pull(Gene),
                # selectLab = .y,
                x = 'avg_log2FC',
                y = 'p_val_adj',
                pCutoff = 0.05,
                FCcutoff = 2,
                pointSize = 1.5,
                labSize = 4.0,
                title = '',
                subtitle = "",
                titleLabSize = 0,
                subtitleLabSize = 0,
                axisLabSize = 22,
                drawConnectors = TRUE,
                # boxedLabels = TRUE,
                # legendLabels =c( "p_val_adj"),
                gridlines.minor = FALSE
                ) +
    theme(legend.position = "none"))

}) 
vlcano

##Save the above image for plot grid in functional_GO_n_lit_comparison.RMD for manuscript
# saveRDS(fh_fl_vlcano, "plots/figS26/fh_fl_vlcano.RDS")
```


```{r, warning=FALSE, message=FALSE}
gns_list <- map(commited_top_cluster_DE_gns, ~tapply(.x$gene, .x$cluster, c))
```
###############################END ## DE between the merged 2 committed clusters and the rest of the clusters merged ######################

##########END# COMPARISON WITH AP2G ACTIVATED SEXUAL RING GENES ##############

############# START SCTtransform #######################################

```{r, warning=FALSE, message=FALSE}
# run sctransform
v3sex_sct <- SCTransform(v3sex_sct, verbose = FALSE, variable.features.n = 2000, vst.flavor = "v2")
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3sex_sct <- RunPCA(v3sex_sct, verbose = FALSE, npcs = 30)

```


```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

v3sex_sct_int <- IntegrateLayers(
  object = v3sex_sct,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 83,
)


```



```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3sex_sct_int <- RunUMAP(v3sex_sct_int, dims = 1:8, verbose = FALSE, n.components = 3, seed.use = 867)
v3sex_sct_int <- FindNeighbors(v3sex_sct_int, dims = 1:8, verbose = FALSE)
v3sex_sct_int <- FindClusters(v3sex_sct_int, verbose = FALSE, resolution = 0.2)
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(v3sex_sct_int, reduction = "umap", c(1,2))

```

```{r, warning=FALSE, message=FALSE}
# ap2-g pfs16 gexp02 gene_count
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-1222600", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-0406200", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-1102500", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-0406500", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "nFeature_RNA", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "nCount_RNA", order = T)


```

############# END SCTtransform #######################################



```{r}
# Idents(v3sex) <- "stagenew"
# ls <- subset(v3sex, idents = c("late stalk"))

# lsm <- as.data.frame(colSums(v3sex@assays$RNA$data[ap2g_gns_up %in% rownames(v3sex),]))
# 
# colnames(lsm) <- "lsm"
# lsm$rank <- c(1:nrow(lsm))
# mcells <- lsm[order(lsm$lsm, decreasing = TRUE),]
# mcells$rank <- c(1:nrow(mcells))
# mcellsx <- mcells[which(mcells$lsm >= mean(lsm$lsm)),] %>% rownames()
# fcellsx <- tail(rownames(mcells), length(mcellsx))
# 
# lsf <- as.data.frame(colSums(v3sex@assays$RNA$data[ap2g_gns_down %in% rownames(v3sex),]))
# 
# colnames(lsf) <- "lsf"
# lsf$rank <- c(1:nrow(lsf))
# fcells <- lsf[order(lsf$lsf, decreasing = TRUE),]
# fcells$rank <- c(1:nrow(fcells))
# fcellsy <- fcells[which(fcells$lsf >= mean(lsf$lsf)),] %>% rownames()
# mcellsy <- tail(rownames(fcells), length(fcellsy))
# 
# 
# base::intersect(mcellsx, fcellsy) %>% length()
# mcellsz <- base::setdiff(mcellsx,fcellsy)
# fcellsz <- base::setdiff(fcellsy,mcellsx)
```

```{r}
v3sex@meta.data$fm <- "other"
# v3sex@meta.data[which(v3sex@meta.data$stagenew == "late stalk"),]$fm <- "other late stak"
v3sex@meta.data[which(colnames(v3sex) %in% mcellsz),]$fm <- "male-like"
v3sex@meta.data[which(colnames(v3sex) %in% fcellsz),]$fm <- "female-like"

```


```{r}
defm <- FindMarkers(v3sex, ident.1 = mcellsz, ident.2 = fcellsz, only.pos = FALSE, test.use = "MAST")

# defm$geneid2 <- rownames(defm)

```

```{r, warning=FALSE, message=FALSE}
defm_cln_anot <- defm %>% rownames_to_column("gene") %>% filter(p_val_adj < 0.05) %>% arrange(p_val_adj) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

defm_cln_anot
```

```{r}
defm <- left_join(defm, pfdb56, by = "geneid2")
#defm <- defm[which(defm$p_val_adj < 0.05),]
defm <- defm %>% mutate(genename2 =
                     case_when(genename == 'N/A' ~ geneid2,
                     TRUE ~ genename))
defm$updown <- "other"
defm[which(defm$avg_log2FC > 0.4 & defm$p_val_adj < 0.05),]$updown <- "up"
defm[which(defm$avg_log2FC < -0.4 & defm$p_val_adj < 0.05),]$updown <- "down"
table(defm$updown)
  
```


############# END SK21 gene likely commited vs non-commited #############


############# START SCTtransform #######################################

```{r, warning=FALSE, message=FALSE}
## Add those without rescued cells to the object above created after merging
v3_m2_asex_cr_filt_no_resc <- v[non_resc_dons]
all_asex_merged_list <- c(v3_m2_asex_cr_filt_no_resc, all_asex_rna)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- merge(all_asex_merged_list[[1]], all_asex_merged_list[2:length(all_asex_merged_list)])
```

```{r, warning=FALSE, message=FALSE}
v3sex_sct <- v3sex
v3sex_sct[["SCT"]] <- NULL
v3sex_sct[["RNA"]]$data  <- NULL
v3sex_sct[["RNA"]]$scale.data  <- NULL
```

```{r, warning=FALSE, message=FALSE}
v3sex_sct[["RNA"]]$counts <- as(object = v3sex_sct[["RNA"]]$counts, Class = "dgCMatrix")

```

```{r, warning=FALSE, message=FALSE}
v3sex_sct[["RNA"]] <- split(v3sex_sct[["RNA"]], f = v3sex_sct$orig.ident)
v3sex_sct
```

```{r, warning=FALSE, message=FALSE}
# remove low n donors
table(v3sex_sct$orig.ident)
# v3sex_sct <- v3sex_sct[, !v3sex_sct$orig.ident %in% c("MSC48.mrna", "MSC67.mrna", "MSC14.mrna")]
v3sex_sct <- v3sex_sct[, !v3sex_sct$orig.ident %in% c("MSC48.mrna", "MSC67.mrna", "MSC14.mrna", "MSC55.mrna")]
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
v3sex_sct <- SCTransform(v3sex_sct, verbose = FALSE, variable.features.n = 2000, vst.flavor = "v2")
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3sex_sct <- RunPCA(v3sex_sct, verbose = FALSE, npcs = 30)

```


```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

v3sex_sct_int <- IntegrateLayers(
  object = v3sex_sct,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 83,
)


```



```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
v3sex_sct_int <- RunUMAP(v3sex_sct_int, dims = 1:8, verbose = FALSE, n.components = 3, seed.use = 867)
v3sex_sct_int <- FindNeighbors(v3sex_sct_int, dims = 1:8, verbose = FALSE)
v3sex_sct_int <- FindClusters(v3sex_sct_int, verbose = FALSE, resolution = 0.2)
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(v3sex_sct_int, reduction = "umap", c(1,2))

```

```{r, warning=FALSE, message=FALSE}
# ap2-g pfs16 gexp02 gene_count
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-1222600", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-0406200", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-1102500", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "PF3D7-0406500", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "nFeature_RNA", order = T)
FeaturePlot(v3sex_sct_int, reduction = "umap",  label = T, features = "nCount_RNA", order = T)


```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3sex_sct_int@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= v3sex_sct_int@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r}
v3sex_sct_int <- PrepSCTFindMarkers(v3sex_sct_int)

```


```{r}
v3sex_sct_de <- FindAllMarkers(v3sex_sct_int, assay = "SCT", verbose = FALSE, test.use = "MAST")

```


```{r, warning=FALSE, message=FALSE}
v3sex_sct_de_anot <- v3sex_sct_de %>% dplyr::filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

v3sex_sct_de_anot
```


```{r, warning=FALSE, message=FALSE}
v3sex_sct_de_anot %>% dplyr::filter(cluster == "0" & avg_log2FC >0)
v3sex_sct_de_anot %>% dplyr::filter(cluster == "1" & avg_log2FC >0)
v3sex_sct_de_anot %>% dplyr::filter(cluster == "2" & avg_log2FC >0)
v3sex_sct_de_anot %>% dplyr::filter(cluster == "3" & avg_log2FC >0)
```

```{r}

head(v3sex_sct_de, n = 15)


```

```{r}

gogo()

```
############# END SCTtransform #######################################

############# START Average expression of genes #######################



```{r}

## Get the genes expressed in atleast 25% of cells in fhe,fle,mhe and mle

cb_asex_gn_exp <- AggregateExpression(v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs[, v3_m2_asex_cr_filt_norm_harmony_jnd_fine_clustrs$harmony_pc_man_clusters %in% c("13", "5", "12", "2")], group.by = "donor", assays = "RNA", return.seurat = T) 

cb_asex_gn_exp_df <- cb_asex_gn_exp@assays$RNA$data

cb_asex_gn_exp_df

```


```{r, fig.width=10, fig.height=10}

## correlation between genes by stage
cb_asex_gn_exp_df %>%
  data.frame() %>%
  # column_to_rownames('gene') %>% 
  select(where(~!all(is.na(.x)))) %>%
  ggpairs(lower = list(continuous = GGally::wrap(lowerfun)),
          upper = list(continuous = wrap("cor", size = 5)),
          progress = FALSE)

```



```{r, warning=FALSE, message=FALSE}

asex_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, top_asex_m14, assay = 'RNA') %>%  data.frame %>%  setNames('asx_avg_cln')  )

mle_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, top_mLE_m13_m14, assay = 'RNA') %>%  data.frame %>%  setNames('mle_avg_cln')  )

fle_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, top_fLE_m13_m14, assay = 'RNA') %>%  data.frame %>%  setNames('fle_avg_cln')  )

asex2_avg_exp_cln <- list(all_resc_sct_comp_m21_m22)  %>%
  pmap(~makeSetScore(..1, asex_2_m70, assay = 'RNA') %>%  data.frame %>%  setNames('asx2_avg_cln')  )


markers_avg_exp_cln <- pmap(list(asex_avg_exp_cln, mle_avg_exp_cln, fle_avg_exp_cln, asex2_avg_exp_cln), ~ bind_cols(..1, ..2, ..3, ..4))
# markers_avg_exp[[1]]

```


```{r, warning=FALSE, message=FALSE}

all_resc_sct_comp_m21_m22 <- list(all_resc_sct_comp_m21_m22, markers_avg_exp_cln) %>% pmap(~AddMetaData(..1, ..2))
```

Visualization of clusters with markers

```{r, warning=FALSE, message=FALSE, eval=T}
## cells that contain 2 or more ap2g UMI
imap(all_resc_sct_comp_m21_m22, ~DimPlot(.x, label = T)+ labs(title = .y))

```

```{r, warning=FALSE, message=FALSE, fig.width=13, fig.height=4}
imap(all_resc_sct_comp_m21_m22, ~FeaturePlot(.x, features = c("asx_avg_cln", "mle_avg_cln", "fle_avg_cln", "asx2_avg_cln"),  ncol = 4) + plot_annotation(title = .y))

```

####################### END Average expression of genes #######################


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_cr_filt_norm_harmony@reductions[["umap.harmony_PC_man"]]@cell.embeddings),
        x = ~umapharmonypcman_1, y = ~umapharmonypcman_2, z = ~umapharmonypcman_3,
        color= v3_m2_asex_cr_filt_sct_int@meta.data$sR_fine_stg,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_cr_filt_norm_harmony@reductions[["integrated.harmony"]]@cell.embeddings),
        x = ~harmony_1, y = ~harmony_2, z = ~harmony_3,
        color= v3_m2_asex_cr_filt_sct_int@meta.data$sR_fine_stg,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 0.1, 
        type = 'scatter3d'
)

```



################## START DE findmarkers


```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(v3_m2_asex_cr_filt_norm_harmony) <- "RNA"
v3_m2_asex_cr_filt_norm_harmony <- JoinLayers(v3_m2_asex_cr_filt_norm_harmony)

```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony[["RNA"]]$counts <- as(object = v3_m2_asex_cr_filt_norm_harmony[["RNA"]]$counts, Class = "dgCMatrix")
v3_m2_asex_cr_filt_norm_harmony[["RNA"]]$data <- as(object = v3_m2_asex_cr_filt_norm_harmony[["RNA"]]$data, Class = "dgCMatrix")
```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony[["RNA"]]["counts"]
```


```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_norm_harmony, reduction = "umap.harmony_PC_man", group.by = "harmony_pc_man_clusters", order = T, pt.size = 0.01, label = T) 

```

```{r, warning=FALSE, message=FALSE, fig.width=14, fig.height=4}
VlnPlot(v3_m2_asex_cr_filt_norm_harmony, features = c("nFeature_RNA", "nCount_RNA"), group.by = "harmony_pc_man_clusters",pt.size = 0.001)

```

```{r, warning=FALSE, message=FALSE}
## DE
Idents(v3_m2_asex_cr_filt_norm_harmony) <- "harmony_pc_man_clusters"
de_gns_fmall_cln <- FindAllMarkers(v3_m2_asex_cr_filt_norm_harmony)
# de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% slice_head(n=5)
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_cln_anot %>% filter(avg_log2FC > 0, cluster == .x))

```



```{r, warning=FALSE, message=FALSE}
## Dormant (d5MDPs) genes after artemisinin - https://www-nature-com.ezp.lib.cam.ac.uk/articles/s41467-024-51846-0
d5MDP_up_gns <- c("PF3D7_1430900", "PF3D7_0928000", "PF3D7_0217100", "PF3D7_0412100", "PF3D7_1341100") %>% str_replace(., "_", "-")

d5MDP_down_gns<- c("PF3D7_0202500", "PF3D7_1102800", "PF3D7_1401400", "PF3D7_0202000") %>% str_replace(., "_", "-")


```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_up_gns)
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_down_gns)
```


Test cold temperature activation
```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

# Cluster 4 highly expressess truRNA which is switched on when temperature is lowered from 37C to 26C activates!!! https://www.sciencedirect.com/science/article/pii/S0378111923003578
# https://pmc.ncbi.nlm.nih.gov/articles/PMC11597781/

```



################## END DE findmarkers


Slingshot
```{r, warning=FALSE, message=FALSE}
DefaultAssay(v3_m2_asex_cr_filt_norm_harmony) <- "RNA"
```

```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_norm_harmony, reduction = "umap.harmony_PC_man", group.by = "harmony_pc_man_clusters", order = T, pt.size = 0.01) 

```

```{r, warning=FALSE, message=FALSE}
v3_m2_asex_cr_filt_norm_harmony <- FindClusters(v3_m2_asex_cr_filt_norm_harmony, resolution = 0.2, cluster.name = "harmony_clusters_ss") 
```


```{r, warning=FALSE, message=FALSE}
DimPlot(v3_m2_asex_cr_filt_norm_harmony, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_ss", order = T, pt.size = 0.01, label=T) 

```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(v3_m2_asex_cr_filt_norm_harmony@reductions[["umap.harmony_PC_man"]]@cell.embeddings),
        x = ~umapharmony_1, y = ~umapharmony_2, z = ~umapharmony_3,
        color= v3_m2_asex_cr_filt_norm_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, eval=T}
# gogo()
v3_m2_asex_cr_filt_norm_harmony_sce <- as.SingleCellExperiment(v3_m2_asex_cr_filt_norm_harmony)

```

```{r, eval=T}
rowData(v3_m2_asex_cr_filt_norm_harmony_sce)$feature_symbol <- rownames(v3_m2_asex_cr_filt_norm_harmony_sce)
```

!!!! NOTE -BEEN DOING SLINGSHOT WRONG. LOGCOUNTS SHOULD HAVE data values from seurat
```{r, eval=T}
# gogo()
# Seurat_object <- m2_asex_harmony_res

## Convert seurat object to SCE
v3_m2_asex_cr_filt_norm_harmony_sce_ss <- slingshot(v3_m2_asex_cr_filt_norm_harmony_sce,
            clusterLabels = "harmony_clusters_ss",
            reducedDim = "UMAP.HARMONY_PC_MAN",
            start.clus = 1,
            end.clus = 3,
            # shrink =1,
            # stretch = 1,
            # allow.break = F
            )


```


```{r}
##Prepare slingshot dataset (SDS) object
v3_m2_asex_cr_filt_norm_harmony_sce_ss_sds <- SlingshotDataSet(v3_m2_asex_cr_filt_norm_harmony_sce_ss)

# plot(reducedDims(v3_m2_asex_cr_filt_norm_harmony_sce_ss)$UMAP[,1:2], col = brewer.pal(9,"Set1")[as.factor(v3_m2_asex_cr_filt_norm_harmony_sce_ss$stage)], pch=16, asp = 0.7)
plot(reducedDims(v3_m2_asex_cr_filt_norm_harmony_sce_ss)$UMAP.HARMONY_PC_MAN[,c(1,2)], col = brewer.pal(9,"Set1")[base::as.factor(v3_m2_asex_cr_filt_norm_harmony_sce_ss$harmony_clusters_ss)], pch=16, asp = 0.7, )

lines(v3_m2_asex_cr_filt_norm_harmony_sce_ss_sds, lwd=2, col="black",dims = c(1,2))
```


```{r}
## Add pseudotime metadata to the object
v3_m2_asex_cr_filt_norm_harmony_pt <- AddMetaData(v3_m2_asex_cr_filt_norm_harmony, 
                                                      data.frame(v3_m2_asex_cr_filt_norm_harmony_sce_ss@colData[, c("slingPseudotime_1"), drop = F]) #%>%
                                                        # rename("pseudotime" = slingPseudotime_1)
                                                      ) 

```


```{r}
FeaturePlot(v3_m2_asex_cr_filt_norm_harmony_pt, features = "slingPseudotime_1", reduction = "umap.harmony_PC_man")

```



```{r, eval = FALSE}
## Save pseudotime objects
saveRDS(v3_m2_asex_cr_filt_norm_harmony_pt, "data/processed/Pf/v3_m2_asex_cr_filt_norm_harmony_pt3.RDS")
saveRDS(v3_m2_asex_cr_filt_norm_harmony_sce_ss, "data/processed/Pf/v3_m2_asex_cr_filt_norm_harmony_sce_ss3.RDS")
saveRDS(v3_m2_asex_cr_filt_norm_harmony_sce_ss_sds, "data/processed/Pf/v3_m2_asex_cr_filt_norm_harmony_sce_ss_sds3.RDS")

### GOTO m22_raw_tseq_MAST_intradon_DE for tradeseq
```

######################## START Test age between asympto and sympto ########################
```{r, fig.width=15, fig.height=5}
donor_mdata_slct <- readRDS( "data/processed/Pf/m2_all/donor_mdata_slct.RDS")

```


```{r}
## Save pseudotime objects
v3_m2_asex_cr_filt_norm_harmony_pt <- readRDS("data/processed/Pf/v3_m2_asex_cr_filt_norm_harmony_pt.RDS")

```


```{r, fig.width=15, fig.height=5}

v3_m2_asex_cr_filt_norm_harmony_pt_mdata <-  v3_m2_asex_cr_filt_norm_harmony_pt[, !v3_m2_asex_cr_filt_norm_harmony_pt$orig.ident == "MSC55.mrna"] 
# v3_m2_asex_cr_filt_norm_harmony_pt_mdata <-  v3_m2_asex_cr_filt_norm_harmony_pt

v3_m2_asex_cr_filt_norm_harmony_pt_mdata <-  v3_m2_asex_cr_filt_norm_harmony_pt_mdata@meta.data %>%
  mutate(short_name = str_remove(donor, "_SLO|_MACS")) %>%
  left_join(., donor_mdata_slct, by = c("short_name" = "short_name")) %>%
  select(short_name, sympto_groups, symptomatic, condition, fever, temperature_c) %>% 
  AddMetaData(v3_m2_asex_cr_filt_norm_harmony_pt_mdata, .)
```


```{r, fig.width=15, fig.height=5}
DimPlot(v3_m2_asex_cr_filt_norm_harmony_pt_mdata, split.by = "sympto_groups")
```


```{r}
v3_m2_asex_cr_filt_norm_harmony_pt_mdata@meta.data %>%
  ggplot(aes(x = sympto_groups, y = slingPseudotime_1)) +
  geom_boxplot() 
```


```{r, fig.width=16, fig.height=3}
v3_m2_asex_cr_filt_norm_harmony_pt_mdata@meta.data %>%
  filter(!Strain_DN %in% c("Doublet", "Negative")) %>%
  ggplot(aes(x = donor, y = slingPseudotime_1, colour = Strain_DN)) +
  geom_boxplot() +
  facet_nested(. ~ sympto_groups + donor, space = "free_x", scales = "free_x")
```

```{r, fig.width=16, fig.height=3}
v3_m2_asex_cr_filt_norm_harmony_pt_mdata@meta.data %>%
  filter(!Strain_DN %in% c("Doublet", "Negative")) %>%
  ggplot(aes(x = donor, y = slingPseudotime_1, colour = Strain_DN)) +
  geom_sina(size = 0.001) +
  facet_nested(. ~ sympto_groups + donor, space = "free_x", scales = "free_x")
```


```{r, fig.width=16, fig.height=3}
gogo()
```

######################## END Test age between asympto and sympto ########################



######################## START integrate cellranger filtered with rescued asexuals ########################




```{r, warning=FALSE, message=FALSE, eval=F}
# run sctransform
all_resc_asex_rna <-  all_resc_asex_sct

```

```{r, warning=FALSE, message=FALSE, eval=F}
## change default assay to RNA for getting cells with ap2g reads
for (i in 1:length(all_resc_asex_rna)) {
  DefaultAssay(all_resc_asex_rna[[i]]) <- "RNA"
  all_resc_asex_rna[[i]]@assays$SCT <- NULL
  all_resc_asex_rna[[i]]@project.name <- "SeuratProject"
}

all_resc_asex_rna
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
v3_m2_asex_cr_filt_list <-  SplitObject(v3_m2_asex_cr_filt, split.by = "donor")
```


```{r, warning=FALSE, message=FALSE}
## Donors with cells in the rescued object
resc_dons <- names(v3_m2_asex_cr_filt_list)[names(v3_m2_asex_cr_filt_list) %in% names(all_resc_asex_rna)]
non_resc_dons <- names(v3_m2_asex_cr_filt_list)[!names(v3_m2_asex_cr_filt_list) %in% names(all_resc_asex_rna)]

```

```{r, warning=FALSE, message=FALSE}
## Donors with cells in the rescued object
all_asex_rna <- list()
```


```{r, warning=FALSE, message=FALSE}
## for those with rescued cells merge these with cell ranger qcd cells
for (i in resc_dons) {
  all_asex_rna[i] <- merge(v3_m2_asex_cr_filt_list[[i]], all_resc_asex_rna[[i]])
  
}

all_asex_rna
```

```{r, warning=FALSE, message=FALSE}
## Add those without rescued cells to the object above created after merging
v3_m2_asex_cr_filt_no_resc <- v3_m2_asex_cr_filt_list[non_resc_dons]
all_asex_merged_list <- c(v3_m2_asex_cr_filt_no_resc, all_asex_rna)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- merge(all_asex_merged_list[[1]], all_asex_merged_list[2:length(all_asex_merged_list)])
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged <- JoinLayers(all_asex_merged)
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged[["RNA"]] <- split(all_asex_merged[["RNA"]], f = all_asex_merged$donor)
all_asex_merged
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
all_asex_merged <- SCTransform(all_asex_merged, verbose = FALSE, variable.features.n = 200)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged <- RunPCA(all_asex_merged, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int <- IntegrateLayers(
  object = all_asex_merged,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int <- FindNeighbors(all_asex_merged_int, dims = 1:6, reduction = "integrated.dr")
all_asex_merged_int <- FindClusters(all_asex_merged_int, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int <- RunUMAP(all_asex_merged_int, dims = 1:6, verbose = FALSE, n.components = 3, seed.use = 867)
all_asex_merged_int <- FindNeighbors(all_asex_merged_int, dims = 1:6, verbose = FALSE)
all_asex_merged_int <- FindClusters(all_asex_merged_int, verbose = FALSE, resolution = 0.1)
```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_asex_merged_int, "data/processed/Pf/m2_all/all_asex_merged_int.RDS")
```

```{r, warning=FALSE, message=FALSE}
# all_asex_merged_int <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int.RDS")
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, reduction = "pca")
DimPlot(all_asex_merged_int, label = T)
DimPlot(all_asex_merged_int, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int, group.by = "donor")
FeaturePlot(all_asex_merged_int, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int, features = "nCount_RNA")
VlnPlot(all_asex_merged_int, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(2,3))
DimPlot(all_asex_merged_int, label = T, dims = c(2,3))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, dims = c(1,3))

```


```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T, group.by = "stage", dims = c(1,3))
DimPlot(all_asex_merged_int, label = T, dims = c(1,3))

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int) <- "RNA"
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int) <- "SCT"
```



```{r, warning=FALSE, message=FALSE}
## ap2g
all_asex_merged_int_ap2g <- subset(all_asex_merged_int, subset = `PF3D7-1222600` > 0, slot = 'counts')

table(all_asex_merged_int_ap2g$donor)
```

```{r, warning=FALSE, message=FALSE}
# ## change ident to ap2g
# for (i in 1:length(all_asex_merged_int)) {
#   Idents(all_asex_merged_int[[i]]) <- "ap2_clst_finer"
# }

```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int) <- "RNA"
all_asex_merged_int <- JoinLayers(all_asex_merged_int)

```



```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int)
de_gns_fmall_all <- de_gns_fmall_all %>% arrange(p_val_adj)

```


```{r, warning=FALSE, message=FALSE}
de_gns_sig_anot <- de_gns_fmall_all %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_anot
```

```{r, warning=FALSE, message=FALSE}
VlnPlot(all_asex_merged_int, features = c("nFeature_RNA", "nCount_RNA"))

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int, label = T)
```




```{r, warning=FALSE, message=FALSE, eval=TRUE}
all_asex_merged_int@meta.data %>% filter(seurat_clusters == 4) %>% count(donor, seurat_clusters)

```


```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

# Cluster 4 highly expressess truRNA which is switched on when temperature is lowered from 37C to 26C activates!!! https://www.sciencedirect.com/science/article/pii/S0378111923003578
# https://pmc.ncbi.nlm.nih.gov/articles/PMC11597781/

```

```{r, warning=FALSE, message=FALSE}
gogo()
```

############ SCT2 -without activated asexuals
Remove the low temperature activate asexuals and cluster the rest
```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv <- all_asex_merged_int[, all_asex_merged_int$seurat_clusters != 4]
DimPlot(all_asex_merged_int_no_actv, label = T)
```


```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv[["RNA"]] <- split(all_asex_merged_int_no_actv[["RNA"]], f = all_asex_merged_int_no_actv$donor)
all_asex_merged_int_no_actv
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
all_asex_merged_int_no_actv <- SCTransform(all_asex_merged_int_no_actv, verbose = FALSE, variable.features.n = 100)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv <- RunPCA(all_asex_merged_int_no_actv, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged_int_no_actv, reduction = "pca")

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int_no_actv <- IntegrateLayers(
  object = all_asex_merged_int_no_actv,
  method = RPCAIntegration,
  normalization.method = "SCT",
  verbose = F,
  k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv <- FindNeighbors(all_asex_merged_int_no_actv, dims = 1:5, reduction = "integrated.dr")
all_asex_merged_int_no_actv <- FindClusters(all_asex_merged_int_no_actv, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv <- RunUMAP(all_asex_merged_int_no_actv, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867)
all_asex_merged_int_no_actv <- FindNeighbors(all_asex_merged_int_no_actv, dims = 1:5, verbose = FALSE)
all_asex_merged_int_no_actv <- FindClusters(all_asex_merged_int_no_actv, verbose = FALSE, resolution = 0.1)
```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_asex_merged_int_no_actv, "data/processed/Pf/m2_all/all_asex_merged_int_no_actv.RDS")
```

```{r, warning=FALSE, message=FALSE}
# all_asex_merged_int_no_actv <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int_no_actv.RDS")
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv, label = T, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv, label = T)
DimPlot(all_asex_merged_int_no_actv, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int_no_actv, group.by = "donor")
FeaturePlot(all_asex_merged_int_no_actv, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int_no_actv, features = "nCount_RNA")
VlnPlot(all_asex_merged_int_no_actv, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv) <- "RNA"
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int_no_actv) <- "SCT"
```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv) <- "RNA"
all_asex_merged_int_no_actv <- JoinLayers(all_asex_merged_int_no_actv)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_cln <- FindAllMarkers(all_asex_merged_int_no_actv)
de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv, label = T)
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

```

```{r, warning=FALSE, message=FALSE}
## Dormant (d5MDPs) genes after artemisinin - https://www-nature-com.ezp.lib.cam.ac.uk/articles/s41467-024-51846-0
d5MDP_up_gns <- c("PF3D7_1430900", "PF3D7_0928000", "PF3D7_0217100", "PF3D7_0412100", "PF3D7_1341100") %>% str_replace(., "_", "-")

d5MDP_down_gns<- c("PF3D7_0202500", "PF3D7_1102800", "PF3D7_1401400", "PF3D7_0202000") %>% str_replace(., "_", "-")


```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_up_gns)
de_gns_sig_cln_anot %>% filter(gene %in% d5MDP_down_gns)
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE}
VlnPlot(all_asex_merged_int, features = c("nFeature_RNA", "nCount_RNA"))

```

############ Harmony -without activated asexuals
Remove the low temperature activate asexuals and cluster the rest
```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony <- all_asex_merged_int[, all_asex_merged_int$seurat_clusters != 4]
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
```


```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony[["RNA"]] <- split(all_asex_merged_int_no_actv_harmony[["RNA"]], f = all_asex_merged_int_no_actv_harmony$donor)
all_asex_merged_int_no_actv_harmony
```

```{r, warning=FALSE, message=FALSE}
# run sctransform
# all_asex_merged_int_no_actv_harmony <- SCTransform(all_asex_merged_int_no_actv_harmony, verbose = FALSE, variable.features.n = 100)

all_asex_merged_int_no_actv_harmony <- NormalizeData(all_asex_merged_int_no_actv_harmony)
all_asex_merged_int_no_actv_harmony <- FindVariableFeatures(all_asex_merged_int_no_actv_harmony, selection.method = "vst", nfeatures = 300)
all_asex_merged_int_no_actv_harmony <- ScaleData(all_asex_merged_int_no_actv_harmony, verbose = F)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv_harmony <- RunPCA(all_asex_merged_int_no_actv_harmony, verbose = FALSE, npcs = 20)

```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, reduction = "pca")
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.5)
```


```{r, warning=FALSE, message=FALSE}
# run sctransform
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "pca")
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "pca", group.by = "stage")


```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv@reductions[["pca"]]@cell.embeddings),
        x = ~PC_1, y = ~PC_2, z = ~PC_3,
        color= all_asex_merged_int_no_actv@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
options(future.globals.maxSize = 1e+09)

all_asex_merged_int_no_actv_harmony <- IntegrateLayers(
  object = all_asex_merged_int_no_actv_harmony,
  method = HarmonyIntegration,
  orig.reduction = "pca", 
  new.reduction = "harmony",
  # normalization.method = "SCT",
  verbose = F,
  # k.weight = 98,
)


```

```{r, warning=FALSE, message=FALSE}

all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, reduction = "harmony")
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.5)
```

```{r, warning=FALSE, message=FALSE}
# These are now standard steps in the Seurat workflow for visualization and clustering
all_asex_merged_int_no_actv_harmony <- RunUMAP(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, n.components = 3, seed.use = 867, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindNeighbors(all_asex_merged_int_no_actv_harmony, dims = 1:5, verbose = FALSE, reduction = "harmony")
# all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, verbose = FALSE, resolution = 0.1)
```


```{r, warning=FALSE, message=FALSE}
saveRDS(all_asex_merged_int_no_actv_harmony, "data/processed/Pf/m2_all/all_asex_merged_int_no_actv_harmony.RDS")
```

```{r, warning=FALSE, message=FALSE}
# all_asex_merged_int_no_actv_harmony <- readRDS("data/processed/Pf/m2_all/all_asex_merged_int_no_actv_harmony.RDS")
```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, label = T, reduction = "harmony")
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
DimPlot(all_asex_merged_int_no_actv_harmony, label = T, group.by = "stage") + scale_color_manual(values = sp_fld_colors)
DimPlot(all_asex_merged_int_no_actv_harmony, group.by = "donor")
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "nFeature_RNA")
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "nCount_RNA")
VlnPlot(all_asex_merged_int_no_actv_harmony, features = "nCount_RNA", group.by = "seurat_clusters")

```

```{r, warning=FALSE, message=FALSE}
## ap2g
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "pca", order = T, pt.size = 0.5) 
FeaturePlot(all_asex_merged_int_no_actv_harmony, features = "PF3D7-1222600", reduction = "harmony", order = T, pt.size = 0.5) 
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "SCT"
```
```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "harmony", order = T, pt.size = 0.5) 

```



```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["umap"]]@cell.embeddings),
        x = ~umap_1, y = ~umap_2, z = ~umap_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["harmony"]]@cell.embeddings),
        x = ~harmony_1, y = ~harmony_2, z = ~harmony_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```

```{r, warning=FALSE, message=FALSE}
## ap2g
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
all_asex_merged_int_no_actv_harmony <- JoinLayers(all_asex_merged_int_no_actv_harmony)

```


```{r, warning=FALSE, message=FALSE}
## ap2g
# de_gns_fmall_all <- FindAllMarkers(all_asex_merged_int, test.use = "MAST")
de_gns_fmall_cln <- FindAllMarkers(all_asex_merged_int_no_actv_harmony)
de_gns_fmall_cln <- de_gns_fmall_all %>% arrange(p_val_adj)

```

```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, label = T)
```

```{r, warning=FALSE, message=FALSE}
de_gns_sig_cln_anot <- de_gns_fmall_cln %>% filter(p_val_adj < 0.05) %>% group_by(cluster) %>% arrange(p_val_adj, .by_group = T) %>% left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")], by = c("gene" = "gene_id"))

de_gns_sig_cln_anot
```

```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC > 0, cluster == .x))

```


```{r, warning=FALSE, message=FALSE, eval=TRUE}
map(unique(de_gns_sig_cln_anot$cluster), ~de_gns_sig_anot %>% filter(avg_log2FC < 0, cluster == .x))

```

Slingshot
```{r, warning=FALSE, message=FALSE}
DefaultAssay(all_asex_merged_int_no_actv_harmony) <- "RNA"
```

```{r, warning=FALSE, message=FALSE}
all_asex_merged_int_no_actv_harmony <- FindClusters(all_asex_merged_int_no_actv_harmony, resolution = 0.05, cluster.name = "harmony_clusters_ss") 
```


```{r, warning=FALSE, message=FALSE}
DimPlot(all_asex_merged_int_no_actv_harmony, reduction = "harmony", group.by = "harmony_clusters_ss", order = T, pt.size = 0.5) 

```


```{r, warning=FALSE, message=FALSE, eval=T}
# Display the combined datasets as an interactive 3D plot

plot_ly(data = data.frame(all_asex_merged_int_no_actv_harmony@reductions[["harmony"]]@cell.embeddings),
        x = ~harmony_1, y = ~harmony_2, z = ~harmony_3,
        color= all_asex_merged_int_no_actv_harmony@meta.data$seurat_clusters,
        # colors = c('blue', 'red'),
        opacity = 0.8,
        size = 1, 
        type = 'scatter3d'
)

```


```{r, eval=T}
# gogo()
all_asex_merged_int_no_actv_harmony_sce <- as.SingleCellExperiment(all_asex_merged_int_no_actv_harmony)

```

```{r, eval=T}
rowData(all_asex_merged_int_no_actv_harmony_sce)$feature_symbol <- rownames(all_asex_merged_int_no_actv_harmony_sce)
```

!!!! NOTE -BEEN DOING SLINGSHOT WRONG. LOGCOUNTS SHOULD HAVE data values from seurat
```{r, eval=T}
# gogo()
# Seurat_object <- m2_asex_harmony_res

## Convert seurat object to SCE
all_asex_merged_int_no_actv_harmony_sce_ss <- slingshot(all_asex_merged_int_no_actv_harmony_sce,
            clusterLabels = "harmony_clusters_ss",
            reducedDim = "HARMONY",
            start.clus = 1,
            end.clus = 0,
            # shrink =1,
            # stretch = 1,
            # allow.break = F
            )


```


```{r}
##Prepare slingshot dataset (SDS) object
all_asex_merged_int_no_actv_harmony_sce_ss_sds <- SlingshotDataSet(all_asex_merged_int_no_actv_harmony_sce_ss)

# plot(reducedDims(all_asex_merged_int_no_actv_harmony_sce_ss)$UMAP[,1:2], col = brewer.pal(9,"Set1")[as.factor(all_asex_merged_int_no_actv_harmony_sce_ss$stage)], pch=16, asp = 0.7)
plot(reducedDims(all_asex_merged_int_no_actv_harmony_sce_ss)$HARMONY[,c(1,2)], col = brewer.pal(9,"Set1")[as.factor(all_asex_merged_int_no_actv_harmony_sce_ss$seurat_clusters)], pch=16, asp = 0.7, )

lines(all_asex_merged_int_no_actv_harmony_sce_ss_sds, lwd=2, col="black",dims = c(1,2))
```


```{r}
## Add pseudotime metadata to the object
all_asex_merged_int_no_actv_harmony_pt <- AddMetaData(all_asex_merged_int_no_actv_harmony, 
                                                      data.frame(all_asex_merged_int_no_actv_harmony_sce_ss@colData[, c("slingPseudotime_1"), drop = F]) #%>%
                                                        # rename("pseudotime" = slingPseudotime_1)
                                                      ) 

```

```{r}
## Save pseudotime objects
saveRDS(all_asex_merged_int_no_actv_harmony_pt, "data/processed/Pf/all_asex_merged_int_no_actv_harmony_pt.RDS")
saveRDS(all_asex_merged_int_no_actv_harmony_sce_ss, "data/processed/Pf/all_asex_merged_int_no_actv_harmony_sce_ss.RDS")
saveRDS(all_asex_merged_int_no_actv_harmony_sce_ss_sds, "data/processed/Pf/all_asex_merged_int_no_actv_harmony_sce_ss_sds.RDS")

```

