---
title: "correlation2bulk_dsets"
output: html_document
date: "2024-07-18"
---
---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(BiocParallel)
  library(ComplexHeatmap)
  library(readxl)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  conflicts_prefer(base::intersect)
  conflicts_prefer(base::setdiff)
  conflicts_prefer(base::union)
  
})
```


```{r}
sample_set = "Mali2"

```

#  integration of all MSC


```{r setup}
# knitr::opts_knit$set(root.dir = paste0("/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/", sample_set))
knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))
```



### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk//plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
```


Male and female fertility genes https://www.cell.com/cell-systems/fulltext/S2405-4712(24)00305-3
```{r}
sayers_Pb_fertility = readxl::read_xlsx("../published_dsets/sayers_PbFertilityGenes_2024/fertility_genes_screen.xlsx", sheet = "(A) Fertility screen results", .name_repair = "universal", na = c("", "NA")) %>% rename_all(., ~str_remove(., "\\.\\."))
```

```{r}
sayers_Pb_fertility_cln <- sayers_Pb_fertility %>% select(Pberghei.ANKA.gene.ID, Pfalciparum.3D7.orthologue..ID, Female.fertilitylog2FC...input.sample.log2.barcode.relative.abundance.is.subtracted.from.output.sample.log2.barcode.relative.abundance.and.then.normalised.to.WT.controls., Female.p.valuelikelihood.that.fertility..log2FC..occurs.by.chance.where.a.p.value..0.05.indicates.statistical.significance., Female.fertility.phenotypemutants.with.a.fertility..log2FC..plus.2xSD.of.less.than..2.are.called.Reduced.and.greater.than..2.are.called.Not.reduced., Male.fertilitylog2FC...input.sample.log2.barcode.relative.abundance.is.subtracted.from.output.sample.log2.barcode.relative.abundance.then.normalised.to.WT.controls., Male.p.valuelikelihood.that.fertility..log2FC..occurs.by.chance.where.a.p.value..0.05.indicates.statistical.significance., Male.fertility.phenotypemutants.with.a.fertility..log2FC..plus.2xSD.of.less.than..2.are.called.Reduced.and.greater.than..2.are.called.Not.reduced., contains("gametocyte.log2FCPMID..36634679"), contains("phenotypePMID..36634679")) %>% rename_all(., ~str_remove_all(., "...input.sample.*|likelihood..*|mutants.*|PMID.*")) %>% mutate(gene_id = str_replace(Pfalciparum.3D7.orthologue..ID, "_", "-"))
```

```{r}
all_mrks <- readRDS("data/processed/Pf/m2_all/all_mrks.RDS")
```


```{r}
all_mrks %>% str()
```


```{r}
## https://www.cell.com/cell-systems/fulltext/S2405-4712(24)00305-3
all_mrks %>% bind_rows(., .id = "donor") %>% left_join(., Pf3D7_genes_plasmoDB[, c("Gene", "gene_id")], by = c("gene"="gene_id")) %>% group_by(donor, cluster) %>% arrange(p_val_adj, .by_group = T) %>% dplyr::filter(cluster == "Female LE") %>% group_by(gene)  %>% add_count(gene, name = "gene_cnt" ) %>% arrange(desc(gene_cnt), gene) %>% filter(avg_log2FC > 0 & p_val_adj < 0.05)
```


```{r}
## https://www.cell.com/cell-systems/fulltext/S2405-4712(24)00305-3
le_gns <- all_mrks %>% bind_rows(., .id = "donor") %>% left_join(., Pf3D7_genes_plasmoDB[, c("Gene", "gene_id")], by = c("gene"="gene_id")) %>% group_by(donor, cluster) %>% arrange(p_val_adj, .by_group = T) %>% dplyr::filter(cluster == "Female LE") %>% group_by(gene)  %>% add_count(gene, name = "gene_cnt" ) %>% arrange(desc(gene_cnt), gene) %>% filter(avg_log2FC > 0 & p_val_adj < 0.05)
```


```{r}
msc_w_cln_strns <- readRDS("data/processed/Pf/m2_all/msc_w_cln_strns.RDS")
```

```{r}
msc_w_cln_strns 
```


```{r}
le_he_mrks <- imap(msc_w_cln_strns[!names(msc_w_cln_strns) %in% c("MSC12")], ~{
  
  print(.y)
  seurobj <- .x
  le <- c("Female LE", "Male LE")
  he <- c("Female", "Male")
  
  map2(le, he, possibly(~{
    seurobj[["RNA"]]$data <- as(object = seurobj[["RNA"]]$data, Class = "dgCMatrix")
    Idents(seurobj) <- "stage"
    FindMarkers(object = seurobj, ident.1 = .x, ident.2 = .y)
    # FindMarkers(object = seurobj, cells.1 = colnames(seurobj[,seurobj$StagePrelim == "Female LE"]), cells.2 = colnames(seurobj[,seurobj$StagePrelim == "Female"]))
  }, NULL))
  })
```


```{r}
fle_fhe_mrks <- map(le_he_mrks, ~.x[[1]])
fle_fhe_mrks <- fle_fhe_mrks[!unlist(map(fle_fhe_mrks, is.null))]
fle_fhe_mrks <- map(fle_fhe_mrks, ~.x %>% rownames_to_column("gene_id")) %>% bind_rows(., .id = "donor")
```

```{r}

mle_mhe_mrks <- map(le_he_mrks, ~.x[[2]])
mle_mhe_mrks <- mle_mhe_mrks[!unlist(map(mle_mhe_mrks, is.null))]
mle_mhe_mrks <- map(mle_mhe_mrks, ~.x %>% rownames_to_column("gene_id")) %>% bind_rows(., .id = "donor")
```



```{r}
le_he_mrks_sig <- map2(list(fle_fhe_mrks, mle_mhe_mrks), c("f", "m"), ~{.x %>% filter(p_val_adj < 0.05) %>% mutate(dir = ifelse(avg_log2FC > 0, paste0(.y, "le_up"), paste0(.y, "he_up"))) %>% left_join(., Pf3D7_genes_plasmoDB[, c("Gene", "gene_id")], by = c("gene_id"="gene_id")) %>% group_by(donor) %>% arrange(p_val_adj, .by_group = T) %>% group_by(gene_id)  %>% add_count(gene_id, name = "gene_cnt" ) %>% arrange(desc(gene_cnt), gene_id, p_val_adj)}) %>% set_names(c("f", "m")) 
```


```{r}
fle_fhe_mrks_sig_deets <- map(le_he_mrks_sig, ~.x %>% group_by(gene_id) %>% add_count(dir, name = "dir_support" ) %>% mutate(dir_support_prop = round(dir_support/gene_cnt, digits = 2)) %>% distinct(gene_id,dir, .keep_all = T) %>% group_by(gene_id) %>% mutate("opposing_dir" = n_distinct(dir)) %>% ungroup() %>% arrange(desc(opposing_dir), gene_id, p_val_adj))
```


```{r}
fle_fhe_mrks_sig_deets[['m']] %>% filter(opposing_dir == 1 & dir_support > 1)
```

```{r}
fl_dir_nms <- c("fle_up", "fhe_up", "mle_up", "mhe_up")
dir_prop_thres <- c(0.7, 0.7, 1, 1)
dir_count_thres <- c(5, 5, 1, 1)

le_he_sig_gns <- pmap(list(rep(fle_fhe_mrks_sig_deets, each =2), fl_dir_nms, dir_prop_thres, dir_count_thres), ~..1 %>% filter(dir == ..2 & dir_support_prop >= ..3 & dir_support > ..4)) %>% set_names(fl_dir_nms)
```

```{r}
map(le_he_sig_gns, ~arrange(.x, dir_support_prop))
```



```{r}
sayers_Pb_fertility_cln_LEup <- pmap(list(le_he_sig_gns, names(le_he_sig_gns), rep(c("Female.fertility.phenotype", "Male.fertility.phenotype"), each = 2)), ~{
  sayers_Pb_fertility_cln %>% 
  filter(!!rlang::sym(..3) == "Reduced") %>% 
  left_join(..1, fle_up_gns, by = "gene_id") %>% 
  mutate(ovlp = case_when(is.na(`Pfalciparum.3D7.orthologue..ID`) ~ "no ortholog", 
                          !is.na(`Pfalciparum.3D7.orthologue..ID`) & is.na(donor) ~ paste0("absent in ", ..2), 
                          !is.na(`Pfalciparum.3D7.orthologue..ID`) & !is.na(donor) ~ "overlap"))
})
```


```{r}
map(sayers_Pb_fertility_cln_LEup, ~count(.x, ovlp))

```



```{r}
map(sayers_Pb_fertility_cln_LEup, ~.x %>%
  ggplot(aes(x = Female.fertilitylog2FC, y = avg_log2FC)) +
  geom_point() +
  geom_smooth())
```


```{r}
map(sayers_Pb_fertility_cln_LEup, ~.x %>%
  ggplot(aes(y = Female.fertilitylog2FC, x = gene_cnt)) +
  geom_point() )
```

```{r}
map(sayers_Pb_fertility_cln_LEup, ~.x %>%
  ggplot(aes(x = gene_cnt)) +
  geom_bar() )
```


Male and female fertility genes https://www.cell.com/cell-systems/fulltext/S2405-4712(24)00305-3
```{r}
sayers_Pb_male_egress = readxl::read_xlsx("../published_dsets/sayers_PbFertilityGenes_2024/male_fertility_egree_motility.xlsx", sheet = "Male motility and egress screen", .name_repair = "universal", na = c("", "NA")) %>% rename_all(., ~str_remove(., "\\.\\."))
```



Male and female fertility genes https://pubmed.ncbi.nlm.nih.gov/28126901/
```{r}
miao_f_proteome = readxl::read_xlsx("../published_dsets/miao_PfFemMaleproteomes_2017/all_proteomes_frm_study.xlsx", sheet = "I. Female All", .name_repair = "universal", na = c("", "NA")) %>% rename_all(., ~str_remove(., "\\.\\.")) %>% mutate(gene_id = str_replace(Gene.ID, "_", "-"))
```



```{r}
le_he_sig_gns[["fle_up"]] %>% inner_join(., miao_f_proteome, by = "gene_id")

```

Male and female fertility genes https://pubmed.ncbi.nlm.nih.gov/28126901/
```{r}
miao_f_proteome = readxl::read_xlsx("../published_dsets/miao_PfFemMaleproteomes_2017/all_proteomes_frm_study.xlsx", sheet = "I. Female All", .name_repair = "universal", na = c("", "NA")) %>% rename_all(., ~str_remove(., "\\.\\."))
```

## Hart P.yoelli CCR4 knockouts gene DE

#### P.berghei annotations 
```{r, message=FALSE}
##ortholog file to translate to falciparum IDs
ortho <- read_csv("../../Pf3D7_genomes_gtfs_indexs/ortho4.csv")

```

```{r}
hart_male_tr_gns <- read_xlsx('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/Hart_deadenylases_male_gam/DE_due_pyccr4_deltn.xlsx', sheet = 'P adjusted Significant') %>% rename('gene_id'=GeneID...1) %>% left_join(., ortho, by=c('gene_id'='yoelii') ) 

hart_gns <- hart_male_tr_gns%>% filter(!is.na(falciparum)) %>% dplyr::pull(falciparum) %>% str_replace(., '_', '-')

# a= mfet_gns_3.13.14_don_colps %>% filter(fh_fl_only == 'y') %>% pull(gene_id) 
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))
# 
# a= mfet_gns_3.13.14_don_colps %>% filter(fh_only == 'y') %>% pull(gene_id) 
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))
# 
# a= mfet_gns_3.13.14_don_colps %>% filter(mh_ml_only == 'y') %>% pull(gene_id) 
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))
# 
# a= mfet_gns_3.13.14_don_colps %>% filter(mh_only == 'y') %>% pull(gene_id) 
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))

```



```{r}

# a= mfet_gns_3.13.14_don_colps %>% filter(fh_fl_only == 'y') %>% pull(gene_id)
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))
# 
# a= mfet_gns_3.13.14_don_colps %>% filter(fh_only == 'y') %>% pull(gene_id)
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))
# 
# a= mfet_gns_3.13.14_don_colps %>% filter(mh_ml_only == 'y') %>% pull(gene_id)
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))
# 
# a= mfet_gns_3.13.14_don_colps %>% filter(mh_only == 'y') %>% pull(gene_id)
# table(a %in% hart_gns)
# prop.table(table(a %in% hart_gns))

```


```{r}
map(le_he_sig_gns, ~.x %>% filter(gene_id %in% hart_gns) %>%  dim())
```


#### Average expression of 511 Lasonder translationally repressed genes in MSC {.tabset}
```{r, message=FALSE}
lasonder_tr <- read_xlsx('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/lasonder/supp_512_TR_genes.xlsx', skip = 4)
tr_gns <- lasonder_tr %>% dplyr::rename('gene_id' = `Gene ID...1`) %>% pull(gene_id) %>% str_replace(.,'_', '-')
tr_gns <- tr_gns[str_detect(tr_gns, 'PF3D7-')]
```


```{r}
map(le_he_sig_gns, ~.x %>% filter(gene_id %in% tr_gns) %>%  dim())
```

Degraded genes in dogga et al - PF3D7-0530800, PF3D7-0624900, PF3D7-0715400, PF3D7-1338800, PF3D7-1423900, PF3D7-1446100, PF3D7-1466500
```{r}
degraded_gns_dogga <- c("PF3D7-0530800", "PF3D7-0624900", "PF3D7-0715400", "PF3D7-1338800", "PF3D7-1423900", "PF3D7-1446100", "PF3D7-1466500")
table(degraded_gns_dogga %in% tr_gns) 
```

#### Rios widespread release of TR genes 2024 plos

```{r, message=FALSE}
rios_tr <- read_xlsx('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/published_dsets/rios_releaseOfTRGeneHostToVector_2024/journal.ppat.1012823.s010.xlsx', sheet = "TR198 List")

rios_tr_gns <- rios_tr %>% left_join(., ortho, by=c('Gene ID'='yoelii') ) %>% filter(!is.na(falciparum)) %>% dplyr::pull(falciparum) %>% str_replace(., '_', '-')

# rios_tr_gns <- rios_tr_gns %>% dplyr::rename('gene_id' = `Gene ID...1`) %>% pull(gene_id) %>% str_replace(.,'_', '-')
# tr_gns_gns <- tr_gns[str_detect(tr_gns, 'PF3D7-')]
```

```{r}
degraded_gns_dogga <- c("PF3D7-0530800", "PF3D7-0624900", "PF3D7-0715400", "PF3D7-1338800", "PF3D7-1423900", "PF3D7-1446100", "PF3D7-1466500")
table(degraded_gns_dogga %in% rios_tr_gns) 
```

```{r}
map(le_he_sig_gns, ~.x %>% filter(gene_id %in% rios_tr_gns) %>%  dim())
```


##### Female LE, HE, Male LE, HE absence/present genes (detected in 25% of the population) overrepresentation test {.tabset}

```{r}
## Copy the gene lists to make do GO overrepresentation test
mfet_gns_3.13.14_don_colps<- readRDS("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali1_thesis/data/processed/DB52e_star/MSC14/mf_le_he/mfet_gns_3.13.14_don_colps.RDS")
mfet_gns_3.13.14_lst<- readRDS("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali1_thesis/data/processed/DB52e_star/MSC14/mf_le_he/mfet_gns_3.13.14_lst.RDS")
```


```{r}
## Copy the gene lists to make do GO overrepresentation test
mfet_gns_3.13.14_don_colps
```


```{r}
## Copy the gene lists to make do GO overrepresentation test
mfet_gns_3.13.14_don_colps %>% count(fh_only, fh_fl_only, F_genes, FLE_genes)
```

```{r}
## Core genes in FHE that are also core genes in FLE
fhe_fle_gns <- mfet_gns_3.13.14_don_colps %>% filter(fh_fl_only == 'y') %>% pull(gene_id)

## Core genes in FHE that are only unique to FHE 
fhe_only_gns <- mfet_gns_3.13.14_don_colps %>% filter(fh_only == 'y') %>% pull(gene_id)

## Core genes in FHE that plus core genes in FLE
fhe_plus_fle_gns <- mfet_gns_3.13.14_don_colps %>% filter(fh_only == 'y' | fh_fl_only == 'y') %>% pull(gene_id)

```


```{r}
fhe_fle_gns
```


```{r}
all_genes <- Pf3D7_genes_plasmoDB %>% pull(gene_id)
```


```{r}
# Load necessary package
# library(stats)

# Create contingency table
a <- length(intersect(fhe_fle_gns, tr_gns))  # In gene_list AND pathway
a
b <- length(setdiff(fhe_fle_gns, tr_gns))    # In gene_list but NOT pathway
b
c <- length(setdiff(tr_gns, fhe_fle_gns))    # In pathway but NOT gene_list
c
d <- length(setdiff(all_genes, union(fhe_fle_gns, tr_gns))) # Neither in gene_list nor pathway
d
  
```

```{r}
# Create 2x2 table
contingency_table <- matrix(c(a, b, c, d), nrow = 2, byrow = TRUE)
contingency_table
```


```{r}
matrix(c(90, 110, 210, 590), nrow = 2)
```

```{r}
# Perform Fisher's Exact Test
test_result <- fisher.test(contingency_table)
  
# Return odds ratio and p-value
test_result$estimate
test_result$p.value

```


```{r}
# Function for Fisher's Exact Test
enrichment_test <- function(gene_list, all_genes, pathway_genes) {
  # Create contingency table
  a <- length(intersect(gene_list, pathway_genes))  # In gene_list AND pathway
  b <- length(setdiff(gene_list, pathway_genes))    # In gene_list but NOT pathway
  c <- length(setdiff(pathway_genes, gene_list))    # In pathway but NOT gene_list
  d <- length(setdiff(all_genes, union(gene_list, pathway_genes))) # Neither in gene_list nor pathway
  
  # Create 2x2 table
  contingency_table <- matrix(c(a, b, c, d), nrow = 2, byrow = TRUE)
  
  # Perform Fisher's Exact Test
  test_result <- fisher.test(contingency_table)
  
  
  OR <- test_result$estimate  # Odds Ratio
  SE <- sqrt(1/a + 1/b + 1/c + 1/d)  # Standard Error of log OR
  
  # Return odds ratio and p-value
  return(list(OR = OR, SE = SE, P_Value = test_result$p.value))
  
  
  # return(list(Odds_Ratio = test_result$estimate, P_Value = test_result$p.value))
}

```

Enrichment for lasonder genes
```{r}
# Run enrichment test for gene_list1 and gene_list2
fhe_fle_enrichment <- enrichment_test(fhe_fle_gns, all_genes, tr_gns)
fhe_only_enrichment <- enrichment_test(fhe_only_gns, all_genes, tr_gns)
fhe_plus_fle_enrichment <- enrichment_test(fhe_plus_fle_gns, all_genes, tr_gns)

# Print results
cat("Gene List 1: Odds Ratio =", fhe_fle_enrichment$OR, ", P-value =", fhe_fle_enrichment$P_Value, "\n")
cat("Gene List 2: Odds Ratio =", fhe_only_enrichment$OR, ", P-value =", fhe_only_enrichment$P_Value, "\n")
cat("Gene List 2: Odds Ratio =", fhe_plus_fle_enrichment$OR, ", P-value =", fhe_plus_fle_enrichment$P_Value, "\n")

```

```{r}
fold_enrichment = fhe_fle_enrichment$OR/fhe_only_enrichment$OR
fold_enrichment
```

Enrichment for Rios genes
```{r}
# Run enrichment test for gene_list1 and gene_list2
fhe_fle_enrichment_rios <- enrichment_test(fhe_fle_gns, all_genes, rios_tr_gns)
fhe_only_enrichment_rios <- enrichment_test(fhe_only_gns, all_genes, rios_tr_gns)
fhe_plus_fle_enrichment_rios <- enrichment_test(fhe_plus_fle_gns, all_genes, rios_tr_gns)

# Print results
cat("Gene List 1: Odds Ratio =", fhe_fle_enrichment_rios$OR, ", P-value =", fhe_fle_enrichment_rios$P_Value, "\n")
cat("Gene List 2: Odds Ratio =", fhe_only_enrichment_rios$OR, ", P-value =", fhe_only_enrichment_rios$P_Value, "\n")
cat("Gene List 2: Odds Ratio =", fhe_plus_fle_enrichment_rios$OR, ", P-value =", fhe_plus_fle_enrichment_rios$P_Value, "\n")

```

```{r}
fold_enrichment_rios = fhe_fle_enrichment_rios$OR/fhe_only_enrichment_rios$OR
fold_enrichment_rios
```

```{r}

# compare two odds ratios using a Z-test
log_OR1 <- log(fhe_fle_enrichment$OR)
log_OR2 <- log(fhe_only_enrichment$OR)
  
SE1 <- fhe_fle_enrichment$SE
SE2 <- fhe_only_enrichment$SE
  
# Compute Z-score
Z <- (log_OR1 - log_OR2) / sqrt(SE1^2 + SE2^2)
  
# Compute two-tailed p-value
P_Value <- 2 * (1 - pnorm(abs(Z)))

# Print results
cat("Odds Ratio Gene List 1:", fhe_fle_enrichment$OR, "\n")
cat("Odds Ratio Gene List 2:", fhe_only_enrichment$OR, "\n")
cat("Z-Score:", Z, "\n")
cat("P-Value:", P_Value, "\n")


```



```{r}

# Function to compare two odds ratios using a Z-test
compare_enrichment <- function(gene_list1, gene_list2, all_genes, pathway_genes) {
  result1 <- enrichment_test(gene_list1, all_genes, pathway_genes)
  result2 <- enrichment_test(gene_list2, all_genes, pathway_genes)
  
  log_OR1 <- log(result1$OR)
  log_OR2 <- log(result2$OR)
  
  SE1 <- result1$SE
  SE2 <- result2$SE
  
  # Compute Z-score
  Z <- (log_OR1 - log_OR2) / sqrt(SE1^2 + SE2^2)
  
  # Compute two-tailed p-value
  P_Value <- 2 * (1 - pnorm(abs(Z)))
  
  return(list(Z_Score = Z, P_Value = P_Value, OR1 = result1$OR, OR2 = result2$OR, SE1 = result1$SE, SE2 = result2$SE))
}

```

Lasonder genes 
```{r}

# Run enrichment comparison
comparison_result <- compare_enrichment(fhe_fle_gns, fhe_only_gns, all_genes, tr_gns)

# Print results
cat("Odds Ratio Gene List 1:", comparison_result$OR1, "\n")
cat("Odds Ratio Gene List 2:", comparison_result$OR2, "\n")
cat("Z-Score:", comparison_result$Z_Score, "\n")
cat("P-Value:", comparison_result$P_Value, "\n")
```

Rios genes 
```{r}

# Run enrichment comparison
comparison_result_rios <- compare_enrichment(fhe_fle_gns, fhe_only_gns, all_genes, rios_tr_gns)

# Print results
cat("Odds Ratio Gene List 1:", comparison_result_rios$OR1, "\n")
cat("Odds Ratio Gene List 2:", comparison_result_rios$OR2, "\n")
cat("Z-Score:", comparison_result_rios$Z_Score, "\n")
cat("P-Value:", comparison_result_rios$P_Value, "\n")
```

```{r}
# Example data (replace with actual OR values and confidence intervals)
data <- data.frame(
  GeneList = c("fhe_fle_gns", "fhe_only_gns"),
  OR = c(comparison_result$OR1, comparison_result$OR2),  # Odds Ratios
  SE = c(comparison_result$SE1, comparison_result$SE2)
)

# Calculate confidence intervals
data$Lower_CI <- exp(log(data$OR) - 1.96 * data$SE)
data$Upper_CI <- exp(log(data$OR) + 1.96 * data$SE)

# Plot
ggplot(data, aes(x = GeneList, y = OR)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  geom_errorbar(aes(ymin = Lower_CI, ymax = Upper_CI), width = 0.2) +
  ylab("Odds Ratio (Enrichment)") +
  xlab("Gene List") +
  ggtitle("Pathway Enrichment Comparison") +
  theme_minimal()

```

```{r}
# Create a data frame with counts
prop_data <- data.frame(
  GeneList = factor(rep(c("Female LE \n retained genes", "Female LE \n lost genes"), each = 2), levels = c("Female LE \n retained genes", "Female LE \n lost genes")),
  Category = rep(c("TR genes", "Non-TR Genes"), 2),
  Count = c(
    length(intersect(fhe_fle_gns, tr_gns)),  # Pathway genes in gene_list1
    length(setdiff(fhe_fle_gns, tr_gns)),    # Other genes in gene_list1
    length(intersect(fhe_only_gns, tr_gns)),  # Pathway genes in gene_list2
    length(setdiff(fhe_only_gns, tr_gns))     # Other genes in gene_list2
  )
)

prop_data
```


```{r, fig.width= 4, fig.height=4.5}
# Plot
ggplot(prop_data, aes(x = GeneList, y = Count, fill = Category)) +
  geom_bar(stat = "identity", position = "fill") +  # "fill" scales to proportion
  ylab("Proportion of Genes") +
  xlab("Female genes") +
  # ggtitle("Proportion of Lasonder TR Genes") +
  scale_fill_manual(values = c("lightgray", "steelblue")) +
  theme_classic() +
  theme(text = element_text(size = 19),
        legend.position = "bottom") +
  guides(fill = guide_legend(title.position = "top"))

```
```{r, fig.width= 6.5, fig.height=4.5}
# Plot
ggplot(prop_data, aes(x = GeneList, y = Count, fill = Category)) +
  geom_bar(stat = "identity", position = "dodge") +  # "fill" scales to proportion
  ylab("Proportion of Genes") +
  xlab("Female genes") +
  # ggtitle("Proportion of Lasonder TR Genes") +
  scale_fill_manual(values = c("lightgray", "steelblue")) +
  theme_classic() +
  theme(text = element_text(size = 19))

```

```{r}
ggplot(prop_data, aes(x = GeneList, y = Count, fill = Category)) +
  geom_bar(stat = "identity", position = "dodge") +  # "dodge" places bars side by side
  ylab("Gene Count") +
  xlab("Gene List") +
  ggtitle("Comparison of Pathway Gene Counts in Gene Lists") +
  scale_fill_manual(values = c("steelblue", "lightgray")) +
  theme_minimal()

```


```{r}
# Create a data frame with counts
prop_data_rios <- data.frame(
  GeneList = factor(rep(c("Female LE \n retained genes", "Female LE \n lost genes"), each = 2), levels = c("Female LE \n retained genes", "Female LE \n lost genes")),
  Category = rep(c("TR genes", "Non-TR Genes"), 2),
  Count = c(
    length(intersect(fhe_fle_gns, rios_tr_gns)),  # Pathway genes in gene_list1
    length(setdiff(fhe_fle_gns, rios_tr_gns)),    # Other genes in gene_list1
    length(intersect(fhe_only_gns, rios_tr_gns)),  # Pathway genes in gene_list2
    length(setdiff(fhe_only_gns, rios_tr_gns))     # Other genes in gene_list2
  )
)

prop_data_rios
```


```{r, fig.width= 4, fig.height=4.5}
# Plot
ggplot(prop_data_rios, aes(x = GeneList, y = Count, fill = Category)) +
  geom_bar(stat = "identity", position = "fill") +  # "fill" scales to proportion
  ylab("Proportion of Genes") +
  xlab("Female genes") +
  # ggtitle("Proportion of Lasonder TR Genes") +
  scale_fill_manual(values = c("lightgray", "steelblue")) +
  theme_classic() +
  theme(text = element_text(size = 17),
        legend.position = "bottom") +
  guides(fill = guide_legend(title.position = "top"))

```

```{r, message=FALSE}
f_canonical_loss <- read_xlsx('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali1_thesis/data/raw/from_sk21/figs24_f.xlsx', sheet = "933")
f_all_loss <- read_xlsx('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali1_thesis/data/raw/from_sk21/figs24_f.xlsx', sheet = "Sheet2")


f_canonical_loss_gns <- f_canonical_loss %>% dplyr::pull(`Gene ID`) %>% str_replace(., '_', '-')
f_all_loss_gns <- f_all_loss %>% dplyr::pull(`Gene ID`) %>% str_replace(., '_', '-')


```


```{r, message=FALSE}
f_canonical_loss_gns
f_all_loss_gns
```



```{r, fig.height=3.8, fig.width=3.8}
# M48, M49, M60
ggvenn(list("All_lost" = f_all_loss_gns,
            "Canonical_lost" = f_canonical_loss_gns,
            "TR" = tr_gns),
       show_percentage = F,
       text_size = 6)


```

```{r, fig.height=3.8, fig.width=3.8}

ggvenn(list("All_lost" = f_all_loss_gns,
            "Canonical_lost" = f_canonical_loss_gns,
            "FL_FH_Core" = fhe_fle_gns,
            "TR" = tr_gns),
       show_percentage = F,
       text_size = 6)


```


```{r, fig.height=3.8, fig.width=3.8}
ggvenn(list("All_lost" = f_all_loss_gns,
            "Canonical_lost" = f_canonical_loss_gns,
            "FH_only_core" = fhe_only_gns,
            "TR" = tr_gns),
       show_percentage = F,
       text_size = 6)


```



26 male genes bound by RNA binding proteins in P.yoeli and retain introns upon knockout of Rbpm1

PY17X_1311800
PY17X_1323900
PY17X_1357300
PY17X_1335600
PY17X_1122300
PY17X_1109100
PY17X_0521800
PY17X_0523500
PY17X_1452900
PY17X_1241500
PY17X_0302800
PY17X_0721100
PY17X_1333900
PY17X_0510800
PY17X_0204100
PY17X_0919000
PY17X_0603800
PY17X_0105800
PY17X_1216400
PY17X_0415900
PY17X_0833600
PY17X_0508900
PY17X_1450400
PY17X_1341200
PY17X_1305400
PY17X_1320300

https://www-nature-com.ezp.lib.cam.ac.uk/articles/s41467-024-49002-9#Sec46
Guan et al., 2024
```{r, fig.height=3.8, fig.width=3.8}
guan_26_gns <- c("PY17X_1311800","PY17X_1323900","PY17X_1357300","PY17X_1335600","PY17X_1122300","PY17X_1109100","PY17X_0521800","PY17X_0523500","PY17X_1452900","PY17X_1241500","PY17X_0302800","PY17X_0721100","PY17X_1333900","PY17X_0510800","PY17X_0204100","PY17X_0919000","PY17X_0603800","PY17X_0105800","PY17X_1216400","PY17X_0415900","PY17X_0833600","PY17X_0508900","PY17X_1450400","PY17X_1341200","PY17X_1305400","PY17X_1320300")


```

```{r, fig.height=3.8, fig.width=3.8}
pf_guan_ortho <- ortho %>% filter(yoelii %in% guan_26_gns)

```


```{r, fig.height=3.8, fig.width=3.8}
 pf_guan_ortho %>% select(falciparum, yoelii)

```


```{r, fig.height=3.8, fig.width=3.8}
 

```

```{r, fig.height=3.8, fig.width=3.8}
 

```

