---
title: "Integration of donor only and with lab"
date: "2024-01-05"
author: "Jesse Rop"
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  # library(sctransform)
  # library(rgl)
  library(ggrepel)
  # library(ggpubr)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  # library(ggnewscale)
  library(tidyr)
  library(cowplot)
  # library(scDblFinder)
  library(SeuratData)
  library(SeuratDisk)
  # library(scmap)
  library(SingleCellExperiment)
  library(pheatmap)
  library(plotly)
  # library(Hmisc)
  # library(rmarkdown)
  library(RColorBrewer)
  # library(MAST)
  library(ggvenn)
  # library("biomaRt")
  library(conflicted)
  # library(SoupX)
  library(MetBrewer)
  # library(dreamlet)
  # library(muscat)
  # library(ExperimentHub)
  # library(zenith)
  library(scater)
  library(slingshot)
  library(harmony)
  library(ggforce)
  # library(miloR)
  library(scater)
  library(scran)
  library(patchwork)
  # library(scRNAseq)
  library(scuttle)
  library(irlba)
  library(BiocParallel)
  library(tradeSeq)
  library(ComplexHeatmap)
  library(circlize)
  library(ggh4x)
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflicts_prefer(dplyr::count)
  conflicts_prefer(base::as.data.frame)
  
})
```


```{r setup}
sample_set = "Mali2"

knitr::opts_knit$set(root.dir = paste0("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/", sample_set))

```



```{r }
save_dir <- "data/processed/Pf/m2_all/strain_DE_afm/"
```

#  integration of all MSC

### Common functions

#### Source commong functions, varriables and tables
```{r}
##Boxplot labeling function
source("../multipurpose_scripts_lnk/seurat_sce/seu_sce_common_fns.R")
source("../multipurpose_scripts_lnk/plotting_common_fns.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/pf_common_vars.R")
source("/nfs/users/nfs_j/jr35/Mali2/mali22_code/14_strain_DE/strain_DE_common_fns.R")
```

```{r}
# optimum_k_qcd <- c("MSC33" =8, "MSC48" =4, "MSC49" =6, "MSC55" =7, "MSC60" =7, "MSC66" =8, "MSC38" =4, "MSC40" =8, "MSC50_MACS" =1, "MSC53" =1, "MSC57" =1, "MSC24" =1, "MSC37" =5, "MSC39" =4, "MSC45" =1, "MSC50_SLO" =1, "MSC54" =1, "MSC67" =1, "MSC68" =3, "MSC70" =1)
# 
# optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]
```

Integrated asexual
```{r, message=F, warning=F, eval=T}
## Get file list
fm_leiden_w_donors_pt_list_files <- Sys.glob("data/processed/Pf/m2_all/integration/fm_leiden_w_donors_*_pt.RDS")

fm_leiden_w_donors_pt_list_files
```

```{r, message=F, warning=F, eval=T}
## Subset to candidate ones
fm_leiden_w_donors_pt_list_files <- fm_leiden_w_donors_pt_list_files[str_detect(fm_leiden_w_donors_pt_list_files, "all_m_mLE|all_m|w_pt5_f_fLE|w_pt7_f|w_pt7_f_fLE")]

fm_leiden_w_donors_pt_list_files
```

```{r, message=F, warning=F, eval=T}
## read filtered barcodes
fm_leiden_w_donors_pt <- map(fm_leiden_w_donors_pt_list_files, ~.x %>% 
  set_names(nm = (str_remove_all(.,'data/processed/Pf/m2_all/integration/fm_leiden_w_donors_|_pt.RDS')))) %>% 
  unlist(., recursive = F) %>% 
  # .[sample_name] %>%
  map(readRDS)

```

```{r, warning=FALSE, message=FALSE}
imap(fm_leiden_w_donors_pt, ~FeaturePlot(.x, features = c("slingPseudotime_1"), reduction = "umap.harmony_PC_man", pt.size = 0.01) + labs(title = .y))
```


```{r, warning=FALSE, message=FALSE}
# afm_leiden_w_donors_pt <- fm_leiden_w_donors_pt[c("w_pt7_f", "all_m")]
```

```{r}
## read asexual
v3_m2_asex_cr_filt_norm_harmony_pt <- readRDS("data/processed/Pf/asex_noM55_jcC_harmony_no_comtd_pt3.RDS")
```

```{r, warning=FALSE, message=FALSE}
afm_leiden_w_donors_pt <- c(v3_m2_asex_cr_filt_norm_harmony_pt, fm_leiden_w_donors_pt[c("w_pt7_f", "all_m")])
```


```{r, warning=FALSE, message=FALSE}
names(afm_leiden_w_donors_pt) <- c("a", "f", "m")
```


```{r, message=FALSE, warning=FALSE}
# Save for tradeseq and plotting 
saveRDS(afm_leiden_w_donors_pt, paste0(save_dir, 'afm_leiden_w_donors_pt.RDS'))

```

```{r, warning=FALSE, message=FALSE}
imap(afm_leiden_w_donors_pt, ~FeaturePlot(.x, features = c("slingPseudotime_1"), reduction = "umap.harmony_PC_man", pt.size = 0.01) + labs(title = .y))
```

```{r}
## Creating new stage variables which are need to fit in with old code
afm_leiden_w_donors_pt <- map(afm_leiden_w_donors_pt, ~.x@meta.data %>% 
  mutate(#across(donor, ~str_remove_all(.x, "C$")),
         sample_id = donor,
         donor_nm = str_remove_all(donor, "_SLO|_MACS|C$"),
         condition = factor(condition, levels = c("Asymptomatic", "Symptomatic")), 
         Symptomatic = condition,
         # across(Symptomatic, ~factor(str_replace_all(.,  c("Yes" = "Symptomatic", "No" = "Asymptomatic")), levels=c("Symptomatic", "Asymptomatic"))),
         StageFL = stage,
         StageFL_ND = stage,
         dset = dset_labmrg,
         # StrainDon = paste0(str_remove(donor, "SC"), "_", Strain_DN),
         StrainDon = paste0(str_remove(donor_nm, "SC"), "_", Strain_DN),
         pseudotime = slingPseudotime_1,
         pseudotime_fld_only = slingPseudotime_1,
         # stage = stage_fld,
         strain = factor(Strain_DN, levels = strain_nms_lvls),
         dset = "Field",
         dsetC = "Field") %>% 
  select(StageFL, StageFL_ND, dset,StrainDon,pseudotime, pseudotime_fld_only,  strain, donor_nm, condition, Symptomatic, donor, dset, dsetC) %>% 
  AddMetaData(.x, .)
)
```



```{r}
map(afm_leiden_w_donors_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "harmony_clusters_ss" ))
```


```{r}
map(afm_leiden_w_donors_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "stage" ))

```

```{r}
map(afm_leiden_w_donors_pt, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man", features = "slingPseudotime_1"))
```

```{r, warning=FALSE, message=FALSE}
# ap2-g pfs16 gexp02 gene_count
map(afm_leiden_w_donors_pt, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man",  label = T, features = "PF3D7-1222600", order = T))
map(afm_leiden_w_donors_pt, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man",  label = T, features = "PF3D7-0406200", order = T))
map(afm_leiden_w_donors_pt, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man",  label = T, features = "PF3D7-1102500", order = T))
map(afm_leiden_w_donors_pt, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man",  label = T, features = "nFeature_RNA", order = T))
map(afm_leiden_w_donors_pt, ~FeaturePlot(.x, reduction = "umap.harmony_PC_man",  label = T, features = "nCount_RNA", order = T))


```

```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
## Name these  pseudotime chunks as downsampling_xxx since they are for downsampling

bal_df <- map(afm_leiden_w_donors_pt, ~ptime_chunk_calc_df_fn(sc_df = .x@meta.data, ptime = "slingPseudotime_1", chunk_num = 20, p_cat_nm = "downsampling_ptime_cat", p_bal_cat_nm = "downsampling_ptime_bal_cat"))
```

```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks
afm_leiden_w_donors_pt <- map2(afm_leiden_w_donors_pt, bal_df, ~AddMetaData(.x, .y) )
```

```{r}
## subset where field parasites are for slingshot
map(afm_leiden_w_donors_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "downsampling_ptime_cat" ))
map(afm_leiden_w_donors_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "downsampling_ptime_bal_cat" ))
```

```{r}
# Colours for plotting pseudotime chunks
ptc_cols <- c("#068B4CFF", "#4FB0C0FF", "#74DC09FF", "#3377ff", "#E769ACFF", 
"#8A1708FF", "#048461FF", "#3BDF3FFF", "#5559B5FF", "#557406FF", 
"#2A158CFF", "#B48606FF", "#CA2F4DFF", "#E75FFDFF", "#E35655FF", 
"#42CFC3FF", "#1BA0B9FF", "#D481E8FF", "#958FEBFF", "#4A7406FF", 
"#540B9CFF") %>% set_names(paste0("PT.", 1:21))
```

```{r}
map(afm_leiden_w_donors_pt, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "downsampling_ptime_bal_cat", cols = ptc_cols, label = T ) +
  theme(legend.position = "none"))
```

```{r}
imap(afm_leiden_w_donors_pt, ~saveRDS(.x, paste0(save_dir, "afm_leiden_w_donors_pt_downsampling_chunks_", .y, ".RDS")))
```


```{r}
# Count the donors with atleast 2 strains each with atleast 100 cells
don_100n <- imap(afm_leiden_w_donors_pt, ~.x@meta.data %>% count(orig.ident, StrainDon) %>% filter(n>100) %>% group_by(orig.ident) %>% filter(n_distinct(StrainDon) > 1) %>% pull(orig.ident) %>% unique() %>% str_remove(., ".mrna"))
```


```{r}
## Get only the asexuals
# m2_asex_seu <- map(msc_w_cln_strns[c("MSC49", "MSC60")], ~.x[,.x@meta.data$stage %in% c("Early trophozoite", "Late ring")] %>% AddMetaData(., ))
# dons_bal_sset <- c("M49_", "M60_", "M33_", "M66_", "M55_", "M49_|M55_")
# dons_bal_sset <- c("M49_", "M60_", "M33_", "M66_")
# dons_bal_sset <- c("M49C_", "M66C_")

# dons_bal_sset_nms <- str_remove_all(str_replace_all(dons_bal_sset, "M", "MSC"), "_$|C_$|\\|")
dons_bal_sset_nms <- don_100n

# Retain donors with atleast 2 strains each with atleast 100 cells for each stage
m2_afm_seu <- map2(afm_leiden_w_donors_pt, dons_bal_sset_nms, ~.x[, str_detect(.x@meta.data$orig.ident, paste(.y, collapse = "|")) & !str_detect(.x@meta.data$StrainDon, "Negative|Doublet")])

```

### Balance (subsample) the strain cells within pseudotime chunks


```{r, warning=FALSE, message=FALSE}
set.seed(989778)
# ds_grp <- rep('StrainDon', length(dons_bal_sset_nms))

# balanced_cells_tbl <- map2(m2_afm_seu, ds_grp, ~balanced_cells_fn(sc_df = .x@colData, dsample_condtn = .y , ptime_strategy = c("donor", "ptime_bal_cat"), multpl = 1, don_yes_no = "no"))
balanced_cells_tbl <- map(m2_afm_seu, ~balanced_cells_fn(sc_df = .x@meta.data, dsample_condtn = "strain" , ptime_strategy = c("donor", "downsampling_ptime_bal_cat"), multpl = 1, don_yes_no = "no"))
```




```{r, warning=FALSE, message=FALSE}
######################### !!!!!NOTE - Unbalanced alternative to increase power replacing balanced with unbalanced - switch to return balanced or viceversa 
# m2_afm_bal_seu <- m2_afm_seu
m2_afm_bal_seu <- map2(m2_afm_seu, balanced_cells_tbl, ~.x[, .y$cell_bc])

```

```{r}
m2_afm_bal_seu[[1]]@meta.data %>% data.frame() %>% dplyr::count(downsampling_ptime_bal_cat, StrainDon)

```


```{r, warning=FALSE, message=FALSE}
## Calculate pseudotime chunks for DE
bal_df2 <- map(m2_afm_seu, ~ptime_chunk_calc_df_fn(sc_df = .x@meta.data, ptime = "slingPseudotime_1", chunk_num = 7, p_cat_nm = "ptime_cat", p_bal_cat_nm = "ptime_bal_cat"))

```


```{r, warning=FALSE, message=FALSE}
## Add pseudotime chunks to metadata
m2_afm_bal_seu <- map2(m2_afm_bal_seu, bal_df2, ~AddMetaData(.x, .y) )

```


```{r}
m2_afm_bal_seu[[1]]@meta.data %>% count(ptime_bal_cat, donor, strain)
```

```{r}
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_cat" , label = T ))
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" , label = T))
```


```{r}
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" , label = T))
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" , label = T))
```

```{r}
map(m2_afm_bal_seu, ~.x@meta.data %>% count(ptime_bal_cat, donor, strain))
```

```{r}
## Subset to remove within each pseudotime chunk any strain with less than 10 cells to ensure there are no pseudotime chunks where some strains are overwhelmingly underrepresented
m2_afm_bal_seu <- map(m2_afm_bal_seu, ~.x[, .x@meta.data %>% 
                                               rownames_to_column('bcode') %>% 
                                               group_by(ptime_bal_cat, donor, strain) %>% 
                                               # filter(n() >=20) %>%
                                               filter(n() >=10) %>%
                                               group_by(ptime_bal_cat, donor) %>% 
                                               filter(n_distinct(strain) >1) %>%
                                               # filter(n() >=100) %>% 
                                               pull(bcode)]
)


```

```{r}
m2_afm_bal_seu[[1]]@meta.data %>% count(donor,ptime_bal_cat,  strain) %>% group_by(donor,ptime_bal_cat) %>% mutate(strn_n = n_distinct(strain)) %>% filter(strn_n >1)
```


```{r}
m2_afm_bal_seu[[1]]@meta.data %>% count( donor, strain) %>% group_by( donor) %>% mutate(strn_n = n_distinct(strain)) %>% filter(strn_n >1)
```


```{r}
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" , label = T))
```

```{r, fig.width=8, fig.height=3}
map(m2_afm_bal_seu, ~.x@meta.data %>% 
  count( donor,  strain) %>% 
  mutate(across(donor, ~str_remove_all(., "SC|SLO|C|_"))) %>% 
  group_by( donor) %>% 
  mutate(strn_n = n_distinct(strain)) %>% 
  filter(strn_n >1) %>%
  ggplot(aes(x = strain, y = n)) +
  geom_col() +
  # geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.45))) +
  facet_nested_wrap(vars(donor), nrow = 1, scales = "free_x") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        strip.text = element_text(size = 14),
        text = element_text(size = 14)) 
)
```

Remove the strain that are overal less than 200 in a donor and retain donors where atleast two strains are available for DE
```{r}
m2_afm_bal_seu <- map(m2_afm_bal_seu, ~.x[, .x@meta.data %>% 
                                               rownames_to_column('bcode') %>% 
                                               group_by(donor, strain) %>% 
                                               # filter(n() >=20) %>%
                                               filter(n() >=200) %>%
                                               group_by(donor) %>% 
                                               filter(n_distinct(strain) >1) %>%
                                               # filter(n() >=100) %>% 
                                               pull(bcode)]
)


```

```{r, fig.width=8, fig.height=3}
map(m2_afm_bal_seu, ~.x@meta.data %>% 
  count( donor,  strain) %>% 
  mutate(across(donor, ~str_remove_all(., "SC|SLO|C|_"))) %>% 
  group_by( donor) %>% 
  mutate(strn_n = n_distinct(strain)) %>% 
  filter(strn_n >1) %>%
  ggplot(aes(x = strain, y = n)) +
  geom_col() +
  # geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.45))) +
  facet_nested_wrap(vars(donor), nrow = 1, scales = "free_x") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        strip.text = element_text(size = 14),
        text = element_text(size = 14)) 
)
```



```{r}
# rand_color(n = 9, transparency = 0.2) %>% dput()
ptc_cols2 <- c("#8FBE68CC", "#85F3D6CC", "#92C70ACC", "#CC0D28CC", "#958FEBFF", 
"#0C9592CC", "#CAE37DCC", "#116575CC") %>% set_names(paste0("PT.", 2:9))
```

```{r}
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat", cols = ptc_cols2, label = T ) +
  theme(legend.position = "none"))

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=3}
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat", label.size = 4, label.box = T, label = T) +
  theme_classic() +
  scale_color_manual(values = ptc_cols2) +
  scale_fill_manual(values = ptc_cols2) +
  labs(title = "", x = "UMAP1", y= "UMAP2") +
  theme(text = element_text(size = 17),
        legend.position = "none",
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) 
)


```

```{r}
m2_afm_bal_seu[[1]]@meta.data %>% count( ptime_bal_cat) 

```

```{r, fig.width=16, fig.height=3}
map(m2_afm_bal_seu, ~.x@meta.data %>% 
  count( donor, ptime_bal_cat, strain) %>% 
  mutate(across(donor, ~str_remove_all(., "SC|SLO|C|_")),
         across(ptime_bal_cat, ~str_remove_all(., "\\."))) %>% 
  group_by( donor) %>% 
  mutate(strn_n = n_distinct(strain)) %>% 
  filter(strn_n >1) %>%
  ggplot(aes(x = strain, y = n)) +
  geom_col() +
  # geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90)+
  facet_nested_wrap(vars(donor,ptime_bal_cat), nrow = 1, scales = "free_x") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        strip.text = element_text(size = 14),
        text = element_text(size = 14)) 
)
```

```{r, fig.width=8, fig.height=3}
map(m2_afm_bal_seu, ~.x@meta.data %>% 
  count( donor,  strain) %>% 
  mutate(across(donor, ~str_remove_all(., "SC|SLO|C|_"))) %>% 
  group_by( donor) %>% 
  mutate(strn_n = n_distinct(strain)) %>% 
  filter(strn_n >1) %>%
  ggplot(aes(x = strain, y = n)) +
  geom_col() +
  # geom_col(position = position_dodge2()) + #width=0.4, position = position_dodge(width=0.5) 
  geom_text(aes(label=n), hjust=-0.05, vjust= 0.5, position = position_dodge(width = .9), size = 5, show.legend = F, fontface ='bold', angle = 90)+
  scale_y_continuous(expand = expansion(mult = c(0.01, 0.45))) +
  facet_nested_wrap(vars(donor), nrow = 1, scales = "free_x") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90),
        strip.text = element_text(size = 14),
        text = element_text(size = 14)) 
)
```

```{r, eval=FALSE}
map(m2_afm_bal_seu, ~DimPlot(.x, reduction = "umap.harmony_PC_man", group.by = "ptime_bal_cat" ))
```





## MAST DE adjusting for pseudotime

MAST DE on the entire dataset before subsampling. Subsampling is not necessary for MAST and also for tradeseq
### Prepare dataset
Normalizing and scaling the data in the subseted seurat objects - may have been done before on the larger parent object but just repeating since will not change anything as it starts from the raw counts.

Adding pseudotime metadata from the single cell experiment object 
Scaling is for visualization


```{r, message=FALSE, warning=FALSE}
m2_afm_bal_seu <- map(m2_afm_bal_seu, ~JoinLayers(.x, layers = c("counts", "data", "scale.data")))
```


```{r, message=FALSE, warning=FALSE}
# Add to metadata the dataset/conditions and ptime cat
m2_afm_bal_seu <- map(m2_afm_bal_seu, ~AddMetaData(.x, 
                                                     .x@meta.data %>% mutate(don_ptime = paste0(donor, "_", ptime_bal_cat)) %>% select(don_ptime)) 
                       )

```

```{r, message=FALSE, warning=FALSE}
# Save for tradeseq and plotting 
saveRDS(m2_afm_bal_seu, paste0(save_dir, 'm2_afm_bal_seu.RDS'))

```

```{r, message=FALSE, warning=FALSE}
# all_afm_de_bal_nosct <- SplitObject(m2_afm_bal_seu, split.by = "don_ptime")
all_afm_de_bal_nosct <- map(m2_afm_bal_seu, ~SplitObject(.x, split.by = "donor")) %>% 
  unlist(recursive = T)
```

```{r, message=FALSE, warning=FALSE}
all_gns = map(all_afm_de_bal_nosct, ~unique(rownames(.x)))

afm_m2_bal_seu <- map2(all_afm_de_bal_nosct,all_gns, ~{
  .x %>%
         NormalizeData(., normalization.method = "LogNormalize", scale.factor = 10000, verbose = FALSE) %>%
         ScaleData(., features = .y,  verbose = FALSE)
  })

```

### Actual MAST DE between pairwise strains and fieldvslab 

MAST DE adjusting for pseudotime as a latent variable

```{r, message=FALSE, warning=FALSE}
de_idents <- rep("strain", length = length(afm_m2_bal_seu))
afm_m2_bal_mast <- map2(afm_m2_bal_seu, de_idents, ~SetIdent(.x, value=.y))

```

```{r, message=FALSE, warning=FALSE}
saveRDS(afm_m2_bal_mast, paste0(save_dir, 'afm_m2_bal_mast.RDS'))

```



```{r, warning=FALSE, message=FALSE}
for (nm in names(afm_m2_bal_mast)) {
  afm_m2_bal_mast[[nm]][["RNA"]]$data <- as(afm_m2_bal_mast[[nm]][["RNA"]]$data, Class = "dgCMatrix")
}
```

```{r, message=FALSE, warning=FALSE, eval=T}
##!! NB Rerun only when the DE inputs change - takes looooonng
go()
## Mast DE of dataset with male low (ml) included - shouldnt have any contribution as each stage handled differently
## Including lab and field strain DE
## !!!min.cells.group added for mh in the strain_dset where there's only balancing in lab VS field - we need to remove later

# pts <- "slingPseudotime_1"
pts <- c("ptime_bal_cat", "slingPseudotime_1")
# pts_all<- c(rep(pts,1))
pts_all<- rep(list(pts),length(afm_m2_bal_mast))
# pts_all<- pts

### !!! NOTE - RUNNING FOR NECESSARY DONORS, WILL REVERT LATER
afm_pairwise_mkrs_seu_clustr <- list(afm_m2_bal_mast, pts_all, de_idents, names(afm_m2_bal_mast)) %>%
# ap2g_ub_pairwise_mkrs_pt_ml <- list(list(m2_bal_mast_noM55M54), pts, de_idents, names(m2_bal_mast)) %>%
  pmap(.,  ~{
    # ide = ..3
    a = combn(levels(droplevels(..1@meta.data[,..3])), 2,simplify = F)
    # a = combn(unique(..1@meta.data[,..3]), 2,simplify = F)
    obj = ..1
    pseudos = ..2
    
    
    print(class(obj))
    print(pseudos)
    
    print(paste("running ...", ..4))
    map(a, ~print(paste(.x[1], "n ", .x[2])))
    
    # map(a, ~FindMarkers(obj, ident.1 = .x[1], ident.2 = .x[2], latent.vars = pseudos, logfc.threshold = 0, test.use = 'MAST', min.cells.group = 1) %>% rownames_to_column('gene_id'))
    map(a, ~FindMarkers(obj, 
                        ident.1 = .x[1], 
                        ident.2 = .x[2], 
                        # latent.vars = c("donor", pseudos), 
                        latent.vars = c(pseudos), 
                        logfc.threshold = 0, 
                        test.use = 'MAST', 
                        min.cells.group = 1) %>% 
          rownames_to_column('gene_id')) %>%
      set_names(map(a, ~paste(., collapse = '_vs_') %>% str_remove_all(' '))) %>%
      bind_rows(., .id = 'pw_comparison')
    
  }#, 'missing genes')
  )

saveRDS(afm_pairwise_mkrs_seu_clustr, paste0(save_dir, 'afm_pairwise_mkrs_seu_clustr.RDS'))
# saveRDS(ap2g_ub_pairwise_mkrs_pt_ml_don,'data/processed/Pf/m2_all/ap2g_ub_pairwise_mkrs_pt_ml_don.RDS')
# c(ap2g_ub_pairwise_mkrs_pt_ml_lf, ap2g_ub_pairwise_mkrs_pt_ml)



```


```{r, message=FALSE, warning=FALSE}
afm_pairwise_mkrs_seu_clustr<- readRDS(paste0(save_dir, 'afm_pairwise_mkrs_seu_clustr.RDS'))

```


```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot <- map(afm_pairwise_mkrs_seu_clustr, ~.x %>%  left_join(., Pf3D7_genes_plasmoDB[,c("gene_id", "Gene")]))

afm_pairwise_mkrs_seu_clustr_anot
```


```{r, warning=FALSE, message=FALSE}
# Another amplification of 3x on chromosome 12
# was observed immediately before GTP cyclohydrolase I (pfgch1,
# PF3D7_1224000) across two genes (PF3D7_1223800-PF3D7_1223900), which
# was also in accordance to previous studies 
# https://link.springer.com/article/10.1186/gb-2009-10-2-r21
# https://escholarship.org/content/qt1bc2p97r/qt1bc2p97r_noSplash_c86b54d668a56b24f34dfe70cb02191e.pdf?t=o8x13q
```


```{r, warning=FALSE, message=FALSE}
map(afm_pairwise_mkrs_seu_clustr_anot, ~.x  %>% filter(p_val_adj < 0.05 & avg_log2FC > 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in gam positive cells

```


```{r, warning=FALSE, message=FALSE}
imap(afm_pairwise_mkrs_seu_clustr_anot, ~.x %>% filter(p_val_adj < 0.05 & avg_log2FC < 0))

## https://www-nature-com.ezp.lib.cam.ac.uk/articles/nature24280.pdf - PF3D7-0406500 and HDA1 are high in gam positive cells

```


```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot_all <- afm_pairwise_mkrs_seu_clustr_anot %>% 
  bind_rows(., .id = "sample") %>%
  filter(p_val_adj < 0.05) %>%
  group_by(gene_id) %>%
  mutate(across(sample, ~str_remove(., "_ptime")),
         donor = str_remove_all(sample, "w.*\\.|_SLOC|_MACS|_SLO"),
         gene_DE_sample_prevalence = n_distinct(sample, pw_comparison),
         gene_DE_donor_prevalence = n_distinct(donor, pw_comparison)) %>%
  ungroup() %>%
  mutate(gene_DE_sample_prevalence_cat = cut(gene_DE_sample_prevalence, breaks = c(3), labels = c("Low", "Medium", "High"), right = FALSE), 
         gene_DE_donor_prevalence_cat = cut(gene_DE_donor_prevalence, breaks = c(3), labels = c("Low", "Medium", "High"), right = FALSE))  %>% 
  filter(!str_detect(Gene, "VAR_"))
```


```{r, warning=FALSE, message=FALSE}
write_csv(afm_pairwise_mkrs_seu_clustr_anot_all, "../colleagues/chiyun/afm_pairwise_mkrs_seu_clustr_anot.csv")
```

```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot_all  %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T)
```

```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot_all  %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T)
```


```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot_all %>% 
  filter(!str_detect(Gene, "VAR_")) %>%
  filter(Gene == "MDR1")
```


```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot_all %>% 
  filter(!str_detect(Gene, "VAR_")) %>%
  distinct(Gene, .keep_all = T)
```


```{r, warning=FALSE, message=FALSE}
## cells that contain 2 or more gam UMI
c("PF3D7-0113600", "PF3D7-1301800", "PF3D7-1222600")
```


```{r, warning=FALSE, message=FALSE}
afm_pairwise_mkrs_seu_clustr_anot[[1]] %>% filter(gene_id %in% c("PF3D7-0113600", "PF3D7-1301800", "PF3D7-1222600"))
```



```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
afm_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(p_val_adj < 1e-10 ) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.1) + 
  geom_hline(yintercept = seq(-12,12,1)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 28))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
## malariagen_CNVs gch1, crt, mdr1, plasmepsin 2/3, hrp2, hrp3

malariagen_CNVs <- c("PF3D7-1224000", "PF3D7-0709000", "PF3D7-0523000", "PF3D7-1408000", "PF3D7-0831800", "PF3D7-1372200")

```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
afm_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(gene_id %in% malariagen_CNVs) 

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
afm_pairwise_mkrs_seu_clustr_anot_all  %>%
  filter(gene_id %in% malariagen_CNVs) %>%
  filter(p_val_adj < 0.0001 ) %>% 
  ggplot(aes(y = avg_log2FC, x = 1)) +
  geom_violin() +
  geom_sina(size = 0.1) + 
  geom_hline(yintercept = seq(-7,7,1)) +
  facet_wrap(Gene ~ ., nrow = 1) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        text = element_text(size = 28))
```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
likely_cnvs <- map(afm_pairwise_mkrs_seu_clustr_anot, ~.x  %>%
  filter(gene_id %in% malariagen_CNVs) %>%
  filter(p_val_adj < 0.0001 ) %>% 
  filter(abs(avg_log2FC)>1) 
)

```


```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
likely_cnvs_don <- map(likely_cnvs, possibly(~.x  %>% 
    distinct(donor, gene_id, pw_comparison, .keep_all = T) %>%
  mutate(labl = paste0(donor, "_", Gene, "_", pw_comparison)) %>%
  select(donor, gene_id, Gene, pw_comparison, labl, avg_log2FC)
))
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
imap(likely_cnvs_don, possibly(~FeaturePlot(subset(m2_afm_bal_seu[[1]], subset = donor == .y), features = .x, split.by = "strain", reduction = "umap.harmony_PC_man", order = T)))
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
likely_cnvs_don
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
pmap(list(likely_cnvs_don$donor, likely_cnvs_don$gene_id, likely_cnvs_don$labl), possibly(~DotPlot(subset(afm_leiden_w_donors_pt, subset = donor == ..1), features = ..2, group.by = "strain") + labs(title = ..3)))
```

```{r, warning=FALSE, message=FALSE, fig.width=20, fig.height=3.5}
pmap(list(likely_cnvs_don$donor, likely_cnvs_don$gene_id, likely_cnvs_don$labl), possibly(~FeaturePlot(subset(afm_leiden_w_donors_pt, subset = donor == ..1), features = ..2, split.by = "strain", reduction = "umap.harmony_PC_man", order = F) + labs(title = ..3)))
```


```{r, warning=FALSE, message=FALSE}
top_prevalent_gns <- map(afm_pairwise_mkrs_seu_clustr_anot , possibly(~.x  %>%
  filter(p_val_adj < 1e-20 ) %>% 
  filter(abs(avg_log2FC)>1) %>%
  arrange(desc(gene_DE_donor_prevalence), gene_id) %>%
  # filter(gene_DE_sample_prevalence >= 5) %>%
  filter(gene_DE_donor_prevalence >= 5) %>%
  distinct(Gene, .keep_all = T) %>% 
  filter(donor != "MSC55") %>%
  arrange(desc(abs(avg_log2FC)))
))
```

```{r, warning=FALSE, message=FALSE}
gogo()
```

```{r, warning=FALSE, message=FALSE}
# as.list(split(top_prevalent_gns[,"donor"], top_prevalent_gns[,"gene_id"]))
top_gns_list <- tapply(top_prevalent_gns$gene_id, top_prevalent_gns$donor, c)
```

```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
imap(top_gns_list, possibly(~DotPlot(subset(afm_leiden_w_donors_pt, subset = donor == .y), features = .x, group.by = "strain") + labs(title = .y)))
```


```{r, warning=FALSE, message=FALSE, fig.width=10, fig.height=3.5}
imap(top_gns_list, possibly(~DotPlot(subset(afm_leiden_w_donors_pt, subset = donor == .y), features = .x, group.by = "strain") + labs(title = .y)))
```

```{r, warning=FALSE, message=FALSE, fig.width=5, fig.height=5}
afm_m2_bal_mast
```
